/* Generated by Cython 0.21 */

#define PY_SSIZE_T_CLEAN
#ifndef CYTHON_USE_PYLONG_INTERNALS
#ifdef PYLONG_BITS_IN_DIGIT
#define CYTHON_USE_PYLONG_INTERNALS 0
#else
#include "pyconfig.h"
#ifdef PYLONG_BITS_IN_DIGIT
#define CYTHON_USE_PYLONG_INTERNALS 1
#else
#define CYTHON_USE_PYLONG_INTERNALS 0
#endif
#endif
#endif
#include "Python.h"
#ifndef Py_PYTHON_H
    #error Python headers needed to compile C extensions, please install development version of Python.
#elif PY_VERSION_HEX < 0x02060000 || (0x03000000 <= PY_VERSION_HEX && PY_VERSION_HEX < 0x03020000)
    #error Cython requires Python 2.6+ or Python 3.2+.
#else
#define CYTHON_ABI "0_21"
#include <stddef.h>
#ifndef offsetof
#define offsetof(type, member) ( (size_t) & ((type*)0) -> member )
#endif
#if !defined(WIN32) && !defined(MS_WINDOWS)
  #ifndef __stdcall
    #define __stdcall
  #endif
  #ifndef __cdecl
    #define __cdecl
  #endif
  #ifndef __fastcall
    #define __fastcall
  #endif
#endif
#ifndef DL_IMPORT
  #define DL_IMPORT(t) t
#endif
#ifndef DL_EXPORT
  #define DL_EXPORT(t) t
#endif
#ifndef PY_LONG_LONG
  #define PY_LONG_LONG LONG_LONG
#endif
#ifndef Py_HUGE_VAL
  #define Py_HUGE_VAL HUGE_VAL
#endif
#ifdef PYPY_VERSION
#define CYTHON_COMPILING_IN_PYPY 1
#define CYTHON_COMPILING_IN_CPYTHON 0
#else
#define CYTHON_COMPILING_IN_PYPY 0
#define CYTHON_COMPILING_IN_CPYTHON 1
#endif
#if CYTHON_COMPILING_IN_PYPY && PY_VERSION_HEX < 0x02070600
#define Py_OptimizeFlag 0
#endif
#define __PYX_BUILD_PY_SSIZE_T "n"
#define CYTHON_FORMAT_SSIZE_T "z"
#if PY_MAJOR_VERSION < 3
  #define __Pyx_BUILTIN_MODULE_NAME "__builtin__"
  #define __Pyx_PyCode_New(a, k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos) \
          PyCode_New(a+k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos)
  #define __Pyx_DefaultClassType PyClass_Type
#else
  #define __Pyx_BUILTIN_MODULE_NAME "builtins"
  #define __Pyx_PyCode_New(a, k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos) \
          PyCode_New(a, k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos)
  #define __Pyx_DefaultClassType PyType_Type
#endif
#if PY_MAJOR_VERSION >= 3
  #define Py_TPFLAGS_CHECKTYPES 0
  #define Py_TPFLAGS_HAVE_INDEX 0
#endif
#if PY_MAJOR_VERSION >= 3
  #define Py_TPFLAGS_HAVE_NEWBUFFER 0
#endif
#if PY_VERSION_HEX < 0x030400a1 && !defined(Py_TPFLAGS_HAVE_FINALIZE)
  #define Py_TPFLAGS_HAVE_FINALIZE 0
#endif
#if PY_VERSION_HEX > 0x03030000 && defined(PyUnicode_KIND)
  #define CYTHON_PEP393_ENABLED 1
  #define __Pyx_PyUnicode_READY(op)       (likely(PyUnicode_IS_READY(op)) ? \
                                              0 : _PyUnicode_Ready((PyObject *)(op)))
  #define __Pyx_PyUnicode_GET_LENGTH(u)   PyUnicode_GET_LENGTH(u)
  #define __Pyx_PyUnicode_READ_CHAR(u, i) PyUnicode_READ_CHAR(u, i)
  #define __Pyx_PyUnicode_KIND(u)         PyUnicode_KIND(u)
  #define __Pyx_PyUnicode_DATA(u)         PyUnicode_DATA(u)
  #define __Pyx_PyUnicode_READ(k, d, i)   PyUnicode_READ(k, d, i)
#else
  #define CYTHON_PEP393_ENABLED 0
  #define __Pyx_PyUnicode_READY(op)       (0)
  #define __Pyx_PyUnicode_GET_LENGTH(u)   PyUnicode_GET_SIZE(u)
  #define __Pyx_PyUnicode_READ_CHAR(u, i) ((Py_UCS4)(PyUnicode_AS_UNICODE(u)[i]))
  #define __Pyx_PyUnicode_KIND(u)         (sizeof(Py_UNICODE))
  #define __Pyx_PyUnicode_DATA(u)         ((void*)PyUnicode_AS_UNICODE(u))
  #define __Pyx_PyUnicode_READ(k, d, i)   ((void)(k), (Py_UCS4)(((Py_UNICODE*)d)[i]))
#endif
#if CYTHON_COMPILING_IN_PYPY
  #define __Pyx_PyUnicode_Concat(a, b)      PyNumber_Add(a, b)
  #define __Pyx_PyUnicode_ConcatSafe(a, b)  PyNumber_Add(a, b)
#else
  #define __Pyx_PyUnicode_Concat(a, b)      PyUnicode_Concat(a, b)
  #define __Pyx_PyUnicode_ConcatSafe(a, b)  ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
      PyNumber_Add(a, b) : __Pyx_PyUnicode_Concat(a, b))
#endif
#define __Pyx_PyString_FormatSafe(a, b)   ((unlikely((a) == Py_None)) ? PyNumber_Remainder(a, b) : __Pyx_PyString_Format(a, b))
#define __Pyx_PyUnicode_FormatSafe(a, b)  ((unlikely((a) == Py_None)) ? PyNumber_Remainder(a, b) : PyUnicode_Format(a, b))
#if PY_MAJOR_VERSION >= 3
  #define __Pyx_PyString_Format(a, b)  PyUnicode_Format(a, b)
#else
  #define __Pyx_PyString_Format(a, b)  PyString_Format(a, b)
#endif
#if PY_MAJOR_VERSION >= 3
  #define PyBaseString_Type            PyUnicode_Type
  #define PyStringObject               PyUnicodeObject
  #define PyString_Type                PyUnicode_Type
  #define PyString_Check               PyUnicode_Check
  #define PyString_CheckExact          PyUnicode_CheckExact
#endif
#if PY_MAJOR_VERSION >= 3
  #define __Pyx_PyBaseString_Check(obj) PyUnicode_Check(obj)
  #define __Pyx_PyBaseString_CheckExact(obj) PyUnicode_CheckExact(obj)
#else
  #define __Pyx_PyBaseString_Check(obj) (PyString_Check(obj) || PyUnicode_Check(obj))
  #define __Pyx_PyBaseString_CheckExact(obj) (PyString_CheckExact(obj) || PyUnicode_CheckExact(obj))
#endif
#ifndef PySet_CheckExact
  #define PySet_CheckExact(obj)        (Py_TYPE(obj) == &PySet_Type)
#endif
#define __Pyx_TypeCheck(obj, type) PyObject_TypeCheck(obj, (PyTypeObject *)type)
#if PY_MAJOR_VERSION >= 3
  #define PyIntObject                  PyLongObject
  #define PyInt_Type                   PyLong_Type
  #define PyInt_Check(op)              PyLong_Check(op)
  #define PyInt_CheckExact(op)         PyLong_CheckExact(op)
  #define PyInt_FromString             PyLong_FromString
  #define PyInt_FromUnicode            PyLong_FromUnicode
  #define PyInt_FromLong               PyLong_FromLong
  #define PyInt_FromSize_t             PyLong_FromSize_t
  #define PyInt_FromSsize_t            PyLong_FromSsize_t
  #define PyInt_AsLong                 PyLong_AsLong
  #define PyInt_AS_LONG                PyLong_AS_LONG
  #define PyInt_AsSsize_t              PyLong_AsSsize_t
  #define PyInt_AsUnsignedLongMask     PyLong_AsUnsignedLongMask
  #define PyInt_AsUnsignedLongLongMask PyLong_AsUnsignedLongLongMask
  #define PyNumber_Int                 PyNumber_Long
#endif
#if PY_MAJOR_VERSION >= 3
  #define PyBoolObject                 PyLongObject
#endif
#if PY_VERSION_HEX < 0x030200A4
  typedef long Py_hash_t;
  #define __Pyx_PyInt_FromHash_t PyInt_FromLong
  #define __Pyx_PyInt_AsHash_t   PyInt_AsLong
#else
  #define __Pyx_PyInt_FromHash_t PyInt_FromSsize_t
  #define __Pyx_PyInt_AsHash_t   PyInt_AsSsize_t
#endif
#if PY_MAJOR_VERSION >= 3
  #define PyMethod_New(func, self, klass) ((self) ? PyMethod_New(func, self) : PyInstanceMethod_New(func))
#endif
#ifndef CYTHON_INLINE
  #if defined(__GNUC__)
    #define CYTHON_INLINE __inline__
  #elif defined(_MSC_VER)
    #define CYTHON_INLINE __inline
  #elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
    #define CYTHON_INLINE inline
  #else
    #define CYTHON_INLINE
  #endif
#endif
#ifndef CYTHON_RESTRICT
  #if defined(__GNUC__)
    #define CYTHON_RESTRICT __restrict__
  #elif defined(_MSC_VER) && _MSC_VER >= 1400
    #define CYTHON_RESTRICT __restrict
  #elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
    #define CYTHON_RESTRICT restrict
  #else
    #define CYTHON_RESTRICT
  #endif
#endif
#ifdef NAN
#define __PYX_NAN() ((float) NAN)
#else
static CYTHON_INLINE float __PYX_NAN() {
  /* Initialize NaN. The sign is irrelevant, an exponent with all bits 1 and
   a nonzero mantissa means NaN. If the first bit in the mantissa is 1, it is
   a quiet NaN. */
  float value;
  memset(&value, 0xFF, sizeof(value));
  return value;
}
#endif
#ifdef __cplusplus
template<typename T>
void __Pyx_call_destructor(T* x) {
    x->~T();
}
#endif


#if PY_MAJOR_VERSION >= 3
  #define __Pyx_PyNumber_Divide(x,y)         PyNumber_TrueDivide(x,y)
  #define __Pyx_PyNumber_InPlaceDivide(x,y)  PyNumber_InPlaceTrueDivide(x,y)
#else
  #define __Pyx_PyNumber_Divide(x,y)         PyNumber_Divide(x,y)
  #define __Pyx_PyNumber_InPlaceDivide(x,y)  PyNumber_InPlaceDivide(x,y)
#endif

#ifndef __PYX_EXTERN_C
  #ifdef __cplusplus
    #define __PYX_EXTERN_C extern "C"
  #else
    #define __PYX_EXTERN_C extern
  #endif
#endif

#if defined(WIN32) || defined(MS_WINDOWS)
#define _USE_MATH_DEFINES
#endif
#include <math.h>
#define __PYX_HAVE__rmgpy__molecule__group
#define __PYX_HAVE_API__rmgpy__molecule__group
#ifdef _OPENMP
#include <omp.h>
#endif /* _OPENMP */

#ifdef PYREX_WITHOUT_ASSERTIONS
#define CYTHON_WITHOUT_ASSERTIONS
#endif

#ifndef CYTHON_UNUSED
# if defined(__GNUC__)
#   if !(defined(__cplusplus)) || (__GNUC__ > 3 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 4))
#     define CYTHON_UNUSED __attribute__ ((__unused__))
#   else
#     define CYTHON_UNUSED
#   endif
# elif defined(__ICC) || (defined(__INTEL_COMPILER) && !defined(_MSC_VER))
#   define CYTHON_UNUSED __attribute__ ((__unused__))
# else
#   define CYTHON_UNUSED
# endif
#endif
typedef struct {PyObject **p; char *s; const Py_ssize_t n; const char* encoding;
                const char is_unicode; const char is_str; const char intern; } __Pyx_StringTabEntry;

#define __PYX_DEFAULT_STRING_ENCODING_IS_ASCII 0
#define __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT 0
#define __PYX_DEFAULT_STRING_ENCODING ""
#define __Pyx_PyObject_FromString __Pyx_PyBytes_FromString
#define __Pyx_PyObject_FromStringAndSize __Pyx_PyBytes_FromStringAndSize
#define __Pyx_fits_Py_ssize_t(v, type, is_signed)  (    \
    (sizeof(type) < sizeof(Py_ssize_t))  ||             \
    (sizeof(type) > sizeof(Py_ssize_t) &&               \
          likely(v < (type)PY_SSIZE_T_MAX ||            \
                 v == (type)PY_SSIZE_T_MAX)  &&         \
          (!is_signed || likely(v > (type)PY_SSIZE_T_MIN ||       \
                                v == (type)PY_SSIZE_T_MIN)))  ||  \
    (sizeof(type) == sizeof(Py_ssize_t) &&              \
          (is_signed || likely(v < (type)PY_SSIZE_T_MAX ||        \
                               v == (type)PY_SSIZE_T_MAX)))  )
static CYTHON_INLINE char* __Pyx_PyObject_AsString(PyObject*);
static CYTHON_INLINE char* __Pyx_PyObject_AsStringAndSize(PyObject*, Py_ssize_t* length);
#define __Pyx_PyByteArray_FromString(s) PyByteArray_FromStringAndSize((const char*)s, strlen((const char*)s))
#define __Pyx_PyByteArray_FromStringAndSize(s, l) PyByteArray_FromStringAndSize((const char*)s, l)
#define __Pyx_PyBytes_FromString        PyBytes_FromString
#define __Pyx_PyBytes_FromStringAndSize PyBytes_FromStringAndSize
static CYTHON_INLINE PyObject* __Pyx_PyUnicode_FromString(const char*);
#if PY_MAJOR_VERSION < 3
    #define __Pyx_PyStr_FromString        __Pyx_PyBytes_FromString
    #define __Pyx_PyStr_FromStringAndSize __Pyx_PyBytes_FromStringAndSize
#else
    #define __Pyx_PyStr_FromString        __Pyx_PyUnicode_FromString
    #define __Pyx_PyStr_FromStringAndSize __Pyx_PyUnicode_FromStringAndSize
#endif
#define __Pyx_PyObject_AsSString(s)    ((signed char*) __Pyx_PyObject_AsString(s))
#define __Pyx_PyObject_AsUString(s)    ((unsigned char*) __Pyx_PyObject_AsString(s))
#define __Pyx_PyObject_FromUString(s)  __Pyx_PyObject_FromString((const char*)s)
#define __Pyx_PyBytes_FromUString(s)   __Pyx_PyBytes_FromString((const char*)s)
#define __Pyx_PyByteArray_FromUString(s)   __Pyx_PyByteArray_FromString((const char*)s)
#define __Pyx_PyStr_FromUString(s)     __Pyx_PyStr_FromString((const char*)s)
#define __Pyx_PyUnicode_FromUString(s) __Pyx_PyUnicode_FromString((const char*)s)
#if PY_MAJOR_VERSION < 3
static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)
{
    const Py_UNICODE *u_end = u;
    while (*u_end++) ;
    return (size_t)(u_end - u - 1);
}
#else
#define __Pyx_Py_UNICODE_strlen Py_UNICODE_strlen
#endif
#define __Pyx_PyUnicode_FromUnicode(u)       PyUnicode_FromUnicode(u, __Pyx_Py_UNICODE_strlen(u))
#define __Pyx_PyUnicode_FromUnicodeAndLength PyUnicode_FromUnicode
#define __Pyx_PyUnicode_AsUnicode            PyUnicode_AsUnicode
#define __Pyx_Owned_Py_None(b) (Py_INCREF(Py_None), Py_None)
#define __Pyx_PyBool_FromLong(b) ((b) ? (Py_INCREF(Py_True), Py_True) : (Py_INCREF(Py_False), Py_False))
static CYTHON_INLINE int __Pyx_PyObject_IsTrue(PyObject*);
static CYTHON_INLINE PyObject* __Pyx_PyNumber_Int(PyObject* x);
static CYTHON_INLINE Py_ssize_t __Pyx_PyIndex_AsSsize_t(PyObject*);
static CYTHON_INLINE PyObject * __Pyx_PyInt_FromSize_t(size_t);
#if CYTHON_COMPILING_IN_CPYTHON
#define __pyx_PyFloat_AsDouble(x) (PyFloat_CheckExact(x) ? PyFloat_AS_DOUBLE(x) : PyFloat_AsDouble(x))
#else
#define __pyx_PyFloat_AsDouble(x) PyFloat_AsDouble(x)
#endif
#define __pyx_PyFloat_AsFloat(x) ((float) __pyx_PyFloat_AsDouble(x))
#if PY_MAJOR_VERSION < 3 && __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
static int __Pyx_sys_getdefaultencoding_not_ascii;
static int __Pyx_init_sys_getdefaultencoding_params(void) {
    PyObject* sys;
    PyObject* default_encoding = NULL;
    PyObject* ascii_chars_u = NULL;
    PyObject* ascii_chars_b = NULL;
    const char* default_encoding_c;
    sys = PyImport_ImportModule("sys");
    if (!sys) goto bad;
    default_encoding = PyObject_CallMethod(sys, (char*) (const char*) "getdefaultencoding", NULL);
    Py_DECREF(sys);
    if (!default_encoding) goto bad;
    default_encoding_c = PyBytes_AsString(default_encoding);
    if (!default_encoding_c) goto bad;
    if (strcmp(default_encoding_c, "ascii") == 0) {
        __Pyx_sys_getdefaultencoding_not_ascii = 0;
    } else {
        char ascii_chars[128];
        int c;
        for (c = 0; c < 128; c++) {
            ascii_chars[c] = c;
        }
        __Pyx_sys_getdefaultencoding_not_ascii = 1;
        ascii_chars_u = PyUnicode_DecodeASCII(ascii_chars, 128, NULL);
        if (!ascii_chars_u) goto bad;
        ascii_chars_b = PyUnicode_AsEncodedString(ascii_chars_u, default_encoding_c, NULL);
        if (!ascii_chars_b || !PyBytes_Check(ascii_chars_b) || memcmp(ascii_chars, PyBytes_AS_STRING(ascii_chars_b), 128) != 0) {
            PyErr_Format(
                PyExc_ValueError,
                "This module compiled with c_string_encoding=ascii, but default encoding '%.200s' is not a superset of ascii.",
                default_encoding_c);
            goto bad;
        }
        Py_DECREF(ascii_chars_u);
        Py_DECREF(ascii_chars_b);
    }
    Py_DECREF(default_encoding);
    return 0;
bad:
    Py_XDECREF(default_encoding);
    Py_XDECREF(ascii_chars_u);
    Py_XDECREF(ascii_chars_b);
    return -1;
}
#endif
#if __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT && PY_MAJOR_VERSION >= 3
#define __Pyx_PyUnicode_FromStringAndSize(c_str, size) PyUnicode_DecodeUTF8(c_str, size, NULL)
#else
#define __Pyx_PyUnicode_FromStringAndSize(c_str, size) PyUnicode_Decode(c_str, size, __PYX_DEFAULT_STRING_ENCODING, NULL)
#if __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT
static char* __PYX_DEFAULT_STRING_ENCODING;
static int __Pyx_init_sys_getdefaultencoding_params(void) {
    PyObject* sys;
    PyObject* default_encoding = NULL;
    char* default_encoding_c;
    sys = PyImport_ImportModule("sys");
    if (!sys) goto bad;
    default_encoding = PyObject_CallMethod(sys, (char*) (const char*) "getdefaultencoding", NULL);
    Py_DECREF(sys);
    if (!default_encoding) goto bad;
    default_encoding_c = PyBytes_AsString(default_encoding);
    if (!default_encoding_c) goto bad;
    __PYX_DEFAULT_STRING_ENCODING = (char*) malloc(strlen(default_encoding_c));
    if (!__PYX_DEFAULT_STRING_ENCODING) goto bad;
    strcpy(__PYX_DEFAULT_STRING_ENCODING, default_encoding_c);
    Py_DECREF(default_encoding);
    return 0;
bad:
    Py_XDECREF(default_encoding);
    return -1;
}
#endif
#endif


/* Test for GCC > 2.95 */
#if defined(__GNUC__)     && (__GNUC__ > 2 || (__GNUC__ == 2 && (__GNUC_MINOR__ > 95)))
  #define likely(x)   __builtin_expect(!!(x), 1)
  #define unlikely(x) __builtin_expect(!!(x), 0)
#else /* !__GNUC__ or GCC < 2.95 */
  #define likely(x)   (x)
  #define unlikely(x) (x)
#endif /* __GNUC__ */

static PyObject *__pyx_m;
static PyObject *__pyx_d;
static PyObject *__pyx_b;
static PyObject *__pyx_empty_tuple;
static PyObject *__pyx_empty_bytes;
static int __pyx_lineno;
static int __pyx_clineno = 0;
static const char * __pyx_cfilenm= __FILE__;
static const char *__pyx_filename;


static const char *__pyx_f[] = {
  "rmgpy/molecule/group.py",
  "rmgpy/molecule/group.pxd",
  "rmgpy/molecule/atomtype.pxd",
};

/*--- Type declarations ---*/
struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex;
struct __pyx_obj_5rmgpy_8molecule_5graph_Edge;
struct __pyx_obj_5rmgpy_8molecule_5graph_Graph;
struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType;
struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom;
struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond;
struct __pyx_obj_5rmgpy_8molecule_5group_Group;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms;

/* "rmgpy/molecule/graph.pxd":90
 *     cpdef updateConnectivityValues(self)
 * 
 *     cpdef Graph copy(self, bint deep=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef Graph merge(self, Graph other)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy {
  int __pyx_n;
  int deep;
};

/* "rmgpy/molecule/graph.pxd":100
 *     cpdef sortVertices(self)
 * 
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/graph.pxd":102
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/graph.pxd":104
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/graph.pxd":106
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isCyclic(self) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms {
  int __pyx_n;
  PyObject *initialMap;
};
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_copy;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findIsomorphism;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms;

/* "rmgpy/molecule/group.pxd":121
 *     cpdef sortAtoms(self)
 * 
 *     cpdef Graph copy(self, bint deep=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef clearLabeledAtoms(self)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_copy {
  int __pyx_n;
  int deep;
};

/* "rmgpy/molecule/group.pxd":133
 *     cpdef fromAdjacencyList(self, str adjlist)
 * 
 *     cpdef toAdjacencyList(self, str label=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef updateFingerprint(self)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList {
  int __pyx_n;
  PyObject *label;
};

/* "rmgpy/molecule/group.pxd":137
 *     cpdef updateFingerprint(self)
 * 
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/group.pxd":139
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findIsomorphism {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/group.pxd":141
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/group.pxd":143
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isIdentical(self, Graph other)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/graph.pxd":27
 * ################################################################################
 * 
 * cdef class Vertex(object):             # <<<<<<<<<<<<<<
 * 
 *     cdef public dict edges
 */
struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex {
  PyObject_HEAD
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex *__pyx_vtab;
  PyObject *edges;
  short connectivity1;
  short connectivity2;
  short connectivity3;
  short sortingLabel;
  int terminal;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *mapping;
  int ignore;
};


/* "rmgpy/molecule/graph.pxd":54
 * ################################################################################
 * 
 * cdef class Edge(object):             # <<<<<<<<<<<<<<
 * 
 *     cdef public Vertex vertex1, vertex2
 */
struct __pyx_obj_5rmgpy_8molecule_5graph_Edge {
  PyObject_HEAD
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge *__pyx_vtab;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *vertex1;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *vertex2;
};


/* "rmgpy/molecule/graph.pxd":68
 * ################################################################################
 * 
 * cdef class Graph:             # <<<<<<<<<<<<<<
 * 
 *     cdef public list vertices
 */
struct __pyx_obj_5rmgpy_8molecule_5graph_Graph {
  PyObject_HEAD
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph *__pyx_vtab;
  PyObject *vertices;
};


/* "rmgpy/molecule/atomtype.pxd":27
 * ################################################################################
 * 
 * cdef class AtomType:             # <<<<<<<<<<<<<<
 * 
 *     cdef public str label
 */
struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType {
  PyObject_HEAD
  struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *__pyx_vtab;
  PyObject *label;
  PyObject *generic;
  PyObject *specific;
  PyObject *incrementBond;
  PyObject *decrementBond;
  PyObject *formBond;
  PyObject *breakBond;
  PyObject *incrementRadical;
  PyObject *decrementRadical;
  PyObject *incrementLonePair;
  PyObject *decrementLonePair;
  PyObject *single;
  PyObject *allDouble;
  PyObject *rDouble;
  PyObject *oDouble;
  PyObject *sDouble;
  PyObject *triple;
  PyObject *benzene;
  PyObject *lonePairs;
};


/* "rmgpy/molecule/group.pxd":32
 * ################################################################################
 * 
 * cdef class GroupAtom(Vertex):             # <<<<<<<<<<<<<<
 * 
 *     cdef public list atomType
 */
struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex __pyx_base;
  PyObject *atomType;
  PyObject *radicalElectrons;
  PyObject *charge;
  PyObject *label;
  PyObject *lonePairs;
};


/* "rmgpy/molecule/group.pxd":67
 * ################################################################################
 * 
 * cdef class GroupBond(Edge):             # <<<<<<<<<<<<<<
 * 
 *     cdef public list order
 */
struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge __pyx_base;
  PyObject *order;
};


/* "rmgpy/molecule/group.pxd":91
 * ################################################################################
 * 
 * cdef class Group(Graph):             # <<<<<<<<<<<<<<
 * 
 *     cdef public list multiplicity
 */
struct __pyx_obj_5rmgpy_8molecule_5group_Group {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph __pyx_base;
  PyObject *multiplicity;
  short carbonCount;
  short nitrogenCount;
  short oxygenCount;
  short sulfurCount;
  short radicalCount;
};



/* "rmgpy/molecule/graph.pxd":27
 * ################################################################################
 * 
 * cdef class Vertex(object):             # <<<<<<<<<<<<<<
 * 
 *     cdef public dict edges
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *(*copy)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*equivalent)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*isSpecificCaseOf)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  PyObject *(*resetConnectivityValues)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex *__pyx_vtabptr_5rmgpy_8molecule_5graph_Vertex;


/* "rmgpy/molecule/graph.pxd":54
 * ################################################################################
 * 
 * cdef class Edge(object):             # <<<<<<<<<<<<<<
 * 
 *     cdef public Vertex vertex1, vertex2
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *(*copy)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  int (*equivalent)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  int (*isSpecificCaseOf)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *(*getOtherVertex)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge *__pyx_vtabptr_5rmgpy_8molecule_5graph_Edge;


/* "rmgpy/molecule/graph.pxd":68
 * ################################################################################
 * 
 * cdef class Graph:             # <<<<<<<<<<<<<<
 * 
 *     cdef public list vertices
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *(*addVertex)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *(*addEdge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  PyObject *(*getEdges)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *(*getEdge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*hasVertex)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*hasEdge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  PyObject *(*removeVertex)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  PyObject *(*removeEdge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  PyObject *(*updateConnectivityValues)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *(*copy)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy *__pyx_optional_args);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *(*merge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*split)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*resetConnectivityValues)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*sortVertices)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  int (*isIsomorphic)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic *__pyx_optional_args);
  PyObject *(*findIsomorphism)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism *__pyx_optional_args);
  int (*isSubgraphIsomorphic)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic *__pyx_optional_args);
  PyObject *(*findSubgraphIsomorphisms)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms *__pyx_optional_args);
  int (*isCyclic)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  int (*isVertexInCycle)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*isEdgeInCycle)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  int (*__pyx___isChainInCycle)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*getAllCyclicVertices)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getAllPolycyclicVertices)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getPolycyclicRings)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getMonocyclicRings)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getDisparateRings)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getAllCycles)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  PyObject *(*__pyx___exploreCyclesRecursively)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, PyObject *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*getSmallestSetOfSmallestRings)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  int (*isMappingValid)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, PyObject *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph *__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph;


/* "rmgpy/molecule/atomtype.pxd":27
 * ################################################################################
 * 
 * cdef class AtomType:             # <<<<<<<<<<<<<<
 * 
 *     cdef public str label
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType {
  int (*isSpecificCaseOf)(struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, int __pyx_skip_dispatch);
  int (*equivalent)(struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, int __pyx_skip_dispatch);
  PyObject *(*getFeatures)(struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *__pyx_vtabptr_5rmgpy_8molecule_8atomtype_AtomType;


/* "rmgpy/molecule/group.py":53
 * ################################################################################
 * 
 * class GroupAtom(Vertex):             # <<<<<<<<<<<<<<
 *     """
 *     An atom group. This class is based on the :class:`Atom` class, except that
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex __pyx_base;
  PyObject *(*__pyx___changeBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*__pyx___formBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*__pyx___breakBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*__pyx___gainRadical)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*__pyx___loseRadical)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*__pyx___gainPair)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*__pyx___losePair)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*applyAction)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch);
  int (*isOxygen)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  int (*isSulfur)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *__pyx_vtabptr_5rmgpy_8molecule_5group_GroupAtom;


/* "rmgpy/molecule/group.py":411
 * ################################################################################
 * 
 * class GroupBond(Edge):             # <<<<<<<<<<<<<<
 *     """
 *     A bond group. This class is based on the :class:`Bond` class, except that
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupBond {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge __pyx_base;
  PyObject *(*__pyx___changeBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, short, int __pyx_skip_dispatch);
  int (*isSingle)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  int (*isDouble)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  int (*isTriple)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  int (*isBenzene)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  PyObject *(*applyAction)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, PyObject *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupBond *__pyx_vtabptr_5rmgpy_8molecule_5group_GroupBond;


/* "rmgpy/molecule/group.py":576
 * ################################################################################
 * 
 * class Group(Graph):             # <<<<<<<<<<<<<<
 *     """
 *     A representation of a molecular substructure group using a graph data
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph __pyx_base;
  PyObject *(*addAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  PyObject *(*addBond)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  PyObject *(*getBonds)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *(*getBond)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  int (*hasAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  int (*hasBond)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  PyObject *(*removeAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  PyObject *(*removeBond)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  PyObject *(*sortAtoms)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  PyObject *(*clearLabeledAtoms)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*containsLabeledAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *(*getLabeledAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*getLabeledAtoms)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  PyObject *(*fromAdjacencyList)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*toAdjacencyList)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList *__pyx_optional_args);
  PyObject *(*updateFingerprint)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*isIdentical)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  int (*standardizeAtomType)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*addExplicitLigands)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*standardizeGroup)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *__pyx_vtabptr_5rmgpy_8molecule_5group_Group;
#ifndef CYTHON_REFNANNY
  #define CYTHON_REFNANNY 0
#endif
#if CYTHON_REFNANNY
  typedef struct {
    void (*INCREF)(void*, PyObject*, int);
    void (*DECREF)(void*, PyObject*, int);
    void (*GOTREF)(void*, PyObject*, int);
    void (*GIVEREF)(void*, PyObject*, int);
    void* (*SetupContext)(const char*, int, const char*);
    void (*FinishContext)(void**);
  } __Pyx_RefNannyAPIStruct;
  static __Pyx_RefNannyAPIStruct *__Pyx_RefNanny = NULL;
  static __Pyx_RefNannyAPIStruct *__Pyx_RefNannyImportAPI(const char *modname);
  #define __Pyx_RefNannyDeclarations void *__pyx_refnanny = NULL;
#ifdef WITH_THREAD
  #define __Pyx_RefNannySetupContext(name, acquire_gil) \
          if (acquire_gil) { \
              PyGILState_STATE __pyx_gilstate_save = PyGILState_Ensure(); \
              __pyx_refnanny = __Pyx_RefNanny->SetupContext((name), __LINE__, __FILE__); \
              PyGILState_Release(__pyx_gilstate_save); \
          } else { \
              __pyx_refnanny = __Pyx_RefNanny->SetupContext((name), __LINE__, __FILE__); \
          }
#else
  #define __Pyx_RefNannySetupContext(name, acquire_gil) \
          __pyx_refnanny = __Pyx_RefNanny->SetupContext((name), __LINE__, __FILE__)
#endif
  #define __Pyx_RefNannyFinishContext() \
          __Pyx_RefNanny->FinishContext(&__pyx_refnanny)
  #define __Pyx_INCREF(r)  __Pyx_RefNanny->INCREF(__pyx_refnanny, (PyObject *)(r), __LINE__)
  #define __Pyx_DECREF(r)  __Pyx_RefNanny->DECREF(__pyx_refnanny, (PyObject *)(r), __LINE__)
  #define __Pyx_GOTREF(r)  __Pyx_RefNanny->GOTREF(__pyx_refnanny, (PyObject *)(r), __LINE__)
  #define __Pyx_GIVEREF(r) __Pyx_RefNanny->GIVEREF(__pyx_refnanny, (PyObject *)(r), __LINE__)
  #define __Pyx_XINCREF(r)  do { if((r) != NULL) {__Pyx_INCREF(r); }} while(0)
  #define __Pyx_XDECREF(r)  do { if((r) != NULL) {__Pyx_DECREF(r); }} while(0)
  #define __Pyx_XGOTREF(r)  do { if((r) != NULL) {__Pyx_GOTREF(r); }} while(0)
  #define __Pyx_XGIVEREF(r) do { if((r) != NULL) {__Pyx_GIVEREF(r);}} while(0)
#else
  #define __Pyx_RefNannyDeclarations
  #define __Pyx_RefNannySetupContext(name, acquire_gil)
  #define __Pyx_RefNannyFinishContext()
  #define __Pyx_INCREF(r) Py_INCREF(r)
  #define __Pyx_DECREF(r) Py_DECREF(r)
  #define __Pyx_GOTREF(r)
  #define __Pyx_GIVEREF(r)
  #define __Pyx_XINCREF(r) Py_XINCREF(r)
  #define __Pyx_XDECREF(r) Py_XDECREF(r)
  #define __Pyx_XGOTREF(r)
  #define __Pyx_XGIVEREF(r)
#endif
#define __Pyx_XDECREF_SET(r, v) do {                            \
        PyObject *tmp = (PyObject *) r;                         \
        r = v; __Pyx_XDECREF(tmp);                              \
    } while (0)
#define __Pyx_DECREF_SET(r, v) do {                             \
        PyObject *tmp = (PyObject *) r;                         \
        r = v; __Pyx_DECREF(tmp);                               \
    } while (0)
#define __Pyx_CLEAR(r)    do { PyObject* tmp = ((PyObject*)(r)); r = NULL; __Pyx_DECREF(tmp);} while(0)
#define __Pyx_XCLEAR(r)   do { if((r) != NULL) {PyObject* tmp = ((PyObject*)(r)); r = NULL; __Pyx_DECREF(tmp);}} while(0)

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_GetAttrStr(PyObject* obj, PyObject* attr_name) {
    PyTypeObject* tp = Py_TYPE(obj);
    if (likely(tp->tp_getattro))
        return tp->tp_getattro(obj, attr_name);
#if PY_MAJOR_VERSION < 3
    if (likely(tp->tp_getattr))
        return tp->tp_getattr(obj, PyString_AS_STRING(attr_name));
#endif
    return PyObject_GetAttr(obj, attr_name);
}
#else
#define __Pyx_PyObject_GetAttrStr(o,n) PyObject_GetAttr(o,n)
#endif

static PyObject *__Pyx_GetBuiltinName(PyObject *name);

static void __Pyx_RaiseDoubleKeywordsError(const char* func_name, PyObject* kw_name);

static int __Pyx_ParseOptionalKeywords(PyObject *kwds, PyObject **argnames[], \
    PyObject *kwds2, PyObject *values[], Py_ssize_t num_pos_args, \
    const char* function_name);

static void __Pyx_RaiseArgtupleInvalid(const char* func_name, int exact,
    Py_ssize_t num_min, Py_ssize_t num_max, Py_ssize_t num_found);

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_Call(PyObject *func, PyObject *arg, PyObject *kw);
#else
#define __Pyx_PyObject_Call(func, arg, kw) PyObject_Call(func, arg, kw)
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallMethO(PyObject *func, PyObject *arg);
#endif

static CYTHON_INLINE PyObject* __Pyx_PyObject_CallOneArg(PyObject *func, PyObject *arg);

#define __Pyx_GetItemInt(o, i, type, is_signed, to_py_func, is_list, wraparound, boundscheck) \
    (__Pyx_fits_Py_ssize_t(i, type, is_signed) ? \
    __Pyx_GetItemInt_Fast(o, (Py_ssize_t)i, is_list, wraparound, boundscheck) : \
    (is_list ? (PyErr_SetString(PyExc_IndexError, "list index out of range"), (PyObject*)NULL) : \
               __Pyx_GetItemInt_Generic(o, to_py_func(i))))
#define __Pyx_GetItemInt_List(o, i, type, is_signed, to_py_func, is_list, wraparound, boundscheck) \
    (__Pyx_fits_Py_ssize_t(i, type, is_signed) ? \
    __Pyx_GetItemInt_List_Fast(o, (Py_ssize_t)i, wraparound, boundscheck) : \
    (PyErr_SetString(PyExc_IndexError, "list index out of range"), (PyObject*)NULL))
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_List_Fast(PyObject *o, Py_ssize_t i,
                                                              int wraparound, int boundscheck);
#define __Pyx_GetItemInt_Tuple(o, i, type, is_signed, to_py_func, is_list, wraparound, boundscheck) \
    (__Pyx_fits_Py_ssize_t(i, type, is_signed) ? \
    __Pyx_GetItemInt_Tuple_Fast(o, (Py_ssize_t)i, wraparound, boundscheck) : \
    (PyErr_SetString(PyExc_IndexError, "tuple index out of range"), (PyObject*)NULL))
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Tuple_Fast(PyObject *o, Py_ssize_t i,
                                                              int wraparound, int boundscheck);
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Generic(PyObject *o, PyObject* j);
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Fast(PyObject *o, Py_ssize_t i,
                                                     int is_list, int wraparound, int boundscheck);

static CYTHON_INLINE PyObject *__Pyx_GetModuleGlobalName(PyObject *name);

#define __Pyx_SetItemInt(o, i, v, type, is_signed, to_py_func, is_list, wraparound, boundscheck) \
    (__Pyx_fits_Py_ssize_t(i, type, is_signed) ? \
    __Pyx_SetItemInt_Fast(o, (Py_ssize_t)i, v, is_list, wraparound, boundscheck) : \
    (is_list ? (PyErr_SetString(PyExc_IndexError, "list assignment index out of range"), -1) : \
               __Pyx_SetItemInt_Generic(o, to_py_func(i), v)))
static CYTHON_INLINE int __Pyx_SetItemInt_Generic(PyObject *o, PyObject *j, PyObject *v);
static CYTHON_INLINE int __Pyx_SetItemInt_Fast(PyObject *o, Py_ssize_t i, PyObject *v,
                                               int is_list, int wraparound, int boundscheck);

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE int __Pyx_ListComp_Append(PyObject* list, PyObject* x) {
    PyListObject* L = (PyListObject*) list;
    Py_ssize_t len = Py_SIZE(list);
    if (likely(L->allocated > len)) {
        Py_INCREF(x);
        PyList_SET_ITEM(list, len, x);
        Py_SIZE(list) = len+1;
        return 0;
    }
    return PyList_Append(list, x);
}
#else
#define __Pyx_ListComp_Append(L,x) PyList_Append(L,x)
#endif

#if PY_MAJOR_VERSION < 3
#define __Pyx_PyString_Join __Pyx_PyBytes_Join
#define __Pyx_PyBaseString_Join(s, v) (PyUnicode_CheckExact(s) ? PyUnicode_Join(s, v) : __Pyx_PyBytes_Join(s, v))
#else
#define __Pyx_PyString_Join PyUnicode_Join
#define __Pyx_PyBaseString_Join PyUnicode_Join
#endif
#if CYTHON_COMPILING_IN_CPYTHON
    #if PY_MAJOR_VERSION < 3
    #define __Pyx_PyBytes_Join _PyString_Join
    #else
    #define __Pyx_PyBytes_Join _PyBytes_Join
    #endif
#else
static CYTHON_INLINE PyObject* __Pyx_PyBytes_Join(PyObject* sep, PyObject* values);
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallNoArg(PyObject *func);
#else
#define __Pyx_PyObject_CallNoArg(func) __Pyx_PyObject_Call(func, __pyx_empty_tuple, NULL)
#endif

static CYTHON_INLINE int __Pyx_TypeTest(PyObject *obj, PyTypeObject *type);

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyList_GetSlice(PyObject* src, Py_ssize_t start, Py_ssize_t stop);
static CYTHON_INLINE PyObject* __Pyx_PyTuple_GetSlice(PyObject* src, Py_ssize_t start, Py_ssize_t stop);
#else
#define __Pyx_PyList_GetSlice(seq, start, stop)   PySequence_GetSlice(seq, start, stop)
#define __Pyx_PyTuple_GetSlice(seq, start, stop)  PySequence_GetSlice(seq, start, stop)
#endif

static CYTHON_INLINE int __Pyx_PyList_Extend(PyObject* L, PyObject* v) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyObject* none = _PyList_Extend((PyListObject*)L, v);
    if (unlikely(!none))
        return -1;
    Py_DECREF(none);
    return 0;
#else
    return PyList_SetSlice(L, PY_SSIZE_T_MAX, PY_SSIZE_T_MAX, v);
#endif
}

static CYTHON_INLINE void __Pyx_ErrRestore(PyObject *type, PyObject *value, PyObject *tb);
static CYTHON_INLINE void __Pyx_ErrFetch(PyObject **type, PyObject **value, PyObject **tb);

static void __Pyx_Raise(PyObject *type, PyObject *value, PyObject *tb, PyObject *cause);

#include <string.h>

static CYTHON_INLINE int __Pyx_PyBytes_Equals(PyObject* s1, PyObject* s2, int equals);

static CYTHON_INLINE int __Pyx_PyUnicode_Equals(PyObject* s1, PyObject* s2, int equals);

#if PY_MAJOR_VERSION >= 3
#define __Pyx_PyString_Equals __Pyx_PyUnicode_Equals
#else
#define __Pyx_PyString_Equals __Pyx_PyBytes_Equals
#endif

static CYTHON_INLINE int __Pyx_ArgTypeTest(PyObject *obj, PyTypeObject *type, int none_allowed,
    const char *name, int exact);

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE int __Pyx_PyList_Append(PyObject* list, PyObject* x) {
    PyListObject* L = (PyListObject*) list;
    Py_ssize_t len = Py_SIZE(list);
    if (likely(L->allocated > len) & likely(len > (L->allocated >> 1))) {
        Py_INCREF(x);
        PyList_SET_ITEM(list, len, x);
        Py_SIZE(list) = len+1;
        return 0;
    }
    return PyList_Append(list, x);
}
#else
#define __Pyx_PyList_Append(L,x) PyList_Append(L,x)
#endif

static PyObject* __Pyx_PyObject_CallMethod1(PyObject* obj, PyObject* method_name, PyObject* arg);

static CYTHON_INLINE int __Pyx_PyObject_Append(PyObject* L, PyObject* x);

static CYTHON_INLINE int __Pyx_PySequence_Contains(PyObject* item, PyObject* seq, int eq) {
    int result = PySequence_Contains(seq, item);
    return unlikely(result < 0) ? result : (result == (eq == Py_EQ));
}

static void __Pyx_WriteUnraisable(const char *name, int clineno,
                                  int lineno, const char *filename,
                                  int full_traceback);

#if CYTHON_COMPILING_IN_CPYTHON
#define __Pyx_PyObject_DelAttrStr(o,n) __Pyx_PyObject_SetAttrStr(o,n,NULL)
static CYTHON_INLINE int __Pyx_PyObject_SetAttrStr(PyObject* obj, PyObject* attr_name, PyObject* value) {
    PyTypeObject* tp = Py_TYPE(obj);
    if (likely(tp->tp_setattro))
        return tp->tp_setattro(obj, attr_name, value);
#if PY_MAJOR_VERSION < 3
    if (likely(tp->tp_setattr))
        return tp->tp_setattr(obj, PyString_AS_STRING(attr_name), value);
#endif
    return PyObject_SetAttr(obj, attr_name, value);
}
#else
#define __Pyx_PyObject_DelAttrStr(o,n)   PyObject_DelAttr(o,n)
#define __Pyx_PyObject_SetAttrStr(o,n,v) PyObject_SetAttr(o,n,v)
#endif

static CYTHON_INLINE int __Pyx_PyDict_Contains(PyObject* item, PyObject* dict, int eq) {
    int result = PyDict_Contains(dict, item);
    return unlikely(result < 0) ? result : (result == (eq == Py_EQ));
}

#if PY_MAJOR_VERSION >= 3
static PyObject *__Pyx_PyDict_GetItem(PyObject *d, PyObject* key) {
    PyObject *value;
    value = PyDict_GetItemWithError(d, key);
    if (unlikely(!value)) {
        if (!PyErr_Occurred()) {
            PyObject* args = PyTuple_Pack(1, key);
            if (likely(args))
                PyErr_SetObject(PyExc_KeyError, args);
            Py_XDECREF(args);
        }
        return NULL;
    }
    Py_INCREF(value);
    return value;
}
#else
    #define __Pyx_PyDict_GetItem(d, key) PyObject_GetItem(d, key)
#endif

static PyObject* __Pyx_ImportFrom(PyObject* module, PyObject* name);

static CYTHON_INLINE void __Pyx_RaiseTooManyValuesError(Py_ssize_t expected);

static CYTHON_INLINE void __Pyx_RaiseNeedMoreValuesError(Py_ssize_t index);

static CYTHON_INLINE int __Pyx_IterFinish(void);

static int __Pyx_IternextUnpackEndCheck(PyObject *retval, Py_ssize_t expected);

static PyObject* __Pyx_PyObject_CallMethod0(PyObject* obj, PyObject* method_name);

static CYTHON_INLINE void __Pyx_RaiseNoneNotIterableError(void);

static void __Pyx_UnpackTupleError(PyObject *, Py_ssize_t index);

static CYTHON_INLINE int __Pyx_unpack_tuple2(PyObject* tuple, PyObject** value1, PyObject** value2,
                                             int is_tuple, int has_known_size, int decref_tuple);

static CYTHON_INLINE PyObject* __Pyx_dict_iterator(PyObject* dict, int is_dict, PyObject* method_name,
                                                   Py_ssize_t* p_orig_length, int* p_is_dict);
static CYTHON_INLINE int __Pyx_dict_iter_next(PyObject* dict_or_iter, Py_ssize_t orig_length, Py_ssize_t* ppos,
                                              PyObject** pkey, PyObject** pvalue, PyObject** pitem, int is_dict);

static CYTHON_INLINE void __Pyx_RaiseUnboundLocalError(const char *varname);

static void __Pyx_call_next_tp_dealloc(PyObject* obj, destructor current_tp_dealloc);

static int __Pyx_call_next_tp_traverse(PyObject* obj, visitproc v, void *a, traverseproc current_tp_traverse);

static void __Pyx_call_next_tp_clear(PyObject* obj, inquiry current_tp_dealloc);

static void* __Pyx_GetVtable(PyObject *dict);

static int __Pyx_SetVtable(PyObject *dict, void *vtable);

static PyObject *__Pyx_CalculateMetaclass(PyTypeObject *metaclass, PyObject *bases);

static PyObject *__Pyx_Py3MetaclassPrepare(PyObject *metaclass, PyObject *bases, PyObject *name, PyObject *qualname,
                                           PyObject *mkw, PyObject *modname, PyObject *doc);
static PyObject *__Pyx_Py3ClassCreate(PyObject *metaclass, PyObject *name, PyObject *bases, PyObject *dict,
                                      PyObject *mkw, int calculate_metaclass, int allow_py2_metaclass);

static PyObject *__Pyx_GetNameInClass(PyObject *nmspace, PyObject *name);

typedef struct {
    int code_line;
    PyCodeObject* code_object;
} __Pyx_CodeObjectCacheEntry;
struct __Pyx_CodeObjectCache {
    int count;
    int max_count;
    __Pyx_CodeObjectCacheEntry* entries;
};
static struct __Pyx_CodeObjectCache __pyx_code_cache = {0,0,NULL};
static int __pyx_bisect_code_objects(__Pyx_CodeObjectCacheEntry* entries, int count, int code_line);
static PyCodeObject *__pyx_find_code_object(int code_line);
static void __pyx_insert_code_object(int code_line, PyCodeObject* code_object);

static void __Pyx_AddTraceback(const char *funcname, int c_line,
                               int py_line, const char *filename);

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_short(short value);

static CYTHON_INLINE short __Pyx_PyInt_As_short(PyObject *);

static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level);

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_long(long value);

static CYTHON_INLINE int __Pyx_PyInt_As_int(PyObject *);

static CYTHON_INLINE long __Pyx_PyInt_As_long(PyObject *);

static int __Pyx_check_binary_version(void);

#if !defined(__Pyx_PyIdentifier_FromString)
#if PY_MAJOR_VERSION < 3
  #define __Pyx_PyIdentifier_FromString(s) PyString_FromString(s)
#else
  #define __Pyx_PyIdentifier_FromString(s) PyUnicode_FromString(s)
#endif
#endif

static PyObject *__Pyx_ImportModule(const char *name);

static PyTypeObject *__Pyx_ImportType(const char *module_name, const char *class_name, size_t size, int strict);

static int __Pyx_InitStrings(__Pyx_StringTabEntry *t);

static struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_order, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___formBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___breakBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___loseRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainPair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___losePair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_action, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isOxygen(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSulfur(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSingle(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isDouble(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isTriple(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond___changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, short __pyx_v_order, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_action, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_addAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_addBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_getBonds(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_f_5rmgpy_8molecule_5group_5Group_getBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_hasAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_hasBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_removeAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_removeBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_sortAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_f_5rmgpy_8molecule_5group_5Group_copy(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_copy *__pyx_optional_args); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_f_5rmgpy_8molecule_5group_5Group_merge(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_split(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_clearLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_containsLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_fromAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_adjlist, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_toAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList *__pyx_optional_args); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_updateFingerprint(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isIsomorphic *__pyx_optional_args); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_findIsomorphism(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findIsomorphism *__pyx_optional_args); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic *__pyx_optional_args); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms *__pyx_optional_args); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isIdentical(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_standardizeAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_addExplicitLigands(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_standardizeGroup(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/

/* Module declarations from 'rmgpy.molecule.graph' */
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5graph_Vertex = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5graph_Edge = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5graph_Graph = 0;

/* Module declarations from 'rmgpy.molecule.atomtype' */
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType = 0;

/* Module declarations from 'cython' */

/* Module declarations from 'rmgpy.molecule.group' */
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5group_GroupBond = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5group_Group = 0;
#define __Pyx_MODULE_NAME "rmgpy.molecule.group"
int __pyx_module_is_main_rmgpy__molecule__group = 0;

/* Implementation of 'rmgpy.molecule.group' */
static PyObject *__pyx_builtin_Exception;
static PyObject *__pyx_builtin_property;
static PyObject *__pyx_builtin_range;
static PyObject *__pyx_builtin_any;
static PyObject *__pyx_builtin_zip;
static PyObject *__pyx_builtin_all;
static PyObject *__pyx_builtin_ValueError;
static PyObject *__pyx_builtin_TypeError;
static PyObject *__pyx_builtin_enumerate;
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom___init__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_atomType, PyObject *__pyx_v_radicalElectrons, PyObject *__pyx_v_charge, PyObject *__pyx_v_label, PyObject *__pyx_v_lonePairs); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_d); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6__str__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8__repr__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_10bonds(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_12copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16__formBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_26__losePair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_28applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_action); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_30equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond___init__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_atom1, PyObject *__pyx_v_atom2, PyObject *__pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_2__str__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_4__repr__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_6__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_8copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_10isSingle(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_12isDouble(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_14isTriple(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_16isBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_18__changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, short __pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_20applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_action); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_22equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group___init__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_atoms, PyObject *__pyx_v_multiplicity); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_2__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_4__getAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_6__setAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_atoms); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_8addAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_10addBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12getBonds(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_14getBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_16hasAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18hasBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_20removeAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_22removeBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_24sortAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_26copy(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_v_deep); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_28update(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_30merge(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_32split(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_38getLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_adjlist); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_44toAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_46updateFingerprint(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_48isIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_50findIsomorphism(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_56isIdentical(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_58standardizeAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_60addExplicitLigands(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_62standardizeGroup(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_GroupAtom(PyTypeObject *t, PyObject *a, PyObject *k); /*proto*/
static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_GroupBond(PyTypeObject *t, PyObject *a, PyObject *k); /*proto*/
static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_Group(PyTypeObject *t, PyObject *a, PyObject *k); /*proto*/
static char __pyx_k_[] = "";
static char __pyx_k_0[] = "[{0}]";
static char __pyx_k_B[] = "B";
static char __pyx_k_C[] = "C";
static char __pyx_k_D[] = "D";
static char __pyx_k_N[] = "N";
static char __pyx_k_O[] = "O";
static char __pyx_k_S[] = "S";
static char __pyx_k_T[] = "T";
static char __pyx_k_CO[] = "CO";
static char __pyx_k_CS[] = "CS";
static char __pyx_k_Oa[] = "Oa";
static char __pyx_k_Od[] = "Od";
static char __pyx_k_Sa[] = "Sa";
static char __pyx_k_Sd[] = "Sd";
static char __pyx_k_Si[] = "Si";
static char __pyx_k__2[] = ",";
static char __pyx_k_all[] = "all";
static char __pyx_k_any[] = "any";
static char __pyx_k_doc[] = "__doc__";
static char __pyx_k_zip[] = "zip";
static char __pyx_k_Edge[] = "Edge";
static char __pyx_k_copy[] = "copy";
static char __pyx_k_deep[] = "deep";
static char __pyx_k_init[] = "__init__";
static char __pyx_k_join[] = "join";
static char __pyx_k_main[] = "__main__";
static char __pyx_k_test[] = "__test__";
static char __pyx_k_Graph[] = "Graph";
static char __pyx_k_atom1[] = "atom1";
static char __pyx_k_atom2[] = "atom2";
static char __pyx_k_atoms[] = "atoms";
static char __pyx_k_bonds[] = "bonds";
static char __pyx_k_class[] = "__class__";
static char __pyx_k_edges[] = "edges";
static char __pyx_k_graph[] = "graph";
static char __pyx_k_group[] = "group";
static char __pyx_k_label[] = "label";
static char __pyx_k_merge[] = "merge";
static char __pyx_k_order[] = "order";
static char __pyx_k_other[] = "other";
static char __pyx_k_range[] = "range";
static char __pyx_k_split[] = "split";
static char __pyx_k_upper[] = "upper";
static char __pyx_k_Vertex[] = "Vertex";
static char __pyx_k_append[] = "append";
static char __pyx_k_charge[] = "charge";
static char __pyx_k_format[] = "format";
static char __pyx_k_import[] = "__import__";
static char __pyx_k_module[] = "__module__";
static char __pyx_k_update[] = "update";
static char __pyx_k_values[] = "values";
static char __pyx_k_addAtom[] = "addAtom";
static char __pyx_k_addBond[] = "addBond";
static char __pyx_k_adjlist[] = "adjlist";
static char __pyx_k_getBond[] = "getBond";
static char __pyx_k_hasAtom[] = "hasAtom";
static char __pyx_k_hasBond[] = "hasBond";
static char __pyx_k_prepare[] = "__prepare__";
static char __pyx_k_atomType[] = "atomType";
static char __pyx_k_atomtype[] = "atomtype";
static char __pyx_k_formBond[] = "__formBond";
static char __pyx_k_gainPair[] = "__gainPair";
static char __pyx_k_getAtoms[] = "__getAtoms";
static char __pyx_k_getBonds[] = "getBonds";
static char __pyx_k_isDouble[] = "isDouble";
static char __pyx_k_isOxygen[] = "isOxygen";
static char __pyx_k_isSingle[] = "isSingle";
static char __pyx_k_isSulfur[] = "isSulfur";
static char __pyx_k_isTriple[] = "isTriple";
static char __pyx_k_losePair[] = "__losePair";
static char __pyx_k_property[] = "property";
static char __pyx_k_qualname[] = "__qualname__";
static char __pyx_k_setAtoms[] = "__setAtoms";
static char __pyx_k_specific[] = "specific";
static char __pyx_k_vertices[] = "vertices";
static char __pyx_k_Exception[] = "Exception";
static char __pyx_k_FORM_BOND[] = "FORM_BOND";
static char __pyx_k_GAIN_PAIR[] = "GAIN_PAIR";
static char __pyx_k_LOSE_PAIR[] = "LOSE_PAIR";
static char __pyx_k_TypeError[] = "TypeError";
static char __pyx_k_atomTypes[] = "atomTypes";
static char __pyx_k_breakBond[] = "__breakBond";
static char __pyx_k_enumerate[] = "enumerate";
static char __pyx_k_isBenzene[] = "isBenzene";
static char __pyx_k_iteritems[] = "iteritems";
static char __pyx_k_lonePairs[] = "lonePairs";
static char __pyx_k_metaclass[] = "__metaclass__";
static char __pyx_k_sortAtoms[] = "sortAtoms";
static char __pyx_k_BREAK_BOND[] = "BREAK_BOND";
static char __pyx_k_ValueError[] = "ValueError";
static char __pyx_k_changeBond[] = "__changeBond";
static char __pyx_k_equivalent[] = "equivalent";
static char __pyx_k_formBond_2[] = "formBond";
static char __pyx_k_initialMap[] = "initialMap";
static char __pyx_k_pyx_vtable[] = "__pyx_vtable__";
static char __pyx_k_removeAtom[] = "removeAtom";
static char __pyx_k_removeBond[] = "removeBond";
static char __pyx_k_ActionError[] = "ActionError";
static char __pyx_k_CHANGE_BOND[] = "CHANGE_BOND";
static char __pyx_k_allElements[] = "allElements";
static char __pyx_k_applyAction[] = "applyAction";
static char __pyx_k_breakBond_2[] = "breakBond";
static char __pyx_k_gainRadical[] = "__gainRadical";
static char __pyx_k_getFeatures[] = "getFeatures";
static char __pyx_k_isIdentical[] = "isIdentical";
static char __pyx_k_loseRadical[] = "__loseRadical";
static char __pyx_k_GAIN_RADICAL[] = "GAIN_RADICAL";
static char __pyx_k_LOSE_RADICAL[] = "LOSE_RADICAL";
static char __pyx_k_isIsomorphic[] = "isIsomorphic";
static char __pyx_k_multiplicity[] = "multiplicity";
static char __pyx_k_nonSpecifics[] = "nonSpecifics";
static char __pyx_k_sortingLabel[] = "sortingLabel";
static char __pyx_k_GroupAtom_0_s[] = "<GroupAtom {0!s}>";
static char __pyx_k_GroupBond_0_r[] = "<GroupBond {0!r}>";
static char __pyx_k_connectivity1[] = "connectivity1";
static char __pyx_k_connectivity2[] = "connectivity2";
static char __pyx_k_connectivity3[] = "connectivity3";
static char __pyx_k_decrementBond[] = "decrementBond";
static char __pyx_k_incrementBond[] = "incrementBond";
static char __pyx_k_getLabeledAtom[] = "getLabeledAtom";
static char __pyx_k_findIsomorphism[] = "findIsomorphism";
static char __pyx_k_getLabeledAtoms[] = "getLabeledAtoms";
static char __pyx_k_toAdjacencyList[] = "toAdjacencyList";
static char __pyx_k_decrementRadical[] = "decrementRadical";
static char __pyx_k_incrementRadical[] = "incrementRadical";
static char __pyx_k_isSpecificCaseOf[] = "isSpecificCaseOf";
static char __pyx_k_radicalElectrons[] = "radicalElectrons";
static char __pyx_k_standardizeGroup[] = "standardizeGroup";
static char __pyx_k_clearLabeledAtoms[] = "clearLabeledAtoms";
static char __pyx_k_decrementLonePair[] = "decrementLonePair";
static char __pyx_k_fromAdjacencyList[] = "fromAdjacencyList";
static char __pyx_k_incrementLonePair[] = "incrementLonePair";
static char __pyx_k_updateFingerprint[] = "updateFingerprint";
static char __pyx_k_addExplicitLigands[] = "addExplicitLigands";
static char __pyx_k_containsLabeledAtom[] = "containsLabeledAtom";
static char __pyx_k_standardizeAtomType[] = "standardizeAtomType";
static char __pyx_k_isSubgraphIsomorphic[] = "isSubgraphIsomorphic";
static char __pyx_k_rmgpy_molecule_group[] = "rmgpy.molecule.group";
static char __pyx_k_findSubgraphIsomorphisms[] = "findSubgraphIsomorphisms";
static char __pyx_k_An_exception_class_for_errors_t[] = "\n    An exception class for errors that occur while applying reaction recipe\n    actions. Pass a string describing the circumstances that caused the\n    exceptional behavior.\n    ";
static char __pyx_k_No_atom_in_the_functional_group[] = "No atom in the functional group has the label \"{0}\".";
static char __pyx_k_This_module_provides_classes_an[] = "\nThis module provides classes and methods for working with molecular substructure\ngroups. These enable molecules to be searched for common motifs (e.g.\nreaction sites).\n";
static char __pyx_k_Got_a_0_object_for_parameter_oth[] = "Got a {0} object for parameter \"other\", when a Group object is required.";
static char __pyx_k_Unable_to_update_GroupAtom_Inval[] = "Unable to update GroupAtom: Invalid action {0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t[] = "Unable to update GroupAtom due to CHANGE_BOND action: Invalid order \"{0}\".";
static char __pyx_k_Unable_to_update_GroupBond_Inval[] = "Unable to update GroupBond: Invalid action {0}\".";
static char __pyx_k_Unable_to_update_GroupBond_due_t[] = "Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order \"{0}\" in set {1}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_2[] = "Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_3[] = "Unable to update GroupAtom due to FORM_BOND action: Invalid order \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_4[] = "Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_5[] = "Unable to update GroupAtom due to BREAK_BOND action: Invalid order \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_6[] = "Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_7[] = "Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_8[] = "Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_9[] = "Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupBond_due_t_2[] = "Unable to update GroupBond due to CHANGE_BOND action: Invalid order \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_10[] = "Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_11[] = "Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_12[] = "Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set \"{0}\".";
static PyObject *__pyx_kp_s_;
static PyObject *__pyx_kp_s_0;
static PyObject *__pyx_n_s_ActionError;
static PyObject *__pyx_kp_s_An_exception_class_for_errors_t;
static PyObject *__pyx_n_s_B;
static PyObject *__pyx_n_s_BREAK_BOND;
static PyObject *__pyx_n_s_C;
static PyObject *__pyx_n_s_CHANGE_BOND;
static PyObject *__pyx_n_s_CO;
static PyObject *__pyx_n_s_CS;
static PyObject *__pyx_n_s_D;
static PyObject *__pyx_n_s_Edge;
static PyObject *__pyx_n_s_Exception;
static PyObject *__pyx_n_s_FORM_BOND;
static PyObject *__pyx_n_s_GAIN_PAIR;
static PyObject *__pyx_n_s_GAIN_RADICAL;
static PyObject *__pyx_kp_s_Got_a_0_object_for_parameter_oth;
static PyObject *__pyx_n_s_Graph;
static PyObject *__pyx_kp_s_GroupAtom_0_s;
static PyObject *__pyx_kp_s_GroupBond_0_r;
static PyObject *__pyx_n_s_LOSE_PAIR;
static PyObject *__pyx_n_s_LOSE_RADICAL;
static PyObject *__pyx_n_s_N;
static PyObject *__pyx_kp_s_No_atom_in_the_functional_group;
static PyObject *__pyx_n_s_O;
static PyObject *__pyx_n_s_Oa;
static PyObject *__pyx_n_s_Od;
static PyObject *__pyx_n_s_S;
static PyObject *__pyx_n_s_Sa;
static PyObject *__pyx_n_s_Sd;
static PyObject *__pyx_n_s_Si;
static PyObject *__pyx_n_s_T;
static PyObject *__pyx_n_s_TypeError;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_Inval;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_10;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_11;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_12;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_2;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_3;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_4;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_5;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_6;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_7;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_8;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_9;
static PyObject *__pyx_kp_s_Unable_to_update_GroupBond_Inval;
static PyObject *__pyx_kp_s_Unable_to_update_GroupBond_due_t;
static PyObject *__pyx_kp_s_Unable_to_update_GroupBond_due_t_2;
static PyObject *__pyx_n_s_ValueError;
static PyObject *__pyx_n_s_Vertex;
static PyObject *__pyx_kp_s__2;
static PyObject *__pyx_n_s_addAtom;
static PyObject *__pyx_n_s_addBond;
static PyObject *__pyx_n_s_addExplicitLigands;
static PyObject *__pyx_n_s_adjlist;
static PyObject *__pyx_n_s_all;
static PyObject *__pyx_n_s_allElements;
static PyObject *__pyx_n_s_any;
static PyObject *__pyx_n_s_append;
static PyObject *__pyx_n_s_applyAction;
static PyObject *__pyx_n_s_atom1;
static PyObject *__pyx_n_s_atom2;
static PyObject *__pyx_n_s_atomType;
static PyObject *__pyx_n_s_atomTypes;
static PyObject *__pyx_n_s_atoms;
static PyObject *__pyx_n_s_atomtype;
static PyObject *__pyx_n_s_bonds;
static PyObject *__pyx_n_s_breakBond;
static PyObject *__pyx_n_s_breakBond_2;
static PyObject *__pyx_n_s_changeBond;
static PyObject *__pyx_n_s_charge;
static PyObject *__pyx_n_s_class;
static PyObject *__pyx_n_s_clearLabeledAtoms;
static PyObject *__pyx_n_s_connectivity1;
static PyObject *__pyx_n_s_connectivity2;
static PyObject *__pyx_n_s_connectivity3;
static PyObject *__pyx_n_s_containsLabeledAtom;
static PyObject *__pyx_n_s_copy;
static PyObject *__pyx_n_s_decrementBond;
static PyObject *__pyx_n_s_decrementLonePair;
static PyObject *__pyx_n_s_decrementRadical;
static PyObject *__pyx_n_s_deep;
static PyObject *__pyx_n_s_doc;
static PyObject *__pyx_n_s_edges;
static PyObject *__pyx_n_s_enumerate;
static PyObject *__pyx_n_s_equivalent;
static PyObject *__pyx_n_s_findIsomorphism;
static PyObject *__pyx_n_s_findSubgraphIsomorphisms;
static PyObject *__pyx_n_s_formBond;
static PyObject *__pyx_n_s_formBond_2;
static PyObject *__pyx_n_s_format;
static PyObject *__pyx_n_s_fromAdjacencyList;
static PyObject *__pyx_n_s_gainPair;
static PyObject *__pyx_n_s_gainRadical;
static PyObject *__pyx_n_s_getAtoms;
static PyObject *__pyx_n_s_getBond;
static PyObject *__pyx_n_s_getBonds;
static PyObject *__pyx_n_s_getFeatures;
static PyObject *__pyx_n_s_getLabeledAtom;
static PyObject *__pyx_n_s_getLabeledAtoms;
static PyObject *__pyx_n_s_graph;
static PyObject *__pyx_n_s_group;
static PyObject *__pyx_n_s_hasAtom;
static PyObject *__pyx_n_s_hasBond;
static PyObject *__pyx_n_s_import;
static PyObject *__pyx_n_s_incrementBond;
static PyObject *__pyx_n_s_incrementLonePair;
static PyObject *__pyx_n_s_incrementRadical;
static PyObject *__pyx_n_s_init;
static PyObject *__pyx_n_s_initialMap;
static PyObject *__pyx_n_s_isBenzene;
static PyObject *__pyx_n_s_isDouble;
static PyObject *__pyx_n_s_isIdentical;
static PyObject *__pyx_n_s_isIsomorphic;
static PyObject *__pyx_n_s_isOxygen;
static PyObject *__pyx_n_s_isSingle;
static PyObject *__pyx_n_s_isSpecificCaseOf;
static PyObject *__pyx_n_s_isSubgraphIsomorphic;
static PyObject *__pyx_n_s_isSulfur;
static PyObject *__pyx_n_s_isTriple;
static PyObject *__pyx_n_s_iteritems;
static PyObject *__pyx_n_s_join;
static PyObject *__pyx_n_s_label;
static PyObject *__pyx_n_s_lonePairs;
static PyObject *__pyx_n_s_losePair;
static PyObject *__pyx_n_s_loseRadical;
static PyObject *__pyx_n_s_main;
static PyObject *__pyx_n_s_merge;
static PyObject *__pyx_n_s_metaclass;
static PyObject *__pyx_n_s_module;
static PyObject *__pyx_n_s_multiplicity;
static PyObject *__pyx_n_s_nonSpecifics;
static PyObject *__pyx_n_s_order;
static PyObject *__pyx_n_s_other;
static PyObject *__pyx_n_s_prepare;
static PyObject *__pyx_n_s_property;
static PyObject *__pyx_n_s_pyx_vtable;
static PyObject *__pyx_n_s_qualname;
static PyObject *__pyx_n_s_radicalElectrons;
static PyObject *__pyx_n_s_range;
static PyObject *__pyx_n_s_removeAtom;
static PyObject *__pyx_n_s_removeBond;
static PyObject *__pyx_n_s_rmgpy_molecule_group;
static PyObject *__pyx_n_s_setAtoms;
static PyObject *__pyx_n_s_sortAtoms;
static PyObject *__pyx_n_s_sortingLabel;
static PyObject *__pyx_n_s_specific;
static PyObject *__pyx_n_s_split;
static PyObject *__pyx_n_s_standardizeAtomType;
static PyObject *__pyx_n_s_standardizeGroup;
static PyObject *__pyx_n_s_test;
static PyObject *__pyx_n_s_toAdjacencyList;
static PyObject *__pyx_n_s_update;
static PyObject *__pyx_n_s_updateFingerprint;
static PyObject *__pyx_n_s_upper;
static PyObject *__pyx_n_s_values;
static PyObject *__pyx_n_s_vertices;
static PyObject *__pyx_n_s_zip;
static PyObject *__pyx_int_0;
static PyObject *__pyx_int_1;
static PyObject *__pyx_int_2;
static PyObject *__pyx_int_3;
static PyObject *__pyx_int_4;

/* "rmgpy/molecule/group.py":76
 *     """
 * 
 *     def __init__(self, atomType=None, radicalElectrons=None, charge=None, label='', lonePairs=None):             # <<<<<<<<<<<<<<
 *         Vertex.__init__(self)
 *         self.atomType = atomType or []
 */

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_atomType = 0;
  PyObject *__pyx_v_radicalElectrons = 0;
  PyObject *__pyx_v_charge = 0;
  PyObject *__pyx_v_label = 0;
  PyObject *__pyx_v_lonePairs = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__init__ (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atomType,&__pyx_n_s_radicalElectrons,&__pyx_n_s_charge,&__pyx_n_s_label,&__pyx_n_s_lonePairs,0};
    PyObject* values[5] = {0,0,0,0,0};
    values[0] = ((PyObject *)Py_None);
    values[1] = ((PyObject *)Py_None);
    values[2] = ((PyObject *)Py_None);
    values[3] = ((PyObject *)__pyx_kp_s_);
    values[4] = ((PyObject *)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
        case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atomType);
          if (value) { values[0] = value; kw_args--; }
        }
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_radicalElectrons);
          if (value) { values[1] = value; kw_args--; }
        }
        case  2:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_charge);
          if (value) { values[2] = value; kw_args--; }
        }
        case  3:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_label);
          if (value) { values[3] = value; kw_args--; }
        }
        case  4:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_lonePairs);
          if (value) { values[4] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__init__") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 76; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
        case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_atomType = values[0];
    __pyx_v_radicalElectrons = values[1];
    __pyx_v_charge = values[2];
    __pyx_v_label = values[3];
    __pyx_v_lonePairs = values[4];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("__init__", 0, 0, 5, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 76; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return -1;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom___init__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), __pyx_v_atomType, __pyx_v_radicalElectrons, __pyx_v_charge, __pyx_v_label, __pyx_v_lonePairs);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom___init__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_atomType, PyObject *__pyx_v_radicalElectrons, PyObject *__pyx_v_charge, PyObject *__pyx_v_label, PyObject *__pyx_v_lonePairs) {
  Py_ssize_t __pyx_v_index;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_t_8;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__init__", 0);

  /* "rmgpy/molecule/group.py":77
 * 
 *     def __init__(self, atomType=None, radicalElectrons=None, charge=None, label='', lonePairs=None):
 *         Vertex.__init__(self)             # <<<<<<<<<<<<<<
 *         self.atomType = atomType or []
 *         for index in range(len(self.atomType)):
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)), __pyx_n_s_init); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 77; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, ((PyObject *)__pyx_v_self)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 77; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 77; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(((PyObject *)__pyx_v_self));
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, ((PyObject *)__pyx_v_self));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 77; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":78
 *     def __init__(self, atomType=None, radicalElectrons=None, charge=None, label='', lonePairs=None):
 *         Vertex.__init__(self)
 *         self.atomType = atomType or []             # <<<<<<<<<<<<<<
 *         for index in range(len(self.atomType)):
 *             if isinstance(self.atomType[index], str):
 */
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_v_atomType); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 78; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_5) {
    goto __pyx_L4_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_atomType))||((__pyx_v_atomType) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_atomType)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 78; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_atomType);
    __pyx_t_1 = __pyx_v_atomType;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 78; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_1 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L3_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":79
 *         Vertex.__init__(self)
 *         self.atomType = atomType or []
 *         for index in range(len(self.atomType)):             # <<<<<<<<<<<<<<
 *             if isinstance(self.atomType[index], str):
 *                 self.atomType[index] = atomTypes[self.atomType[index]]
 */
  __pyx_t_1 = __pyx_v_self->atomType;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 79; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_6 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 79; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (__pyx_t_7 = 0; __pyx_t_7 < __pyx_t_6; __pyx_t_7+=1) {
    __pyx_v_index = __pyx_t_7;

    /* "rmgpy/molecule/group.py":80
 *         self.atomType = atomType or []
 *         for index in range(len(self.atomType)):
 *             if isinstance(self.atomType[index], str):             # <<<<<<<<<<<<<<
 *                 self.atomType[index] = atomTypes[self.atomType[index]]
 *         self.radicalElectrons = radicalElectrons or []
 */
    if (unlikely(__pyx_v_self->atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 80; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->atomType, __pyx_v_index, Py_ssize_t, 1, PyInt_FromSsize_t, 1, 1, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 80; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_5 = PyString_Check(__pyx_t_1); 
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_8 = (__pyx_t_5 != 0);
    if (__pyx_t_8) {

      /* "rmgpy/molecule/group.py":81
 *         for index in range(len(self.atomType)):
 *             if isinstance(self.atomType[index], str):
 *                 self.atomType[index] = atomTypes[self.atomType[index]]             # <<<<<<<<<<<<<<
 *         self.radicalElectrons = radicalElectrons or []
 *         self.charge = charge or []
 */
      __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 81; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      if (unlikely(__pyx_v_self->atomType == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 81; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_self->atomType, __pyx_v_index, Py_ssize_t, 1, PyInt_FromSsize_t, 1, 1, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 81; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_4 = PyObject_GetItem(__pyx_t_1, __pyx_t_2); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 81; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (unlikely(__pyx_v_self->atomType == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 81; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      if (unlikely(__Pyx_SetItemInt(__pyx_v_self->atomType, __pyx_v_index, __pyx_t_4, Py_ssize_t, 1, PyInt_FromSsize_t, 1, 1, 1) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 81; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      goto __pyx_L7;
    }
    __pyx_L7:;
  }

  /* "rmgpy/molecule/group.py":82
 *             if isinstance(self.atomType[index], str):
 *                 self.atomType[index] = atomTypes[self.atomType[index]]
 *         self.radicalElectrons = radicalElectrons or []             # <<<<<<<<<<<<<<
 *         self.charge = charge or []
 *         self.label = label
 */
  __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_v_radicalElectrons); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 82; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_8) {
    goto __pyx_L9_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_radicalElectrons))||((__pyx_v_radicalElectrons) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_radicalElectrons)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 82; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_radicalElectrons);
    __pyx_t_4 = __pyx_v_radicalElectrons;
    goto __pyx_L8_bool_binop_done;
  }
  __pyx_L9_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 82; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_4 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L8_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_4);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = ((PyObject*)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":83
 *                 self.atomType[index] = atomTypes[self.atomType[index]]
 *         self.radicalElectrons = radicalElectrons or []
 *         self.charge = charge or []             # <<<<<<<<<<<<<<
 *         self.label = label
 *         self.lonePairs = lonePairs or []
 */
  __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_v_charge); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 83; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_8) {
    goto __pyx_L11_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_charge))||((__pyx_v_charge) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_charge)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 83; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_charge);
    __pyx_t_4 = __pyx_v_charge;
    goto __pyx_L10_bool_binop_done;
  }
  __pyx_L11_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 83; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_4 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L10_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_4);
  __Pyx_GOTREF(__pyx_v_self->charge);
  __Pyx_DECREF(__pyx_v_self->charge);
  __pyx_v_self->charge = ((PyObject*)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":84
 *         self.radicalElectrons = radicalElectrons or []
 *         self.charge = charge or []
 *         self.label = label             # <<<<<<<<<<<<<<
 *         self.lonePairs = lonePairs or []
 * 
 */
  if (!(likely(PyString_CheckExact(__pyx_v_label))||((__pyx_v_label) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_v_label)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 84; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_4 = __pyx_v_label;
  __Pyx_INCREF(__pyx_t_4);
  __Pyx_GIVEREF(__pyx_t_4);
  __Pyx_GOTREF(__pyx_v_self->label);
  __Pyx_DECREF(__pyx_v_self->label);
  __pyx_v_self->label = ((PyObject*)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":85
 *         self.charge = charge or []
 *         self.label = label
 *         self.lonePairs = lonePairs or []             # <<<<<<<<<<<<<<
 * 
 *     def __reduce__(self):
 */
  __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_v_lonePairs); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 85; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_8) {
    goto __pyx_L13_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_lonePairs))||((__pyx_v_lonePairs) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_lonePairs)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 85; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_lonePairs);
    __pyx_t_4 = __pyx_v_lonePairs;
    goto __pyx_L12_bool_binop_done;
  }
  __pyx_L13_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 85; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_4 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L12_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_4);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":76
 *     """
 * 
 *     def __init__(self, atomType=None, radicalElectrons=None, charge=None, label='', lonePairs=None):             # <<<<<<<<<<<<<<
 *         Vertex.__init__(self)
 *         self.atomType = atomType or []
 */

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":87
 *         self.lonePairs = lonePairs or []
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_3__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__[] = "\n        A helper function used when pickling an object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_3__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__reduce__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_v_d = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_v_a = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_t_3;
  int __pyx_t_4;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__reduce__", 0);

  /* "rmgpy/molecule/group.py":91
 *         A helper function used when pickling an object.
 *         """
 *         d = {             # <<<<<<<<<<<<<<
 *             'edges': self.edges,
 *             'connectivity1': self.connectivity1,
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 91; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);

  /* "rmgpy/molecule/group.py":92
 *         """
 *         d = {
 *             'edges': self.edges,             # <<<<<<<<<<<<<<
 *             'connectivity1': self.connectivity1,
 *             'connectivity2': self.connectivity2,
 */
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_edges, __pyx_v_self->__pyx_base.edges) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 91; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":93
 *         d = {
 *             'edges': self.edges,
 *             'connectivity1': self.connectivity1,             # <<<<<<<<<<<<<<
 *             'connectivity2': self.connectivity2,
 *             'connectivity3': self.connectivity3,
 */
  __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_self->__pyx_base.connectivity1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 93; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_connectivity1, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 91; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":94
 *             'edges': self.edges,
 *             'connectivity1': self.connectivity1,
 *             'connectivity2': self.connectivity2,             # <<<<<<<<<<<<<<
 *             'connectivity3': self.connectivity3,
 *             'sortingLabel': self.sortingLabel,
 */
  __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_self->__pyx_base.connectivity2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 94; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_connectivity2, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 91; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":95
 *             'connectivity1': self.connectivity1,
 *             'connectivity2': self.connectivity2,
 *             'connectivity3': self.connectivity3,             # <<<<<<<<<<<<<<
 *             'sortingLabel': self.sortingLabel,
 *         }
 */
  __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_self->__pyx_base.connectivity3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 95; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_connectivity3, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 91; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":96
 *             'connectivity2': self.connectivity2,
 *             'connectivity3': self.connectivity3,
 *             'sortingLabel': self.sortingLabel,             # <<<<<<<<<<<<<<
 *         }
 *         atomType = self.atomType
 */
  __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_self->__pyx_base.sortingLabel); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 96; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_sortingLabel, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 91; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_d = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":98
 *             'sortingLabel': self.sortingLabel,
 *         }
 *         atomType = self.atomType             # <<<<<<<<<<<<<<
 *         if atomType is not None:
 *             atomType = [a.label for a in atomType]
 */
  __pyx_t_1 = __pyx_v_self->atomType;
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":99
 *         }
 *         atomType = self.atomType
 *         if atomType is not None:             # <<<<<<<<<<<<<<
 *             atomType = [a.label for a in atomType]
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)
 */
  __pyx_t_3 = (__pyx_v_atomType != ((PyObject*)Py_None));
  __pyx_t_4 = (__pyx_t_3 != 0);
  if (__pyx_t_4) {

    /* "rmgpy/molecule/group.py":100
 *         atomType = self.atomType
 *         if atomType is not None:
 *             atomType = [a.label for a in atomType]             # <<<<<<<<<<<<<<
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)
 * 
 */
    __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (unlikely(__pyx_v_atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_5 = 0;
    for (;;) {
      if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_6 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_6); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_6 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_a, __pyx_t_6);
      __pyx_t_6 = 0;
      __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_v_a, __pyx_n_s_label); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_6))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF_SET(__pyx_v_atomType, ((PyObject*)__pyx_t_1));
    __pyx_t_1 = 0;
    goto __pyx_L3;
  }
  __pyx_L3:;

  /* "rmgpy/molecule/group.py":101
 *         if atomType is not None:
 *             atomType = [a.label for a in atomType]
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)             # <<<<<<<<<<<<<<
 * 
 *     def __setstate__(self, d):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(5); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_atomType);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_atomType);
  __Pyx_GIVEREF(__pyx_v_atomType);
  __Pyx_INCREF(__pyx_v_self->radicalElectrons);
  PyTuple_SET_ITEM(__pyx_t_1, 1, __pyx_v_self->radicalElectrons);
  __Pyx_GIVEREF(__pyx_v_self->radicalElectrons);
  __Pyx_INCREF(__pyx_v_self->charge);
  PyTuple_SET_ITEM(__pyx_t_1, 2, __pyx_v_self->charge);
  __Pyx_GIVEREF(__pyx_v_self->charge);
  __Pyx_INCREF(__pyx_v_self->label);
  PyTuple_SET_ITEM(__pyx_t_1, 3, __pyx_v_self->label);
  __Pyx_GIVEREF(__pyx_v_self->label);
  __Pyx_INCREF(__pyx_v_self->lonePairs);
  PyTuple_SET_ITEM(__pyx_t_1, 4, __pyx_v_self->lonePairs);
  __Pyx_GIVEREF(__pyx_v_self->lonePairs);
  __pyx_t_2 = PyTuple_New(3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)));
  PyTuple_SET_ITEM(__pyx_t_2, 0, ((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)));
  __Pyx_GIVEREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)));
  PyTuple_SET_ITEM(__pyx_t_2, 1, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_d);
  PyTuple_SET_ITEM(__pyx_t_2, 2, __pyx_v_d);
  __Pyx_GIVEREF(__pyx_v_d);
  __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":87
 *         self.lonePairs = lonePairs or []
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__reduce__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_d);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XDECREF(__pyx_v_a);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":103
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)
 * 
 *     def __setstate__(self, d):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when unpickling an object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5__setstate__(PyObject *__pyx_v_self, PyObject *__pyx_v_d); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__[] = "\n        A helper function used when unpickling an object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5__setstate__(PyObject *__pyx_v_self, PyObject *__pyx_v_d) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__setstate__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_d));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_d) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  short __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__setstate__", 0);

  /* "rmgpy/molecule/group.py":107
 *         A helper function used when unpickling an object.
 *         """
 *         self.edges = d['edges']             # <<<<<<<<<<<<<<
 *         self.connectivity1 = d['connectivity1']
 *         self.connectivity2 = d['connectivity2']
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_edges); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 107; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  if (!(likely(PyDict_CheckExact(__pyx_t_1))||((__pyx_t_1) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "dict", Py_TYPE(__pyx_t_1)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 107; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->__pyx_base.edges);
  __Pyx_DECREF(__pyx_v_self->__pyx_base.edges);
  __pyx_v_self->__pyx_base.edges = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":108
 *         """
 *         self.edges = d['edges']
 *         self.connectivity1 = d['connectivity1']             # <<<<<<<<<<<<<<
 *         self.connectivity2 = d['connectivity2']
 *         self.connectivity3 = d['connectivity3']
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_connectivity1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 108; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_2 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 108; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_self->__pyx_base.connectivity1 = __pyx_t_2;

  /* "rmgpy/molecule/group.py":109
 *         self.edges = d['edges']
 *         self.connectivity1 = d['connectivity1']
 *         self.connectivity2 = d['connectivity2']             # <<<<<<<<<<<<<<
 *         self.connectivity3 = d['connectivity3']
 *         self.sortingLabel = d['sortingLabel']
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_connectivity2); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 109; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_2 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 109; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_self->__pyx_base.connectivity2 = __pyx_t_2;

  /* "rmgpy/molecule/group.py":110
 *         self.connectivity1 = d['connectivity1']
 *         self.connectivity2 = d['connectivity2']
 *         self.connectivity3 = d['connectivity3']             # <<<<<<<<<<<<<<
 *         self.sortingLabel = d['sortingLabel']
 * 
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_connectivity3); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 110; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_2 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 110; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_self->__pyx_base.connectivity3 = __pyx_t_2;

  /* "rmgpy/molecule/group.py":111
 *         self.connectivity2 = d['connectivity2']
 *         self.connectivity3 = d['connectivity3']
 *         self.sortingLabel = d['sortingLabel']             # <<<<<<<<<<<<<<
 * 
 *     def __str__(self):
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_sortingLabel); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 111; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_2 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_self->__pyx_base.sortingLabel = __pyx_t_2;

  /* "rmgpy/molecule/group.py":103
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)
 * 
 *     def __setstate__(self, d):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when unpickling an object.
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__setstate__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":113
 *         self.sortingLabel = d['sortingLabel']
 * 
 *     def __str__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a human-readable string representation of the object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_7__str__(PyObject *__pyx_v_self); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_6__str__[] = "\n        Return a human-readable string representation of the object.\n        ";
#if CYTHON_COMPILING_IN_CPYTHON
struct wrapperbase __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_6__str__;
#endif
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_7__str__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__str__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6__str__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6__str__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_v_a = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  PyObject *__pyx_t_7 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__str__", 0);

  /* "rmgpy/molecule/group.py":117
 *         Return a human-readable string representation of the object.
 *         """
 *         return '[{0}]'.format(','.join([repr(a.label) for a in self.atomType]))             # <<<<<<<<<<<<<<
 * 
 *     def __repr__(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_0, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_4 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_4); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_4)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_6 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_5); __Pyx_INCREF(__pyx_t_6); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_6 = PySequence_ITEM(__pyx_t_4, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_a, __pyx_t_6);
    __pyx_t_6 = 0;
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_v_a, __pyx_n_s_label); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_7 = PyObject_Repr(__pyx_t_6); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_3, (PyObject*)__pyx_t_7))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  }
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_t_4 = __Pyx_PyString_Join(__pyx_kp_s__2, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_7 = PyTuple_New(1+1); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    PyTuple_SET_ITEM(__pyx_t_7, 0+1, __pyx_t_4);
    __Pyx_GIVEREF(__pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_7, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 117; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":113
 *         self.sortingLabel = d['sortingLabel']
 * 
 *     def __str__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a human-readable string representation of the object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_7);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__str__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_a);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":119
 *         return '[{0}]'.format(','.join([repr(a.label) for a in self.atomType]))
 * 
 *     def __repr__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a representation that can be used to reconstruct the object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9__repr__(PyObject *__pyx_v_self); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_8__repr__[] = "\n        Return a representation that can be used to reconstruct the object.\n        ";
#if CYTHON_COMPILING_IN_CPYTHON
struct wrapperbase __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_8__repr__;
#endif
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9__repr__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__repr__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8__repr__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8__repr__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__repr__", 0);

  /* "rmgpy/molecule/group.py":123
 *         Return a representation that can be used to reconstruct the object.
 *         """
 *         return "<GroupAtom {0!s}>".format(self)             # <<<<<<<<<<<<<<
 * 
 *     @property
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_GroupAtom_0_s, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 123; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, ((PyObject *)__pyx_v_self)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 123; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 123; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(((PyObject *)__pyx_v_self));
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, ((PyObject *)__pyx_v_self));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 123; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":119
 *         return '[{0}]'.format(','.join([repr(a.label) for a in self.atomType]))
 * 
 *     def __repr__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a representation that can be used to reconstruct the object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__repr__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":126
 * 
 *     @property
 *     def bonds(self): return self.edges             # <<<<<<<<<<<<<<
 * 
 *     def copy(self):
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_11bonds(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_11bonds(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("bonds (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_10bonds(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_10bonds(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("bonds", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->__pyx_base.edges);
  __pyx_r = __pyx_v_self->__pyx_base.edges;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":128
 *     def bonds(self): return self.edges
 * 
 *     def copy(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a deep copy of the :class:`GroupAtom` object. Modifying the
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_copy); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5graph_Vertex))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":133
 *         attributes of the copy will not affect the original.
 *         """
 *         return GroupAtom(self.atomType[:], self.radicalElectrons[:], self.charge[:], self.label)             # <<<<<<<<<<<<<<
 * 
 *     def __changeBond(self, order):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 133; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_PyList_GetSlice(__pyx_v_self->atomType, 0, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 133; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 133; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __Pyx_PyList_GetSlice(__pyx_v_self->radicalElectrons, 0, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 133; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (unlikely(__pyx_v_self->charge == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 133; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_3 = __Pyx_PyList_GetSlice(__pyx_v_self->charge, 0, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 133; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_4 = PyTuple_New(4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 133; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  PyTuple_SET_ITEM(__pyx_t_4, 1, __pyx_t_2);
  __Pyx_GIVEREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_4, 2, __pyx_t_3);
  __Pyx_GIVEREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_v_self->label);
  PyTuple_SET_ITEM(__pyx_t_4, 3, __pyx_v_self->label);
  __Pyx_GIVEREF(__pyx_v_self->label);
  __pyx_t_1 = 0;
  __pyx_t_2 = 0;
  __pyx_t_3 = 0;
  __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_t_4, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 133; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_t_3);
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":128
 *     def bonds(self): return self.edges
 * 
 *     def copy(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a deep copy of the :class:`GroupAtom` object. Modifying the
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_12copy[] = "\n        Return a deep copy of the :class:`GroupAtom` object. Modifying the\n        attributes of the copy will not affect the original.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("copy (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_12copy(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_12copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_copy(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":135
 *         return GroupAtom(self.atomType[:], self.radicalElectrons[:], self.charge[:], self.label)
 * 
 *     def __changeBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a CHANGE_BOND action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_order, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  int __pyx_t_8;
  PyObject *__pyx_t_9 = NULL;
  PyObject *__pyx_t_10 = NULL;
  int __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__changeBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_changeBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 135; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_order); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 135; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 135; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 135; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 135; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":141
 *         in bond order, and should be 1 or -1.
 *         """
 *         atomType = []             # <<<<<<<<<<<<<<
 *         for atom in self.atomType:
 *             if order == 1:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 141; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":142
 *         """
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 142; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 142; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 142; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":145
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)
 *             elif order == -1:             # <<<<<<<<<<<<<<
 *                 atomType.extend(atom.decrementBond)
 *             else:
 */
    switch (__pyx_v_order) {

      /* "rmgpy/molecule/group.py":143
 *         atomType = []
 *         for atom in self.atomType:
 *             if order == 1:             # <<<<<<<<<<<<<<
 *                 atomType.extend(atom.incrementBond)
 *             elif order == -1:
 */
      case 1:

      /* "rmgpy/molecule/group.py":144
 *         for atom in self.atomType:
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)             # <<<<<<<<<<<<<<
 *             elif order == -1:
 *                 atomType.extend(atom.decrementBond)
 */
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_incrementBond); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_8 = __Pyx_PyList_Extend(__pyx_v_atomType, __pyx_t_2); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      break;

      /* "rmgpy/molecule/group.py":145
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)
 *             elif order == -1:             # <<<<<<<<<<<<<<
 *                 atomType.extend(atom.decrementBond)
 *             else:
 */
      case -1:

      /* "rmgpy/molecule/group.py":146
 *                 atomType.extend(atom.incrementBond)
 *             elif order == -1:
 *                 atomType.extend(atom.decrementBond)             # <<<<<<<<<<<<<<
 *             else:
 *                 raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 */
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_decrementBond); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_8 = __Pyx_PyList_Extend(__pyx_v_atomType, __pyx_t_2); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      break;
      default:

      /* "rmgpy/molecule/group.py":148
 *                 atomType.extend(atom.decrementBond)
 *             else:
 *                 raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Invalid order "{0}".'.format(order))             # <<<<<<<<<<<<<<
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_5 = __Pyx_PyInt_From_short(__pyx_v_order); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      __pyx_t_9 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_9 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_9)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_9);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_9) {
        __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        __Pyx_GOTREF(__pyx_t_6);
      } else {
        __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_9); __Pyx_GIVEREF(__pyx_t_9); __pyx_t_9 = NULL;
        PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_t_5);
        __Pyx_GIVEREF(__pyx_t_5);
        __pyx_t_5 = 0;
        __pyx_t_6 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_10, NULL); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_3 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_3)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_3);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_3) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_6); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_t_6);
        __Pyx_GIVEREF(__pyx_t_6);
        __pyx_t_6 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_10, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 148; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      break;
    }

    /* "rmgpy/molecule/group.py":142
 *         """
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":149
 *             else:
 *                 raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 *         if len(atomType) == 0:             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 */
  __pyx_t_7 = PyList_GET_SIZE(__pyx_v_atomType); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 149; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_11 = ((__pyx_t_7 == 0) != 0);
  if (__pyx_t_11) {

    /* "rmgpy/molecule/group.py":150
 *                 raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_10 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_2, __pyx_n_s_format); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_10);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_10))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_10);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_10);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_10, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_10, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_3 = PyTuple_New(1+1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_3, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_10, __pyx_t_3, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    }
    __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
    __pyx_t_10 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_10)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_10);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_10) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_3 = PyTuple_New(1+1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_10); __Pyx_GIVEREF(__pyx_t_10); __pyx_t_10 = NULL;
      PyTuple_SET_ITEM(__pyx_t_3, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_3, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 150; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":152
 *             raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))             # <<<<<<<<<<<<<<
 * 
 *     def __formBond(self, order):
 */
  __pyx_t_1 = PySet_New(__pyx_v_atomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 152; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 152; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyList_Type))), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 152; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":135
 *         return GroupAtom(self.atomType[:], self.radicalElectrons[:], self.charge[:], self.label)
 * 
 *     def __changeBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a CHANGE_BOND action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_9);
  __Pyx_XDECREF(__pyx_t_10);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond[] = "\n        Update the atom group as a result of applying a CHANGE_BOND action,\n        where `order` specifies whether the bond is incremented or decremented\n        in bond order, and should be 1 or -1.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order) {
  short __pyx_v_order;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__changeBond (wrapper)", 0);
  assert(__pyx_arg_order); {
    __pyx_v_order = __Pyx_PyInt_As_short(__pyx_arg_order); if (unlikely((__pyx_v_order == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 135; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_order));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_order) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__changeBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___changeBond(__pyx_v_self, __pyx_v_order, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 135; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":154
 *         self.atomType = list(set(atomType))
 * 
 *     def __formBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a FORM_BOND action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___formBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  PyObject *__pyx_t_8 = NULL;
  Py_ssize_t __pyx_t_9;
  int __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__formBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_formBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 154; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_order); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 154; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 154; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_order);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_order);
        __Pyx_GIVEREF(__pyx_v_order);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 154; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":160
 *         'S' (since we only allow forming of single bonds).
 *         """
 *         if order != 'S':             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_order, __pyx_n_s_S, Py_NE)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 160; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = (__pyx_t_6 != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":161
 *         """
 *         if order != 'S':
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))             # <<<<<<<<<<<<<<
 *         atomType = []
 *         for atom in self.atomType:
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_3, __pyx_n_s_format); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_5);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_5, function);
      }
    }
    if (!__pyx_t_4) {
      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_v_order); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      __Pyx_INCREF(__pyx_v_order);
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_v_order);
      __Pyx_GIVEREF(__pyx_v_order);
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_8, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":162
 *         if order != 'S':
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []             # <<<<<<<<<<<<<<
 *         for atom in self.atomType:
 *             atomType.extend(atom.formBond)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 162; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":163
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             atomType.extend(atom.formBond)
 *         if len(atomType) == 0:
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 163; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_9 = 0;
  for (;;) {
    if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_9); __Pyx_INCREF(__pyx_t_2); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 163; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 163; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":164
 *         atomType = []
 *         for atom in self.atomType:
 *             atomType.extend(atom.formBond)             # <<<<<<<<<<<<<<
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_formBond_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 164; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_10 = __Pyx_PyList_Extend(__pyx_v_atomType, __pyx_t_2); if (unlikely(__pyx_t_10 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 164; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":163
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             atomType.extend(atom.formBond)
 *         if len(atomType) == 0:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":165
 *         for atom in self.atomType:
 *             atomType.extend(atom.formBond)
 *         if len(atomType) == 0:             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 */
  __pyx_t_9 = PyList_GET_SIZE(__pyx_v_atomType); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = ((__pyx_t_9 == 0) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":166
 *             atomType.extend(atom.formBond)
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_4, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_self->atomType); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_4, NULL); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_8); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_8);
      __Pyx_GIVEREF(__pyx_t_8);
      __pyx_t_8 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":168
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))             # <<<<<<<<<<<<<<
 * 
 *     def __breakBond(self, order):
 */
  __pyx_t_1 = PySet_New(__pyx_v_atomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyList_Type))), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":154
 *         self.atomType = list(set(atomType))
 * 
 *     def __formBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a FORM_BOND action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_8);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__formBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_16__formBond[] = "\n        Update the atom group as a result of applying a FORM_BOND action,\n        where `order` specifies the order of the forming bond, and should be\n        'S' (since we only allow forming of single bonds).\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__formBond (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_order), (&PyString_Type), 1, "order", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 154; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16__formBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject*)__pyx_v_order));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16__formBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__formBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___formBond(__pyx_v_self, __pyx_v_order, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 154; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__formBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":170
 *         self.atomType = list(set(atomType))
 * 
 *     def __breakBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a BREAK_BOND action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___breakBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  PyObject *__pyx_t_8 = NULL;
  Py_ssize_t __pyx_t_9;
  int __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__breakBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_breakBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_order); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_order);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_order);
        __Pyx_GIVEREF(__pyx_v_order);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":176
 *         'S' (since we only allow breaking of single bonds).
 *         """
 *         if order != 'S':             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_order, __pyx_n_s_S, Py_NE)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 176; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = (__pyx_t_6 != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":177
 *         """
 *         if order != 'S':
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))             # <<<<<<<<<<<<<<
 *         atomType = []
 *         for atom in self.atomType:
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_5, __pyx_n_s_format); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_5);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_5, function);
      }
    }
    if (!__pyx_t_4) {
      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_v_order); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      __Pyx_INCREF(__pyx_v_order);
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_v_order);
      __Pyx_GIVEREF(__pyx_v_order);
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_8, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":178
 *         if order != 'S':
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []             # <<<<<<<<<<<<<<
 *         for atom in self.atomType:
 *             atomType.extend(atom.breakBond)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 178; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":179
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             atomType.extend(atom.breakBond)
 *         if len(atomType) == 0:
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 179; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_9 = 0;
  for (;;) {
    if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_9); __Pyx_INCREF(__pyx_t_2); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 179; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 179; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":180
 *         atomType = []
 *         for atom in self.atomType:
 *             atomType.extend(atom.breakBond)             # <<<<<<<<<<<<<<
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_breakBond_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 180; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_10 = __Pyx_PyList_Extend(__pyx_v_atomType, __pyx_t_2); if (unlikely(__pyx_t_10 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 180; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":179
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             atomType.extend(atom.breakBond)
 *         if len(atomType) == 0:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":181
 *         for atom in self.atomType:
 *             atomType.extend(atom.breakBond)
 *         if len(atomType) == 0:             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 */
  __pyx_t_9 = PyList_GET_SIZE(__pyx_v_atomType); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 181; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = ((__pyx_t_9 == 0) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":182
 *             atomType.extend(atom.breakBond)
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_6, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_self->atomType); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_4, NULL); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_8); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_8);
      __Pyx_GIVEREF(__pyx_t_8);
      __pyx_t_8 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 182; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":184
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))             # <<<<<<<<<<<<<<
 * 
 *     def __gainRadical(self, radical):
 */
  __pyx_t_1 = PySet_New(__pyx_v_atomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 184; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 184; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyList_Type))), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 184; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":170
 *         self.atomType = list(set(atomType))
 * 
 *     def __breakBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a BREAK_BOND action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_8);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__breakBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond[] = "\n        Update the atom group as a result of applying a BREAK_BOND action,\n        where `order` specifies the order of the breaking bond, and should be\n        'S' (since we only allow breaking of single bonds).\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__breakBond (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_order), (&PyString_Type), 1, "order", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject*)__pyx_v_order));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__breakBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___breakBond(__pyx_v_self, __pyx_v_order, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__breakBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":186
 *         self.atomType = list(set(atomType))
 * 
 *     def __gainRadical(self, radical):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a GAIN_RADICAL action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_radicalElectrons = NULL;
  PyObject *__pyx_v_electron = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  int __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__gainRadical", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_gainRadical); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 186; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_radical); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 186; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 186; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 186; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 186; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":191
 *         where `radical` specifies the number of radical electrons to add.
 *         """
 *         radicalElectrons = []             # <<<<<<<<<<<<<<
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 191; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_radicalElectrons = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":192
 *         """
 *         radicalElectrons = []
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_4); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomType, __pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomType, __pyx_n_s_incrementRadical); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = PyObject_Length(__pyx_t_4); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyBool_FromLong((__pyx_t_8 == 0)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_4))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":193
 *         radicalElectrons = []
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         for electron in self.radicalElectrons:
 *             radicalElectrons.append(electron + radical)
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_7, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_6, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":194
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             radicalElectrons.append(electron + radical)
 *         # Set the new radical electron counts
 */
  if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 194; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 194; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 194; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_electron, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":195
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:
 *             radicalElectrons.append(electron + radical)             # <<<<<<<<<<<<<<
 *         # Set the new radical electron counts
 *         self.radicalElectrons = radicalElectrons
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_radical); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 195; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Add(__pyx_v_electron, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 195; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_10 = __Pyx_PyList_Append(__pyx_v_radicalElectrons, __pyx_t_5); if (unlikely(__pyx_t_10 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 195; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;

    /* "rmgpy/molecule/group.py":194
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             radicalElectrons.append(electron + radical)
 *         # Set the new radical electron counts
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":197
 *             radicalElectrons.append(electron + radical)
 *         # Set the new radical electron counts
 *         self.radicalElectrons = radicalElectrons             # <<<<<<<<<<<<<<
 * 
 *     def __loseRadical(self, radical):
 */
  __Pyx_INCREF(__pyx_v_radicalElectrons);
  __Pyx_GIVEREF(__pyx_v_radicalElectrons);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = __pyx_v_radicalElectrons;

  /* "rmgpy/molecule/group.py":186
 *         self.atomType = list(set(atomType))
 * 
 *     def __gainRadical(self, radical):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a GAIN_RADICAL action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_radicalElectrons);
  __Pyx_XDECREF(__pyx_v_electron);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical[] = "\n        Update the atom group as a result of applying a GAIN_RADICAL action,\n        where `radical` specifies the number of radical electrons to add.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical) {
  short __pyx_v_radical;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__gainRadical (wrapper)", 0);
  assert(__pyx_arg_radical); {
    __pyx_v_radical = __Pyx_PyInt_As_short(__pyx_arg_radical); if (unlikely((__pyx_v_radical == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 186; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_radical));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__gainRadical", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainRadical(__pyx_v_self, __pyx_v_radical, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 186; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":199
 *         self.radicalElectrons = radicalElectrons
 * 
 *     def __loseRadical(self, radical):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a LOSE_RADICAL action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___loseRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_radicalElectrons = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_pairs = NULL;
  PyObject *__pyx_v_electron = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  PyObject *__pyx_t_10 = NULL;
  int __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__loseRadical", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_loseRadical); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 199; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_radical); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 199; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 199; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 199; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 199; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":204
 *         where `radical` specifies the number of radical electrons to remove.
 *         """
 *         radicalElectrons = []             # <<<<<<<<<<<<<<
 *         pairs = set()
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_radicalElectrons = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":205
 *         """
 *         radicalElectrons = []
 *         pairs = set()             # <<<<<<<<<<<<<<
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
  __pyx_t_1 = PySet_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 205; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_pairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":206
 *         radicalElectrons = []
 *         pairs = set()
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_4); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomType, __pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomType, __pyx_n_s_decrementRadical); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = PyObject_Length(__pyx_t_4); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyBool_FromLong((__pyx_t_8 == 0)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_4))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":207
 *         pairs = set()
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         for electron in self.radicalElectrons:
 *             electron = electron - radical
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_8, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_6, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":208
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             electron = electron - radical
 *             if electron < 0:
 */
  if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 208; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 208; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 208; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_electron, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":209
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:
 *             electron = electron - radical             # <<<<<<<<<<<<<<
 *             if electron < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set "{0}".'.format(self.radicalElectrons))
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_radical); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Subtract(__pyx_v_electron, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF_SET(__pyx_v_electron, __pyx_t_5);
    __pyx_t_5 = 0;

    /* "rmgpy/molecule/group.py":210
 *         for electron in self.radicalElectrons:
 *             electron = electron - radical
 *             if electron < 0:             # <<<<<<<<<<<<<<
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set "{0}".'.format(self.radicalElectrons))
 *             radicalElectrons.append(electron)
 */
    __pyx_t_5 = PyObject_RichCompare(__pyx_v_electron, __pyx_int_0, Py_LT); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    if (__pyx_t_9) {

      /* "rmgpy/molecule/group.py":211
 *             electron = electron - radical
 *             if electron < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set "{0}".'.format(self.radicalElectrons))             # <<<<<<<<<<<<<<
 *             radicalElectrons.append(electron)
 * 
 */
      __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_9, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __pyx_t_3 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
        __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
        if (likely(__pyx_t_3)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
          __Pyx_INCREF(__pyx_t_3);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_6, function);
        }
      }
      if (!__pyx_t_3) {
        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->radicalElectrons); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
      } else {
        __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        __Pyx_INCREF(__pyx_v_self->radicalElectrons);
        PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_v_self->radicalElectrons);
        __Pyx_GIVEREF(__pyx_v_self->radicalElectrons);
        __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_10, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      }
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __pyx_t_6 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
        __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
        if (likely(__pyx_t_6)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
          __Pyx_INCREF(__pyx_t_6);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_2, function);
        }
      }
      if (!__pyx_t_6) {
        __pyx_t_5 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_GOTREF(__pyx_t_5);
      } else {
        __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
        PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_t_4);
        __Pyx_GIVEREF(__pyx_t_4);
        __pyx_t_4 = 0;
        __pyx_t_5 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_10, NULL); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      }
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_Raise(__pyx_t_5, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "rmgpy/molecule/group.py":212
 *             if electron < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set "{0}".'.format(self.radicalElectrons))
 *             radicalElectrons.append(electron)             # <<<<<<<<<<<<<<
 * 
 *         # Set the new radical electron counts
 */
    __pyx_t_11 = __Pyx_PyList_Append(__pyx_v_radicalElectrons, __pyx_v_electron); if (unlikely(__pyx_t_11 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 212; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

    /* "rmgpy/molecule/group.py":208
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             electron = electron - radical
 *             if electron < 0:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":215
 * 
 *         # Set the new radical electron counts
 *         self.radicalElectrons = radicalElectrons             # <<<<<<<<<<<<<<
 * 
 *     def __gainPair(self, pair):
 */
  __Pyx_INCREF(__pyx_v_radicalElectrons);
  __Pyx_GIVEREF(__pyx_v_radicalElectrons);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = __pyx_v_radicalElectrons;

  /* "rmgpy/molecule/group.py":199
 *         self.radicalElectrons = radicalElectrons
 * 
 *     def __loseRadical(self, radical):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a LOSE_RADICAL action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_10);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__loseRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_radicalElectrons);
  __Pyx_XDECREF(__pyx_v_pairs);
  __Pyx_XDECREF(__pyx_v_electron);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical[] = "\n        Update the atom group as a result of applying a LOSE_RADICAL action,\n        where `radical` specifies the number of radical electrons to remove.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical) {
  short __pyx_v_radical;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__loseRadical (wrapper)", 0);
  assert(__pyx_arg_radical); {
    __pyx_v_radical = __Pyx_PyInt_As_short(__pyx_arg_radical); if (unlikely((__pyx_v_radical == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 199; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__loseRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_radical));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__loseRadical", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___loseRadical(__pyx_v_self, __pyx_v_radical, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 199; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__loseRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":217
 *         self.radicalElectrons = radicalElectrons
 * 
 *     def __gainPair(self, pair):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a GAIN_PAIR action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainPair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_lonePairs = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  PyObject *(*__pyx_t_10)(PyObject *);
  int __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__gainPair", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_gainPair); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":222
 *         where `pair` specifies the number of lone electron pairs to add.
 *         """
 *         lonePairs = []             # <<<<<<<<<<<<<<
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_lonePairs = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":223
 *         """
 *         lonePairs = []
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_4); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomType, __pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomType, __pyx_n_s_incrementLonePair); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = PyObject_Length(__pyx_t_4); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyBool_FromLong((__pyx_t_8 == 0)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_4))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":224
 *         lonePairs = []
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         for lonePairs in zip(self.lonePairs):
 *             lonePairs.append(lonePairs + pair)
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_10, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_6, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 224; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":225
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):             # <<<<<<<<<<<<<<
 *             lonePairs.append(lonePairs + pair)
 *         # Set the new lone electron pair count
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_self->lonePairs);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_self->lonePairs);
  __Pyx_GIVEREF(__pyx_v_self->lonePairs);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
    __pyx_t_10 = NULL;
  } else {
    __pyx_t_7 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_10 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_10)) {
      if (likely(PyList_CheckExact(__pyx_t_1))) {
        if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_7 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_10(__pyx_t_1);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_DECREF_SET(__pyx_v_lonePairs, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":226
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):
 *             lonePairs.append(lonePairs + pair)             # <<<<<<<<<<<<<<
 *         # Set the new lone electron pair count
 *         self.lonePairs = lonePairs
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 226; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Add(__pyx_v_lonePairs, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 226; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_11 = __Pyx_PyObject_Append(__pyx_v_lonePairs, __pyx_t_5); if (unlikely(__pyx_t_11 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 226; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;

    /* "rmgpy/molecule/group.py":225
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):             # <<<<<<<<<<<<<<
 *             lonePairs.append(lonePairs + pair)
 *         # Set the new lone electron pair count
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":228
 *             lonePairs.append(lonePairs + pair)
 *         # Set the new lone electron pair count
 *         self.lonePairs = lonePairs             # <<<<<<<<<<<<<<
 * 
 *     def __losePair(self, pair):
 */
  if (!(likely(PyList_CheckExact(__pyx_v_lonePairs))||((__pyx_v_lonePairs) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_lonePairs)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_lonePairs;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":217
 *         self.radicalElectrons = radicalElectrons
 * 
 *     def __gainPair(self, pair):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a GAIN_PAIR action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainPair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_lonePairs);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair[] = "\n        Update the atom group as a result of applying a GAIN_PAIR action,\n        where `pair` specifies the number of lone electron pairs to add.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair) {
  short __pyx_v_pair;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__gainPair (wrapper)", 0);
  assert(__pyx_arg_pair); {
    __pyx_v_pair = __Pyx_PyInt_As_short(__pyx_arg_pair); if (unlikely((__pyx_v_pair == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainPair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_pair));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__gainPair", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainPair(__pyx_v_self, __pyx_v_pair, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainPair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":230
 *         self.lonePairs = lonePairs
 * 
 *     def __losePair(self, pair):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a LOSE_PAIR action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___losePair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_lonePairs = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  PyObject *(*__pyx_t_10)(PyObject *);
  PyObject *__pyx_t_11 = NULL;
  int __pyx_t_12;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__losePair", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_losePair); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":235
 *         where `pair` specifies the number of lone electron pairs to remove.
 *         """
 *         lonePairs = []             # <<<<<<<<<<<<<<
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_lonePairs = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":236
 *         """
 *         lonePairs = []
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_4); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomType, __pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomType, __pyx_n_s_decrementLonePair); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = PyObject_Length(__pyx_t_4); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyBool_FromLong((__pyx_t_8 == 0)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_4))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":237
 *         lonePairs = []
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         for lonePairs in zip(self.lonePairs):
 *             if lonePairs - pair < 0:
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_11, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_6, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":238
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):             # <<<<<<<<<<<<<<
 *             if lonePairs - pair < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_self->lonePairs);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_self->lonePairs);
  __Pyx_GIVEREF(__pyx_v_self->lonePairs);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
    __pyx_t_10 = NULL;
  } else {
    __pyx_t_7 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_10 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_10)) {
      if (likely(PyList_CheckExact(__pyx_t_1))) {
        if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_7 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_10(__pyx_t_1);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 238; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_DECREF_SET(__pyx_v_lonePairs, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":239
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):
 *             if lonePairs - pair < 0:             # <<<<<<<<<<<<<<
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))
 *             lonePairs.append(lonePairs - pair)
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 239; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Subtract(__pyx_v_lonePairs, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 239; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyObject_RichCompare(__pyx_t_5, __pyx_int_0, Py_LT); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 239; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 239; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_9) {

      /* "rmgpy/molecule/group.py":240
 *         for lonePairs in zip(self.lonePairs):
 *             if lonePairs - pair < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))             # <<<<<<<<<<<<<<
 *             lonePairs.append(lonePairs - pair)
 *         # Set the new lone electron pair count
 */
      __pyx_t_5 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_12, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __pyx_t_3 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
        __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
        if (likely(__pyx_t_3)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
          __Pyx_INCREF(__pyx_t_3);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_6, function);
        }
      }
      if (!__pyx_t_3) {
        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->lonePairs); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
      } else {
        __pyx_t_11 = PyTuple_New(1+1); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        __Pyx_INCREF(__pyx_v_self->lonePairs);
        PyTuple_SET_ITEM(__pyx_t_11, 0+1, __pyx_v_self->lonePairs);
        __Pyx_GIVEREF(__pyx_v_self->lonePairs);
        __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_11, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      }
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __pyx_t_6 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_5))) {
        __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_5);
        if (likely(__pyx_t_6)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
          __Pyx_INCREF(__pyx_t_6);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_5, function);
        }
      }
      if (!__pyx_t_6) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_11 = PyTuple_New(1+1); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
        PyTuple_SET_ITEM(__pyx_t_11, 0+1, __pyx_t_4);
        __Pyx_GIVEREF(__pyx_t_4);
        __pyx_t_4 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_11, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      }
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "rmgpy/molecule/group.py":241
 *             if lonePairs - pair < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))
 *             lonePairs.append(lonePairs - pair)             # <<<<<<<<<<<<<<
 *         # Set the new lone electron pair count
 *         self.lonePairs = lonePairs
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Subtract(__pyx_v_lonePairs, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_12 = __Pyx_PyObject_Append(__pyx_v_lonePairs, __pyx_t_5); if (unlikely(__pyx_t_12 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;

    /* "rmgpy/molecule/group.py":238
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):             # <<<<<<<<<<<<<<
 *             if lonePairs - pair < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":243
 *             lonePairs.append(lonePairs - pair)
 *         # Set the new lone electron pair count
 *         self.lonePairs = lonePairs             # <<<<<<<<<<<<<<
 * 
 *     def applyAction(self, action):
 */
  if (!(likely(PyList_CheckExact(__pyx_v_lonePairs))||((__pyx_v_lonePairs) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_lonePairs)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 243; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_lonePairs;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":230
 *         self.lonePairs = lonePairs
 * 
 *     def __losePair(self, pair):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a LOSE_PAIR action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_11);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__losePair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_lonePairs);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_26__losePair[] = "\n        Update the atom group as a result of applying a LOSE_PAIR action,\n        where `pair` specifies the number of lone electron pairs to remove.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair) {
  short __pyx_v_pair;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__losePair (wrapper)", 0);
  assert(__pyx_arg_pair); {
    __pyx_v_pair = __Pyx_PyInt_As_short(__pyx_arg_pair); if (unlikely((__pyx_v_pair == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 230; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__losePair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_26__losePair(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_pair));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_26__losePair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__losePair", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___losePair(__pyx_v_self, __pyx_v_pair, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__losePair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":245
 *         self.lonePairs = lonePairs
 * 
 *     def applyAction(self, action):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying `action`, a tuple
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_action, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_act = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  short __pyx_t_7;
  PyObject *__pyx_t_8 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("applyAction", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_applyAction); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 245; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_action); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 245; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 245; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_action);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_action);
        __Pyx_GIVEREF(__pyx_v_action);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 245; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":252
 *         :ref:`here <reaction-recipe-actions>`.
 *         """
 *         act = action[0].upper()             # <<<<<<<<<<<<<<
 *         if act == 'CHANGE_BOND':
 *             self.__changeBond(action[2])
 */
  if (unlikely(__pyx_v_action == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_action, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 252; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_upper); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (__pyx_t_2) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_v_act = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":253
 *         """
 *         act = action[0].upper()
 *         if act == 'CHANGE_BOND':             # <<<<<<<<<<<<<<
 *             self.__changeBond(action[2])
 *         elif act == 'FORM_BOND':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_CHANGE_BOND, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 253; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":254
 *         act = action[0].upper()
 *         if act == 'CHANGE_BOND':
 *             self.__changeBond(action[2])             # <<<<<<<<<<<<<<
 *         elif act == 'FORM_BOND':
 *             self.__formBond(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 254; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 254; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 254; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___changeBond(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 254; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":255
 *         if act == 'CHANGE_BOND':
 *             self.__changeBond(action[2])
 *         elif act == 'FORM_BOND':             # <<<<<<<<<<<<<<
 *             self.__formBond(action[2])
 *         elif act == 'BREAK_BOND':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_FORM_BOND, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 255; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":256
 *             self.__changeBond(action[2])
 *         elif act == 'FORM_BOND':
 *             self.__formBond(action[2])             # <<<<<<<<<<<<<<
 *         elif act == 'BREAK_BOND':
 *             self.__breakBond(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    if (!(likely(PyString_CheckExact(__pyx_t_1))||((__pyx_t_1) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_1)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_3 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___formBond(__pyx_v_self, ((PyObject*)__pyx_t_1), 0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":257
 *         elif act == 'FORM_BOND':
 *             self.__formBond(action[2])
 *         elif act == 'BREAK_BOND':             # <<<<<<<<<<<<<<
 *             self.__breakBond(action[2])
 *         elif act == 'GAIN_RADICAL':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_BREAK_BOND, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":258
 *             self.__formBond(action[2])
 *         elif act == 'BREAK_BOND':
 *             self.__breakBond(action[2])             # <<<<<<<<<<<<<<
 *         elif act == 'GAIN_RADICAL':
 *             self.__gainRadical(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 258; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_3 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 258; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    if (!(likely(PyString_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 258; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___breakBond(__pyx_v_self, ((PyObject*)__pyx_t_3), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 258; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":259
 *         elif act == 'BREAK_BOND':
 *             self.__breakBond(action[2])
 *         elif act == 'GAIN_RADICAL':             # <<<<<<<<<<<<<<
 *             self.__gainRadical(action[2])
 *         elif act == 'LOSE_RADICAL':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_GAIN_RADICAL, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 259; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":260
 *             self.__breakBond(action[2])
 *         elif act == 'GAIN_RADICAL':
 *             self.__gainRadical(action[2])             # <<<<<<<<<<<<<<
 *         elif act == 'LOSE_RADICAL':
 *             self.__loseRadical(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 260; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 260; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 260; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___gainRadical(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 260; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":261
 *         elif act == 'GAIN_RADICAL':
 *             self.__gainRadical(action[2])
 *         elif act == 'LOSE_RADICAL':             # <<<<<<<<<<<<<<
 *             self.__loseRadical(action[2])
 *         elif action[0].upper() == 'GAIN_PAIR':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_LOSE_RADICAL, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 261; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":262
 *             self.__gainRadical(action[2])
 *         elif act == 'LOSE_RADICAL':
 *             self.__loseRadical(action[2])             # <<<<<<<<<<<<<<
 *         elif action[0].upper() == 'GAIN_PAIR':
 *             self.__gainPair(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 262; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 262; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 262; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___loseRadical(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 262; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":263
 *         elif act == 'LOSE_RADICAL':
 *             self.__loseRadical(action[2])
 *         elif action[0].upper() == 'GAIN_PAIR':             # <<<<<<<<<<<<<<
 *             self.__gainPair(action[2])
 *         elif action[0].upper() == 'LOSE_PAIR':
 */
  if (unlikely(__pyx_v_action == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_3 = __Pyx_GetItemInt_List(__pyx_v_action, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_upper); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_GAIN_PAIR, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":264
 *             self.__loseRadical(action[2])
 *         elif action[0].upper() == 'GAIN_PAIR':
 *             self.__gainPair(action[2])             # <<<<<<<<<<<<<<
 *         elif action[0].upper() == 'LOSE_PAIR':
 *             self.__losePair(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 264; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 264; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 264; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___gainPair(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 264; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":265
 *         elif action[0].upper() == 'GAIN_PAIR':
 *             self.__gainPair(action[2])
 *         elif action[0].upper() == 'LOSE_PAIR':             # <<<<<<<<<<<<<<
 *             self.__losePair(action[2])
 *         else:
 */
  if (unlikely(__pyx_v_action == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_action, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_upper); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (__pyx_t_2) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_LOSE_PAIR, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":266
 *             self.__gainPair(action[2])
 *         elif action[0].upper() == 'LOSE_PAIR':
 *             self.__losePair(action[2])             # <<<<<<<<<<<<<<
 *         else:
 *             raise ActionError('Unable to update GroupAtom: Invalid action {0}".'.format(action))
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 266; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 266; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 266; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___losePair(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 266; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":268
 *             self.__losePair(action[2])
 *         else:
 *             raise ActionError('Unable to update GroupAtom: Invalid action {0}".'.format(action))             # <<<<<<<<<<<<<<
 * 
 *     def equivalent(self, other):
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_Inval, __pyx_n_s_format); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_5);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_5, function);
      }
    }
    if (!__pyx_t_4) {
      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_v_action); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      __Pyx_INCREF(__pyx_v_action);
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_v_action);
      __Pyx_GIVEREF(__pyx_v_action);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_t_2);
      __Pyx_GIVEREF(__pyx_t_2);
      __pyx_t_2 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_8, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_L3:;

  /* "rmgpy/molecule/group.py":245
 *         self.lonePairs = lonePairs
 * 
 *     def applyAction(self, action):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying `action`, a tuple
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_8);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.applyAction", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_act);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_28applyAction[] = "\n        Update the atom group as a result of applying `action`, a tuple\n        containing the name of the reaction recipe action along with any\n        required parameters. The available actions can be found\n        :ref:`here <reaction-recipe-actions>`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("applyAction (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_action), (&PyList_Type), 1, "action", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 245; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_28applyAction(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject*)__pyx_v_action));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_28applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_action) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("applyAction", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_applyAction(__pyx_v_self, __pyx_v_action, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 245; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.applyAction", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":270
 *             raise ActionError('Unable to update GroupAtom: Invalid action {0}".'.format(action))
 * 
 *     def equivalent(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_group = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_atomType1 = 0;
  short __pyx_v_radical1;
  short __pyx_v_radical2;
  short __pyx_v_lp1;
  short __pyx_v_lp2;
  short __pyx_v_charge1;
  short __pyx_v_charge2;
  PyObject *__pyx_v_atomType2 = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  short __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("equivalent", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_equivalent); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":279
 *         """
 *         cython.declare(group=GroupAtom)
 *         if not isinstance(other, GroupAtom):             # <<<<<<<<<<<<<<
 *             # Let the equivalent method of other handle it
 *             # We expect self to be an Atom object, but can't test for it here
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":283
 *             # We expect self to be an Atom object, but can't test for it here
 *             # because that would create an import cycle
 *             return other.equivalent(self)             # <<<<<<<<<<<<<<
 *         group=other
 * 
 */
    __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_other->__pyx_vtab)->equivalent(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_self), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_r = __pyx_t_7;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":284
 *             # because that would create an import cycle
 *             return other.equivalent(self)
 *         group=other             # <<<<<<<<<<<<<<
 * 
 *         cython.declare(atomType1=AtomType, atomtype2=AtomType, radical1=cython.short, radical2=cython.short,
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 284; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_group = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":290
 *         # Compare two atom groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:             # <<<<<<<<<<<<<<
 *             for atomType2 in group.atomType:
 *                 if atomType1.equivalent(atomType2): break
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 290; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 290; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 290; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 290; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":291
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:
 *             for atomType2 in group.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    if (unlikely(__pyx_v_group->atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 291; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_group->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 291; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 291; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_atomType2, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":292
 *         for atomType1 in self.atomType:
 *             for atomType2 in group.atomType:
 *                 if atomType1.equivalent(atomType2): break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      if (!(likely(((__pyx_v_atomType2) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_atomType2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 292; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = (((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType1->__pyx_vtab)->equivalent(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType2), 0) != 0);
      if (__pyx_t_7) {
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":291
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:
 *             for atomType2 in group.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":294
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":291
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:
 *             for atomType2 in group.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":290
 *         # Compare two atom groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:             # <<<<<<<<<<<<<<
 *             for atomType2 in group.atomType:
 *                 if atomType1.equivalent(atomType2): break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":295
 *             else:
 *                 return False
 *         for atomType1 in group.atomType:             # <<<<<<<<<<<<<<
 *             for atomType2 in self.atomType:
 *                 if atomType1.equivalent(atomType2): break
 */
  if (unlikely(__pyx_v_group->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 295; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_group->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 295; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 295; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 295; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":296
 *                 return False
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    if (unlikely(__pyx_v_self->atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 296; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 296; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 296; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_atomType2, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":297
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:
 *                 if atomType1.equivalent(atomType2): break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      if (!(likely(((__pyx_v_atomType2) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_atomType2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 297; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = (((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType1->__pyx_vtab)->equivalent(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType2), 0) != 0);
      if (__pyx_t_7) {
        goto __pyx_L13_break;
      }

      /* "rmgpy/molecule/group.py":296
 *                 return False
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":299
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         for radical1 in self.radicalElectrons:
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":296
 *                 return False
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    __pyx_L13_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":295
 *             else:
 *                 return False
 *         for atomType1 in group.atomType:             # <<<<<<<<<<<<<<
 *             for atomType2 in self.atomType:
 *                 if atomType1.equivalent(atomType2): break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":301
 *                 return False
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         for radical1 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:
 */
  if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 301; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 301; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 301; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 301; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_radical1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":302
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         for radical1 in self.radicalElectrons:
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.             # <<<<<<<<<<<<<<
 *                 for radical2  in group.radicalElectrons:
 *                     if radical1 == radical2: break
 */
    __pyx_t_7 = (__pyx_v_group->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_group->radicalElectrons) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":303
 *         for radical1 in self.radicalElectrons:
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      if (unlikely(__pyx_v_group->radicalElectrons == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 303; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_group->radicalElectrons; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 303; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 303; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 303; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_radical2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":304
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:
 *                     if radical1 == radical2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_radical1 == __pyx_v_radical2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L20_break;
        }

        /* "rmgpy/molecule/group.py":303
 *         for radical1 in self.radicalElectrons:
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":306
 *                     if radical1 == radical2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":303
 *         for radical1 in self.radicalElectrons:
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      __pyx_L20_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L18;
    }
    __pyx_L18:;

    /* "rmgpy/molecule/group.py":301
 *                 return False
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         for radical1 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":307
 *                 else:
 *                     return False
 *         for radical1 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:
 */
  if (unlikely(__pyx_v_group->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 307; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_group->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 307; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 307; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 307; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_radical1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":308
 *                     return False
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                 for radical2 in self.radicalElectrons:
 *                     if radical1 == radical2: break
 */
    __pyx_t_7 = (__pyx_v_self->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_self->radicalElectrons) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":309
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 309; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 309; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 309; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 309; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_radical2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":310
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:
 *                     if radical1 == radical2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_radical1 == __pyx_v_radical2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L27_break;
        }

        /* "rmgpy/molecule/group.py":309
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":312
 *                     if radical1 == radical2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":309
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      __pyx_L27_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L25;
    }
    __pyx_L25:;

    /* "rmgpy/molecule/group.py":307
 *                 else:
 *                     return False
 *         for radical1 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":313
 *                 else:
 *                     return False
 *         for lp1 in self.lonePairs:             # <<<<<<<<<<<<<<
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:
 */
  if (unlikely(__pyx_v_self->lonePairs == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 313; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->lonePairs; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 313; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 313; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 313; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_lp1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":314
 *                     return False
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:             # <<<<<<<<<<<<<<
 *                 for lp2 in group.lonePairs:
 *                     if lp1 == lp2: break
 */
    __pyx_t_7 = (__pyx_v_group->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_group->lonePairs) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":315
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                     if lp1 == lp2: break
 *                 else:
 */
      if (unlikely(__pyx_v_group->lonePairs == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 315; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_group->lonePairs; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 315; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 315; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 315; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_lp2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":316
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:
 *                     if lp1 == lp2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_lp1 == __pyx_v_lp2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L34_break;
        }

        /* "rmgpy/molecule/group.py":315
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                     if lp1 == lp2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":318
 *                     if lp1 == lp2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         #Each charge in self must have an equivalent in other (and vice versa)
 *         for charge1 in self.charge:
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":315
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                     if lp1 == lp2: break
 *                 else:
 */
      __pyx_L34_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L32;
    }
    __pyx_L32:;

    /* "rmgpy/molecule/group.py":313
 *                 else:
 *                     return False
 *         for lp1 in self.lonePairs:             # <<<<<<<<<<<<<<
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":320
 *                     return False
 *         #Each charge in self must have an equivalent in other (and vice versa)
 *         for charge1 in self.charge:             # <<<<<<<<<<<<<<
 *             if group.charge:
 *                 for charge2 in group.charge:
 */
  if (unlikely(__pyx_v_self->charge == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->charge; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_charge1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":321
 *         #Each charge in self must have an equivalent in other (and vice versa)
 *         for charge1 in self.charge:
 *             if group.charge:             # <<<<<<<<<<<<<<
 *                 for charge2 in group.charge:
 *                     if charge1 == charge2: break
 */
    __pyx_t_7 = (__pyx_v_group->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_group->charge) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":322
 *         for charge1 in self.charge:
 *             if group.charge:
 *                 for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      if (unlikely(__pyx_v_group->charge == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 322; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_group->charge; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 322; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 322; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 322; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_charge2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":323
 *             if group.charge:
 *                 for charge2 in group.charge:
 *                     if charge1 == charge2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_charge1 == __pyx_v_charge2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L41_break;
        }

        /* "rmgpy/molecule/group.py":322
 *         for charge1 in self.charge:
 *             if group.charge:
 *                 for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":325
 *                     if charge1 == charge2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         for charge1 in group.charge:
 *             if self.charge:
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":322
 *         for charge1 in self.charge:
 *             if group.charge:
 *                 for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      __pyx_L41_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L39;
    }
    __pyx_L39:;

    /* "rmgpy/molecule/group.py":320
 *                     return False
 *         #Each charge in self must have an equivalent in other (and vice versa)
 *         for charge1 in self.charge:             # <<<<<<<<<<<<<<
 *             if group.charge:
 *                 for charge2 in group.charge:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":326
 *                 else:
 *                     return False
 *         for charge1 in group.charge:             # <<<<<<<<<<<<<<
 *             if self.charge:
 *                 for charge2 in self.charge:
 */
  if (unlikely(__pyx_v_group->charge == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 326; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_group->charge; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 326; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 326; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 326; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_charge1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":327
 *                     return False
 *         for charge1 in group.charge:
 *             if self.charge:             # <<<<<<<<<<<<<<
 *                 for charge2 in self.charge:
 *                     if charge1 == charge2: break
 */
    __pyx_t_7 = (__pyx_v_self->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_self->charge) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":328
 *         for charge1 in group.charge:
 *             if self.charge:
 *                 for charge2 in self.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      if (unlikely(__pyx_v_self->charge == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_self->charge; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_charge2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":329
 *             if self.charge:
 *                 for charge2 in self.charge:
 *                     if charge1 == charge2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_charge1 == __pyx_v_charge2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L48_break;
        }

        /* "rmgpy/molecule/group.py":328
 *         for charge1 in group.charge:
 *             if self.charge:
 *                 for charge2 in self.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":331
 *                     if charge1 == charge2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         # Otherwise the two atom groups are equivalent
 *         return True
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":328
 *         for charge1 in group.charge:
 *             if self.charge:
 *                 for charge2 in self.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      __pyx_L48_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L46;
    }
    __pyx_L46:;

    /* "rmgpy/molecule/group.py":326
 *                 else:
 *                     return False
 *         for charge1 in group.charge:             # <<<<<<<<<<<<<<
 *             if self.charge:
 *                 for charge2 in self.charge:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":333
 *                     return False
 *         # Otherwise the two atom groups are equivalent
 *         return True             # <<<<<<<<<<<<<<
 * 
 *     def isSpecificCaseOf(self, other):
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":270
 *             raise ActionError('Unable to update GroupAtom: Invalid action {0}".'.format(action))
 * 
 *     def equivalent(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.equivalent", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_group);
  __Pyx_XDECREF((PyObject *)__pyx_v_atomType1);
  __Pyx_XDECREF(__pyx_v_atomType2);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_30equivalent[] = "\n        Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,\n        where `other` can be either an :class:`Atom` or an :class:`GroupAtom`\n        object. When comparing two :class:`GroupAtom` objects, this function\n        respects wildcards, e.g. ``R!H`` is equivalent to ``C``.\n        \n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("equivalent (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Vertex, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_30equivalent(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_30equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("equivalent", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_equivalent(__pyx_v_self, __pyx_v_other, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.equivalent", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":335
 *         return True
 * 
 *     def isSpecificCaseOf(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is the same as `self` or is a more
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_group = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_atomType1 = 0;
  short __pyx_v_radical1;
  short __pyx_v_radical2;
  short __pyx_v_lp1;
  short __pyx_v_lp2;
  short __pyx_v_charge1;
  short __pyx_v_charge2;
  PyObject *__pyx_v_atomType2 = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  short __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSpecificCaseOf", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSpecificCaseOf); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":342
 *         """
 *         cython.declare(group=GroupAtom)
 *         if not isinstance(other, GroupAtom):             # <<<<<<<<<<<<<<
 *             # Let the isSpecificCaseOf method of other handle it
 *             # We expect self to be an Atom object, but can't test for it here
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":346
 *             # We expect self to be an Atom object, but can't test for it here
 *             # because that would create an import cycle
 *             return other.isSpecificCaseOf(self)             # <<<<<<<<<<<<<<
 *         group=other
 * 
 */
    __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_other->__pyx_vtab)->isSpecificCaseOf(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_self), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_r = __pyx_t_7;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":347
 *             # because that would create an import cycle
 *             return other.isSpecificCaseOf(self)
 *         group=other             # <<<<<<<<<<<<<<
 * 
 *         cython.declare(atomType1=AtomType, atomtype2=AtomType, radical1=cython.short, radical2=cython.short,
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 347; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_group = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":353
 *         # Compare two atom groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match             # <<<<<<<<<<<<<<
 *             for atomType2 in group.atomType: # can match any of these
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 353; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 353; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 353; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 353; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":354
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match
 *             for atomType2 in group.atomType: # can match any of these             # <<<<<<<<<<<<<<
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 *             else:
 */
    if (unlikely(__pyx_v_group->atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_group->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_atomType2, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":355
 *         for atomType1 in self.atomType: # all these must match
 *             for atomType2 in group.atomType: # can match any of these
 *                 if atomType1.isSpecificCaseOf(atomType2): break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      if (!(likely(((__pyx_v_atomType2) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_atomType2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 355; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = (((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType1->__pyx_vtab)->isSpecificCaseOf(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType2), 0) != 0);
      if (__pyx_t_7) {
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":354
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match
 *             for atomType2 in group.atomType: # can match any of these             # <<<<<<<<<<<<<<
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":357
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         if self.radicalElectrons:
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":354
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match
 *             for atomType2 in group.atomType: # can match any of these             # <<<<<<<<<<<<<<
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 *             else:
 */
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":353
 *         # Compare two atom groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match             # <<<<<<<<<<<<<<
 *             for atomType2 in group.atomType: # can match any of these
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":359
 *                 return False
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         if self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:
 */
  __pyx_t_7 = (__pyx_v_self->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_self->radicalElectrons) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":360
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         if self.radicalElectrons:
 *             for radical1 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:
 */
    if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 360; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
    for (;;) {
      if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 360; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 360; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 360; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_radical1 = __pyx_t_10;

      /* "rmgpy/molecule/group.py":361
 *         if self.radicalElectrons:
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     for radical2 in group.radicalElectrons:
 *                         if radical1 == radical2: break
 */
      __pyx_t_7 = (__pyx_v_group->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_group->radicalElectrons) != 0);
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":362
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                         if radical1 == radical2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->radicalElectrons == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 362; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->radicalElectrons; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
        for (;;) {
          if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 362; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 362; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 362; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_v_radical2 = __pyx_t_10;

          /* "rmgpy/molecule/group.py":363
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:
 *                         if radical1 == radical2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return False
 */
          __pyx_t_7 = ((__pyx_v_radical1 == __pyx_v_radical2) != 0);
          if (__pyx_t_7) {
            goto __pyx_L15_break;
          }

          /* "rmgpy/molecule/group.py":362
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                         if radical1 == radical2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":365
 *                         if radical1 == radical2: break
 *                     else:
 *                         return False             # <<<<<<<<<<<<<<
 *         else:
 *             if group.radicalElectrons: return False
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":362
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                         if radical1 == radical2: break
 *                     else:
 */
        __pyx_L15_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L13;
      }
      __pyx_L13:;

      /* "rmgpy/molecule/group.py":360
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         if self.radicalElectrons:
 *             for radical1 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L10;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":367
 *                         return False
 *         else:
 *             if group.radicalElectrons: return False             # <<<<<<<<<<<<<<
 *         if self.lonePairs:
 *             for lp1 in self.lonePairs:
 */
    __pyx_t_7 = (__pyx_v_group->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_group->radicalElectrons) != 0);
    if (__pyx_t_7) {
      __pyx_r = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L10:;

  /* "rmgpy/molecule/group.py":368
 *         else:
 *             if group.radicalElectrons: return False
 *         if self.lonePairs:             # <<<<<<<<<<<<<<
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:
 */
  __pyx_t_7 = (__pyx_v_self->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_self->lonePairs) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":369
 *             if group.radicalElectrons: return False
 *         if self.lonePairs:
 *             for lp1 in self.lonePairs:             # <<<<<<<<<<<<<<
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:
 */
    if (unlikely(__pyx_v_self->lonePairs == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 369; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->lonePairs; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
    for (;;) {
      if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 369; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 369; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 369; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_lp1 = __pyx_t_10;

      /* "rmgpy/molecule/group.py":370
 *         if self.lonePairs:
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:             # <<<<<<<<<<<<<<
 *                     for lp2 in group.lonePairs:
 *                         if lp1 == lp2: break
 */
      __pyx_t_7 = (__pyx_v_group->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_group->lonePairs) != 0);
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":371
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                         if lp1 == lp2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->lonePairs == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->lonePairs; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
        for (;;) {
          if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_v_lp2 = __pyx_t_10;

          /* "rmgpy/molecule/group.py":372
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:
 *                         if lp1 == lp2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return False
 */
          __pyx_t_7 = ((__pyx_v_lp1 == __pyx_v_lp2) != 0);
          if (__pyx_t_7) {
            goto __pyx_L24_break;
          }

          /* "rmgpy/molecule/group.py":371
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                         if lp1 == lp2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":374
 *                         if lp1 == lp2: break
 *                     else:
 *                         return False             # <<<<<<<<<<<<<<
 *         else:
 *             if group.lonePairs: return False
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":371
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                         if lp1 == lp2: break
 *                     else:
 */
        __pyx_L24_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L22;
      }
      __pyx_L22:;

      /* "rmgpy/molecule/group.py":369
 *             if group.radicalElectrons: return False
 *         if self.lonePairs:
 *             for lp1 in self.lonePairs:             # <<<<<<<<<<<<<<
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L19;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":376
 *                         return False
 *         else:
 *             if group.lonePairs: return False             # <<<<<<<<<<<<<<
 *         #Each charge in self must have an equivalent in other
 *         if self.charge:
 */
    __pyx_t_7 = (__pyx_v_group->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_group->lonePairs) != 0);
    if (__pyx_t_7) {
      __pyx_r = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L19:;

  /* "rmgpy/molecule/group.py":378
 *             if group.lonePairs: return False
 *         #Each charge in self must have an equivalent in other
 *         if self.charge:             # <<<<<<<<<<<<<<
 *             for charge1 in self.charge:
 *                 if group.charge:
 */
  __pyx_t_7 = (__pyx_v_self->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_self->charge) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":379
 *         #Each charge in self must have an equivalent in other
 *         if self.charge:
 *             for charge1 in self.charge:             # <<<<<<<<<<<<<<
 *                 if group.charge:
 *                     for charge2 in group.charge:
 */
    if (unlikely(__pyx_v_self->charge == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 379; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->charge; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
    for (;;) {
      if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 379; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 379; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 379; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_charge1 = __pyx_t_10;

      /* "rmgpy/molecule/group.py":380
 *         if self.charge:
 *             for charge1 in self.charge:
 *                 if group.charge:             # <<<<<<<<<<<<<<
 *                     for charge2 in group.charge:
 *                         if charge1 == charge2: break
 */
      __pyx_t_7 = (__pyx_v_group->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_group->charge) != 0);
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":381
 *             for charge1 in self.charge:
 *                 if group.charge:
 *                     for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                         if charge1 == charge2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->charge == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 381; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->charge; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
        for (;;) {
          if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 381; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 381; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 381; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_v_charge2 = __pyx_t_10;

          /* "rmgpy/molecule/group.py":382
 *                 if group.charge:
 *                     for charge2 in group.charge:
 *                         if charge1 == charge2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return False
 */
          __pyx_t_7 = ((__pyx_v_charge1 == __pyx_v_charge2) != 0);
          if (__pyx_t_7) {
            goto __pyx_L33_break;
          }

          /* "rmgpy/molecule/group.py":381
 *             for charge1 in self.charge:
 *                 if group.charge:
 *                     for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                         if charge1 == charge2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":384
 *                         if charge1 == charge2: break
 *                     else:
 *                         return False             # <<<<<<<<<<<<<<
 *         else:
 *             if group.charge: return False
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":381
 *             for charge1 in self.charge:
 *                 if group.charge:
 *                     for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                         if charge1 == charge2: break
 *                     else:
 */
        __pyx_L33_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L31;
      }
      __pyx_L31:;

      /* "rmgpy/molecule/group.py":379
 *         #Each charge in self must have an equivalent in other
 *         if self.charge:
 *             for charge1 in self.charge:             # <<<<<<<<<<<<<<
 *                 if group.charge:
 *                     for charge2 in group.charge:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L28;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":386
 *                         return False
 *         else:
 *             if group.charge: return False             # <<<<<<<<<<<<<<
 *         # Otherwise self is in fact a specific case of other
 *         return True
 */
    __pyx_t_7 = (__pyx_v_group->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_group->charge) != 0);
    if (__pyx_t_7) {
      __pyx_r = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L28:;

  /* "rmgpy/molecule/group.py":388
 *             if group.charge: return False
 *         # Otherwise self is in fact a specific case of other
 *         return True             # <<<<<<<<<<<<<<
 * 
 *     def isOxygen(self):
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":335
 *         return True
 * 
 *     def isSpecificCaseOf(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is the same as `self` or is a more
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.isSpecificCaseOf", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_group);
  __Pyx_XDECREF((PyObject *)__pyx_v_atomType1);
  __Pyx_XDECREF(__pyx_v_atomType2);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf[] = "\n        Returns ``True`` if `other` is the same as `self` or is a more\n        specific case of `self`. Returns ``False`` if some of `self` is not\n        included in `other` or they are mutually exclusive. \n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSpecificCaseOf (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Vertex, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSpecificCaseOf", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSpecificCaseOf(__pyx_v_self, __pyx_v_other, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.isSpecificCaseOf", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":390
 *         return True
 * 
 *     def isOxygen(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom represents an oxygen atom or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isOxygen(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_allOxygens = NULL;
  PyObject *__pyx_v_checkList = NULL;
  PyObject *__pyx_v_x = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isOxygen", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isOxygen); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":395
 *         not.
 *         """
 *         allOxygens = [atomTypes['O']] + atomTypes['O'].specific             # <<<<<<<<<<<<<<
 *         checkList=[x in allOxygens for x in self.atomType]
 * 
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 395; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_O); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 395; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 395; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_t_2);
  __Pyx_GIVEREF(__pyx_t_2);
  __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 395; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_O); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 395; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_specific); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 395; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = PyNumber_Add(__pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 395; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_allOxygens = __pyx_t_3;
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":396
 *         """
 *         allOxygens = [atomTypes['O']] + atomTypes['O'].specific
 *         checkList=[x in allOxygens for x in self.atomType]             # <<<<<<<<<<<<<<
 * 
 *         return all(checkList)
 */
  __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 396; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 396; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 396; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 396; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_x, __pyx_v_allOxygens, Py_EQ)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 396; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_t_5); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 396; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_3, (PyObject*)__pyx_t_1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 396; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_checkList = ((PyObject*)__pyx_t_3);
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":398
 *         checkList=[x in allOxygens for x in self.atomType]
 * 
 *         return all(checkList)             # <<<<<<<<<<<<<<
 * 
 *     def isSulfur(self):
 */
  __pyx_t_3 = PyTuple_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 398; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_v_checkList);
  PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_v_checkList);
  __Pyx_GIVEREF(__pyx_v_checkList);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_all, __pyx_t_3, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 398; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 398; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":390
 *         return True
 * 
 *     def isOxygen(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom represents an oxygen atom or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.GroupAtom.isOxygen", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_allOxygens);
  __Pyx_XDECREF(__pyx_v_checkList);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen[] = "\n        Return ``True`` if the atom represents an oxygen atom or ``False`` if\n        not.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isOxygen (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isOxygen", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isOxygen(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.isOxygen", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":400
 *         return all(checkList)
 * 
 *     def isSulfur(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom represents an sulfur atom or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSulfur(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_allSulfur = NULL;
  PyObject *__pyx_v_checkList = NULL;
  PyObject *__pyx_v_x = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSulfur", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSulfur); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 400; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 400; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 400; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 400; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":405
 *         not.
 *         """
 *         allSulfur = [atomTypes['S']] + atomTypes['S'].specific             # <<<<<<<<<<<<<<
 *         checkList=[x in allSulfur for x in self.atomType]
 * 
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 405; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_S); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 405; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 405; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_t_2);
  __Pyx_GIVEREF(__pyx_t_2);
  __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 405; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_S); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 405; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_specific); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 405; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = PyNumber_Add(__pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 405; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_allSulfur = __pyx_t_3;
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":406
 *         """
 *         allSulfur = [atomTypes['S']] + atomTypes['S'].specific
 *         checkList=[x in allSulfur for x in self.atomType]             # <<<<<<<<<<<<<<
 * 
 *         return all(checkList)
 */
  __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_x, __pyx_v_allSulfur, Py_EQ)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_t_5); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_3, (PyObject*)__pyx_t_1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_checkList = ((PyObject*)__pyx_t_3);
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":408
 *         checkList=[x in allSulfur for x in self.atomType]
 * 
 *         return all(checkList)             # <<<<<<<<<<<<<<
 * ################################################################################
 * 
 */
  __pyx_t_3 = PyTuple_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 408; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_v_checkList);
  PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_v_checkList);
  __Pyx_GIVEREF(__pyx_v_checkList);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_all, __pyx_t_3, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 408; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 408; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":400
 *         return all(checkList)
 * 
 *     def isSulfur(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom represents an sulfur atom or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.GroupAtom.isSulfur", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_allSulfur);
  __Pyx_XDECREF(__pyx_v_checkList);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur[] = "\n        Return ``True`` if the atom represents an sulfur atom or ``False`` if\n        not.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSulfur (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSulfur", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSulfur(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 400; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.isSulfur", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":34
 * cdef class GroupAtom(Vertex):
 * 
 *     cdef public list atomType             # <<<<<<<<<<<<<<
 *     cdef public list radicalElectrons
 *     cdef public list charge
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->atomType);
  __pyx_r = __pyx_v_self->atomType;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 34; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.atomType.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":35
 * 
 *     cdef public list atomType
 *     cdef public list radicalElectrons             # <<<<<<<<<<<<<<
 *     cdef public list charge
 *     cdef public str label
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->radicalElectrons);
  __pyx_r = __pyx_v_self->radicalElectrons;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 35; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.radicalElectrons.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":36
 *     cdef public list atomType
 *     cdef public list radicalElectrons
 *     cdef public list charge             # <<<<<<<<<<<<<<
 *     cdef public str label
 *     cdef public list lonePairs
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->charge);
  __pyx_r = __pyx_v_self->charge;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 36; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->charge);
  __Pyx_DECREF(__pyx_v_self->charge);
  __pyx_v_self->charge = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.charge.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->charge);
  __Pyx_DECREF(__pyx_v_self->charge);
  __pyx_v_self->charge = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":37
 *     cdef public list radicalElectrons
 *     cdef public list charge
 *     cdef public str label             # <<<<<<<<<<<<<<
 *     cdef public list lonePairs
 * 
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->label);
  __pyx_r = __pyx_v_self->label;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyString_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 37; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->label);
  __Pyx_DECREF(__pyx_v_self->label);
  __pyx_v_self->label = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.label.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->label);
  __Pyx_DECREF(__pyx_v_self->label);
  __pyx_v_self->label = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":38
 *     cdef public list charge
 *     cdef public str label
 *     cdef public list lonePairs             # <<<<<<<<<<<<<<
 * 
 *     cpdef Vertex copy(self)
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->lonePairs);
  __pyx_r = __pyx_v_self->lonePairs;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 38; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.lonePairs.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":427
 *     """
 * 
 *     def __init__(self, atom1, atom2, order=None):             # <<<<<<<<<<<<<<
 *         Edge.__init__(self, atom1, atom2)
 *         self.order = order or []
 */

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_atom1 = 0;
  PyObject *__pyx_v_atom2 = 0;
  PyObject *__pyx_v_order = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__init__ (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atom1,&__pyx_n_s_atom2,&__pyx_n_s_order,0};
    PyObject* values[3] = {0,0,0};
    values[2] = ((PyObject *)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom1)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom2)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("__init__", 0, 2, 3, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 427; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
        case  2:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_order);
          if (value) { values[2] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__init__") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 427; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_atom1 = values[0];
    __pyx_v_atom2 = values[1];
    __pyx_v_order = values[2];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("__init__", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 427; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return -1;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond___init__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), __pyx_v_atom1, __pyx_v_atom2, __pyx_v_order);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond___init__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_atom1, PyObject *__pyx_v_atom2, PyObject *__pyx_v_order) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  Py_ssize_t __pyx_t_4;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__init__", 0);

  /* "rmgpy/molecule/group.py":428
 * 
 *     def __init__(self, atom1, atom2, order=None):
 *         Edge.__init__(self, atom1, atom2)             # <<<<<<<<<<<<<<
 *         self.order = order or []
 * 
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5graph_Edge)), __pyx_n_s_init); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  __pyx_t_4 = 0;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
      __pyx_t_4 = 1;
    }
  }
  __pyx_t_5 = PyTuple_New(3+__pyx_t_4); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  if (__pyx_t_3) {
    PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
  }
  __Pyx_INCREF(((PyObject *)__pyx_v_self));
  PyTuple_SET_ITEM(__pyx_t_5, 0+__pyx_t_4, ((PyObject *)__pyx_v_self));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
  __Pyx_INCREF(__pyx_v_atom1);
  PyTuple_SET_ITEM(__pyx_t_5, 1+__pyx_t_4, __pyx_v_atom1);
  __Pyx_GIVEREF(__pyx_v_atom1);
  __Pyx_INCREF(__pyx_v_atom2);
  PyTuple_SET_ITEM(__pyx_t_5, 2+__pyx_t_4, __pyx_v_atom2);
  __Pyx_GIVEREF(__pyx_v_atom2);
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":429
 *     def __init__(self, atom1, atom2, order=None):
 *         Edge.__init__(self, atom1, atom2)
 *         self.order = order or []             # <<<<<<<<<<<<<<
 * 
 *     def __str__(self):
 */
  __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_v_order); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 429; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_6) {
    goto __pyx_L4_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_order))||((__pyx_v_order) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_order)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 429; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_order);
    __pyx_t_1 = __pyx_v_order;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 429; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_1 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L3_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->order);
  __Pyx_DECREF(__pyx_v_self->order);
  __pyx_v_self->order = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":427
 *     """
 * 
 *     def __init__(self, atom1, atom2, order=None):             # <<<<<<<<<<<<<<
 *         Edge.__init__(self, atom1, atom2)
 *         self.order = order or []
 */

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":431
 *         self.order = order or []
 * 
 *     def __str__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a human-readable string representation of the object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_3__str__(PyObject *__pyx_v_self); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_2__str__[] = "\n        Return a human-readable string representation of the object.\n        ";
#if CYTHON_COMPILING_IN_CPYTHON
struct wrapperbase __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_2__str__;
#endif
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_3__str__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__str__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_2__str__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_2__str__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__str__", 0);

  /* "rmgpy/molecule/group.py":435
 *         Return a human-readable string representation of the object.
 *         """
 *         return str(self.order)             # <<<<<<<<<<<<<<
 * 
 *     def __repr__(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 435; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_self->order);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_self->order);
  __Pyx_GIVEREF(__pyx_v_self->order);
  __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyString_Type))), __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 435; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":431
 *         self.order = order or []
 * 
 *     def __str__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a human-readable string representation of the object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__str__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":437
 *         return str(self.order)
 * 
 *     def __repr__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a representation that can be used to reconstruct the object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5__repr__(PyObject *__pyx_v_self); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_4__repr__[] = "\n        Return a representation that can be used to reconstruct the object.\n        ";
#if CYTHON_COMPILING_IN_CPYTHON
struct wrapperbase __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_4__repr__;
#endif
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5__repr__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__repr__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_4__repr__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_4__repr__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__repr__", 0);

  /* "rmgpy/molecule/group.py":441
 *         Return a representation that can be used to reconstruct the object.
 *         """
 *         return "<GroupBond {0!r}>".format(self.order)             # <<<<<<<<<<<<<<
 * 
 *     def __reduce__(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_GroupBond_0_r, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 441; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_v_self->order); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 441; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 441; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(__pyx_v_self->order);
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_v_self->order);
    __Pyx_GIVEREF(__pyx_v_self->order);
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 441; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":437
 *         return str(self.order)
 * 
 *     def __repr__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a representation that can be used to reconstruct the object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__repr__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":443
 *         return "<GroupBond {0!r}>".format(self.order)
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_7__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_6__reduce__[] = "\n        A helper function used when pickling an object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_7__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__reduce__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_6__reduce__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_6__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__reduce__", 0);

  /* "rmgpy/molecule/group.py":447
 *         A helper function used when pickling an object.
 *         """
 *         return (GroupBond, (self.vertex1, self.vertex2, self.order))             # <<<<<<<<<<<<<<
 * 
 *     def copy(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 447; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  PyTuple_SET_ITEM(__pyx_t_1, 0, ((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  __Pyx_INCREF(((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  PyTuple_SET_ITEM(__pyx_t_1, 1, ((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  __Pyx_INCREF(__pyx_v_self->order);
  PyTuple_SET_ITEM(__pyx_t_1, 2, __pyx_v_self->order);
  __Pyx_GIVEREF(__pyx_v_self->order);
  __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 447; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)));
  PyTuple_SET_ITEM(__pyx_t_2, 0, ((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)));
  __Pyx_GIVEREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)));
  PyTuple_SET_ITEM(__pyx_t_2, 1, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":443
 *         return "<GroupBond {0!r}>".format(self.order)
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__reduce__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":449
 *         return (GroupBond, (self.vertex1, self.vertex2, self.order))
 * 
 *     def copy(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a deep copy of the :class:`GroupBond` object. Modifying the
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_copy); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 449; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 449; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 449; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5graph_Edge))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 449; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":454
 *         attributes of the copy will not affect the original.
 *         """
 *         return GroupBond(self.vertex1, self.vertex2, self.order[:])             # <<<<<<<<<<<<<<
 * 
 *     def isSingle(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_PyList_GetSlice(__pyx_v_self->order, 0, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  PyTuple_SET_ITEM(__pyx_t_2, 0, ((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  __Pyx_INCREF(((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  PyTuple_SET_ITEM(__pyx_t_2, 1, ((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  PyTuple_SET_ITEM(__pyx_t_2, 2, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":449
 *         return (GroupBond, (self.vertex1, self.vertex2, self.order))
 * 
 *     def copy(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a deep copy of the :class:`GroupBond` object. Modifying the
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_8copy[] = "\n        Return a deep copy of the :class:`GroupBond` object. Modifying the\n        attributes of the copy will not affect the original.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("copy (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_8copy(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_8copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_9GroupBond_copy(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 449; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":456
 *         return GroupBond(self.vertex1, self.vertex2, self.order[:])
 * 
 *     def isSingle(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a single bond or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSingle(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSingle", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSingle); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 456; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 456; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 456; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 456; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":461
 *         not. Bonds with any wildcards will return  ``False``.
 *         """
 *         return self.order[0] == 'S' and len(self.order) == 1             # <<<<<<<<<<<<<<
 * 
 *     def isDouble(self):
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 461; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 461; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_S, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 461; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {
    goto __pyx_L4_next_and;
  } else {
    __pyx_t_5 = __pyx_t_6;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_and:;
  __pyx_t_1 = __pyx_v_self->order;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 461; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 461; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_6 = ((__pyx_t_7 == 1) != 0);
  __pyx_t_5 = __pyx_t_6;
  __pyx_L3_bool_binop_done:;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":456
 *         return GroupBond(self.vertex1, self.vertex2, self.order[:])
 * 
 *     def isSingle(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a single bond or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isSingle", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_10isSingle[] = "\n        Return ``True`` if the bond represents a single bond or ``False`` if\n        not. Bonds with any wildcards will return  ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSingle (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_10isSingle(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_10isSingle(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSingle", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSingle(__pyx_v_self, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 456; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 456; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isSingle", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":463
 *         return self.order[0] == 'S' and len(self.order) == 1
 * 
 *     def isDouble(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a double bond or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isDouble(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isDouble", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isDouble); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 463; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 463; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 463; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 463; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":468
 *         not. Bonds with any wildcards will return  ``False``.
 *         """
 *         return self.order[0] == 'D' and len(self.order) == 1             # <<<<<<<<<<<<<<
 * 
 *     def isTriple(self):
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_D, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {
    goto __pyx_L4_next_and;
  } else {
    __pyx_t_5 = __pyx_t_6;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_and:;
  __pyx_t_1 = __pyx_v_self->order;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_6 = ((__pyx_t_7 == 1) != 0);
  __pyx_t_5 = __pyx_t_6;
  __pyx_L3_bool_binop_done:;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":463
 *         return self.order[0] == 'S' and len(self.order) == 1
 * 
 *     def isDouble(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a double bond or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isDouble", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_12isDouble[] = "\n        Return ``True`` if the bond represents a double bond or ``False`` if\n        not. Bonds with any wildcards will return  ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isDouble (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_12isDouble(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_12isDouble(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isDouble", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isDouble(__pyx_v_self, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 463; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 463; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isDouble", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":470
 *         return self.order[0] == 'D' and len(self.order) == 1
 * 
 *     def isTriple(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a triple bond or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isTriple(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isTriple", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isTriple); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 470; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 470; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 470; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 470; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":475
 *         not. Bonds with any wildcards will return  ``False``.
 *         """
 *         return self.order[0] == 'T' and len(self.order) == 1             # <<<<<<<<<<<<<<
 * 
 *     def isBenzene(self):
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 475; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 475; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_T, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 475; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {
    goto __pyx_L4_next_and;
  } else {
    __pyx_t_5 = __pyx_t_6;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_and:;
  __pyx_t_1 = __pyx_v_self->order;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 475; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 475; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_6 = ((__pyx_t_7 == 1) != 0);
  __pyx_t_5 = __pyx_t_6;
  __pyx_L3_bool_binop_done:;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":470
 *         return self.order[0] == 'D' and len(self.order) == 1
 * 
 *     def isTriple(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a triple bond or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isTriple", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_14isTriple[] = "\n        Return ``True`` if the bond represents a triple bond or ``False`` if\n        not. Bonds with any wildcards will return  ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isTriple (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_14isTriple(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_14isTriple(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isTriple", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isTriple(__pyx_v_self, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 470; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 470; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isTriple", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":477
 *         return self.order[0] == 'T' and len(self.order) == 1
 * 
 *     def isBenzene(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a benzene bond or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isBenzene", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 477; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 477; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 477; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 477; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":482
 *         not. Bonds with any wildcards will return  ``False``.
 *         """
 *         return self.order[0] == 'B' and len(self.order) == 1             # <<<<<<<<<<<<<<
 * 
 *     def __changeBond(self, order):
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_B, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {
    goto __pyx_L4_next_and;
  } else {
    __pyx_t_5 = __pyx_t_6;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_and:;
  __pyx_t_1 = __pyx_v_self->order;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_6 = ((__pyx_t_7 == 1) != 0);
  __pyx_t_5 = __pyx_t_6;
  __pyx_L3_bool_binop_done:;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":477
 *         return self.order[0] == 'T' and len(self.order) == 1
 * 
 *     def isBenzene(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a benzene bond or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isBenzene", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_16isBenzene[] = "\n        Return ``True`` if the bond represents a benzene bond or ``False`` if\n        not. Bonds with any wildcards will return  ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isBenzene (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_16isBenzene(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_16isBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isBenzene", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isBenzene(__pyx_v_self, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 477; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 477; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isBenzene", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":484
 *         return self.order[0] == 'B' and len(self.order) == 1
 * 
 *     def __changeBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the bond group as a result of applying a CHANGE_BOND action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond___changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, short __pyx_v_order, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_newOrder = NULL;
  PyObject *__pyx_v_bond = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  int __pyx_t_8;
  int __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  PyObject *__pyx_t_11 = NULL;
  PyObject *__pyx_t_12 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__changeBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_changeBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_order); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":490
 *         in bond order, and should be 1 or -1.
 *         """
 *         newOrder = []             # <<<<<<<<<<<<<<
 *         for bond in self.order:
 *             if order == 1:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 490; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_newOrder = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":491
 *         """
 *         newOrder = []
 *         for bond in self.order:             # <<<<<<<<<<<<<<
 *             if order == 1:
 *                 if bond == 'S':         newOrder.append('D')
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 491; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->order; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 491; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 491; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_bond, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":497
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 *             elif order == -1:             # <<<<<<<<<<<<<<
 *                 if bond == 'D':         newOrder.append('S')
 *                 elif bond == 'T':       newOrder.append('D')
 */
    switch (__pyx_v_order) {

      /* "rmgpy/molecule/group.py":492
 *         newOrder = []
 *         for bond in self.order:
 *             if order == 1:             # <<<<<<<<<<<<<<
 *                 if bond == 'S':         newOrder.append('D')
 *                 elif bond == 'D':       newOrder.append('T')
 */
      case 1:

      /* "rmgpy/molecule/group.py":493
 *         for bond in self.order:
 *             if order == 1:
 *                 if bond == 'S':         newOrder.append('D')             # <<<<<<<<<<<<<<
 *                 elif bond == 'D':       newOrder.append('T')
 *                 else:
 */
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_v_bond, __pyx_n_s_S, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_8) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_newOrder, __pyx_n_s_D); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L5;
      }

      /* "rmgpy/molecule/group.py":494
 *             if order == 1:
 *                 if bond == 'S':         newOrder.append('D')
 *                 elif bond == 'D':       newOrder.append('T')             # <<<<<<<<<<<<<<
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 */
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_v_bond, __pyx_n_s_D, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 494; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_8) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_newOrder, __pyx_n_s_T); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 494; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L5;
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":496
 *                 elif bond == 'D':       newOrder.append('T')
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))             # <<<<<<<<<<<<<<
 *             elif order == -1:
 *                 if bond == 'D':         newOrder.append('S')
 */
        __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupBond_due_t, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_5 = NULL;
        __pyx_t_10 = 0;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
          __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
          if (likely(__pyx_t_5)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
            __Pyx_INCREF(__pyx_t_5);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_3, function);
            __pyx_t_10 = 1;
          }
        }
        __pyx_t_11 = PyTuple_New(2+__pyx_t_10); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        if (__pyx_t_5) {
          PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        }
        __Pyx_INCREF(__pyx_v_bond);
        PyTuple_SET_ITEM(__pyx_t_11, 0+__pyx_t_10, __pyx_v_bond);
        __Pyx_GIVEREF(__pyx_v_bond);
        __Pyx_INCREF(__pyx_v_self->order);
        PyTuple_SET_ITEM(__pyx_t_11, 1+__pyx_t_10, __pyx_v_self->order);
        __Pyx_GIVEREF(__pyx_v_self->order);
        __pyx_t_6 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_11, NULL); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_3 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
          __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_4);
          if (likely(__pyx_t_3)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_3);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_4, function);
          }
        }
        if (!__pyx_t_3) {
          __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_6); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
          __Pyx_GOTREF(__pyx_t_2);
        } else {
          __pyx_t_11 = PyTuple_New(1+1); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_11);
          PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
          PyTuple_SET_ITEM(__pyx_t_11, 0+1, __pyx_t_6);
          __Pyx_GIVEREF(__pyx_t_6);
          __pyx_t_6 = 0;
          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_11, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
        }
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_Raise(__pyx_t_2, 0, 0, 0);
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_L5:;
      break;

      /* "rmgpy/molecule/group.py":497
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 *             elif order == -1:             # <<<<<<<<<<<<<<
 *                 if bond == 'D':         newOrder.append('S')
 *                 elif bond == 'T':       newOrder.append('D')
 */
      case -1:

      /* "rmgpy/molecule/group.py":498
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 *             elif order == -1:
 *                 if bond == 'D':         newOrder.append('S')             # <<<<<<<<<<<<<<
 *                 elif bond == 'T':       newOrder.append('D')
 *                 else:
 */
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_v_bond, __pyx_n_s_D, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 498; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_8) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_newOrder, __pyx_n_s_S); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 498; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L6;
      }

      /* "rmgpy/molecule/group.py":499
 *             elif order == -1:
 *                 if bond == 'D':         newOrder.append('S')
 *                 elif bond == 'T':       newOrder.append('D')             # <<<<<<<<<<<<<<
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 */
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_v_bond, __pyx_n_s_T, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 499; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_8) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_newOrder, __pyx_n_s_D); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 499; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L6;
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":501
 *                 elif bond == 'T':       newOrder.append('D')
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))             # <<<<<<<<<<<<<<
 *             else:
 *                 raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 */
        __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupBond_due_t, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        __pyx_t_3 = NULL;
        __pyx_t_10 = 0;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
          __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
          if (likely(__pyx_t_3)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
            __Pyx_INCREF(__pyx_t_3);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_6, function);
            __pyx_t_10 = 1;
          }
        }
        __pyx_t_5 = PyTuple_New(2+__pyx_t_10); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        if (__pyx_t_3) {
          PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        }
        __Pyx_INCREF(__pyx_v_bond);
        PyTuple_SET_ITEM(__pyx_t_5, 0+__pyx_t_10, __pyx_v_bond);
        __Pyx_GIVEREF(__pyx_v_bond);
        __Pyx_INCREF(__pyx_v_self->order);
        PyTuple_SET_ITEM(__pyx_t_5, 1+__pyx_t_10, __pyx_v_self->order);
        __Pyx_GIVEREF(__pyx_v_self->order);
        __pyx_t_11 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
        __pyx_t_6 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
          __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_4);
          if (likely(__pyx_t_6)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_6);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_4, function);
          }
        }
        if (!__pyx_t_6) {
          __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_11); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
          __Pyx_GOTREF(__pyx_t_2);
        } else {
          __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_5);
          PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
          PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_11);
          __Pyx_GIVEREF(__pyx_t_11);
          __pyx_t_11 = 0;
          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        }
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_Raise(__pyx_t_2, 0, 0, 0);
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_L6:;
      break;
      default:

      /* "rmgpy/molecule/group.py":503
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 *             else:
 *                 raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid order "{0}".'.format(order))             # <<<<<<<<<<<<<<
 *         # Set the new bond orders, removing any duplicates
 *         self.order = list(set(newOrder))
 */
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_11 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupBond_due_t_2, __pyx_n_s_format); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_6 = __Pyx_PyInt_From_short(__pyx_v_order); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __pyx_t_3 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_11))) {
        __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_11);
        if (likely(__pyx_t_3)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_11);
          __Pyx_INCREF(__pyx_t_3);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_11, function);
        }
      }
      if (!__pyx_t_3) {
        __pyx_t_5 = __Pyx_PyObject_CallOneArg(__pyx_t_11, __pyx_t_6); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
        __Pyx_GOTREF(__pyx_t_5);
      } else {
        __pyx_t_12 = PyTuple_New(1+1); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_12);
        PyTuple_SET_ITEM(__pyx_t_12, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        PyTuple_SET_ITEM(__pyx_t_12, 0+1, __pyx_t_6);
        __Pyx_GIVEREF(__pyx_t_6);
        __pyx_t_6 = 0;
        __pyx_t_5 = __Pyx_PyObject_Call(__pyx_t_11, __pyx_t_12, NULL); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      }
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __pyx_t_11 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_11 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_11)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_11);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_11) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_5); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_12 = PyTuple_New(1+1); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_12);
        PyTuple_SET_ITEM(__pyx_t_12, 0, __pyx_t_11); __Pyx_GIVEREF(__pyx_t_11); __pyx_t_11 = NULL;
        PyTuple_SET_ITEM(__pyx_t_12, 0+1, __pyx_t_5);
        __Pyx_GIVEREF(__pyx_t_5);
        __pyx_t_5 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_12, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      break;
    }

    /* "rmgpy/molecule/group.py":491
 *         """
 *         newOrder = []
 *         for bond in self.order:             # <<<<<<<<<<<<<<
 *             if order == 1:
 *                 if bond == 'S':         newOrder.append('D')
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":505
 *                 raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 *         # Set the new bond orders, removing any duplicates
 *         self.order = list(set(newOrder))             # <<<<<<<<<<<<<<
 * 
 *     def applyAction(self, action):
 */
  __pyx_t_1 = PySet_New(__pyx_v_newOrder); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 505; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 505; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyList_Type))), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 505; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->order);
  __Pyx_DECREF(__pyx_v_self->order);
  __pyx_v_self->order = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":484
 *         return self.order[0] == 'B' and len(self.order) == 1
 * 
 *     def __changeBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the bond group as a result of applying a CHANGE_BOND action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_11);
  __Pyx_XDECREF(__pyx_t_12);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_newOrder);
  __Pyx_XDECREF(__pyx_v_bond);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_18__changeBond[] = "\n        Update the bond group as a result of applying a CHANGE_BOND action,\n        where `order` specifies whether the bond is incremented or decremented\n        in bond order, and should be 1 or -1.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order) {
  short __pyx_v_order;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__changeBond (wrapper)", 0);
  assert(__pyx_arg_order); {
    __pyx_v_order = __Pyx_PyInt_As_short(__pyx_arg_order); if (unlikely((__pyx_v_order == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_18__changeBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((short)__pyx_v_order));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_18__changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, short __pyx_v_order) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__changeBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond___changeBond(__pyx_v_self, __pyx_v_order, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":507
 *         self.order = list(set(newOrder))
 * 
 *     def applyAction(self, action):             # <<<<<<<<<<<<<<
 *         """
 *         Update the bond group as a result of applying `action`, a tuple
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_action, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  short __pyx_t_7;
  PyObject *__pyx_t_8 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("applyAction", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_applyAction); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 507; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_action); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 507; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 507; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_action);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_action);
        __Pyx_GIVEREF(__pyx_v_action);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 507; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":514
 *         :ref:`here <reaction-recipe-actions>`.
 *         """
 *         if action[0].upper() == 'CHANGE_BOND':             # <<<<<<<<<<<<<<
 *             self.__changeBond(action[2])
 *         else:
 */
  if (unlikely(__pyx_v_action == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_action, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_upper); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (__pyx_t_2) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_CHANGE_BOND, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":515
 *         """
 *         if action[0].upper() == 'CHANGE_BOND':
 *             self.__changeBond(action[2])             # <<<<<<<<<<<<<<
 *         else:
 *             raise ActionError('Unable to update GroupBond: Invalid action {0}".'.format(action))
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 515; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___changeBond(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":517
 *             self.__changeBond(action[2])
 *         else:
 *             raise ActionError('Unable to update GroupBond: Invalid action {0}".'.format(action))             # <<<<<<<<<<<<<<
 * 
 *     def equivalent(self, other):
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupBond_Inval, __pyx_n_s_format); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_5);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_5, function);
      }
    }
    if (!__pyx_t_4) {
      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_v_action); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      __Pyx_INCREF(__pyx_v_action);
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_v_action);
      __Pyx_GIVEREF(__pyx_v_action);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_t_2);
      __Pyx_GIVEREF(__pyx_t_2);
      __pyx_t_2 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_8, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 517; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_L3:;

  /* "rmgpy/molecule/group.py":507
 *         self.order = list(set(newOrder))
 * 
 *     def applyAction(self, action):             # <<<<<<<<<<<<<<
 *         """
 *         Update the bond group as a result of applying `action`, a tuple
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_8);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.applyAction", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_20applyAction[] = "\n        Update the bond group as a result of applying `action`, a tuple\n        containing the name of the reaction recipe action along with any\n        required parameters. The available actions can be found\n        :ref:`here <reaction-recipe-actions>`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("applyAction (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_action), (&PyList_Type), 1, "action", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 507; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_20applyAction(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((PyObject*)__pyx_v_action));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_20applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_action) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("applyAction", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_applyAction(__pyx_v_self, __pyx_v_action, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 507; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.applyAction", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":519
 *             raise ActionError('Unable to update GroupBond: Invalid action {0}".'.format(action))
 * 
 *     def equivalent(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_gb = 0;
  PyObject *__pyx_v_order1 = 0;
  PyObject *__pyx_v_order2 = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("equivalent", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_equivalent); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":526
 *         """
 *         cython.declare(gb=GroupBond)
 *         if not isinstance(other, GroupBond):             # <<<<<<<<<<<<<<
 *             # Let the equivalent method of other handle it
 *             # We expect self to be a Bond object, but can't test for it here
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":530
 *             # We expect self to be a Bond object, but can't test for it here
 *             # because that would create an import cycle
 *             return other.equivalent(self)             # <<<<<<<<<<<<<<
 *         gb = other
 * 
 */
    __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge *)__pyx_v_other->__pyx_vtab)->equivalent(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_self), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 530; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_r = __pyx_t_7;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":531
 *             # because that would create an import cycle
 *             return other.equivalent(self)
 *         gb = other             # <<<<<<<<<<<<<<
 * 
 *         cython.declare(order1=str, order2=str)
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_GroupBond))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 531; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_gb = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":536
 *         # Compare two bond groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:             # <<<<<<<<<<<<<<
 *             for order2 in gb.order:
 *                 if order1 == order2: break
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 536; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->order; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 536; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 536; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(PyString_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 536; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_order1, ((PyObject*)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":537
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:
 *             for order2 in gb.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    if (unlikely(__pyx_v_gb->order == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 537; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_gb->order; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 537; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 537; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      if (!(likely(PyString_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 537; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_XDECREF_SET(__pyx_v_order2, ((PyObject*)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":538
 *         for order1 in self.order:
 *             for order2 in gb.order:
 *                 if order1 == order2: break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_v_order1, __pyx_v_order2, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 538; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_6 = (__pyx_t_7 != 0);
      if (__pyx_t_6) {
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":537
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:
 *             for order2 in gb.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":540
 *                 if order1 == order2: break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         for order1 in gb.order:
 *             for order2 in self.order:
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":537
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:
 *             for order2 in gb.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":536
 *         # Compare two bond groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:             # <<<<<<<<<<<<<<
 *             for order2 in gb.order:
 *                 if order1 == order2: break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":541
 *             else:
 *                 return False
 *         for order1 in gb.order:             # <<<<<<<<<<<<<<
 *             for order2 in self.order:
 *                 if order1 == order2: break
 */
  if (unlikely(__pyx_v_gb->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 541; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_gb->order; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 541; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 541; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(PyString_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 541; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_order1, ((PyObject*)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":542
 *                 return False
 *         for order1 in gb.order:
 *             for order2 in self.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    if (unlikely(__pyx_v_self->order == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 542; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_self->order; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 542; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 542; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      if (!(likely(PyString_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 542; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_XDECREF_SET(__pyx_v_order2, ((PyObject*)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":543
 *         for order1 in gb.order:
 *             for order2 in self.order:
 *                 if order1 == order2: break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_order1, __pyx_v_order2, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 543; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = (__pyx_t_6 != 0);
      if (__pyx_t_7) {
        goto __pyx_L13_break;
      }

      /* "rmgpy/molecule/group.py":542
 *                 return False
 *         for order1 in gb.order:
 *             for order2 in self.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":545
 *                 if order1 == order2: break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         # Otherwise the two bond groups are equivalent
 *         return True
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":542
 *                 return False
 *         for order1 in gb.order:
 *             for order2 in self.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    __pyx_L13_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":541
 *             else:
 *                 return False
 *         for order1 in gb.order:             # <<<<<<<<<<<<<<
 *             for order2 in self.order:
 *                 if order1 == order2: break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":547
 *                 return False
 *         # Otherwise the two bond groups are equivalent
 *         return True             # <<<<<<<<<<<<<<
 * 
 *     def isSpecificCaseOf(self, other):
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":519
 *             raise ActionError('Unable to update GroupBond: Invalid action {0}".'.format(action))
 * 
 *     def equivalent(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.equivalent", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_gb);
  __Pyx_XDECREF(__pyx_v_order1);
  __Pyx_XDECREF(__pyx_v_order2);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_22equivalent[] = "\n        Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,\n        where `other` can be either an :class:`Bond` or an :class:`GroupBond`\n        object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("equivalent (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Edge, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_22equivalent(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_22equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("equivalent", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_equivalent(__pyx_v_self, __pyx_v_other, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.equivalent", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":549
 *         return True
 * 
 *     def isSpecificCaseOf(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is the same as `self` or is a more
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_gb = 0;
  PyObject *__pyx_v_order1 = 0;
  PyObject *__pyx_v_order2 = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSpecificCaseOf", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSpecificCaseOf); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 549; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 549; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 549; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 549; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 549; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":556
 *         """
 *         cython.declare(gb=GroupBond)
 *         if not isinstance(other, GroupBond):             # <<<<<<<<<<<<<<
 *             # Let the isSpecificCaseOf method of other handle it
 *             # We expect self to be a Bond object, but can't test for it here
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":560
 *             # We expect self to be a Bond object, but can't test for it here
 *             # because that would create an import cycle
 *             return other.isSpecificCaseOf(self)             # <<<<<<<<<<<<<<
 *         gb = other
 * 
 */
    __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge *)__pyx_v_other->__pyx_vtab)->isSpecificCaseOf(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_self), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 560; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_r = __pyx_t_7;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":561
 *             # because that would create an import cycle
 *             return other.isSpecificCaseOf(self)
 *         gb = other             # <<<<<<<<<<<<<<
 * 
 *         cython.declare(order1=str, order2=str)
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_GroupBond))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 561; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_gb = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":566
 *         # Compare two bond groups for equivalence
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match             # <<<<<<<<<<<<<<
 *             for order2 in gb.order: # can match any of these
 *                 if order1 == order2: break
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 566; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->order; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 566; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 566; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(PyString_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 566; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_order1, ((PyObject*)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":567
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match
 *             for order2 in gb.order: # can match any of these             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    if (unlikely(__pyx_v_gb->order == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_gb->order; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      if (!(likely(PyString_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_XDECREF_SET(__pyx_v_order2, ((PyObject*)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":568
 *         for order1 in self.order: # all these must match
 *             for order2 in gb.order: # can match any of these
 *                 if order1 == order2: break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_v_order1, __pyx_v_order2, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 568; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_6 = (__pyx_t_7 != 0);
      if (__pyx_t_6) {
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":567
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match
 *             for order2 in gb.order: # can match any of these             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":570
 *                 if order1 == order2: break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         # Otherwise self is in fact a specific case of other
 *         return True
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":567
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match
 *             for order2 in gb.order: # can match any of these             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":566
 *         # Compare two bond groups for equivalence
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match             # <<<<<<<<<<<<<<
 *             for order2 in gb.order: # can match any of these
 *                 if order1 == order2: break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":572
 *                 return False
 *         # Otherwise self is in fact a specific case of other
 *         return True             # <<<<<<<<<<<<<<
 * 
 * ################################################################################
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":549
 *         return True
 * 
 *     def isSpecificCaseOf(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is the same as `self` or is a more
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isSpecificCaseOf", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_gb);
  __Pyx_XDECREF(__pyx_v_order1);
  __Pyx_XDECREF(__pyx_v_order2);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf[] = "\n        Returns ``True`` if `other` is the same as `self` or is a more\n        specific case of `self`. Returns ``False`` if some of `self` is not\n        included in `other` or they are mutually exclusive.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSpecificCaseOf (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Edge, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 549; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSpecificCaseOf", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSpecificCaseOf(__pyx_v_self, __pyx_v_other, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 549; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 549; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isSpecificCaseOf", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":69
 * cdef class GroupBond(Edge):
 * 
 *     cdef public list order             # <<<<<<<<<<<<<<
 * 
 *     cpdef Edge copy(self)
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->order);
  __pyx_r = __pyx_v_self->order;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 69; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->order);
  __Pyx_DECREF(__pyx_v_self->order);
  __pyx_v_self->order = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.order.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->order);
  __Pyx_DECREF(__pyx_v_self->order);
  __pyx_v_self->order = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":585
 *     """
 * 
 *     def __init__(self, atoms=None, multiplicity=None):             # <<<<<<<<<<<<<<
 *         Graph.__init__(self, atoms)
 *         self.multiplicity = multiplicity if multiplicity else []
 */

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_atoms = 0;
  PyObject *__pyx_v_multiplicity = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__init__ (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atoms,&__pyx_n_s_multiplicity,0};
    PyObject* values[2] = {0,0};
    values[0] = ((PyObject *)Py_None);
    values[1] = ((PyObject *)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atoms);
          if (value) { values[0] = value; kw_args--; }
        }
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_multiplicity);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__init__") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 585; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_atoms = values[0];
    __pyx_v_multiplicity = values[1];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("__init__", 0, 0, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 585; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return -1;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group___init__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_atoms, __pyx_v_multiplicity);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group___init__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_atoms, PyObject *__pyx_v_multiplicity) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  Py_ssize_t __pyx_t_4;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__init__", 0);

  /* "rmgpy/molecule/group.py":586
 * 
 *     def __init__(self, atoms=None, multiplicity=None):
 *         Graph.__init__(self, atoms)             # <<<<<<<<<<<<<<
 *         self.multiplicity = multiplicity if multiplicity else []
 *         self.update()
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5graph_Graph)), __pyx_n_s_init); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 586; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  __pyx_t_4 = 0;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
      __pyx_t_4 = 1;
    }
  }
  __pyx_t_5 = PyTuple_New(2+__pyx_t_4); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 586; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  if (__pyx_t_3) {
    PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
  }
  __Pyx_INCREF(((PyObject *)__pyx_v_self));
  PyTuple_SET_ITEM(__pyx_t_5, 0+__pyx_t_4, ((PyObject *)__pyx_v_self));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
  __Pyx_INCREF(__pyx_v_atoms);
  PyTuple_SET_ITEM(__pyx_t_5, 1+__pyx_t_4, __pyx_v_atoms);
  __Pyx_GIVEREF(__pyx_v_atoms);
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 586; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":587
 *     def __init__(self, atoms=None, multiplicity=None):
 *         Graph.__init__(self, atoms)
 *         self.multiplicity = multiplicity if multiplicity else []             # <<<<<<<<<<<<<<
 *         self.update()
 * 
 */
  __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_v_multiplicity); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 587; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {
    if (!(likely(PyList_CheckExact(__pyx_v_multiplicity))||((__pyx_v_multiplicity) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_multiplicity)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 587; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_multiplicity);
    __pyx_t_1 = __pyx_v_multiplicity;
  } else {
    __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 587; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = __pyx_t_2;
    __pyx_t_2 = 0;
  }
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->multiplicity);
  __Pyx_DECREF(__pyx_v_self->multiplicity);
  __pyx_v_self->multiplicity = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":588
 *         Graph.__init__(self, atoms)
 *         self.multiplicity = multiplicity if multiplicity else []
 *         self.update()             # <<<<<<<<<<<<<<
 * 
 *     def __reduce__(self):
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_update); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_5 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_5)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_5);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (__pyx_t_5) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_5); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":585
 *     """
 * 
 *     def __init__(self, atoms=None, multiplicity=None):             # <<<<<<<<<<<<<<
 *         Graph.__init__(self, atoms)
 *         self.multiplicity = multiplicity if multiplicity else []
 */

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":590
 *         self.update()
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_3__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_2__reduce__[] = "\n        A helper function used when pickling an object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_3__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__reduce__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_2__reduce__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_2__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__reduce__", 0);

  /* "rmgpy/molecule/group.py":594
 *         A helper function used when pickling an object.
 *         """
 *         return (Group, (self.vertices,))             # <<<<<<<<<<<<<<
 * 
 *     def __getAtoms(self): return self.vertices
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 594; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_self->__pyx_base.vertices);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_self->__pyx_base.vertices);
  __Pyx_GIVEREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 594; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)));
  PyTuple_SET_ITEM(__pyx_t_2, 0, ((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)));
  __Pyx_GIVEREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)));
  PyTuple_SET_ITEM(__pyx_t_2, 1, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":590
 *         self.update()
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.__reduce__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":596
 *         return (Group, (self.vertices,))
 * 
 *     def __getAtoms(self): return self.vertices             # <<<<<<<<<<<<<<
 *     def __setAtoms(self, atoms): self.vertices = atoms
 *     atoms = property(__getAtoms, __setAtoms)
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_5__getAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_5__getAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__getAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_4__getAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_4__getAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__getAtoms", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_r = __pyx_v_self->__pyx_base.vertices;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":597
 * 
 *     def __getAtoms(self): return self.vertices
 *     def __setAtoms(self, atoms): self.vertices = atoms             # <<<<<<<<<<<<<<
 *     atoms = property(__getAtoms, __setAtoms)
 * 
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_7__setAtoms(PyObject *__pyx_v_self, PyObject *__pyx_v_atoms); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_7__setAtoms(PyObject *__pyx_v_self, PyObject *__pyx_v_atoms) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__setAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_6__setAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_atoms));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_6__setAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_atoms) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__setAtoms", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_atoms))||((__pyx_v_atoms) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_atoms)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 597; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_atoms;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->__pyx_base.vertices);
  __Pyx_DECREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_v_self->__pyx_base.vertices = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.__setAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":600
 *     atoms = property(__getAtoms, __setAtoms)
 * 
 *     def addAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Add an `atom` to the graph. The atom is initialized with no bonds.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_addAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_addAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_atom)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_atom));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_atom));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_atom));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":604
 *         Add an `atom` to the graph. The atom is initialized with no bonds.
 *         """
 *         return self.addVertex(atom)             # <<<<<<<<<<<<<<
 * 
 *     def addBond(self, bond):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.addVertex(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom), 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 604; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":600
 *     atoms = property(__getAtoms, __setAtoms)
 * 
 *     def addAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Add an `atom` to the graph. The atom is initialized with no bonds.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_8addAtom[] = "\n        Add an `atom` to the graph. The atom is initialized with no bonds.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_8addAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_8addAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_addAtom(__pyx_v_self, __pyx_v_atom, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":606
 *         return self.addVertex(atom)
 * 
 *     def addBond(self, bond):             # <<<<<<<<<<<<<<
 *         """
 *         Add a `bond` to the graph as an edge connecting the two atoms `atom1`
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_addBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_addBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 606; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_bond)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 606; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 606; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_bond));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_bond));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_bond));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 606; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":611
 *         and `atom2`.
 *         """
 *         return self.addEdge(bond)             # <<<<<<<<<<<<<<
 * 
 *     def getBonds(self, atom):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.addEdge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_bond), 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 611; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":606
 *         return self.addVertex(atom)
 * 
 *     def addBond(self, bond):             # <<<<<<<<<<<<<<
 *         """
 *         Add a `bond` to the graph as an edge connecting the two atoms `atom1`
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_10addBond[] = "\n        Add a `bond` to the graph as an edge connecting the two atoms `atom1`\n        and `atom2`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addBond (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_bond), __pyx_ptype_5rmgpy_8molecule_5group_GroupBond, 1, "bond", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 606; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_10addBond(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_bond));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_10addBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_addBond(__pyx_v_self, __pyx_v_bond, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 606; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":613
 *         return self.addEdge(bond)
 * 
 *     def getBonds(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Return a list of the bonds involving the specified `atom`.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_getBonds(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBonds", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getBonds); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_atom)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_atom));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_atom));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_atom));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyDict_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "dict", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":617
 *         Return a list of the bonds involving the specified `atom`.
 *         """
 *         return self.getEdges(atom)             # <<<<<<<<<<<<<<
 * 
 *     def getBond(self, atom1, atom2):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.getEdges(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 617; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":613
 *         return self.addEdge(bond)
 * 
 *     def getBonds(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Return a list of the bonds involving the specified `atom`.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBonds", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_12getBonds[] = "\n        Return a list of the bonds involving the specified `atom`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds(PyObject *__pyx_v_self, PyObject *__pyx_v_atom) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getBonds (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12getBonds(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12getBonds(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBonds", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_getBonds(__pyx_v_self, __pyx_v_atom, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBonds", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":619
 *         return self.getEdges(atom)
 * 
 *     def getBond(self, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Returns the bond connecting atoms `atom1` and `atom2`.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_f_5rmgpy_8molecule_5group_5Group_getBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_atom1));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_atom1));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom1));
      __Pyx_INCREF(((PyObject *)__pyx_v_atom2));
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, ((PyObject *)__pyx_v_atom2));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom2));
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5group_GroupBond))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":623
 *         Returns the bond connecting atoms `atom1` and `atom2`.
 *         """
 *         return self.getEdge(atom1, atom2)             # <<<<<<<<<<<<<<
 * 
 *     def hasAtom(self, atom):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  __pyx_t_1 = ((PyObject *)((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.getEdge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom1), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom2), 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 623; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (!(likely(((__pyx_t_1) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_1, __pyx_ptype_5rmgpy_8molecule_5group_GroupBond))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 623; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":619
 *         return self.getEdges(atom)
 * 
 *     def getBond(self, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Returns the bond connecting atoms `atom1` and `atom2`.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_14getBond[] = "\n        Returns the bond connecting atoms `atom1` and `atom2`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1 = 0;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2 = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getBond (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atom1,&__pyx_n_s_atom2,0};
    PyObject* values[2] = {0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom1)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom2)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("getBond", 1, 2, 2, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "getBond") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
    }
    __pyx_v_atom1 = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)values[0]);
    __pyx_v_atom2 = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("getBond", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom1), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom1", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom2), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom2", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_14getBond(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_atom1, __pyx_v_atom2);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_14getBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_getBond(__pyx_v_self, __pyx_v_atom1, __pyx_v_atom2, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":625
 *         return self.getEdge(atom1, atom2)
 * 
 *     def hasAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `atom` is an atom in the graph, or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_hasAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_hasAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_atom)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_atom));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_atom));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_atom));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":630
 *         not.
 *         """
 *         return self.hasVertex(atom)             # <<<<<<<<<<<<<<
 * 
 *     def hasBond(self, atom1, atom2):
 */
  __pyx_t_6 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.hasVertex(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom), 0); if (unlikely(__pyx_t_6 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_t_6;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":625
 *         return self.getEdge(atom1, atom2)
 * 
 *     def hasAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `atom` is an atom in the graph, or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.hasAtom", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_16hasAtom[] = "\n        Returns ``True`` if `atom` is an atom in the graph, or ``False`` if\n        not.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("hasAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_16hasAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_16hasAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_hasAtom(__pyx_v_self, __pyx_v_atom, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.hasAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":632
 *         return self.hasVertex(atom)
 * 
 *     def hasBond(self, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if atoms `atom1` and `atom2` are connected
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_hasBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_hasBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_atom1));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_atom1));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom1));
      __Pyx_INCREF(((PyObject *)__pyx_v_atom2));
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, ((PyObject *)__pyx_v_atom2));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom2));
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_7 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_7;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":637
 *         by an bond, or ``False`` if not.
 *         """
 *         return self.hasEdge(atom1, atom2)             # <<<<<<<<<<<<<<
 * 
 *     def removeAtom(self, atom):
 */
  __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.hasEdge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom1), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom2), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_t_7;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":632
 *         return self.hasVertex(atom)
 * 
 *     def hasBond(self, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if atoms `atom1` and `atom2` are connected
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.hasBond", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_18hasBond[] = "\n        Returns ``True`` if atoms `atom1` and `atom2` are connected\n        by an bond, or ``False`` if not.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1 = 0;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2 = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("hasBond (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atom1,&__pyx_n_s_atom2,0};
    PyObject* values[2] = {0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom1)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom2)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("hasBond", 1, 2, 2, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "hasBond") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
    }
    __pyx_v_atom1 = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)values[0]);
    __pyx_v_atom2 = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("hasBond", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.hasBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom1), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom1", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom2), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom2", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_18hasBond(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_atom1, __pyx_v_atom2);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18hasBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_hasBond(__pyx_v_self, __pyx_v_atom1, __pyx_v_atom2, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.hasBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":639
 *         return self.hasEdge(atom1, atom2)
 * 
 *     def removeAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Remove `atom` and all bonds associated with it from the graph. Does
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_removeAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("removeAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_removeAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_atom)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_atom));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_atom));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_atom));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":645
 *         removal.
 *         """
 *         return self.removeVertex(atom)             # <<<<<<<<<<<<<<
 * 
 *     def removeBond(self, bond):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.removeVertex(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 645; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":639
 *         return self.hasEdge(atom1, atom2)
 * 
 *     def removeAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Remove `atom` and all bonds associated with it from the graph. Does
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.removeAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_20removeAtom[] = "\n        Remove `atom` and all bonds associated with it from the graph. Does\n        not remove atoms that no longer have any bonds as a result of this\n        removal.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("removeAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_20removeAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_20removeAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("removeAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_removeAtom(__pyx_v_self, __pyx_v_atom, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.removeAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":647
 *         return self.removeVertex(atom)
 * 
 *     def removeBond(self, bond):             # <<<<<<<<<<<<<<
 *         """
 *         Remove the bond between atoms `atom1` and `atom2` from the graph.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_removeBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("removeBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_removeBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_bond)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_bond));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_bond));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_bond));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":653
 *         this removal.
 *         """
 *         return self.removeEdge(bond)             # <<<<<<<<<<<<<<
 * 
 *     def sortAtoms(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.removeEdge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_bond), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 653; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":647
 *         return self.removeVertex(atom)
 * 
 *     def removeBond(self, bond):             # <<<<<<<<<<<<<<
 *         """
 *         Remove the bond between atoms `atom1` and `atom2` from the graph.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.removeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_22removeBond[] = "\n        Remove the bond between atoms `atom1` and `atom2` from the graph.\n        Does not remove atoms that no longer have any bonds as a result of\n        this removal.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("removeBond (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_bond), __pyx_ptype_5rmgpy_8molecule_5group_GroupBond, 1, "bond", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_22removeBond(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_bond));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_22removeBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("removeBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_removeBond(__pyx_v_self, __pyx_v_bond, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.removeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":655
 *         return self.removeEdge(bond)
 * 
 *     def sortAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Sort the atoms in the graph. This can make certain operations, e.g.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_sortAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("sortAtoms", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_sortAtoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":660
 *         the isomorphism functions, much more efficient.
 *         """
 *         return self.sortVertices()             # <<<<<<<<<<<<<<
 * 
 *     def copy(self, deep=False):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.sortVertices(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 660; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":655
 *         return self.removeEdge(bond)
 * 
 *     def sortAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Sort the atoms in the graph. This can make certain operations, e.g.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.sortAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_24sortAtoms[] = "\n        Sort the atoms in the graph. This can make certain operations, e.g.\n        the isomorphism functions, much more efficient.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("sortAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_24sortAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_24sortAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("sortAtoms", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_sortAtoms(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.sortAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":662
 *         return self.sortVertices()
 * 
 *     def copy(self, deep=False):             # <<<<<<<<<<<<<<
 *         """
 *         Create a copy of the current graph. If `deep` is ``True``, a deep copy
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_f_5rmgpy_8molecule_5group_5Group_copy(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_copy *__pyx_optional_args) {
  int __pyx_v_deep = ((int)0);
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_other = 0;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_g = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_deep = __pyx_optional_args->deep;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_copy); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_deep); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5graph_Graph))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":670
 *         """
 *         other = cython.declare(Group)
 *         g = Graph.copy(self, deep)             # <<<<<<<<<<<<<<
 *         other = Group(g.vertices)
 *         return other
 */
  __pyx_t_7.__pyx_n = 1;
  __pyx_t_7.deep = __pyx_v_deep;
  __pyx_t_1 = ((PyObject *)__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->copy(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 1, &__pyx_t_7)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 670; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_g = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":671
 *         other = cython.declare(Group)
 *         g = Graph.copy(self, deep)
 *         other = Group(g.vertices)             # <<<<<<<<<<<<<<
 *         return other
 * 
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 671; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_g->vertices);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_g->vertices);
  __Pyx_GIVEREF(__pyx_v_g->vertices);
  __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)), __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 671; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":672
 *         g = Graph.copy(self, deep)
 *         other = Group(g.vertices)
 *         return other             # <<<<<<<<<<<<<<
 * 
 *     def update(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  __Pyx_INCREF(((PyObject *)__pyx_v_other));
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_other);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":662
 *         return self.sortVertices()
 * 
 *     def copy(self, deep=False):             # <<<<<<<<<<<<<<
 *         """
 *         Create a copy of the current graph. If `deep` is ``True``, a deep copy
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_other);
  __Pyx_XDECREF((PyObject *)__pyx_v_g);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_26copy[] = "\n        Create a copy of the current graph. If `deep` is ``True``, a deep copy\n        is made: copies of the vertices and edges are used in the new graph.\n        If `deep` is ``False`` or not specified, a shallow copy is made: the\n        original vertices and edges are used in the new graph.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  int __pyx_v_deep;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("copy (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_deep,0};
    PyObject* values[1] = {0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_deep);
          if (value) { values[0] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "copy") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    if (values[0]) {
      __pyx_v_deep = __Pyx_PyObject_IsTrue(values[0]); if (unlikely((__pyx_v_deep == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
    } else {
      __pyx_v_deep = ((int)0);
    }
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("copy", 0, 0, 1, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_26copy(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_deep);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_26copy(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_v_deep) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.deep = __pyx_v_deep;
  __pyx_t_1 = ((PyObject *)__pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.copy(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 1, &__pyx_t_2)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 662; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":674
 *         return other
 * 
 *     def update(self):             # <<<<<<<<<<<<<<
 * 
 *         self.updateConnectivityValues()
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_29update(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_29update(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("update (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_28update(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_28update(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("update", 0);

  /* "rmgpy/molecule/group.py":676
 *     def update(self):
 * 
 *         self.updateConnectivityValues()             # <<<<<<<<<<<<<<
 *         self.updateFingerprint()
 * 
 */
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.updateConnectivityValues(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 676; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":677
 * 
 *         self.updateConnectivityValues()
 *         self.updateFingerprint()             # <<<<<<<<<<<<<<
 * 
 * 
 */
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->updateFingerprint(__pyx_v_self, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 677; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":674
 *         return other
 * 
 *     def update(self):             # <<<<<<<<<<<<<<
 * 
 *         self.updateConnectivityValues()
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.update", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":680
 * 
 * 
 *     def merge(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Merge two groups so as to store them in a single
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_f_5rmgpy_8molecule_5group_5Group_merge(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_g = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_molecule = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("merge", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_merge); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 680; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 680; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 680; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 680; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5graph_Graph))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 680; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":686
 *         object is returned.
 *         """
 *         g = Graph.merge(self, other)             # <<<<<<<<<<<<<<
 *         molecule = Group(atoms=g.vertices)
 *         return molecule
 */
  __pyx_t_1 = ((PyObject *)__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->merge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 686; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_g = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":687
 *         """
 *         g = Graph.merge(self, other)
 *         molecule = Group(atoms=g.vertices)             # <<<<<<<<<<<<<<
 *         return molecule
 * 
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 687; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_atoms, __pyx_v_g->vertices) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 687; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)), __pyx_empty_tuple, __pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 687; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_molecule = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":688
 *         g = Graph.merge(self, other)
 *         molecule = Group(atoms=g.vertices)
 *         return molecule             # <<<<<<<<<<<<<<
 * 
 *     def split(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  __Pyx_INCREF(((PyObject *)__pyx_v_molecule));
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_molecule);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":680
 * 
 * 
 *     def merge(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Merge two groups so as to store them in a single
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.merge", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_g);
  __Pyx_XDECREF((PyObject *)__pyx_v_molecule);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_30merge[] = "\n        Merge two groups so as to store them in a single\n        :class:`Group` object. The merged :class:`Group`\n        object is returned.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("merge (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 680; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_30merge(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_30merge(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("merge", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_merge(__pyx_v_self, __pyx_v_other, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 680; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.merge", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":690
 *         return molecule
 * 
 *     def split(self):             # <<<<<<<<<<<<<<
 *         """
 *         Convert a single :class:`Group` object containing two or more
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_33split(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_split(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_graphs = NULL;
  PyObject *__pyx_v_molecules = NULL;
  PyObject *__pyx_v_g = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_molecule = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("split", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_split); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 690; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_33split)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 690; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 690; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyList_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 690; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":695
 *         unconnected groups into separate class:`Group` objects.
 *         """
 *         graphs = Graph.split(self)             # <<<<<<<<<<<<<<
 *         molecules = []
 *         for g in graphs:
 */
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->split(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 695; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_graphs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":696
 *         """
 *         graphs = Graph.split(self)
 *         molecules = []             # <<<<<<<<<<<<<<
 *         for g in graphs:
 *             molecule = Group(atoms=g.vertices)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_molecules = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":697
 *         graphs = Graph.split(self)
 *         molecules = []
 *         for g in graphs:             # <<<<<<<<<<<<<<
 *             molecule = Group(atoms=g.vertices)
 *             molecules.append(molecule)
 */
  if (unlikely(__pyx_v_graphs == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 697; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_graphs; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 697; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 697; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_g, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":698
 *         molecules = []
 *         for g in graphs:
 *             molecule = Group(atoms=g.vertices)             # <<<<<<<<<<<<<<
 *             molecules.append(molecule)
 *         return molecules
 */
    __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 698; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_g, __pyx_n_s_vertices); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 698; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_atoms, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 698; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)), __pyx_empty_tuple, __pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 698; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_XDECREF_SET(__pyx_v_molecule, ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_3));
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":699
 *         for g in graphs:
 *             molecule = Group(atoms=g.vertices)
 *             molecules.append(molecule)             # <<<<<<<<<<<<<<
 *         return molecules
 * 
 */
    __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_molecules, ((PyObject *)__pyx_v_molecule)); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 699; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

    /* "rmgpy/molecule/group.py":697
 *         graphs = Graph.split(self)
 *         molecules = []
 *         for g in graphs:             # <<<<<<<<<<<<<<
 *             molecule = Group(atoms=g.vertices)
 *             molecules.append(molecule)
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":700
 *             molecule = Group(atoms=g.vertices)
 *             molecules.append(molecule)
 *         return molecules             # <<<<<<<<<<<<<<
 * 
 *     def clearLabeledAtoms(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_molecules);
  __pyx_r = __pyx_v_molecules;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":690
 *         return molecule
 * 
 *     def split(self):             # <<<<<<<<<<<<<<
 *         """
 *         Convert a single :class:`Group` object containing two or more
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.split", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_graphs);
  __Pyx_XDECREF(__pyx_v_molecules);
  __Pyx_XDECREF(__pyx_v_g);
  __Pyx_XDECREF((PyObject *)__pyx_v_molecule);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_33split(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_32split[] = "\n        Convert a single :class:`Group` object containing two or more\n        unconnected groups into separate class:`Group` objects.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_33split(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("split (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_32split(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_32split(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("split", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_split(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 690; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.split", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":702
 *         return molecules
 * 
 *     def clearLabeledAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Remove the labels from all atoms in the molecular group.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_clearLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("clearLabeledAtoms", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_clearLabeledAtoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":706
 *         Remove the labels from all atoms in the molecular group.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             atom.label = ''
 * 
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 706; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 706; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 706; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":707
 *         """
 *         for atom in self.vertices:
 *             atom.label = ''             # <<<<<<<<<<<<<<
 * 
 *     def containsLabeledAtom(self, label):
 */
    if (__Pyx_PyObject_SetAttrStr(__pyx_v_atom, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 707; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

    /* "rmgpy/molecule/group.py":706
 *         Remove the labels from all atoms in the molecular group.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             atom.label = ''
 * 
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":702
 *         return molecules
 * 
 *     def clearLabeledAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Remove the labels from all atoms in the molecular group.
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.clearLabeledAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms[] = "\n        Remove the labels from all atoms in the molecular group.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("clearLabeledAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("clearLabeledAtoms", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_clearLabeledAtoms(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.clearLabeledAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":709
 *             atom.label = ''
 * 
 *     def containsLabeledAtom(self, label):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the group contains an atom with the label
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_containsLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atom = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("containsLabeledAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_containsLabeledAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 709; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 709; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 709; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_label);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_label);
        __Pyx_GIVEREF(__pyx_v_label);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 709; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 709; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":714
 *         `label` and ``False`` otherwise.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label == label: return True
 *         return False
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":715
 *         """
 *         for atom in self.vertices:
 *             if atom.label == label: return True             # <<<<<<<<<<<<<<
 *         return False
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 715; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_v_label, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 715; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_6) {
      __pyx_r = 1;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":714
 *         `label` and ``False`` otherwise.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label == label: return True
 *         return False
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":716
 *         for atom in self.vertices:
 *             if atom.label == label: return True
 *         return False             # <<<<<<<<<<<<<<
 * 
 *     def getLabeledAtom(self, label):
 */
  __pyx_r = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":709
 *             atom.label = ''
 * 
 *     def containsLabeledAtom(self, label):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the group contains an atom with the label
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.containsLabeledAtom", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom[] = "\n        Return ``True`` if the group contains an atom with the label\n        `label` and ``False`` otherwise.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("containsLabeledAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_label), (&PyString_Type), 1, "label", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 709; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject*)__pyx_v_label));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("containsLabeledAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_containsLabeledAtom(__pyx_v_self, __pyx_v_label, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 709; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.containsLabeledAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":718
 *         return False
 * 
 *     def getLabeledAtom(self, label):             # <<<<<<<<<<<<<<
 *         """
 *         Return the atom in the group that is labeled with the given `label`.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atom = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  Py_ssize_t __pyx_t_6;
  int __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getLabeledAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getLabeledAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_label);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_label);
        __Pyx_GIVEREF(__pyx_v_label);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":723
 *         Raises :class:`ValueError` if no atom in the group has that label.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label == label: return atom
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 723; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 723; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 723; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":724
 *         """
 *         for atom in self.vertices:
 *             if atom.label == label: return atom             # <<<<<<<<<<<<<<
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 724; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_v_label, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 724; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_7) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      if (!(likely(((__pyx_v_atom) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_atom, __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 724; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_INCREF(__pyx_v_atom);
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":723
 *         Raises :class:`ValueError` if no atom in the group has that label.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label == label: return atom
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":725
 *         for atom in self.vertices:
 *             if atom.label == label: return atom
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))             # <<<<<<<<<<<<<<
 * 
 *     def getLabeledAtoms(self):
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_No_atom_in_the_functional_group, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_v_label); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(__pyx_v_label);
    PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_label);
    __Pyx_GIVEREF(__pyx_v_label);
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_ValueError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_Raise(__pyx_t_1, 0, 0, 0);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  {__pyx_filename = __pyx_f[0]; __pyx_lineno = 725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":718
 *         return False
 * 
 *     def getLabeledAtom(self, label):             # <<<<<<<<<<<<<<
 *         """
 *         Return the atom in the group that is labeled with the given `label`.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getLabeledAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_38getLabeledAtom[] = "\n        Return the atom in the group that is labeled with the given `label`.\n        Raises :class:`ValueError` if no atom in the group has that label.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getLabeledAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_label), (&PyString_Type), 1, "label", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_38getLabeledAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject*)__pyx_v_label));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_38getLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getLabeledAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtom(__pyx_v_self, __pyx_v_label, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getLabeledAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":727
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 * 
 *     def getLabeledAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return the labeled atoms as a ``dict`` with the keys being the labels
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_labeled = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  int __pyx_t_6;
  int __pyx_t_7;
  int __pyx_t_8;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getLabeledAtoms", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getLabeledAtoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyDict_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "dict", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":733
 *         same label, the value is converted to a list of these atoms.
 *         """
 *         labeled = {}             # <<<<<<<<<<<<<<
 *         for atom in self.vertices:
 *             if atom.label != '':
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 733; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_labeled = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":734
 *         """
 *         labeled = {}
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label != '':
 *                 if atom.label in labeled:
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 734; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 734; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 734; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":735
 *         labeled = {}
 *         for atom in self.vertices:
 *             if atom.label != '':             # <<<<<<<<<<<<<<
 *                 if atom.label in labeled:
 *                     if isinstance(labeled[atom.label],list):
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 735; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_kp_s_, Py_NE)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 735; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_6) {

      /* "rmgpy/molecule/group.py":736
 *         for atom in self.vertices:
 *             if atom.label != '':
 *                 if atom.label in labeled:             # <<<<<<<<<<<<<<
 *                     if isinstance(labeled[atom.label],list):
 *                         labeled[atom.label].append(atom)
 */
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 736; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_6 = (__Pyx_PyDict_Contains(__pyx_t_2, __pyx_v_labeled, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 736; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_7 = (__pyx_t_6 != 0);
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":737
 *             if atom.label != '':
 *                 if atom.label in labeled:
 *                     if isinstance(labeled[atom.label],list):             # <<<<<<<<<<<<<<
 *                         labeled[atom.label].append(atom)
 *                     else:
 */
        __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 737; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __pyx_t_3 = __Pyx_PyDict_GetItem(__pyx_v_labeled, __pyx_t_2); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 737; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_3);
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        __pyx_t_7 = PyList_Check(__pyx_t_3); 
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_6 = (__pyx_t_7 != 0);
        if (__pyx_t_6) {

          /* "rmgpy/molecule/group.py":738
 *                 if atom.label in labeled:
 *                     if isinstance(labeled[atom.label],list):
 *                         labeled[atom.label].append(atom)             # <<<<<<<<<<<<<<
 *                     else:
 *                         labeled[atom.label] = [labeled[atom.label]]
 */
          __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 738; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          __pyx_t_2 = __Pyx_PyDict_GetItem(__pyx_v_labeled, __pyx_t_3); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 738; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_2);
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_t_8 = __Pyx_PyObject_Append(__pyx_t_2, __pyx_v_atom); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 738; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L7;
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":740
 *                         labeled[atom.label].append(atom)
 *                     else:
 *                         labeled[atom.label] = [labeled[atom.label]]             # <<<<<<<<<<<<<<
 *                         labeled[atom.label].append(atom)
 *                 else:
 */
          __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __pyx_t_3 = __Pyx_PyDict_GetItem(__pyx_v_labeled, __pyx_t_2); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 740; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          PyList_SET_ITEM(__pyx_t_2, 0, __pyx_t_3);
          __Pyx_GIVEREF(__pyx_t_3);
          __pyx_t_3 = 0;
          __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          if (unlikely(PyDict_SetItem(__pyx_v_labeled, __pyx_t_3, __pyx_t_2) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

          /* "rmgpy/molecule/group.py":741
 *                     else:
 *                         labeled[atom.label] = [labeled[atom.label]]
 *                         labeled[atom.label].append(atom)             # <<<<<<<<<<<<<<
 *                 else:
 *                     labeled[atom.label] = atom
 */
          __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 741; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __pyx_t_3 = __Pyx_PyDict_GetItem(__pyx_v_labeled, __pyx_t_2); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 741; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          __pyx_t_8 = __Pyx_PyObject_Append(__pyx_t_3, __pyx_v_atom); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 741; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        }
        __pyx_L7:;
        goto __pyx_L6;
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":743
 *                         labeled[atom.label].append(atom)
 *                 else:
 *                     labeled[atom.label] = atom             # <<<<<<<<<<<<<<
 *         return labeled
 * 
 */
        __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        if (unlikely(PyDict_SetItem(__pyx_v_labeled, __pyx_t_3, __pyx_v_atom) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      }
      __pyx_L6:;
      goto __pyx_L5;
    }
    __pyx_L5:;

    /* "rmgpy/molecule/group.py":734
 *         """
 *         labeled = {}
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label != '':
 *                 if atom.label in labeled:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":744
 *                 else:
 *                     labeled[atom.label] = atom
 *         return labeled             # <<<<<<<<<<<<<<
 * 
 *     def fromAdjacencyList(self, adjlist):
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_labeled);
  __pyx_r = __pyx_v_labeled;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":727
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 * 
 *     def getLabeledAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return the labeled atoms as a ``dict`` with the keys being the labels
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getLabeledAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_labeled);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms[] = "\n        Return the labeled atoms as a ``dict`` with the keys being the labels\n        and the values the atoms themselves. If two or more atoms have the\n        same label, the value is converted to a list of these atoms.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getLabeledAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getLabeledAtoms", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtoms(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getLabeledAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":746
 *         return labeled
 * 
 *     def fromAdjacencyList(self, adjlist):             # <<<<<<<<<<<<<<
 *         """
 *         Convert a string adjacency list `adjlist` to a molecular structure.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_v_adjlist); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_fromAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_adjlist, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_fromAdjacencyList = NULL;
  PyObject *__pyx_v_multiplicity = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *(*__pyx_t_6)(PyObject *);
  int __pyx_t_7;
  int __pyx_t_8;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("fromAdjacencyList", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_fromAdjacencyList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 746; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_adjlist); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 746; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 746; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_adjlist);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_adjlist);
        __Pyx_GIVEREF(__pyx_v_adjlist);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 746; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":752
 *         ``False``.
 *         """
 *         from .adjlist import fromAdjacencyList             # <<<<<<<<<<<<<<
 *         self.vertices, multiplicity = fromAdjacencyList(adjlist, group=True)
 *         if multiplicity is not None:
 */
  __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 752; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_n_s_fromAdjacencyList);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_n_s_fromAdjacencyList);
  __Pyx_GIVEREF(__pyx_n_s_fromAdjacencyList);
  __pyx_t_2 = __Pyx_Import(__pyx_n_s_adjlist, __pyx_t_1, 1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 752; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_ImportFrom(__pyx_t_2, __pyx_n_s_fromAdjacencyList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 752; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_fromAdjacencyList = __pyx_t_1;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":753
 *         """
 *         from .adjlist import fromAdjacencyList
 *         self.vertices, multiplicity = fromAdjacencyList(adjlist, group=True)             # <<<<<<<<<<<<<<
 *         if multiplicity is not None:
 *             self.multiplicity = multiplicity
 */
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_v_adjlist);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_adjlist);
  __Pyx_GIVEREF(__pyx_v_adjlist);
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_group, Py_True) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = __Pyx_PyObject_Call(__pyx_v_fromAdjacencyList, __pyx_t_2, __pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if ((likely(PyTuple_CheckExact(__pyx_t_3))) || (PyList_CheckExact(__pyx_t_3))) {
    PyObject* sequence = __pyx_t_3;
    #if CYTHON_COMPILING_IN_CPYTHON
    Py_ssize_t size = Py_SIZE(sequence);
    #else
    Py_ssize_t size = PySequence_Size(sequence);
    #endif
    if (unlikely(size != 2)) {
      if (size > 2) __Pyx_RaiseTooManyValuesError(2);
      else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    #if CYTHON_COMPILING_IN_CPYTHON
    if (likely(PyTuple_CheckExact(sequence))) {
      __pyx_t_1 = PyTuple_GET_ITEM(sequence, 0); 
      __pyx_t_2 = PyTuple_GET_ITEM(sequence, 1); 
    } else {
      __pyx_t_1 = PyList_GET_ITEM(sequence, 0); 
      __pyx_t_2 = PyList_GET_ITEM(sequence, 1); 
    }
    __Pyx_INCREF(__pyx_t_1);
    __Pyx_INCREF(__pyx_t_2);
    #else
    __pyx_t_1 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_2 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    #endif
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  } else {
    Py_ssize_t index = -1;
    __pyx_t_5 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_6 = Py_TYPE(__pyx_t_5)->tp_iternext;
    index = 0; __pyx_t_1 = __pyx_t_6(__pyx_t_5); if (unlikely(!__pyx_t_1)) goto __pyx_L3_unpacking_failed;
    __Pyx_GOTREF(__pyx_t_1);
    index = 1; __pyx_t_2 = __pyx_t_6(__pyx_t_5); if (unlikely(!__pyx_t_2)) goto __pyx_L3_unpacking_failed;
    __Pyx_GOTREF(__pyx_t_2);
    if (__Pyx_IternextUnpackEndCheck(__pyx_t_6(__pyx_t_5), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_6 = NULL;
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    goto __pyx_L4_unpacking_done;
    __pyx_L3_unpacking_failed:;
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_6 = NULL;
    if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_L4_unpacking_done:;
  }
  if (!(likely(PyList_CheckExact(__pyx_t_1))||((__pyx_t_1) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_t_1)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->__pyx_base.vertices);
  __Pyx_DECREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_v_self->__pyx_base.vertices = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_v_multiplicity = __pyx_t_2;
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":754
 *         from .adjlist import fromAdjacencyList
 *         self.vertices, multiplicity = fromAdjacencyList(adjlist, group=True)
 *         if multiplicity is not None:             # <<<<<<<<<<<<<<
 *             self.multiplicity = multiplicity
 *         self.update()
 */
  __pyx_t_7 = (__pyx_v_multiplicity != Py_None);
  __pyx_t_8 = (__pyx_t_7 != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":755
 *         self.vertices, multiplicity = fromAdjacencyList(adjlist, group=True)
 *         if multiplicity is not None:
 *             self.multiplicity = multiplicity             # <<<<<<<<<<<<<<
 *         self.update()
 *         return self
 */
    if (!(likely(PyList_CheckExact(__pyx_v_multiplicity))||((__pyx_v_multiplicity) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_multiplicity)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 755; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_3 = __pyx_v_multiplicity;
    __Pyx_INCREF(__pyx_t_3);
    __Pyx_GIVEREF(__pyx_t_3);
    __Pyx_GOTREF(__pyx_v_self->multiplicity);
    __Pyx_DECREF(__pyx_v_self->multiplicity);
    __pyx_v_self->multiplicity = ((PyObject*)__pyx_t_3);
    __pyx_t_3 = 0;
    goto __pyx_L5;
  }
  __pyx_L5:;

  /* "rmgpy/molecule/group.py":756
 *         if multiplicity is not None:
 *             self.multiplicity = multiplicity
 *         self.update()             # <<<<<<<<<<<<<<
 *         return self
 * 
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_update); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 756; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_1 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_1 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_1)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_1);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (__pyx_t_1) {
    __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 756; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  } else {
    __pyx_t_3 = __Pyx_PyObject_CallNoArg(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 756; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":757
 *             self.multiplicity = multiplicity
 *         self.update()
 *         return self             # <<<<<<<<<<<<<<
 * 
 *     def toAdjacencyList(self, label=''):
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(((PyObject *)__pyx_v_self));
  __pyx_r = ((PyObject *)__pyx_v_self);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":746
 *         return labeled
 * 
 *     def fromAdjacencyList(self, adjlist):             # <<<<<<<<<<<<<<
 *         """
 *         Convert a string adjacency list `adjlist` to a molecular structure.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.fromAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_fromAdjacencyList);
  __Pyx_XDECREF(__pyx_v_multiplicity);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_v_adjlist); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList[] = "\n        Convert a string adjacency list `adjlist` to a molecular structure.\n        Skips the first line (assuming it's a label) unless `withLabel` is\n        ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_v_adjlist) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("fromAdjacencyList (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_adjlist), (&PyString_Type), 1, "adjlist", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 746; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject*)__pyx_v_adjlist));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_adjlist) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("fromAdjacencyList", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_fromAdjacencyList(__pyx_v_self, __pyx_v_adjlist, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 746; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.fromAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":759
 *         return self
 * 
 *     def toAdjacencyList(self, label=''):             # <<<<<<<<<<<<<<
 *         """
 *         Convert the molecular structure to a string adjacency list.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_toAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList *__pyx_optional_args) {
  PyObject *__pyx_v_label = ((PyObject*)__pyx_kp_s_);
  PyObject *__pyx_v_toAdjacencyList = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("toAdjacencyList", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_label = __pyx_optional_args->label;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_toAdjacencyList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 759; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 759; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 759; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_label);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_label);
        __Pyx_GIVEREF(__pyx_v_label);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 759; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":763
 *         Convert the molecular structure to a string adjacency list.
 *         """
 *         from .adjlist import toAdjacencyList             # <<<<<<<<<<<<<<
 *         return toAdjacencyList(self.vertices, multiplicity=self.multiplicity, label='', group=True)
 * 
 */
  __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 763; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_n_s_toAdjacencyList);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_n_s_toAdjacencyList);
  __Pyx_GIVEREF(__pyx_n_s_toAdjacencyList);
  __pyx_t_2 = __Pyx_Import(__pyx_n_s_adjlist, __pyx_t_1, 1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 763; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_ImportFrom(__pyx_t_2, __pyx_n_s_toAdjacencyList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 763; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_toAdjacencyList = __pyx_t_1;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":764
 *         """
 *         from .adjlist import toAdjacencyList
 *         return toAdjacencyList(self.vertices, multiplicity=self.multiplicity, label='', group=True)             # <<<<<<<<<<<<<<
 * 
 * 
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 764; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_v_self->__pyx_base.vertices);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_self->__pyx_base.vertices);
  __Pyx_GIVEREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 764; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_multiplicity, __pyx_v_self->multiplicity) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 764; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 764; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_group, Py_True) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 764; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = __Pyx_PyObject_Call(__pyx_v_toAdjacencyList, __pyx_t_2, __pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 764; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_r = __pyx_t_3;
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":759
 *         return self
 * 
 *     def toAdjacencyList(self, label=''):             # <<<<<<<<<<<<<<
 *         """
 *         Convert the molecular structure to a string adjacency list.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.toAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_toAdjacencyList);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_44toAdjacencyList[] = "\n        Convert the molecular structure to a string adjacency list.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_label = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("toAdjacencyList (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_label,0};
    PyObject* values[1] = {0};
    values[0] = ((PyObject*)__pyx_kp_s_);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_label);
          if (value) { values[0] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "toAdjacencyList") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 759; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_label = ((PyObject*)values[0]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("toAdjacencyList", 0, 0, 1, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 759; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.toAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_label), (&PyString_Type), 1, "label", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 759; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_44toAdjacencyList(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_label);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_44toAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("toAdjacencyList", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.label = __pyx_v_label;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->toAdjacencyList(__pyx_v_self, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 759; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.toAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":767
 * 
 * 
 *     def updateFingerprint(self):             # <<<<<<<<<<<<<<
 *         """
 *         Update the molecular fingerprint used to accelerate the subgraph
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_updateFingerprint(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_atomType = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_carbon = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_nitrogen = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_oxygen = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_sulfur = 0;
  int __pyx_v_isCarbon;
  int __pyx_v_isNitrogen;
  int __pyx_v_isOxygen;
  int __pyx_v_isSulfur;
  int __pyx_v_radical;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_t_7;
  int __pyx_t_8;
  int __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("updateFingerprint", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_updateFingerprint); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 767; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 767; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 767; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":776
 *         cython.declare(isCarbon=cython.bint, isNitrogen=cython.bint, isOxygen=cython.bint, isSulfur=cython.bint, radical=cython.int)
 * 
 *         carbon   = atomTypes['C']             # <<<<<<<<<<<<<<
 *         nitrogen = atomTypes['N']
 *         oxygen   = atomTypes['O']
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 776; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_C); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 776; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 776; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_carbon = ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":777
 * 
 *         carbon   = atomTypes['C']
 *         nitrogen = atomTypes['N']             # <<<<<<<<<<<<<<
 *         oxygen   = atomTypes['O']
 *         sulfur   = atomTypes['S']
 */
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_1 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_N); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  if (!(likely(((__pyx_t_1) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_1, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_nitrogen = ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":778
 *         carbon   = atomTypes['C']
 *         nitrogen = atomTypes['N']
 *         oxygen   = atomTypes['O']             # <<<<<<<<<<<<<<
 *         sulfur   = atomTypes['S']
 * 
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 778; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_O); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 778; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 778; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_oxygen = ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":779
 *         nitrogen = atomTypes['N']
 *         oxygen   = atomTypes['O']
 *         sulfur   = atomTypes['S']             # <<<<<<<<<<<<<<
 * 
 *         self.carbonCount   = 0
 */
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 779; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_1 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_S); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 779; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  if (!(likely(((__pyx_t_1) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_1, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 779; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_sulfur = ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":781
 *         sulfur   = atomTypes['S']
 * 
 *         self.carbonCount   = 0             # <<<<<<<<<<<<<<
 *         self.nitrogenCount = 0
 *         self.oxygenCount   = 0
 */
  __pyx_v_self->carbonCount = 0;

  /* "rmgpy/molecule/group.py":782
 * 
 *         self.carbonCount   = 0
 *         self.nitrogenCount = 0             # <<<<<<<<<<<<<<
 *         self.oxygenCount   = 0
 *         self.sulfurCount   = 0
 */
  __pyx_v_self->nitrogenCount = 0;

  /* "rmgpy/molecule/group.py":783
 *         self.carbonCount   = 0
 *         self.nitrogenCount = 0
 *         self.oxygenCount   = 0             # <<<<<<<<<<<<<<
 *         self.sulfurCount   = 0
 *         self.radicalCount  = 0
 */
  __pyx_v_self->oxygenCount = 0;

  /* "rmgpy/molecule/group.py":784
 *         self.nitrogenCount = 0
 *         self.oxygenCount   = 0
 *         self.sulfurCount   = 0             # <<<<<<<<<<<<<<
 *         self.radicalCount  = 0
 *         for atom in self.vertices:
 */
  __pyx_v_self->sulfurCount = 0;

  /* "rmgpy/molecule/group.py":785
 *         self.oxygenCount   = 0
 *         self.sulfurCount   = 0
 *         self.radicalCount  = 0             # <<<<<<<<<<<<<<
 *         for atom in self.vertices:
 *             if len(atom.atomType) == 1:
 */
  __pyx_v_self->radicalCount = 0;

  /* "rmgpy/molecule/group.py":786
 *         self.sulfurCount   = 0
 *         self.radicalCount  = 0
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if len(atom.atomType) == 1:
 *                 atomType   = atom.atomType[0]
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 786; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 786; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 786; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 786; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_atom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":787
 *         self.radicalCount  = 0
 *         for atom in self.vertices:
 *             if len(atom.atomType) == 1:             # <<<<<<<<<<<<<<
 *                 atomType   = atom.atomType[0]
 *                 isCarbon   = atomType.equivalent(carbon)
 */
    __pyx_t_2 = __pyx_v_atom->atomType;
    __Pyx_INCREF(__pyx_t_2);
    if (unlikely(__pyx_t_2 == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 787; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_6 = PyList_GET_SIZE(__pyx_t_2); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 787; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_7 = ((__pyx_t_6 == 1) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":788
 *         for atom in self.vertices:
 *             if len(atom.atomType) == 1:
 *                 atomType   = atom.atomType[0]             # <<<<<<<<<<<<<<
 *                 isCarbon   = atomType.equivalent(carbon)
 *                 isNitrogen = atomType.equivalent(nitrogen)
 */
      if (unlikely(__pyx_v_atom->atomType == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 788; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_atom->atomType, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 788; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 788; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_XDECREF_SET(__pyx_v_atomType, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2));
      __pyx_t_2 = 0;

      /* "rmgpy/molecule/group.py":789
 *             if len(atom.atomType) == 1:
 *                 atomType   = atom.atomType[0]
 *                 isCarbon   = atomType.equivalent(carbon)             # <<<<<<<<<<<<<<
 *                 isNitrogen = atomType.equivalent(nitrogen)
 *                 isOxygen   = atomType.equivalent(oxygen)
 */
      __pyx_v_isCarbon = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType->__pyx_vtab)->equivalent(__pyx_v_atomType, __pyx_v_carbon, 0);

      /* "rmgpy/molecule/group.py":790
 *                 atomType   = atom.atomType[0]
 *                 isCarbon   = atomType.equivalent(carbon)
 *                 isNitrogen = atomType.equivalent(nitrogen)             # <<<<<<<<<<<<<<
 *                 isOxygen   = atomType.equivalent(oxygen)
 *                 isSulfur   = atomType.equivalent(sulfur)
 */
      __pyx_v_isNitrogen = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType->__pyx_vtab)->equivalent(__pyx_v_atomType, __pyx_v_nitrogen, 0);

      /* "rmgpy/molecule/group.py":791
 *                 isCarbon   = atomType.equivalent(carbon)
 *                 isNitrogen = atomType.equivalent(nitrogen)
 *                 isOxygen   = atomType.equivalent(oxygen)             # <<<<<<<<<<<<<<
 *                 isSulfur   = atomType.equivalent(sulfur)
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:
 */
      __pyx_v_isOxygen = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType->__pyx_vtab)->equivalent(__pyx_v_atomType, __pyx_v_oxygen, 0);

      /* "rmgpy/molecule/group.py":792
 *                 isNitrogen = atomType.equivalent(nitrogen)
 *                 isOxygen   = atomType.equivalent(oxygen)
 *                 isSulfur   = atomType.equivalent(sulfur)             # <<<<<<<<<<<<<<
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:
 *                     self.carbonCount += 1
 */
      __pyx_v_isSulfur = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType->__pyx_vtab)->equivalent(__pyx_v_atomType, __pyx_v_sulfur, 0);

      /* "rmgpy/molecule/group.py":793
 *                 isOxygen   = atomType.equivalent(oxygen)
 *                 isSulfur   = atomType.equivalent(sulfur)
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:             # <<<<<<<<<<<<<<
 *                     self.carbonCount += 1
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:
 */
      __pyx_t_8 = (__pyx_v_isCarbon != 0);
      if (__pyx_t_8) {
        goto __pyx_L8_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L7_bool_binop_done;
      }
      __pyx_L8_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isNitrogen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L9_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L7_bool_binop_done;
      }
      __pyx_L9_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isOxygen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L10_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L7_bool_binop_done;
      }
      __pyx_L10_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isSulfur != 0)) != 0);
      __pyx_t_7 = __pyx_t_8;
      __pyx_L7_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":794
 *                 isSulfur   = atomType.equivalent(sulfur)
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:
 *                     self.carbonCount += 1             # <<<<<<<<<<<<<<
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:
 *                     self.nitrogenCount += 1
 */
        __pyx_v_self->carbonCount = (__pyx_v_self->carbonCount + 1);
        goto __pyx_L6;
      }

      /* "rmgpy/molecule/group.py":795
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:
 *                     self.carbonCount += 1
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:             # <<<<<<<<<<<<<<
 *                     self.nitrogenCount += 1
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:
 */
      __pyx_t_8 = (__pyx_v_isNitrogen != 0);
      if (__pyx_t_8) {
        goto __pyx_L12_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L11_bool_binop_done;
      }
      __pyx_L12_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isCarbon != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L13_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L11_bool_binop_done;
      }
      __pyx_L13_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isOxygen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L14_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L11_bool_binop_done;
      }
      __pyx_L14_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isSulfur != 0)) != 0);
      __pyx_t_7 = __pyx_t_8;
      __pyx_L11_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":796
 *                     self.carbonCount += 1
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:
 *                     self.nitrogenCount += 1             # <<<<<<<<<<<<<<
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:
 *                     self.oxygenCount += 1
 */
        __pyx_v_self->nitrogenCount = (__pyx_v_self->nitrogenCount + 1);
        goto __pyx_L6;
      }

      /* "rmgpy/molecule/group.py":797
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:
 *                     self.nitrogenCount += 1
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:             # <<<<<<<<<<<<<<
 *                     self.oxygenCount += 1
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:
 */
      __pyx_t_8 = (__pyx_v_isOxygen != 0);
      if (__pyx_t_8) {
        goto __pyx_L16_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L15_bool_binop_done;
      }
      __pyx_L16_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isCarbon != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L17_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L15_bool_binop_done;
      }
      __pyx_L17_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isNitrogen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L18_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L15_bool_binop_done;
      }
      __pyx_L18_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isSulfur != 0)) != 0);
      __pyx_t_7 = __pyx_t_8;
      __pyx_L15_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":798
 *                     self.nitrogenCount += 1
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:
 *                     self.oxygenCount += 1             # <<<<<<<<<<<<<<
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:
 *                     self.sulfurCount += 1
 */
        __pyx_v_self->oxygenCount = (__pyx_v_self->oxygenCount + 1);
        goto __pyx_L6;
      }

      /* "rmgpy/molecule/group.py":799
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:
 *                     self.oxygenCount += 1
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:             # <<<<<<<<<<<<<<
 *                     self.sulfurCount += 1
 *             if len(atom.radicalElectrons) == 1:
 */
      __pyx_t_8 = (__pyx_v_isSulfur != 0);
      if (__pyx_t_8) {
        goto __pyx_L20_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L19_bool_binop_done;
      }
      __pyx_L20_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isCarbon != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L21_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L19_bool_binop_done;
      }
      __pyx_L21_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isNitrogen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L22_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L19_bool_binop_done;
      }
      __pyx_L22_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isOxygen != 0)) != 0);
      __pyx_t_7 = __pyx_t_8;
      __pyx_L19_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":800
 *                     self.oxygenCount += 1
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:
 *                     self.sulfurCount += 1             # <<<<<<<<<<<<<<
 *             if len(atom.radicalElectrons) == 1:
 *                 radical = atom.radicalElectrons[0]
 */
        __pyx_v_self->sulfurCount = (__pyx_v_self->sulfurCount + 1);
        goto __pyx_L6;
      }
      __pyx_L6:;
      goto __pyx_L5;
    }
    __pyx_L5:;

    /* "rmgpy/molecule/group.py":801
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:
 *                     self.sulfurCount += 1
 *             if len(atom.radicalElectrons) == 1:             # <<<<<<<<<<<<<<
 *                 radical = atom.radicalElectrons[0]
 *                 self.radicalCount += radical
 */
    __pyx_t_2 = __pyx_v_atom->radicalElectrons;
    __Pyx_INCREF(__pyx_t_2);
    if (unlikely(__pyx_t_2 == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 801; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_6 = PyList_GET_SIZE(__pyx_t_2); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 801; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_7 = ((__pyx_t_6 == 1) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":802
 *                     self.sulfurCount += 1
 *             if len(atom.radicalElectrons) == 1:
 *                 radical = atom.radicalElectrons[0]             # <<<<<<<<<<<<<<
 *                 self.radicalCount += radical
 * 
 */
      if (unlikely(__pyx_v_atom->radicalElectrons == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 802; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_atom->radicalElectrons, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 802; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_9 = __Pyx_PyInt_As_int(__pyx_t_2); if (unlikely((__pyx_t_9 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 802; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_radical = __pyx_t_9;

      /* "rmgpy/molecule/group.py":803
 *             if len(atom.radicalElectrons) == 1:
 *                 radical = atom.radicalElectrons[0]
 *                 self.radicalCount += radical             # <<<<<<<<<<<<<<
 * 
 *     def isIsomorphic(self, other, initialMap=None):
 */
      __pyx_v_self->radicalCount = (__pyx_v_self->radicalCount + __pyx_v_radical);
      goto __pyx_L23;
    }
    __pyx_L23:;

    /* "rmgpy/molecule/group.py":786
 *         self.sulfurCount   = 0
 *         self.radicalCount  = 0
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if len(atom.atomType) == 1:
 *                 atomType   = atom.atomType[0]
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":767
 * 
 * 
 *     def updateFingerprint(self):             # <<<<<<<<<<<<<<
 *         """
 *         Update the molecular fingerprint used to accelerate the subgraph
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.updateFingerprint", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_atom);
  __Pyx_XDECREF((PyObject *)__pyx_v_atomType);
  __Pyx_XDECREF((PyObject *)__pyx_v_carbon);
  __Pyx_XDECREF((PyObject *)__pyx_v_nitrogen);
  __Pyx_XDECREF((PyObject *)__pyx_v_oxygen);
  __Pyx_XDECREF((PyObject *)__pyx_v_sulfur);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_46updateFingerprint[] = "\n        Update the molecular fingerprint used to accelerate the subgraph\n        isomorphism checks.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("updateFingerprint (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_46updateFingerprint(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_46updateFingerprint(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("updateFingerprint", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_updateFingerprint(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 767; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.updateFingerprint", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":805
 *                 self.radicalCount += radical
 * 
 *     def isIsomorphic(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if two graphs are isomorphic and ``False``
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isIsomorphic *__pyx_optional_args) {
  PyObject *__pyx_v_initialMap = ((PyObject*)Py_None);
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_t_8;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isIsomorphic", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_initialMap = __pyx_optional_args->initialMap;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isIsomorphic); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_other));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_other));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
      __Pyx_INCREF(__pyx_v_initialMap);
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_initialMap);
      __Pyx_GIVEREF(__pyx_v_initialMap);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_7 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_7;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":815
 *         # It only makes sense to compare a Group to a Group for full
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # Do the isomorphism comparison
 */
  __pyx_t_7 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_8 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":816
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         # Do the isomorphism comparison
 *         return Graph.isIsomorphic(self, other, initialMap)
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":818
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # Do the isomorphism comparison
 *         return Graph.isIsomorphic(self, other, initialMap)             # <<<<<<<<<<<<<<
 * 
 *     def findIsomorphism(self, other, initialMap=None):
 */
  __pyx_t_9.__pyx_n = 1;
  __pyx_t_9.initialMap = __pyx_v_initialMap;
  __pyx_t_8 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->isIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_9); if (unlikely(__pyx_t_8 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 818; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_t_8;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":805
 *                 self.radicalCount += radical
 * 
 *     def isIsomorphic(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if two graphs are isomorphic and ``False``
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_48isIsomorphic[] = "\n        Returns ``True`` if two graphs are isomorphic and ``False``\n        otherwise. The `initialMap` attribute can be used to specify a required\n        mapping from `self` to `other` (i.e. the atoms of `self` are the keys,\n        while the atoms of `other` are the values). The `other` parameter must\n        be a :class:`Group` object, or a :class:`TypeError` is raised.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other = 0;
  PyObject *__pyx_v_initialMap = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isIsomorphic (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_other,&__pyx_n_s_initialMap,0};
    PyObject* values[2] = {0,0};
    values[1] = ((PyObject*)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_other)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_initialMap);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "isIsomorphic") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)values[0]);
    __pyx_v_initialMap = ((PyObject*)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("isIsomorphic", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_initialMap), (&PyDict_Type), 1, "initialMap", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_48isIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_other, __pyx_v_initialMap);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_48isIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic __pyx_t_2;
  PyObject *__pyx_t_3 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isIsomorphic", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.isIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_2); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 805; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_r = __pyx_t_3;
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":820
 *         return Graph.isIsomorphic(self, other, initialMap)
 * 
 *     def findIsomorphism(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is isomorphic and ``False``
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_findIsomorphism(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findIsomorphism *__pyx_optional_args) {
  PyObject *__pyx_v_initialMap = ((PyObject*)Py_None);
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_t_8;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("findIsomorphism", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_initialMap = __pyx_optional_args->initialMap;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_findIsomorphism); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_other));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_other));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
      __Pyx_INCREF(__pyx_v_initialMap);
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_initialMap);
      __Pyx_GIVEREF(__pyx_v_initialMap);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyList_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":832
 *         # It only makes sense to compare a Group to a Group for full
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # Do the isomorphism comparison
 */
  __pyx_t_7 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_8 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":833
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         # Do the isomorphism comparison
 *         return Graph.findIsomorphism(self, other, initialMap)
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":835
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # Do the isomorphism comparison
 *         return Graph.findIsomorphism(self, other, initialMap)             # <<<<<<<<<<<<<<
 * 
 *     def isSubgraphIsomorphic(self, other, initialMap=None):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_9.__pyx_n = 1;
  __pyx_t_9.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->findIsomorphism(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_9); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 835; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":820
 *         return Graph.isIsomorphic(self, other, initialMap)
 * 
 *     def findIsomorphism(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is isomorphic and ``False``
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findIsomorphism", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_50findIsomorphism[] = "\n        Returns ``True`` if `other` is isomorphic and ``False``\n        otherwise, and the matching mapping. The `initialMap` attribute can be\n        used to specify a required mapping from `self` to `other` (i.e. the\n        atoms of `self` are the keys, while the atoms of `other` are the\n        values). The returned mapping also uses the atoms of `self` for the keys\n        and the atoms of `other` for the values. The `other` parameter must\n        be a :class:`Group` object, or a :class:`TypeError` is raised.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other = 0;
  PyObject *__pyx_v_initialMap = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("findIsomorphism (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_other,&__pyx_n_s_initialMap,0};
    PyObject* values[2] = {0,0};
    values[1] = ((PyObject*)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_other)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_initialMap);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "findIsomorphism") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)values[0]);
    __pyx_v_initialMap = ((PyObject*)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("findIsomorphism", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findIsomorphism", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_initialMap), (&PyDict_Type), 1, "initialMap", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_50findIsomorphism(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_other, __pyx_v_initialMap);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_50findIsomorphism(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("findIsomorphism", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.findIsomorphism(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findIsomorphism", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":837
 *         return Graph.findIsomorphism(self, other, initialMap)
 * 
 *     def isSubgraphIsomorphic(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is subgraph isomorphic and ``False``
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic *__pyx_optional_args) {
  PyObject *__pyx_v_initialMap = ((PyObject*)Py_None);
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_group = 0;
  short __pyx_v_mult1;
  short __pyx_v_mult2;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_t_8;
  short __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSubgraphIsomorphic", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_initialMap = __pyx_optional_args->initialMap;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSubgraphIsomorphic); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_other));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_other));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
      __Pyx_INCREF(__pyx_v_initialMap);
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_initialMap);
      __Pyx_GIVEREF(__pyx_v_initialMap);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_7 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_7;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":850
 *         # It only makes sense to compare a Group to a Group for subgraph
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         group = other
 */
  __pyx_t_7 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_8 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":851
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         group = other
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 851; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 851; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 851; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 851; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 851; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 851; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 851; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 851; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":852
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         group = other             # <<<<<<<<<<<<<<
 * 
 *         if self.multiplicity:
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 852; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_group = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":854
 *         group = other
 * 
 *         if self.multiplicity:             # <<<<<<<<<<<<<<
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 */
  __pyx_t_8 = (__pyx_v_self->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_self->multiplicity) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":855
 * 
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:             # <<<<<<<<<<<<<<
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 */
    if (unlikely(__pyx_v_self->multiplicity == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->multiplicity; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
    for (;;) {
      if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_9 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_9 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_mult1 = __pyx_t_9;

      /* "rmgpy/molecule/group.py":856
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:             # <<<<<<<<<<<<<<
 *                     for mult2 in group.multiplicity:
 *                         if mult1 == mult2: break
 */
      __pyx_t_8 = (__pyx_v_group->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_group->multiplicity) != 0);
      if (__pyx_t_8) {

        /* "rmgpy/molecule/group.py":857
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->multiplicity == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->multiplicity; __Pyx_INCREF(__pyx_t_2); __pyx_t_10 = 0;
        for (;;) {
          if (__pyx_t_10 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_10); __Pyx_INCREF(__pyx_t_4); __pyx_t_10++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_10); __pyx_t_10++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_9 = __Pyx_PyInt_As_short(__pyx_t_4); if (unlikely((__pyx_t_9 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
          __pyx_v_mult2 = __pyx_t_9;

          /* "rmgpy/molecule/group.py":858
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 *                         if mult1 == mult2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return False
 */
          __pyx_t_8 = ((__pyx_v_mult1 == __pyx_v_mult2) != 0);
          if (__pyx_t_8) {
            goto __pyx_L9_break;
          }

          /* "rmgpy/molecule/group.py":857
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":860
 *                         if mult1 == mult2: break
 *                     else:
 *                         return False             # <<<<<<<<<<<<<<
 *         else:
 *             if group.multiplicity: return False
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":857
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        __pyx_L9_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L7;
      }
      __pyx_L7:;

      /* "rmgpy/molecule/group.py":855
 * 
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:             # <<<<<<<<<<<<<<
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L4;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":862
 *                         return False
 *         else:
 *             if group.multiplicity: return False             # <<<<<<<<<<<<<<
 *         # Do the isomorphism comparison
 *         return Graph.isSubgraphIsomorphic(self, other, initialMap)
 */
    __pyx_t_8 = (__pyx_v_group->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_group->multiplicity) != 0);
    if (__pyx_t_8) {
      __pyx_r = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L4:;

  /* "rmgpy/molecule/group.py":864
 *             if group.multiplicity: return False
 *         # Do the isomorphism comparison
 *         return Graph.isSubgraphIsomorphic(self, other, initialMap)             # <<<<<<<<<<<<<<
 * 
 *     def findSubgraphIsomorphisms(self, other, initialMap=None):
 */
  __pyx_t_11.__pyx_n = 1;
  __pyx_t_11.initialMap = __pyx_v_initialMap;
  __pyx_t_8 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->isSubgraphIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_11); if (unlikely(__pyx_t_8 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 864; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_t_8;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":837
 *         return Graph.findIsomorphism(self, other, initialMap)
 * 
 *     def isSubgraphIsomorphic(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is subgraph isomorphic and ``False``
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isSubgraphIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_group);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic[] = "\n        Returns ``True`` if `other` is subgraph isomorphic and ``False``\n        otherwise. In other words, return ``True`` if self is more specific than other.\n        The `initialMap` attribute can be used to specify a required\n        mapping from `self` to `other` (i.e. the atoms of `self` are the keys,\n        while the atoms of `other` are the values). The `other` parameter must\n        be a :class:`Group` object, or a :class:`TypeError` is raised.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other = 0;
  PyObject *__pyx_v_initialMap = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSubgraphIsomorphic (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_other,&__pyx_n_s_initialMap,0};
    PyObject* values[2] = {0,0};
    values[1] = ((PyObject*)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_other)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_initialMap);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "isSubgraphIsomorphic") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)values[0]);
    __pyx_v_initialMap = ((PyObject*)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("isSubgraphIsomorphic", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isSubgraphIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_initialMap), (&PyDict_Type), 1, "initialMap", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_other, __pyx_v_initialMap);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic __pyx_t_2;
  PyObject *__pyx_t_3 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSubgraphIsomorphic", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.isSubgraphIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_2); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_r = __pyx_t_3;
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isSubgraphIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":866
 *         return Graph.isSubgraphIsomorphic(self, other, initialMap)
 * 
 *     def findSubgraphIsomorphisms(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is subgraph isomorphic and ``False``
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms *__pyx_optional_args) {
  PyObject *__pyx_v_initialMap = ((PyObject*)Py_None);
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_group = 0;
  short __pyx_v_mult1;
  short __pyx_v_mult2;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_t_8;
  short __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("findSubgraphIsomorphisms", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_initialMap = __pyx_optional_args->initialMap;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_findSubgraphIsomorphisms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_other));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_other));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
      __Pyx_INCREF(__pyx_v_initialMap);
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_initialMap);
      __Pyx_GIVEREF(__pyx_v_initialMap);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyList_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":883
 *         # It only makes sense to compare a Group to a Group for subgraph
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         group = other
 */
  __pyx_t_7 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_8 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":884
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         group = other
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 884; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 884; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 884; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 884; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 884; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 884; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 884; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 884; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":885
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         group = other             # <<<<<<<<<<<<<<
 * 
 *         if self.multiplicity:
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 885; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_group = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":887
 *         group = other
 * 
 *         if self.multiplicity:             # <<<<<<<<<<<<<<
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 */
  __pyx_t_8 = (__pyx_v_self->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_self->multiplicity) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":888
 * 
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:             # <<<<<<<<<<<<<<
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 */
    if (unlikely(__pyx_v_self->multiplicity == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 888; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->multiplicity; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
    for (;;) {
      if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 888; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 888; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_9 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_9 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 888; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_mult1 = __pyx_t_9;

      /* "rmgpy/molecule/group.py":889
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:             # <<<<<<<<<<<<<<
 *                     for mult2 in group.multiplicity:
 *                         if mult1 == mult2: break
 */
      __pyx_t_8 = (__pyx_v_group->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_group->multiplicity) != 0);
      if (__pyx_t_8) {

        /* "rmgpy/molecule/group.py":890
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->multiplicity == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 890; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->multiplicity; __Pyx_INCREF(__pyx_t_2); __pyx_t_10 = 0;
        for (;;) {
          if (__pyx_t_10 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_10); __Pyx_INCREF(__pyx_t_4); __pyx_t_10++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 890; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_10); __pyx_t_10++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 890; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_9 = __Pyx_PyInt_As_short(__pyx_t_4); if (unlikely((__pyx_t_9 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 890; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
          __pyx_v_mult2 = __pyx_t_9;

          /* "rmgpy/molecule/group.py":891
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 *                         if mult1 == mult2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return []
 */
          __pyx_t_8 = ((__pyx_v_mult1 == __pyx_v_mult2) != 0);
          if (__pyx_t_8) {
            goto __pyx_L9_break;
          }

          /* "rmgpy/molecule/group.py":890
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":893
 *                         if mult1 == mult2: break
 *                     else:
 *                         return []             # <<<<<<<<<<<<<<
 *         else:
 *             if group.multiplicity: return []
 */
          __Pyx_XDECREF(__pyx_r);
          __pyx_t_4 = PyList_New(0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 893; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_4);
          __pyx_r = ((PyObject*)__pyx_t_4);
          __pyx_t_4 = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":890
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        __pyx_L9_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L7;
      }
      __pyx_L7:;

      /* "rmgpy/molecule/group.py":888
 * 
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:             # <<<<<<<<<<<<<<
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L4;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":895
 *                         return []
 *         else:
 *             if group.multiplicity: return []             # <<<<<<<<<<<<<<
 * 
 *         # Do the isomorphism comparison
 */
    __pyx_t_8 = (__pyx_v_group->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_group->multiplicity) != 0);
    if (__pyx_t_8) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 895; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_r = ((PyObject*)__pyx_t_1);
      __pyx_t_1 = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L4:;

  /* "rmgpy/molecule/group.py":898
 * 
 *         # Do the isomorphism comparison
 *         return Graph.findSubgraphIsomorphisms(self, other, initialMap)             # <<<<<<<<<<<<<<
 * 
 *     def isIdentical(self, other):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_11.__pyx_n = 1;
  __pyx_t_11.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->findSubgraphIsomorphisms(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_11); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 898; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":866
 *         return Graph.isSubgraphIsomorphic(self, other, initialMap)
 * 
 *     def findSubgraphIsomorphisms(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is subgraph isomorphic and ``False``
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findSubgraphIsomorphisms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_group);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms[] = "\n        Returns ``True`` if `other` is subgraph isomorphic and ``False``\n        otherwise. In other words, return ``True`` is self is more specific than other.\n        Also returns the lists all of valid mappings. The\n        `initialMap` attribute can be used to specify a required mapping from\n        `self` to `other` (i.e. the atoms of `self` are the keys, while the\n        atoms of `other` are the values). The returned mappings also use the\n        atoms of `self` for the keys and the atoms of `other` for the values.\n        The `other` parameter must be a :class:`Group` object, or a\n        :class:`TypeError` is raised.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other = 0;
  PyObject *__pyx_v_initialMap = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("findSubgraphIsomorphisms (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_other,&__pyx_n_s_initialMap,0};
    PyObject* values[2] = {0,0};
    values[1] = ((PyObject*)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_other)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_initialMap);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "findSubgraphIsomorphisms") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)values[0]);
    __pyx_v_initialMap = ((PyObject*)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("findSubgraphIsomorphisms", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findSubgraphIsomorphisms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_initialMap), (&PyDict_Type), 1, "initialMap", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_other, __pyx_v_initialMap);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("findSubgraphIsomorphisms", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.findSubgraphIsomorphisms(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findSubgraphIsomorphisms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":900
 *         return Graph.findSubgraphIsomorphisms(self, other, initialMap)
 * 
 *     def isIdentical(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is identical and ``False`` otherwise.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isIdentical(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isIdentical", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isIdentical); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 900; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 900; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 900; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 900; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 900; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":909
 *         # It only makes sense to compare a Group to a Group for full
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # An identical group is always a child of itself and
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":910
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         # An identical group is always a child of itself and
 *         # is the only case where that is true. Therefore
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 910; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 910; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 910; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 910; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 910; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 910; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 910; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 910; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":915
 *         # if we do both directions of isSubgraphIsmorphic, we need
 *         # to get True twice for it to be identical
 *         if not self.isSubgraphIsomorphic(other):             # <<<<<<<<<<<<<<
 *             return False
 *         elif not other.isSubgraphIsomorphic(self):
 */
  __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.isSubgraphIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 0, NULL); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 915; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_6 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":916
 *         # to get True twice for it to be identical
 *         if not self.isSubgraphIsomorphic(other):
 *             return False             # <<<<<<<<<<<<<<
 *         elif not other.isSubgraphIsomorphic(self):
 *             return False
 */
    __pyx_r = 0;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":917
 *         if not self.isSubgraphIsomorphic(other):
 *             return False
 *         elif not other.isSubgraphIsomorphic(self):             # <<<<<<<<<<<<<<
 *             return False
 *         else:
 */
  __pyx_t_6 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph *)__pyx_v_other->__pyx_vtab)->isSubgraphIsomorphic(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 0, NULL); if (unlikely(__pyx_t_6 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 917; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":918
 *             return False
 *         elif not other.isSubgraphIsomorphic(self):
 *             return False             # <<<<<<<<<<<<<<
 *         else:
 *             return True
 */
    __pyx_r = 0;
    goto __pyx_L0;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":920
 *             return False
 *         else:
 *             return True             # <<<<<<<<<<<<<<
 * 
 *     def standardizeAtomType(self):
 */
    __pyx_r = 1;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":900
 *         return Graph.findSubgraphIsomorphisms(self, other, initialMap)
 * 
 *     def isIdentical(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is identical and ``False`` otherwise.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.isIdentical", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_56isIdentical[] = "\n        Returns ``True`` if `other` is identical and ``False`` otherwise.\n        The function `isIsomorphic` respects wildcards, while this function\n        does not, make it more useful for checking groups to groups (as\n        opposed to molecules to groups)\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isIdentical (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 900; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_56isIdentical(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_56isIdentical(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isIdentical", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_isIdentical(__pyx_v_self, __pyx_v_other, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 900; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isIdentical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":922
 *             return True
 * 
 *     def standardizeAtomType(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function changes the atomTypes in a group if the atom must
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_59standardizeAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_standardizeAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_v_modified;
  PyObject *__pyx_v_valency = NULL;
  PyObject *__pyx_v_specifics = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_index = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_claimedAtomType = NULL;
  PyObject *__pyx_v_newAtomType = NULL;
  PyObject *__pyx_v_element = NULL;
  PyObject *__pyx_v_elementLabel = NULL;
  PyObject *__pyx_v_groupFeatures = NULL;
  PyObject *__pyx_v_single = NULL;
  PyObject *__pyx_v_allDouble = NULL;
  PyObject *__pyx_v_triple = NULL;
  PyObject *__pyx_v_benzene = NULL;
  PyObject *__pyx_v_bondValency = NULL;
  PyObject *__pyx_v_filledValency = NULL;
  PyObject *__pyx_v_specificAtomType = NULL;
  PyObject *__pyx_v_atomtypeFeatureList = NULL;
  PyObject *__pyx_v_molFeature = NULL;
  PyObject *__pyx_v_atomtypeFeature = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  PyObject *(*__pyx_t_7)(PyObject *);
  int __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  PyObject *__pyx_t_10 = NULL;
  PyObject *__pyx_t_11 = NULL;
  int __pyx_t_12;
  PyObject *__pyx_t_13 = NULL;
  PyObject *(*__pyx_t_14)(PyObject *);
  Py_ssize_t __pyx_t_15;
  PyObject *(*__pyx_t_16)(PyObject *);
  PyObject *__pyx_t_17 = NULL;
  PyObject *(*__pyx_t_18)(PyObject *);
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("standardizeAtomType", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_standardizeAtomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 922; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_59standardizeAtomType)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 922; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 922; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 922; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":940
 *         """
 * 
 *         modified = False             # <<<<<<<<<<<<<<
 * 
 *         #dictionary of element to expected valency
 */
  __pyx_v_modified = 0;

  /* "rmgpy/molecule/group.py":943
 * 
 *         #dictionary of element to expected valency
 *         valency = {atomTypes['C'] : 4,             # <<<<<<<<<<<<<<
 *                    atomTypes['O'] : 2,
 *                    atomTypes['S']: 2,
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 943; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 943; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_C); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 943; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  if (PyDict_SetItem(__pyx_t_1, __pyx_t_3, __pyx_int_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 943; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":944
 *         #dictionary of element to expected valency
 *         valency = {atomTypes['C'] : 4,
 *                    atomTypes['O'] : 2,             # <<<<<<<<<<<<<<
 *                    atomTypes['S']: 2,
 *                    atomTypes['Si']:4
 */
  __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 944; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_O); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 944; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  if (PyDict_SetItem(__pyx_t_1, __pyx_t_2, __pyx_int_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 943; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":945
 *         valency = {atomTypes['C'] : 4,
 *                    atomTypes['O'] : 2,
 *                    atomTypes['S']: 2,             # <<<<<<<<<<<<<<
 *                    atomTypes['Si']:4
 *                    }
 */
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 945; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_S); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 945; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  if (PyDict_SetItem(__pyx_t_1, __pyx_t_3, __pyx_int_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 943; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":946
 *                    atomTypes['O'] : 2,
 *                    atomTypes['S']: 2,
 *                    atomTypes['Si']:4             # <<<<<<<<<<<<<<
 *                    }
 * 
 */
  __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_Si); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  if (PyDict_SetItem(__pyx_t_1, __pyx_t_2, __pyx_int_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 943; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_valency = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":950
 * 
 *         #list of :class:AtomType which are elements with more sub-divided atomtypes beneath them
 *         specifics= [elementLabel for elementLabel in allElements if elementLabel not in nonSpecifics]             # <<<<<<<<<<<<<<
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_allElements); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_3 = __pyx_t_2; __Pyx_INCREF(__pyx_t_3); __pyx_t_6 = 0;
    __pyx_t_7 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_7 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_7)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_7(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_elementLabel, __pyx_t_2);
    __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_nonSpecifics); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_elementLabel, __pyx_t_2, Py_NE)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_8 = (__pyx_t_5 != 0);
    if (__pyx_t_8) {
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_v_elementLabel))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L5;
    }
    __pyx_L5:;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_v_specifics = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":951
 *         #list of :class:AtomType which are elements with more sub-divided atomtypes beneath them
 *         specifics= [elementLabel for elementLabel in allElements if elementLabel not in nonSpecifics]
 *         for index, atom in enumerate(self.atoms):             # <<<<<<<<<<<<<<
 *             claimedAtomType = atom.atomType[0]
 *             newAtomType = None
 */
  __Pyx_INCREF(__pyx_int_0);
  __pyx_t_1 = __pyx_int_0;
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (likely(PyList_CheckExact(__pyx_t_3)) || PyTuple_CheckExact(__pyx_t_3)) {
    __pyx_t_2 = __pyx_t_3; __Pyx_INCREF(__pyx_t_2); __pyx_t_6 = 0;
    __pyx_t_7 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_7 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  for (;;) {
    if (likely(!__pyx_t_7)) {
      if (likely(PyList_CheckExact(__pyx_t_2))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_3); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_3); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_3 = __pyx_t_7(__pyx_t_2);
      if (unlikely(!__pyx_t_3)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_3);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_3);
    __pyx_t_3 = 0;
    __Pyx_INCREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_1);
    __pyx_t_3 = PyNumber_Add(__pyx_t_1, __pyx_int_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_1);
    __pyx_t_1 = __pyx_t_3;
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":952
 *         specifics= [elementLabel for elementLabel in allElements if elementLabel not in nonSpecifics]
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]             # <<<<<<<<<<<<<<
 *             newAtomType = None
 *             element = None
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 952; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = __Pyx_GetItemInt(__pyx_t_3, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 952; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_claimedAtomType, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":953
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]
 *             newAtomType = None             # <<<<<<<<<<<<<<
 *             element = None
 *             #Ignore elements that do not have more than one atomtype
 */
    __Pyx_INCREF(Py_None);
    __Pyx_XDECREF_SET(__pyx_v_newAtomType, Py_None);

    /* "rmgpy/molecule/group.py":954
 *             claimedAtomType = atom.atomType[0]
 *             newAtomType = None
 *             element = None             # <<<<<<<<<<<<<<
 *             #Ignore elements that do not have more than one atomtype
 *             if claimedAtomType.label in nonSpecifics: continue
 */
    __Pyx_INCREF(Py_None);
    __Pyx_XDECREF_SET(__pyx_v_element, Py_None);

    /* "rmgpy/molecule/group.py":956
 *             element = None
 *             #Ignore elements that do not have more than one atomtype
 *             if claimedAtomType.label in nonSpecifics: continue             # <<<<<<<<<<<<<<
 *             for elementLabel in specifics:
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 */
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_claimedAtomType, __pyx_n_s_label); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 956; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_nonSpecifics); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 956; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_8 = (__Pyx_PySequence_Contains(__pyx_t_4, __pyx_t_3, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 956; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_5 = (__pyx_t_8 != 0);
    if (__pyx_t_5) {
      goto __pyx_L6_continue;
    }

    /* "rmgpy/molecule/group.py":957
 *             #Ignore elements that do not have more than one atomtype
 *             if claimedAtomType.label in nonSpecifics: continue
 *             for elementLabel in specifics:             # <<<<<<<<<<<<<<
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 *                     element = atomTypes[elementLabel]
 */
    __pyx_t_3 = __pyx_v_specifics; __Pyx_INCREF(__pyx_t_3); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_3)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_4 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_9); __Pyx_INCREF(__pyx_t_4); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 957; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_4 = PySequence_ITEM(__pyx_t_3, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 957; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_elementLabel, __pyx_t_4);
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":958
 *             if claimedAtomType.label in nonSpecifics: continue
 *             for elementLabel in specifics:
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:             # <<<<<<<<<<<<<<
 *                     element = atomTypes[elementLabel]
 *                     break
 */
      __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_claimedAtomType, __pyx_n_s_label); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_10 = PyObject_RichCompare(__pyx_t_4, __pyx_v_elementLabel, Py_EQ); __Pyx_XGOTREF(__pyx_t_10); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_t_10); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      if (!__pyx_t_8) {
        goto __pyx_L13_next_or;
      } else {
        __pyx_t_5 = __pyx_t_8;
        goto __pyx_L12_bool_binop_done;
      }
      __pyx_L13_next_or:;
      __pyx_t_10 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_10);
      __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_claimedAtomType, __pyx_n_s_label); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_11 = PyObject_GetItem(__pyx_t_10, __pyx_t_4); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_11);
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_10 = PyObject_GetItem(__pyx_t_4, __pyx_v_elementLabel); if (unlikely(__pyx_t_10 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_10);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_t_10, __pyx_n_s_specific); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      __pyx_t_8 = (__Pyx_PySequence_Contains(__pyx_t_11, __pyx_t_4, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 958; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_12 = (__pyx_t_8 != 0);
      __pyx_t_5 = __pyx_t_12;
      __pyx_L12_bool_binop_done:;
      if (__pyx_t_5) {

        /* "rmgpy/molecule/group.py":959
 *             for elementLabel in specifics:
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 *                     element = atomTypes[elementLabel]             # <<<<<<<<<<<<<<
 *                     break
 * 
 */
        __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 959; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_11 = PyObject_GetItem(__pyx_t_4, __pyx_v_elementLabel); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 959; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_11);
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_DECREF_SET(__pyx_v_element, __pyx_t_11);
        __pyx_t_11 = 0;

        /* "rmgpy/molecule/group.py":960
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 *                     element = atomTypes[elementLabel]
 *                     break             # <<<<<<<<<<<<<<
 * 
 *             #claimedAtomType is not in one of the specified elements
 */
        goto __pyx_L10_break;
      }

      /* "rmgpy/molecule/group.py":957
 *             #Ignore elements that do not have more than one atomtype
 *             if claimedAtomType.label in nonSpecifics: continue
 *             for elementLabel in specifics:             # <<<<<<<<<<<<<<
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 *                     element = atomTypes[elementLabel]
 */
    }
    __pyx_L10_break:;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":963
 * 
 *             #claimedAtomType is not in one of the specified elements
 *             if not element: continue             # <<<<<<<<<<<<<<
 *             #Don't standardize atomtypes for nitrogen for now. My feeling is that
 *             # the work on the nitrogen atomtypes is still incomplete
 */
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_v_element); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_12 = ((!__pyx_t_5) != 0);
    if (__pyx_t_12) {
      goto __pyx_L6_continue;
    }

    /* "rmgpy/molecule/group.py":966
 *             #Don't standardize atomtypes for nitrogen for now. My feeling is that
 *             # the work on the nitrogen atomtypes is still incomplete
 *             elif element is atomTypes['N']: continue             # <<<<<<<<<<<<<<
 * 
 *             groupFeatures = getFeatures(atom, atom.bonds)
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 966; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_11 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_N); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 966; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_11);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_12 = (__pyx_v_element == __pyx_t_11);
    __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
    __pyx_t_5 = (__pyx_t_12 != 0);
    if (__pyx_t_5) {
      goto __pyx_L6_continue;
    }

    /* "rmgpy/molecule/group.py":968
 *             elif element is atomTypes['N']: continue
 * 
 *             groupFeatures = getFeatures(atom, atom.bonds)             # <<<<<<<<<<<<<<
 * 
 *             single = groupFeatures[0]
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_getFeatures); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 968; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 968; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_10 = NULL;
    __pyx_t_9 = 0;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_10)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_10);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
        __pyx_t_9 = 1;
      }
    }
    __pyx_t_13 = PyTuple_New(2+__pyx_t_9); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 968; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_13);
    if (__pyx_t_10) {
      PyTuple_SET_ITEM(__pyx_t_13, 0, __pyx_t_10); __Pyx_GIVEREF(__pyx_t_10); __pyx_t_10 = NULL;
    }
    __Pyx_INCREF(__pyx_v_atom);
    PyTuple_SET_ITEM(__pyx_t_13, 0+__pyx_t_9, __pyx_v_atom);
    __Pyx_GIVEREF(__pyx_v_atom);
    PyTuple_SET_ITEM(__pyx_t_13, 1+__pyx_t_9, __pyx_t_4);
    __Pyx_GIVEREF(__pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_11 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_13, NULL); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 968; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_11);
    __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_groupFeatures, __pyx_t_11);
    __pyx_t_11 = 0;

    /* "rmgpy/molecule/group.py":970
 *             groupFeatures = getFeatures(atom, atom.bonds)
 * 
 *             single = groupFeatures[0]             # <<<<<<<<<<<<<<
 *             allDouble = groupFeatures[1]
 *             triple = groupFeatures[5]
 */
    __pyx_t_11 = __Pyx_GetItemInt(__pyx_v_groupFeatures, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 970; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_11);
    __Pyx_XDECREF_SET(__pyx_v_single, __pyx_t_11);
    __pyx_t_11 = 0;

    /* "rmgpy/molecule/group.py":971
 * 
 *             single = groupFeatures[0]
 *             allDouble = groupFeatures[1]             # <<<<<<<<<<<<<<
 *             triple = groupFeatures[5]
 *             benzene = groupFeatures[6]
 */
    __pyx_t_11 = __Pyx_GetItemInt(__pyx_v_groupFeatures, 1, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 971; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_11);
    __Pyx_XDECREF_SET(__pyx_v_allDouble, __pyx_t_11);
    __pyx_t_11 = 0;

    /* "rmgpy/molecule/group.py":972
 *             single = groupFeatures[0]
 *             allDouble = groupFeatures[1]
 *             triple = groupFeatures[5]             # <<<<<<<<<<<<<<
 *             benzene = groupFeatures[6]
 *             if benzene == 3:
 */
    __pyx_t_11 = __Pyx_GetItemInt(__pyx_v_groupFeatures, 5, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 972; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_11);
    __Pyx_XDECREF_SET(__pyx_v_triple, __pyx_t_11);
    __pyx_t_11 = 0;

    /* "rmgpy/molecule/group.py":973
 *             allDouble = groupFeatures[1]
 *             triple = groupFeatures[5]
 *             benzene = groupFeatures[6]             # <<<<<<<<<<<<<<
 *             if benzene == 3:
 *                 bondValency = single + 2 * allDouble + 3 * triple + 4.0/3.0 * benzene
 */
    __pyx_t_11 = __Pyx_GetItemInt(__pyx_v_groupFeatures, 6, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 973; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_11);
    __Pyx_XDECREF_SET(__pyx_v_benzene, __pyx_t_11);
    __pyx_t_11 = 0;

    /* "rmgpy/molecule/group.py":974
 *             triple = groupFeatures[5]
 *             benzene = groupFeatures[6]
 *             if benzene == 3:             # <<<<<<<<<<<<<<
 *                 bondValency = single + 2 * allDouble + 3 * triple + 4.0/3.0 * benzene
 *             else:
 */
    __pyx_t_11 = PyObject_RichCompare(__pyx_v_benzene, __pyx_int_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_11); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 974; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_11); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 974; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":975
 *             benzene = groupFeatures[6]
 *             if benzene == 3:
 *                 bondValency = single + 2 * allDouble + 3 * triple + 4.0/3.0 * benzene             # <<<<<<<<<<<<<<
 *             else:
 *                 bondValency =  single + 2 * allDouble + 3 * triple + 3.0/2.0 * benzene
 */
      __pyx_t_11 = PyNumber_Multiply(__pyx_int_2, __pyx_v_allDouble); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 975; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_3 = PyNumber_Add(__pyx_v_single, __pyx_t_11); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 975; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __pyx_t_11 = PyNumber_Multiply(__pyx_int_3, __pyx_v_triple); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 975; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_13 = PyNumber_Add(__pyx_t_3, __pyx_t_11); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 975; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_13);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __pyx_t_11 = PyFloat_FromDouble((4.0 / 3.0)); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 975; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_3 = PyNumber_Multiply(__pyx_t_11, __pyx_v_benzene); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 975; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __pyx_t_11 = PyNumber_Add(__pyx_t_13, __pyx_t_3); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 975; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_bondValency, __pyx_t_11);
      __pyx_t_11 = 0;
      goto __pyx_L15;
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":977
 *                 bondValency = single + 2 * allDouble + 3 * triple + 4.0/3.0 * benzene
 *             else:
 *                 bondValency =  single + 2 * allDouble + 3 * triple + 3.0/2.0 * benzene             # <<<<<<<<<<<<<<
 *             filledValency =  atom.radicalElectrons[0] + bondValency
 * 
 */
      __pyx_t_11 = PyNumber_Multiply(__pyx_int_2, __pyx_v_allDouble); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 977; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_3 = PyNumber_Add(__pyx_v_single, __pyx_t_11); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 977; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __pyx_t_11 = PyNumber_Multiply(__pyx_int_3, __pyx_v_triple); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 977; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_13 = PyNumber_Add(__pyx_t_3, __pyx_t_11); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 977; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_13);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __pyx_t_11 = PyFloat_FromDouble((3.0 / 2.0)); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 977; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_3 = PyNumber_Multiply(__pyx_t_11, __pyx_v_benzene); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 977; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __pyx_t_11 = PyNumber_Add(__pyx_t_13, __pyx_t_3); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 977; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_bondValency, __pyx_t_11);
      __pyx_t_11 = 0;
    }
    __pyx_L15:;

    /* "rmgpy/molecule/group.py":978
 *             else:
 *                 bondValency =  single + 2 * allDouble + 3 * triple + 3.0/2.0 * benzene
 *             filledValency =  atom.radicalElectrons[0] + bondValency             # <<<<<<<<<<<<<<
 * 
 *             #For an atomtype to be known for certain, the valency must be filled
 */
    __pyx_t_11 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_radicalElectrons); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 978; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_11);
    __pyx_t_3 = __Pyx_GetItemInt(__pyx_t_11, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 978; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
    __pyx_t_11 = PyNumber_Add(__pyx_t_3, __pyx_v_bondValency); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 978; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_11);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_filledValency, __pyx_t_11);
    __pyx_t_11 = 0;

    /* "rmgpy/molecule/group.py":982
 *             #For an atomtype to be known for certain, the valency must be filled
 *             #within 1 of the total valency available
 *             if filledValency >= valency[element] - 1:             # <<<<<<<<<<<<<<
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 */
    __pyx_t_11 = __Pyx_PyDict_GetItem(__pyx_v_valency, __pyx_v_element); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 982; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_11);
    __pyx_t_3 = PyNumber_Subtract(__pyx_t_11, __pyx_int_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 982; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
    __pyx_t_11 = PyObject_RichCompare(__pyx_v_filledValency, __pyx_t_3, Py_GE); __Pyx_XGOTREF(__pyx_t_11); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 982; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_11); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 982; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":983
 *             #within 1 of the total valency available
 *             if filledValency >= valency[element] - 1:
 *                 for specificAtomType in element.specific:             # <<<<<<<<<<<<<<
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 */
      __pyx_t_11 = __Pyx_PyObject_GetAttrStr(__pyx_v_element, __pyx_n_s_specific); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 983; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      if (likely(PyList_CheckExact(__pyx_t_11)) || PyTuple_CheckExact(__pyx_t_11)) {
        __pyx_t_3 = __pyx_t_11; __Pyx_INCREF(__pyx_t_3); __pyx_t_9 = 0;
        __pyx_t_14 = NULL;
      } else {
        __pyx_t_9 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_11); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 983; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_14 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_14)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 983; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      for (;;) {
        if (likely(!__pyx_t_14)) {
          if (likely(PyList_CheckExact(__pyx_t_3))) {
            if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_3)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_11 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_9); __Pyx_INCREF(__pyx_t_11); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 983; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_11 = PySequence_ITEM(__pyx_t_3, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 983; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          } else {
            if (__pyx_t_9 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_11 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_9); __Pyx_INCREF(__pyx_t_11); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 983; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_11 = PySequence_ITEM(__pyx_t_3, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 983; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          }
        } else {
          __pyx_t_11 = __pyx_t_14(__pyx_t_3);
          if (unlikely(!__pyx_t_11)) {
            PyObject* exc_type = PyErr_Occurred();
            if (exc_type) {
              if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
              else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 983; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            break;
          }
          __Pyx_GOTREF(__pyx_t_11);
        }
        __Pyx_XDECREF_SET(__pyx_v_specificAtomType, __pyx_t_11);
        __pyx_t_11 = 0;

        /* "rmgpy/molecule/group.py":984
 *             if filledValency >= valency[element] - 1:
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()             # <<<<<<<<<<<<<<
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 *                         if atomtypeFeature == []:
 */
        __pyx_t_13 = __Pyx_PyObject_GetAttrStr(__pyx_v_specificAtomType, __pyx_n_s_getFeatures); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 984; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_13);
        __pyx_t_4 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_13))) {
          __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_13);
          if (likely(__pyx_t_4)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_13);
            __Pyx_INCREF(__pyx_t_4);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_13, function);
          }
        }
        if (__pyx_t_4) {
          __pyx_t_11 = __Pyx_PyObject_CallOneArg(__pyx_t_13, __pyx_t_4); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 984; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        } else {
          __pyx_t_11 = __Pyx_PyObject_CallNoArg(__pyx_t_13); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 984; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_11);
        __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
        __Pyx_XDECREF_SET(__pyx_v_atomtypeFeatureList, __pyx_t_11);
        __pyx_t_11 = 0;

        /* "rmgpy/molecule/group.py":985
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):             # <<<<<<<<<<<<<<
 *                         if atomtypeFeature == []:
 *                             continue
 */
        __pyx_t_11 = PyTuple_New(2); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        __Pyx_INCREF(__pyx_v_groupFeatures);
        PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_v_groupFeatures);
        __Pyx_GIVEREF(__pyx_v_groupFeatures);
        __Pyx_INCREF(__pyx_v_atomtypeFeatureList);
        PyTuple_SET_ITEM(__pyx_t_11, 1, __pyx_v_atomtypeFeatureList);
        __Pyx_GIVEREF(__pyx_v_atomtypeFeatureList);
        __pyx_t_13 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_11, NULL); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_13);
        __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
        if (likely(PyList_CheckExact(__pyx_t_13)) || PyTuple_CheckExact(__pyx_t_13)) {
          __pyx_t_11 = __pyx_t_13; __Pyx_INCREF(__pyx_t_11); __pyx_t_15 = 0;
          __pyx_t_16 = NULL;
        } else {
          __pyx_t_15 = -1; __pyx_t_11 = PyObject_GetIter(__pyx_t_13); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_11);
          __pyx_t_16 = Py_TYPE(__pyx_t_11)->tp_iternext; if (unlikely(!__pyx_t_16)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
        for (;;) {
          if (likely(!__pyx_t_16)) {
            if (likely(PyList_CheckExact(__pyx_t_11))) {
              if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_11)) break;
              #if CYTHON_COMPILING_IN_CPYTHON
              __pyx_t_13 = PyList_GET_ITEM(__pyx_t_11, __pyx_t_15); __Pyx_INCREF(__pyx_t_13); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              #else
              __pyx_t_13 = PySequence_ITEM(__pyx_t_11, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              #endif
            } else {
              if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_11)) break;
              #if CYTHON_COMPILING_IN_CPYTHON
              __pyx_t_13 = PyTuple_GET_ITEM(__pyx_t_11, __pyx_t_15); __Pyx_INCREF(__pyx_t_13); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              #else
              __pyx_t_13 = PySequence_ITEM(__pyx_t_11, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              #endif
            }
          } else {
            __pyx_t_13 = __pyx_t_16(__pyx_t_11);
            if (unlikely(!__pyx_t_13)) {
              PyObject* exc_type = PyErr_Occurred();
              if (exc_type) {
                if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
                else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              }
              break;
            }
            __Pyx_GOTREF(__pyx_t_13);
          }
          if ((likely(PyTuple_CheckExact(__pyx_t_13))) || (PyList_CheckExact(__pyx_t_13))) {
            PyObject* sequence = __pyx_t_13;
            #if CYTHON_COMPILING_IN_CPYTHON
            Py_ssize_t size = Py_SIZE(sequence);
            #else
            Py_ssize_t size = PySequence_Size(sequence);
            #endif
            if (unlikely(size != 2)) {
              if (size > 2) __Pyx_RaiseTooManyValuesError(2);
              else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
              {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            #if CYTHON_COMPILING_IN_CPYTHON
            if (likely(PyTuple_CheckExact(sequence))) {
              __pyx_t_4 = PyTuple_GET_ITEM(sequence, 0); 
              __pyx_t_10 = PyTuple_GET_ITEM(sequence, 1); 
            } else {
              __pyx_t_4 = PyList_GET_ITEM(sequence, 0); 
              __pyx_t_10 = PyList_GET_ITEM(sequence, 1); 
            }
            __Pyx_INCREF(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_10);
            #else
            __pyx_t_4 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_4);
            __pyx_t_10 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_10);
            #endif
            __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
          } else {
            Py_ssize_t index = -1;
            __pyx_t_17 = PyObject_GetIter(__pyx_t_13); if (unlikely(!__pyx_t_17)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_17);
            __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
            __pyx_t_18 = Py_TYPE(__pyx_t_17)->tp_iternext;
            index = 0; __pyx_t_4 = __pyx_t_18(__pyx_t_17); if (unlikely(!__pyx_t_4)) goto __pyx_L21_unpacking_failed;
            __Pyx_GOTREF(__pyx_t_4);
            index = 1; __pyx_t_10 = __pyx_t_18(__pyx_t_17); if (unlikely(!__pyx_t_10)) goto __pyx_L21_unpacking_failed;
            __Pyx_GOTREF(__pyx_t_10);
            if (__Pyx_IternextUnpackEndCheck(__pyx_t_18(__pyx_t_17), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __pyx_t_18 = NULL;
            __Pyx_DECREF(__pyx_t_17); __pyx_t_17 = 0;
            goto __pyx_L22_unpacking_done;
            __pyx_L21_unpacking_failed:;
            __Pyx_DECREF(__pyx_t_17); __pyx_t_17 = 0;
            __pyx_t_18 = NULL;
            if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
            {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __pyx_L22_unpacking_done:;
          }
          __Pyx_XDECREF_SET(__pyx_v_molFeature, __pyx_t_4);
          __pyx_t_4 = 0;
          __Pyx_XDECREF_SET(__pyx_v_atomtypeFeature, __pyx_t_10);
          __pyx_t_10 = 0;

          /* "rmgpy/molecule/group.py":986
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 *                         if atomtypeFeature == []:             # <<<<<<<<<<<<<<
 *                             continue
 *                         elif molFeature not in atomtypeFeature:
 */
          __pyx_t_13 = PyList_New(0); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 986; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_13);
          __pyx_t_10 = PyObject_RichCompare(__pyx_v_atomtypeFeature, __pyx_t_13, Py_EQ); __Pyx_XGOTREF(__pyx_t_10); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 986; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
          __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_10); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 986; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
          if (__pyx_t_5) {

            /* "rmgpy/molecule/group.py":987
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 *                         if atomtypeFeature == []:
 *                             continue             # <<<<<<<<<<<<<<
 *                         elif molFeature not in atomtypeFeature:
 *                             break
 */
            goto __pyx_L19_continue;
          }

          /* "rmgpy/molecule/group.py":988
 *                         if atomtypeFeature == []:
 *                             continue
 *                         elif molFeature not in atomtypeFeature:             # <<<<<<<<<<<<<<
 *                             break
 *                     else:
 */
          __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_molFeature, __pyx_v_atomtypeFeature, Py_NE)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 988; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __pyx_t_12 = (__pyx_t_5 != 0);
          if (__pyx_t_12) {

            /* "rmgpy/molecule/group.py":989
 *                             continue
 *                         elif molFeature not in atomtypeFeature:
 *                             break             # <<<<<<<<<<<<<<
 *                     else:
 *                         if specificAtomType is atomTypes['Oa'] or specificAtomType is atomTypes['Sa']:
 */
            goto __pyx_L20_break;
          }

          /* "rmgpy/molecule/group.py":985
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):             # <<<<<<<<<<<<<<
 *                         if atomtypeFeature == []:
 *                             continue
 */
          __pyx_L19_continue:;
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":991
 *                             break
 *                     else:
 *                         if specificAtomType is atomTypes['Oa'] or specificAtomType is atomTypes['Sa']:             # <<<<<<<<<<<<<<
 *                             if atom.lonePairs == 3 or atom.radicalElectrons == 2:
 *                                 newAtomType = specificAtomType
 */
          __pyx_t_10 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 991; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_10);
          __pyx_t_13 = PyObject_GetItem(__pyx_t_10, __pyx_n_s_Oa); if (unlikely(__pyx_t_13 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 991; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_13);
          __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
          __pyx_t_5 = (__pyx_v_specificAtomType == __pyx_t_13);
          __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
          __pyx_t_8 = (__pyx_t_5 != 0);
          if (!__pyx_t_8) {
            goto __pyx_L27_next_or;
          } else {
            __pyx_t_12 = __pyx_t_8;
            goto __pyx_L26_bool_binop_done;
          }
          __pyx_L27_next_or:;
          __pyx_t_13 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 991; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_13);
          __pyx_t_10 = PyObject_GetItem(__pyx_t_13, __pyx_n_s_Sa); if (unlikely(__pyx_t_10 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 991; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_10);
          __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
          __pyx_t_8 = (__pyx_v_specificAtomType == __pyx_t_10);
          __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
          __pyx_t_5 = (__pyx_t_8 != 0);
          __pyx_t_12 = __pyx_t_5;
          __pyx_L26_bool_binop_done:;
          if (__pyx_t_12) {

            /* "rmgpy/molecule/group.py":992
 *                     else:
 *                         if specificAtomType is atomTypes['Oa'] or specificAtomType is atomTypes['Sa']:
 *                             if atom.lonePairs == 3 or atom.radicalElectrons == 2:             # <<<<<<<<<<<<<<
 *                                 newAtomType = specificAtomType
 *                                 break
 */
            __pyx_t_10 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_lonePairs); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 992; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_10);
            __pyx_t_13 = PyObject_RichCompare(__pyx_t_10, __pyx_int_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_13); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 992; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
            __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_13); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 992; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
            if (!__pyx_t_5) {
              goto __pyx_L30_next_or;
            } else {
              __pyx_t_12 = __pyx_t_5;
              goto __pyx_L29_bool_binop_done;
            }
            __pyx_L30_next_or:;
            __pyx_t_13 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_radicalElectrons); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 992; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_13);
            __pyx_t_10 = PyObject_RichCompare(__pyx_t_13, __pyx_int_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_10); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 992; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
            __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_10); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 992; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
            __pyx_t_12 = __pyx_t_5;
            __pyx_L29_bool_binop_done:;
            if (__pyx_t_12) {

              /* "rmgpy/molecule/group.py":993
 *                         if specificAtomType is atomTypes['Oa'] or specificAtomType is atomTypes['Sa']:
 *                             if atom.lonePairs == 3 or atom.radicalElectrons == 2:
 *                                 newAtomType = specificAtomType             # <<<<<<<<<<<<<<
 *                                 break
 *                         else:
 */
              __Pyx_INCREF(__pyx_v_specificAtomType);
              __Pyx_DECREF_SET(__pyx_v_newAtomType, __pyx_v_specificAtomType);

              /* "rmgpy/molecule/group.py":994
 *                             if atom.lonePairs == 3 or atom.radicalElectrons == 2:
 *                                 newAtomType = specificAtomType
 *                                 break             # <<<<<<<<<<<<<<
 *                         else:
 *                             newAtomType = specificAtomType
 */
              goto __pyx_L18_break;
            }
            goto __pyx_L25;
          }
          /*else*/ {

            /* "rmgpy/molecule/group.py":996
 *                                 break
 *                         else:
 *                             newAtomType = specificAtomType             # <<<<<<<<<<<<<<
 *                             break
 * 
 */
            __Pyx_INCREF(__pyx_v_specificAtomType);
            __Pyx_DECREF_SET(__pyx_v_newAtomType, __pyx_v_specificAtomType);

            /* "rmgpy/molecule/group.py":997
 *                         else:
 *                             newAtomType = specificAtomType
 *                             break             # <<<<<<<<<<<<<<
 * 
 *             #set the new atom type if the algorithm found one
 */
            goto __pyx_L18_break;
          }
          __pyx_L25:;
        }

        /* "rmgpy/molecule/group.py":985
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):             # <<<<<<<<<<<<<<
 *                         if atomtypeFeature == []:
 *                             continue
 */
        __pyx_L20_break:;
        __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;

        /* "rmgpy/molecule/group.py":983
 *             #within 1 of the total valency available
 *             if filledValency >= valency[element] - 1:
 *                 for specificAtomType in element.specific:             # <<<<<<<<<<<<<<
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 */
      }
      __pyx_L18_break:;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      goto __pyx_L16;
    }
    __pyx_L16:;

    /* "rmgpy/molecule/group.py":1000
 * 
 *             #set the new atom type if the algorithm found one
 *             if newAtomType and newAtomType is not claimedAtomType:             # <<<<<<<<<<<<<<
 *                 atom.atomType[0] = newAtomType
 *                 modified = True
 */
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_v_newAtomType); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1000; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (__pyx_t_5) {
      goto __pyx_L33_next_and;
    } else {
      __pyx_t_12 = __pyx_t_5;
      goto __pyx_L32_bool_binop_done;
    }
    __pyx_L33_next_and:;
    __pyx_t_5 = (__pyx_v_newAtomType != __pyx_v_claimedAtomType);
    __pyx_t_8 = (__pyx_t_5 != 0);
    __pyx_t_12 = __pyx_t_8;
    __pyx_L32_bool_binop_done:;
    if (__pyx_t_12) {

      /* "rmgpy/molecule/group.py":1001
 *             #set the new atom type if the algorithm found one
 *             if newAtomType and newAtomType is not claimedAtomType:
 *                 atom.atomType[0] = newAtomType             # <<<<<<<<<<<<<<
 *                 modified = True
 * 
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1001; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      if (unlikely(__Pyx_SetItemInt(__pyx_t_3, 0, __pyx_v_newAtomType, long, 1, __Pyx_PyInt_From_long, 0, 0, 1) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1001; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1002
 *             if newAtomType and newAtomType is not claimedAtomType:
 *                 atom.atomType[0] = newAtomType
 *                 modified = True             # <<<<<<<<<<<<<<
 * 
 *         return modified
 */
      __pyx_v_modified = 1;
      goto __pyx_L31;
    }
    __pyx_L31:;

    /* "rmgpy/molecule/group.py":951
 *         #list of :class:AtomType which are elements with more sub-divided atomtypes beneath them
 *         specifics= [elementLabel for elementLabel in allElements if elementLabel not in nonSpecifics]
 *         for index, atom in enumerate(self.atoms):             # <<<<<<<<<<<<<<
 *             claimedAtomType = atom.atomType[0]
 *             newAtomType = None
 */
    __pyx_L6_continue:;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1004
 *                 modified = True
 * 
 *         return modified             # <<<<<<<<<<<<<<
 * 
 *     def addExplicitLigands(self):
 */
  __pyx_r = __pyx_v_modified;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":922
 *             return True
 * 
 *     def standardizeAtomType(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function changes the atomTypes in a group if the atom must
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_10);
  __Pyx_XDECREF(__pyx_t_11);
  __Pyx_XDECREF(__pyx_t_13);
  __Pyx_XDECREF(__pyx_t_17);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.standardizeAtomType", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_valency);
  __Pyx_XDECREF(__pyx_v_specifics);
  __Pyx_XDECREF(__pyx_v_index);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_claimedAtomType);
  __Pyx_XDECREF(__pyx_v_newAtomType);
  __Pyx_XDECREF(__pyx_v_element);
  __Pyx_XDECREF(__pyx_v_elementLabel);
  __Pyx_XDECREF(__pyx_v_groupFeatures);
  __Pyx_XDECREF(__pyx_v_single);
  __Pyx_XDECREF(__pyx_v_allDouble);
  __Pyx_XDECREF(__pyx_v_triple);
  __Pyx_XDECREF(__pyx_v_benzene);
  __Pyx_XDECREF(__pyx_v_bondValency);
  __Pyx_XDECREF(__pyx_v_filledValency);
  __Pyx_XDECREF(__pyx_v_specificAtomType);
  __Pyx_XDECREF(__pyx_v_atomtypeFeatureList);
  __Pyx_XDECREF(__pyx_v_molFeature);
  __Pyx_XDECREF(__pyx_v_atomtypeFeature);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_59standardizeAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_58standardizeAtomType[] = "\n        This function changes the atomTypes in a group if the atom must\n        be a specific atomType based on its bonds and valency.\n\n        Currently only standardizes oxygen, carbon and sulfur atomTypes\n\n        We also only check when there is exactly one atomType,\n        one bondType, one radical setting.\n        For any group where there are wildcards or multiple attributes,\n        we cannot apply this check.\n\n        In the case where the atomType is ambigious based on bonds\n        and valency, this function will not change the type.\n\n        Returns a 'True' if the group was modified otherwise returns 'False'\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_59standardizeAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("standardizeAtomType (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_58standardizeAtomType(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_58standardizeAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("standardizeAtomType", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_standardizeAtomType(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 922; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.standardizeAtomType", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1006
 *         return modified
 * 
 *     def addExplicitLigands(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function Od/Sd ligand to CO or CS atomtypes if they are not already there.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_61addExplicitLigands(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_addExplicitLigands(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_v_modified;
  PyObject *__pyx_v_atomsToAddTo = NULL;
  PyObject *__pyx_v_index = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_claimedAtomType = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_v_atomIndex = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_newAtom = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_newBond = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  PyObject *(*__pyx_t_7)(PyObject *);
  int __pyx_t_8;
  int __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  Py_ssize_t __pyx_t_11;
  int __pyx_t_12;
  PyObject *__pyx_t_13 = NULL;
  int __pyx_t_14;
  PyObject *__pyx_t_15 = NULL;
  int __pyx_t_16;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addExplicitLigands", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_addExplicitLigands); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1006; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_61addExplicitLigands)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1006; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1006; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1006; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1013
 *         """
 * 
 *         modified = False             # <<<<<<<<<<<<<<
 * 
 *         atomsToAddTo=[]
 */
  __pyx_v_modified = 0;

  /* "rmgpy/molecule/group.py":1015
 *         modified = False
 * 
 *         atomsToAddTo=[]             # <<<<<<<<<<<<<<
 * 
 *         for index, atom in enumerate(self.atoms):
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1015; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomsToAddTo = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1017
 *         atomsToAddTo=[]
 * 
 *         for index, atom in enumerate(self.atoms):             # <<<<<<<<<<<<<<
 *             claimedAtomType = atom.atomType[0]
 *             if claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:
 */
  __Pyx_INCREF(__pyx_int_0);
  __pyx_t_1 = __pyx_int_0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_3 = __pyx_t_2; __Pyx_INCREF(__pyx_t_3); __pyx_t_6 = 0;
    __pyx_t_7 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_7 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_7)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_7(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;
    __Pyx_INCREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_1);
    __pyx_t_2 = PyNumber_Add(__pyx_t_1, __pyx_int_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1017; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1);
    __pyx_t_1 = __pyx_t_2;
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1018
 * 
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]             # <<<<<<<<<<<<<<
 *             if claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:
 *                 for atom2, bond12 in atom.bonds.iteritems():
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1018; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_4 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1018; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_XDECREF_SET(__pyx_v_claimedAtomType, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1019
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]
 *             if claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:             # <<<<<<<<<<<<<<
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isDouble():
 */
    __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1019; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_2 = PyObject_GetItem(__pyx_t_4, __pyx_n_s_CO); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1019; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_8 = (__pyx_v_claimedAtomType == __pyx_t_2);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_9 = (__pyx_t_8 != 0);
    if (!__pyx_t_9) {
      goto __pyx_L7_next_or;
    } else {
      __pyx_t_5 = __pyx_t_9;
      goto __pyx_L6_bool_binop_done;
    }
    __pyx_L7_next_or:;
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1019; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_4 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_CS); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1019; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_9 = (__pyx_v_claimedAtomType == __pyx_t_4);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_8 = (__pyx_t_9 != 0);
    __pyx_t_5 = __pyx_t_8;
    __pyx_L6_bool_binop_done:;
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1020
 *             claimedAtomType = atom.atomType[0]
 *             if claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:
 *                 for atom2, bond12 in atom.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                     if bond12.isDouble():
 *                         break
 */
      __pyx_t_10 = 0;
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1020; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      if (unlikely(__pyx_t_2 == Py_None)) {
        PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1020; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_13 = __Pyx_dict_iterator(__pyx_t_2, 0, __pyx_n_s_iteritems, (&__pyx_t_11), (&__pyx_t_12)); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1020; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_13);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_XDECREF(__pyx_t_4);
      __pyx_t_4 = __pyx_t_13;
      __pyx_t_13 = 0;
      while (1) {
        __pyx_t_14 = __Pyx_dict_iter_next(__pyx_t_4, __pyx_t_11, &__pyx_t_10, &__pyx_t_13, &__pyx_t_2, NULL, __pyx_t_12);
        if (unlikely(__pyx_t_14 == 0)) break;
        if (unlikely(__pyx_t_14 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1020; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_13);
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_13);
        __pyx_t_13 = 0;
        __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_2);
        __pyx_t_2 = 0;

        /* "rmgpy/molecule/group.py":1021
 *             if claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isDouble():             # <<<<<<<<<<<<<<
 *                         break
 *                 else: atomsToAddTo.append(index)
 */
        __pyx_t_13 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isDouble); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1021; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_13);
        __pyx_t_15 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_13))) {
          __pyx_t_15 = PyMethod_GET_SELF(__pyx_t_13);
          if (likely(__pyx_t_15)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_13);
            __Pyx_INCREF(__pyx_t_15);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_13, function);
          }
        }
        if (__pyx_t_15) {
          __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_13, __pyx_t_15); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1021; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
        } else {
          __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_13); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1021; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
        __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1021; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        if (__pyx_t_5) {

          /* "rmgpy/molecule/group.py":1022
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isDouble():
 *                         break             # <<<<<<<<<<<<<<
 *                 else: atomsToAddTo.append(index)
 * 
 */
          goto __pyx_L9_break;
        }
      }

      /* "rmgpy/molecule/group.py":1023
 *                     if bond12.isDouble():
 *                         break
 *                 else: atomsToAddTo.append(index)             # <<<<<<<<<<<<<<
 * 
 *         for atomIndex in atomsToAddTo:
 */
      /*else*/ {
        __pyx_t_16 = __Pyx_PyList_Append(__pyx_v_atomsToAddTo, __pyx_v_index); if (unlikely(__pyx_t_16 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1023; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_L9_break:;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      goto __pyx_L5;
    }
    __pyx_L5:;

    /* "rmgpy/molecule/group.py":1017
 *         atomsToAddTo=[]
 * 
 *         for index, atom in enumerate(self.atoms):             # <<<<<<<<<<<<<<
 *             claimedAtomType = atom.atomType[0]
 *             if claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:
 */
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1025
 *                 else: atomsToAddTo.append(index)
 * 
 *         for atomIndex in atomsToAddTo:             # <<<<<<<<<<<<<<
 *             modified = True
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 */
  __pyx_t_1 = __pyx_v_atomsToAddTo; __Pyx_INCREF(__pyx_t_1); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_3 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_6); __Pyx_INCREF(__pyx_t_3); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1025; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1025; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomIndex, __pyx_t_3);
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1026
 * 
 *         for atomIndex in atomsToAddTo:
 *             modified = True             # <<<<<<<<<<<<<<
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
    __pyx_v_modified = 1;

    /* "rmgpy/molecule/group.py":1027
 *         for atomIndex in atomsToAddTo:
 *             modified = True
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1027; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = PyObject_GetItem(__pyx_t_3, __pyx_v_atomIndex); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1027; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_4, __pyx_n_s_atomType); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1027; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_GetItemInt(__pyx_t_3, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1027; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1027; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_2 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_CO); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1027; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_5 = (__pyx_t_4 == __pyx_t_2);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_8 = (__pyx_t_5 != 0);
    if (__pyx_t_8) {

      /* "rmgpy/molecule/group.py":1028
 *             modified = True
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
      __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_GetItem(__pyx_t_4, __pyx_n_s_Od); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyList_SET_ITEM(__pyx_t_4, 0, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_atomType, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_INCREF(__pyx_int_0);
      PyList_SET_ITEM(__pyx_t_4, 0, __pyx_int_0);
      __Pyx_GIVEREF(__pyx_int_0);
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_radicalElectrons, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_charge, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_2); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_4));
      __pyx_t_4 = 0;
      goto __pyx_L13;
    }

    /* "rmgpy/molecule/group.py":1029
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             self.addAtom(newAtom)
 */
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1029; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_2 = PyObject_GetItem(__pyx_t_4, __pyx_v_atomIndex); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1029; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1029; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_GetItemInt(__pyx_t_4, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1029; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1029; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_3 = PyObject_GetItem(__pyx_t_4, __pyx_n_s_CS); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1029; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_8 = (__pyx_t_2 == __pyx_t_3);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_5 = (__pyx_t_8 != 0);
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1030
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *             self.addAtom(newAtom)
 *             newBond = GroupBond(self.atoms[atomIndex], newAtom, order=['D'])
 */
      __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_4 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_Sd); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      PyList_SET_ITEM(__pyx_t_2, 0, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_atomType, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_INCREF(__pyx_int_0);
      PyList_SET_ITEM(__pyx_t_2, 0, __pyx_int_0);
      __Pyx_GIVEREF(__pyx_int_0);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_radicalElectrons, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_charge, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_2));
      __pyx_t_2 = 0;
      goto __pyx_L13;
    }
    __pyx_L13:;

    /* "rmgpy/molecule/group.py":1031
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             self.addAtom(newAtom)             # <<<<<<<<<<<<<<
 *             newBond = GroupBond(self.atoms[atomIndex], newAtom, order=['D'])
 *             self.addBond(newBond)
 */
    if (unlikely(!__pyx_v_newAtom)) { __Pyx_RaiseUnboundLocalError("newAtom"); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1031; __pyx_clineno = __LINE__; goto __pyx_L1_error;} }
    __pyx_t_2 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->addAtom(__pyx_v_self, __pyx_v_newAtom, 0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1031; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1032
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             self.addAtom(newAtom)
 *             newBond = GroupBond(self.atoms[atomIndex], newAtom, order=['D'])             # <<<<<<<<<<<<<<
 *             self.addBond(newBond)
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1032; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_v_atomIndex); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1032; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (unlikely(!__pyx_v_newAtom)) { __Pyx_RaiseUnboundLocalError("newAtom"); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1032; __pyx_clineno = __LINE__; goto __pyx_L1_error;} }
    __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1032; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_3);
    __Pyx_GIVEREF(__pyx_t_3);
    __Pyx_INCREF(((PyObject *)__pyx_v_newAtom));
    PyTuple_SET_ITEM(__pyx_t_2, 1, ((PyObject *)__pyx_v_newAtom));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_newAtom));
    __pyx_t_3 = 0;
    __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1032; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1032; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_INCREF(__pyx_n_s_D);
    PyList_SET_ITEM(__pyx_t_4, 0, __pyx_n_s_D);
    __Pyx_GIVEREF(__pyx_n_s_D);
    if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_order, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1032; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1032; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_newBond, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_4));
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1033
 *             self.addAtom(newAtom)
 *             newBond = GroupBond(self.atoms[atomIndex], newAtom, order=['D'])
 *             self.addBond(newBond)             # <<<<<<<<<<<<<<
 * 
 *         return modified
 */
    __pyx_t_4 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->addBond(__pyx_v_self, __pyx_v_newBond, 0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1033; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1025
 *                 else: atomsToAddTo.append(index)
 * 
 *         for atomIndex in atomsToAddTo:             # <<<<<<<<<<<<<<
 *             modified = True
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1035
 *             self.addBond(newBond)
 * 
 *         return modified             # <<<<<<<<<<<<<<
 * 
 *     def standardizeGroup(self):
 */
  __pyx_r = __pyx_v_modified;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1006
 *         return modified
 * 
 *     def addExplicitLigands(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function Od/Sd ligand to CO or CS atomtypes if they are not already there.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_13);
  __Pyx_XDECREF(__pyx_t_15);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.addExplicitLigands", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomsToAddTo);
  __Pyx_XDECREF(__pyx_v_index);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_claimedAtomType);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XDECREF(__pyx_v_atomIndex);
  __Pyx_XDECREF((PyObject *)__pyx_v_newAtom);
  __Pyx_XDECREF((PyObject *)__pyx_v_newBond);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_61addExplicitLigands(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_60addExplicitLigands[] = "\n        This function Od/Sd ligand to CO or CS atomtypes if they are not already there.\n\n        Returns a 'True' if the group was modified otherwise returns 'False'\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_61addExplicitLigands(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addExplicitLigands (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_60addExplicitLigands(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_60addExplicitLigands(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addExplicitLigands", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_addExplicitLigands(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1006; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addExplicitLigands", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1037
 *         return modified
 * 
 *     def standardizeGroup(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function modifies groups to make them have a standard AdjList form.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_63standardizeGroup(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_standardizeGroup(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_v_modified;
  int __pyx_v_viableToCheck;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_bond = NULL;
  PyObject *__pyx_v_checkList = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  PyObject *(*__pyx_t_7)(PyObject *);
  Py_ssize_t __pyx_t_8;
  PyObject *(*__pyx_t_9)(PyObject *);
  Py_ssize_t __pyx_t_10;
  int __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("standardizeGroup", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_standardizeGroup); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1037; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_63standardizeGroup)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1037; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1037; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1037; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1051
 *         """
 * 
 *         modified = False             # <<<<<<<<<<<<<<
 * 
 *         #see if this is a group we can check, must not have any OR groups in
 */
  __pyx_v_modified = 0;

  /* "rmgpy/molecule/group.py":1055
 *         #see if this is a group we can check, must not have any OR groups in
 *         #its bonds or atomtypes
 *         viableToCheck = True             # <<<<<<<<<<<<<<
 *         for atom in self.atoms:
 *             if len(atom.atomType) > 1:
 */
  __pyx_v_viableToCheck = 1;

  /* "rmgpy/molecule/group.py":1056
 *         #its bonds or atomtypes
 *         viableToCheck = True
 *         for atom in self.atoms:             # <<<<<<<<<<<<<<
 *             if len(atom.atomType) > 1:
 *                 viableToCheck = False
 */
  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1056; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
    __pyx_t_2 = __pyx_t_1; __Pyx_INCREF(__pyx_t_2); __pyx_t_6 = 0;
    __pyx_t_7 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1056; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_7 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1056; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (;;) {
    if (likely(!__pyx_t_7)) {
      if (likely(PyList_CheckExact(__pyx_t_2))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1056; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1056; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1056; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1056; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_1 = __pyx_t_7(__pyx_t_2);
      if (unlikely(!__pyx_t_1)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1056; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_1);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1057
 *         viableToCheck = True
 *         for atom in self.atoms:
 *             if len(atom.atomType) > 1:             # <<<<<<<<<<<<<<
 *                 viableToCheck = False
 *                 break
 */
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1057; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_8 = PyObject_Length(__pyx_t_1); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1057; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_5 = ((__pyx_t_8 > 1) != 0);
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1058
 *         for atom in self.atoms:
 *             if len(atom.atomType) > 1:
 *                 viableToCheck = False             # <<<<<<<<<<<<<<
 *                 break
 *             elif len(atom.radicalElectrons) > 1:
 */
      __pyx_v_viableToCheck = 0;

      /* "rmgpy/molecule/group.py":1059
 *             if len(atom.atomType) > 1:
 *                 viableToCheck = False
 *                 break             # <<<<<<<<<<<<<<
 *             elif len(atom.radicalElectrons) > 1:
 *                 viableToCheck = False
 */
      goto __pyx_L4_break;
    }

    /* "rmgpy/molecule/group.py":1060
 *                 viableToCheck = False
 *                 break
 *             elif len(atom.radicalElectrons) > 1:             # <<<<<<<<<<<<<<
 *                 viableToCheck = False
 *                 break
 */
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_radicalElectrons); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1060; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_8 = PyObject_Length(__pyx_t_1); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1060; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_5 = ((__pyx_t_8 > 1) != 0);
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1061
 *                 break
 *             elif len(atom.radicalElectrons) > 1:
 *                 viableToCheck = False             # <<<<<<<<<<<<<<
 *                 break
 *             elif len(atom.lonePairs) > 1:
 */
      __pyx_v_viableToCheck = 0;

      /* "rmgpy/molecule/group.py":1062
 *             elif len(atom.radicalElectrons) > 1:
 *                 viableToCheck = False
 *                 break             # <<<<<<<<<<<<<<
 *             elif len(atom.lonePairs) > 1:
 *                 viableToCheck = False
 */
      goto __pyx_L4_break;
    }

    /* "rmgpy/molecule/group.py":1063
 *                 viableToCheck = False
 *                 break
 *             elif len(atom.lonePairs) > 1:             # <<<<<<<<<<<<<<
 *                 viableToCheck = False
 *                 break
 */
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_lonePairs); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1063; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_8 = PyObject_Length(__pyx_t_1); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1063; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_5 = ((__pyx_t_8 > 1) != 0);
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1064
 *                 break
 *             elif len(atom.lonePairs) > 1:
 *                 viableToCheck = False             # <<<<<<<<<<<<<<
 *                 break
 *             for bond in atom.bonds.values():
 */
      __pyx_v_viableToCheck = 0;

      /* "rmgpy/molecule/group.py":1065
 *             elif len(atom.lonePairs) > 1:
 *                 viableToCheck = False
 *                 break             # <<<<<<<<<<<<<<
 *             for bond in atom.bonds.values():
 *                 if len(bond.order) > 1:
 */
      goto __pyx_L4_break;
    }

    /* "rmgpy/molecule/group.py":1066
 *                 viableToCheck = False
 *                 break
 *             for bond in atom.bonds.values():             # <<<<<<<<<<<<<<
 *                 if len(bond.order) > 1:
 *                     viableToCheck = False
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_values); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_4);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_4, function);
      }
    }
    if (__pyx_t_3) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    } else {
      __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
      __pyx_t_4 = __pyx_t_1; __Pyx_INCREF(__pyx_t_4); __pyx_t_8 = 0;
      __pyx_t_9 = NULL;
    } else {
      __pyx_t_8 = -1; __pyx_t_4 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_9 = Py_TYPE(__pyx_t_4)->tp_iternext; if (unlikely(!__pyx_t_9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    for (;;) {
      if (likely(!__pyx_t_9)) {
        if (likely(PyList_CheckExact(__pyx_t_4))) {
          if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_4)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_1 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_8); __Pyx_INCREF(__pyx_t_1); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_1 = PySequence_ITEM(__pyx_t_4, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        } else {
          if (__pyx_t_8 >= PyTuple_GET_SIZE(__pyx_t_4)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_4, __pyx_t_8); __Pyx_INCREF(__pyx_t_1); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_1 = PySequence_ITEM(__pyx_t_4, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        }
      } else {
        __pyx_t_1 = __pyx_t_9(__pyx_t_4);
        if (unlikely(!__pyx_t_1)) {
          PyObject* exc_type = PyErr_Occurred();
          if (exc_type) {
            if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
            else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1066; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          break;
        }
        __Pyx_GOTREF(__pyx_t_1);
      }
      __Pyx_XDECREF_SET(__pyx_v_bond, __pyx_t_1);
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1067
 *                 break
 *             for bond in atom.bonds.values():
 *                 if len(bond.order) > 1:             # <<<<<<<<<<<<<<
 *                     viableToCheck = False
 *                     break
 */
      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond, __pyx_n_s_order); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1067; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_10 = PyObject_Length(__pyx_t_1); if (unlikely(__pyx_t_10 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1067; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_5 = ((__pyx_t_10 > 1) != 0);
      if (__pyx_t_5) {

        /* "rmgpy/molecule/group.py":1068
 *             for bond in atom.bonds.values():
 *                 if len(bond.order) > 1:
 *                     viableToCheck = False             # <<<<<<<<<<<<<<
 *                     break
 *         if not viableToCheck: return modified
 */
        __pyx_v_viableToCheck = 0;

        /* "rmgpy/molecule/group.py":1069
 *                 if len(bond.order) > 1:
 *                     viableToCheck = False
 *                     break             # <<<<<<<<<<<<<<
 *         if not viableToCheck: return modified
 * 
 */
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":1066
 *                 viableToCheck = False
 *                 break
 *             for bond in atom.bonds.values():             # <<<<<<<<<<<<<<
 *                 if len(bond.order) > 1:
 *                     viableToCheck = False
 */
    }
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1056
 *         #its bonds or atomtypes
 *         viableToCheck = True
 *         for atom in self.atoms:             # <<<<<<<<<<<<<<
 *             if len(atom.atomType) > 1:
 *                 viableToCheck = False
 */
  }
  __pyx_L4_break:;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1070
 *                     viableToCheck = False
 *                     break
 *         if not viableToCheck: return modified             # <<<<<<<<<<<<<<
 * 
 *         #If viable then we apply current conventions:
 */
  __pyx_t_5 = ((!(__pyx_v_viableToCheck != 0)) != 0);
  if (__pyx_t_5) {
    __pyx_r = __pyx_v_modified;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":1073
 * 
 *         #If viable then we apply current conventions:
 *         checkList=[]             # <<<<<<<<<<<<<<
 *         checkList.append(self.standardizeAtomType())
 *         checkList.append(self.addExplicitLigands())
 */
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1073; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_v_checkList = ((PyObject*)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1074
 *         #If viable then we apply current conventions:
 *         checkList=[]
 *         checkList.append(self.standardizeAtomType())             # <<<<<<<<<<<<<<
 *         checkList.append(self.addExplicitLigands())
 * 
 */
  __pyx_t_2 = __Pyx_PyBool_FromLong(((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->standardizeAtomType(__pyx_v_self, 0)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1074; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_11 = __Pyx_PyList_Append(__pyx_v_checkList, __pyx_t_2); if (unlikely(__pyx_t_11 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1074; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1075
 *         checkList=[]
 *         checkList.append(self.standardizeAtomType())
 *         checkList.append(self.addExplicitLigands())             # <<<<<<<<<<<<<<
 * 
 *         return any(checkList)
 */
  __pyx_t_2 = __Pyx_PyBool_FromLong(((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->addExplicitLigands(__pyx_v_self, 0)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1075; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_11 = __Pyx_PyList_Append(__pyx_v_checkList, __pyx_t_2); if (unlikely(__pyx_t_11 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1075; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1077
 *         checkList.append(self.addExplicitLigands())
 * 
 *         return any(checkList)             # <<<<<<<<<<<<<<
 */
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1077; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_v_checkList);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_checkList);
  __Pyx_GIVEREF(__pyx_v_checkList);
  __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1077; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1077; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1037
 *         return modified
 * 
 *     def standardizeGroup(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function modifies groups to make them have a standard AdjList form.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.standardizeGroup", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_bond);
  __Pyx_XDECREF(__pyx_v_checkList);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_63standardizeGroup(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_62standardizeGroup[] = "\n        This function modifies groups to make them have a standard AdjList form.\n\n        Currently it makes atomtypes as specific as possible and makes CO/CS atomtypes\n        have explicit Od/Sd ligands. Other functions can be added as necessary\n\n        We also only check when there is exactly one atomType, one bondType, one\n        radical setting. For any group where there are wildcards or multiple\n        attributes, we do not apply this check.\n\n        Returns a 'True' if the group was modified otherwise returns 'False'\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_63standardizeGroup(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("standardizeGroup (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_62standardizeGroup(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_62standardizeGroup(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("standardizeGroup", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_standardizeGroup(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1037; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.standardizeGroup", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":93
 * cdef class Group(Graph):
 * 
 *     cdef public list multiplicity             # <<<<<<<<<<<<<<
 * 
 *     # These read-only attribues act as a "fingerprint" for accelerating
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->multiplicity);
  __pyx_r = __pyx_v_self->multiplicity;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 93; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->multiplicity);
  __Pyx_DECREF(__pyx_v_self->multiplicity);
  __pyx_v_self->multiplicity = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.multiplicity.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->multiplicity);
  __Pyx_DECREF(__pyx_v_self->multiplicity);
  __pyx_v_self->multiplicity = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":97
 *     # These read-only attribues act as a "fingerprint" for accelerating
 *     # subgraph isomorphism checks
 *     cdef public short carbonCount             # <<<<<<<<<<<<<<
 *     cdef public short nitrogenCount
 *     cdef public short oxygenCount
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->carbonCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 97; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.carbonCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 97; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->carbonCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.carbonCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":98
 *     # subgraph isomorphism checks
 *     cdef public short carbonCount
 *     cdef public short nitrogenCount             # <<<<<<<<<<<<<<
 *     cdef public short oxygenCount
 *     cdef public short sulfurCount
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->nitrogenCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 98; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.nitrogenCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 98; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->nitrogenCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.nitrogenCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":99
 *     cdef public short carbonCount
 *     cdef public short nitrogenCount
 *     cdef public short oxygenCount             # <<<<<<<<<<<<<<
 *     cdef public short sulfurCount
 *     cdef public short radicalCount
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->oxygenCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 99; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.oxygenCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 99; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->oxygenCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.oxygenCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":100
 *     cdef public short nitrogenCount
 *     cdef public short oxygenCount
 *     cdef public short sulfurCount             # <<<<<<<<<<<<<<
 *     cdef public short radicalCount
 * 
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->sulfurCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.sulfurCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->sulfurCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.sulfurCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":101
 *     cdef public short oxygenCount
 *     cdef public short sulfurCount
 *     cdef public short radicalCount             # <<<<<<<<<<<<<<
 * 
 *     cpdef addAtom(self, GroupAtom atom)
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->radicalCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.radicalCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->radicalCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.radicalCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom;

static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_GroupAtom(PyTypeObject *t, PyObject *a, PyObject *k) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *p;
  PyObject *o = __pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_new(t, a, k);
  if (unlikely(!o)) return 0;
  p = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)o);
  p->__pyx_base.__pyx_vtab = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex*)__pyx_vtabptr_5rmgpy_8molecule_5group_GroupAtom;
  p->atomType = ((PyObject*)Py_None); Py_INCREF(Py_None);
  p->radicalElectrons = ((PyObject*)Py_None); Py_INCREF(Py_None);
  p->charge = ((PyObject*)Py_None); Py_INCREF(Py_None);
  p->label = ((PyObject*)Py_None); Py_INCREF(Py_None);
  p->lonePairs = ((PyObject*)Py_None); Py_INCREF(Py_None);
  return o;
}

static void __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupAtom(PyObject *o) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)o;
  #if PY_VERSION_HEX >= 0x030400a1
  if (unlikely(Py_TYPE(o)->tp_finalize) && !_PyGC_FINALIZED(o)) {
    if (PyObject_CallFinalizerFromDealloc(o)) return;
  }
  #endif
  PyObject_GC_UnTrack(o);
  Py_CLEAR(p->atomType);
  Py_CLEAR(p->radicalElectrons);
  Py_CLEAR(p->charge);
  Py_CLEAR(p->label);
  Py_CLEAR(p->lonePairs);
  PyObject_GC_Track(o);
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)) __pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_dealloc(o); else __Pyx_call_next_tp_dealloc(o, __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupAtom);
}

static int __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupAtom(PyObject *o, visitproc v, void *a) {
  int e;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)o;
  e = ((likely(__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)) ? ((__pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_traverse) ? __pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_traverse(o, v, a) : 0) : __Pyx_call_next_tp_traverse(o, v, a, __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupAtom)); if (e) return e;
  if (p->atomType) {
    e = (*v)(p->atomType, a); if (e) return e;
  }
  if (p->radicalElectrons) {
    e = (*v)(p->radicalElectrons, a); if (e) return e;
  }
  if (p->charge) {
    e = (*v)(p->charge, a); if (e) return e;
  }
  if (p->lonePairs) {
    e = (*v)(p->lonePairs, a); if (e) return e;
  }
  return 0;
}

static int __pyx_tp_clear_5rmgpy_8molecule_5group_GroupAtom(PyObject *o) {
  PyObject* tmp;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)o;
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)) { if (__pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_clear) __pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_clear(o); } else __Pyx_call_next_tp_clear(o, __pyx_tp_clear_5rmgpy_8molecule_5group_GroupAtom);
  tmp = ((PyObject*)p->atomType);
  p->atomType = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  tmp = ((PyObject*)p->radicalElectrons);
  p->radicalElectrons = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  tmp = ((PyObject*)p->charge);
  p->charge = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  tmp = ((PyObject*)p->lonePairs);
  p->lonePairs = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  return 0;
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_atomType(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_atomType(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_radicalElectrons(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_radicalElectrons(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_charge(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_charge(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_label(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_label(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_lonePairs(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_lonePairs(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_5__del__(o);
  }
}

static PyMethodDef __pyx_methods_5rmgpy_8molecule_5group_GroupAtom[] = {
  {"__reduce__", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_3__reduce__, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__},
  {"__setstate__", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5__setstate__, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__},
  {"bonds", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_11bonds, METH_NOARGS, 0},
  {"copy", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_12copy},
  {"__changeBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond},
  {"__formBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_16__formBond},
  {"__breakBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond},
  {"__gainRadical", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical},
  {"__loseRadical", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical},
  {"__gainPair", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair},
  {"__losePair", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_26__losePair},
  {"applyAction", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_28applyAction},
  {"equivalent", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_30equivalent},
  {"isSpecificCaseOf", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf},
  {"isOxygen", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen},
  {"isSulfur", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur},
  {0, 0, 0, 0}
};

static struct PyGetSetDef __pyx_getsets_5rmgpy_8molecule_5group_GroupAtom[] = {
  {(char *)"atomType", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_atomType, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_atomType, 0, 0},
  {(char *)"radicalElectrons", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_radicalElectrons, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_radicalElectrons, 0, 0},
  {(char *)"charge", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_charge, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_charge, 0, 0},
  {(char *)"label", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_label, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_label, 0, 0},
  {(char *)"lonePairs", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_lonePairs, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_lonePairs, 0, 0},
  {0, 0, 0, 0, 0}
};

static PyTypeObject __pyx_type_5rmgpy_8molecule_5group_GroupAtom = {
  PyVarObject_HEAD_INIT(0, 0)
  "rmgpy.molecule.group.GroupAtom", /*tp_name*/
  sizeof(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom), /*tp_basicsize*/
  0, /*tp_itemsize*/
  __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupAtom, /*tp_dealloc*/
  0, /*tp_print*/
  0, /*tp_getattr*/
  0, /*tp_setattr*/
  #if PY_MAJOR_VERSION < 3
  0, /*tp_compare*/
  #else
  0, /*reserved*/
  #endif
  __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9__repr__, /*tp_repr*/
  0, /*tp_as_number*/
  0, /*tp_as_sequence*/
  0, /*tp_as_mapping*/
  0, /*tp_hash*/
  0, /*tp_call*/
  __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_7__str__, /*tp_str*/
  0, /*tp_getattro*/
  0, /*tp_setattro*/
  0, /*tp_as_buffer*/
  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE|Py_TPFLAGS_HAVE_GC, /*tp_flags*/
  "\n    An atom group. This class is based on the :class:`Atom` class, except that\n    it uses :ref:`atom types <atom-types>` instead of elements, and all\n    attributes are lists rather than individual values. The attributes are:\n\n    =================== =================== ====================================\n    Attribute           Type                Description\n    =================== =================== ====================================\n    `atomType`          ``list``            The allowed atom types (as :class:`AtomType` objects)\n    `radicalElectrons`  ``list``            The allowed numbers of radical electrons (as short integers)\n    `charge`            ``list``            The allowed formal charges (as short integers)\n    `label`             ``str``             A string label that can be used to tag individual atoms\n    `lonePairs`         ``list``            The number of lone electron pairs\n    =================== =================== ====================================\n\n    Each list represents a logical OR construct, i.e. an atom will match the\n    group if it matches *any* item in the list. However, the\n    `radicalElectrons`, and `charge` attributes are linked\n    such that an atom must match values from the same index in each of these in\n    order to match.\n    ", /*tp_doc*/
  __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupAtom, /*tp_traverse*/
  __pyx_tp_clear_5rmgpy_8molecule_5group_GroupAtom, /*tp_clear*/
  0, /*tp_richcompare*/
  0, /*tp_weaklistoffset*/
  0, /*tp_iter*/
  0, /*tp_iternext*/
  __pyx_methods_5rmgpy_8molecule_5group_GroupAtom, /*tp_methods*/
  0, /*tp_members*/
  __pyx_getsets_5rmgpy_8molecule_5group_GroupAtom, /*tp_getset*/
  0, /*tp_base*/
  0, /*tp_dict*/
  0, /*tp_descr_get*/
  0, /*tp_descr_set*/
  0, /*tp_dictoffset*/
  __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_1__init__, /*tp_init*/
  0, /*tp_alloc*/
  __pyx_tp_new_5rmgpy_8molecule_5group_GroupAtom, /*tp_new*/
  0, /*tp_free*/
  0, /*tp_is_gc*/
  0, /*tp_bases*/
  0, /*tp_mro*/
  0, /*tp_cache*/
  0, /*tp_subclasses*/
  0, /*tp_weaklist*/
  0, /*tp_del*/
  0, /*tp_version_tag*/
  #if PY_VERSION_HEX >= 0x030400a1
  0, /*tp_finalize*/
  #endif
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupBond __pyx_vtable_5rmgpy_8molecule_5group_GroupBond;

static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_GroupBond(PyTypeObject *t, PyObject *a, PyObject *k) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *p;
  PyObject *o = __pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_new(t, a, k);
  if (unlikely(!o)) return 0;
  p = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)o);
  p->__pyx_base.__pyx_vtab = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge*)__pyx_vtabptr_5rmgpy_8molecule_5group_GroupBond;
  p->order = ((PyObject*)Py_None); Py_INCREF(Py_None);
  return o;
}

static void __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupBond(PyObject *o) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)o;
  #if PY_VERSION_HEX >= 0x030400a1
  if (unlikely(Py_TYPE(o)->tp_finalize) && !_PyGC_FINALIZED(o)) {
    if (PyObject_CallFinalizerFromDealloc(o)) return;
  }
  #endif
  PyObject_GC_UnTrack(o);
  Py_CLEAR(p->order);
  PyObject_GC_Track(o);
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Edge)) __pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_dealloc(o); else __Pyx_call_next_tp_dealloc(o, __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupBond);
}

static int __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupBond(PyObject *o, visitproc v, void *a) {
  int e;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)o;
  e = ((likely(__pyx_ptype_5rmgpy_8molecule_5graph_Edge)) ? ((__pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_traverse) ? __pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_traverse(o, v, a) : 0) : __Pyx_call_next_tp_traverse(o, v, a, __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupBond)); if (e) return e;
  if (p->order) {
    e = (*v)(p->order, a); if (e) return e;
  }
  return 0;
}

static int __pyx_tp_clear_5rmgpy_8molecule_5group_GroupBond(PyObject *o) {
  PyObject* tmp;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)o;
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Edge)) { if (__pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_clear) __pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_clear(o); } else __Pyx_call_next_tp_clear(o, __pyx_tp_clear_5rmgpy_8molecule_5group_GroupBond);
  tmp = ((PyObject*)p->order);
  p->order = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  return 0;
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupBond_order(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupBond_order(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_5__del__(o);
  }
}

static PyMethodDef __pyx_methods_5rmgpy_8molecule_5group_GroupBond[] = {
  {"__reduce__", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_7__reduce__, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_6__reduce__},
  {"copy", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_8copy},
  {"isSingle", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_10isSingle},
  {"isDouble", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_12isDouble},
  {"isTriple", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_14isTriple},
  {"isBenzene", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_16isBenzene},
  {"__changeBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_18__changeBond},
  {"applyAction", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_20applyAction},
  {"equivalent", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_22equivalent},
  {"isSpecificCaseOf", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf},
  {0, 0, 0, 0}
};

static struct PyGetSetDef __pyx_getsets_5rmgpy_8molecule_5group_GroupBond[] = {
  {(char *)"order", __pyx_getprop_5rmgpy_8molecule_5group_9GroupBond_order, __pyx_setprop_5rmgpy_8molecule_5group_9GroupBond_order, 0, 0},
  {0, 0, 0, 0, 0}
};

static PyTypeObject __pyx_type_5rmgpy_8molecule_5group_GroupBond = {
  PyVarObject_HEAD_INIT(0, 0)
  "rmgpy.molecule.group.GroupBond", /*tp_name*/
  sizeof(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond), /*tp_basicsize*/
  0, /*tp_itemsize*/
  __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupBond, /*tp_dealloc*/
  0, /*tp_print*/
  0, /*tp_getattr*/
  0, /*tp_setattr*/
  #if PY_MAJOR_VERSION < 3
  0, /*tp_compare*/
  #else
  0, /*reserved*/
  #endif
  __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5__repr__, /*tp_repr*/
  0, /*tp_as_number*/
  0, /*tp_as_sequence*/
  0, /*tp_as_mapping*/
  0, /*tp_hash*/
  0, /*tp_call*/
  __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_3__str__, /*tp_str*/
  0, /*tp_getattro*/
  0, /*tp_setattro*/
  0, /*tp_as_buffer*/
  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE|Py_TPFLAGS_HAVE_GC, /*tp_flags*/
  "\n    A bond group. This class is based on the :class:`Bond` class, except that\n    all attributes are lists rather than individual values. The allowed bond\n    types are given :ref:`here <bond-types>`. The attributes are:\n\n    =================== =================== ====================================\n    Attribute           Type                Description\n    =================== =================== ====================================\n    `order`             ``list``            The allowed bond orders (as character strings)\n    =================== =================== ====================================\n\n    Each list represents a logical OR construct, i.e. a bond will match the\n    group if it matches *any* item in the list.\n    ", /*tp_doc*/
  __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupBond, /*tp_traverse*/
  __pyx_tp_clear_5rmgpy_8molecule_5group_GroupBond, /*tp_clear*/
  0, /*tp_richcompare*/
  0, /*tp_weaklistoffset*/
  0, /*tp_iter*/
  0, /*tp_iternext*/
  __pyx_methods_5rmgpy_8molecule_5group_GroupBond, /*tp_methods*/
  0, /*tp_members*/
  __pyx_getsets_5rmgpy_8molecule_5group_GroupBond, /*tp_getset*/
  0, /*tp_base*/
  0, /*tp_dict*/
  0, /*tp_descr_get*/
  0, /*tp_descr_set*/
  0, /*tp_dictoffset*/
  __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_1__init__, /*tp_init*/
  0, /*tp_alloc*/
  __pyx_tp_new_5rmgpy_8molecule_5group_GroupBond, /*tp_new*/
  0, /*tp_free*/
  0, /*tp_is_gc*/
  0, /*tp_bases*/
  0, /*tp_mro*/
  0, /*tp_cache*/
  0, /*tp_subclasses*/
  0, /*tp_weaklist*/
  0, /*tp_del*/
  0, /*tp_version_tag*/
  #if PY_VERSION_HEX >= 0x030400a1
  0, /*tp_finalize*/
  #endif
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group __pyx_vtable_5rmgpy_8molecule_5group_Group;

static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_Group(PyTypeObject *t, PyObject *a, PyObject *k) {
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *p;
  PyObject *o = __pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_new(t, a, k);
  if (unlikely(!o)) return 0;
  p = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)o);
  p->__pyx_base.__pyx_vtab = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph*)__pyx_vtabptr_5rmgpy_8molecule_5group_Group;
  p->multiplicity = ((PyObject*)Py_None); Py_INCREF(Py_None);
  return o;
}

static void __pyx_tp_dealloc_5rmgpy_8molecule_5group_Group(PyObject *o) {
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *p = (struct __pyx_obj_5rmgpy_8molecule_5group_Group *)o;
  #if PY_VERSION_HEX >= 0x030400a1
  if (unlikely(Py_TYPE(o)->tp_finalize) && !_PyGC_FINALIZED(o)) {
    if (PyObject_CallFinalizerFromDealloc(o)) return;
  }
  #endif
  PyObject_GC_UnTrack(o);
  Py_CLEAR(p->multiplicity);
  PyObject_GC_Track(o);
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Graph)) __pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_dealloc(o); else __Pyx_call_next_tp_dealloc(o, __pyx_tp_dealloc_5rmgpy_8molecule_5group_Group);
}

static int __pyx_tp_traverse_5rmgpy_8molecule_5group_Group(PyObject *o, visitproc v, void *a) {
  int e;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *p = (struct __pyx_obj_5rmgpy_8molecule_5group_Group *)o;
  e = ((likely(__pyx_ptype_5rmgpy_8molecule_5graph_Graph)) ? ((__pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_traverse) ? __pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_traverse(o, v, a) : 0) : __Pyx_call_next_tp_traverse(o, v, a, __pyx_tp_traverse_5rmgpy_8molecule_5group_Group)); if (e) return e;
  if (p->multiplicity) {
    e = (*v)(p->multiplicity, a); if (e) return e;
  }
  return 0;
}

static int __pyx_tp_clear_5rmgpy_8molecule_5group_Group(PyObject *o) {
  PyObject* tmp;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *p = (struct __pyx_obj_5rmgpy_8molecule_5group_Group *)o;
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Graph)) { if (__pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_clear) __pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_clear(o); } else __Pyx_call_next_tp_clear(o, __pyx_tp_clear_5rmgpy_8molecule_5group_Group);
  tmp = ((PyObject*)p->multiplicity);
  p->multiplicity = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  return 0;
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_multiplicity(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_multiplicity(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_carbonCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_carbonCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_nitrogenCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_nitrogenCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_oxygenCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_oxygenCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_sulfurCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_sulfurCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_radicalCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_radicalCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyMethodDef __pyx_methods_5rmgpy_8molecule_5group_Group[] = {
  {"__reduce__", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_3__reduce__, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_2__reduce__},
  {"__getAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_5__getAtoms, METH_NOARGS, 0},
  {"__setAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_7__setAtoms, METH_O, 0},
  {"addAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_8addAtom},
  {"addBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_10addBond},
  {"getBonds", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_12getBonds},
  {"getBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_14getBond},
  {"hasAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_16hasAtom},
  {"hasBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_18hasBond},
  {"removeAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_20removeAtom},
  {"removeBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_22removeBond},
  {"sortAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_24sortAtoms},
  {"copy", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_26copy},
  {"update", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_29update, METH_NOARGS, 0},
  {"merge", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_30merge},
  {"split", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_33split, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_32split},
  {"clearLabeledAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms},
  {"containsLabeledAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom},
  {"getLabeledAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_38getLabeledAtom},
  {"getLabeledAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms},
  {"fromAdjacencyList", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList},
  {"toAdjacencyList", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_44toAdjacencyList},
  {"updateFingerprint", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_46updateFingerprint},
  {"isIsomorphic", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_48isIsomorphic},
  {"findIsomorphism", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_50findIsomorphism},
  {"isSubgraphIsomorphic", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic},
  {"findSubgraphIsomorphisms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms},
  {"isIdentical", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_56isIdentical},
  {"standardizeAtomType", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_59standardizeAtomType, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_58standardizeAtomType},
  {"addExplicitLigands", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_61addExplicitLigands, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_60addExplicitLigands},
  {"standardizeGroup", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_63standardizeGroup, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_62standardizeGroup},
  {0, 0, 0, 0}
};

static struct PyGetSetDef __pyx_getsets_5rmgpy_8molecule_5group_Group[] = {
  {(char *)"multiplicity", __pyx_getprop_5rmgpy_8molecule_5group_5Group_multiplicity, __pyx_setprop_5rmgpy_8molecule_5group_5Group_multiplicity, 0, 0},
  {(char *)"carbonCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_carbonCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_carbonCount, 0, 0},
  {(char *)"nitrogenCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_nitrogenCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_nitrogenCount, 0, 0},
  {(char *)"oxygenCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_oxygenCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_oxygenCount, 0, 0},
  {(char *)"sulfurCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_sulfurCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_sulfurCount, 0, 0},
  {(char *)"radicalCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_radicalCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_radicalCount, 0, 0},
  {0, 0, 0, 0, 0}
};

static PyTypeObject __pyx_type_5rmgpy_8molecule_5group_Group = {
  PyVarObject_HEAD_INIT(0, 0)
  "rmgpy.molecule.group.Group", /*tp_name*/
  sizeof(struct __pyx_obj_5rmgpy_8molecule_5group_Group), /*tp_basicsize*/
  0, /*tp_itemsize*/
  __pyx_tp_dealloc_5rmgpy_8molecule_5group_Group, /*tp_dealloc*/
  0, /*tp_print*/
  0, /*tp_getattr*/
  0, /*tp_setattr*/
  #if PY_MAJOR_VERSION < 3
  0, /*tp_compare*/
  #else
  0, /*reserved*/
  #endif
  0, /*tp_repr*/
  0, /*tp_as_number*/
  0, /*tp_as_sequence*/
  0, /*tp_as_mapping*/
  0, /*tp_hash*/
  0, /*tp_call*/
  0, /*tp_str*/
  0, /*tp_getattro*/
  0, /*tp_setattro*/
  0, /*tp_as_buffer*/
  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE|Py_TPFLAGS_HAVE_GC, /*tp_flags*/
  "\n    A representation of a molecular substructure group using a graph data\n    type, extending the :class:`Graph` class. The `atoms` and `bonds` attributes\n    are aliases for the `vertices` and `edges` attributes, and store \n    :class:`GroupAtom` and :class:`GroupBond` objects, respectively.\n    Corresponding alias methods have also been provided.\n    ", /*tp_doc*/
  __pyx_tp_traverse_5rmgpy_8molecule_5group_Group, /*tp_traverse*/
  __pyx_tp_clear_5rmgpy_8molecule_5group_Group, /*tp_clear*/
  0, /*tp_richcompare*/
  0, /*tp_weaklistoffset*/
  0, /*tp_iter*/
  0, /*tp_iternext*/
  __pyx_methods_5rmgpy_8molecule_5group_Group, /*tp_methods*/
  0, /*tp_members*/
  __pyx_getsets_5rmgpy_8molecule_5group_Group, /*tp_getset*/
  0, /*tp_base*/
  0, /*tp_dict*/
  0, /*tp_descr_get*/
  0, /*tp_descr_set*/
  0, /*tp_dictoffset*/
  __pyx_pw_5rmgpy_8molecule_5group_5Group_1__init__, /*tp_init*/
  0, /*tp_alloc*/
  __pyx_tp_new_5rmgpy_8molecule_5group_Group, /*tp_new*/
  0, /*tp_free*/
  0, /*tp_is_gc*/
  0, /*tp_bases*/
  0, /*tp_mro*/
  0, /*tp_cache*/
  0, /*tp_subclasses*/
  0, /*tp_weaklist*/
  0, /*tp_del*/
  0, /*tp_version_tag*/
  #if PY_VERSION_HEX >= 0x030400a1
  0, /*tp_finalize*/
  #endif
};

static PyMethodDef __pyx_methods[] = {
  {0, 0, 0, 0}
};

#if PY_MAJOR_VERSION >= 3
static struct PyModuleDef __pyx_moduledef = {
  #if PY_VERSION_HEX < 0x03020000
    { PyObject_HEAD_INIT(NULL) NULL, 0, NULL },
  #else
    PyModuleDef_HEAD_INIT,
  #endif
    "group",
    __pyx_k_This_module_provides_classes_an, /* m_doc */
    -1, /* m_size */
    __pyx_methods /* m_methods */,
    NULL, /* m_reload */
    NULL, /* m_traverse */
    NULL, /* m_clear */
    NULL /* m_free */
};
#endif

static __Pyx_StringTabEntry __pyx_string_tab[] = {
  {&__pyx_kp_s_, __pyx_k_, sizeof(__pyx_k_), 0, 0, 1, 0},
  {&__pyx_kp_s_0, __pyx_k_0, sizeof(__pyx_k_0), 0, 0, 1, 0},
  {&__pyx_n_s_ActionError, __pyx_k_ActionError, sizeof(__pyx_k_ActionError), 0, 0, 1, 1},
  {&__pyx_kp_s_An_exception_class_for_errors_t, __pyx_k_An_exception_class_for_errors_t, sizeof(__pyx_k_An_exception_class_for_errors_t), 0, 0, 1, 0},
  {&__pyx_n_s_B, __pyx_k_B, sizeof(__pyx_k_B), 0, 0, 1, 1},
  {&__pyx_n_s_BREAK_BOND, __pyx_k_BREAK_BOND, sizeof(__pyx_k_BREAK_BOND), 0, 0, 1, 1},
  {&__pyx_n_s_C, __pyx_k_C, sizeof(__pyx_k_C), 0, 0, 1, 1},
  {&__pyx_n_s_CHANGE_BOND, __pyx_k_CHANGE_BOND, sizeof(__pyx_k_CHANGE_BOND), 0, 0, 1, 1},
  {&__pyx_n_s_CO, __pyx_k_CO, sizeof(__pyx_k_CO), 0, 0, 1, 1},
  {&__pyx_n_s_CS, __pyx_k_CS, sizeof(__pyx_k_CS), 0, 0, 1, 1},
  {&__pyx_n_s_D, __pyx_k_D, sizeof(__pyx_k_D), 0, 0, 1, 1},
  {&__pyx_n_s_Edge, __pyx_k_Edge, sizeof(__pyx_k_Edge), 0, 0, 1, 1},
  {&__pyx_n_s_Exception, __pyx_k_Exception, sizeof(__pyx_k_Exception), 0, 0, 1, 1},
  {&__pyx_n_s_FORM_BOND, __pyx_k_FORM_BOND, sizeof(__pyx_k_FORM_BOND), 0, 0, 1, 1},
  {&__pyx_n_s_GAIN_PAIR, __pyx_k_GAIN_PAIR, sizeof(__pyx_k_GAIN_PAIR), 0, 0, 1, 1},
  {&__pyx_n_s_GAIN_RADICAL, __pyx_k_GAIN_RADICAL, sizeof(__pyx_k_GAIN_RADICAL), 0, 0, 1, 1},
  {&__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_k_Got_a_0_object_for_parameter_oth, sizeof(__pyx_k_Got_a_0_object_for_parameter_oth), 0, 0, 1, 0},
  {&__pyx_n_s_Graph, __pyx_k_Graph, sizeof(__pyx_k_Graph), 0, 0, 1, 1},
  {&__pyx_kp_s_GroupAtom_0_s, __pyx_k_GroupAtom_0_s, sizeof(__pyx_k_GroupAtom_0_s), 0, 0, 1, 0},
  {&__pyx_kp_s_GroupBond_0_r, __pyx_k_GroupBond_0_r, sizeof(__pyx_k_GroupBond_0_r), 0, 0, 1, 0},
  {&__pyx_n_s_LOSE_PAIR, __pyx_k_LOSE_PAIR, sizeof(__pyx_k_LOSE_PAIR), 0, 0, 1, 1},
  {&__pyx_n_s_LOSE_RADICAL, __pyx_k_LOSE_RADICAL, sizeof(__pyx_k_LOSE_RADICAL), 0, 0, 1, 1},
  {&__pyx_n_s_N, __pyx_k_N, sizeof(__pyx_k_N), 0, 0, 1, 1},
  {&__pyx_kp_s_No_atom_in_the_functional_group, __pyx_k_No_atom_in_the_functional_group, sizeof(__pyx_k_No_atom_in_the_functional_group), 0, 0, 1, 0},
  {&__pyx_n_s_O, __pyx_k_O, sizeof(__pyx_k_O), 0, 0, 1, 1},
  {&__pyx_n_s_Oa, __pyx_k_Oa, sizeof(__pyx_k_Oa), 0, 0, 1, 1},
  {&__pyx_n_s_Od, __pyx_k_Od, sizeof(__pyx_k_Od), 0, 0, 1, 1},
  {&__pyx_n_s_S, __pyx_k_S, sizeof(__pyx_k_S), 0, 0, 1, 1},
  {&__pyx_n_s_Sa, __pyx_k_Sa, sizeof(__pyx_k_Sa), 0, 0, 1, 1},
  {&__pyx_n_s_Sd, __pyx_k_Sd, sizeof(__pyx_k_Sd), 0, 0, 1, 1},
  {&__pyx_n_s_Si, __pyx_k_Si, sizeof(__pyx_k_Si), 0, 0, 1, 1},
  {&__pyx_n_s_T, __pyx_k_T, sizeof(__pyx_k_T), 0, 0, 1, 1},
  {&__pyx_n_s_TypeError, __pyx_k_TypeError, sizeof(__pyx_k_TypeError), 0, 0, 1, 1},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_Inval, __pyx_k_Unable_to_update_GroupAtom_Inval, sizeof(__pyx_k_Unable_to_update_GroupAtom_Inval), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t, __pyx_k_Unable_to_update_GroupAtom_due_t, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_10, __pyx_k_Unable_to_update_GroupAtom_due_t_10, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_10), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_11, __pyx_k_Unable_to_update_GroupAtom_due_t_11, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_11), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_12, __pyx_k_Unable_to_update_GroupAtom_due_t_12, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_12), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_2, __pyx_k_Unable_to_update_GroupAtom_due_t_2, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_2), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_3, __pyx_k_Unable_to_update_GroupAtom_due_t_3, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_3), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_4, __pyx_k_Unable_to_update_GroupAtom_due_t_4, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_4), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_5, __pyx_k_Unable_to_update_GroupAtom_due_t_5, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_5), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_6, __pyx_k_Unable_to_update_GroupAtom_due_t_6, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_6), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_7, __pyx_k_Unable_to_update_GroupAtom_due_t_7, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_7), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_8, __pyx_k_Unable_to_update_GroupAtom_due_t_8, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_8), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_9, __pyx_k_Unable_to_update_GroupAtom_due_t_9, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_9), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupBond_Inval, __pyx_k_Unable_to_update_GroupBond_Inval, sizeof(__pyx_k_Unable_to_update_GroupBond_Inval), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupBond_due_t, __pyx_k_Unable_to_update_GroupBond_due_t, sizeof(__pyx_k_Unable_to_update_GroupBond_due_t), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupBond_due_t_2, __pyx_k_Unable_to_update_GroupBond_due_t_2, sizeof(__pyx_k_Unable_to_update_GroupBond_due_t_2), 0, 0, 1, 0},
  {&__pyx_n_s_ValueError, __pyx_k_ValueError, sizeof(__pyx_k_ValueError), 0, 0, 1, 1},
  {&__pyx_n_s_Vertex, __pyx_k_Vertex, sizeof(__pyx_k_Vertex), 0, 0, 1, 1},
  {&__pyx_kp_s__2, __pyx_k__2, sizeof(__pyx_k__2), 0, 0, 1, 0},
  {&__pyx_n_s_addAtom, __pyx_k_addAtom, sizeof(__pyx_k_addAtom), 0, 0, 1, 1},
  {&__pyx_n_s_addBond, __pyx_k_addBond, sizeof(__pyx_k_addBond), 0, 0, 1, 1},
  {&__pyx_n_s_addExplicitLigands, __pyx_k_addExplicitLigands, sizeof(__pyx_k_addExplicitLigands), 0, 0, 1, 1},
  {&__pyx_n_s_adjlist, __pyx_k_adjlist, sizeof(__pyx_k_adjlist), 0, 0, 1, 1},
  {&__pyx_n_s_all, __pyx_k_all, sizeof(__pyx_k_all), 0, 0, 1, 1},
  {&__pyx_n_s_allElements, __pyx_k_allElements, sizeof(__pyx_k_allElements), 0, 0, 1, 1},
  {&__pyx_n_s_any, __pyx_k_any, sizeof(__pyx_k_any), 0, 0, 1, 1},
  {&__pyx_n_s_append, __pyx_k_append, sizeof(__pyx_k_append), 0, 0, 1, 1},
  {&__pyx_n_s_applyAction, __pyx_k_applyAction, sizeof(__pyx_k_applyAction), 0, 0, 1, 1},
  {&__pyx_n_s_atom1, __pyx_k_atom1, sizeof(__pyx_k_atom1), 0, 0, 1, 1},
  {&__pyx_n_s_atom2, __pyx_k_atom2, sizeof(__pyx_k_atom2), 0, 0, 1, 1},
  {&__pyx_n_s_atomType, __pyx_k_atomType, sizeof(__pyx_k_atomType), 0, 0, 1, 1},
  {&__pyx_n_s_atomTypes, __pyx_k_atomTypes, sizeof(__pyx_k_atomTypes), 0, 0, 1, 1},
  {&__pyx_n_s_atoms, __pyx_k_atoms, sizeof(__pyx_k_atoms), 0, 0, 1, 1},
  {&__pyx_n_s_atomtype, __pyx_k_atomtype, sizeof(__pyx_k_atomtype), 0, 0, 1, 1},
  {&__pyx_n_s_bonds, __pyx_k_bonds, sizeof(__pyx_k_bonds), 0, 0, 1, 1},
  {&__pyx_n_s_breakBond, __pyx_k_breakBond, sizeof(__pyx_k_breakBond), 0, 0, 1, 1},
  {&__pyx_n_s_breakBond_2, __pyx_k_breakBond_2, sizeof(__pyx_k_breakBond_2), 0, 0, 1, 1},
  {&__pyx_n_s_changeBond, __pyx_k_changeBond, sizeof(__pyx_k_changeBond), 0, 0, 1, 1},
  {&__pyx_n_s_charge, __pyx_k_charge, sizeof(__pyx_k_charge), 0, 0, 1, 1},
  {&__pyx_n_s_class, __pyx_k_class, sizeof(__pyx_k_class), 0, 0, 1, 1},
  {&__pyx_n_s_clearLabeledAtoms, __pyx_k_clearLabeledAtoms, sizeof(__pyx_k_clearLabeledAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_connectivity1, __pyx_k_connectivity1, sizeof(__pyx_k_connectivity1), 0, 0, 1, 1},
  {&__pyx_n_s_connectivity2, __pyx_k_connectivity2, sizeof(__pyx_k_connectivity2), 0, 0, 1, 1},
  {&__pyx_n_s_connectivity3, __pyx_k_connectivity3, sizeof(__pyx_k_connectivity3), 0, 0, 1, 1},
  {&__pyx_n_s_containsLabeledAtom, __pyx_k_containsLabeledAtom, sizeof(__pyx_k_containsLabeledAtom), 0, 0, 1, 1},
  {&__pyx_n_s_copy, __pyx_k_copy, sizeof(__pyx_k_copy), 0, 0, 1, 1},
  {&__pyx_n_s_decrementBond, __pyx_k_decrementBond, sizeof(__pyx_k_decrementBond), 0, 0, 1, 1},
  {&__pyx_n_s_decrementLonePair, __pyx_k_decrementLonePair, sizeof(__pyx_k_decrementLonePair), 0, 0, 1, 1},
  {&__pyx_n_s_decrementRadical, __pyx_k_decrementRadical, sizeof(__pyx_k_decrementRadical), 0, 0, 1, 1},
  {&__pyx_n_s_deep, __pyx_k_deep, sizeof(__pyx_k_deep), 0, 0, 1, 1},
  {&__pyx_n_s_doc, __pyx_k_doc, sizeof(__pyx_k_doc), 0, 0, 1, 1},
  {&__pyx_n_s_edges, __pyx_k_edges, sizeof(__pyx_k_edges), 0, 0, 1, 1},
  {&__pyx_n_s_enumerate, __pyx_k_enumerate, sizeof(__pyx_k_enumerate), 0, 0, 1, 1},
  {&__pyx_n_s_equivalent, __pyx_k_equivalent, sizeof(__pyx_k_equivalent), 0, 0, 1, 1},
  {&__pyx_n_s_findIsomorphism, __pyx_k_findIsomorphism, sizeof(__pyx_k_findIsomorphism), 0, 0, 1, 1},
  {&__pyx_n_s_findSubgraphIsomorphisms, __pyx_k_findSubgraphIsomorphisms, sizeof(__pyx_k_findSubgraphIsomorphisms), 0, 0, 1, 1},
  {&__pyx_n_s_formBond, __pyx_k_formBond, sizeof(__pyx_k_formBond), 0, 0, 1, 1},
  {&__pyx_n_s_formBond_2, __pyx_k_formBond_2, sizeof(__pyx_k_formBond_2), 0, 0, 1, 1},
  {&__pyx_n_s_format, __pyx_k_format, sizeof(__pyx_k_format), 0, 0, 1, 1},
  {&__pyx_n_s_fromAdjacencyList, __pyx_k_fromAdjacencyList, sizeof(__pyx_k_fromAdjacencyList), 0, 0, 1, 1},
  {&__pyx_n_s_gainPair, __pyx_k_gainPair, sizeof(__pyx_k_gainPair), 0, 0, 1, 1},
  {&__pyx_n_s_gainRadical, __pyx_k_gainRadical, sizeof(__pyx_k_gainRadical), 0, 0, 1, 1},
  {&__pyx_n_s_getAtoms, __pyx_k_getAtoms, sizeof(__pyx_k_getAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_getBond, __pyx_k_getBond, sizeof(__pyx_k_getBond), 0, 0, 1, 1},
  {&__pyx_n_s_getBonds, __pyx_k_getBonds, sizeof(__pyx_k_getBonds), 0, 0, 1, 1},
  {&__pyx_n_s_getFeatures, __pyx_k_getFeatures, sizeof(__pyx_k_getFeatures), 0, 0, 1, 1},
  {&__pyx_n_s_getLabeledAtom, __pyx_k_getLabeledAtom, sizeof(__pyx_k_getLabeledAtom), 0, 0, 1, 1},
  {&__pyx_n_s_getLabeledAtoms, __pyx_k_getLabeledAtoms, sizeof(__pyx_k_getLabeledAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_graph, __pyx_k_graph, sizeof(__pyx_k_graph), 0, 0, 1, 1},
  {&__pyx_n_s_group, __pyx_k_group, sizeof(__pyx_k_group), 0, 0, 1, 1},
  {&__pyx_n_s_hasAtom, __pyx_k_hasAtom, sizeof(__pyx_k_hasAtom), 0, 0, 1, 1},
  {&__pyx_n_s_hasBond, __pyx_k_hasBond, sizeof(__pyx_k_hasBond), 0, 0, 1, 1},
  {&__pyx_n_s_import, __pyx_k_import, sizeof(__pyx_k_import), 0, 0, 1, 1},
  {&__pyx_n_s_incrementBond, __pyx_k_incrementBond, sizeof(__pyx_k_incrementBond), 0, 0, 1, 1},
  {&__pyx_n_s_incrementLonePair, __pyx_k_incrementLonePair, sizeof(__pyx_k_incrementLonePair), 0, 0, 1, 1},
  {&__pyx_n_s_incrementRadical, __pyx_k_incrementRadical, sizeof(__pyx_k_incrementRadical), 0, 0, 1, 1},
  {&__pyx_n_s_init, __pyx_k_init, sizeof(__pyx_k_init), 0, 0, 1, 1},
  {&__pyx_n_s_initialMap, __pyx_k_initialMap, sizeof(__pyx_k_initialMap), 0, 0, 1, 1},
  {&__pyx_n_s_isBenzene, __pyx_k_isBenzene, sizeof(__pyx_k_isBenzene), 0, 0, 1, 1},
  {&__pyx_n_s_isDouble, __pyx_k_isDouble, sizeof(__pyx_k_isDouble), 0, 0, 1, 1},
  {&__pyx_n_s_isIdentical, __pyx_k_isIdentical, sizeof(__pyx_k_isIdentical), 0, 0, 1, 1},
  {&__pyx_n_s_isIsomorphic, __pyx_k_isIsomorphic, sizeof(__pyx_k_isIsomorphic), 0, 0, 1, 1},
  {&__pyx_n_s_isOxygen, __pyx_k_isOxygen, sizeof(__pyx_k_isOxygen), 0, 0, 1, 1},
  {&__pyx_n_s_isSingle, __pyx_k_isSingle, sizeof(__pyx_k_isSingle), 0, 0, 1, 1},
  {&__pyx_n_s_isSpecificCaseOf, __pyx_k_isSpecificCaseOf, sizeof(__pyx_k_isSpecificCaseOf), 0, 0, 1, 1},
  {&__pyx_n_s_isSubgraphIsomorphic, __pyx_k_isSubgraphIsomorphic, sizeof(__pyx_k_isSubgraphIsomorphic), 0, 0, 1, 1},
  {&__pyx_n_s_isSulfur, __pyx_k_isSulfur, sizeof(__pyx_k_isSulfur), 0, 0, 1, 1},
  {&__pyx_n_s_isTriple, __pyx_k_isTriple, sizeof(__pyx_k_isTriple), 0, 0, 1, 1},
  {&__pyx_n_s_iteritems, __pyx_k_iteritems, sizeof(__pyx_k_iteritems), 0, 0, 1, 1},
  {&__pyx_n_s_join, __pyx_k_join, sizeof(__pyx_k_join), 0, 0, 1, 1},
  {&__pyx_n_s_label, __pyx_k_label, sizeof(__pyx_k_label), 0, 0, 1, 1},
  {&__pyx_n_s_lonePairs, __pyx_k_lonePairs, sizeof(__pyx_k_lonePairs), 0, 0, 1, 1},
  {&__pyx_n_s_losePair, __pyx_k_losePair, sizeof(__pyx_k_losePair), 0, 0, 1, 1},
  {&__pyx_n_s_loseRadical, __pyx_k_loseRadical, sizeof(__pyx_k_loseRadical), 0, 0, 1, 1},
  {&__pyx_n_s_main, __pyx_k_main, sizeof(__pyx_k_main), 0, 0, 1, 1},
  {&__pyx_n_s_merge, __pyx_k_merge, sizeof(__pyx_k_merge), 0, 0, 1, 1},
  {&__pyx_n_s_metaclass, __pyx_k_metaclass, sizeof(__pyx_k_metaclass), 0, 0, 1, 1},
  {&__pyx_n_s_module, __pyx_k_module, sizeof(__pyx_k_module), 0, 0, 1, 1},
  {&__pyx_n_s_multiplicity, __pyx_k_multiplicity, sizeof(__pyx_k_multiplicity), 0, 0, 1, 1},
  {&__pyx_n_s_nonSpecifics, __pyx_k_nonSpecifics, sizeof(__pyx_k_nonSpecifics), 0, 0, 1, 1},
  {&__pyx_n_s_order, __pyx_k_order, sizeof(__pyx_k_order), 0, 0, 1, 1},
  {&__pyx_n_s_other, __pyx_k_other, sizeof(__pyx_k_other), 0, 0, 1, 1},
  {&__pyx_n_s_prepare, __pyx_k_prepare, sizeof(__pyx_k_prepare), 0, 0, 1, 1},
  {&__pyx_n_s_property, __pyx_k_property, sizeof(__pyx_k_property), 0, 0, 1, 1},
  {&__pyx_n_s_pyx_vtable, __pyx_k_pyx_vtable, sizeof(__pyx_k_pyx_vtable), 0, 0, 1, 1},
  {&__pyx_n_s_qualname, __pyx_k_qualname, sizeof(__pyx_k_qualname), 0, 0, 1, 1},
  {&__pyx_n_s_radicalElectrons, __pyx_k_radicalElectrons, sizeof(__pyx_k_radicalElectrons), 0, 0, 1, 1},
  {&__pyx_n_s_range, __pyx_k_range, sizeof(__pyx_k_range), 0, 0, 1, 1},
  {&__pyx_n_s_removeAtom, __pyx_k_removeAtom, sizeof(__pyx_k_removeAtom), 0, 0, 1, 1},
  {&__pyx_n_s_removeBond, __pyx_k_removeBond, sizeof(__pyx_k_removeBond), 0, 0, 1, 1},
  {&__pyx_n_s_rmgpy_molecule_group, __pyx_k_rmgpy_molecule_group, sizeof(__pyx_k_rmgpy_molecule_group), 0, 0, 1, 1},
  {&__pyx_n_s_setAtoms, __pyx_k_setAtoms, sizeof(__pyx_k_setAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_sortAtoms, __pyx_k_sortAtoms, sizeof(__pyx_k_sortAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_sortingLabel, __pyx_k_sortingLabel, sizeof(__pyx_k_sortingLabel), 0, 0, 1, 1},
  {&__pyx_n_s_specific, __pyx_k_specific, sizeof(__pyx_k_specific), 0, 0, 1, 1},
  {&__pyx_n_s_split, __pyx_k_split, sizeof(__pyx_k_split), 0, 0, 1, 1},
  {&__pyx_n_s_standardizeAtomType, __pyx_k_standardizeAtomType, sizeof(__pyx_k_standardizeAtomType), 0, 0, 1, 1},
  {&__pyx_n_s_standardizeGroup, __pyx_k_standardizeGroup, sizeof(__pyx_k_standardizeGroup), 0, 0, 1, 1},
  {&__pyx_n_s_test, __pyx_k_test, sizeof(__pyx_k_test), 0, 0, 1, 1},
  {&__pyx_n_s_toAdjacencyList, __pyx_k_toAdjacencyList, sizeof(__pyx_k_toAdjacencyList), 0, 0, 1, 1},
  {&__pyx_n_s_update, __pyx_k_update, sizeof(__pyx_k_update), 0, 0, 1, 1},
  {&__pyx_n_s_updateFingerprint, __pyx_k_updateFingerprint, sizeof(__pyx_k_updateFingerprint), 0, 0, 1, 1},
  {&__pyx_n_s_upper, __pyx_k_upper, sizeof(__pyx_k_upper), 0, 0, 1, 1},
  {&__pyx_n_s_values, __pyx_k_values, sizeof(__pyx_k_values), 0, 0, 1, 1},
  {&__pyx_n_s_vertices, __pyx_k_vertices, sizeof(__pyx_k_vertices), 0, 0, 1, 1},
  {&__pyx_n_s_zip, __pyx_k_zip, sizeof(__pyx_k_zip), 0, 0, 1, 1},
  {0, 0, 0, 0, 0, 0, 0}
};
static int __Pyx_InitCachedBuiltins(void) {
  __pyx_builtin_Exception = __Pyx_GetBuiltinName(__pyx_n_s_Exception); if (!__pyx_builtin_Exception) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 43; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_property = __Pyx_GetBuiltinName(__pyx_n_s_property); if (!__pyx_builtin_property) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 125; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_range = __Pyx_GetBuiltinName(__pyx_n_s_range); if (!__pyx_builtin_range) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 79; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_any = __Pyx_GetBuiltinName(__pyx_n_s_any); if (!__pyx_builtin_any) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_zip = __Pyx_GetBuiltinName(__pyx_n_s_zip); if (!__pyx_builtin_zip) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_all = __Pyx_GetBuiltinName(__pyx_n_s_all); if (!__pyx_builtin_all) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 398; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_ValueError = __Pyx_GetBuiltinName(__pyx_n_s_ValueError); if (!__pyx_builtin_ValueError) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_TypeError = __Pyx_GetBuiltinName(__pyx_n_s_TypeError); if (!__pyx_builtin_TypeError) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_enumerate = __Pyx_GetBuiltinName(__pyx_n_s_enumerate); if (!__pyx_builtin_enumerate) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 951; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  return 0;
  __pyx_L1_error:;
  return -1;
}

static int __Pyx_InitCachedConstants(void) {
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__Pyx_InitCachedConstants", 0);
  __Pyx_RefNannyFinishContext();
  return 0;
}

static int __Pyx_InitGlobals(void) {
  if (__Pyx_InitStrings(__pyx_string_tab) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __pyx_int_0 = PyInt_FromLong(0); if (unlikely(!__pyx_int_0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_1 = PyInt_FromLong(1); if (unlikely(!__pyx_int_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_2 = PyInt_FromLong(2); if (unlikely(!__pyx_int_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_3 = PyInt_FromLong(3); if (unlikely(!__pyx_int_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_4 = PyInt_FromLong(4); if (unlikely(!__pyx_int_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  return 0;
  __pyx_L1_error:;
  return -1;
}

#if PY_MAJOR_VERSION < 3
PyMODINIT_FUNC initgroup(void); /*proto*/
PyMODINIT_FUNC initgroup(void)
#else
PyMODINIT_FUNC PyInit_group(void); /*proto*/
PyMODINIT_FUNC PyInit_group(void)
#endif
{
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannyDeclarations
  #if CYTHON_REFNANNY
  __Pyx_RefNanny = __Pyx_RefNannyImportAPI("refnanny");
  if (!__Pyx_RefNanny) {
      PyErr_Clear();
      __Pyx_RefNanny = __Pyx_RefNannyImportAPI("Cython.Runtime.refnanny");
      if (!__Pyx_RefNanny)
          Py_FatalError("failed to import 'refnanny' module");
  }
  #endif
  __Pyx_RefNannySetupContext("PyMODINIT_FUNC PyInit_group(void)", 0);
  if ( __Pyx_check_binary_version() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_empty_tuple = PyTuple_New(0); if (unlikely(!__pyx_empty_tuple)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_empty_bytes = PyBytes_FromStringAndSize("", 0); if (unlikely(!__pyx_empty_bytes)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #ifdef __Pyx_CyFunction_USED
  if (__Pyx_CyFunction_init() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #endif
  #ifdef __Pyx_FusedFunction_USED
  if (__pyx_FusedFunction_init() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #endif
  #ifdef __Pyx_Generator_USED
  if (__pyx_Generator_init() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #endif
  /*--- Library function declarations ---*/
  /*--- Threads initialization code ---*/
  #if defined(__PYX_FORCE_INIT_THREADS) && __PYX_FORCE_INIT_THREADS
  #ifdef WITH_THREAD /* Python build with threading support? */
  PyEval_InitThreads();
  #endif
  #endif
  /*--- Module creation code ---*/
  #if PY_MAJOR_VERSION < 3
  __pyx_m = Py_InitModule4("group", __pyx_methods, __pyx_k_This_module_provides_classes_an, 0, PYTHON_API_VERSION); Py_XINCREF(__pyx_m);
  #else
  __pyx_m = PyModule_Create(&__pyx_moduledef);
  #endif
  if (unlikely(!__pyx_m)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_d = PyModule_GetDict(__pyx_m); if (unlikely(!__pyx_d)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  Py_INCREF(__pyx_d);
  __pyx_b = PyImport_AddModule(__Pyx_BUILTIN_MODULE_NAME); if (unlikely(!__pyx_b)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #if CYTHON_COMPILING_IN_PYPY
  Py_INCREF(__pyx_b);
  #endif
  if (PyObject_SetAttrString(__pyx_m, "__builtins__", __pyx_b) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  /*--- Initialize various global constants etc. ---*/
  if (unlikely(__Pyx_InitGlobals() < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #if PY_MAJOR_VERSION < 3 && (__PYX_DEFAULT_STRING_ENCODING_IS_ASCII || __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT)
  if (__Pyx_init_sys_getdefaultencoding_params() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #endif
  if (__pyx_module_is_main_rmgpy__molecule__group) {
    if (PyObject_SetAttrString(__pyx_m, "__name__", __pyx_n_s_main) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  }
  #if PY_MAJOR_VERSION >= 3
  {
    PyObject *modules = PyImport_GetModuleDict(); if (unlikely(!modules)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (!PyDict_GetItemString(modules, "rmgpy.molecule.group")) {
      if (unlikely(PyDict_SetItemString(modules, "rmgpy.molecule.group", __pyx_m) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
  }
  #endif
  /*--- Builtin init code ---*/
  if (unlikely(__Pyx_InitCachedBuiltins() < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  /*--- Constants init code ---*/
  if (unlikely(__Pyx_InitCachedConstants() < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  /*--- Global init code ---*/
  /*--- Variable export code ---*/
  /*--- Function export code ---*/
  /*--- Type init code ---*/
  __pyx_ptype_5rmgpy_8molecule_5graph_Vertex = __Pyx_ImportType("rmgpy.molecule.graph", "Vertex", sizeof(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5graph_Vertex = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_5graph_Vertex)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5group_GroupAtom = &__pyx_vtable_5rmgpy_8molecule_5group_GroupAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx_base = *__pyx_vtabptr_5rmgpy_8molecule_5graph_Vertex;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx_base.copy = (struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_copy;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx_base.equivalent = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_equivalent;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx_base.isSpecificCaseOf = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSpecificCaseOf;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___changeBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___changeBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___formBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___formBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___breakBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___breakBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___gainRadical = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainRadical;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___loseRadical = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___loseRadical;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___gainPair = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainPair;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___losePair = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___losePair;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.applyAction = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_applyAction;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.isOxygen = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isOxygen;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.isSulfur = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSulfur;
  __pyx_type_5rmgpy_8molecule_5group_GroupAtom.tp_base = __pyx_ptype_5rmgpy_8molecule_5graph_Vertex;
  if (PyType_Ready(&__pyx_type_5rmgpy_8molecule_5group_GroupAtom) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 53; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_type_5rmgpy_8molecule_5group_GroupAtom.tp_print = 0;
  #if CYTHON_COMPILING_IN_CPYTHON
  {
    PyObject *wrapper = PyObject_GetAttrString((PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupAtom, "__str__"); if (unlikely(!wrapper)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 53; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (Py_TYPE(wrapper) == &PyWrapperDescr_Type) {
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_6__str__ = *((PyWrapperDescrObject *)wrapper)->d_base;
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_6__str__.doc = __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_6__str__;
      ((PyWrapperDescrObject *)wrapper)->d_base = &__pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_6__str__;
    }
  }
  #endif
  #if CYTHON_COMPILING_IN_CPYTHON
  {
    PyObject *wrapper = PyObject_GetAttrString((PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupAtom, "__repr__"); if (unlikely(!wrapper)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 53; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (Py_TYPE(wrapper) == &PyWrapperDescr_Type) {
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_8__repr__ = *((PyWrapperDescrObject *)wrapper)->d_base;
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_8__repr__.doc = __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_8__repr__;
      ((PyWrapperDescrObject *)wrapper)->d_base = &__pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_8__repr__;
    }
  }
  #endif
  if (__Pyx_SetVtable(__pyx_type_5rmgpy_8molecule_5group_GroupAtom.tp_dict, __pyx_vtabptr_5rmgpy_8molecule_5group_GroupAtom) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 53; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyObject_SetAttrString(__pyx_m, "GroupAtom", (PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupAtom) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 53; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom = &__pyx_type_5rmgpy_8molecule_5group_GroupAtom;
  __pyx_ptype_5rmgpy_8molecule_5graph_Edge = __Pyx_ImportType("rmgpy.molecule.graph", "Edge", sizeof(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_5graph_Edge)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5graph_Edge = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_5graph_Edge)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5group_GroupBond = &__pyx_vtable_5rmgpy_8molecule_5group_GroupBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx_base = *__pyx_vtabptr_5rmgpy_8molecule_5graph_Edge;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx_base.copy = (struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_copy;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx_base.equivalent = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_equivalent;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx_base.isSpecificCaseOf = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSpecificCaseOf;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx___changeBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond___changeBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.isSingle = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSingle;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.isDouble = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isDouble;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.isTriple = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isTriple;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.isBenzene = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isBenzene;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.applyAction = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_applyAction;
  __pyx_type_5rmgpy_8molecule_5group_GroupBond.tp_base = __pyx_ptype_5rmgpy_8molecule_5graph_Edge;
  if (PyType_Ready(&__pyx_type_5rmgpy_8molecule_5group_GroupBond) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_type_5rmgpy_8molecule_5group_GroupBond.tp_print = 0;
  #if CYTHON_COMPILING_IN_CPYTHON
  {
    PyObject *wrapper = PyObject_GetAttrString((PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupBond, "__str__"); if (unlikely(!wrapper)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (Py_TYPE(wrapper) == &PyWrapperDescr_Type) {
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_2__str__ = *((PyWrapperDescrObject *)wrapper)->d_base;
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_2__str__.doc = __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_2__str__;
      ((PyWrapperDescrObject *)wrapper)->d_base = &__pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_2__str__;
    }
  }
  #endif
  #if CYTHON_COMPILING_IN_CPYTHON
  {
    PyObject *wrapper = PyObject_GetAttrString((PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupBond, "__repr__"); if (unlikely(!wrapper)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (Py_TYPE(wrapper) == &PyWrapperDescr_Type) {
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_4__repr__ = *((PyWrapperDescrObject *)wrapper)->d_base;
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_4__repr__.doc = __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_4__repr__;
      ((PyWrapperDescrObject *)wrapper)->d_base = &__pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_4__repr__;
    }
  }
  #endif
  if (__Pyx_SetVtable(__pyx_type_5rmgpy_8molecule_5group_GroupBond.tp_dict, __pyx_vtabptr_5rmgpy_8molecule_5group_GroupBond) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyObject_SetAttrString(__pyx_m, "GroupBond", (PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupBond) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_5group_GroupBond = &__pyx_type_5rmgpy_8molecule_5group_GroupBond;
  __pyx_ptype_5rmgpy_8molecule_5graph_Graph = __Pyx_ImportType("rmgpy.molecule.graph", "Graph", sizeof(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_5graph_Graph)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5group_Group = &__pyx_vtable_5rmgpy_8molecule_5group_Group;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base = *__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.copy = (struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_copy;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.merge = (struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_merge;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.split = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_split;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.isIsomorphic = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_isIsomorphic;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.findIsomorphism = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_findIsomorphism;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.isSubgraphIsomorphic = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.findSubgraphIsomorphisms = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.addAtom = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_addAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.addBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_addBond;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.getBonds = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_getBonds;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.getBond = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_getBond;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.hasAtom = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_hasAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.hasBond = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_hasBond;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.removeAtom = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_removeAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.removeBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_removeBond;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.sortAtoms = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_sortAtoms;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.clearLabeledAtoms = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_clearLabeledAtoms;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.containsLabeledAtom = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_containsLabeledAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.getLabeledAtom = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.getLabeledAtoms = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtoms;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.fromAdjacencyList = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_fromAdjacencyList;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.toAdjacencyList = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_toAdjacencyList;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.updateFingerprint = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_updateFingerprint;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.isIdentical = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_isIdentical;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.standardizeAtomType = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_standardizeAtomType;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.addExplicitLigands = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_addExplicitLigands;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.standardizeGroup = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_standardizeGroup;
  __pyx_type_5rmgpy_8molecule_5group_Group.tp_base = __pyx_ptype_5rmgpy_8molecule_5graph_Graph;
  if (PyType_Ready(&__pyx_type_5rmgpy_8molecule_5group_Group) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 576; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_type_5rmgpy_8molecule_5group_Group.tp_print = 0;
  if (__Pyx_SetVtable(__pyx_type_5rmgpy_8molecule_5group_Group.tp_dict, __pyx_vtabptr_5rmgpy_8molecule_5group_Group) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 576; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyObject_SetAttrString(__pyx_m, "Group", (PyObject *)&__pyx_type_5rmgpy_8molecule_5group_Group) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 576; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_5group_Group = &__pyx_type_5rmgpy_8molecule_5group_Group;
  /*--- Type import code ---*/
  __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType = __Pyx_ImportType("rmgpy.molecule.atomtype", "AtomType", sizeof(struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 27; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_8atomtype_AtomType = (struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_8atomtype_AtomType)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 27; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  /*--- Variable import code ---*/
  /*--- Function import code ---*/
  /*--- Execution code ---*/

  /* "rmgpy/molecule/group.py":38
 * import cython
 * 
 * from .graph import Vertex, Edge, Graph             # <<<<<<<<<<<<<<
 * from .atomtype import atomTypes, allElements, nonSpecifics, getFeatures
 * 
 */
  __pyx_t_1 = PyList_New(3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 38; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_n_s_Vertex);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_n_s_Vertex);
  __Pyx_GIVEREF(__pyx_n_s_Vertex);
  __Pyx_INCREF(__pyx_n_s_Edge);
  PyList_SET_ITEM(__pyx_t_1, 1, __pyx_n_s_Edge);
  __Pyx_GIVEREF(__pyx_n_s_Edge);
  __Pyx_INCREF(__pyx_n_s_Graph);
  PyList_SET_ITEM(__pyx_t_1, 2, __pyx_n_s_Graph);
  __Pyx_GIVEREF(__pyx_n_s_Graph);
  __pyx_t_2 = __Pyx_Import(__pyx_n_s_graph, __pyx_t_1, 1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 38; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":39
 * 
 * from .graph import Vertex, Edge, Graph
 * from .atomtype import atomTypes, allElements, nonSpecifics, getFeatures             # <<<<<<<<<<<<<<
 * 
 * ################################################################################
 */
  __pyx_t_2 = PyList_New(4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_n_s_atomTypes);
  PyList_SET_ITEM(__pyx_t_2, 0, __pyx_n_s_atomTypes);
  __Pyx_GIVEREF(__pyx_n_s_atomTypes);
  __Pyx_INCREF(__pyx_n_s_allElements);
  PyList_SET_ITEM(__pyx_t_2, 1, __pyx_n_s_allElements);
  __Pyx_GIVEREF(__pyx_n_s_allElements);
  __Pyx_INCREF(__pyx_n_s_nonSpecifics);
  PyList_SET_ITEM(__pyx_t_2, 2, __pyx_n_s_nonSpecifics);
  __Pyx_GIVEREF(__pyx_n_s_nonSpecifics);
  __Pyx_INCREF(__pyx_n_s_getFeatures);
  PyList_SET_ITEM(__pyx_t_2, 3, __pyx_n_s_getFeatures);
  __Pyx_GIVEREF(__pyx_n_s_getFeatures);
  __pyx_t_1 = __Pyx_Import(__pyx_n_s_atomtype, __pyx_t_2, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_ImportFrom(__pyx_t_1, __pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_atomTypes, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_ImportFrom(__pyx_t_1, __pyx_n_s_allElements); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_allElements, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_ImportFrom(__pyx_t_1, __pyx_n_s_nonSpecifics); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_nonSpecifics, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_ImportFrom(__pyx_t_1, __pyx_n_s_getFeatures); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_getFeatures, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":43
 * ################################################################################
 * 
 * class ActionError(Exception):             # <<<<<<<<<<<<<<
 *     """
 *     An exception class for errors that occur while applying reaction recipe
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 43; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_builtin_Exception);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_builtin_Exception);
  __Pyx_GIVEREF(__pyx_builtin_Exception);
  __pyx_t_2 = __Pyx_CalculateMetaclass(NULL, __pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 43; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_Py3MetaclassPrepare(__pyx_t_2, __pyx_t_1, __pyx_n_s_ActionError, __pyx_n_s_ActionError, (PyObject *) NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_kp_s_An_exception_class_for_errors_t); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 43; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_4 = __Pyx_Py3ClassCreate(__pyx_t_2, __pyx_n_s_ActionError, __pyx_t_1, __pyx_t_3, NULL, 0, 1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 43; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_ActionError, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 43; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":126
 * 
 *     @property
 *     def bonds(self): return self.edges             # <<<<<<<<<<<<<<
 * 
 *     def copy(self):
 */
  __pyx_t_1 = __Pyx_GetNameInClass((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 126; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);

  /* "rmgpy/molecule/group.py":125
 *         return "<GroupAtom {0!s}>".format(self)
 * 
 *     @property             # <<<<<<<<<<<<<<
 *     def bonds(self): return self.edges
 * 
 */
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 125; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_property, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 125; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  if (PyDict_SetItem((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom->tp_dict, __pyx_n_s_bonds, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 126; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  PyType_Modified(__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom);

  /* "rmgpy/molecule/group.py":598
 *     def __getAtoms(self): return self.vertices
 *     def __setAtoms(self, atoms): self.vertices = atoms
 *     atoms = property(__getAtoms, __setAtoms)             # <<<<<<<<<<<<<<
 * 
 *     def addAtom(self, atom):
 */
  __pyx_t_1 = __Pyx_GetNameInClass((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_Group, __pyx_n_s_getAtoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 598; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_GetNameInClass((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_Group, __pyx_n_s_setAtoms); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 598; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyTuple_New(2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 598; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  PyTuple_SET_ITEM(__pyx_t_3, 1, __pyx_t_2);
  __Pyx_GIVEREF(__pyx_t_2);
  __pyx_t_1 = 0;
  __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_property, __pyx_t_3, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 598; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  if (PyDict_SetItem((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_Group->tp_dict, __pyx_n_s_atoms, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 598; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  PyType_Modified(__pyx_ptype_5rmgpy_8molecule_5group_Group);

  /* "rmgpy/molecule/group.py":1
 * #!/usr/bin/env python             # <<<<<<<<<<<<<<
 * # encoding: utf-8
 * 
 */
  __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_test, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /*--- Wrapped vars code ---*/

  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  if (__pyx_m) {
    if (__pyx_d) {
      __Pyx_AddTraceback("init rmgpy.molecule.group", __pyx_clineno, __pyx_lineno, __pyx_filename);
      Py_DECREF(__pyx_d); __pyx_d = 0;
    }
    Py_DECREF(__pyx_m); __pyx_m = 0;
  } else if (!PyErr_Occurred()) {
    PyErr_SetString(PyExc_ImportError, "init rmgpy.molecule.group");
  }
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  #if PY_MAJOR_VERSION < 3
  return;
  #else
  return __pyx_m;
  #endif
}

/* Runtime support code */
#if CYTHON_REFNANNY
static __Pyx_RefNannyAPIStruct *__Pyx_RefNannyImportAPI(const char *modname) {
    PyObject *m = NULL, *p = NULL;
    void *r = NULL;
    m = PyImport_ImportModule((char *)modname);
    if (!m) goto end;
    p = PyObject_GetAttrString(m, (char *)"RefNannyAPI");
    if (!p) goto end;
    r = PyLong_AsVoidPtr(p);
end:
    Py_XDECREF(p);
    Py_XDECREF(m);
    return (__Pyx_RefNannyAPIStruct *)r;
}
#endif

static PyObject *__Pyx_GetBuiltinName(PyObject *name) {
    PyObject* result = __Pyx_PyObject_GetAttrStr(__pyx_b, name);
    if (unlikely(!result)) {
        PyErr_Format(PyExc_NameError,
#if PY_MAJOR_VERSION >= 3
            "name '%U' is not defined", name);
#else
            "name '%.200s' is not defined", PyString_AS_STRING(name));
#endif
    }
    return result;
}

static void __Pyx_RaiseDoubleKeywordsError(
    const char* func_name,
    PyObject* kw_name)
{
    PyErr_Format(PyExc_TypeError,
        #if PY_MAJOR_VERSION >= 3
        "%s() got multiple values for keyword argument '%U'", func_name, kw_name);
        #else
        "%s() got multiple values for keyword argument '%s'", func_name,
        PyString_AsString(kw_name));
        #endif
}

static int __Pyx_ParseOptionalKeywords(
    PyObject *kwds,
    PyObject **argnames[],
    PyObject *kwds2,
    PyObject *values[],
    Py_ssize_t num_pos_args,
    const char* function_name)
{
    PyObject *key = 0, *value = 0;
    Py_ssize_t pos = 0;
    PyObject*** name;
    PyObject*** first_kw_arg = argnames + num_pos_args;
    while (PyDict_Next(kwds, &pos, &key, &value)) {
        name = first_kw_arg;
        while (*name && (**name != key)) name++;
        if (*name) {
            values[name-argnames] = value;
            continue;
        }
        name = first_kw_arg;
        #if PY_MAJOR_VERSION < 3
        if (likely(PyString_CheckExact(key)) || likely(PyString_Check(key))) {
            while (*name) {
                if ((CYTHON_COMPILING_IN_PYPY || PyString_GET_SIZE(**name) == PyString_GET_SIZE(key))
                        && _PyString_Eq(**name, key)) {
                    values[name-argnames] = value;
                    break;
                }
                name++;
            }
            if (*name) continue;
            else {
                PyObject*** argname = argnames;
                while (argname != first_kw_arg) {
                    if ((**argname == key) || (
                            (CYTHON_COMPILING_IN_PYPY || PyString_GET_SIZE(**argname) == PyString_GET_SIZE(key))
                             && _PyString_Eq(**argname, key))) {
                        goto arg_passed_twice;
                    }
                    argname++;
                }
            }
        } else
        #endif
        if (likely(PyUnicode_Check(key))) {
            while (*name) {
                int cmp = (**name == key) ? 0 :
                #if !CYTHON_COMPILING_IN_PYPY && PY_MAJOR_VERSION >= 3
                    (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :
                #endif
                    PyUnicode_Compare(**name, key);
                if (cmp < 0 && unlikely(PyErr_Occurred())) goto bad;
                if (cmp == 0) {
                    values[name-argnames] = value;
                    break;
                }
                name++;
            }
            if (*name) continue;
            else {
                PyObject*** argname = argnames;
                while (argname != first_kw_arg) {
                    int cmp = (**argname == key) ? 0 :
                    #if !CYTHON_COMPILING_IN_PYPY && PY_MAJOR_VERSION >= 3
                        (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :
                    #endif
                        PyUnicode_Compare(**argname, key);
                    if (cmp < 0 && unlikely(PyErr_Occurred())) goto bad;
                    if (cmp == 0) goto arg_passed_twice;
                    argname++;
                }
            }
        } else
            goto invalid_keyword_type;
        if (kwds2) {
            if (unlikely(PyDict_SetItem(kwds2, key, value))) goto bad;
        } else {
            goto invalid_keyword;
        }
    }
    return 0;
arg_passed_twice:
    __Pyx_RaiseDoubleKeywordsError(function_name, key);
    goto bad;
invalid_keyword_type:
    PyErr_Format(PyExc_TypeError,
        "%.200s() keywords must be strings", function_name);
    goto bad;
invalid_keyword:
    PyErr_Format(PyExc_TypeError,
    #if PY_MAJOR_VERSION < 3
        "%.200s() got an unexpected keyword argument '%.200s'",
        function_name, PyString_AsString(key));
    #else
        "%s() got an unexpected keyword argument '%U'",
        function_name, key);
    #endif
bad:
    return -1;
}

static void __Pyx_RaiseArgtupleInvalid(
    const char* func_name,
    int exact,
    Py_ssize_t num_min,
    Py_ssize_t num_max,
    Py_ssize_t num_found)
{
    Py_ssize_t num_expected;
    const char *more_or_less;
    if (num_found < num_min) {
        num_expected = num_min;
        more_or_less = "at least";
    } else {
        num_expected = num_max;
        more_or_less = "at most";
    }
    if (exact) {
        more_or_less = "exactly";
    }
    PyErr_Format(PyExc_TypeError,
                 "%.200s() takes %.8s %" CYTHON_FORMAT_SSIZE_T "d positional argument%.1s (%" CYTHON_FORMAT_SSIZE_T "d given)",
                 func_name, more_or_less, num_expected,
                 (num_expected == 1) ? "" : "s", num_found);
}

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_Call(PyObject *func, PyObject *arg, PyObject *kw) {
    PyObject *result;
    ternaryfunc call = func->ob_type->tp_call;
    if (unlikely(!call))
        return PyObject_Call(func, arg, kw);
    if (unlikely(Py_EnterRecursiveCall((char*)" while calling a Python object")))
        return NULL;
    result = (*call)(func, arg, kw);
    Py_LeaveRecursiveCall();
    if (unlikely(!result) && unlikely(!PyErr_Occurred())) {
        PyErr_SetString(
            PyExc_SystemError,
            "NULL result without error in PyObject_Call");
    }
    return result;
}
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallMethO(PyObject *func, PyObject *arg) {
    PyObject *self, *result;
    PyCFunction cfunc;
    cfunc = PyCFunction_GET_FUNCTION(func);
    self = PyCFunction_GET_SELF(func);
    if (unlikely(Py_EnterRecursiveCall((char*)" while calling a Python object")))
        return NULL;
    result = cfunc(self, arg);
    Py_LeaveRecursiveCall();
    if (unlikely(!result) && unlikely(!PyErr_Occurred())) {
        PyErr_SetString(
            PyExc_SystemError,
            "NULL result without error in PyObject_Call");
    }
    return result;
}
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static PyObject* __Pyx__PyObject_CallOneArg(PyObject *func, PyObject *arg) {
    PyObject *result;
    PyObject *args = PyTuple_New(1);
    if (unlikely(!args)) return NULL;
    Py_INCREF(arg);
    PyTuple_SET_ITEM(args, 0, arg);
    result = __Pyx_PyObject_Call(func, args, NULL);
    Py_DECREF(args);
    return result;
}
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallOneArg(PyObject *func, PyObject *arg) {
#ifdef __Pyx_CyFunction_USED
    if (likely(PyCFunction_Check(func) || PyObject_TypeCheck(func, __pyx_CyFunctionType))) {
#else
    if (likely(PyCFunction_Check(func))) {
#endif
        if (likely(PyCFunction_GET_FLAGS(func) & METH_O)) {
            return __Pyx_PyObject_CallMethO(func, arg);
        }
    }
    return __Pyx__PyObject_CallOneArg(func, arg);
}
#else
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallOneArg(PyObject *func, PyObject *arg) {
    PyObject* args = PyTuple_Pack(1, arg);
    return (likely(args)) ? __Pyx_PyObject_Call(func, args, NULL) : NULL;
}
#endif

static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Generic(PyObject *o, PyObject* j) {
    PyObject *r;
    if (!j) return NULL;
    r = PyObject_GetItem(o, j);
    Py_DECREF(j);
    return r;
}
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_List_Fast(PyObject *o, Py_ssize_t i,
                                                              int wraparound, int boundscheck) {
#if CYTHON_COMPILING_IN_CPYTHON
    if (wraparound & unlikely(i < 0)) i += PyList_GET_SIZE(o);
    if ((!boundscheck) || likely((0 <= i) & (i < PyList_GET_SIZE(o)))) {
        PyObject *r = PyList_GET_ITEM(o, i);
        Py_INCREF(r);
        return r;
    }
    return __Pyx_GetItemInt_Generic(o, PyInt_FromSsize_t(i));
#else
    return PySequence_GetItem(o, i);
#endif
}
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Tuple_Fast(PyObject *o, Py_ssize_t i,
                                                              int wraparound, int boundscheck) {
#if CYTHON_COMPILING_IN_CPYTHON
    if (wraparound & unlikely(i < 0)) i += PyTuple_GET_SIZE(o);
    if ((!boundscheck) || likely((0 <= i) & (i < PyTuple_GET_SIZE(o)))) {
        PyObject *r = PyTuple_GET_ITEM(o, i);
        Py_INCREF(r);
        return r;
    }
    return __Pyx_GetItemInt_Generic(o, PyInt_FromSsize_t(i));
#else
    return PySequence_GetItem(o, i);
#endif
}
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Fast(PyObject *o, Py_ssize_t i,
                                                     int is_list, int wraparound, int boundscheck) {
#if CYTHON_COMPILING_IN_CPYTHON
    if (is_list || PyList_CheckExact(o)) {
        Py_ssize_t n = ((!wraparound) | likely(i >= 0)) ? i : i + PyList_GET_SIZE(o);
        if ((!boundscheck) || (likely((n >= 0) & (n < PyList_GET_SIZE(o))))) {
            PyObject *r = PyList_GET_ITEM(o, n);
            Py_INCREF(r);
            return r;
        }
    }
    else if (PyTuple_CheckExact(o)) {
        Py_ssize_t n = ((!wraparound) | likely(i >= 0)) ? i : i + PyTuple_GET_SIZE(o);
        if ((!boundscheck) || likely((n >= 0) & (n < PyTuple_GET_SIZE(o)))) {
            PyObject *r = PyTuple_GET_ITEM(o, n);
            Py_INCREF(r);
            return r;
        }
    } else {
        PySequenceMethods *m = Py_TYPE(o)->tp_as_sequence;
        if (likely(m && m->sq_item)) {
            if (wraparound && unlikely(i < 0) && likely(m->sq_length)) {
                Py_ssize_t l = m->sq_length(o);
                if (likely(l >= 0)) {
                    i += l;
                } else {
                    if (PyErr_ExceptionMatches(PyExc_OverflowError))
                        PyErr_Clear();
                    else
                        return NULL;
                }
            }
            return m->sq_item(o, i);
        }
    }
#else
    if (is_list || PySequence_Check(o)) {
        return PySequence_GetItem(o, i);
    }
#endif
    return __Pyx_GetItemInt_Generic(o, PyInt_FromSsize_t(i));
}

static CYTHON_INLINE PyObject *__Pyx_GetModuleGlobalName(PyObject *name) {
    PyObject *result;
#if CYTHON_COMPILING_IN_CPYTHON
    result = PyDict_GetItem(__pyx_d, name);
    if (likely(result)) {
        Py_INCREF(result);
    } else {
#else
    result = PyObject_GetItem(__pyx_d, name);
    if (!result) {
        PyErr_Clear();
#endif
        result = __Pyx_GetBuiltinName(name);
    }
    return result;
}

static CYTHON_INLINE int __Pyx_SetItemInt_Generic(PyObject *o, PyObject *j, PyObject *v) {
    int r;
    if (!j) return -1;
    r = PyObject_SetItem(o, j, v);
    Py_DECREF(j);
    return r;
}
static CYTHON_INLINE int __Pyx_SetItemInt_Fast(PyObject *o, Py_ssize_t i, PyObject *v,
                                               int is_list, int wraparound, int boundscheck) {
#if CYTHON_COMPILING_IN_CPYTHON
    if (is_list || PyList_CheckExact(o)) {
        Py_ssize_t n = (!wraparound) ? i : ((likely(i >= 0)) ? i : i + PyList_GET_SIZE(o));
        if ((!boundscheck) || likely((n >= 0) & (n < PyList_GET_SIZE(o)))) {
            PyObject* old = PyList_GET_ITEM(o, n);
            Py_INCREF(v);
            PyList_SET_ITEM(o, n, v);
            Py_DECREF(old);
            return 1;
        }
    } else {
        PySequenceMethods *m = Py_TYPE(o)->tp_as_sequence;
        if (likely(m && m->sq_ass_item)) {
            if (wraparound && unlikely(i < 0) && likely(m->sq_length)) {
                Py_ssize_t l = m->sq_length(o);
                if (likely(l >= 0)) {
                    i += l;
                } else {
                    if (PyErr_ExceptionMatches(PyExc_OverflowError))
                        PyErr_Clear();
                    else
                        return -1;
                }
            }
            return m->sq_ass_item(o, i, v);
        }
    }
#else
#if CYTHON_COMPILING_IN_PYPY
    if (is_list || (PySequence_Check(o) && !PyDict_Check(o))) {
#else
    if (is_list || PySequence_Check(o)) {
#endif
        return PySequence_SetItem(o, i, v);
    }
#endif
    return __Pyx_SetItemInt_Generic(o, PyInt_FromSsize_t(i), v);
}

#if !CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyBytes_Join(PyObject* sep, PyObject* values) {
    return PyObject_CallMethodObjArgs(sep, __pyx_n_s_join, values, NULL);
}
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallNoArg(PyObject *func) {
#ifdef __Pyx_CyFunction_USED
    if (likely(PyCFunction_Check(func) || PyObject_TypeCheck(func, __pyx_CyFunctionType))) {
#else
    if (likely(PyCFunction_Check(func))) {
#endif
        if (likely(PyCFunction_GET_FLAGS(func) & METH_NOARGS)) {
            return __Pyx_PyObject_CallMethO(func, NULL);
        }
    }
    return __Pyx_PyObject_Call(func, __pyx_empty_tuple, NULL);
}
#endif

static CYTHON_INLINE int __Pyx_TypeTest(PyObject *obj, PyTypeObject *type) {
    if (unlikely(!type)) {
        PyErr_SetString(PyExc_SystemError, "Missing type object");
        return 0;
    }
    if (likely(PyObject_TypeCheck(obj, type)))
        return 1;
    PyErr_Format(PyExc_TypeError, "Cannot convert %.200s to %.200s",
                 Py_TYPE(obj)->tp_name, type->tp_name);
    return 0;
}

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE void __Pyx_crop_slice(Py_ssize_t* _start, Py_ssize_t* _stop, Py_ssize_t* _length) {
    Py_ssize_t start = *_start, stop = *_stop, length = *_length;
    if (start < 0) {
        start += length;
        if (start < 0)
            start = 0;
    }
    if (stop < 0)
        stop += length;
    else if (stop > length)
        stop = length;
    *_length = stop - start;
    *_start = start;
    *_stop = stop;
}
static CYTHON_INLINE void __Pyx_copy_object_array(PyObject** CYTHON_RESTRICT src, PyObject** CYTHON_RESTRICT dest, Py_ssize_t length) {
    PyObject *v;
    Py_ssize_t i;
    for (i = 0; i < length; i++) {
        v = dest[i] = src[i];
        Py_INCREF(v);
    }
}
static CYTHON_INLINE PyObject* __Pyx_PyList_GetSlice(
            PyObject* src, Py_ssize_t start, Py_ssize_t stop) {
    PyObject* dest;
    Py_ssize_t length = PyList_GET_SIZE(src);
    __Pyx_crop_slice(&start, &stop, &length);
    if (unlikely(length <= 0))
        return PyList_New(0);
    dest = PyList_New(length);
    if (unlikely(!dest))
        return NULL;
    __Pyx_copy_object_array(
        ((PyListObject*)src)->ob_item + start,
        ((PyListObject*)dest)->ob_item,
        length);
    return dest;
}
static CYTHON_INLINE PyObject* __Pyx_PyTuple_GetSlice(
            PyObject* src, Py_ssize_t start, Py_ssize_t stop) {
    PyObject* dest;
    Py_ssize_t length = PyTuple_GET_SIZE(src);
    __Pyx_crop_slice(&start, &stop, &length);
    if (unlikely(length <= 0))
        return PyTuple_New(0);
    dest = PyTuple_New(length);
    if (unlikely(!dest))
        return NULL;
    __Pyx_copy_object_array(
        ((PyTupleObject*)src)->ob_item + start,
        ((PyTupleObject*)dest)->ob_item,
        length);
    return dest;
}
#endif

static CYTHON_INLINE void __Pyx_ErrRestore(PyObject *type, PyObject *value, PyObject *tb) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyObject *tmp_type, *tmp_value, *tmp_tb;
    PyThreadState *tstate = PyThreadState_GET();
    tmp_type = tstate->curexc_type;
    tmp_value = tstate->curexc_value;
    tmp_tb = tstate->curexc_traceback;
    tstate->curexc_type = type;
    tstate->curexc_value = value;
    tstate->curexc_traceback = tb;
    Py_XDECREF(tmp_type);
    Py_XDECREF(tmp_value);
    Py_XDECREF(tmp_tb);
#else
    PyErr_Restore(type, value, tb);
#endif
}
static CYTHON_INLINE void __Pyx_ErrFetch(PyObject **type, PyObject **value, PyObject **tb) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyThreadState *tstate = PyThreadState_GET();
    *type = tstate->curexc_type;
    *value = tstate->curexc_value;
    *tb = tstate->curexc_traceback;
    tstate->curexc_type = 0;
    tstate->curexc_value = 0;
    tstate->curexc_traceback = 0;
#else
    PyErr_Fetch(type, value, tb);
#endif
}

#if PY_MAJOR_VERSION < 3
static void __Pyx_Raise(PyObject *type, PyObject *value, PyObject *tb,
                        CYTHON_UNUSED PyObject *cause) {
    Py_XINCREF(type);
    if (!value || value == Py_None)
        value = NULL;
    else
        Py_INCREF(value);
    if (!tb || tb == Py_None)
        tb = NULL;
    else {
        Py_INCREF(tb);
        if (!PyTraceBack_Check(tb)) {
            PyErr_SetString(PyExc_TypeError,
                "raise: arg 3 must be a traceback or None");
            goto raise_error;
        }
    }
    if (PyType_Check(type)) {
#if CYTHON_COMPILING_IN_PYPY
        if (!value) {
            Py_INCREF(Py_None);
            value = Py_None;
        }
#endif
        PyErr_NormalizeException(&type, &value, &tb);
    } else {
        if (value) {
            PyErr_SetString(PyExc_TypeError,
                "instance exception may not have a separate value");
            goto raise_error;
        }
        value = type;
        type = (PyObject*) Py_TYPE(type);
        Py_INCREF(type);
        if (!PyType_IsSubtype((PyTypeObject *)type, (PyTypeObject *)PyExc_BaseException)) {
            PyErr_SetString(PyExc_TypeError,
                "raise: exception class must be a subclass of BaseException");
            goto raise_error;
        }
    }
    __Pyx_ErrRestore(type, value, tb);
    return;
raise_error:
    Py_XDECREF(value);
    Py_XDECREF(type);
    Py_XDECREF(tb);
    return;
}
#else
static void __Pyx_Raise(PyObject *type, PyObject *value, PyObject *tb, PyObject *cause) {
    PyObject* owned_instance = NULL;
    if (tb == Py_None) {
        tb = 0;
    } else if (tb && !PyTraceBack_Check(tb)) {
        PyErr_SetString(PyExc_TypeError,
            "raise: arg 3 must be a traceback or None");
        goto bad;
    }
    if (value == Py_None)
        value = 0;
    if (PyExceptionInstance_Check(type)) {
        if (value) {
            PyErr_SetString(PyExc_TypeError,
                "instance exception may not have a separate value");
            goto bad;
        }
        value = type;
        type = (PyObject*) Py_TYPE(value);
    } else if (PyExceptionClass_Check(type)) {
        PyObject *instance_class = NULL;
        if (value && PyExceptionInstance_Check(value)) {
            instance_class = (PyObject*) Py_TYPE(value);
            if (instance_class != type) {
                if (PyObject_IsSubclass(instance_class, type)) {
                    type = instance_class;
                } else {
                    instance_class = NULL;
                }
            }
        }
        if (!instance_class) {
            PyObject *args;
            if (!value)
                args = PyTuple_New(0);
            else if (PyTuple_Check(value)) {
                Py_INCREF(value);
                args = value;
            } else
                args = PyTuple_Pack(1, value);
            if (!args)
                goto bad;
            owned_instance = PyObject_Call(type, args, NULL);
            Py_DECREF(args);
            if (!owned_instance)
                goto bad;
            value = owned_instance;
            if (!PyExceptionInstance_Check(value)) {
                PyErr_Format(PyExc_TypeError,
                             "calling %R should have returned an instance of "
                             "BaseException, not %R",
                             type, Py_TYPE(value));
                goto bad;
            }
        }
    } else {
        PyErr_SetString(PyExc_TypeError,
            "raise: exception class must be a subclass of BaseException");
        goto bad;
    }
#if PY_VERSION_HEX >= 0x03030000
    if (cause) {
#else
    if (cause && cause != Py_None) {
#endif
        PyObject *fixed_cause;
        if (cause == Py_None) {
            fixed_cause = NULL;
        } else if (PyExceptionClass_Check(cause)) {
            fixed_cause = PyObject_CallObject(cause, NULL);
            if (fixed_cause == NULL)
                goto bad;
        } else if (PyExceptionInstance_Check(cause)) {
            fixed_cause = cause;
            Py_INCREF(fixed_cause);
        } else {
            PyErr_SetString(PyExc_TypeError,
                            "exception causes must derive from "
                            "BaseException");
            goto bad;
        }
        PyException_SetCause(value, fixed_cause);
    }
    PyErr_SetObject(type, value);
    if (tb) {
        PyThreadState *tstate = PyThreadState_GET();
        PyObject* tmp_tb = tstate->curexc_traceback;
        if (tb != tmp_tb) {
            Py_INCREF(tb);
            tstate->curexc_traceback = tb;
            Py_XDECREF(tmp_tb);
        }
    }
bad:
    Py_XDECREF(owned_instance);
    return;
}
#endif

static CYTHON_INLINE int __Pyx_PyBytes_Equals(PyObject* s1, PyObject* s2, int equals) {
#if CYTHON_COMPILING_IN_PYPY
    return PyObject_RichCompareBool(s1, s2, equals);
#else
    if (s1 == s2) {
        return (equals == Py_EQ);
    } else if (PyBytes_CheckExact(s1) & PyBytes_CheckExact(s2)) {
        const char *ps1, *ps2;
        Py_ssize_t length = PyBytes_GET_SIZE(s1);
        if (length != PyBytes_GET_SIZE(s2))
            return (equals == Py_NE);
        ps1 = PyBytes_AS_STRING(s1);
        ps2 = PyBytes_AS_STRING(s2);
        if (ps1[0] != ps2[0]) {
            return (equals == Py_NE);
        } else if (length == 1) {
            return (equals == Py_EQ);
        } else {
            int result = memcmp(ps1, ps2, (size_t)length);
            return (equals == Py_EQ) ? (result == 0) : (result != 0);
        }
    } else if ((s1 == Py_None) & PyBytes_CheckExact(s2)) {
        return (equals == Py_NE);
    } else if ((s2 == Py_None) & PyBytes_CheckExact(s1)) {
        return (equals == Py_NE);
    } else {
        int result;
        PyObject* py_result = PyObject_RichCompare(s1, s2, equals);
        if (!py_result)
            return -1;
        result = __Pyx_PyObject_IsTrue(py_result);
        Py_DECREF(py_result);
        return result;
    }
#endif
}

static CYTHON_INLINE int __Pyx_PyUnicode_Equals(PyObject* s1, PyObject* s2, int equals) {
#if CYTHON_COMPILING_IN_PYPY
    return PyObject_RichCompareBool(s1, s2, equals);
#else
#if PY_MAJOR_VERSION < 3
    PyObject* owned_ref = NULL;
#endif
    int s1_is_unicode, s2_is_unicode;
    if (s1 == s2) {
        goto return_eq;
    }
    s1_is_unicode = PyUnicode_CheckExact(s1);
    s2_is_unicode = PyUnicode_CheckExact(s2);
#if PY_MAJOR_VERSION < 3
    if ((s1_is_unicode & (!s2_is_unicode)) && PyString_CheckExact(s2)) {
        owned_ref = PyUnicode_FromObject(s2);
        if (unlikely(!owned_ref))
            return -1;
        s2 = owned_ref;
        s2_is_unicode = 1;
    } else if ((s2_is_unicode & (!s1_is_unicode)) && PyString_CheckExact(s1)) {
        owned_ref = PyUnicode_FromObject(s1);
        if (unlikely(!owned_ref))
            return -1;
        s1 = owned_ref;
        s1_is_unicode = 1;
    } else if (((!s2_is_unicode) & (!s1_is_unicode))) {
        return __Pyx_PyBytes_Equals(s1, s2, equals);
    }
#endif
    if (s1_is_unicode & s2_is_unicode) {
        Py_ssize_t length;
        int kind;
        void *data1, *data2;
        if (unlikely(__Pyx_PyUnicode_READY(s1) < 0) || unlikely(__Pyx_PyUnicode_READY(s2) < 0))
            return -1;
        length = __Pyx_PyUnicode_GET_LENGTH(s1);
        if (length != __Pyx_PyUnicode_GET_LENGTH(s2)) {
            goto return_ne;
        }
        kind = __Pyx_PyUnicode_KIND(s1);
        if (kind != __Pyx_PyUnicode_KIND(s2)) {
            goto return_ne;
        }
        data1 = __Pyx_PyUnicode_DATA(s1);
        data2 = __Pyx_PyUnicode_DATA(s2);
        if (__Pyx_PyUnicode_READ(kind, data1, 0) != __Pyx_PyUnicode_READ(kind, data2, 0)) {
            goto return_ne;
        } else if (length == 1) {
            goto return_eq;
        } else {
            int result = memcmp(data1, data2, (size_t)(length * kind));
            #if PY_MAJOR_VERSION < 3
            Py_XDECREF(owned_ref);
            #endif
            return (equals == Py_EQ) ? (result == 0) : (result != 0);
        }
    } else if ((s1 == Py_None) & s2_is_unicode) {
        goto return_ne;
    } else if ((s2 == Py_None) & s1_is_unicode) {
        goto return_ne;
    } else {
        int result;
        PyObject* py_result = PyObject_RichCompare(s1, s2, equals);
        if (!py_result)
            return -1;
        result = __Pyx_PyObject_IsTrue(py_result);
        Py_DECREF(py_result);
        return result;
    }
return_eq:
    #if PY_MAJOR_VERSION < 3
    Py_XDECREF(owned_ref);
    #endif
    return (equals == Py_EQ);
return_ne:
    #if PY_MAJOR_VERSION < 3
    Py_XDECREF(owned_ref);
    #endif
    return (equals == Py_NE);
#endif
}

static void __Pyx_RaiseArgumentTypeInvalid(const char* name, PyObject *obj, PyTypeObject *type) {
    PyErr_Format(PyExc_TypeError,
        "Argument '%.200s' has incorrect type (expected %.200s, got %.200s)",
        name, type->tp_name, Py_TYPE(obj)->tp_name);
}
static CYTHON_INLINE int __Pyx_ArgTypeTest(PyObject *obj, PyTypeObject *type, int none_allowed,
    const char *name, int exact)
{
    if (unlikely(!type)) {
        PyErr_SetString(PyExc_SystemError, "Missing type object");
        return 0;
    }
    if (none_allowed && obj == Py_None) return 1;
    else if (exact) {
        if (likely(Py_TYPE(obj) == type)) return 1;
        #if PY_MAJOR_VERSION == 2
        else if ((type == &PyBaseString_Type) && likely(__Pyx_PyBaseString_CheckExact(obj))) return 1;
        #endif
    }
    else {
        if (likely(PyObject_TypeCheck(obj, type))) return 1;
    }
    __Pyx_RaiseArgumentTypeInvalid(name, obj, type);
    return 0;
}

static PyObject* __Pyx_PyObject_CallMethod1(PyObject* obj, PyObject* method_name, PyObject* arg) {
    PyObject *method, *result = NULL;
    method = __Pyx_PyObject_GetAttrStr(obj, method_name);
    if (unlikely(!method)) goto bad;
#if CYTHON_COMPILING_IN_CPYTHON
    if (likely(PyMethod_Check(method))) {
        PyObject *self = PyMethod_GET_SELF(method);
        if (likely(self)) {
            PyObject *args;
            PyObject *function = PyMethod_GET_FUNCTION(method);
            args = PyTuple_New(2);
            if (unlikely(!args)) goto bad;
            Py_INCREF(self);
            PyTuple_SET_ITEM(args, 0, self);
            Py_INCREF(arg);
            PyTuple_SET_ITEM(args, 1, arg);
            Py_INCREF(function);
            Py_DECREF(method); method = NULL;
            result = __Pyx_PyObject_Call(function, args, NULL);
            Py_DECREF(args);
            Py_DECREF(function);
            return result;
        }
    }
#endif
    result = __Pyx_PyObject_CallOneArg(method, arg);
bad:
    Py_XDECREF(method);
    return result;
}

static CYTHON_INLINE int __Pyx_PyObject_Append(PyObject* L, PyObject* x) {
    if (likely(PyList_CheckExact(L))) {
        if (unlikely(__Pyx_PyList_Append(L, x) < 0)) return -1;
    } else {
        PyObject* retval = __Pyx_PyObject_CallMethod1(L, __pyx_n_s_append, x);
        if (unlikely(!retval))
            return -1;
        Py_DECREF(retval);
    }
    return 0;
}

static void __Pyx_WriteUnraisable(const char *name, CYTHON_UNUSED int clineno,
                                  CYTHON_UNUSED int lineno, CYTHON_UNUSED const char *filename,
                                  int full_traceback) {
    PyObject *old_exc, *old_val, *old_tb;
    PyObject *ctx;
    __Pyx_ErrFetch(&old_exc, &old_val, &old_tb);
    if (full_traceback) {
        Py_XINCREF(old_exc);
        Py_XINCREF(old_val);
        Py_XINCREF(old_tb);
        __Pyx_ErrRestore(old_exc, old_val, old_tb);
        PyErr_PrintEx(1);
    }
    #if PY_MAJOR_VERSION < 3
    ctx = PyString_FromString(name);
    #else
    ctx = PyUnicode_FromString(name);
    #endif
    __Pyx_ErrRestore(old_exc, old_val, old_tb);
    if (!ctx) {
        PyErr_WriteUnraisable(Py_None);
    } else {
        PyErr_WriteUnraisable(ctx);
        Py_DECREF(ctx);
    }
}

static PyObject* __Pyx_ImportFrom(PyObject* module, PyObject* name) {
    PyObject* value = __Pyx_PyObject_GetAttrStr(module, name);
    if (unlikely(!value) && PyErr_ExceptionMatches(PyExc_AttributeError)) {
        PyErr_Format(PyExc_ImportError,
        #if PY_MAJOR_VERSION < 3
            "cannot import name %.230s", PyString_AS_STRING(name));
        #else
            "cannot import name %S", name);
        #endif
    }
    return value;
}

static CYTHON_INLINE void __Pyx_RaiseTooManyValuesError(Py_ssize_t expected) {
    PyErr_Format(PyExc_ValueError,
                 "too many values to unpack (expected %" CYTHON_FORMAT_SSIZE_T "d)", expected);
}

static CYTHON_INLINE void __Pyx_RaiseNeedMoreValuesError(Py_ssize_t index) {
    PyErr_Format(PyExc_ValueError,
                 "need more than %" CYTHON_FORMAT_SSIZE_T "d value%.1s to unpack",
                 index, (index == 1) ? "" : "s");
}

static CYTHON_INLINE int __Pyx_IterFinish(void) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyThreadState *tstate = PyThreadState_GET();
    PyObject* exc_type = tstate->curexc_type;
    if (unlikely(exc_type)) {
        if (likely(exc_type == PyExc_StopIteration) || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration)) {
            PyObject *exc_value, *exc_tb;
            exc_value = tstate->curexc_value;
            exc_tb = tstate->curexc_traceback;
            tstate->curexc_type = 0;
            tstate->curexc_value = 0;
            tstate->curexc_traceback = 0;
            Py_DECREF(exc_type);
            Py_XDECREF(exc_value);
            Py_XDECREF(exc_tb);
            return 0;
        } else {
            return -1;
        }
    }
    return 0;
#else
    if (unlikely(PyErr_Occurred())) {
        if (likely(PyErr_ExceptionMatches(PyExc_StopIteration))) {
            PyErr_Clear();
            return 0;
        } else {
            return -1;
        }
    }
    return 0;
#endif
}

static int __Pyx_IternextUnpackEndCheck(PyObject *retval, Py_ssize_t expected) {
    if (unlikely(retval)) {
        Py_DECREF(retval);
        __Pyx_RaiseTooManyValuesError(expected);
        return -1;
    } else {
        return __Pyx_IterFinish();
    }
    return 0;
}

static PyObject* __Pyx_PyObject_CallMethod0(PyObject* obj, PyObject* method_name) {
    PyObject *method, *result = NULL;
    method = __Pyx_PyObject_GetAttrStr(obj, method_name);
    if (unlikely(!method)) goto bad;
#if CYTHON_COMPILING_IN_CPYTHON
    if (likely(PyMethod_Check(method))) {
        PyObject *self = PyMethod_GET_SELF(method);
        if (likely(self)) {
            PyObject *function = PyMethod_GET_FUNCTION(method);
            result = __Pyx_PyObject_CallOneArg(function, self);
            Py_DECREF(method);
            return result;
        }
    }
#endif
    result = __Pyx_PyObject_CallNoArg(method);
    Py_DECREF(method);
bad:
    return result;
}

static CYTHON_INLINE void __Pyx_RaiseNoneNotIterableError(void) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
}

static void __Pyx_UnpackTupleError(PyObject *t, Py_ssize_t index) {
    if (t == Py_None) {
      __Pyx_RaiseNoneNotIterableError();
    } else if (PyTuple_GET_SIZE(t) < index) {
      __Pyx_RaiseNeedMoreValuesError(PyTuple_GET_SIZE(t));
    } else {
      __Pyx_RaiseTooManyValuesError(index);
    }
}

static CYTHON_INLINE int __Pyx_unpack_tuple2(PyObject* tuple, PyObject** pvalue1, PyObject** pvalue2,
                                             int is_tuple, int has_known_size, int decref_tuple) {
    Py_ssize_t index;
    PyObject *value1 = NULL, *value2 = NULL, *iter = NULL;
    if (!is_tuple && unlikely(!PyTuple_Check(tuple))) {
        iternextfunc iternext;
        iter = PyObject_GetIter(tuple);
        if (unlikely(!iter)) goto bad;
        if (decref_tuple) { Py_DECREF(tuple); tuple = NULL; }
        iternext = Py_TYPE(iter)->tp_iternext;
        value1 = iternext(iter); if (unlikely(!value1)) { index = 0; goto unpacking_failed; }
        value2 = iternext(iter); if (unlikely(!value2)) { index = 1; goto unpacking_failed; }
        if (!has_known_size && unlikely(__Pyx_IternextUnpackEndCheck(iternext(iter), 2))) goto bad;
        Py_DECREF(iter);
    } else {
        if (!has_known_size && unlikely(PyTuple_GET_SIZE(tuple) != 2)) {
            __Pyx_UnpackTupleError(tuple, 2);
            goto bad;
        }
#if CYTHON_COMPILING_IN_PYPY
        value1 = PySequence_ITEM(tuple, 0);
        if (unlikely(!value1)) goto bad;
        value2 = PySequence_ITEM(tuple, 1);
        if (unlikely(!value2)) goto bad;
#else
        value1 = PyTuple_GET_ITEM(tuple, 0);
        value2 = PyTuple_GET_ITEM(tuple, 1);
        Py_INCREF(value1);
        Py_INCREF(value2);
#endif
        if (decref_tuple) { Py_DECREF(tuple); }
    }
    *pvalue1 = value1;
    *pvalue2 = value2;
    return 0;
unpacking_failed:
    if (!has_known_size && __Pyx_IterFinish() == 0)
        __Pyx_RaiseNeedMoreValuesError(index);
bad:
    Py_XDECREF(iter);
    Py_XDECREF(value1);
    Py_XDECREF(value2);
    if (decref_tuple) { Py_XDECREF(tuple); }
    return -1;
}

static CYTHON_INLINE PyObject* __Pyx_dict_iterator(PyObject* iterable, int is_dict, PyObject* method_name,
                                                   Py_ssize_t* p_orig_length, int* p_source_is_dict) {
    is_dict = is_dict || likely(PyDict_CheckExact(iterable));
    *p_source_is_dict = is_dict;
#if !CYTHON_COMPILING_IN_PYPY
    if (is_dict) {
        *p_orig_length = PyDict_Size(iterable);
        Py_INCREF(iterable);
        return iterable;
    }
#endif
    *p_orig_length = 0;
    if (method_name) {
        PyObject* iter;
        iterable = __Pyx_PyObject_CallMethod0(iterable, method_name);
        if (!iterable)
            return NULL;
#if !CYTHON_COMPILING_IN_PYPY
        if (PyTuple_CheckExact(iterable) || PyList_CheckExact(iterable))
            return iterable;
#endif
        iter = PyObject_GetIter(iterable);
        Py_DECREF(iterable);
        return iter;
    }
    return PyObject_GetIter(iterable);
}
static CYTHON_INLINE int __Pyx_dict_iter_next(PyObject* iter_obj, Py_ssize_t orig_length, Py_ssize_t* ppos,
                                              PyObject** pkey, PyObject** pvalue, PyObject** pitem, int source_is_dict) {
    PyObject* next_item;
#if !CYTHON_COMPILING_IN_PYPY
    if (source_is_dict) {
        PyObject *key, *value;
        if (unlikely(orig_length != PyDict_Size(iter_obj))) {
            PyErr_SetString(PyExc_RuntimeError, "dictionary changed size during iteration");
            return -1;
        }
        if (unlikely(!PyDict_Next(iter_obj, ppos, &key, &value))) {
            return 0;
        }
        if (pitem) {
            PyObject* tuple = PyTuple_New(2);
            if (unlikely(!tuple)) {
                return -1;
            }
            Py_INCREF(key);
            Py_INCREF(value);
            PyTuple_SET_ITEM(tuple, 0, key);
            PyTuple_SET_ITEM(tuple, 1, value);
            *pitem = tuple;
        } else {
            if (pkey) {
                Py_INCREF(key);
                *pkey = key;
            }
            if (pvalue) {
                Py_INCREF(value);
                *pvalue = value;
            }
        }
        return 1;
    } else if (PyTuple_CheckExact(iter_obj)) {
        Py_ssize_t pos = *ppos;
        if (unlikely(pos >= PyTuple_GET_SIZE(iter_obj))) return 0;
        *ppos = pos + 1;
        next_item = PyTuple_GET_ITEM(iter_obj, pos);
        Py_INCREF(next_item);
    } else if (PyList_CheckExact(iter_obj)) {
        Py_ssize_t pos = *ppos;
        if (unlikely(pos >= PyList_GET_SIZE(iter_obj))) return 0;
        *ppos = pos + 1;
        next_item = PyList_GET_ITEM(iter_obj, pos);
        Py_INCREF(next_item);
    } else
#endif
    {
        next_item = PyIter_Next(iter_obj);
        if (unlikely(!next_item)) {
            return __Pyx_IterFinish();
        }
    }
    if (pitem) {
        *pitem = next_item;
    } else if (pkey && pvalue) {
        if (__Pyx_unpack_tuple2(next_item, pkey, pvalue, source_is_dict, source_is_dict, 1))
            return -1;
    } else if (pkey) {
        *pkey = next_item;
    } else {
        *pvalue = next_item;
    }
    return 1;
}

static CYTHON_INLINE void __Pyx_RaiseUnboundLocalError(const char *varname) {
    PyErr_Format(PyExc_UnboundLocalError, "local variable '%s' referenced before assignment", varname);
}

static void __Pyx_call_next_tp_dealloc(PyObject* obj, destructor current_tp_dealloc) {
    PyTypeObject* type = Py_TYPE(obj);
    while (type && type->tp_dealloc != current_tp_dealloc)
        type = type->tp_base;
    while (type && type->tp_dealloc == current_tp_dealloc)
        type = type->tp_base;
    if (type)
        type->tp_dealloc(obj);
}

static int __Pyx_call_next_tp_traverse(PyObject* obj, visitproc v, void *a, traverseproc current_tp_traverse) {
    PyTypeObject* type = Py_TYPE(obj);
    while (type && type->tp_traverse != current_tp_traverse)
        type = type->tp_base;
    while (type && type->tp_traverse == current_tp_traverse)
        type = type->tp_base;
    if (type && type->tp_traverse)
        return type->tp_traverse(obj, v, a);
    return 0;
}

static void __Pyx_call_next_tp_clear(PyObject* obj, inquiry current_tp_clear) {
    PyTypeObject* type = Py_TYPE(obj);
    while (type && type->tp_clear != current_tp_clear)
        type = type->tp_base;
    while (type && type->tp_clear == current_tp_clear)
        type = type->tp_base;
    if (type && type->tp_clear)
        type->tp_clear(obj);
}

static void* __Pyx_GetVtable(PyObject *dict) {
    void* ptr;
    PyObject *ob = PyObject_GetItem(dict, __pyx_n_s_pyx_vtable);
    if (!ob)
        goto bad;
#if PY_VERSION_HEX >= 0x02070000
    ptr = PyCapsule_GetPointer(ob, 0);
#else
    ptr = PyCObject_AsVoidPtr(ob);
#endif
    if (!ptr && !PyErr_Occurred())
        PyErr_SetString(PyExc_RuntimeError, "invalid vtable found for imported type");
    Py_DECREF(ob);
    return ptr;
bad:
    Py_XDECREF(ob);
    return NULL;
}

static int __Pyx_SetVtable(PyObject *dict, void *vtable) {
#if PY_VERSION_HEX >= 0x02070000
    PyObject *ob = PyCapsule_New(vtable, 0, 0);
#else
    PyObject *ob = PyCObject_FromVoidPtr(vtable, 0);
#endif
    if (!ob)
        goto bad;
    if (PyDict_SetItem(dict, __pyx_n_s_pyx_vtable, ob) < 0)
        goto bad;
    Py_DECREF(ob);
    return 0;
bad:
    Py_XDECREF(ob);
    return -1;
}

static PyObject *__Pyx_CalculateMetaclass(PyTypeObject *metaclass, PyObject *bases) {
    Py_ssize_t i, nbases = PyTuple_GET_SIZE(bases);
    for (i=0; i < nbases; i++) {
        PyTypeObject *tmptype;
        PyObject *tmp = PyTuple_GET_ITEM(bases, i);
        tmptype = Py_TYPE(tmp);
#if PY_MAJOR_VERSION < 3
        if (tmptype == &PyClass_Type)
            continue;
#endif
        if (!metaclass) {
            metaclass = tmptype;
            continue;
        }
        if (PyType_IsSubtype(metaclass, tmptype))
            continue;
        if (PyType_IsSubtype(tmptype, metaclass)) {
            metaclass = tmptype;
            continue;
        }
        PyErr_SetString(PyExc_TypeError,
                        "metaclass conflict: "
                        "the metaclass of a derived class "
                        "must be a (non-strict) subclass "
                        "of the metaclasses of all its bases");
        return NULL;
    }
    if (!metaclass) {
#if PY_MAJOR_VERSION < 3
        metaclass = &PyClass_Type;
#else
        metaclass = &PyType_Type;
#endif
    }
    Py_INCREF((PyObject*) metaclass);
    return (PyObject*) metaclass;
}

static PyObject *__Pyx_Py3MetaclassPrepare(PyObject *metaclass, PyObject *bases, PyObject *name,
                                           PyObject *qualname, PyObject *mkw, PyObject *modname, PyObject *doc) {
    PyObject *ns;
    if (metaclass) {
        PyObject *prep = __Pyx_PyObject_GetAttrStr(metaclass, __pyx_n_s_prepare);
        if (prep) {
            PyObject *pargs = PyTuple_Pack(2, name, bases);
            if (unlikely(!pargs)) {
                Py_DECREF(prep);
                return NULL;
            }
            ns = PyObject_Call(prep, pargs, mkw);
            Py_DECREF(prep);
            Py_DECREF(pargs);
        } else {
            if (unlikely(!PyErr_ExceptionMatches(PyExc_AttributeError)))
                return NULL;
            PyErr_Clear();
            ns = PyDict_New();
        }
    } else {
        ns = PyDict_New();
    }
    if (unlikely(!ns))
        return NULL;
    if (unlikely(PyObject_SetItem(ns, __pyx_n_s_module, modname) < 0)) goto bad;
    if (unlikely(PyObject_SetItem(ns, __pyx_n_s_qualname, qualname) < 0)) goto bad;
    if (unlikely(doc && PyObject_SetItem(ns, __pyx_n_s_doc, doc) < 0)) goto bad;
    return ns;
bad:
    Py_DECREF(ns);
    return NULL;
}
static PyObject *__Pyx_Py3ClassCreate(PyObject *metaclass, PyObject *name, PyObject *bases,
                                      PyObject *dict, PyObject *mkw,
                                      int calculate_metaclass, int allow_py2_metaclass) {
    PyObject *result, *margs;
    PyObject *owned_metaclass = NULL;
    if (allow_py2_metaclass) {
        owned_metaclass = PyObject_GetItem(dict, __pyx_n_s_metaclass);
        if (owned_metaclass) {
            metaclass = owned_metaclass;
        } else if (likely(PyErr_ExceptionMatches(PyExc_KeyError))) {
            PyErr_Clear();
        } else {
            return NULL;
        }
    }
    if (calculate_metaclass && (!metaclass || PyType_Check(metaclass))) {
        metaclass = __Pyx_CalculateMetaclass((PyTypeObject*) metaclass, bases);
        Py_XDECREF(owned_metaclass);
        if (unlikely(!metaclass))
            return NULL;
        owned_metaclass = metaclass;
    }
    margs = PyTuple_Pack(3, name, bases, dict);
    if (unlikely(!margs)) {
        result = NULL;
    } else {
        result = PyObject_Call(metaclass, margs, mkw);
        Py_DECREF(margs);
    }
    Py_XDECREF(owned_metaclass);
    return result;
}

static PyObject *__Pyx_GetNameInClass(PyObject *nmspace, PyObject *name) {
    PyObject *result;
    result = __Pyx_PyObject_GetAttrStr(nmspace, name);
    if (!result)
        result = __Pyx_GetModuleGlobalName(name);
    return result;
}

static int __pyx_bisect_code_objects(__Pyx_CodeObjectCacheEntry* entries, int count, int code_line) {
    int start = 0, mid = 0, end = count - 1;
    if (end >= 0 && code_line > entries[end].code_line) {
        return count;
    }
    while (start < end) {
        mid = (start + end) / 2;
        if (code_line < entries[mid].code_line) {
            end = mid;
        } else if (code_line > entries[mid].code_line) {
             start = mid + 1;
        } else {
            return mid;
        }
    }
    if (code_line <= entries[mid].code_line) {
        return mid;
    } else {
        return mid + 1;
    }
}
static PyCodeObject *__pyx_find_code_object(int code_line) {
    PyCodeObject* code_object;
    int pos;
    if (unlikely(!code_line) || unlikely(!__pyx_code_cache.entries)) {
        return NULL;
    }
    pos = __pyx_bisect_code_objects(__pyx_code_cache.entries, __pyx_code_cache.count, code_line);
    if (unlikely(pos >= __pyx_code_cache.count) || unlikely(__pyx_code_cache.entries[pos].code_line != code_line)) {
        return NULL;
    }
    code_object = __pyx_code_cache.entries[pos].code_object;
    Py_INCREF(code_object);
    return code_object;
}
static void __pyx_insert_code_object(int code_line, PyCodeObject* code_object) {
    int pos, i;
    __Pyx_CodeObjectCacheEntry* entries = __pyx_code_cache.entries;
    if (unlikely(!code_line)) {
        return;
    }
    if (unlikely(!entries)) {
        entries = (__Pyx_CodeObjectCacheEntry*)PyMem_Malloc(64*sizeof(__Pyx_CodeObjectCacheEntry));
        if (likely(entries)) {
            __pyx_code_cache.entries = entries;
            __pyx_code_cache.max_count = 64;
            __pyx_code_cache.count = 1;
            entries[0].code_line = code_line;
            entries[0].code_object = code_object;
            Py_INCREF(code_object);
        }
        return;
    }
    pos = __pyx_bisect_code_objects(__pyx_code_cache.entries, __pyx_code_cache.count, code_line);
    if ((pos < __pyx_code_cache.count) && unlikely(__pyx_code_cache.entries[pos].code_line == code_line)) {
        PyCodeObject* tmp = entries[pos].code_object;
        entries[pos].code_object = code_object;
        Py_DECREF(tmp);
        return;
    }
    if (__pyx_code_cache.count == __pyx_code_cache.max_count) {
        int new_max = __pyx_code_cache.max_count + 64;
        entries = (__Pyx_CodeObjectCacheEntry*)PyMem_Realloc(
            __pyx_code_cache.entries, (size_t)new_max*sizeof(__Pyx_CodeObjectCacheEntry));
        if (unlikely(!entries)) {
            return;
        }
        __pyx_code_cache.entries = entries;
        __pyx_code_cache.max_count = new_max;
    }
    for (i=__pyx_code_cache.count; i>pos; i--) {
        entries[i] = entries[i-1];
    }
    entries[pos].code_line = code_line;
    entries[pos].code_object = code_object;
    __pyx_code_cache.count++;
    Py_INCREF(code_object);
}

#include "compile.h"
#include "frameobject.h"
#include "traceback.h"
static PyCodeObject* __Pyx_CreateCodeObjectForTraceback(
            const char *funcname, int c_line,
            int py_line, const char *filename) {
    PyCodeObject *py_code = 0;
    PyObject *py_srcfile = 0;
    PyObject *py_funcname = 0;
    #if PY_MAJOR_VERSION < 3
    py_srcfile = PyString_FromString(filename);
    #else
    py_srcfile = PyUnicode_FromString(filename);
    #endif
    if (!py_srcfile) goto bad;
    if (c_line) {
        #if PY_MAJOR_VERSION < 3
        py_funcname = PyString_FromFormat( "%s (%s:%d)", funcname, __pyx_cfilenm, c_line);
        #else
        py_funcname = PyUnicode_FromFormat( "%s (%s:%d)", funcname, __pyx_cfilenm, c_line);
        #endif
    }
    else {
        #if PY_MAJOR_VERSION < 3
        py_funcname = PyString_FromString(funcname);
        #else
        py_funcname = PyUnicode_FromString(funcname);
        #endif
    }
    if (!py_funcname) goto bad;
    py_code = __Pyx_PyCode_New(
        0,
        0,
        0,
        0,
        0,
        __pyx_empty_bytes, /*PyObject *code,*/
        __pyx_empty_tuple, /*PyObject *consts,*/
        __pyx_empty_tuple, /*PyObject *names,*/
        __pyx_empty_tuple, /*PyObject *varnames,*/
        __pyx_empty_tuple, /*PyObject *freevars,*/
        __pyx_empty_tuple, /*PyObject *cellvars,*/
        py_srcfile,   /*PyObject *filename,*/
        py_funcname,  /*PyObject *name,*/
        py_line,
        __pyx_empty_bytes  /*PyObject *lnotab*/
    );
    Py_DECREF(py_srcfile);
    Py_DECREF(py_funcname);
    return py_code;
bad:
    Py_XDECREF(py_srcfile);
    Py_XDECREF(py_funcname);
    return NULL;
}
static void __Pyx_AddTraceback(const char *funcname, int c_line,
                               int py_line, const char *filename) {
    PyCodeObject *py_code = 0;
    PyFrameObject *py_frame = 0;
    py_code = __pyx_find_code_object(c_line ? c_line : py_line);
    if (!py_code) {
        py_code = __Pyx_CreateCodeObjectForTraceback(
            funcname, c_line, py_line, filename);
        if (!py_code) goto bad;
        __pyx_insert_code_object(c_line ? c_line : py_line, py_code);
    }
    py_frame = PyFrame_New(
        PyThreadState_GET(), /*PyThreadState *tstate,*/
        py_code,             /*PyCodeObject *code,*/
        __pyx_d,      /*PyObject *globals,*/
        0                    /*PyObject *locals*/
    );
    if (!py_frame) goto bad;
    py_frame->f_lineno = py_line;
    PyTraceBack_Here(py_frame);
bad:
    Py_XDECREF(py_code);
    Py_XDECREF(py_frame);
}

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_short(short value) {
    const short neg_one = (short) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
    if (is_unsigned) {
        if (sizeof(short) < sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(short) <= sizeof(unsigned long)) {
            return PyLong_FromUnsignedLong((unsigned long) value);
        } else if (sizeof(short) <= sizeof(unsigned long long)) {
            return PyLong_FromUnsignedLongLong((unsigned long long) value);
        }
    } else {
        if (sizeof(short) <= sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(short) <= sizeof(long long)) {
            return PyLong_FromLongLong((long long) value);
        }
    }
    {
        int one = 1; int little = (int)*(unsigned char *)&one;
        unsigned char *bytes = (unsigned char *)&value;
        return _PyLong_FromByteArray(bytes, sizeof(short),
                                     little, !is_unsigned);
    }
}

#define __PYX_VERIFY_RETURN_INT(target_type, func_type, func_value)       \
    {                                                                     \
        func_type value = func_value;                                     \
        if (sizeof(target_type) < sizeof(func_type)) {                    \
            if (unlikely(value != (func_type) (target_type) value)) {     \
                func_type zero = 0;                                       \
                if (is_unsigned && unlikely(value < zero))                \
                    goto raise_neg_overflow;                              \
                else                                                      \
                    goto raise_overflow;                                  \
            }                                                             \
        }                                                                 \
        return (target_type) value;                                       \
    }

#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
  #include "longintrepr.h"
 #endif
#endif

static CYTHON_INLINE short __Pyx_PyInt_As_short(PyObject *x) {
    const short neg_one = (short) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
#if PY_MAJOR_VERSION < 3
    if (likely(PyInt_Check(x))) {
        if (sizeof(short) < sizeof(long)) {
            __PYX_VERIFY_RETURN_INT(short, long, PyInt_AS_LONG(x))
        } else {
            long val = PyInt_AS_LONG(x);
            if (is_unsigned && unlikely(val < 0)) {
                goto raise_neg_overflow;
            }
            return (short) val;
        }
    } else
#endif
    if (likely(PyLong_Check(x))) {
        if (is_unsigned) {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(short, digit, ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (unlikely(Py_SIZE(x) < 0)) {
                goto raise_neg_overflow;
            }
            if (sizeof(short) <= sizeof(unsigned long)) {
                __PYX_VERIFY_RETURN_INT(short, unsigned long, PyLong_AsUnsignedLong(x))
            } else if (sizeof(short) <= sizeof(unsigned long long)) {
                __PYX_VERIFY_RETURN_INT(short, unsigned long long, PyLong_AsUnsignedLongLong(x))
            }
        } else {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(short,  digit, +(((PyLongObject*)x)->ob_digit[0]));
                case -1: __PYX_VERIFY_RETURN_INT(short, sdigit, -(sdigit) ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (sizeof(short) <= sizeof(long)) {
                __PYX_VERIFY_RETURN_INT(short, long, PyLong_AsLong(x))
            } else if (sizeof(short) <= sizeof(long long)) {
                __PYX_VERIFY_RETURN_INT(short, long long, PyLong_AsLongLong(x))
            }
        }
        {
#if CYTHON_COMPILING_IN_PYPY && !defined(_PyLong_AsByteArray)
            PyErr_SetString(PyExc_RuntimeError,
                            "_PyLong_AsByteArray() not available in PyPy, cannot convert large numbers");
#else
            short val;
            PyObject *v = __Pyx_PyNumber_Int(x);
 #if PY_MAJOR_VERSION < 3
            if (likely(v) && !PyLong_Check(v)) {
                PyObject *tmp = v;
                v = PyNumber_Long(tmp);
                Py_DECREF(tmp);
            }
 #endif
            if (likely(v)) {
                int one = 1; int is_little = (int)*(unsigned char *)&one;
                unsigned char *bytes = (unsigned char *)&val;
                int ret = _PyLong_AsByteArray((PyLongObject *)v,
                                              bytes, sizeof(val),
                                              is_little, !is_unsigned);
                Py_DECREF(v);
                if (likely(!ret))
                    return val;
            }
#endif
            return (short) -1;
        }
    } else {
        short val;
        PyObject *tmp = __Pyx_PyNumber_Int(x);
        if (!tmp) return (short) -1;
        val = __Pyx_PyInt_As_short(tmp);
        Py_DECREF(tmp);
        return val;
    }
raise_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "value too large to convert to short");
    return (short) -1;
raise_neg_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "can't convert negative value to short");
    return (short) -1;
}

static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level) {
    PyObject *empty_list = 0;
    PyObject *module = 0;
    PyObject *global_dict = 0;
    PyObject *empty_dict = 0;
    PyObject *list;
    #if PY_VERSION_HEX < 0x03030000
    PyObject *py_import;
    py_import = __Pyx_PyObject_GetAttrStr(__pyx_b, __pyx_n_s_import);
    if (!py_import)
        goto bad;
    #endif
    if (from_list)
        list = from_list;
    else {
        empty_list = PyList_New(0);
        if (!empty_list)
            goto bad;
        list = empty_list;
    }
    global_dict = PyModule_GetDict(__pyx_m);
    if (!global_dict)
        goto bad;
    empty_dict = PyDict_New();
    if (!empty_dict)
        goto bad;
    {
        #if PY_MAJOR_VERSION >= 3
        if (level == -1) {
            if (strchr(__Pyx_MODULE_NAME, '.')) {
                #if PY_VERSION_HEX < 0x03030000
                PyObject *py_level = PyInt_FromLong(1);
                if (!py_level)
                    goto bad;
                module = PyObject_CallFunctionObjArgs(py_import,
                    name, global_dict, empty_dict, list, py_level, NULL);
                Py_DECREF(py_level);
                #else
                module = PyImport_ImportModuleLevelObject(
                    name, global_dict, empty_dict, list, 1);
                #endif
                if (!module) {
                    if (!PyErr_ExceptionMatches(PyExc_ImportError))
                        goto bad;
                    PyErr_Clear();
                }
            }
            level = 0;
        }
        #endif
        if (!module) {
            #if PY_VERSION_HEX < 0x03030000
            PyObject *py_level = PyInt_FromLong(level);
            if (!py_level)
                goto bad;
            module = PyObject_CallFunctionObjArgs(py_import,
                name, global_dict, empty_dict, list, py_level, NULL);
            Py_DECREF(py_level);
            #else
            module = PyImport_ImportModuleLevelObject(
                name, global_dict, empty_dict, list, level);
            #endif
        }
    }
bad:
    #if PY_VERSION_HEX < 0x03030000
    Py_XDECREF(py_import);
    #endif
    Py_XDECREF(empty_list);
    Py_XDECREF(empty_dict);
    return module;
}

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_long(long value) {
    const long neg_one = (long) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
    if (is_unsigned) {
        if (sizeof(long) < sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(long) <= sizeof(unsigned long)) {
            return PyLong_FromUnsignedLong((unsigned long) value);
        } else if (sizeof(long) <= sizeof(unsigned long long)) {
            return PyLong_FromUnsignedLongLong((unsigned long long) value);
        }
    } else {
        if (sizeof(long) <= sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(long) <= sizeof(long long)) {
            return PyLong_FromLongLong((long long) value);
        }
    }
    {
        int one = 1; int little = (int)*(unsigned char *)&one;
        unsigned char *bytes = (unsigned char *)&value;
        return _PyLong_FromByteArray(bytes, sizeof(long),
                                     little, !is_unsigned);
    }
}

static CYTHON_INLINE int __Pyx_PyInt_As_int(PyObject *x) {
    const int neg_one = (int) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
#if PY_MAJOR_VERSION < 3
    if (likely(PyInt_Check(x))) {
        if (sizeof(int) < sizeof(long)) {
            __PYX_VERIFY_RETURN_INT(int, long, PyInt_AS_LONG(x))
        } else {
            long val = PyInt_AS_LONG(x);
            if (is_unsigned && unlikely(val < 0)) {
                goto raise_neg_overflow;
            }
            return (int) val;
        }
    } else
#endif
    if (likely(PyLong_Check(x))) {
        if (is_unsigned) {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(int, digit, ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (unlikely(Py_SIZE(x) < 0)) {
                goto raise_neg_overflow;
            }
            if (sizeof(int) <= sizeof(unsigned long)) {
                __PYX_VERIFY_RETURN_INT(int, unsigned long, PyLong_AsUnsignedLong(x))
            } else if (sizeof(int) <= sizeof(unsigned long long)) {
                __PYX_VERIFY_RETURN_INT(int, unsigned long long, PyLong_AsUnsignedLongLong(x))
            }
        } else {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(int,  digit, +(((PyLongObject*)x)->ob_digit[0]));
                case -1: __PYX_VERIFY_RETURN_INT(int, sdigit, -(sdigit) ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (sizeof(int) <= sizeof(long)) {
                __PYX_VERIFY_RETURN_INT(int, long, PyLong_AsLong(x))
            } else if (sizeof(int) <= sizeof(long long)) {
                __PYX_VERIFY_RETURN_INT(int, long long, PyLong_AsLongLong(x))
            }
        }
        {
#if CYTHON_COMPILING_IN_PYPY && !defined(_PyLong_AsByteArray)
            PyErr_SetString(PyExc_RuntimeError,
                            "_PyLong_AsByteArray() not available in PyPy, cannot convert large numbers");
#else
            int val;
            PyObject *v = __Pyx_PyNumber_Int(x);
 #if PY_MAJOR_VERSION < 3
            if (likely(v) && !PyLong_Check(v)) {
                PyObject *tmp = v;
                v = PyNumber_Long(tmp);
                Py_DECREF(tmp);
            }
 #endif
            if (likely(v)) {
                int one = 1; int is_little = (int)*(unsigned char *)&one;
                unsigned char *bytes = (unsigned char *)&val;
                int ret = _PyLong_AsByteArray((PyLongObject *)v,
                                              bytes, sizeof(val),
                                              is_little, !is_unsigned);
                Py_DECREF(v);
                if (likely(!ret))
                    return val;
            }
#endif
            return (int) -1;
        }
    } else {
        int val;
        PyObject *tmp = __Pyx_PyNumber_Int(x);
        if (!tmp) return (int) -1;
        val = __Pyx_PyInt_As_int(tmp);
        Py_DECREF(tmp);
        return val;
    }
raise_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "value too large to convert to int");
    return (int) -1;
raise_neg_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "can't convert negative value to int");
    return (int) -1;
}

static CYTHON_INLINE long __Pyx_PyInt_As_long(PyObject *x) {
    const long neg_one = (long) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
#if PY_MAJOR_VERSION < 3
    if (likely(PyInt_Check(x))) {
        if (sizeof(long) < sizeof(long)) {
            __PYX_VERIFY_RETURN_INT(long, long, PyInt_AS_LONG(x))
        } else {
            long val = PyInt_AS_LONG(x);
            if (is_unsigned && unlikely(val < 0)) {
                goto raise_neg_overflow;
            }
            return (long) val;
        }
    } else
#endif
    if (likely(PyLong_Check(x))) {
        if (is_unsigned) {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(long, digit, ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (unlikely(Py_SIZE(x) < 0)) {
                goto raise_neg_overflow;
            }
            if (sizeof(long) <= sizeof(unsigned long)) {
                __PYX_VERIFY_RETURN_INT(long, unsigned long, PyLong_AsUnsignedLong(x))
            } else if (sizeof(long) <= sizeof(unsigned long long)) {
                __PYX_VERIFY_RETURN_INT(long, unsigned long long, PyLong_AsUnsignedLongLong(x))
            }
        } else {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(long,  digit, +(((PyLongObject*)x)->ob_digit[0]));
                case -1: __PYX_VERIFY_RETURN_INT(long, sdigit, -(sdigit) ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (sizeof(long) <= sizeof(long)) {
                __PYX_VERIFY_RETURN_INT(long, long, PyLong_AsLong(x))
            } else if (sizeof(long) <= sizeof(long long)) {
                __PYX_VERIFY_RETURN_INT(long, long long, PyLong_AsLongLong(x))
            }
        }
        {
#if CYTHON_COMPILING_IN_PYPY && !defined(_PyLong_AsByteArray)
            PyErr_SetString(PyExc_RuntimeError,
                            "_PyLong_AsByteArray() not available in PyPy, cannot convert large numbers");
#else
            long val;
            PyObject *v = __Pyx_PyNumber_Int(x);
 #if PY_MAJOR_VERSION < 3
            if (likely(v) && !PyLong_Check(v)) {
                PyObject *tmp = v;
                v = PyNumber_Long(tmp);
                Py_DECREF(tmp);
            }
 #endif
            if (likely(v)) {
                int one = 1; int is_little = (int)*(unsigned char *)&one;
                unsigned char *bytes = (unsigned char *)&val;
                int ret = _PyLong_AsByteArray((PyLongObject *)v,
                                              bytes, sizeof(val),
                                              is_little, !is_unsigned);
                Py_DECREF(v);
                if (likely(!ret))
                    return val;
            }
#endif
            return (long) -1;
        }
    } else {
        long val;
        PyObject *tmp = __Pyx_PyNumber_Int(x);
        if (!tmp) return (long) -1;
        val = __Pyx_PyInt_As_long(tmp);
        Py_DECREF(tmp);
        return val;
    }
raise_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "value too large to convert to long");
    return (long) -1;
raise_neg_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "can't convert negative value to long");
    return (long) -1;
}

static int __Pyx_check_binary_version(void) {
    char ctversion[4], rtversion[4];
    PyOS_snprintf(ctversion, 4, "%d.%d", PY_MAJOR_VERSION, PY_MINOR_VERSION);
    PyOS_snprintf(rtversion, 4, "%s", Py_GetVersion());
    if (ctversion[0] != rtversion[0] || ctversion[2] != rtversion[2]) {
        char message[200];
        PyOS_snprintf(message, sizeof(message),
                      "compiletime version %s of module '%.100s' "
                      "does not match runtime version %s",
                      ctversion, __Pyx_MODULE_NAME, rtversion);
        return PyErr_WarnEx(NULL, message, 1);
    }
    return 0;
}

#ifndef __PYX_HAVE_RT_ImportModule
#define __PYX_HAVE_RT_ImportModule
static PyObject *__Pyx_ImportModule(const char *name) {
    PyObject *py_name = 0;
    PyObject *py_module = 0;
    py_name = __Pyx_PyIdentifier_FromString(name);
    if (!py_name)
        goto bad;
    py_module = PyImport_Import(py_name);
    Py_DECREF(py_name);
    return py_module;
bad:
    Py_XDECREF(py_name);
    return 0;
}
#endif

#ifndef __PYX_HAVE_RT_ImportType
#define __PYX_HAVE_RT_ImportType
static PyTypeObject *__Pyx_ImportType(const char *module_name, const char *class_name,
    size_t size, int strict)
{
    PyObject *py_module = 0;
    PyObject *result = 0;
    PyObject *py_name = 0;
    char warning[200];
    Py_ssize_t basicsize;
#ifdef Py_LIMITED_API
    PyObject *py_basicsize;
#endif
    py_module = __Pyx_ImportModule(module_name);
    if (!py_module)
        goto bad;
    py_name = __Pyx_PyIdentifier_FromString(class_name);
    if (!py_name)
        goto bad;
    result = PyObject_GetAttr(py_module, py_name);
    Py_DECREF(py_name);
    py_name = 0;
    Py_DECREF(py_module);
    py_module = 0;
    if (!result)
        goto bad;
    if (!PyType_Check(result)) {
        PyErr_Format(PyExc_TypeError,
            "%.200s.%.200s is not a type object",
            module_name, class_name);
        goto bad;
    }
#ifndef Py_LIMITED_API
    basicsize = ((PyTypeObject *)result)->tp_basicsize;
#else
    py_basicsize = PyObject_GetAttrString(result, "__basicsize__");
    if (!py_basicsize)
        goto bad;
    basicsize = PyLong_AsSsize_t(py_basicsize);
    Py_DECREF(py_basicsize);
    py_basicsize = 0;
    if (basicsize == (Py_ssize_t)-1 && PyErr_Occurred())
        goto bad;
#endif
    if (!strict && (size_t)basicsize > size) {
        PyOS_snprintf(warning, sizeof(warning),
            "%s.%s size changed, may indicate binary incompatibility",
            module_name, class_name);
        if (PyErr_WarnEx(NULL, warning, 0) < 0) goto bad;
    }
    else if ((size_t)basicsize != size) {
        PyErr_Format(PyExc_ValueError,
            "%.200s.%.200s has the wrong size, try recompiling",
            module_name, class_name);
        goto bad;
    }
    return (PyTypeObject *)result;
bad:
    Py_XDECREF(py_module);
    Py_XDECREF(result);
    return NULL;
}
#endif

static int __Pyx_InitStrings(__Pyx_StringTabEntry *t) {
    while (t->p) {
        #if PY_MAJOR_VERSION < 3
        if (t->is_unicode) {
            *t->p = PyUnicode_DecodeUTF8(t->s, t->n - 1, NULL);
        } else if (t->intern) {
            *t->p = PyString_InternFromString(t->s);
        } else {
            *t->p = PyString_FromStringAndSize(t->s, t->n - 1);
        }
        #else
        if (t->is_unicode | t->is_str) {
            if (t->intern) {
                *t->p = PyUnicode_InternFromString(t->s);
            } else if (t->encoding) {
                *t->p = PyUnicode_Decode(t->s, t->n - 1, t->encoding, NULL);
            } else {
                *t->p = PyUnicode_FromStringAndSize(t->s, t->n - 1);
            }
        } else {
            *t->p = PyBytes_FromStringAndSize(t->s, t->n - 1);
        }
        #endif
        if (!*t->p)
            return -1;
        ++t;
    }
    return 0;
}

static CYTHON_INLINE PyObject* __Pyx_PyUnicode_FromString(const char* c_str) {
    return __Pyx_PyUnicode_FromStringAndSize(c_str, (Py_ssize_t)strlen(c_str));
}
static CYTHON_INLINE char* __Pyx_PyObject_AsString(PyObject* o) {
    Py_ssize_t ignore;
    return __Pyx_PyObject_AsStringAndSize(o, &ignore);
}
static CYTHON_INLINE char* __Pyx_PyObject_AsStringAndSize(PyObject* o, Py_ssize_t *length) {
#if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII || __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT
    if (
#if PY_MAJOR_VERSION < 3 && __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
            __Pyx_sys_getdefaultencoding_not_ascii &&
#endif
            PyUnicode_Check(o)) {
#if PY_VERSION_HEX < 0x03030000
        char* defenc_c;
        PyObject* defenc = _PyUnicode_AsDefaultEncodedString(o, NULL);
        if (!defenc) return NULL;
        defenc_c = PyBytes_AS_STRING(defenc);
#if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
        {
            char* end = defenc_c + PyBytes_GET_SIZE(defenc);
            char* c;
            for (c = defenc_c; c < end; c++) {
                if ((unsigned char) (*c) >= 128) {
                    PyUnicode_AsASCIIString(o);
                    return NULL;
                }
            }
        }
#endif
        *length = PyBytes_GET_SIZE(defenc);
        return defenc_c;
#else
        if (__Pyx_PyUnicode_READY(o) == -1) return NULL;
#if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
        if (PyUnicode_IS_ASCII(o)) {
            *length = PyUnicode_GET_LENGTH(o);
            return PyUnicode_AsUTF8(o);
        } else {
            PyUnicode_AsASCIIString(o);
            return NULL;
        }
#else
        return PyUnicode_AsUTF8AndSize(o, length);
#endif
#endif
    } else
#endif
#if !CYTHON_COMPILING_IN_PYPY
    if (PyByteArray_Check(o)) {
        *length = PyByteArray_GET_SIZE(o);
        return PyByteArray_AS_STRING(o);
    } else
#endif
    {
        char* result;
        int r = PyBytes_AsStringAndSize(o, &result, length);
        if (unlikely(r < 0)) {
            return NULL;
        } else {
            return result;
        }
    }
}
static CYTHON_INLINE int __Pyx_PyObject_IsTrue(PyObject* x) {
   int is_true = x == Py_True;
   if (is_true | (x == Py_False) | (x == Py_None)) return is_true;
   else return PyObject_IsTrue(x);
}
static CYTHON_INLINE PyObject* __Pyx_PyNumber_Int(PyObject* x) {
  PyNumberMethods *m;
  const char *name = NULL;
  PyObject *res = NULL;
#if PY_MAJOR_VERSION < 3
  if (PyInt_Check(x) || PyLong_Check(x))
#else
  if (PyLong_Check(x))
#endif
    return Py_INCREF(x), x;
  m = Py_TYPE(x)->tp_as_number;
#if PY_MAJOR_VERSION < 3
  if (m && m->nb_int) {
    name = "int";
    res = PyNumber_Int(x);
  }
  else if (m && m->nb_long) {
    name = "long";
    res = PyNumber_Long(x);
  }
#else
  if (m && m->nb_int) {
    name = "int";
    res = PyNumber_Long(x);
  }
#endif
  if (res) {
#if PY_MAJOR_VERSION < 3
    if (!PyInt_Check(res) && !PyLong_Check(res)) {
#else
    if (!PyLong_Check(res)) {
#endif
      PyErr_Format(PyExc_TypeError,
                   "__%.4s__ returned non-%.4s (type %.200s)",
                   name, name, Py_TYPE(res)->tp_name);
      Py_DECREF(res);
      return NULL;
    }
  }
  else if (!PyErr_Occurred()) {
    PyErr_SetString(PyExc_TypeError,
                    "an integer is required");
  }
  return res;
}
static CYTHON_INLINE Py_ssize_t __Pyx_PyIndex_AsSsize_t(PyObject* b) {
  Py_ssize_t ival;
  PyObject *x;
#if PY_MAJOR_VERSION < 3
  if (likely(PyInt_CheckExact(b)))
      return PyInt_AS_LONG(b);
#endif
  if (likely(PyLong_CheckExact(b))) {
    #if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
     #if CYTHON_USE_PYLONG_INTERNALS
       switch (Py_SIZE(b)) {
       case -1: return -(sdigit)((PyLongObject*)b)->ob_digit[0];
       case  0: return 0;
       case  1: return ((PyLongObject*)b)->ob_digit[0];
       }
     #endif
    #endif
    return PyLong_AsSsize_t(b);
  }
  x = PyNumber_Index(b);
  if (!x) return -1;
  ival = PyInt_AsSsize_t(x);
  Py_DECREF(x);
  return ival;
}
static CYTHON_INLINE PyObject * __Pyx_PyInt_FromSize_t(size_t ival) {
    return PyInt_FromSize_t(ival);
}


#endif /* Py_PYTHON_H */
