/* Generated by Cython 0.21 */

#define PY_SSIZE_T_CLEAN
#ifndef CYTHON_USE_PYLONG_INTERNALS
#ifdef PYLONG_BITS_IN_DIGIT
#define CYTHON_USE_PYLONG_INTERNALS 0
#else
#include "pyconfig.h"
#ifdef PYLONG_BITS_IN_DIGIT
#define CYTHON_USE_PYLONG_INTERNALS 1
#else
#define CYTHON_USE_PYLONG_INTERNALS 0
#endif
#endif
#endif
#include "Python.h"
#ifndef Py_PYTHON_H
    #error Python headers needed to compile C extensions, please install development version of Python.
#elif PY_VERSION_HEX < 0x02060000 || (0x03000000 <= PY_VERSION_HEX && PY_VERSION_HEX < 0x03020000)
    #error Cython requires Python 2.6+ or Python 3.2+.
#else
#define CYTHON_ABI "0_21"
#include <stddef.h>
#ifndef offsetof
#define offsetof(type, member) ( (size_t) & ((type*)0) -> member )
#endif
#if !defined(WIN32) && !defined(MS_WINDOWS)
  #ifndef __stdcall
    #define __stdcall
  #endif
  #ifndef __cdecl
    #define __cdecl
  #endif
  #ifndef __fastcall
    #define __fastcall
  #endif
#endif
#ifndef DL_IMPORT
  #define DL_IMPORT(t) t
#endif
#ifndef DL_EXPORT
  #define DL_EXPORT(t) t
#endif
#ifndef PY_LONG_LONG
  #define PY_LONG_LONG LONG_LONG
#endif
#ifndef Py_HUGE_VAL
  #define Py_HUGE_VAL HUGE_VAL
#endif
#ifdef PYPY_VERSION
#define CYTHON_COMPILING_IN_PYPY 1
#define CYTHON_COMPILING_IN_CPYTHON 0
#else
#define CYTHON_COMPILING_IN_PYPY 0
#define CYTHON_COMPILING_IN_CPYTHON 1
#endif
#if CYTHON_COMPILING_IN_PYPY && PY_VERSION_HEX < 0x02070600
#define Py_OptimizeFlag 0
#endif
#define __PYX_BUILD_PY_SSIZE_T "n"
#define CYTHON_FORMAT_SSIZE_T "z"
#if PY_MAJOR_VERSION < 3
  #define __Pyx_BUILTIN_MODULE_NAME "__builtin__"
  #define __Pyx_PyCode_New(a, k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos) \
          PyCode_New(a+k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos)
  #define __Pyx_DefaultClassType PyClass_Type
#else
  #define __Pyx_BUILTIN_MODULE_NAME "builtins"
  #define __Pyx_PyCode_New(a, k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos) \
          PyCode_New(a, k, l, s, f, code, c, n, v, fv, cell, fn, name, fline, lnos)
  #define __Pyx_DefaultClassType PyType_Type
#endif
#if PY_MAJOR_VERSION >= 3
  #define Py_TPFLAGS_CHECKTYPES 0
  #define Py_TPFLAGS_HAVE_INDEX 0
#endif
#if PY_MAJOR_VERSION >= 3
  #define Py_TPFLAGS_HAVE_NEWBUFFER 0
#endif
#if PY_VERSION_HEX < 0x030400a1 && !defined(Py_TPFLAGS_HAVE_FINALIZE)
  #define Py_TPFLAGS_HAVE_FINALIZE 0
#endif
#if PY_VERSION_HEX > 0x03030000 && defined(PyUnicode_KIND)
  #define CYTHON_PEP393_ENABLED 1
  #define __Pyx_PyUnicode_READY(op)       (likely(PyUnicode_IS_READY(op)) ? \
                                              0 : _PyUnicode_Ready((PyObject *)(op)))
  #define __Pyx_PyUnicode_GET_LENGTH(u)   PyUnicode_GET_LENGTH(u)
  #define __Pyx_PyUnicode_READ_CHAR(u, i) PyUnicode_READ_CHAR(u, i)
  #define __Pyx_PyUnicode_KIND(u)         PyUnicode_KIND(u)
  #define __Pyx_PyUnicode_DATA(u)         PyUnicode_DATA(u)
  #define __Pyx_PyUnicode_READ(k, d, i)   PyUnicode_READ(k, d, i)
#else
  #define CYTHON_PEP393_ENABLED 0
  #define __Pyx_PyUnicode_READY(op)       (0)
  #define __Pyx_PyUnicode_GET_LENGTH(u)   PyUnicode_GET_SIZE(u)
  #define __Pyx_PyUnicode_READ_CHAR(u, i) ((Py_UCS4)(PyUnicode_AS_UNICODE(u)[i]))
  #define __Pyx_PyUnicode_KIND(u)         (sizeof(Py_UNICODE))
  #define __Pyx_PyUnicode_DATA(u)         ((void*)PyUnicode_AS_UNICODE(u))
  #define __Pyx_PyUnicode_READ(k, d, i)   ((void)(k), (Py_UCS4)(((Py_UNICODE*)d)[i]))
#endif
#if CYTHON_COMPILING_IN_PYPY
  #define __Pyx_PyUnicode_Concat(a, b)      PyNumber_Add(a, b)
  #define __Pyx_PyUnicode_ConcatSafe(a, b)  PyNumber_Add(a, b)
#else
  #define __Pyx_PyUnicode_Concat(a, b)      PyUnicode_Concat(a, b)
  #define __Pyx_PyUnicode_ConcatSafe(a, b)  ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
      PyNumber_Add(a, b) : __Pyx_PyUnicode_Concat(a, b))
#endif
#define __Pyx_PyString_FormatSafe(a, b)   ((unlikely((a) == Py_None)) ? PyNumber_Remainder(a, b) : __Pyx_PyString_Format(a, b))
#define __Pyx_PyUnicode_FormatSafe(a, b)  ((unlikely((a) == Py_None)) ? PyNumber_Remainder(a, b) : PyUnicode_Format(a, b))
#if PY_MAJOR_VERSION >= 3
  #define __Pyx_PyString_Format(a, b)  PyUnicode_Format(a, b)
#else
  #define __Pyx_PyString_Format(a, b)  PyString_Format(a, b)
#endif
#if PY_MAJOR_VERSION >= 3
  #define PyBaseString_Type            PyUnicode_Type
  #define PyStringObject               PyUnicodeObject
  #define PyString_Type                PyUnicode_Type
  #define PyString_Check               PyUnicode_Check
  #define PyString_CheckExact          PyUnicode_CheckExact
#endif
#if PY_MAJOR_VERSION >= 3
  #define __Pyx_PyBaseString_Check(obj) PyUnicode_Check(obj)
  #define __Pyx_PyBaseString_CheckExact(obj) PyUnicode_CheckExact(obj)
#else
  #define __Pyx_PyBaseString_Check(obj) (PyString_Check(obj) || PyUnicode_Check(obj))
  #define __Pyx_PyBaseString_CheckExact(obj) (PyString_CheckExact(obj) || PyUnicode_CheckExact(obj))
#endif
#ifndef PySet_CheckExact
  #define PySet_CheckExact(obj)        (Py_TYPE(obj) == &PySet_Type)
#endif
#define __Pyx_TypeCheck(obj, type) PyObject_TypeCheck(obj, (PyTypeObject *)type)
#if PY_MAJOR_VERSION >= 3
  #define PyIntObject                  PyLongObject
  #define PyInt_Type                   PyLong_Type
  #define PyInt_Check(op)              PyLong_Check(op)
  #define PyInt_CheckExact(op)         PyLong_CheckExact(op)
  #define PyInt_FromString             PyLong_FromString
  #define PyInt_FromUnicode            PyLong_FromUnicode
  #define PyInt_FromLong               PyLong_FromLong
  #define PyInt_FromSize_t             PyLong_FromSize_t
  #define PyInt_FromSsize_t            PyLong_FromSsize_t
  #define PyInt_AsLong                 PyLong_AsLong
  #define PyInt_AS_LONG                PyLong_AS_LONG
  #define PyInt_AsSsize_t              PyLong_AsSsize_t
  #define PyInt_AsUnsignedLongMask     PyLong_AsUnsignedLongMask
  #define PyInt_AsUnsignedLongLongMask PyLong_AsUnsignedLongLongMask
  #define PyNumber_Int                 PyNumber_Long
#endif
#if PY_MAJOR_VERSION >= 3
  #define PyBoolObject                 PyLongObject
#endif
#if PY_VERSION_HEX < 0x030200A4
  typedef long Py_hash_t;
  #define __Pyx_PyInt_FromHash_t PyInt_FromLong
  #define __Pyx_PyInt_AsHash_t   PyInt_AsLong
#else
  #define __Pyx_PyInt_FromHash_t PyInt_FromSsize_t
  #define __Pyx_PyInt_AsHash_t   PyInt_AsSsize_t
#endif
#if PY_MAJOR_VERSION >= 3
  #define PyMethod_New(func, self, klass) ((self) ? PyMethod_New(func, self) : PyInstanceMethod_New(func))
#endif
#ifndef CYTHON_INLINE
  #if defined(__GNUC__)
    #define CYTHON_INLINE __inline__
  #elif defined(_MSC_VER)
    #define CYTHON_INLINE __inline
  #elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
    #define CYTHON_INLINE inline
  #else
    #define CYTHON_INLINE
  #endif
#endif
#ifndef CYTHON_RESTRICT
  #if defined(__GNUC__)
    #define CYTHON_RESTRICT __restrict__
  #elif defined(_MSC_VER) && _MSC_VER >= 1400
    #define CYTHON_RESTRICT __restrict
  #elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
    #define CYTHON_RESTRICT restrict
  #else
    #define CYTHON_RESTRICT
  #endif
#endif
#ifdef NAN
#define __PYX_NAN() ((float) NAN)
#else
static CYTHON_INLINE float __PYX_NAN() {
  /* Initialize NaN. The sign is irrelevant, an exponent with all bits 1 and
   a nonzero mantissa means NaN. If the first bit in the mantissa is 1, it is
   a quiet NaN. */
  float value;
  memset(&value, 0xFF, sizeof(value));
  return value;
}
#endif
#ifdef __cplusplus
template<typename T>
void __Pyx_call_destructor(T* x) {
    x->~T();
}
#endif


#if PY_MAJOR_VERSION >= 3
  #define __Pyx_PyNumber_Divide(x,y)         PyNumber_TrueDivide(x,y)
  #define __Pyx_PyNumber_InPlaceDivide(x,y)  PyNumber_InPlaceTrueDivide(x,y)
#else
  #define __Pyx_PyNumber_Divide(x,y)         PyNumber_Divide(x,y)
  #define __Pyx_PyNumber_InPlaceDivide(x,y)  PyNumber_InPlaceDivide(x,y)
#endif

#ifndef __PYX_EXTERN_C
  #ifdef __cplusplus
    #define __PYX_EXTERN_C extern "C"
  #else
    #define __PYX_EXTERN_C extern
  #endif
#endif

#if defined(WIN32) || defined(MS_WINDOWS)
#define _USE_MATH_DEFINES
#endif
#include <math.h>
#define __PYX_HAVE__rmgpy__molecule__group
#define __PYX_HAVE_API__rmgpy__molecule__group
#include "string.h"
#include "stdio.h"
#include "stdlib.h"
#include "numpy/arrayobject.h"
#include "numpy/ufuncobject.h"
#ifdef _OPENMP
#include <omp.h>
#endif /* _OPENMP */

#ifdef PYREX_WITHOUT_ASSERTIONS
#define CYTHON_WITHOUT_ASSERTIONS
#endif

#ifndef CYTHON_UNUSED
# if defined(__GNUC__)
#   if !(defined(__cplusplus)) || (__GNUC__ > 3 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 4))
#     define CYTHON_UNUSED __attribute__ ((__unused__))
#   else
#     define CYTHON_UNUSED
#   endif
# elif defined(__ICC) || (defined(__INTEL_COMPILER) && !defined(_MSC_VER))
#   define CYTHON_UNUSED __attribute__ ((__unused__))
# else
#   define CYTHON_UNUSED
# endif
#endif
typedef struct {PyObject **p; char *s; const Py_ssize_t n; const char* encoding;
                const char is_unicode; const char is_str; const char intern; } __Pyx_StringTabEntry;

#define __PYX_DEFAULT_STRING_ENCODING_IS_ASCII 0
#define __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT 0
#define __PYX_DEFAULT_STRING_ENCODING ""
#define __Pyx_PyObject_FromString __Pyx_PyBytes_FromString
#define __Pyx_PyObject_FromStringAndSize __Pyx_PyBytes_FromStringAndSize
#define __Pyx_fits_Py_ssize_t(v, type, is_signed)  (    \
    (sizeof(type) < sizeof(Py_ssize_t))  ||             \
    (sizeof(type) > sizeof(Py_ssize_t) &&               \
          likely(v < (type)PY_SSIZE_T_MAX ||            \
                 v == (type)PY_SSIZE_T_MAX)  &&         \
          (!is_signed || likely(v > (type)PY_SSIZE_T_MIN ||       \
                                v == (type)PY_SSIZE_T_MIN)))  ||  \
    (sizeof(type) == sizeof(Py_ssize_t) &&              \
          (is_signed || likely(v < (type)PY_SSIZE_T_MAX ||        \
                               v == (type)PY_SSIZE_T_MAX)))  )
static CYTHON_INLINE char* __Pyx_PyObject_AsString(PyObject*);
static CYTHON_INLINE char* __Pyx_PyObject_AsStringAndSize(PyObject*, Py_ssize_t* length);
#define __Pyx_PyByteArray_FromString(s) PyByteArray_FromStringAndSize((const char*)s, strlen((const char*)s))
#define __Pyx_PyByteArray_FromStringAndSize(s, l) PyByteArray_FromStringAndSize((const char*)s, l)
#define __Pyx_PyBytes_FromString        PyBytes_FromString
#define __Pyx_PyBytes_FromStringAndSize PyBytes_FromStringAndSize
static CYTHON_INLINE PyObject* __Pyx_PyUnicode_FromString(const char*);
#if PY_MAJOR_VERSION < 3
    #define __Pyx_PyStr_FromString        __Pyx_PyBytes_FromString
    #define __Pyx_PyStr_FromStringAndSize __Pyx_PyBytes_FromStringAndSize
#else
    #define __Pyx_PyStr_FromString        __Pyx_PyUnicode_FromString
    #define __Pyx_PyStr_FromStringAndSize __Pyx_PyUnicode_FromStringAndSize
#endif
#define __Pyx_PyObject_AsSString(s)    ((signed char*) __Pyx_PyObject_AsString(s))
#define __Pyx_PyObject_AsUString(s)    ((unsigned char*) __Pyx_PyObject_AsString(s))
#define __Pyx_PyObject_FromUString(s)  __Pyx_PyObject_FromString((const char*)s)
#define __Pyx_PyBytes_FromUString(s)   __Pyx_PyBytes_FromString((const char*)s)
#define __Pyx_PyByteArray_FromUString(s)   __Pyx_PyByteArray_FromString((const char*)s)
#define __Pyx_PyStr_FromUString(s)     __Pyx_PyStr_FromString((const char*)s)
#define __Pyx_PyUnicode_FromUString(s) __Pyx_PyUnicode_FromString((const char*)s)
#if PY_MAJOR_VERSION < 3
static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u)
{
    const Py_UNICODE *u_end = u;
    while (*u_end++) ;
    return (size_t)(u_end - u - 1);
}
#else
#define __Pyx_Py_UNICODE_strlen Py_UNICODE_strlen
#endif
#define __Pyx_PyUnicode_FromUnicode(u)       PyUnicode_FromUnicode(u, __Pyx_Py_UNICODE_strlen(u))
#define __Pyx_PyUnicode_FromUnicodeAndLength PyUnicode_FromUnicode
#define __Pyx_PyUnicode_AsUnicode            PyUnicode_AsUnicode
#define __Pyx_Owned_Py_None(b) (Py_INCREF(Py_None), Py_None)
#define __Pyx_PyBool_FromLong(b) ((b) ? (Py_INCREF(Py_True), Py_True) : (Py_INCREF(Py_False), Py_False))
static CYTHON_INLINE int __Pyx_PyObject_IsTrue(PyObject*);
static CYTHON_INLINE PyObject* __Pyx_PyNumber_Int(PyObject* x);
static CYTHON_INLINE Py_ssize_t __Pyx_PyIndex_AsSsize_t(PyObject*);
static CYTHON_INLINE PyObject * __Pyx_PyInt_FromSize_t(size_t);
#if CYTHON_COMPILING_IN_CPYTHON
#define __pyx_PyFloat_AsDouble(x) (PyFloat_CheckExact(x) ? PyFloat_AS_DOUBLE(x) : PyFloat_AsDouble(x))
#else
#define __pyx_PyFloat_AsDouble(x) PyFloat_AsDouble(x)
#endif
#define __pyx_PyFloat_AsFloat(x) ((float) __pyx_PyFloat_AsDouble(x))
#if PY_MAJOR_VERSION < 3 && __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
static int __Pyx_sys_getdefaultencoding_not_ascii;
static int __Pyx_init_sys_getdefaultencoding_params(void) {
    PyObject* sys;
    PyObject* default_encoding = NULL;
    PyObject* ascii_chars_u = NULL;
    PyObject* ascii_chars_b = NULL;
    const char* default_encoding_c;
    sys = PyImport_ImportModule("sys");
    if (!sys) goto bad;
    default_encoding = PyObject_CallMethod(sys, (char*) (const char*) "getdefaultencoding", NULL);
    Py_DECREF(sys);
    if (!default_encoding) goto bad;
    default_encoding_c = PyBytes_AsString(default_encoding);
    if (!default_encoding_c) goto bad;
    if (strcmp(default_encoding_c, "ascii") == 0) {
        __Pyx_sys_getdefaultencoding_not_ascii = 0;
    } else {
        char ascii_chars[128];
        int c;
        for (c = 0; c < 128; c++) {
            ascii_chars[c] = c;
        }
        __Pyx_sys_getdefaultencoding_not_ascii = 1;
        ascii_chars_u = PyUnicode_DecodeASCII(ascii_chars, 128, NULL);
        if (!ascii_chars_u) goto bad;
        ascii_chars_b = PyUnicode_AsEncodedString(ascii_chars_u, default_encoding_c, NULL);
        if (!ascii_chars_b || !PyBytes_Check(ascii_chars_b) || memcmp(ascii_chars, PyBytes_AS_STRING(ascii_chars_b), 128) != 0) {
            PyErr_Format(
                PyExc_ValueError,
                "This module compiled with c_string_encoding=ascii, but default encoding '%.200s' is not a superset of ascii.",
                default_encoding_c);
            goto bad;
        }
        Py_DECREF(ascii_chars_u);
        Py_DECREF(ascii_chars_b);
    }
    Py_DECREF(default_encoding);
    return 0;
bad:
    Py_XDECREF(default_encoding);
    Py_XDECREF(ascii_chars_u);
    Py_XDECREF(ascii_chars_b);
    return -1;
}
#endif
#if __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT && PY_MAJOR_VERSION >= 3
#define __Pyx_PyUnicode_FromStringAndSize(c_str, size) PyUnicode_DecodeUTF8(c_str, size, NULL)
#else
#define __Pyx_PyUnicode_FromStringAndSize(c_str, size) PyUnicode_Decode(c_str, size, __PYX_DEFAULT_STRING_ENCODING, NULL)
#if __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT
static char* __PYX_DEFAULT_STRING_ENCODING;
static int __Pyx_init_sys_getdefaultencoding_params(void) {
    PyObject* sys;
    PyObject* default_encoding = NULL;
    char* default_encoding_c;
    sys = PyImport_ImportModule("sys");
    if (!sys) goto bad;
    default_encoding = PyObject_CallMethod(sys, (char*) (const char*) "getdefaultencoding", NULL);
    Py_DECREF(sys);
    if (!default_encoding) goto bad;
    default_encoding_c = PyBytes_AsString(default_encoding);
    if (!default_encoding_c) goto bad;
    __PYX_DEFAULT_STRING_ENCODING = (char*) malloc(strlen(default_encoding_c));
    if (!__PYX_DEFAULT_STRING_ENCODING) goto bad;
    strcpy(__PYX_DEFAULT_STRING_ENCODING, default_encoding_c);
    Py_DECREF(default_encoding);
    return 0;
bad:
    Py_XDECREF(default_encoding);
    return -1;
}
#endif
#endif


/* Test for GCC > 2.95 */
#if defined(__GNUC__)     && (__GNUC__ > 2 || (__GNUC__ == 2 && (__GNUC_MINOR__ > 95)))
  #define likely(x)   __builtin_expect(!!(x), 1)
  #define unlikely(x) __builtin_expect(!!(x), 0)
#else /* !__GNUC__ or GCC < 2.95 */
  #define likely(x)   (x)
  #define unlikely(x) (x)
#endif /* __GNUC__ */

static PyObject *__pyx_m;
static PyObject *__pyx_d;
static PyObject *__pyx_b;
static PyObject *__pyx_empty_tuple;
static PyObject *__pyx_empty_bytes;
static int __pyx_lineno;
static int __pyx_clineno = 0;
static const char * __pyx_cfilenm= __FILE__;
static const char *__pyx_filename;

#if !defined(CYTHON_CCOMPLEX)
  #if defined(__cplusplus)
    #define CYTHON_CCOMPLEX 1
  #elif defined(_Complex_I)
    #define CYTHON_CCOMPLEX 1
  #else
    #define CYTHON_CCOMPLEX 0
  #endif
#endif
#if CYTHON_CCOMPLEX
  #ifdef __cplusplus
    #include <complex>
  #else
    #include <complex.h>
  #endif
#endif
#if CYTHON_CCOMPLEX && !defined(__cplusplus) && defined(__sun__) && defined(__GNUC__)
  #undef _Complex_I
  #define _Complex_I 1.0fj
#endif


static const char *__pyx_f[] = {
  "rmgpy/molecule/group.py",
  "rmgpy/molecule/group.pxd",
  "__init__.pxd",
  "rmgpy/molecule/atomtype.pxd",
  "rmgpy/molecule/element.pxd",
  "type.pxd",
  "rmgpy/molecule/molecule.pxd",
};

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":723
 * # in Cython to enable them only on the right systems.
 * 
 * ctypedef npy_int8       int8_t             # <<<<<<<<<<<<<<
 * ctypedef npy_int16      int16_t
 * ctypedef npy_int32      int32_t
 */
typedef npy_int8 __pyx_t_5numpy_int8_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":724
 * 
 * ctypedef npy_int8       int8_t
 * ctypedef npy_int16      int16_t             # <<<<<<<<<<<<<<
 * ctypedef npy_int32      int32_t
 * ctypedef npy_int64      int64_t
 */
typedef npy_int16 __pyx_t_5numpy_int16_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":725
 * ctypedef npy_int8       int8_t
 * ctypedef npy_int16      int16_t
 * ctypedef npy_int32      int32_t             # <<<<<<<<<<<<<<
 * ctypedef npy_int64      int64_t
 * #ctypedef npy_int96      int96_t
 */
typedef npy_int32 __pyx_t_5numpy_int32_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":726
 * ctypedef npy_int16      int16_t
 * ctypedef npy_int32      int32_t
 * ctypedef npy_int64      int64_t             # <<<<<<<<<<<<<<
 * #ctypedef npy_int96      int96_t
 * #ctypedef npy_int128     int128_t
 */
typedef npy_int64 __pyx_t_5numpy_int64_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":730
 * #ctypedef npy_int128     int128_t
 * 
 * ctypedef npy_uint8      uint8_t             # <<<<<<<<<<<<<<
 * ctypedef npy_uint16     uint16_t
 * ctypedef npy_uint32     uint32_t
 */
typedef npy_uint8 __pyx_t_5numpy_uint8_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":731
 * 
 * ctypedef npy_uint8      uint8_t
 * ctypedef npy_uint16     uint16_t             # <<<<<<<<<<<<<<
 * ctypedef npy_uint32     uint32_t
 * ctypedef npy_uint64     uint64_t
 */
typedef npy_uint16 __pyx_t_5numpy_uint16_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":732
 * ctypedef npy_uint8      uint8_t
 * ctypedef npy_uint16     uint16_t
 * ctypedef npy_uint32     uint32_t             # <<<<<<<<<<<<<<
 * ctypedef npy_uint64     uint64_t
 * #ctypedef npy_uint96     uint96_t
 */
typedef npy_uint32 __pyx_t_5numpy_uint32_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":733
 * ctypedef npy_uint16     uint16_t
 * ctypedef npy_uint32     uint32_t
 * ctypedef npy_uint64     uint64_t             # <<<<<<<<<<<<<<
 * #ctypedef npy_uint96     uint96_t
 * #ctypedef npy_uint128    uint128_t
 */
typedef npy_uint64 __pyx_t_5numpy_uint64_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":737
 * #ctypedef npy_uint128    uint128_t
 * 
 * ctypedef npy_float32    float32_t             # <<<<<<<<<<<<<<
 * ctypedef npy_float64    float64_t
 * #ctypedef npy_float80    float80_t
 */
typedef npy_float32 __pyx_t_5numpy_float32_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":738
 * 
 * ctypedef npy_float32    float32_t
 * ctypedef npy_float64    float64_t             # <<<<<<<<<<<<<<
 * #ctypedef npy_float80    float80_t
 * #ctypedef npy_float128   float128_t
 */
typedef npy_float64 __pyx_t_5numpy_float64_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":747
 * # The int types are mapped a bit surprising --
 * # numpy.int corresponds to 'l' and numpy.long to 'q'
 * ctypedef npy_long       int_t             # <<<<<<<<<<<<<<
 * ctypedef npy_longlong   long_t
 * ctypedef npy_longlong   longlong_t
 */
typedef npy_long __pyx_t_5numpy_int_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":748
 * # numpy.int corresponds to 'l' and numpy.long to 'q'
 * ctypedef npy_long       int_t
 * ctypedef npy_longlong   long_t             # <<<<<<<<<<<<<<
 * ctypedef npy_longlong   longlong_t
 * 
 */
typedef npy_longlong __pyx_t_5numpy_long_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":749
 * ctypedef npy_long       int_t
 * ctypedef npy_longlong   long_t
 * ctypedef npy_longlong   longlong_t             # <<<<<<<<<<<<<<
 * 
 * ctypedef npy_ulong      uint_t
 */
typedef npy_longlong __pyx_t_5numpy_longlong_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":751
 * ctypedef npy_longlong   longlong_t
 * 
 * ctypedef npy_ulong      uint_t             # <<<<<<<<<<<<<<
 * ctypedef npy_ulonglong  ulong_t
 * ctypedef npy_ulonglong  ulonglong_t
 */
typedef npy_ulong __pyx_t_5numpy_uint_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":752
 * 
 * ctypedef npy_ulong      uint_t
 * ctypedef npy_ulonglong  ulong_t             # <<<<<<<<<<<<<<
 * ctypedef npy_ulonglong  ulonglong_t
 * 
 */
typedef npy_ulonglong __pyx_t_5numpy_ulong_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":753
 * ctypedef npy_ulong      uint_t
 * ctypedef npy_ulonglong  ulong_t
 * ctypedef npy_ulonglong  ulonglong_t             # <<<<<<<<<<<<<<
 * 
 * ctypedef npy_intp       intp_t
 */
typedef npy_ulonglong __pyx_t_5numpy_ulonglong_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":755
 * ctypedef npy_ulonglong  ulonglong_t
 * 
 * ctypedef npy_intp       intp_t             # <<<<<<<<<<<<<<
 * ctypedef npy_uintp      uintp_t
 * 
 */
typedef npy_intp __pyx_t_5numpy_intp_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":756
 * 
 * ctypedef npy_intp       intp_t
 * ctypedef npy_uintp      uintp_t             # <<<<<<<<<<<<<<
 * 
 * ctypedef npy_double     float_t
 */
typedef npy_uintp __pyx_t_5numpy_uintp_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":758
 * ctypedef npy_uintp      uintp_t
 * 
 * ctypedef npy_double     float_t             # <<<<<<<<<<<<<<
 * ctypedef npy_double     double_t
 * ctypedef npy_longdouble longdouble_t
 */
typedef npy_double __pyx_t_5numpy_float_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":759
 * 
 * ctypedef npy_double     float_t
 * ctypedef npy_double     double_t             # <<<<<<<<<<<<<<
 * ctypedef npy_longdouble longdouble_t
 * 
 */
typedef npy_double __pyx_t_5numpy_double_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":760
 * ctypedef npy_double     float_t
 * ctypedef npy_double     double_t
 * ctypedef npy_longdouble longdouble_t             # <<<<<<<<<<<<<<
 * 
 * ctypedef npy_cfloat      cfloat_t
 */
typedef npy_longdouble __pyx_t_5numpy_longdouble_t;
#if CYTHON_CCOMPLEX
  #ifdef __cplusplus
    typedef ::std::complex< float > __pyx_t_float_complex;
  #else
    typedef float _Complex __pyx_t_float_complex;
  #endif
#else
    typedef struct { float real, imag; } __pyx_t_float_complex;
#endif

#if CYTHON_CCOMPLEX
  #ifdef __cplusplus
    typedef ::std::complex< double > __pyx_t_double_complex;
  #else
    typedef double _Complex __pyx_t_double_complex;
  #endif
#else
    typedef struct { double real, imag; } __pyx_t_double_complex;
#endif


/*--- Type declarations ---*/
struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex;
struct __pyx_obj_5rmgpy_8molecule_5graph_Edge;
struct __pyx_obj_5rmgpy_8molecule_5graph_Graph;
struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType;
struct __pyx_obj_5rmgpy_8molecule_7element_Element;
struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom;
struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond;
struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule;
struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom;
struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond;
struct __pyx_obj_5rmgpy_8molecule_5group_Group;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms;

/* "rmgpy/molecule/graph.pxd":90
 *     cpdef updateConnectivityValues(self)
 * 
 *     cpdef Graph copy(self, bint deep=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef Graph merge(self, Graph other)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy {
  int __pyx_n;
  int deep;
};

/* "rmgpy/molecule/graph.pxd":100
 *     cpdef sortVertices(self)
 * 
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/graph.pxd":102
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/graph.pxd":104
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/graph.pxd":106
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isCyclic(self) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms {
  int __pyx_n;
  PyObject *initialMap;
};
struct __pyx_opt_args_5rmgpy_8molecule_7element_getElement;

/* "rmgpy/molecule/element.pxd":36
 *     cdef public int isotope
 * 
 * cpdef Element getElement(value, int isotope=?)             # <<<<<<<<<<<<<<
 */
struct __pyx_opt_args_5rmgpy_8molecule_7element_getElement {
  int __pyx_n;
  int isotope;
};

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":762
 * ctypedef npy_longdouble longdouble_t
 * 
 * ctypedef npy_cfloat      cfloat_t             # <<<<<<<<<<<<<<
 * ctypedef npy_cdouble     cdouble_t
 * ctypedef npy_clongdouble clongdouble_t
 */
typedef npy_cfloat __pyx_t_5numpy_cfloat_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":763
 * 
 * ctypedef npy_cfloat      cfloat_t
 * ctypedef npy_cdouble     cdouble_t             # <<<<<<<<<<<<<<
 * ctypedef npy_clongdouble clongdouble_t
 * 
 */
typedef npy_cdouble __pyx_t_5numpy_cdouble_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":764
 * ctypedef npy_cfloat      cfloat_t
 * ctypedef npy_cdouble     cdouble_t
 * ctypedef npy_clongdouble clongdouble_t             # <<<<<<<<<<<<<<
 * 
 * ctypedef npy_cdouble     complex_t
 */
typedef npy_clongdouble __pyx_t_5numpy_clongdouble_t;

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":766
 * ctypedef npy_clongdouble clongdouble_t
 * 
 * ctypedef npy_cdouble     complex_t             # <<<<<<<<<<<<<<
 * 
 * cdef inline object PyArray_MultiIterNew1(a):
 */
typedef npy_cdouble __pyx_t_5numpy_complex_t;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_getNumAtoms;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_copy;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_isIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_findIsomorphism;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_isSubgraphIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_findSubgraphIsomorphisms;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromInChI;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromSMILES;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromAdjacencyList;
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_toAdjacencyList;

/* "rmgpy/molecule/molecule.pxd":143
 *     cpdef double getMolecularWeight(self)
 * 
 *     cpdef int getNumAtoms(self, str element=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef int getNumberOfRadicalElectrons(self)
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_getNumAtoms {
  int __pyx_n;
  PyObject *element;
};

/* "rmgpy/molecule/molecule.pxd":147
 *     cpdef int getNumberOfRadicalElectrons(self)
 * 
 *     cpdef Graph copy(self, bint deep=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef deleteHydrogens(self)
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_copy {
  int __pyx_n;
  int deep;
};

/* "rmgpy/molecule/molecule.pxd":159
 *     cpdef dict getLabeledAtoms(self)
 * 
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_isIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/molecule.pxd":161
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_findIsomorphism {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/molecule.pxd":163
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_isSubgraphIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/molecule.pxd":165
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isAtomInCycle(self, Atom atom) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_findSubgraphIsomorphisms {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/molecule.pxd":173
 *     cpdef draw(self, str path)
 * 
 *     cpdef fromInChI(self, str inchistr, backend=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef fromSMILES(self, str smilesstr, backend=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromInChI {
  int __pyx_n;
  PyObject *backend;
};

/* "rmgpy/molecule/molecule.pxd":175
 *     cpdef fromInChI(self, str inchistr, backend=?)
 * 
 *     cpdef fromSMILES(self, str smilesstr, backend=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef fromAdjacencyList(self, str adjlist, bint saturateH=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromSMILES {
  int __pyx_n;
  PyObject *backend;
};

/* "rmgpy/molecule/molecule.pxd":177
 *     cpdef fromSMILES(self, str smilesstr, backend=?)
 * 
 *     cpdef fromAdjacencyList(self, str adjlist, bint saturateH=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef fromXYZ(self, numpy.ndarray atomicNums, numpy.ndarray coordinates)
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromAdjacencyList {
  int __pyx_n;
  int saturateH;
};

/* "rmgpy/molecule/molecule.pxd":191
 *     cpdef str toSMILES(self)
 * 
 *     cpdef toAdjacencyList(self, str label=?, bint removeH=?, bint removeLonePairs=?, bint oldStyle=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isLinear(self) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_toAdjacencyList {
  int __pyx_n;
  PyObject *label;
  int removeH;
  int removeLonePairs;
  int oldStyle;
};
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_copy;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findIsomorphism;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic;
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms;
struct __pyx_defaults;
typedef struct __pyx_defaults __pyx_defaults;

/* "rmgpy/molecule/group.pxd":130
 *     cpdef sortAtoms(self)
 * 
 *     cpdef Graph copy(self, bint deep=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef clearLabeledAtoms(self)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_copy {
  int __pyx_n;
  int deep;
};

/* "rmgpy/molecule/group.pxd":142
 *     cpdef fromAdjacencyList(self, str adjlist)
 * 
 *     cpdef toAdjacencyList(self, str label=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef updateFingerprint(self)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList {
  int __pyx_n;
  PyObject *label;
};

/* "rmgpy/molecule/group.pxd":146
 *     cpdef updateFingerprint(self)
 * 
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/group.pxd":148
 *     cpdef bint isIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findIsomorphism {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/group.pxd":150
 *     cpdef list findIsomorphism(self, Graph other, dict initialMap=?)
 * 
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2             # <<<<<<<<<<<<<<
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic {
  int __pyx_n;
  PyObject *initialMap;
};

/* "rmgpy/molecule/group.pxd":152
 *     cpdef bint isSubgraphIsomorphic(self, Graph other, dict initialMap=?) except -2
 * 
 *     cpdef list findSubgraphIsomorphisms(self, Graph other, dict initialMap=?)             # <<<<<<<<<<<<<<
 * 
 *     cpdef bint isIdentical(self, Graph other)
 */
struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms {
  int __pyx_n;
  PyObject *initialMap;
};
struct __pyx_defaults {
  PyObject *__pyx_arg_partners;
};

/* "rmgpy/molecule/graph.pxd":27
 * ################################################################################
 * 
 * cdef class Vertex(object):             # <<<<<<<<<<<<<<
 * 
 *     cdef public dict edges
 */
struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex {
  PyObject_HEAD
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex *__pyx_vtab;
  PyObject *edges;
  short connectivity1;
  short connectivity2;
  short connectivity3;
  short sortingLabel;
  int terminal;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *mapping;
  int ignore;
};


/* "rmgpy/molecule/graph.pxd":54
 * ################################################################################
 * 
 * cdef class Edge(object):             # <<<<<<<<<<<<<<
 * 
 *     cdef public Vertex vertex1, vertex2
 */
struct __pyx_obj_5rmgpy_8molecule_5graph_Edge {
  PyObject_HEAD
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge *__pyx_vtab;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *vertex1;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *vertex2;
};


/* "rmgpy/molecule/graph.pxd":68
 * ################################################################################
 * 
 * cdef class Graph:             # <<<<<<<<<<<<<<
 * 
 *     cdef public list vertices
 */
struct __pyx_obj_5rmgpy_8molecule_5graph_Graph {
  PyObject_HEAD
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph *__pyx_vtab;
  PyObject *vertices;
};


/* "rmgpy/molecule/atomtype.pxd":27
 * ################################################################################
 * 
 * cdef class AtomType:             # <<<<<<<<<<<<<<
 * 
 *     cdef public str label
 */
struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType {
  PyObject_HEAD
  struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *__pyx_vtab;
  PyObject *label;
  PyObject *generic;
  PyObject *specific;
  PyObject *incrementBond;
  PyObject *decrementBond;
  PyObject *formBond;
  PyObject *breakBond;
  PyObject *incrementRadical;
  PyObject *decrementRadical;
  PyObject *incrementLonePair;
  PyObject *decrementLonePair;
  PyObject *single;
  PyObject *allDouble;
  PyObject *rDouble;
  PyObject *oDouble;
  PyObject *sDouble;
  PyObject *triple;
  PyObject *benzene;
  PyObject *lonePairs;
};

struct __pyx_obj_5rmgpy_8molecule_7element_Element {
  PyObject_HEAD
  int number;
  PyObject *name;
  PyObject *symbol;
  float mass;
  float covRadius;
  int isotope;
};


/* "rmgpy/molecule/molecule.pxd":37
 * cdef dict bond_orders
 * 
 * cdef class Atom(Vertex):             # <<<<<<<<<<<<<<
 * 
 *     cdef public Element element
 */
struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex __pyx_base;
  struct __pyx_obj_5rmgpy_8molecule_7element_Element *element;
  short radicalElectrons;
  short charge;
  PyObject *label;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *atomType;
  PyArrayObject *coords;
  short lonePairs;
};


/* "rmgpy/molecule/molecule.pxd":82
 * ################################################################################
 * 
 * cdef class Bond(Edge):             # <<<<<<<<<<<<<<
 * 
 *     cdef public str order
 */
struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge __pyx_base;
  PyObject *order;
};


/* "rmgpy/molecule/molecule.pxd":106
 * ################################################################################
 * 
 * cdef class Molecule(Graph):             # <<<<<<<<<<<<<<
 * 
 *     cdef public bint implicitHydrogens
 */
struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph __pyx_base;
  int implicitHydrogens;
  int symmetryNumber;
  int multiplicity;
  PyObject *rdMol;
  int rdMolConfId;
  PyObject *_fingerprint;
  PyObject *InChI;
  PyObject *props;
};


/* "rmgpy/molecule/group.pxd":33
 * ################################################################################
 * 
 * cdef class GroupAtom(Vertex):             # <<<<<<<<<<<<<<
 * 
 *     cdef public list atomType
 */
struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex __pyx_base;
  PyObject *atomType;
  PyObject *radicalElectrons;
  PyObject *charge;
  PyObject *label;
  PyObject *lonePairs;
};


/* "rmgpy/molecule/group.pxd":74
 * ################################################################################
 * 
 * cdef class GroupBond(Edge):             # <<<<<<<<<<<<<<
 * 
 *     cdef public list order
 */
struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge __pyx_base;
  PyObject *order;
};


/* "rmgpy/molecule/group.pxd":100
 * ################################################################################
 * 
 * cdef class Group(Graph):             # <<<<<<<<<<<<<<
 * 
 *     cdef public list multiplicity
 */
struct __pyx_obj_5rmgpy_8molecule_5group_Group {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph __pyx_base;
  PyObject *multiplicity;
  short carbonCount;
  short nitrogenCount;
  short oxygenCount;
  short sulfurCount;
  short radicalCount;
};



/* "rmgpy/molecule/graph.pxd":27
 * ################################################################################
 * 
 * cdef class Vertex(object):             # <<<<<<<<<<<<<<
 * 
 *     cdef public dict edges
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *(*copy)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*equivalent)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*isSpecificCaseOf)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  PyObject *(*resetConnectivityValues)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex *__pyx_vtabptr_5rmgpy_8molecule_5graph_Vertex;


/* "rmgpy/molecule/graph.pxd":54
 * ################################################################################
 * 
 * cdef class Edge(object):             # <<<<<<<<<<<<<<
 * 
 *     cdef public Vertex vertex1, vertex2
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *(*copy)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  int (*equivalent)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  int (*isSpecificCaseOf)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *(*getOtherVertex)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge *__pyx_vtabptr_5rmgpy_8molecule_5graph_Edge;


/* "rmgpy/molecule/graph.pxd":68
 * ################################################################################
 * 
 * cdef class Graph:             # <<<<<<<<<<<<<<
 * 
 *     cdef public list vertices
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *(*addVertex)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *(*addEdge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  PyObject *(*getEdges)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *(*getEdge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*hasVertex)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*hasEdge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  PyObject *(*removeVertex)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  PyObject *(*removeEdge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  PyObject *(*updateConnectivityValues)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *(*copy)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy *__pyx_optional_args);
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *(*merge)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*split)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*resetConnectivityValues)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*sortVertices)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  int (*isIsomorphic)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic *__pyx_optional_args);
  PyObject *(*findIsomorphism)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism *__pyx_optional_args);
  int (*isSubgraphIsomorphic)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic *__pyx_optional_args);
  PyObject *(*findSubgraphIsomorphisms)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms *__pyx_optional_args);
  int (*isCyclic)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  int (*isVertexInCycle)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  int (*isEdgeInCycle)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch);
  int (*__pyx___isChainInCycle)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*getAllCyclicVertices)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getAllPolycyclicVertices)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getPolycyclicRings)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getMonocyclicRings)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getDisparateRings)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  PyObject *(*getAllCycles)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch);
  PyObject *(*getAllCyclesOfSize)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int, int __pyx_skip_dispatch);
  PyObject *(*__pyx___exploreCyclesRecursively)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, PyObject *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*getSmallestSetOfSmallestRings)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  int (*isMappingValid)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, PyObject *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph *__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph;


/* "rmgpy/molecule/atomtype.pxd":27
 * ################################################################################
 * 
 * cdef class AtomType:             # <<<<<<<<<<<<<<
 * 
 *     cdef public str label
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType {
  int (*isSpecificCaseOf)(struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, int __pyx_skip_dispatch);
  int (*equivalent)(struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, int __pyx_skip_dispatch);
  PyObject *(*getFeatures)(struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *__pyx_vtabptr_5rmgpy_8molecule_8atomtype_AtomType;


/* "rmgpy/molecule/molecule.pxd":37
 * cdef dict bond_orders
 * 
 * cdef class Atom(Vertex):             # <<<<<<<<<<<<<<
 * 
 *     cdef public Element element
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Atom {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex __pyx_base;
  int (*isHydrogen)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  int (*isNonHydrogen)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  int (*isCarbon)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  int (*isOxygen)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  int (*isSulfur)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*incrementRadical)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*decrementRadical)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*setLonePairs)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int, int __pyx_skip_dispatch);
  PyObject *(*incrementLonePairs)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*decrementLonePairs)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*updateCharge)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*setSpinMultiplicity)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int, int __pyx_skip_dispatch);
  PyObject *(*getBondOrdersForAtom)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Atom *__pyx_vtabptr_5rmgpy_8molecule_8molecule_Atom;


/* "rmgpy/molecule/molecule.pxd":82
 * ################################################################################
 * 
 * cdef class Bond(Edge):             # <<<<<<<<<<<<<<
 * 
 *     cdef public str order
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Bond {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge __pyx_base;
  int (*isSingle)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
  int (*isDouble)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
  int (*isTriple)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
  int (*isBenzene)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
  PyObject *(*incrementOrder)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
  PyObject *(*decrementOrder)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Bond *__pyx_vtabptr_5rmgpy_8molecule_8molecule_Bond;


/* "rmgpy/molecule/molecule.pxd":106
 * ################################################################################
 * 
 * cdef class Molecule(Graph):             # <<<<<<<<<<<<<<
 * 
 *     cdef public bint implicitHydrogens
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Molecule {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph __pyx_base;
  PyObject *(*getFingerprint)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*addAtom)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*addBond)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
  PyObject *(*getBonds)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *(*getBond)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  int (*hasAtom)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  int (*hasBond)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*removeAtom)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  PyObject *(*removeBond)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
  PyObject *(*sortAtoms)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*getFormula)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  short (*getRadicalCount)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  double (*getMolecularWeight)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  int (*getNumAtoms)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_getNumAtoms *__pyx_optional_args);
  int (*getNumberOfRadicalElectrons)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*deleteHydrogens)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*clearLabeledAtoms)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  int (*containsLabeledAtom)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, PyObject *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *(*getLabeledAtom)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*getLabeledAtoms)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  int (*isAtomInCycle)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
  int (*isBondInCycle)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *, int __pyx_skip_dispatch);
  PyObject *(*draw)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*fromInChI)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, PyObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromInChI *__pyx_optional_args);
  PyObject *(*fromSMILES)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, PyObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromSMILES *__pyx_optional_args);
  PyObject *(*fromAdjacencyList)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, PyObject *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_fromAdjacencyList *__pyx_optional_args);
  PyObject *(*fromXYZ)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, PyArrayObject *, PyArrayObject *, int __pyx_skip_dispatch);
  PyObject *(*toInChI)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*toAugmentedInChI)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*toInChIKey)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*toAugmentedInChIKey)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*toSMILES)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*toAdjacencyList)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_8molecule_8Molecule_toAdjacencyList *__pyx_optional_args);
  int (*isLinear)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  int (*countInternalRotors)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  double (*calculateCp0)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  double (*calculateCpInf)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*updateAtomTypes)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  int (*isRadical)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  int (*calculateSymmetryNumber)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
  PyObject *(*generateResonanceIsomers)(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Molecule *__pyx_vtabptr_5rmgpy_8molecule_8molecule_Molecule;


/* "rmgpy/molecule/group.py":64
 * ################################################################################
 * 
 * class GroupAtom(Vertex):             # <<<<<<<<<<<<<<
 *     """
 *     An atom group. This class is based on the :class:`Atom` class, except that
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex __pyx_base;
  PyObject *(*__pyx___changeBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*__pyx___formBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*__pyx___breakBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*__pyx___gainRadical)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*__pyx___loseRadical)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*__pyx___gainPair)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*__pyx___losePair)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch);
  PyObject *(*applyAction)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch);
  int (*isOxygen)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  int (*isSulfur)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  int (*hasWildcards)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *(*makeSampleAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  PyObject *(*getBondOrdersForAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *__pyx_vtabptr_5rmgpy_8molecule_5group_GroupAtom;


/* "rmgpy/molecule/group.py":529
 * ################################################################################
 * 
 * class GroupBond(Edge):             # <<<<<<<<<<<<<<
 *     """
 *     A bond group. This class is based on the :class:`Bond` class, except that
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupBond {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge __pyx_base;
  PyObject *(*__pyx___changeBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, short, int __pyx_skip_dispatch);
  int (*isSingle)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  int (*isDouble)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  int (*isTriple)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  int (*isBenzene)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  PyObject *(*applyAction)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*makeBond)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupBond *__pyx_vtabptr_5rmgpy_8molecule_5group_GroupBond;


/* "rmgpy/molecule/group.py":706
 * ################################################################################
 * 
 * class Group(Graph):             # <<<<<<<<<<<<<<
 *     """
 *     A representation of a molecular substructure group using a graph data
 */

struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group {
  struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph __pyx_base;
  PyObject *(*addAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  PyObject *(*addBond)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  PyObject *(*getBonds)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *(*getBond)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  int (*hasAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  int (*hasBond)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  PyObject *(*removeAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch);
  PyObject *(*removeBond)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch);
  PyObject *(*sortAtoms)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  PyObject *(*clearLabeledAtoms)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*containsLabeledAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *(*getLabeledAtom)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*getLabeledAtoms)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  PyObject *(*fromAdjacencyList)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch);
  PyObject *(*toAdjacencyList)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList *__pyx_optional_args);
  PyObject *(*updateFingerprint)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*isIdentical)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch);
  int (*isAromaticRing)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*standardizeAtomType)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*addExplicitLigands)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*standardizeGroup)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *(*addImplicitAtomsFromAtomType)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *(*makeSampleMolecule)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *(*addImplicitBenzene)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
  int (*isBenzeneExplicit)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch);
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *__pyx_vtabptr_5rmgpy_8molecule_5group_Group;
#ifndef CYTHON_REFNANNY
  #define CYTHON_REFNANNY 0
#endif
#if CYTHON_REFNANNY
  typedef struct {
    void (*INCREF)(void*, PyObject*, int);
    void (*DECREF)(void*, PyObject*, int);
    void (*GOTREF)(void*, PyObject*, int);
    void (*GIVEREF)(void*, PyObject*, int);
    void* (*SetupContext)(const char*, int, const char*);
    void (*FinishContext)(void**);
  } __Pyx_RefNannyAPIStruct;
  static __Pyx_RefNannyAPIStruct *__Pyx_RefNanny = NULL;
  static __Pyx_RefNannyAPIStruct *__Pyx_RefNannyImportAPI(const char *modname);
  #define __Pyx_RefNannyDeclarations void *__pyx_refnanny = NULL;
#ifdef WITH_THREAD
  #define __Pyx_RefNannySetupContext(name, acquire_gil) \
          if (acquire_gil) { \
              PyGILState_STATE __pyx_gilstate_save = PyGILState_Ensure(); \
              __pyx_refnanny = __Pyx_RefNanny->SetupContext((name), __LINE__, __FILE__); \
              PyGILState_Release(__pyx_gilstate_save); \
          } else { \
              __pyx_refnanny = __Pyx_RefNanny->SetupContext((name), __LINE__, __FILE__); \
          }
#else
  #define __Pyx_RefNannySetupContext(name, acquire_gil) \
          __pyx_refnanny = __Pyx_RefNanny->SetupContext((name), __LINE__, __FILE__)
#endif
  #define __Pyx_RefNannyFinishContext() \
          __Pyx_RefNanny->FinishContext(&__pyx_refnanny)
  #define __Pyx_INCREF(r)  __Pyx_RefNanny->INCREF(__pyx_refnanny, (PyObject *)(r), __LINE__)
  #define __Pyx_DECREF(r)  __Pyx_RefNanny->DECREF(__pyx_refnanny, (PyObject *)(r), __LINE__)
  #define __Pyx_GOTREF(r)  __Pyx_RefNanny->GOTREF(__pyx_refnanny, (PyObject *)(r), __LINE__)
  #define __Pyx_GIVEREF(r) __Pyx_RefNanny->GIVEREF(__pyx_refnanny, (PyObject *)(r), __LINE__)
  #define __Pyx_XINCREF(r)  do { if((r) != NULL) {__Pyx_INCREF(r); }} while(0)
  #define __Pyx_XDECREF(r)  do { if((r) != NULL) {__Pyx_DECREF(r); }} while(0)
  #define __Pyx_XGOTREF(r)  do { if((r) != NULL) {__Pyx_GOTREF(r); }} while(0)
  #define __Pyx_XGIVEREF(r) do { if((r) != NULL) {__Pyx_GIVEREF(r);}} while(0)
#else
  #define __Pyx_RefNannyDeclarations
  #define __Pyx_RefNannySetupContext(name, acquire_gil)
  #define __Pyx_RefNannyFinishContext()
  #define __Pyx_INCREF(r) Py_INCREF(r)
  #define __Pyx_DECREF(r) Py_DECREF(r)
  #define __Pyx_GOTREF(r)
  #define __Pyx_GIVEREF(r)
  #define __Pyx_XINCREF(r) Py_XINCREF(r)
  #define __Pyx_XDECREF(r) Py_XDECREF(r)
  #define __Pyx_XGOTREF(r)
  #define __Pyx_XGIVEREF(r)
#endif
#define __Pyx_XDECREF_SET(r, v) do {                            \
        PyObject *tmp = (PyObject *) r;                         \
        r = v; __Pyx_XDECREF(tmp);                              \
    } while (0)
#define __Pyx_DECREF_SET(r, v) do {                             \
        PyObject *tmp = (PyObject *) r;                         \
        r = v; __Pyx_DECREF(tmp);                               \
    } while (0)
#define __Pyx_CLEAR(r)    do { PyObject* tmp = ((PyObject*)(r)); r = NULL; __Pyx_DECREF(tmp);} while(0)
#define __Pyx_XCLEAR(r)   do { if((r) != NULL) {PyObject* tmp = ((PyObject*)(r)); r = NULL; __Pyx_DECREF(tmp);}} while(0)

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_GetAttrStr(PyObject* obj, PyObject* attr_name) {
    PyTypeObject* tp = Py_TYPE(obj);
    if (likely(tp->tp_getattro))
        return tp->tp_getattro(obj, attr_name);
#if PY_MAJOR_VERSION < 3
    if (likely(tp->tp_getattr))
        return tp->tp_getattr(obj, PyString_AS_STRING(attr_name));
#endif
    return PyObject_GetAttr(obj, attr_name);
}
#else
#define __Pyx_PyObject_GetAttrStr(o,n) PyObject_GetAttr(o,n)
#endif

static PyObject *__Pyx_GetBuiltinName(PyObject *name);

static void __Pyx_RaiseDoubleKeywordsError(const char* func_name, PyObject* kw_name);

static int __Pyx_ParseOptionalKeywords(PyObject *kwds, PyObject **argnames[], \
    PyObject *kwds2, PyObject *values[], Py_ssize_t num_pos_args, \
    const char* function_name);

static void __Pyx_RaiseArgtupleInvalid(const char* func_name, int exact,
    Py_ssize_t num_min, Py_ssize_t num_max, Py_ssize_t num_found);

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_Call(PyObject *func, PyObject *arg, PyObject *kw);
#else
#define __Pyx_PyObject_Call(func, arg, kw) PyObject_Call(func, arg, kw)
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallMethO(PyObject *func, PyObject *arg);
#endif

static CYTHON_INLINE PyObject* __Pyx_PyObject_CallOneArg(PyObject *func, PyObject *arg);

#define __Pyx_GetItemInt(o, i, type, is_signed, to_py_func, is_list, wraparound, boundscheck) \
    (__Pyx_fits_Py_ssize_t(i, type, is_signed) ? \
    __Pyx_GetItemInt_Fast(o, (Py_ssize_t)i, is_list, wraparound, boundscheck) : \
    (is_list ? (PyErr_SetString(PyExc_IndexError, "list index out of range"), (PyObject*)NULL) : \
               __Pyx_GetItemInt_Generic(o, to_py_func(i))))
#define __Pyx_GetItemInt_List(o, i, type, is_signed, to_py_func, is_list, wraparound, boundscheck) \
    (__Pyx_fits_Py_ssize_t(i, type, is_signed) ? \
    __Pyx_GetItemInt_List_Fast(o, (Py_ssize_t)i, wraparound, boundscheck) : \
    (PyErr_SetString(PyExc_IndexError, "list index out of range"), (PyObject*)NULL))
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_List_Fast(PyObject *o, Py_ssize_t i,
                                                              int wraparound, int boundscheck);
#define __Pyx_GetItemInt_Tuple(o, i, type, is_signed, to_py_func, is_list, wraparound, boundscheck) \
    (__Pyx_fits_Py_ssize_t(i, type, is_signed) ? \
    __Pyx_GetItemInt_Tuple_Fast(o, (Py_ssize_t)i, wraparound, boundscheck) : \
    (PyErr_SetString(PyExc_IndexError, "tuple index out of range"), (PyObject*)NULL))
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Tuple_Fast(PyObject *o, Py_ssize_t i,
                                                              int wraparound, int boundscheck);
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Generic(PyObject *o, PyObject* j);
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Fast(PyObject *o, Py_ssize_t i,
                                                     int is_list, int wraparound, int boundscheck);

static CYTHON_INLINE PyObject *__Pyx_GetModuleGlobalName(PyObject *name);

#define __Pyx_SetItemInt(o, i, v, type, is_signed, to_py_func, is_list, wraparound, boundscheck) \
    (__Pyx_fits_Py_ssize_t(i, type, is_signed) ? \
    __Pyx_SetItemInt_Fast(o, (Py_ssize_t)i, v, is_list, wraparound, boundscheck) : \
    (is_list ? (PyErr_SetString(PyExc_IndexError, "list assignment index out of range"), -1) : \
               __Pyx_SetItemInt_Generic(o, to_py_func(i), v)))
static CYTHON_INLINE int __Pyx_SetItemInt_Generic(PyObject *o, PyObject *j, PyObject *v);
static CYTHON_INLINE int __Pyx_SetItemInt_Fast(PyObject *o, Py_ssize_t i, PyObject *v,
                                               int is_list, int wraparound, int boundscheck);

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE int __Pyx_ListComp_Append(PyObject* list, PyObject* x) {
    PyListObject* L = (PyListObject*) list;
    Py_ssize_t len = Py_SIZE(list);
    if (likely(L->allocated > len)) {
        Py_INCREF(x);
        PyList_SET_ITEM(list, len, x);
        Py_SIZE(list) = len+1;
        return 0;
    }
    return PyList_Append(list, x);
}
#else
#define __Pyx_ListComp_Append(L,x) PyList_Append(L,x)
#endif

#if PY_MAJOR_VERSION < 3
#define __Pyx_PyString_Join __Pyx_PyBytes_Join
#define __Pyx_PyBaseString_Join(s, v) (PyUnicode_CheckExact(s) ? PyUnicode_Join(s, v) : __Pyx_PyBytes_Join(s, v))
#else
#define __Pyx_PyString_Join PyUnicode_Join
#define __Pyx_PyBaseString_Join PyUnicode_Join
#endif
#if CYTHON_COMPILING_IN_CPYTHON
    #if PY_MAJOR_VERSION < 3
    #define __Pyx_PyBytes_Join _PyString_Join
    #else
    #define __Pyx_PyBytes_Join _PyBytes_Join
    #endif
#else
static CYTHON_INLINE PyObject* __Pyx_PyBytes_Join(PyObject* sep, PyObject* values);
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallNoArg(PyObject *func);
#else
#define __Pyx_PyObject_CallNoArg(func) __Pyx_PyObject_Call(func, __pyx_empty_tuple, NULL)
#endif

static CYTHON_INLINE int __Pyx_TypeTest(PyObject *obj, PyTypeObject *type);

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyList_GetSlice(PyObject* src, Py_ssize_t start, Py_ssize_t stop);
static CYTHON_INLINE PyObject* __Pyx_PyTuple_GetSlice(PyObject* src, Py_ssize_t start, Py_ssize_t stop);
#else
#define __Pyx_PyList_GetSlice(seq, start, stop)   PySequence_GetSlice(seq, start, stop)
#define __Pyx_PyTuple_GetSlice(seq, start, stop)  PySequence_GetSlice(seq, start, stop)
#endif

static CYTHON_INLINE int __Pyx_PyList_Extend(PyObject* L, PyObject* v) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyObject* none = _PyList_Extend((PyListObject*)L, v);
    if (unlikely(!none))
        return -1;
    Py_DECREF(none);
    return 0;
#else
    return PyList_SetSlice(L, PY_SSIZE_T_MAX, PY_SSIZE_T_MAX, v);
#endif
}

static CYTHON_INLINE void __Pyx_ErrRestore(PyObject *type, PyObject *value, PyObject *tb);
static CYTHON_INLINE void __Pyx_ErrFetch(PyObject **type, PyObject **value, PyObject **tb);

static void __Pyx_Raise(PyObject *type, PyObject *value, PyObject *tb, PyObject *cause);

#include <string.h>

static CYTHON_INLINE int __Pyx_PyBytes_Equals(PyObject* s1, PyObject* s2, int equals);

static CYTHON_INLINE int __Pyx_PyUnicode_Equals(PyObject* s1, PyObject* s2, int equals);

#if PY_MAJOR_VERSION >= 3
#define __Pyx_PyString_Equals __Pyx_PyUnicode_Equals
#else
#define __Pyx_PyString_Equals __Pyx_PyBytes_Equals
#endif

static CYTHON_INLINE int __Pyx_ArgTypeTest(PyObject *obj, PyTypeObject *type, int none_allowed,
    const char *name, int exact);

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE int __Pyx_PyList_Append(PyObject* list, PyObject* x) {
    PyListObject* L = (PyListObject*) list;
    Py_ssize_t len = Py_SIZE(list);
    if (likely(L->allocated > len) & likely(len > (L->allocated >> 1))) {
        Py_INCREF(x);
        PyList_SET_ITEM(list, len, x);
        Py_SIZE(list) = len+1;
        return 0;
    }
    return PyList_Append(list, x);
}
#else
#define __Pyx_PyList_Append(L,x) PyList_Append(L,x)
#endif

static PyObject* __Pyx_PyObject_CallMethod1(PyObject* obj, PyObject* method_name, PyObject* arg);

static CYTHON_INLINE int __Pyx_PyObject_Append(PyObject* L, PyObject* x);

static CYTHON_INLINE int __Pyx_PySequence_Contains(PyObject* item, PyObject* seq, int eq) {
    int result = PySequence_Contains(seq, item);
    return unlikely(result < 0) ? result : (result == (eq == Py_EQ));
}

static void __Pyx_WriteUnraisable(const char *name, int clineno,
                                  int lineno, const char *filename,
                                  int full_traceback);

#if CYTHON_COMPILING_IN_CPYTHON
#define __Pyx_PyObject_DelAttrStr(o,n) __Pyx_PyObject_SetAttrStr(o,n,NULL)
static CYTHON_INLINE int __Pyx_PyObject_SetAttrStr(PyObject* obj, PyObject* attr_name, PyObject* value) {
    PyTypeObject* tp = Py_TYPE(obj);
    if (likely(tp->tp_setattro))
        return tp->tp_setattro(obj, attr_name, value);
#if PY_MAJOR_VERSION < 3
    if (likely(tp->tp_setattr))
        return tp->tp_setattr(obj, PyString_AS_STRING(attr_name), value);
#endif
    return PyObject_SetAttr(obj, attr_name, value);
}
#else
#define __Pyx_PyObject_DelAttrStr(o,n)   PyObject_DelAttr(o,n)
#define __Pyx_PyObject_SetAttrStr(o,n,v) PyObject_SetAttr(o,n,v)
#endif

#if PY_MAJOR_VERSION >= 3
static PyObject *__Pyx_PyDict_GetItem(PyObject *d, PyObject* key) {
    PyObject *value;
    value = PyDict_GetItemWithError(d, key);
    if (unlikely(!value)) {
        if (!PyErr_Occurred()) {
            PyObject* args = PyTuple_Pack(1, key);
            if (likely(args))
                PyErr_SetObject(PyExc_KeyError, args);
            Py_XDECREF(args);
        }
        return NULL;
    }
    Py_INCREF(value);
    return value;
}
#else
    #define __Pyx_PyDict_GetItem(d, key) PyObject_GetItem(d, key)
#endif

static CYTHON_INLINE int __Pyx_IterFinish(void);

static PyObject* __Pyx_PyObject_CallMethod0(PyObject* obj, PyObject* method_name);

static CYTHON_INLINE void __Pyx_RaiseNeedMoreValuesError(Py_ssize_t index);

static CYTHON_INLINE void __Pyx_RaiseTooManyValuesError(Py_ssize_t expected);

static int __Pyx_IternextUnpackEndCheck(PyObject *retval, Py_ssize_t expected);

static CYTHON_INLINE void __Pyx_RaiseNoneNotIterableError(void);

static void __Pyx_UnpackTupleError(PyObject *, Py_ssize_t index);

static CYTHON_INLINE int __Pyx_unpack_tuple2(PyObject* tuple, PyObject** value1, PyObject** value2,
                                             int is_tuple, int has_known_size, int decref_tuple);

static CYTHON_INLINE PyObject* __Pyx_dict_iterator(PyObject* dict, int is_dict, PyObject* method_name,
                                                   Py_ssize_t* p_orig_length, int* p_is_dict);
static CYTHON_INLINE int __Pyx_dict_iter_next(PyObject* dict_or_iter, Py_ssize_t orig_length, Py_ssize_t* ppos,
                                              PyObject** pkey, PyObject** pvalue, PyObject** pitem, int is_dict);

static CYTHON_INLINE int __Pyx_PyDict_Contains(PyObject* item, PyObject* dict, int eq) {
    int result = PyDict_Contains(dict, item);
    return unlikely(result < 0) ? result : (result == (eq == Py_EQ));
}

static PyObject* __Pyx_ImportFrom(PyObject* module, PyObject* name);

static CYTHON_INLINE void __Pyx_RaiseUnboundLocalError(const char *varname);

#define __Pyx_PyObject_PopIndex(L, ix, is_signed, type, to_py_func) ( \
    (PyList_CheckExact(L) && __Pyx_fits_Py_ssize_t(ix, type, is_signed)) ? \
        __Pyx__PyList_PopIndex(L, ix) : __Pyx__PyObject_PopIndex(L, to_py_func(ix)))
#define __Pyx_PyList_PopIndex(L, ix, is_signed, type, to_py_func) ( \
    __Pyx_fits_Py_ssize_t(ix, type, is_signed) ? \
        __Pyx__PyList_PopIndex(L, ix) : __Pyx__PyObject_PopIndex(L, to_py_func(ix)))
static PyObject* __Pyx__PyList_PopIndex(PyObject* L, Py_ssize_t ix);
static PyObject* __Pyx__PyObject_PopIndex(PyObject* L, PyObject* py_ix);

static CYTHON_INLINE PyObject* __Pyx_PyObject_GetSlice(
        PyObject* obj, Py_ssize_t cstart, Py_ssize_t cstop,
        PyObject** py_start, PyObject** py_stop, PyObject** py_slice,
        int has_cstart, int has_cstop, int wraparound);

static PyTypeObject* __Pyx_FetchCommonType(PyTypeObject* type);

#define __Pyx_CyFunction_USED 1
#include <structmember.h>
#define __Pyx_CYFUNCTION_STATICMETHOD  0x01
#define __Pyx_CYFUNCTION_CLASSMETHOD   0x02
#define __Pyx_CYFUNCTION_CCLASS        0x04
#define __Pyx_CyFunction_GetClosure(f) \
    (((__pyx_CyFunctionObject *) (f))->func_closure)
#define __Pyx_CyFunction_GetClassObj(f) \
    (((__pyx_CyFunctionObject *) (f))->func_classobj)
#define __Pyx_CyFunction_Defaults(type, f) \
    ((type *)(((__pyx_CyFunctionObject *) (f))->defaults))
#define __Pyx_CyFunction_SetDefaultsGetter(f, g) \
    ((__pyx_CyFunctionObject *) (f))->defaults_getter = (g)
typedef struct {
    PyCFunctionObject func;
#if PY_VERSION_HEX < 0x030500A0
    PyObject *func_weakreflist;
#endif
    PyObject *func_dict;
    PyObject *func_name;
    PyObject *func_qualname;
    PyObject *func_doc;
    PyObject *func_globals;
    PyObject *func_code;
    PyObject *func_closure;
    PyObject *func_classobj;
    void *defaults;
    int defaults_pyobjects;
    int flags;
    PyObject *defaults_tuple;
    PyObject *defaults_kwdict;
    PyObject *(*defaults_getter)(PyObject *);
    PyObject *func_annotations;
} __pyx_CyFunctionObject;
static PyTypeObject *__pyx_CyFunctionType = 0;
#define __Pyx_CyFunction_NewEx(ml, flags, qualname, self, module, globals, code) \
    __Pyx_CyFunction_New(__pyx_CyFunctionType, ml, flags, qualname, self, module, globals, code)
static PyObject *__Pyx_CyFunction_New(PyTypeObject *, PyMethodDef *ml,
                                      int flags, PyObject* qualname,
                                      PyObject *self,
                                      PyObject *module, PyObject *globals,
                                      PyObject* code);
static CYTHON_INLINE void *__Pyx_CyFunction_InitDefaults(PyObject *m,
                                                         size_t size,
                                                         int pyobjects);
static CYTHON_INLINE void __Pyx_CyFunction_SetDefaultsTuple(PyObject *m,
                                                            PyObject *tuple);
static CYTHON_INLINE void __Pyx_CyFunction_SetDefaultsKwDict(PyObject *m,
                                                             PyObject *dict);
static CYTHON_INLINE void __Pyx_CyFunction_SetAnnotationsDict(PyObject *m,
                                                              PyObject *dict);
static int __Pyx_CyFunction_init(void);

static CYTHON_INLINE PyObject* __Pyx_PyDict_Values(PyObject* d);

static void __Pyx_call_next_tp_dealloc(PyObject* obj, destructor current_tp_dealloc);

static int __Pyx_call_next_tp_traverse(PyObject* obj, visitproc v, void *a, traverseproc current_tp_traverse);

static void __Pyx_call_next_tp_clear(PyObject* obj, inquiry current_tp_dealloc);

static void* __Pyx_GetVtable(PyObject *dict);

static int __Pyx_SetVtable(PyObject *dict, void *vtable);

static PyObject *__Pyx_CalculateMetaclass(PyTypeObject *metaclass, PyObject *bases);

static PyObject *__Pyx_Py3MetaclassPrepare(PyObject *metaclass, PyObject *bases, PyObject *name, PyObject *qualname,
                                           PyObject *mkw, PyObject *modname, PyObject *doc);
static PyObject *__Pyx_Py3ClassCreate(PyObject *metaclass, PyObject *name, PyObject *bases, PyObject *dict,
                                      PyObject *mkw, int calculate_metaclass, int allow_py2_metaclass);

static PyObject *__Pyx_GetNameInClass(PyObject *nmspace, PyObject *name);

typedef struct {
    int code_line;
    PyCodeObject* code_object;
} __Pyx_CodeObjectCacheEntry;
struct __Pyx_CodeObjectCache {
    int count;
    int max_count;
    __Pyx_CodeObjectCacheEntry* entries;
};
static struct __Pyx_CodeObjectCache __pyx_code_cache = {0,0,NULL};
static int __pyx_bisect_code_objects(__Pyx_CodeObjectCacheEntry* entries, int count, int code_line);
static PyCodeObject *__pyx_find_code_object(int code_line);
static void __pyx_insert_code_object(int code_line, PyCodeObject* code_object);

static void __Pyx_AddTraceback(const char *funcname, int c_line,
                               int py_line, const char *filename);

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_short(short value);

static CYTHON_INLINE short __Pyx_PyInt_As_short(PyObject *);

static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level);

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_long(long value);

static CYTHON_INLINE int __Pyx_PyInt_As_int(PyObject *);

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_int(int value);

#if CYTHON_CCOMPLEX
  #ifdef __cplusplus
    #define __Pyx_CREAL(z) ((z).real())
    #define __Pyx_CIMAG(z) ((z).imag())
  #else
    #define __Pyx_CREAL(z) (__real__(z))
    #define __Pyx_CIMAG(z) (__imag__(z))
  #endif
#else
    #define __Pyx_CREAL(z) ((z).real)
    #define __Pyx_CIMAG(z) ((z).imag)
#endif
#if (defined(_WIN32) || defined(__clang__)) && defined(__cplusplus) && CYTHON_CCOMPLEX
    #define __Pyx_SET_CREAL(z,x) ((z).real(x))
    #define __Pyx_SET_CIMAG(z,y) ((z).imag(y))
#else
    #define __Pyx_SET_CREAL(z,x) __Pyx_CREAL(z) = (x)
    #define __Pyx_SET_CIMAG(z,y) __Pyx_CIMAG(z) = (y)
#endif

static CYTHON_INLINE __pyx_t_float_complex __pyx_t_float_complex_from_parts(float, float);

#if CYTHON_CCOMPLEX
    #define __Pyx_c_eqf(a, b)   ((a)==(b))
    #define __Pyx_c_sumf(a, b)  ((a)+(b))
    #define __Pyx_c_difff(a, b) ((a)-(b))
    #define __Pyx_c_prodf(a, b) ((a)*(b))
    #define __Pyx_c_quotf(a, b) ((a)/(b))
    #define __Pyx_c_negf(a)     (-(a))
  #ifdef __cplusplus
    #define __Pyx_c_is_zerof(z) ((z)==(float)0)
    #define __Pyx_c_conjf(z)    (::std::conj(z))
    #if 1
        #define __Pyx_c_absf(z)     (::std::abs(z))
        #define __Pyx_c_powf(a, b)  (::std::pow(a, b))
    #endif
  #else
    #define __Pyx_c_is_zerof(z) ((z)==0)
    #define __Pyx_c_conjf(z)    (conjf(z))
    #if 1
        #define __Pyx_c_absf(z)     (cabsf(z))
        #define __Pyx_c_powf(a, b)  (cpowf(a, b))
    #endif
 #endif
#else
    static CYTHON_INLINE int __Pyx_c_eqf(__pyx_t_float_complex, __pyx_t_float_complex);
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_sumf(__pyx_t_float_complex, __pyx_t_float_complex);
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_difff(__pyx_t_float_complex, __pyx_t_float_complex);
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_prodf(__pyx_t_float_complex, __pyx_t_float_complex);
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_quotf(__pyx_t_float_complex, __pyx_t_float_complex);
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_negf(__pyx_t_float_complex);
    static CYTHON_INLINE int __Pyx_c_is_zerof(__pyx_t_float_complex);
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_conjf(__pyx_t_float_complex);
    #if 1
        static CYTHON_INLINE float __Pyx_c_absf(__pyx_t_float_complex);
        static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_powf(__pyx_t_float_complex, __pyx_t_float_complex);
    #endif
#endif

static CYTHON_INLINE __pyx_t_double_complex __pyx_t_double_complex_from_parts(double, double);

#if CYTHON_CCOMPLEX
    #define __Pyx_c_eq(a, b)   ((a)==(b))
    #define __Pyx_c_sum(a, b)  ((a)+(b))
    #define __Pyx_c_diff(a, b) ((a)-(b))
    #define __Pyx_c_prod(a, b) ((a)*(b))
    #define __Pyx_c_quot(a, b) ((a)/(b))
    #define __Pyx_c_neg(a)     (-(a))
  #ifdef __cplusplus
    #define __Pyx_c_is_zero(z) ((z)==(double)0)
    #define __Pyx_c_conj(z)    (::std::conj(z))
    #if 1
        #define __Pyx_c_abs(z)     (::std::abs(z))
        #define __Pyx_c_pow(a, b)  (::std::pow(a, b))
    #endif
  #else
    #define __Pyx_c_is_zero(z) ((z)==0)
    #define __Pyx_c_conj(z)    (conj(z))
    #if 1
        #define __Pyx_c_abs(z)     (cabs(z))
        #define __Pyx_c_pow(a, b)  (cpow(a, b))
    #endif
 #endif
#else
    static CYTHON_INLINE int __Pyx_c_eq(__pyx_t_double_complex, __pyx_t_double_complex);
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_sum(__pyx_t_double_complex, __pyx_t_double_complex);
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_diff(__pyx_t_double_complex, __pyx_t_double_complex);
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_prod(__pyx_t_double_complex, __pyx_t_double_complex);
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_quot(__pyx_t_double_complex, __pyx_t_double_complex);
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_neg(__pyx_t_double_complex);
    static CYTHON_INLINE int __Pyx_c_is_zero(__pyx_t_double_complex);
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_conj(__pyx_t_double_complex);
    #if 1
        static CYTHON_INLINE double __Pyx_c_abs(__pyx_t_double_complex);
        static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_pow(__pyx_t_double_complex, __pyx_t_double_complex);
    #endif
#endif

static CYTHON_INLINE long __Pyx_PyInt_As_long(PyObject *);

static int __Pyx_check_binary_version(void);

#if !defined(__Pyx_PyIdentifier_FromString)
#if PY_MAJOR_VERSION < 3
  #define __Pyx_PyIdentifier_FromString(s) PyString_FromString(s)
#else
  #define __Pyx_PyIdentifier_FromString(s) PyUnicode_FromString(s)
#endif
#endif

static PyObject *__Pyx_ImportModule(const char *name);

static PyTypeObject *__Pyx_ImportType(const char *module_name, const char *class_name, size_t size, int strict);

static int __Pyx_ImportVoidPtr(PyObject *module, const char *name, void **p, const char *sig);

static int __Pyx_InitStrings(__Pyx_StringTabEntry *t);

static struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_order, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___formBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___breakBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___loseRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainPair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___losePair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_action, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isOxygen(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSulfur(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_hasWildcards(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_makeSampleAtom(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_getBondOrdersForAtom(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSingle(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isDouble(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isTriple(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond___changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, short __pyx_v_order, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_action, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_makeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *__pyx_v_molecule, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom2, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_addAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_addBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_getBonds(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_f_5rmgpy_8molecule_5group_5Group_getBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_hasAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_hasBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_removeAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_removeBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_sortAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_f_5rmgpy_8molecule_5group_5Group_copy(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_copy *__pyx_optional_args); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_f_5rmgpy_8molecule_5group_5Group_merge(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_split(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_clearLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_containsLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_fromAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_adjlist, int __pyx_skip_dispatch); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_toAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList *__pyx_optional_args); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_updateFingerprint(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isIsomorphic *__pyx_optional_args); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_findIsomorphism(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findIsomorphism *__pyx_optional_args); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic *__pyx_optional_args); /* proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms *__pyx_optional_args); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isIdentical(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isAromaticRing(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_standardizeAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_addExplicitLigands(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_standardizeGroup(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_f_5rmgpy_8molecule_5group_5Group_addImplicitAtomsFromAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_f_5rmgpy_8molecule_5group_5Group_addImplicitBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *__pyx_f_5rmgpy_8molecule_5group_5Group_makeSampleMolecule(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isBenzeneExplicit(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch); /* proto*/

/* Module declarations from 'rmgpy.molecule.graph' */
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5graph_Vertex = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5graph_Edge = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5graph_Graph = 0;

/* Module declarations from 'rmgpy.molecule.atomtype' */
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType = 0;

/* Module declarations from 'rmgpy.molecule.element' */
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_7element_Element = 0;

/* Module declarations from 'rmgpy.constants' */
static double *__pyx_vp_5rmgpy_9constants_pi = 0;
#define __pyx_v_5rmgpy_9constants_pi (*__pyx_vp_5rmgpy_9constants_pi)
static double *__pyx_vp_5rmgpy_9constants_Na = 0;
#define __pyx_v_5rmgpy_9constants_Na (*__pyx_vp_5rmgpy_9constants_Na)
static double *__pyx_vp_5rmgpy_9constants_kB = 0;
#define __pyx_v_5rmgpy_9constants_kB (*__pyx_vp_5rmgpy_9constants_kB)
static double *__pyx_vp_5rmgpy_9constants_R = 0;
#define __pyx_v_5rmgpy_9constants_R (*__pyx_vp_5rmgpy_9constants_R)
static double *__pyx_vp_5rmgpy_9constants_h = 0;
#define __pyx_v_5rmgpy_9constants_h (*__pyx_vp_5rmgpy_9constants_h)
static double *__pyx_vp_5rmgpy_9constants_hbar = 0;
#define __pyx_v_5rmgpy_9constants_hbar (*__pyx_vp_5rmgpy_9constants_hbar)
static double *__pyx_vp_5rmgpy_9constants_c = 0;
#define __pyx_v_5rmgpy_9constants_c (*__pyx_vp_5rmgpy_9constants_c)
static double *__pyx_vp_5rmgpy_9constants_e = 0;
#define __pyx_v_5rmgpy_9constants_e (*__pyx_vp_5rmgpy_9constants_e)
static double *__pyx_vp_5rmgpy_9constants_m_e = 0;
#define __pyx_v_5rmgpy_9constants_m_e (*__pyx_vp_5rmgpy_9constants_m_e)
static double *__pyx_vp_5rmgpy_9constants_m_p = 0;
#define __pyx_v_5rmgpy_9constants_m_p (*__pyx_vp_5rmgpy_9constants_m_p)
static double *__pyx_vp_5rmgpy_9constants_m_n = 0;
#define __pyx_v_5rmgpy_9constants_m_n (*__pyx_vp_5rmgpy_9constants_m_n)
static double *__pyx_vp_5rmgpy_9constants_amu = 0;
#define __pyx_v_5rmgpy_9constants_amu (*__pyx_vp_5rmgpy_9constants_amu)
static double *__pyx_vp_5rmgpy_9constants_a0 = 0;
#define __pyx_v_5rmgpy_9constants_a0 (*__pyx_vp_5rmgpy_9constants_a0)
static double *__pyx_vp_5rmgpy_9constants_E_h = 0;
#define __pyx_v_5rmgpy_9constants_E_h (*__pyx_vp_5rmgpy_9constants_E_h)

/* Module declarations from 'cpython.buffer' */

/* Module declarations from 'cpython.ref' */

/* Module declarations from 'libc.string' */

/* Module declarations from 'libc.stdio' */

/* Module declarations from 'cpython.object' */

/* Module declarations from '__builtin__' */

/* Module declarations from 'cpython.type' */
static PyTypeObject *__pyx_ptype_7cpython_4type_type = 0;

/* Module declarations from 'libc.stdlib' */

/* Module declarations from 'numpy' */

/* Module declarations from 'numpy' */
static PyTypeObject *__pyx_ptype_5numpy_dtype = 0;
static PyTypeObject *__pyx_ptype_5numpy_flatiter = 0;
static PyTypeObject *__pyx_ptype_5numpy_broadcast = 0;
static PyTypeObject *__pyx_ptype_5numpy_ndarray = 0;
static PyTypeObject *__pyx_ptype_5numpy_ufunc = 0;
static CYTHON_INLINE char *__pyx_f_5numpy__util_dtypestring(PyArray_Descr *, char *, char *, int *); /*proto*/

/* Module declarations from 'rmgpy.molecule.molecule' */
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_8molecule_Atom = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_8molecule_Bond = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_8molecule_Molecule = 0;
static PyObject **__pyx_vp_5rmgpy_8molecule_8molecule_bond_orders = 0;
#define __pyx_v_5rmgpy_8molecule_8molecule_bond_orders (*__pyx_vp_5rmgpy_8molecule_8molecule_bond_orders)

/* Module declarations from 'cython' */

/* Module declarations from 'rmgpy.molecule.group' */
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5group_GroupBond = 0;
static PyTypeObject *__pyx_ptype_5rmgpy_8molecule_5group_Group = 0;
#define __Pyx_MODULE_NAME "rmgpy.molecule.group"
int __pyx_module_is_main_rmgpy__molecule__group = 0;

/* Implementation of 'rmgpy.molecule.group' */
static PyObject *__pyx_builtin_Exception;
static PyObject *__pyx_builtin_property;
static PyObject *__pyx_builtin_range;
static PyObject *__pyx_builtin_any;
static PyObject *__pyx_builtin_zip;
static PyObject *__pyx_builtin_all;
static PyObject *__pyx_builtin_ValueError;
static PyObject *__pyx_builtin_TypeError;
static PyObject *__pyx_builtin_enumerate;
static PyObject *__pyx_builtin_RuntimeError;
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom___init__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_atomType, PyObject *__pyx_v_radicalElectrons, PyObject *__pyx_v_charge, PyObject *__pyx_v_label, PyObject *__pyx_v_lonePairs); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_d); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6__str__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8__repr__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_10bonds(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_12copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16__formBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_26__losePair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_28applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_action); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_30equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_38hasWildcards(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_40makeSampleAtom(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_42getBondOrdersForAtom(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond___init__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_atom1, PyObject *__pyx_v_atom2, PyObject *__pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_2__str__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_4__repr__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_6__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_8copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_10isSingle(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_12isDouble(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_14isTriple(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_16isBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_18__changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, short __pyx_v_order); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_20applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_action); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_22equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_26makeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *__pyx_v_molecule, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom2); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group___init__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_atoms, PyObject *__pyx_v_multiplicity); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_2__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_4__getAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_6__setAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_atoms); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_8addAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_10addBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12getBonds(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_14getBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_16hasAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18hasBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_20removeAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_22removeBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_24sortAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_26copy(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_v_deep); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_28update(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_30merge(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_32split(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_38getLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_adjlist); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_44toAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_46updateFingerprint(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_48isIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_50findIsomorphism(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_56isIdentical(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_58isAromaticRing(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_60standardizeAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_62addExplicitLigands(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_64standardizeGroup(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_66addImplicitAtomsFromAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_12__defaults__(CYTHON_UNUSED PyObject *__pyx_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_classifyBenzeneCarbons(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_group, PyObject *__pyx_v_partners); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_2addBenzeneAtomToGroup(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_group, PyObject *__pyx_v_connectingAtom, PyObject *__pyx_v_cbf); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_4sortByConnectivity(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_atomList); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_6checkSet(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_superList, PyObject *__pyx_v_subList); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_8mergeOverlappingBenzeneRings(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_ring1, PyObject *__pyx_v_ring2, PyObject *__pyx_v_od); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_10addCbAtomToRing(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_ring, PyObject *__pyx_v_cbAtom); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_68addImplicitBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_70makeSampleMolecule(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_72isBenzeneExplicit(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self); /* proto */
static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value); /* proto */
static int __pyx_pf_5numpy_7ndarray___getbuffer__(PyArrayObject *__pyx_v_self, Py_buffer *__pyx_v_info, int __pyx_v_flags); /* proto */
static void __pyx_pf_5numpy_7ndarray_2__releasebuffer__(PyArrayObject *__pyx_v_self, Py_buffer *__pyx_v_info); /* proto */
static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_GroupAtom(PyTypeObject *t, PyObject *a, PyObject *k); /*proto*/
static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_GroupBond(PyTypeObject *t, PyObject *a, PyObject *k); /*proto*/
static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_Group(PyTypeObject *t, PyObject *a, PyObject *k); /*proto*/
static char __pyx_k_[] = "";
static char __pyx_k_0[] = "[{0}]";
static char __pyx_k_B[] = "B";
static char __pyx_k_C[] = "C";
static char __pyx_k_D[] = "D";
static char __pyx_k_H[] = "H";
static char __pyx_k_I[] = "I";
static char __pyx_k_L[] = "L";
static char __pyx_k_N[] = "N";
static char __pyx_k_O[] = "O";
static char __pyx_k_Q[] = "Q";
static char __pyx_k_S[] = "S";
static char __pyx_k_T[] = "T";
static char __pyx_k_b[] = "b";
static char __pyx_k_d[] = "d";
static char __pyx_k_f[] = "f";
static char __pyx_k_g[] = "g";
static char __pyx_k_h[] = "h";
static char __pyx_k_i[] = "i";
static char __pyx_k_l[] = "l";
static char __pyx_k_q[] = "q";
static char __pyx_k_x[] = "x";
static char __pyx_k_Ar[] = "Ar";
static char __pyx_k_CO[] = "CO";
static char __pyx_k_CS[] = "CS";
static char __pyx_k_Cb[] = "Cb";
static char __pyx_k_Cl[] = "Cl";
static char __pyx_k_He[] = "He";
static char __pyx_k_Ne[] = "Ne";
static char __pyx_k_Oa[] = "Oa";
static char __pyx_k_Od[] = "Od";
static char __pyx_k_Sa[] = "Sa";
static char __pyx_k_Sd[] = "Sd";
static char __pyx_k_Si[] = "Si";
static char __pyx_k_Zd[] = "Zd";
static char __pyx_k_Zf[] = "Zf";
static char __pyx_k_Zg[] = "Zg";
static char __pyx_k__2[] = ",";
static char __pyx_k_od[] = "od";
static char __pyx_k_x1[] = "x1";
static char __pyx_k_x2[] = "x2";
static char __pyx_k_Cbf[] = "Cbf";
static char __pyx_k_N1d[] = "N1d";
static char __pyx_k_N3b[] = "N3b";
static char __pyx_k_N5b[] = "N5b";
static char __pyx_k_N5d[] = "N5d";
static char __pyx_k_N5s[] = "N5s";
static char __pyx_k_N5t[] = "N5t";
static char __pyx_k_R_H[] = "R!H";
static char __pyx_k__26[] = "*";
static char __pyx_k_all[] = "all";
static char __pyx_k_any[] = "any";
static char __pyx_k_cbf[] = "cbf";
static char __pyx_k_doc[] = "__doc__";
static char __pyx_k_mol[] = "mol";
static char __pyx_k_pop[] = "pop";
static char __pyx_k_zip[] = "zip";
static char __pyx_k_Edge[] = "Edge";
static char __pyx_k_N5dd[] = "N5dd";
static char __pyx_k_atom[] = "atom";
static char __pyx_k_bond[] = "bond";
static char __pyx_k_copy[] = "copy";
static char __pyx_k_deep[] = "deep";
static char __pyx_k_init[] = "__init__";
static char __pyx_k_join[] = "join";
static char __pyx_k_keys[] = "keys";
static char __pyx_k_main[] = "__main__";
static char __pyx_k_ring[] = "ring";
static char __pyx_k_test[] = "__test__";
static char __pyx_k_Graph[] = "Graph";
static char __pyx_k_added[] = "added";
static char __pyx_k_atom1[] = "atom1";
static char __pyx_k_atom2[] = "atom2";
static char __pyx_k_atoms[] = "atoms";
static char __pyx_k_bonds[] = "bonds";
static char __pyx_k_class[] = "__class__";
static char __pyx_k_edges[] = "edges";
static char __pyx_k_graph[] = "graph";
static char __pyx_k_group[] = "group";
static char __pyx_k_label[] = "label";
static char __pyx_k_merge[] = "merge";
static char __pyx_k_order[] = "order";
static char __pyx_k_other[] = "other";
static char __pyx_k_range[] = "range";
static char __pyx_k_ring1[] = "ring1";
static char __pyx_k_ring2[] = "ring2";
static char __pyx_k_split[] = "split";
static char __pyx_k_upper[] = "upper";
static char __pyx_k_Vertex[] = "Vertex";
static char __pyx_k_append[] = "append";
static char __pyx_k_bond12[] = "bond12";
static char __pyx_k_cbAtom[] = "cbAtom";
static char __pyx_k_charge[] = "charge";
static char __pyx_k_format[] = "format";
static char __pyx_k_import[] = "__import__";
static char __pyx_k_module[] = "__module__";
static char __pyx_k_remove[] = "remove";
static char __pyx_k_subSet[] = "subSet";
static char __pyx_k_symbol[] = "symbol";
static char __pyx_k_update[] = "update";
static char __pyx_k_values[] = "values";
static char __pyx_k_addAtom[] = "addAtom";
static char __pyx_k_addBond[] = "addBond";
static char __pyx_k_adjlist[] = "adjlist";
static char __pyx_k_cbfAtom[] = "cbfAtom";
static char __pyx_k_element[] = "element";
static char __pyx_k_fbBonds[] = "fbBonds";
static char __pyx_k_getBond[] = "getBond";
static char __pyx_k_hasAtom[] = "hasAtom";
static char __pyx_k_hasBond[] = "hasBond";
static char __pyx_k_newAtom[] = "newAtom";
static char __pyx_k_newBond[] = "newBond";
static char __pyx_k_newRing[] = "newRing";
static char __pyx_k_prepare[] = "__prepare__";
static char __pyx_k_subList[] = "subList";
static char __pyx_k_atomList[] = "atomList";
static char __pyx_k_atomType[] = "atomType";
static char __pyx_k_atomtype[] = "atomtype";
static char __pyx_k_checkSet[] = "checkSet";
static char __pyx_k_deepcopy[] = "deepcopy";
static char __pyx_k_formBond[] = "__formBond";
static char __pyx_k_gainPair[] = "__gainPair";
static char __pyx_k_getAtoms[] = "__getAtoms";
static char __pyx_k_getBonds[] = "getBonds";
static char __pyx_k_isDouble[] = "isDouble";
static char __pyx_k_isOxygen[] = "isOxygen";
static char __pyx_k_isSingle[] = "isSingle";
static char __pyx_k_isSulfur[] = "isSulfur";
static char __pyx_k_isTriple[] = "isTriple";
static char __pyx_k_losePair[] = "__losePair";
static char __pyx_k_makeBond[] = "makeBond";
static char __pyx_k_molecule[] = "molecule";
static char __pyx_k_partners[] = "partners";
static char __pyx_k_property[] = "property";
static char __pyx_k_qualname[] = "__qualname__";
static char __pyx_k_setAtoms[] = "__setAtoms";
static char __pyx_k_specific[] = "specific";
static char __pyx_k_superSet[] = "superSet";
static char __pyx_k_vertices[] = "vertices";
static char __pyx_k_Exception[] = "Exception";
static char __pyx_k_FORM_BOND[] = "FORM_BOND";
static char __pyx_k_GAIN_PAIR[] = "GAIN_PAIR";
static char __pyx_k_LOSE_PAIR[] = "LOSE_PAIR";
static char __pyx_k_TypeError[] = "TypeError";
static char __pyx_k_atomTypes[] = "atomTypes";
static char __pyx_k_breakBond[] = "__breakBond";
static char __pyx_k_enumerate[] = "enumerate";
static char __pyx_k_isBenzene[] = "isBenzene";
static char __pyx_k_iteritems[] = "iteritems";
static char __pyx_k_lonePairs[] = "lonePairs";
static char __pyx_k_metaclass[] = "__metaclass__";
static char __pyx_k_sortAtoms[] = "sortAtoms";
static char __pyx_k_superList[] = "superList";
static char __pyx_k_BREAK_BOND[] = "BREAK_BOND";
static char __pyx_k_ValueError[] = "ValueError";
static char __pyx_k_cbAtomList[] = "cbAtomList";
static char __pyx_k_changeBond[] = "__changeBond";
static char __pyx_k_equivalent[] = "equivalent";
static char __pyx_k_formBond_2[] = "formBond";
static char __pyx_k_initialMap[] = "initialMap";
static char __pyx_k_issuperset[] = "issuperset";
static char __pyx_k_matchList1[] = "matchList1";
static char __pyx_k_matchList2[] = "matchList2";
static char __pyx_k_matchList3[] = "matchList3";
static char __pyx_k_matchList4[] = "matchList4";
static char __pyx_k_mergedRing[] = "mergedRing";
static char __pyx_k_pyx_vtable[] = "__pyx_vtable__";
static char __pyx_k_removeAtom[] = "removeAtom";
static char __pyx_k_removeBond[] = "removeBond";
static char __pyx_k_ActionError[] = "ActionError";
static char __pyx_k_CHANGE_BOND[] = "CHANGE_BOND";
static char __pyx_k_allElements[] = "allElements";
static char __pyx_k_applyAction[] = "applyAction";
static char __pyx_k_breakBond_2[] = "breakBond";
static char __pyx_k_cbfAtomList[] = "cbfAtomList";
static char __pyx_k_gainRadical[] = "__gainRadical";
static char __pyx_k_getFeatures[] = "getFeatures";
static char __pyx_k_isIdentical[] = "isIdentical";
static char __pyx_k_loseRadical[] = "__loseRadical";
static char __pyx_k_GAIN_RADICAL[] = "GAIN_RADICAL";
static char __pyx_k_LOSE_RADICAL[] = "LOSE_RADICAL";
static char __pyx_k_RuntimeError[] = "RuntimeError";
static char __pyx_k_benzeneBonds[] = "benzeneBonds";
static char __pyx_k_cbfAtomList1[] = "cbfAtomList1";
static char __pyx_k_cbfAtomList2[] = "cbfAtomList2";
static char __pyx_k_hasWildCards[] = "hasWildCards";
static char __pyx_k_hasWildcards[] = "hasWildcards";
static char __pyx_k_isIsomorphic[] = "isIsomorphic";
static char __pyx_k_multiplicity[] = "multiplicity";
static char __pyx_k_nonSpecifics[] = "nonSpecifics";
static char __pyx_k_sortingLabel[] = "sortingLabel";
static char __pyx_k_updateCharge[] = "updateCharge";
static char __pyx_k_GroupAtom_0_s[] = "<GroupAtom {0!s}>";
static char __pyx_k_GroupBond_0_r[] = "<GroupBond {0!r}>";
static char __pyx_k_connectedCbfs[] = "connectedCbfs";
static char __pyx_k_connectivity1[] = "connectivity1";
static char __pyx_k_connectivity2[] = "connectivity2";
static char __pyx_k_connectivity3[] = "connectivity3";
static char __pyx_k_decrementBond[] = "decrementBond";
static char __pyx_k_incrementBond[] = "incrementBond";
static char __pyx_k_PeriodicSystem[] = "PeriodicSystem";
static char __pyx_k_connectingAtom[] = "connectingAtom";
static char __pyx_k_getLabeledAtom[] = "getLabeledAtom";
static char __pyx_k_isAromaticRing[] = "isAromaticRing";
static char __pyx_k_makeSampleAtom[] = "makeSampleAtom";
static char __pyx_k_previousLength[] = "previousLength";
static char __pyx_k_sortedAtomList[] = "sortedAtomList";
static char __pyx_k_addCbAtomToRing[] = "addCbAtomToRing";
static char __pyx_k_findIsomorphism[] = "findIsomorphism";
static char __pyx_k_getLabeledAtoms[] = "getLabeledAtoms";
static char __pyx_k_toAdjacencyList[] = "toAdjacencyList";
static char __pyx_k_decrementRadical[] = "decrementRadical";
static char __pyx_k_incrementRadical[] = "incrementRadical";
static char __pyx_k_isSpecificCaseOf[] = "isSpecificCaseOf";
static char __pyx_k_radicalElectrons[] = "radicalElectrons";
static char __pyx_k_standardizeGroup[] = "standardizeGroup";
static char __pyx_k_clearLabeledAtoms[] = "clearLabeledAtoms";
static char __pyx_k_decrementLonePair[] = "decrementLonePair";
static char __pyx_k_fromAdjacencyList[] = "fromAdjacencyList";
static char __pyx_k_incrementLonePair[] = "incrementLonePair";
static char __pyx_k_isBenzeneExplicit[] = "isBenzeneExplicit";
static char __pyx_k_updateFingerprint[] = "updateFingerprint";
static char __pyx_k_valence_electrons[] = "valence_electrons";
static char __pyx_k_addExplicitLigands[] = "addExplicitLigands";
static char __pyx_k_addImplicitBenzene[] = "addImplicitBenzene";
static char __pyx_k_getAllCyclesOfSize[] = "getAllCyclesOfSize";
static char __pyx_k_makeSampleMolecule[] = "makeSampleMolecule";
static char __pyx_k_sortByConnectivity[] = "sortByConnectivity";
static char __pyx_k_containsLabeledAtom[] = "containsLabeledAtom";
static char __pyx_k_standardizeAtomType[] = "standardizeAtomType";
static char __pyx_k_ImplicitBenzeneError[] = "ImplicitBenzeneError";
static char __pyx_k_getBondOrdersForAtom[] = "getBondOrdersForAtom";
static char __pyx_k_isSubgraphIsomorphic[] = "isSubgraphIsomorphic";
static char __pyx_k_rmgpy_molecule_group[] = "rmgpy.molecule.group";
static char __pyx_k_addBenzeneAtomToGroup[] = "addBenzeneAtomToGroup";
static char __pyx_k_classifyBenzeneCarbons[] = "classifyBenzeneCarbons";
static char __pyx_k_labelsOfCarbonAtomTypes[] = "labelsOfCarbonAtomTypes";
static char __pyx_k_rmgpy_molecule_molecule[] = "rmgpy.molecule.molecule";
static char __pyx_k_findSubgraphIsomorphisms[] = "findSubgraphIsomorphisms";
static char __pyx_k_ndarray_is_not_C_contiguous[] = "ndarray is not C contiguous";
static char __pyx_k_addImplicitAtomsFromAtomType[] = "addImplicitAtomsFromAtomType";
static char __pyx_k_mergeOverlappingBenzeneRings[] = "mergeOverlappingBenzeneRings";
static char __pyx_k_0_has_more_than_three_Cbf_atoms[] = "{0} has more than three Cbf atoms and does not have fully explicit benzene rings.";
static char __pyx_k_An_exception_class_for_errors_t[] = "\n    An exception class for errors that occur while applying reaction recipe\n    actions. Pass a string describing the circumstances that caused the\n    exceptional behavior.\n    ";
static char __pyx_k_An_exception_class_when_encount[] = "\n    An exception class when encountering a group with too many implicit benzene\n    atoms. These groups are hard to create sample molecules and hard for users\n    to interpret. Pass a string describing the limitation.\n    ";
static char __pyx_k_Group_addImplicitBenzene_locals[] = "Group.addImplicitBenzene.<locals>.classifyBenzeneCarbons";
static char __pyx_k_No_atom_in_the_functional_group[] = "No atom in the functional group has the label \"{0}\".";
static char __pyx_k_This_module_provides_classes_an[] = "\nThis module provides classes and methods for working with molecular substructure\ngroups. These enable molecules to be searched for common motifs (e.g.\nreaction sites).\n";
static char __pyx_k_Users_agnes_PycharmProjects_RMG[] = "/Users/agnes/PycharmProjects/RMG-parallel/RMG-Py/rmgpy/molecule/group.py";
static char __pyx_k_unknown_dtype_code_in_numpy_pxd[] = "unknown dtype code in numpy.pxd (%d)";
static char __pyx_k_Cbf_atom_in_0_has_a_bond_with_an[] = "Cbf atom in {0} has a bond with an order other than 'B'";
static char __pyx_k_Format_string_allocated_too_shor[] = "Format string allocated too short, see comment in numpy.pxd";
static char __pyx_k_Got_a_0_object_for_parameter_oth[] = "Got a {0} object for parameter \"other\", when a Group object is required.";
static char __pyx_k_Non_native_byte_order_not_suppor[] = "Non-native byte order not supported";
static char __pyx_k_Unable_to_update_GroupAtom_Inval[] = "Unable to update GroupAtom: Invalid action {0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t[] = "Unable to update GroupAtom due to CHANGE_BOND action: Invalid order \"{0}\".";
static char __pyx_k_Unable_to_update_GroupBond_Inval[] = "Unable to update GroupBond: Invalid action {0}\".";
static char __pyx_k_Unable_to_update_GroupBond_due_t[] = "Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order \"{0}\" in set {1}\".";
static char __pyx_k_ndarray_is_not_Fortran_contiguou[] = "ndarray is not Fortran contiguous";
static char __pyx_k_Group_addImplicitBenzene_locals_2[] = "Group.addImplicitBenzene.<locals>.addBenzeneAtomToGroup";
static char __pyx_k_Group_addImplicitBenzene_locals_3[] = "Group.addImplicitBenzene.<locals>.sortByConnectivity";
static char __pyx_k_Group_addImplicitBenzene_locals_4[] = "Group.addImplicitBenzene.<locals>.checkSet";
static char __pyx_k_Group_addImplicitBenzene_locals_5[] = "Group.addImplicitBenzene.<locals>.mergeOverlappingBenzeneRings";
static char __pyx_k_Group_addImplicitBenzene_locals_6[] = "Group.addImplicitBenzene.<locals>.addCbAtomToRing";
static char __pyx_k_Format_string_allocated_too_shor_2[] = "Format string allocated too short.";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_2[] = "Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_3[] = "Unable to update GroupAtom due to FORM_BOND action: Invalid order \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_4[] = "Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_5[] = "Unable to update GroupAtom due to BREAK_BOND action: Invalid order \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_6[] = "Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_7[] = "Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_8[] = "Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_9[] = "Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupBond_due_t_2[] = "Unable to update GroupBond due to CHANGE_BOND action: Invalid order \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_10[] = "Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_11[] = "Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set \"{0}\".";
static char __pyx_k_Unable_to_update_GroupAtom_due_t_12[] = "Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set \"{0}\".";
static PyObject *__pyx_kp_s_;
static PyObject *__pyx_kp_s_0;
static PyObject *__pyx_kp_s_0_has_more_than_three_Cbf_atoms;
static PyObject *__pyx_n_s_ActionError;
static PyObject *__pyx_kp_s_An_exception_class_for_errors_t;
static PyObject *__pyx_kp_s_An_exception_class_when_encount;
static PyObject *__pyx_n_s_Ar;
static PyObject *__pyx_n_s_B;
static PyObject *__pyx_n_s_BREAK_BOND;
static PyObject *__pyx_n_s_C;
static PyObject *__pyx_n_s_CHANGE_BOND;
static PyObject *__pyx_n_s_CO;
static PyObject *__pyx_n_s_CS;
static PyObject *__pyx_n_s_Cb;
static PyObject *__pyx_n_s_Cbf;
static PyObject *__pyx_kp_s_Cbf_atom_in_0_has_a_bond_with_an;
static PyObject *__pyx_n_s_Cl;
static PyObject *__pyx_n_s_D;
static PyObject *__pyx_n_s_Edge;
static PyObject *__pyx_n_s_Exception;
static PyObject *__pyx_n_s_FORM_BOND;
static PyObject *__pyx_kp_u_Format_string_allocated_too_shor;
static PyObject *__pyx_kp_u_Format_string_allocated_too_shor_2;
static PyObject *__pyx_n_s_GAIN_PAIR;
static PyObject *__pyx_n_s_GAIN_RADICAL;
static PyObject *__pyx_kp_s_Got_a_0_object_for_parameter_oth;
static PyObject *__pyx_n_s_Graph;
static PyObject *__pyx_kp_s_GroupAtom_0_s;
static PyObject *__pyx_kp_s_GroupBond_0_r;
static PyObject *__pyx_n_s_Group_addImplicitBenzene_locals;
static PyObject *__pyx_n_s_Group_addImplicitBenzene_locals_2;
static PyObject *__pyx_n_s_Group_addImplicitBenzene_locals_3;
static PyObject *__pyx_n_s_Group_addImplicitBenzene_locals_4;
static PyObject *__pyx_n_s_Group_addImplicitBenzene_locals_5;
static PyObject *__pyx_n_s_Group_addImplicitBenzene_locals_6;
static PyObject *__pyx_n_s_H;
static PyObject *__pyx_n_s_He;
static PyObject *__pyx_n_s_ImplicitBenzeneError;
static PyObject *__pyx_n_s_LOSE_PAIR;
static PyObject *__pyx_n_s_LOSE_RADICAL;
static PyObject *__pyx_n_s_N;
static PyObject *__pyx_n_s_N1d;
static PyObject *__pyx_n_s_N3b;
static PyObject *__pyx_n_s_N5b;
static PyObject *__pyx_n_s_N5d;
static PyObject *__pyx_n_s_N5dd;
static PyObject *__pyx_n_s_N5s;
static PyObject *__pyx_n_s_N5t;
static PyObject *__pyx_n_s_Ne;
static PyObject *__pyx_kp_s_No_atom_in_the_functional_group;
static PyObject *__pyx_kp_u_Non_native_byte_order_not_suppor;
static PyObject *__pyx_n_s_O;
static PyObject *__pyx_n_s_Oa;
static PyObject *__pyx_n_s_Od;
static PyObject *__pyx_n_s_PeriodicSystem;
static PyObject *__pyx_kp_s_R_H;
static PyObject *__pyx_n_s_RuntimeError;
static PyObject *__pyx_n_s_S;
static PyObject *__pyx_n_s_Sa;
static PyObject *__pyx_n_s_Sd;
static PyObject *__pyx_n_s_Si;
static PyObject *__pyx_n_s_T;
static PyObject *__pyx_n_s_TypeError;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_Inval;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_10;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_11;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_12;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_2;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_3;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_4;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_5;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_6;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_7;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_8;
static PyObject *__pyx_kp_s_Unable_to_update_GroupAtom_due_t_9;
static PyObject *__pyx_kp_s_Unable_to_update_GroupBond_Inval;
static PyObject *__pyx_kp_s_Unable_to_update_GroupBond_due_t;
static PyObject *__pyx_kp_s_Unable_to_update_GroupBond_due_t_2;
static PyObject *__pyx_kp_s_Users_agnes_PycharmProjects_RMG;
static PyObject *__pyx_n_s_ValueError;
static PyObject *__pyx_n_s_Vertex;
static PyObject *__pyx_kp_s__2;
static PyObject *__pyx_n_s__26;
static PyObject *__pyx_n_s_addAtom;
static PyObject *__pyx_n_s_addBenzeneAtomToGroup;
static PyObject *__pyx_n_s_addBond;
static PyObject *__pyx_n_s_addCbAtomToRing;
static PyObject *__pyx_n_s_addExplicitLigands;
static PyObject *__pyx_n_s_addImplicitAtomsFromAtomType;
static PyObject *__pyx_n_s_addImplicitBenzene;
static PyObject *__pyx_n_s_added;
static PyObject *__pyx_n_s_adjlist;
static PyObject *__pyx_n_s_all;
static PyObject *__pyx_n_s_allElements;
static PyObject *__pyx_n_s_any;
static PyObject *__pyx_n_s_append;
static PyObject *__pyx_n_s_applyAction;
static PyObject *__pyx_n_s_atom;
static PyObject *__pyx_n_s_atom1;
static PyObject *__pyx_n_s_atom2;
static PyObject *__pyx_n_s_atomList;
static PyObject *__pyx_n_s_atomType;
static PyObject *__pyx_n_s_atomTypes;
static PyObject *__pyx_n_s_atoms;
static PyObject *__pyx_n_s_atomtype;
static PyObject *__pyx_n_s_benzeneBonds;
static PyObject *__pyx_n_s_bond;
static PyObject *__pyx_n_s_bond12;
static PyObject *__pyx_n_s_bonds;
static PyObject *__pyx_n_s_breakBond;
static PyObject *__pyx_n_s_breakBond_2;
static PyObject *__pyx_n_s_cbAtom;
static PyObject *__pyx_n_s_cbAtomList;
static PyObject *__pyx_n_s_cbf;
static PyObject *__pyx_n_s_cbfAtom;
static PyObject *__pyx_n_s_cbfAtomList;
static PyObject *__pyx_n_s_cbfAtomList1;
static PyObject *__pyx_n_s_cbfAtomList2;
static PyObject *__pyx_n_s_changeBond;
static PyObject *__pyx_n_s_charge;
static PyObject *__pyx_n_s_checkSet;
static PyObject *__pyx_n_s_class;
static PyObject *__pyx_n_s_classifyBenzeneCarbons;
static PyObject *__pyx_n_s_clearLabeledAtoms;
static PyObject *__pyx_n_s_connectedCbfs;
static PyObject *__pyx_n_s_connectingAtom;
static PyObject *__pyx_n_s_connectivity1;
static PyObject *__pyx_n_s_connectivity2;
static PyObject *__pyx_n_s_connectivity3;
static PyObject *__pyx_n_s_containsLabeledAtom;
static PyObject *__pyx_n_s_copy;
static PyObject *__pyx_n_s_decrementBond;
static PyObject *__pyx_n_s_decrementLonePair;
static PyObject *__pyx_n_s_decrementRadical;
static PyObject *__pyx_n_s_deep;
static PyObject *__pyx_n_s_deepcopy;
static PyObject *__pyx_n_s_doc;
static PyObject *__pyx_n_s_edges;
static PyObject *__pyx_n_s_element;
static PyObject *__pyx_n_s_enumerate;
static PyObject *__pyx_n_s_equivalent;
static PyObject *__pyx_n_s_fbBonds;
static PyObject *__pyx_n_s_findIsomorphism;
static PyObject *__pyx_n_s_findSubgraphIsomorphisms;
static PyObject *__pyx_n_s_formBond;
static PyObject *__pyx_n_s_formBond_2;
static PyObject *__pyx_n_s_format;
static PyObject *__pyx_n_s_fromAdjacencyList;
static PyObject *__pyx_n_s_gainPair;
static PyObject *__pyx_n_s_gainRadical;
static PyObject *__pyx_n_s_getAllCyclesOfSize;
static PyObject *__pyx_n_s_getAtoms;
static PyObject *__pyx_n_s_getBond;
static PyObject *__pyx_n_s_getBondOrdersForAtom;
static PyObject *__pyx_n_s_getBonds;
static PyObject *__pyx_n_s_getFeatures;
static PyObject *__pyx_n_s_getLabeledAtom;
static PyObject *__pyx_n_s_getLabeledAtoms;
static PyObject *__pyx_n_s_graph;
static PyObject *__pyx_n_s_group;
static PyObject *__pyx_n_s_hasAtom;
static PyObject *__pyx_n_s_hasBond;
static PyObject *__pyx_n_s_hasWildCards;
static PyObject *__pyx_n_s_hasWildcards;
static PyObject *__pyx_n_s_import;
static PyObject *__pyx_n_s_incrementBond;
static PyObject *__pyx_n_s_incrementLonePair;
static PyObject *__pyx_n_s_incrementRadical;
static PyObject *__pyx_n_s_init;
static PyObject *__pyx_n_s_initialMap;
static PyObject *__pyx_n_s_isAromaticRing;
static PyObject *__pyx_n_s_isBenzene;
static PyObject *__pyx_n_s_isBenzeneExplicit;
static PyObject *__pyx_n_s_isDouble;
static PyObject *__pyx_n_s_isIdentical;
static PyObject *__pyx_n_s_isIsomorphic;
static PyObject *__pyx_n_s_isOxygen;
static PyObject *__pyx_n_s_isSingle;
static PyObject *__pyx_n_s_isSpecificCaseOf;
static PyObject *__pyx_n_s_isSubgraphIsomorphic;
static PyObject *__pyx_n_s_isSulfur;
static PyObject *__pyx_n_s_isTriple;
static PyObject *__pyx_n_s_issuperset;
static PyObject *__pyx_n_s_iteritems;
static PyObject *__pyx_n_s_join;
static PyObject *__pyx_n_s_keys;
static PyObject *__pyx_n_s_label;
static PyObject *__pyx_n_s_labelsOfCarbonAtomTypes;
static PyObject *__pyx_n_s_lonePairs;
static PyObject *__pyx_n_s_losePair;
static PyObject *__pyx_n_s_loseRadical;
static PyObject *__pyx_n_s_main;
static PyObject *__pyx_n_s_makeBond;
static PyObject *__pyx_n_s_makeSampleAtom;
static PyObject *__pyx_n_s_makeSampleMolecule;
static PyObject *__pyx_n_s_matchList1;
static PyObject *__pyx_n_s_matchList2;
static PyObject *__pyx_n_s_matchList3;
static PyObject *__pyx_n_s_matchList4;
static PyObject *__pyx_n_s_merge;
static PyObject *__pyx_n_s_mergeOverlappingBenzeneRings;
static PyObject *__pyx_n_s_mergedRing;
static PyObject *__pyx_n_s_metaclass;
static PyObject *__pyx_n_s_module;
static PyObject *__pyx_n_s_mol;
static PyObject *__pyx_n_s_molecule;
static PyObject *__pyx_n_s_multiplicity;
static PyObject *__pyx_kp_u_ndarray_is_not_C_contiguous;
static PyObject *__pyx_kp_u_ndarray_is_not_Fortran_contiguou;
static PyObject *__pyx_n_s_newAtom;
static PyObject *__pyx_n_s_newBond;
static PyObject *__pyx_n_s_newRing;
static PyObject *__pyx_n_s_nonSpecifics;
static PyObject *__pyx_n_s_od;
static PyObject *__pyx_n_s_order;
static PyObject *__pyx_n_s_other;
static PyObject *__pyx_n_s_partners;
static PyObject *__pyx_n_s_pop;
static PyObject *__pyx_n_s_prepare;
static PyObject *__pyx_n_s_previousLength;
static PyObject *__pyx_n_s_property;
static PyObject *__pyx_n_s_pyx_vtable;
static PyObject *__pyx_n_s_qualname;
static PyObject *__pyx_n_s_radicalElectrons;
static PyObject *__pyx_n_s_range;
static PyObject *__pyx_n_s_remove;
static PyObject *__pyx_n_s_removeAtom;
static PyObject *__pyx_n_s_removeBond;
static PyObject *__pyx_n_s_ring;
static PyObject *__pyx_n_s_ring1;
static PyObject *__pyx_n_s_ring2;
static PyObject *__pyx_n_s_rmgpy_molecule_group;
static PyObject *__pyx_n_s_rmgpy_molecule_molecule;
static PyObject *__pyx_n_s_setAtoms;
static PyObject *__pyx_n_s_sortAtoms;
static PyObject *__pyx_n_s_sortByConnectivity;
static PyObject *__pyx_n_s_sortedAtomList;
static PyObject *__pyx_n_s_sortingLabel;
static PyObject *__pyx_n_s_specific;
static PyObject *__pyx_n_s_split;
static PyObject *__pyx_n_s_standardizeAtomType;
static PyObject *__pyx_n_s_standardizeGroup;
static PyObject *__pyx_n_s_subList;
static PyObject *__pyx_n_s_subSet;
static PyObject *__pyx_n_s_superList;
static PyObject *__pyx_n_s_superSet;
static PyObject *__pyx_n_s_symbol;
static PyObject *__pyx_n_s_test;
static PyObject *__pyx_n_s_toAdjacencyList;
static PyObject *__pyx_kp_u_unknown_dtype_code_in_numpy_pxd;
static PyObject *__pyx_n_s_update;
static PyObject *__pyx_n_s_updateCharge;
static PyObject *__pyx_n_s_updateFingerprint;
static PyObject *__pyx_n_s_upper;
static PyObject *__pyx_n_s_valence_electrons;
static PyObject *__pyx_n_s_values;
static PyObject *__pyx_n_s_vertices;
static PyObject *__pyx_n_s_x;
static PyObject *__pyx_n_s_x1;
static PyObject *__pyx_n_s_x2;
static PyObject *__pyx_n_s_zip;
static PyObject *__pyx_float_2_0;
static PyObject *__pyx_float_3_0;
static PyObject *__pyx_int_0;
static PyObject *__pyx_int_1;
static PyObject *__pyx_int_2;
static PyObject *__pyx_int_3;
static PyObject *__pyx_int_4;
static PyObject *__pyx_int_6;
static PyObject *__pyx_int_neg_1;
static PyObject *__pyx_int_neg_100;
static PyObject *__pyx_tuple__3;
static PyObject *__pyx_tuple__4;
static PyObject *__pyx_tuple__6;
static PyObject *__pyx_tuple__8;
static PyObject *__pyx_tuple__9;
static PyObject *__pyx_tuple__11;
static PyObject *__pyx_tuple__13;
static PyObject *__pyx_tuple__15;
static PyObject *__pyx_tuple__17;
static PyObject *__pyx_tuple__18;
static PyObject *__pyx_tuple__19;
static PyObject *__pyx_tuple__20;
static PyObject *__pyx_tuple__21;
static PyObject *__pyx_tuple__22;
static PyObject *__pyx_tuple__23;
static PyObject *__pyx_tuple__24;
static PyObject *__pyx_tuple__25;
static PyObject *__pyx_codeobj__5;
static PyObject *__pyx_codeobj__7;
static PyObject *__pyx_codeobj__10;
static PyObject *__pyx_codeobj__12;
static PyObject *__pyx_codeobj__14;
static PyObject *__pyx_codeobj__16;

/* "rmgpy/molecule/group.py":87
 *     """
 * 
 *     def __init__(self, atomType=None, radicalElectrons=None, charge=None, label='', lonePairs=None):             # <<<<<<<<<<<<<<
 *         Vertex.__init__(self)
 *         self.atomType = atomType or []
 */

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_atomType = 0;
  PyObject *__pyx_v_radicalElectrons = 0;
  PyObject *__pyx_v_charge = 0;
  PyObject *__pyx_v_label = 0;
  PyObject *__pyx_v_lonePairs = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__init__ (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atomType,&__pyx_n_s_radicalElectrons,&__pyx_n_s_charge,&__pyx_n_s_label,&__pyx_n_s_lonePairs,0};
    PyObject* values[5] = {0,0,0,0,0};
    values[0] = ((PyObject *)Py_None);
    values[1] = ((PyObject *)Py_None);
    values[2] = ((PyObject *)Py_None);
    values[3] = ((PyObject *)__pyx_kp_s_);
    values[4] = ((PyObject *)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
        case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atomType);
          if (value) { values[0] = value; kw_args--; }
        }
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_radicalElectrons);
          if (value) { values[1] = value; kw_args--; }
        }
        case  2:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_charge);
          if (value) { values[2] = value; kw_args--; }
        }
        case  3:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_label);
          if (value) { values[3] = value; kw_args--; }
        }
        case  4:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_lonePairs);
          if (value) { values[4] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__init__") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 87; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  5: values[4] = PyTuple_GET_ITEM(__pyx_args, 4);
        case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_atomType = values[0];
    __pyx_v_radicalElectrons = values[1];
    __pyx_v_charge = values[2];
    __pyx_v_label = values[3];
    __pyx_v_lonePairs = values[4];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("__init__", 0, 0, 5, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 87; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return -1;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom___init__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), __pyx_v_atomType, __pyx_v_radicalElectrons, __pyx_v_charge, __pyx_v_label, __pyx_v_lonePairs);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom___init__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_atomType, PyObject *__pyx_v_radicalElectrons, PyObject *__pyx_v_charge, PyObject *__pyx_v_label, PyObject *__pyx_v_lonePairs) {
  Py_ssize_t __pyx_v_index;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_t_8;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__init__", 0);

  /* "rmgpy/molecule/group.py":88
 * 
 *     def __init__(self, atomType=None, radicalElectrons=None, charge=None, label='', lonePairs=None):
 *         Vertex.__init__(self)             # <<<<<<<<<<<<<<
 *         self.atomType = atomType or []
 *         for index in range(len(self.atomType)):
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)), __pyx_n_s_init); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 88; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, ((PyObject *)__pyx_v_self)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 88; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 88; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(((PyObject *)__pyx_v_self));
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, ((PyObject *)__pyx_v_self));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 88; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":89
 *     def __init__(self, atomType=None, radicalElectrons=None, charge=None, label='', lonePairs=None):
 *         Vertex.__init__(self)
 *         self.atomType = atomType or []             # <<<<<<<<<<<<<<
 *         for index in range(len(self.atomType)):
 *             if isinstance(self.atomType[index], str):
 */
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_v_atomType); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 89; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_5) {
    goto __pyx_L4_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_atomType))||((__pyx_v_atomType) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_atomType)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 89; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_atomType);
    __pyx_t_1 = __pyx_v_atomType;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 89; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_1 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L3_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":90
 *         Vertex.__init__(self)
 *         self.atomType = atomType or []
 *         for index in range(len(self.atomType)):             # <<<<<<<<<<<<<<
 *             if isinstance(self.atomType[index], str):
 *                 self.atomType[index] = atomTypes[self.atomType[index]]
 */
  __pyx_t_1 = __pyx_v_self->atomType;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 90; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_6 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 90; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (__pyx_t_7 = 0; __pyx_t_7 < __pyx_t_6; __pyx_t_7+=1) {
    __pyx_v_index = __pyx_t_7;

    /* "rmgpy/molecule/group.py":91
 *         self.atomType = atomType or []
 *         for index in range(len(self.atomType)):
 *             if isinstance(self.atomType[index], str):             # <<<<<<<<<<<<<<
 *                 self.atomType[index] = atomTypes[self.atomType[index]]
 *         self.radicalElectrons = radicalElectrons or []
 */
    if (unlikely(__pyx_v_self->atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 91; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->atomType, __pyx_v_index, Py_ssize_t, 1, PyInt_FromSsize_t, 1, 1, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 91; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_5 = PyString_Check(__pyx_t_1); 
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_8 = (__pyx_t_5 != 0);
    if (__pyx_t_8) {

      /* "rmgpy/molecule/group.py":92
 *         for index in range(len(self.atomType)):
 *             if isinstance(self.atomType[index], str):
 *                 self.atomType[index] = atomTypes[self.atomType[index]]             # <<<<<<<<<<<<<<
 *         self.radicalElectrons = radicalElectrons or []
 *         self.charge = charge or []
 */
      __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 92; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      if (unlikely(__pyx_v_self->atomType == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 92; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_self->atomType, __pyx_v_index, Py_ssize_t, 1, PyInt_FromSsize_t, 1, 1, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 92; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_4 = PyObject_GetItem(__pyx_t_1, __pyx_t_2); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 92; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (unlikely(__pyx_v_self->atomType == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 92; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      if (unlikely(__Pyx_SetItemInt(__pyx_v_self->atomType, __pyx_v_index, __pyx_t_4, Py_ssize_t, 1, PyInt_FromSsize_t, 1, 1, 1) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 92; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      goto __pyx_L7;
    }
    __pyx_L7:;
  }

  /* "rmgpy/molecule/group.py":93
 *             if isinstance(self.atomType[index], str):
 *                 self.atomType[index] = atomTypes[self.atomType[index]]
 *         self.radicalElectrons = radicalElectrons or []             # <<<<<<<<<<<<<<
 *         self.charge = charge or []
 *         self.label = label
 */
  __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_v_radicalElectrons); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 93; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_8) {
    goto __pyx_L9_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_radicalElectrons))||((__pyx_v_radicalElectrons) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_radicalElectrons)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 93; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_radicalElectrons);
    __pyx_t_4 = __pyx_v_radicalElectrons;
    goto __pyx_L8_bool_binop_done;
  }
  __pyx_L9_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 93; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_4 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L8_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_4);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = ((PyObject*)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":94
 *                 self.atomType[index] = atomTypes[self.atomType[index]]
 *         self.radicalElectrons = radicalElectrons or []
 *         self.charge = charge or []             # <<<<<<<<<<<<<<
 *         self.label = label
 *         self.lonePairs = lonePairs or []
 */
  __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_v_charge); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 94; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_8) {
    goto __pyx_L11_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_charge))||((__pyx_v_charge) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_charge)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 94; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_charge);
    __pyx_t_4 = __pyx_v_charge;
    goto __pyx_L10_bool_binop_done;
  }
  __pyx_L11_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 94; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_4 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L10_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_4);
  __Pyx_GOTREF(__pyx_v_self->charge);
  __Pyx_DECREF(__pyx_v_self->charge);
  __pyx_v_self->charge = ((PyObject*)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":95
 *         self.radicalElectrons = radicalElectrons or []
 *         self.charge = charge or []
 *         self.label = label             # <<<<<<<<<<<<<<
 *         self.lonePairs = lonePairs or []
 * 
 */
  if (!(likely(PyString_CheckExact(__pyx_v_label))||((__pyx_v_label) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_v_label)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 95; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_4 = __pyx_v_label;
  __Pyx_INCREF(__pyx_t_4);
  __Pyx_GIVEREF(__pyx_t_4);
  __Pyx_GOTREF(__pyx_v_self->label);
  __Pyx_DECREF(__pyx_v_self->label);
  __pyx_v_self->label = ((PyObject*)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":96
 *         self.charge = charge or []
 *         self.label = label
 *         self.lonePairs = lonePairs or []             # <<<<<<<<<<<<<<
 * 
 *     def __reduce__(self):
 */
  __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_v_lonePairs); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 96; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_8) {
    goto __pyx_L13_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_lonePairs))||((__pyx_v_lonePairs) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_lonePairs)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 96; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_lonePairs);
    __pyx_t_4 = __pyx_v_lonePairs;
    goto __pyx_L12_bool_binop_done;
  }
  __pyx_L13_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 96; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_4 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L12_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_4);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":87
 *     """
 * 
 *     def __init__(self, atomType=None, radicalElectrons=None, charge=None, label='', lonePairs=None):             # <<<<<<<<<<<<<<
 *         Vertex.__init__(self)
 *         self.atomType = atomType or []
 */

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":98
 *         self.lonePairs = lonePairs or []
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_3__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__[] = "\n        A helper function used when pickling an object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_3__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__reduce__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_v_d = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_v_a = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_t_3;
  int __pyx_t_4;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__reduce__", 0);

  /* "rmgpy/molecule/group.py":102
 *         A helper function used when pickling an object.
 *         """
 *         d = {             # <<<<<<<<<<<<<<
 *             'edges': self.edges,
 *             'connectivity1': self.connectivity1,
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 102; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);

  /* "rmgpy/molecule/group.py":103
 *         """
 *         d = {
 *             'edges': self.edges,             # <<<<<<<<<<<<<<
 *             'connectivity1': self.connectivity1,
 *             'connectivity2': self.connectivity2,
 */
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_edges, __pyx_v_self->__pyx_base.edges) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 102; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":104
 *         d = {
 *             'edges': self.edges,
 *             'connectivity1': self.connectivity1,             # <<<<<<<<<<<<<<
 *             'connectivity2': self.connectivity2,
 *             'connectivity3': self.connectivity3,
 */
  __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_self->__pyx_base.connectivity1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 104; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_connectivity1, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 102; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":105
 *             'edges': self.edges,
 *             'connectivity1': self.connectivity1,
 *             'connectivity2': self.connectivity2,             # <<<<<<<<<<<<<<
 *             'connectivity3': self.connectivity3,
 *             'sortingLabel': self.sortingLabel,
 */
  __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_self->__pyx_base.connectivity2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 105; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_connectivity2, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 102; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":106
 *             'connectivity1': self.connectivity1,
 *             'connectivity2': self.connectivity2,
 *             'connectivity3': self.connectivity3,             # <<<<<<<<<<<<<<
 *             'sortingLabel': self.sortingLabel,
 *         }
 */
  __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_self->__pyx_base.connectivity3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 106; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_connectivity3, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 102; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":107
 *             'connectivity2': self.connectivity2,
 *             'connectivity3': self.connectivity3,
 *             'sortingLabel': self.sortingLabel,             # <<<<<<<<<<<<<<
 *         }
 *         atomType = self.atomType
 */
  __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_self->__pyx_base.sortingLabel); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 107; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_sortingLabel, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 102; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_d = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":109
 *             'sortingLabel': self.sortingLabel,
 *         }
 *         atomType = self.atomType             # <<<<<<<<<<<<<<
 *         if atomType is not None:
 *             atomType = [a.label for a in atomType]
 */
  __pyx_t_1 = __pyx_v_self->atomType;
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":110
 *         }
 *         atomType = self.atomType
 *         if atomType is not None:             # <<<<<<<<<<<<<<
 *             atomType = [a.label for a in atomType]
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)
 */
  __pyx_t_3 = (__pyx_v_atomType != ((PyObject*)Py_None));
  __pyx_t_4 = (__pyx_t_3 != 0);
  if (__pyx_t_4) {

    /* "rmgpy/molecule/group.py":111
 *         atomType = self.atomType
 *         if atomType is not None:
 *             atomType = [a.label for a in atomType]             # <<<<<<<<<<<<<<
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)
 * 
 */
    __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (unlikely(__pyx_v_atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_5 = 0;
    for (;;) {
      if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_6 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_6); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_6 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_a, __pyx_t_6);
      __pyx_t_6 = 0;
      __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_v_a, __pyx_n_s_label); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_6))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF_SET(__pyx_v_atomType, ((PyObject*)__pyx_t_1));
    __pyx_t_1 = 0;
    goto __pyx_L3;
  }
  __pyx_L3:;

  /* "rmgpy/molecule/group.py":112
 *         if atomType is not None:
 *             atomType = [a.label for a in atomType]
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)             # <<<<<<<<<<<<<<
 * 
 *     def __setstate__(self, d):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(5); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 112; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_atomType);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_atomType);
  __Pyx_GIVEREF(__pyx_v_atomType);
  __Pyx_INCREF(__pyx_v_self->radicalElectrons);
  PyTuple_SET_ITEM(__pyx_t_1, 1, __pyx_v_self->radicalElectrons);
  __Pyx_GIVEREF(__pyx_v_self->radicalElectrons);
  __Pyx_INCREF(__pyx_v_self->charge);
  PyTuple_SET_ITEM(__pyx_t_1, 2, __pyx_v_self->charge);
  __Pyx_GIVEREF(__pyx_v_self->charge);
  __Pyx_INCREF(__pyx_v_self->label);
  PyTuple_SET_ITEM(__pyx_t_1, 3, __pyx_v_self->label);
  __Pyx_GIVEREF(__pyx_v_self->label);
  __Pyx_INCREF(__pyx_v_self->lonePairs);
  PyTuple_SET_ITEM(__pyx_t_1, 4, __pyx_v_self->lonePairs);
  __Pyx_GIVEREF(__pyx_v_self->lonePairs);
  __pyx_t_2 = PyTuple_New(3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 112; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)));
  PyTuple_SET_ITEM(__pyx_t_2, 0, ((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)));
  __Pyx_GIVEREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)));
  PyTuple_SET_ITEM(__pyx_t_2, 1, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_d);
  PyTuple_SET_ITEM(__pyx_t_2, 2, __pyx_v_d);
  __Pyx_GIVEREF(__pyx_v_d);
  __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":98
 *         self.lonePairs = lonePairs or []
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__reduce__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_d);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XDECREF(__pyx_v_a);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":114
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)
 * 
 *     def __setstate__(self, d):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when unpickling an object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5__setstate__(PyObject *__pyx_v_self, PyObject *__pyx_v_d); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__[] = "\n        A helper function used when unpickling an object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5__setstate__(PyObject *__pyx_v_self, PyObject *__pyx_v_d) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__setstate__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_d));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_d) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  short __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__setstate__", 0);

  /* "rmgpy/molecule/group.py":118
 *         A helper function used when unpickling an object.
 *         """
 *         self.edges = d['edges']             # <<<<<<<<<<<<<<
 *         self.connectivity1 = d['connectivity1']
 *         self.connectivity2 = d['connectivity2']
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_edges); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 118; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  if (!(likely(PyDict_CheckExact(__pyx_t_1))||((__pyx_t_1) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "dict", Py_TYPE(__pyx_t_1)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 118; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->__pyx_base.edges);
  __Pyx_DECREF(__pyx_v_self->__pyx_base.edges);
  __pyx_v_self->__pyx_base.edges = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":119
 *         """
 *         self.edges = d['edges']
 *         self.connectivity1 = d['connectivity1']             # <<<<<<<<<<<<<<
 *         self.connectivity2 = d['connectivity2']
 *         self.connectivity3 = d['connectivity3']
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_connectivity1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 119; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_2 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_self->__pyx_base.connectivity1 = __pyx_t_2;

  /* "rmgpy/molecule/group.py":120
 *         self.edges = d['edges']
 *         self.connectivity1 = d['connectivity1']
 *         self.connectivity2 = d['connectivity2']             # <<<<<<<<<<<<<<
 *         self.connectivity3 = d['connectivity3']
 *         self.sortingLabel = d['sortingLabel']
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_connectivity2); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 120; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_2 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 120; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_self->__pyx_base.connectivity2 = __pyx_t_2;

  /* "rmgpy/molecule/group.py":121
 *         self.connectivity1 = d['connectivity1']
 *         self.connectivity2 = d['connectivity2']
 *         self.connectivity3 = d['connectivity3']             # <<<<<<<<<<<<<<
 *         self.sortingLabel = d['sortingLabel']
 * 
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_connectivity3); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 121; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_2 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_self->__pyx_base.connectivity3 = __pyx_t_2;

  /* "rmgpy/molecule/group.py":122
 *         self.connectivity2 = d['connectivity2']
 *         self.connectivity3 = d['connectivity3']
 *         self.sortingLabel = d['sortingLabel']             # <<<<<<<<<<<<<<
 * 
 *     def __str__(self):
 */
  __pyx_t_1 = PyObject_GetItem(__pyx_v_d, __pyx_n_s_sortingLabel); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 122; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_2 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 122; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_self->__pyx_base.sortingLabel = __pyx_t_2;

  /* "rmgpy/molecule/group.py":114
 *         return (GroupAtom, (atomType, self.radicalElectrons, self.charge, self.label, self.lonePairs), d)
 * 
 *     def __setstate__(self, d):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when unpickling an object.
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__setstate__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":124
 *         self.sortingLabel = d['sortingLabel']
 * 
 *     def __str__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a human-readable string representation of the object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_7__str__(PyObject *__pyx_v_self); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_6__str__[] = "\n        Return a human-readable string representation of the object.\n        ";
#if CYTHON_COMPILING_IN_CPYTHON
struct wrapperbase __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_6__str__;
#endif
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_7__str__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__str__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6__str__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6__str__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_v_a = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  PyObject *__pyx_t_7 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__str__", 0);

  /* "rmgpy/molecule/group.py":128
 *         Return a human-readable string representation of the object.
 *         """
 *         return '[{0}]'.format(','.join([repr(a.label) for a in self.atomType]))             # <<<<<<<<<<<<<<
 * 
 *     def __repr__(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_0, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_4 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_4); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_4)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_6 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_5); __Pyx_INCREF(__pyx_t_6); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_6 = PySequence_ITEM(__pyx_t_4, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_a, __pyx_t_6);
    __pyx_t_6 = 0;
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_v_a, __pyx_n_s_label); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_7 = PyObject_Repr(__pyx_t_6); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_3, (PyObject*)__pyx_t_7))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  }
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_t_4 = __Pyx_PyString_Join(__pyx_kp_s__2, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_7 = PyTuple_New(1+1); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    PyTuple_SET_ITEM(__pyx_t_7, 0+1, __pyx_t_4);
    __Pyx_GIVEREF(__pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_7, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":124
 *         self.sortingLabel = d['sortingLabel']
 * 
 *     def __str__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a human-readable string representation of the object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_7);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__str__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_a);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":130
 *         return '[{0}]'.format(','.join([repr(a.label) for a in self.atomType]))
 * 
 *     def __repr__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a representation that can be used to reconstruct the object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9__repr__(PyObject *__pyx_v_self); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_8__repr__[] = "\n        Return a representation that can be used to reconstruct the object.\n        ";
#if CYTHON_COMPILING_IN_CPYTHON
struct wrapperbase __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_8__repr__;
#endif
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9__repr__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__repr__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8__repr__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8__repr__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__repr__", 0);

  /* "rmgpy/molecule/group.py":134
 *         Return a representation that can be used to reconstruct the object.
 *         """
 *         return "<GroupAtom {0!s}>".format(self)             # <<<<<<<<<<<<<<
 * 
 *     @property
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_GroupAtom_0_s, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 134; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, ((PyObject *)__pyx_v_self)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 134; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 134; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(((PyObject *)__pyx_v_self));
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, ((PyObject *)__pyx_v_self));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 134; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":130
 *         return '[{0}]'.format(','.join([repr(a.label) for a in self.atomType]))
 * 
 *     def __repr__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a representation that can be used to reconstruct the object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__repr__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":137
 * 
 *     @property
 *     def bonds(self): return self.edges             # <<<<<<<<<<<<<<
 * 
 *     def copy(self):
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_11bonds(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_11bonds(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("bonds (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_10bonds(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_10bonds(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("bonds", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->__pyx_base.edges);
  __pyx_r = __pyx_v_self->__pyx_base.edges;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":139
 *     def bonds(self): return self.edges
 * 
 *     def copy(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a deep copy of the :class:`GroupAtom` object. Modifying the
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_copy); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 139; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 139; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 139; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5graph_Vertex))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 139; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":144
 *         attributes of the copy will not affect the original.
 *         """
 *         return GroupAtom(self.atomType[:], self.radicalElectrons[:], self.charge[:], self.label)             # <<<<<<<<<<<<<<
 * 
 *     def __changeBond(self, order):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_PyList_GetSlice(__pyx_v_self->atomType, 0, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __Pyx_PyList_GetSlice(__pyx_v_self->radicalElectrons, 0, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (unlikely(__pyx_v_self->charge == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_3 = __Pyx_PyList_GetSlice(__pyx_v_self->charge, 0, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_4 = PyTuple_New(4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  PyTuple_SET_ITEM(__pyx_t_4, 1, __pyx_t_2);
  __Pyx_GIVEREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_4, 2, __pyx_t_3);
  __Pyx_GIVEREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_v_self->label);
  PyTuple_SET_ITEM(__pyx_t_4, 3, __pyx_v_self->label);
  __Pyx_GIVEREF(__pyx_v_self->label);
  __pyx_t_1 = 0;
  __pyx_t_2 = 0;
  __pyx_t_3 = 0;
  __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_t_4, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 144; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_t_3);
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":139
 *     def bonds(self): return self.edges
 * 
 *     def copy(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a deep copy of the :class:`GroupAtom` object. Modifying the
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_12copy[] = "\n        Return a deep copy of the :class:`GroupAtom` object. Modifying the\n        attributes of the copy will not affect the original.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("copy (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_12copy(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_12copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_copy(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 139; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":146
 *         return GroupAtom(self.atomType[:], self.radicalElectrons[:], self.charge[:], self.label)
 * 
 *     def __changeBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a CHANGE_BOND action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_order, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  int __pyx_t_8;
  PyObject *__pyx_t_9 = NULL;
  PyObject *__pyx_t_10 = NULL;
  int __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__changeBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_changeBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_order); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":152
 *         in bond order, and should be 1 or -1.
 *         """
 *         atomType = []             # <<<<<<<<<<<<<<
 *         for atom in self.atomType:
 *             if order == 1:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 152; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":153
 *         """
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":156
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)
 *             elif order == -1:             # <<<<<<<<<<<<<<
 *                 atomType.extend(atom.decrementBond)
 *             else:
 */
    switch (__pyx_v_order) {

      /* "rmgpy/molecule/group.py":154
 *         atomType = []
 *         for atom in self.atomType:
 *             if order == 1:             # <<<<<<<<<<<<<<
 *                 atomType.extend(atom.incrementBond)
 *             elif order == -1:
 */
      case 1:

      /* "rmgpy/molecule/group.py":155
 *         for atom in self.atomType:
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)             # <<<<<<<<<<<<<<
 *             elif order == -1:
 *                 atomType.extend(atom.decrementBond)
 */
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_incrementBond); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 155; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_8 = __Pyx_PyList_Extend(__pyx_v_atomType, __pyx_t_2); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 155; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      break;

      /* "rmgpy/molecule/group.py":156
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)
 *             elif order == -1:             # <<<<<<<<<<<<<<
 *                 atomType.extend(atom.decrementBond)
 *             else:
 */
      case -1:

      /* "rmgpy/molecule/group.py":157
 *                 atomType.extend(atom.incrementBond)
 *             elif order == -1:
 *                 atomType.extend(atom.decrementBond)             # <<<<<<<<<<<<<<
 *             else:
 *                 raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 */
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_decrementBond); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 157; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_8 = __Pyx_PyList_Extend(__pyx_v_atomType, __pyx_t_2); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 157; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      break;
      default:

      /* "rmgpy/molecule/group.py":159
 *                 atomType.extend(atom.decrementBond)
 *             else:
 *                 raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Invalid order "{0}".'.format(order))             # <<<<<<<<<<<<<<
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_5 = __Pyx_PyInt_From_short(__pyx_v_order); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      __pyx_t_9 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_9 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_9)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_9);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_9) {
        __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        __Pyx_GOTREF(__pyx_t_6);
      } else {
        __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_9); __Pyx_GIVEREF(__pyx_t_9); __pyx_t_9 = NULL;
        PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_t_5);
        __Pyx_GIVEREF(__pyx_t_5);
        __pyx_t_5 = 0;
        __pyx_t_6 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_10, NULL); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_3 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_3)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_3);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_3) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_6); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_t_6);
        __Pyx_GIVEREF(__pyx_t_6);
        __pyx_t_6 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_10, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      break;
    }

    /* "rmgpy/molecule/group.py":153
 *         """
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             if order == 1:
 *                 atomType.extend(atom.incrementBond)
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":160
 *             else:
 *                 raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 *         if len(atomType) == 0:             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 */
  __pyx_t_7 = PyList_GET_SIZE(__pyx_v_atomType); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 160; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_11 = ((__pyx_t_7 == 0) != 0);
  if (__pyx_t_11) {

    /* "rmgpy/molecule/group.py":161
 *                 raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_10 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_2, __pyx_n_s_format); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_10);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_10))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_10);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_10);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_10, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_10, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_3 = PyTuple_New(1+1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_3, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_10, __pyx_t_3, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    }
    __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
    __pyx_t_10 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_10)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_10);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_10) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_3 = PyTuple_New(1+1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_10); __Pyx_GIVEREF(__pyx_t_10); __pyx_t_10 = NULL;
      PyTuple_SET_ITEM(__pyx_t_3, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_3, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":163
 *             raise ActionError('Unable to update GroupAtom due to CHANGE_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))             # <<<<<<<<<<<<<<
 * 
 *     def __formBond(self, order):
 */
  __pyx_t_1 = PySet_New(__pyx_v_atomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 163; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 163; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyList_Type))), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 163; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":146
 *         return GroupAtom(self.atomType[:], self.radicalElectrons[:], self.charge[:], self.label)
 * 
 *     def __changeBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a CHANGE_BOND action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_9);
  __Pyx_XDECREF(__pyx_t_10);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond[] = "\n        Update the atom group as a result of applying a CHANGE_BOND action,\n        where `order` specifies whether the bond is incremented or decremented\n        in bond order, and should be 1 or -1.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order) {
  short __pyx_v_order;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__changeBond (wrapper)", 0);
  assert(__pyx_arg_order); {
    __pyx_v_order = __Pyx_PyInt_As_short(__pyx_arg_order); if (unlikely((__pyx_v_order == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_order));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_order) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__changeBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___changeBond(__pyx_v_self, __pyx_v_order, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 146; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":165
 *         self.atomType = list(set(atomType))
 * 
 *     def __formBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a FORM_BOND action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___formBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  PyObject *__pyx_t_8 = NULL;
  Py_ssize_t __pyx_t_9;
  int __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__formBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_formBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_order); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_order);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_order);
        __Pyx_GIVEREF(__pyx_v_order);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":171
 *         'S' (since we only allow forming of single bonds).
 *         """
 *         if order != 'S':             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_order, __pyx_n_s_S, Py_NE)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 171; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = (__pyx_t_6 != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":172
 *         """
 *         if order != 'S':
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))             # <<<<<<<<<<<<<<
 *         atomType = []
 *         for atom in self.atomType:
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_3, __pyx_n_s_format); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_5);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_5, function);
      }
    }
    if (!__pyx_t_4) {
      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_v_order); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      __Pyx_INCREF(__pyx_v_order);
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_v_order);
      __Pyx_GIVEREF(__pyx_v_order);
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_8, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 172; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":173
 *         if order != 'S':
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []             # <<<<<<<<<<<<<<
 *         for atom in self.atomType:
 *             atomType.extend(atom.formBond)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 173; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":174
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             atomType.extend(atom.formBond)
 *         if len(atomType) == 0:
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 174; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_9 = 0;
  for (;;) {
    if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_9); __Pyx_INCREF(__pyx_t_2); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 174; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 174; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":175
 *         atomType = []
 *         for atom in self.atomType:
 *             atomType.extend(atom.formBond)             # <<<<<<<<<<<<<<
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_formBond_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 175; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_10 = __Pyx_PyList_Extend(__pyx_v_atomType, __pyx_t_2); if (unlikely(__pyx_t_10 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 175; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":174
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             atomType.extend(atom.formBond)
 *         if len(atomType) == 0:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":176
 *         for atom in self.atomType:
 *             atomType.extend(atom.formBond)
 *         if len(atomType) == 0:             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 */
  __pyx_t_9 = PyList_GET_SIZE(__pyx_v_atomType); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 176; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = ((__pyx_t_9 == 0) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":177
 *             atomType.extend(atom.formBond)
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_4, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_self->atomType); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_4, NULL); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_8); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_8);
      __Pyx_GIVEREF(__pyx_t_8);
      __pyx_t_8 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 177; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":179
 *             raise ActionError('Unable to update GroupAtom due to FORM_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))             # <<<<<<<<<<<<<<
 * 
 *     def __breakBond(self, order):
 */
  __pyx_t_1 = PySet_New(__pyx_v_atomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 179; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 179; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyList_Type))), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 179; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":165
 *         self.atomType = list(set(atomType))
 * 
 *     def __formBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a FORM_BOND action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_8);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__formBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_16__formBond[] = "\n        Update the atom group as a result of applying a FORM_BOND action,\n        where `order` specifies the order of the forming bond, and should be\n        'S' (since we only allow forming of single bonds).\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__formBond (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_order), (&PyString_Type), 1, "order", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16__formBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject*)__pyx_v_order));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16__formBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__formBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___formBond(__pyx_v_self, __pyx_v_order, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__formBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":181
 *         self.atomType = list(set(atomType))
 * 
 *     def __breakBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a BREAK_BOND action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___breakBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  PyObject *__pyx_t_8 = NULL;
  Py_ssize_t __pyx_t_9;
  int __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__breakBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_breakBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 181; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_order); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 181; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 181; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_order);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_order);
        __Pyx_GIVEREF(__pyx_v_order);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 181; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":187
 *         'S' (since we only allow breaking of single bonds).
 *         """
 *         if order != 'S':             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_order, __pyx_n_s_S, Py_NE)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 187; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = (__pyx_t_6 != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":188
 *         """
 *         if order != 'S':
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))             # <<<<<<<<<<<<<<
 *         atomType = []
 *         for atom in self.atomType:
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_5, __pyx_n_s_format); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_5);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_5, function);
      }
    }
    if (!__pyx_t_4) {
      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_v_order); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      __Pyx_INCREF(__pyx_v_order);
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_v_order);
      __Pyx_GIVEREF(__pyx_v_order);
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_8, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":189
 *         if order != 'S':
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []             # <<<<<<<<<<<<<<
 *         for atom in self.atomType:
 *             atomType.extend(atom.breakBond)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 189; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":190
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             atomType.extend(atom.breakBond)
 *         if len(atomType) == 0:
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 190; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_9 = 0;
  for (;;) {
    if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_9); __Pyx_INCREF(__pyx_t_2); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 190; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 190; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":191
 *         atomType = []
 *         for atom in self.atomType:
 *             atomType.extend(atom.breakBond)             # <<<<<<<<<<<<<<
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_breakBond_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 191; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_10 = __Pyx_PyList_Extend(__pyx_v_atomType, __pyx_t_2); if (unlikely(__pyx_t_10 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 191; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":190
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Invalid order "{0}".'.format(order))
 *         atomType = []
 *         for atom in self.atomType:             # <<<<<<<<<<<<<<
 *             atomType.extend(atom.breakBond)
 *         if len(atomType) == 0:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":192
 *         for atom in self.atomType:
 *             atomType.extend(atom.breakBond)
 *         if len(atomType) == 0:             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 */
  __pyx_t_9 = PyList_GET_SIZE(__pyx_v_atomType); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 192; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = ((__pyx_t_9 == 0) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":193
 *             atomType.extend(atom.breakBond)
 *         if len(atomType) == 0:
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_6, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_8 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_self->atomType); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_8 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_4, NULL); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_8); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_8);
      __Pyx_GIVEREF(__pyx_t_8);
      __pyx_t_8 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 193; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":195
 *             raise ActionError('Unable to update GroupAtom due to BREAK_BOND action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         # Set the new atom types, removing any duplicates
 *         self.atomType = list(set(atomType))             # <<<<<<<<<<<<<<
 * 
 *     def __gainRadical(self, radical):
 */
  __pyx_t_1 = PySet_New(__pyx_v_atomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 195; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 195; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyList_Type))), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 195; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":181
 *         self.atomType = list(set(atomType))
 * 
 *     def __breakBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a BREAK_BOND action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_8);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__breakBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond[] = "\n        Update the atom group as a result of applying a BREAK_BOND action,\n        where `order` specifies the order of the breaking bond, and should be\n        'S' (since we only allow breaking of single bonds).\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond(PyObject *__pyx_v_self, PyObject *__pyx_v_order) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__breakBond (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_order), (&PyString_Type), 1, "order", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 181; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject*)__pyx_v_order));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_order) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__breakBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___breakBond(__pyx_v_self, __pyx_v_order, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 181; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__breakBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":197
 *         self.atomType = list(set(atomType))
 * 
 *     def __gainRadical(self, radical):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a GAIN_RADICAL action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_radicalElectrons = NULL;
  PyObject *__pyx_v_electron = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  int __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__gainRadical", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_gainRadical); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 197; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_radical); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 197; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 197; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 197; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 197; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":202
 *         where `radical` specifies the number of radical electrons to add.
 *         """
 *         radicalElectrons = []             # <<<<<<<<<<<<<<
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 202; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_radicalElectrons = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":203
 *         """
 *         radicalElectrons = []
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_4); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomType, __pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomType, __pyx_n_s_incrementRadical); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = PyObject_Length(__pyx_t_4); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyBool_FromLong((__pyx_t_8 == 0)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_4))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":204
 *         radicalElectrons = []
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         for electron in self.radicalElectrons:
 *             radicalElectrons.append(electron + radical)
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_7, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_6, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":205
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             radicalElectrons.append(electron + radical)
 *         # Set the new radical electron counts
 */
  if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 205; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 205; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 205; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_electron, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":206
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:
 *             radicalElectrons.append(electron + radical)             # <<<<<<<<<<<<<<
 *         # Set the new radical electron counts
 *         self.radicalElectrons = radicalElectrons
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_radical); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Add(__pyx_v_electron, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_10 = __Pyx_PyList_Append(__pyx_v_radicalElectrons, __pyx_t_5); if (unlikely(__pyx_t_10 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 206; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;

    /* "rmgpy/molecule/group.py":205
 *         if any([len(atomType.incrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             radicalElectrons.append(electron + radical)
 *         # Set the new radical electron counts
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":208
 *             radicalElectrons.append(electron + radical)
 *         # Set the new radical electron counts
 *         self.radicalElectrons = radicalElectrons             # <<<<<<<<<<<<<<
 * 
 *     def __loseRadical(self, radical):
 */
  __Pyx_INCREF(__pyx_v_radicalElectrons);
  __Pyx_GIVEREF(__pyx_v_radicalElectrons);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = __pyx_v_radicalElectrons;

  /* "rmgpy/molecule/group.py":197
 *         self.atomType = list(set(atomType))
 * 
 *     def __gainRadical(self, radical):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a GAIN_RADICAL action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_radicalElectrons);
  __Pyx_XDECREF(__pyx_v_electron);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical[] = "\n        Update the atom group as a result of applying a GAIN_RADICAL action,\n        where `radical` specifies the number of radical electrons to add.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical) {
  short __pyx_v_radical;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__gainRadical (wrapper)", 0);
  assert(__pyx_arg_radical); {
    __pyx_v_radical = __Pyx_PyInt_As_short(__pyx_arg_radical); if (unlikely((__pyx_v_radical == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 197; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_radical));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__gainRadical", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainRadical(__pyx_v_self, __pyx_v_radical, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 197; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":210
 *         self.radicalElectrons = radicalElectrons
 * 
 *     def __loseRadical(self, radical):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a LOSE_RADICAL action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___loseRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_radicalElectrons = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_pairs = NULL;
  PyObject *__pyx_v_electron = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  PyObject *__pyx_t_10 = NULL;
  int __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__loseRadical", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_loseRadical); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_radical); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":215
 *         where `radical` specifies the number of radical electrons to remove.
 *         """
 *         radicalElectrons = []             # <<<<<<<<<<<<<<
 *         pairs = set()
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 215; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_radicalElectrons = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":216
 *         """
 *         radicalElectrons = []
 *         pairs = set()             # <<<<<<<<<<<<<<
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
  __pyx_t_1 = PySet_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 216; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_pairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":217
 *         radicalElectrons = []
 *         pairs = set()
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_4); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomType, __pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomType, __pyx_n_s_decrementRadical); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = PyObject_Length(__pyx_t_4); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyBool_FromLong((__pyx_t_8 == 0)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_4))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 217; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":218
 *         pairs = set()
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         for electron in self.radicalElectrons:
 *             electron = electron - radical
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_8, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_6, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 218; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":219
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             electron = electron - radical
 *             if electron < 0:
 */
  if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 219; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 219; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 219; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_electron, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":220
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:
 *             electron = electron - radical             # <<<<<<<<<<<<<<
 *             if electron < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set "{0}".'.format(self.radicalElectrons))
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_radical); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 220; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Subtract(__pyx_v_electron, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 220; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF_SET(__pyx_v_electron, __pyx_t_5);
    __pyx_t_5 = 0;

    /* "rmgpy/molecule/group.py":221
 *         for electron in self.radicalElectrons:
 *             electron = electron - radical
 *             if electron < 0:             # <<<<<<<<<<<<<<
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set "{0}".'.format(self.radicalElectrons))
 *             radicalElectrons.append(electron)
 */
    __pyx_t_5 = PyObject_RichCompare(__pyx_v_electron, __pyx_int_0, Py_LT); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 221; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 221; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    if (__pyx_t_9) {

      /* "rmgpy/molecule/group.py":222
 *             electron = electron - radical
 *             if electron < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set "{0}".'.format(self.radicalElectrons))             # <<<<<<<<<<<<<<
 *             radicalElectrons.append(electron)
 * 
 */
      __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_9, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __pyx_t_3 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
        __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
        if (likely(__pyx_t_3)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
          __Pyx_INCREF(__pyx_t_3);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_6, function);
        }
      }
      if (!__pyx_t_3) {
        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->radicalElectrons); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
      } else {
        __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        __Pyx_INCREF(__pyx_v_self->radicalElectrons);
        PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_v_self->radicalElectrons);
        __Pyx_GIVEREF(__pyx_v_self->radicalElectrons);
        __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_10, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      }
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __pyx_t_6 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
        __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
        if (likely(__pyx_t_6)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
          __Pyx_INCREF(__pyx_t_6);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_2, function);
        }
      }
      if (!__pyx_t_6) {
        __pyx_t_5 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_GOTREF(__pyx_t_5);
      } else {
        __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
        PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_t_4);
        __Pyx_GIVEREF(__pyx_t_4);
        __pyx_t_4 = 0;
        __pyx_t_5 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_10, NULL); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      }
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_Raise(__pyx_t_5, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 222; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "rmgpy/molecule/group.py":223
 *             if electron < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Invalid radical electron set "{0}".'.format(self.radicalElectrons))
 *             radicalElectrons.append(electron)             # <<<<<<<<<<<<<<
 * 
 *         # Set the new radical electron counts
 */
    __pyx_t_11 = __Pyx_PyList_Append(__pyx_v_radicalElectrons, __pyx_v_electron); if (unlikely(__pyx_t_11 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

    /* "rmgpy/molecule/group.py":219
 *         if any([len(atomType.decrementRadical) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_RADICAL action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for electron in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             electron = electron - radical
 *             if electron < 0:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":226
 * 
 *         # Set the new radical electron counts
 *         self.radicalElectrons = radicalElectrons             # <<<<<<<<<<<<<<
 * 
 *     def __gainPair(self, pair):
 */
  __Pyx_INCREF(__pyx_v_radicalElectrons);
  __Pyx_GIVEREF(__pyx_v_radicalElectrons);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = __pyx_v_radicalElectrons;

  /* "rmgpy/molecule/group.py":210
 *         self.radicalElectrons = radicalElectrons
 * 
 *     def __loseRadical(self, radical):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a LOSE_RADICAL action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_10);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__loseRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_radicalElectrons);
  __Pyx_XDECREF(__pyx_v_pairs);
  __Pyx_XDECREF(__pyx_v_electron);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical[] = "\n        Update the atom group as a result of applying a LOSE_RADICAL action,\n        where `radical` specifies the number of radical electrons to remove.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical(PyObject *__pyx_v_self, PyObject *__pyx_arg_radical) {
  short __pyx_v_radical;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__loseRadical (wrapper)", 0);
  assert(__pyx_arg_radical); {
    __pyx_v_radical = __Pyx_PyInt_As_short(__pyx_arg_radical); if (unlikely((__pyx_v_radical == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__loseRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_radical));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_radical) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__loseRadical", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___loseRadical(__pyx_v_self, __pyx_v_radical, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__loseRadical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":228
 *         self.radicalElectrons = radicalElectrons
 * 
 *     def __gainPair(self, pair):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a GAIN_PAIR action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainPair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_lonePairs = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  PyObject *(*__pyx_t_10)(PyObject *);
  int __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__gainPair", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_gainPair); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":233
 *         where `pair` specifies the number of lone electron pairs to add.
 *         """
 *         lonePairs = []             # <<<<<<<<<<<<<<
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 233; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_lonePairs = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":234
 *         """
 *         lonePairs = []
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_4); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomType, __pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomType, __pyx_n_s_incrementLonePair); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = PyObject_Length(__pyx_t_4); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyBool_FromLong((__pyx_t_8 == 0)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_4))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":235
 *         lonePairs = []
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         for lonePairs in zip(self.lonePairs):
 *             lonePairs.append(lonePairs + pair)
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_10, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_6, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":236
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):             # <<<<<<<<<<<<<<
 *             lonePairs.append(lonePairs + pair)
 *         # Set the new lone electron pair count
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_self->lonePairs);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_self->lonePairs);
  __Pyx_GIVEREF(__pyx_v_self->lonePairs);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
    __pyx_t_10 = NULL;
  } else {
    __pyx_t_7 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_10 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_10)) {
      if (likely(PyList_CheckExact(__pyx_t_1))) {
        if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_7 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_10(__pyx_t_1);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_DECREF_SET(__pyx_v_lonePairs, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":237
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):
 *             lonePairs.append(lonePairs + pair)             # <<<<<<<<<<<<<<
 *         # Set the new lone electron pair count
 *         self.lonePairs = lonePairs
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Add(__pyx_v_lonePairs, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_11 = __Pyx_PyObject_Append(__pyx_v_lonePairs, __pyx_t_5); if (unlikely(__pyx_t_11 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 237; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;

    /* "rmgpy/molecule/group.py":236
 *         if any([len(atomType.incrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to GAIN_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):             # <<<<<<<<<<<<<<
 *             lonePairs.append(lonePairs + pair)
 *         # Set the new lone electron pair count
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":239
 *             lonePairs.append(lonePairs + pair)
 *         # Set the new lone electron pair count
 *         self.lonePairs = lonePairs             # <<<<<<<<<<<<<<
 * 
 *     def __losePair(self, pair):
 */
  if (!(likely(PyList_CheckExact(__pyx_v_lonePairs))||((__pyx_v_lonePairs) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_lonePairs)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 239; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_lonePairs;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":228
 *         self.radicalElectrons = radicalElectrons
 * 
 *     def __gainPair(self, pair):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a GAIN_PAIR action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainPair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_lonePairs);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair[] = "\n        Update the atom group as a result of applying a GAIN_PAIR action,\n        where `pair` specifies the number of lone electron pairs to add.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair) {
  short __pyx_v_pair;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__gainPair (wrapper)", 0);
  assert(__pyx_arg_pair); {
    __pyx_v_pair = __Pyx_PyInt_As_short(__pyx_arg_pair); if (unlikely((__pyx_v_pair == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 228; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainPair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_pair));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__gainPair", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainPair(__pyx_v_self, __pyx_v_pair, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__gainPair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":241
 *         self.lonePairs = lonePairs
 * 
 *     def __losePair(self, pair):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a LOSE_PAIR action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___losePair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_lonePairs = NULL;
  PyObject *__pyx_v_atomType = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  PyObject *(*__pyx_t_10)(PyObject *);
  PyObject *__pyx_t_11 = NULL;
  int __pyx_t_12;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__losePair", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_losePair); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":246
 *         where `pair` specifies the number of lone electron pairs to remove.
 *         """
 *         lonePairs = []             # <<<<<<<<<<<<<<
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_lonePairs = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":247
 *         """
 *         lonePairs = []
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):             # <<<<<<<<<<<<<<
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_4); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomType, __pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomType, __pyx_n_s_decrementLonePair); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = PyObject_Length(__pyx_t_4); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyBool_FromLong((__pyx_t_8 == 0)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_4))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":248
 *         lonePairs = []
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))             # <<<<<<<<<<<<<<
 *         for lonePairs in zip(self.lonePairs):
 *             if lonePairs - pair < 0:
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_11, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_6, function);
      }
    }
    if (!__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
      __Pyx_INCREF(__pyx_v_self->atomType);
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_self->atomType);
      __Pyx_GIVEREF(__pyx_v_self->atomType);
      __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":249
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):             # <<<<<<<<<<<<<<
 *             if lonePairs - pair < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_self->lonePairs);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_self->lonePairs);
  __Pyx_GIVEREF(__pyx_v_self->lonePairs);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
    __pyx_t_10 = NULL;
  } else {
    __pyx_t_7 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_10 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_10)) {
      if (likely(PyList_CheckExact(__pyx_t_1))) {
        if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_7 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_10(__pyx_t_1);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_DECREF_SET(__pyx_v_lonePairs, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":250
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):
 *             if lonePairs - pair < 0:             # <<<<<<<<<<<<<<
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))
 *             lonePairs.append(lonePairs - pair)
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 250; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Subtract(__pyx_v_lonePairs, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 250; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyObject_RichCompare(__pyx_t_5, __pyx_int_0, Py_LT); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 250; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 250; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_9) {

      /* "rmgpy/molecule/group.py":251
 *         for lonePairs in zip(self.lonePairs):
 *             if lonePairs - pair < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))             # <<<<<<<<<<<<<<
 *             lonePairs.append(lonePairs - pair)
 *         # Set the new lone electron pair count
 */
      __pyx_t_5 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_due_t_12, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __pyx_t_3 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
        __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
        if (likely(__pyx_t_3)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
          __Pyx_INCREF(__pyx_t_3);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_6, function);
        }
      }
      if (!__pyx_t_3) {
        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_6, __pyx_v_self->lonePairs); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
      } else {
        __pyx_t_11 = PyTuple_New(1+1); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        __Pyx_INCREF(__pyx_v_self->lonePairs);
        PyTuple_SET_ITEM(__pyx_t_11, 0+1, __pyx_v_self->lonePairs);
        __Pyx_GIVEREF(__pyx_v_self->lonePairs);
        __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_11, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      }
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __pyx_t_6 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_5))) {
        __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_5);
        if (likely(__pyx_t_6)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
          __Pyx_INCREF(__pyx_t_6);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_5, function);
        }
      }
      if (!__pyx_t_6) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_11 = PyTuple_New(1+1); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
        PyTuple_SET_ITEM(__pyx_t_11, 0+1, __pyx_t_4);
        __Pyx_GIVEREF(__pyx_t_4);
        __pyx_t_4 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_11, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      }
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "rmgpy/molecule/group.py":252
 *             if lonePairs - pair < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))
 *             lonePairs.append(lonePairs - pair)             # <<<<<<<<<<<<<<
 *         # Set the new lone electron pair count
 *         self.lonePairs = lonePairs
 */
    __pyx_t_2 = __Pyx_PyInt_From_short(__pyx_v_pair); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = PyNumber_Subtract(__pyx_v_lonePairs, __pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_12 = __Pyx_PyObject_Append(__pyx_v_lonePairs, __pyx_t_5); if (unlikely(__pyx_t_12 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;

    /* "rmgpy/molecule/group.py":249
 *         if any([len(atomType.decrementLonePair) == 0 for atomType in self.atomType]):
 *             raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Unknown atom type produced from set "{0}".'.format(self.atomType))
 *         for lonePairs in zip(self.lonePairs):             # <<<<<<<<<<<<<<
 *             if lonePairs - pair < 0:
 *                 raise ActionError('Unable to update GroupAtom due to LOSE_PAIR action: Invalid lone electron pairs set "{0}".'.format(self.lonePairs))
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":254
 *             lonePairs.append(lonePairs - pair)
 *         # Set the new lone electron pair count
 *         self.lonePairs = lonePairs             # <<<<<<<<<<<<<<
 * 
 *     def applyAction(self, action):
 */
  if (!(likely(PyList_CheckExact(__pyx_v_lonePairs))||((__pyx_v_lonePairs) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_lonePairs)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 254; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_lonePairs;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":241
 *         self.lonePairs = lonePairs
 * 
 *     def __losePair(self, pair):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying a LOSE_PAIR action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_11);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__losePair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_lonePairs);
  __Pyx_XDECREF(__pyx_v_atomType);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_26__losePair[] = "\n        Update the atom group as a result of applying a LOSE_PAIR action,\n        where `pair` specifies the number of lone electron pairs to remove.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair(PyObject *__pyx_v_self, PyObject *__pyx_arg_pair) {
  short __pyx_v_pair;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__losePair (wrapper)", 0);
  assert(__pyx_arg_pair); {
    __pyx_v_pair = __Pyx_PyInt_As_short(__pyx_arg_pair); if (unlikely((__pyx_v_pair == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__losePair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_26__losePair(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((short)__pyx_v_pair));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_26__losePair(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, short __pyx_v_pair) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__losePair", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom___losePair(__pyx_v_self, __pyx_v_pair, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.__losePair", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":256
 *         self.lonePairs = lonePairs
 * 
 *     def applyAction(self, action):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying `action`, a tuple
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_action, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_act = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  short __pyx_t_7;
  PyObject *__pyx_t_8 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("applyAction", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_applyAction); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_action); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_action);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_action);
        __Pyx_GIVEREF(__pyx_v_action);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":263
 *         :ref:`here <reaction-recipe-actions>`.
 *         """
 *         act = action[0].upper()             # <<<<<<<<<<<<<<
 *         if act == 'CHANGE_BOND':
 *             self.__changeBond(action[2])
 */
  if (unlikely(__pyx_v_action == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_action, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_upper); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (__pyx_t_2) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 263; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_v_act = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":264
 *         """
 *         act = action[0].upper()
 *         if act == 'CHANGE_BOND':             # <<<<<<<<<<<<<<
 *             self.__changeBond(action[2])
 *         elif act == 'FORM_BOND':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_CHANGE_BOND, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 264; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":265
 *         act = action[0].upper()
 *         if act == 'CHANGE_BOND':
 *             self.__changeBond(action[2])             # <<<<<<<<<<<<<<
 *         elif act == 'FORM_BOND':
 *             self.__formBond(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___changeBond(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":266
 *         if act == 'CHANGE_BOND':
 *             self.__changeBond(action[2])
 *         elif act == 'FORM_BOND':             # <<<<<<<<<<<<<<
 *             self.__formBond(action[2])
 *         elif act == 'BREAK_BOND':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_FORM_BOND, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 266; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":267
 *             self.__changeBond(action[2])
 *         elif act == 'FORM_BOND':
 *             self.__formBond(action[2])             # <<<<<<<<<<<<<<
 *         elif act == 'BREAK_BOND':
 *             self.__breakBond(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 267; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 267; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    if (!(likely(PyString_CheckExact(__pyx_t_1))||((__pyx_t_1) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_1)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 267; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_3 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___formBond(__pyx_v_self, ((PyObject*)__pyx_t_1), 0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 267; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":268
 *         elif act == 'FORM_BOND':
 *             self.__formBond(action[2])
 *         elif act == 'BREAK_BOND':             # <<<<<<<<<<<<<<
 *             self.__breakBond(action[2])
 *         elif act == 'GAIN_RADICAL':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_BREAK_BOND, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 268; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":269
 *             self.__formBond(action[2])
 *         elif act == 'BREAK_BOND':
 *             self.__breakBond(action[2])             # <<<<<<<<<<<<<<
 *         elif act == 'GAIN_RADICAL':
 *             self.__gainRadical(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 269; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_3 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 269; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    if (!(likely(PyString_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 269; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___breakBond(__pyx_v_self, ((PyObject*)__pyx_t_3), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 269; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":270
 *         elif act == 'BREAK_BOND':
 *             self.__breakBond(action[2])
 *         elif act == 'GAIN_RADICAL':             # <<<<<<<<<<<<<<
 *             self.__gainRadical(action[2])
 *         elif act == 'LOSE_RADICAL':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_GAIN_RADICAL, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 270; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":271
 *             self.__breakBond(action[2])
 *         elif act == 'GAIN_RADICAL':
 *             self.__gainRadical(action[2])             # <<<<<<<<<<<<<<
 *         elif act == 'LOSE_RADICAL':
 *             self.__loseRadical(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 271; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 271; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 271; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___gainRadical(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 271; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":272
 *         elif act == 'GAIN_RADICAL':
 *             self.__gainRadical(action[2])
 *         elif act == 'LOSE_RADICAL':             # <<<<<<<<<<<<<<
 *             self.__loseRadical(action[2])
 *         elif action[0].upper() == 'GAIN_PAIR':
 */
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_act, __pyx_n_s_LOSE_RADICAL, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 272; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":273
 *             self.__gainRadical(action[2])
 *         elif act == 'LOSE_RADICAL':
 *             self.__loseRadical(action[2])             # <<<<<<<<<<<<<<
 *         elif action[0].upper() == 'GAIN_PAIR':
 *             self.__gainPair(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 273; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 273; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 273; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___loseRadical(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 273; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":274
 *         elif act == 'LOSE_RADICAL':
 *             self.__loseRadical(action[2])
 *         elif action[0].upper() == 'GAIN_PAIR':             # <<<<<<<<<<<<<<
 *             self.__gainPair(action[2])
 *         elif action[0].upper() == 'LOSE_PAIR':
 */
  if (unlikely(__pyx_v_action == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 274; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_3 = __Pyx_GetItemInt_List(__pyx_v_action, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 274; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_upper); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 274; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 274; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 274; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_GAIN_PAIR, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 274; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":275
 *             self.__loseRadical(action[2])
 *         elif action[0].upper() == 'GAIN_PAIR':
 *             self.__gainPair(action[2])             # <<<<<<<<<<<<<<
 *         elif action[0].upper() == 'LOSE_PAIR':
 *             self.__losePair(action[2])
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 275; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 275; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 275; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___gainPair(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 275; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }

  /* "rmgpy/molecule/group.py":276
 *         elif action[0].upper() == 'GAIN_PAIR':
 *             self.__gainPair(action[2])
 *         elif action[0].upper() == 'LOSE_PAIR':             # <<<<<<<<<<<<<<
 *             self.__losePair(action[2])
 *         else:
 */
  if (unlikely(__pyx_v_action == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_action, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_upper); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (__pyx_t_2) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_LOSE_PAIR, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":277
 *             self.__gainPair(action[2])
 *         elif action[0].upper() == 'LOSE_PAIR':
 *             self.__losePair(action[2])             # <<<<<<<<<<<<<<
 *         else:
 *             raise ActionError('Unable to update GroupAtom: Invalid action {0}".'.format(action))
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 277; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 277; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 277; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___losePair(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 277; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":279
 *             self.__losePair(action[2])
 *         else:
 *             raise ActionError('Unable to update GroupAtom: Invalid action {0}".'.format(action))             # <<<<<<<<<<<<<<
 * 
 *     def equivalent(self, other):
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupAtom_Inval, __pyx_n_s_format); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_5);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_5, function);
      }
    }
    if (!__pyx_t_4) {
      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_v_action); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      __Pyx_INCREF(__pyx_v_action);
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_v_action);
      __Pyx_GIVEREF(__pyx_v_action);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_t_2);
      __Pyx_GIVEREF(__pyx_t_2);
      __pyx_t_2 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_8, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 279; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_L3:;

  /* "rmgpy/molecule/group.py":256
 *         self.lonePairs = lonePairs
 * 
 *     def applyAction(self, action):             # <<<<<<<<<<<<<<
 *         """
 *         Update the atom group as a result of applying `action`, a tuple
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_8);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.applyAction", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_act);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_28applyAction[] = "\n        Update the atom group as a result of applying `action`, a tuple\n        containing the name of the reaction recipe action along with any\n        required parameters. The available actions can be found\n        :ref:`here <reaction-recipe-actions>`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("applyAction (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_action), (&PyList_Type), 1, "action", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_28applyAction(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject*)__pyx_v_action));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_28applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_action) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("applyAction", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_applyAction(__pyx_v_self, __pyx_v_action, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.applyAction", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":281
 *             raise ActionError('Unable to update GroupAtom: Invalid action {0}".'.format(action))
 * 
 *     def equivalent(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_group = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_atomType1 = 0;
  short __pyx_v_radical1;
  short __pyx_v_radical2;
  short __pyx_v_lp1;
  short __pyx_v_lp2;
  short __pyx_v_charge1;
  short __pyx_v_charge2;
  PyObject *__pyx_v_atomType2 = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  short __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("equivalent", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_equivalent); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 281; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 281; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 281; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 281; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 281; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":290
 *         """
 *         cython.declare(group=GroupAtom)
 *         if not isinstance(other, GroupAtom):             # <<<<<<<<<<<<<<
 *             # Let the equivalent method of other handle it
 *             # We expect self to be an Atom object, but can't test for it here
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":294
 *             # We expect self to be an Atom object, but can't test for it here
 *             # because that would create an import cycle
 *             return other.equivalent(self)             # <<<<<<<<<<<<<<
 *         group=other
 * 
 */
    __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_other->__pyx_vtab)->equivalent(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_self), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 294; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_r = __pyx_t_7;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":295
 *             # because that would create an import cycle
 *             return other.equivalent(self)
 *         group=other             # <<<<<<<<<<<<<<
 * 
 *         cython.declare(atomType1=AtomType, atomtype2=AtomType, radical1=cython.short, radical2=cython.short,
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 295; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_group = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":301
 *         # Compare two atom groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:             # <<<<<<<<<<<<<<
 *             for atomType2 in group.atomType:
 *                 if atomType1.equivalent(atomType2): break
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 301; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 301; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 301; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 301; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":302
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:
 *             for atomType2 in group.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    if (unlikely(__pyx_v_group->atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_group->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_atomType2, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":303
 *         for atomType1 in self.atomType:
 *             for atomType2 in group.atomType:
 *                 if atomType1.equivalent(atomType2): break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      if (!(likely(((__pyx_v_atomType2) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_atomType2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 303; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = (((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType1->__pyx_vtab)->equivalent(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType2), 0) != 0);
      if (__pyx_t_7) {
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":302
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:
 *             for atomType2 in group.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":305
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":302
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:
 *             for atomType2 in group.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":301
 *         # Compare two atom groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType:             # <<<<<<<<<<<<<<
 *             for atomType2 in group.atomType:
 *                 if atomType1.equivalent(atomType2): break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":306
 *             else:
 *                 return False
 *         for atomType1 in group.atomType:             # <<<<<<<<<<<<<<
 *             for atomType2 in self.atomType:
 *                 if atomType1.equivalent(atomType2): break
 */
  if (unlikely(__pyx_v_group->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 306; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_group->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 306; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 306; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 306; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":307
 *                 return False
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    if (unlikely(__pyx_v_self->atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 307; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 307; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 307; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_atomType2, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":308
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:
 *                 if atomType1.equivalent(atomType2): break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      if (!(likely(((__pyx_v_atomType2) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_atomType2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 308; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = (((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType1->__pyx_vtab)->equivalent(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType2), 0) != 0);
      if (__pyx_t_7) {
        goto __pyx_L13_break;
      }

      /* "rmgpy/molecule/group.py":307
 *                 return False
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":310
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         for radical1 in self.radicalElectrons:
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":307
 *                 return False
 *         for atomType1 in group.atomType:
 *             for atomType2 in self.atomType:             # <<<<<<<<<<<<<<
 *                 if atomType1.equivalent(atomType2): break
 *             else:
 */
    __pyx_L13_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":306
 *             else:
 *                 return False
 *         for atomType1 in group.atomType:             # <<<<<<<<<<<<<<
 *             for atomType2 in self.atomType:
 *                 if atomType1.equivalent(atomType2): break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":312
 *                 return False
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         for radical1 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:
 */
  if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 312; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 312; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 312; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 312; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_radical1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":313
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         for radical1 in self.radicalElectrons:
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.             # <<<<<<<<<<<<<<
 *                 for radical2  in group.radicalElectrons:
 *                     if radical1 == radical2: break
 */
    __pyx_t_7 = (__pyx_v_group->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_group->radicalElectrons) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":314
 *         for radical1 in self.radicalElectrons:
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      if (unlikely(__pyx_v_group->radicalElectrons == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 314; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_group->radicalElectrons; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 314; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 314; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 314; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_radical2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":315
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:
 *                     if radical1 == radical2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_radical1 == __pyx_v_radical2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L20_break;
        }

        /* "rmgpy/molecule/group.py":314
 *         for radical1 in self.radicalElectrons:
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":317
 *                     if radical1 == radical2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":314
 *         for radical1 in self.radicalElectrons:
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      __pyx_L20_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L18;
    }
    __pyx_L18:;

    /* "rmgpy/molecule/group.py":312
 *                 return False
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         for radical1 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             if group.radicalElectrons:  # Only check if the list is non-empty.  An empty list indicates a wildcard.
 *                 for radical2  in group.radicalElectrons:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":318
 *                 else:
 *                     return False
 *         for radical1 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:
 */
  if (unlikely(__pyx_v_group->radicalElectrons == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 318; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_group->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 318; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 318; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 318; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_radical1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":319
 *                     return False
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                 for radical2 in self.radicalElectrons:
 *                     if radical1 == radical2: break
 */
    __pyx_t_7 = (__pyx_v_self->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_self->radicalElectrons) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":320
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_radical2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":321
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:
 *                     if radical1 == radical2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_radical1 == __pyx_v_radical2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L27_break;
        }

        /* "rmgpy/molecule/group.py":320
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":323
 *                     if radical1 == radical2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":320
 *         for radical1 in group.radicalElectrons:
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     if radical1 == radical2: break
 *                 else:
 */
      __pyx_L27_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L25;
    }
    __pyx_L25:;

    /* "rmgpy/molecule/group.py":318
 *                 else:
 *                     return False
 *         for radical1 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *             if self.radicalElectrons:
 *                 for radical2 in self.radicalElectrons:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":324
 *                 else:
 *                     return False
 *         for lp1 in self.lonePairs:             # <<<<<<<<<<<<<<
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:
 */
  if (unlikely(__pyx_v_self->lonePairs == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 324; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->lonePairs; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 324; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 324; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 324; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_lp1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":325
 *                     return False
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:             # <<<<<<<<<<<<<<
 *                 for lp2 in group.lonePairs:
 *                     if lp1 == lp2: break
 */
    __pyx_t_7 = (__pyx_v_group->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_group->lonePairs) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":326
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                     if lp1 == lp2: break
 *                 else:
 */
      if (unlikely(__pyx_v_group->lonePairs == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 326; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_group->lonePairs; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 326; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 326; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 326; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_lp2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":327
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:
 *                     if lp1 == lp2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_lp1 == __pyx_v_lp2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L34_break;
        }

        /* "rmgpy/molecule/group.py":326
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                     if lp1 == lp2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":329
 *                     if lp1 == lp2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         #Each charge in self must have an equivalent in other (and vice versa)
 *         for charge1 in self.charge:
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":326
 *         for lp1 in self.lonePairs:
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                     if lp1 == lp2: break
 *                 else:
 */
      __pyx_L34_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L32;
    }
    __pyx_L32:;

    /* "rmgpy/molecule/group.py":324
 *                 else:
 *                     return False
 *         for lp1 in self.lonePairs:             # <<<<<<<<<<<<<<
 *             if group.lonePairs:
 *                 for lp2 in group.lonePairs:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":331
 *                     return False
 *         #Each charge in self must have an equivalent in other (and vice versa)
 *         for charge1 in self.charge:             # <<<<<<<<<<<<<<
 *             if group.charge:
 *                 for charge2 in group.charge:
 */
  if (unlikely(__pyx_v_self->charge == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 331; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->charge; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 331; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 331; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 331; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_charge1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":332
 *         #Each charge in self must have an equivalent in other (and vice versa)
 *         for charge1 in self.charge:
 *             if group.charge:             # <<<<<<<<<<<<<<
 *                 for charge2 in group.charge:
 *                     if charge1 == charge2: break
 */
    __pyx_t_7 = (__pyx_v_group->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_group->charge) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":333
 *         for charge1 in self.charge:
 *             if group.charge:
 *                 for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      if (unlikely(__pyx_v_group->charge == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 333; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_group->charge; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 333; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 333; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 333; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_charge2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":334
 *             if group.charge:
 *                 for charge2 in group.charge:
 *                     if charge1 == charge2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_charge1 == __pyx_v_charge2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L41_break;
        }

        /* "rmgpy/molecule/group.py":333
 *         for charge1 in self.charge:
 *             if group.charge:
 *                 for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":336
 *                     if charge1 == charge2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         for charge1 in group.charge:
 *             if self.charge:
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":333
 *         for charge1 in self.charge:
 *             if group.charge:
 *                 for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      __pyx_L41_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L39;
    }
    __pyx_L39:;

    /* "rmgpy/molecule/group.py":331
 *                     return False
 *         #Each charge in self must have an equivalent in other (and vice versa)
 *         for charge1 in self.charge:             # <<<<<<<<<<<<<<
 *             if group.charge:
 *                 for charge2 in group.charge:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":337
 *                 else:
 *                     return False
 *         for charge1 in group.charge:             # <<<<<<<<<<<<<<
 *             if self.charge:
 *                 for charge2 in self.charge:
 */
  if (unlikely(__pyx_v_group->charge == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_group->charge; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_v_charge1 = __pyx_t_10;

    /* "rmgpy/molecule/group.py":338
 *                     return False
 *         for charge1 in group.charge:
 *             if self.charge:             # <<<<<<<<<<<<<<
 *                 for charge2 in self.charge:
 *                     if charge1 == charge2: break
 */
    __pyx_t_7 = (__pyx_v_self->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_self->charge) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":339
 *         for charge1 in group.charge:
 *             if self.charge:
 *                 for charge2 in self.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      if (unlikely(__pyx_v_self->charge == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 339; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __pyx_v_self->charge; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
      for (;;) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 339; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 339; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 339; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_v_charge2 = __pyx_t_10;

        /* "rmgpy/molecule/group.py":340
 *             if self.charge:
 *                 for charge2 in self.charge:
 *                     if charge1 == charge2: break             # <<<<<<<<<<<<<<
 *                 else:
 *                     return False
 */
        __pyx_t_7 = ((__pyx_v_charge1 == __pyx_v_charge2) != 0);
        if (__pyx_t_7) {
          goto __pyx_L48_break;
        }

        /* "rmgpy/molecule/group.py":339
 *         for charge1 in group.charge:
 *             if self.charge:
 *                 for charge2 in self.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":342
 *                     if charge1 == charge2: break
 *                 else:
 *                     return False             # <<<<<<<<<<<<<<
 *         # Otherwise the two atom groups are equivalent
 *         return True
 */
        __pyx_r = 0;
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L0;
      }

      /* "rmgpy/molecule/group.py":339
 *         for charge1 in group.charge:
 *             if self.charge:
 *                 for charge2 in self.charge:             # <<<<<<<<<<<<<<
 *                     if charge1 == charge2: break
 *                 else:
 */
      __pyx_L48_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L46;
    }
    __pyx_L46:;

    /* "rmgpy/molecule/group.py":337
 *                 else:
 *                     return False
 *         for charge1 in group.charge:             # <<<<<<<<<<<<<<
 *             if self.charge:
 *                 for charge2 in self.charge:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":344
 *                     return False
 *         # Otherwise the two atom groups are equivalent
 *         return True             # <<<<<<<<<<<<<<
 * 
 *     def isSpecificCaseOf(self, other):
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":281
 *             raise ActionError('Unable to update GroupAtom: Invalid action {0}".'.format(action))
 * 
 *     def equivalent(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.equivalent", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_group);
  __Pyx_XDECREF((PyObject *)__pyx_v_atomType1);
  __Pyx_XDECREF(__pyx_v_atomType2);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_30equivalent[] = "\n        Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,\n        where `other` can be either an :class:`Atom` or an :class:`GroupAtom`\n        object. When comparing two :class:`GroupAtom` objects, this function\n        respects wildcards, e.g. ``R!H`` is equivalent to ``C``.\n        \n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("equivalent (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Vertex, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 281; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_30equivalent(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_30equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("equivalent", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_equivalent(__pyx_v_self, __pyx_v_other, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 281; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 281; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.equivalent", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":346
 *         return True
 * 
 *     def isSpecificCaseOf(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is the same as `self` or is a more
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_group = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_atomType1 = 0;
  short __pyx_v_radical1;
  short __pyx_v_radical2;
  short __pyx_v_lp1;
  short __pyx_v_lp2;
  short __pyx_v_charge1;
  short __pyx_v_charge2;
  PyObject *__pyx_v_atomType2 = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  short __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSpecificCaseOf", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSpecificCaseOf); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":353
 *         """
 *         cython.declare(group=GroupAtom)
 *         if not isinstance(other, GroupAtom):             # <<<<<<<<<<<<<<
 *             # Let the isSpecificCaseOf method of other handle it
 *             # We expect self to be an Atom object, but can't test for it here
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":357
 *             # We expect self to be an Atom object, but can't test for it here
 *             # because that would create an import cycle
 *             return other.isSpecificCaseOf(self)             # <<<<<<<<<<<<<<
 *         group=other
 * 
 */
    __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_other->__pyx_vtab)->isSpecificCaseOf(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_self), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 357; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_r = __pyx_t_7;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":358
 *             # because that would create an import cycle
 *             return other.isSpecificCaseOf(self)
 *         group=other             # <<<<<<<<<<<<<<
 * 
 *         cython.declare(atomType1=AtomType, atomtype2=AtomType, radical1=cython.short, radical2=cython.short,
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 358; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_group = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":364
 *         # Compare two atom groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match             # <<<<<<<<<<<<<<
 *             for atomType2 in group.atomType: # can match any of these
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 364; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 364; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 364; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 364; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":365
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match
 *             for atomType2 in group.atomType: # can match any of these             # <<<<<<<<<<<<<<
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 *             else:
 */
    if (unlikely(__pyx_v_group->atomType == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 365; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_group->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 365; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 365; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_atomType2, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":366
 *         for atomType1 in self.atomType: # all these must match
 *             for atomType2 in group.atomType: # can match any of these
 *                 if atomType1.isSpecificCaseOf(atomType2): break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      if (!(likely(((__pyx_v_atomType2) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_atomType2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 366; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = (((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType1->__pyx_vtab)->isSpecificCaseOf(__pyx_v_atomType1, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType2), 0) != 0);
      if (__pyx_t_7) {
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":365
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match
 *             for atomType2 in group.atomType: # can match any of these             # <<<<<<<<<<<<<<
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":368
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         if self.radicalElectrons:
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":365
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match
 *             for atomType2 in group.atomType: # can match any of these             # <<<<<<<<<<<<<<
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 *             else:
 */
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":364
 *         # Compare two atom groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for atomType1 in self.atomType: # all these must match             # <<<<<<<<<<<<<<
 *             for atomType2 in group.atomType: # can match any of these
 *                 if atomType1.isSpecificCaseOf(atomType2): break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":370
 *                 return False
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         if self.radicalElectrons:             # <<<<<<<<<<<<<<
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:
 */
  __pyx_t_7 = (__pyx_v_self->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_self->radicalElectrons) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":371
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         if self.radicalElectrons:
 *             for radical1 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:
 */
    if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->radicalElectrons; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
    for (;;) {
      if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_radical1 = __pyx_t_10;

      /* "rmgpy/molecule/group.py":372
 *         if self.radicalElectrons:
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                     for radical2 in group.radicalElectrons:
 *                         if radical1 == radical2: break
 */
      __pyx_t_7 = (__pyx_v_group->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_group->radicalElectrons) != 0);
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":373
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                         if radical1 == radical2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->radicalElectrons == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 373; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->radicalElectrons; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
        for (;;) {
          if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 373; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 373; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 373; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_v_radical2 = __pyx_t_10;

          /* "rmgpy/molecule/group.py":374
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:
 *                         if radical1 == radical2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return False
 */
          __pyx_t_7 = ((__pyx_v_radical1 == __pyx_v_radical2) != 0);
          if (__pyx_t_7) {
            goto __pyx_L15_break;
          }

          /* "rmgpy/molecule/group.py":373
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                         if radical1 == radical2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":376
 *                         if radical1 == radical2: break
 *                     else:
 *                         return False             # <<<<<<<<<<<<<<
 *         else:
 *             if group.radicalElectrons: return False
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":373
 *             for radical1 in self.radicalElectrons:
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:             # <<<<<<<<<<<<<<
 *                         if radical1 == radical2: break
 *                     else:
 */
        __pyx_L15_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L13;
      }
      __pyx_L13:;

      /* "rmgpy/molecule/group.py":371
 *         # Each free radical electron state in self must have an equivalent in other (and vice versa)
 *         if self.radicalElectrons:
 *             for radical1 in self.radicalElectrons:             # <<<<<<<<<<<<<<
 *                 if group.radicalElectrons:
 *                     for radical2 in group.radicalElectrons:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L10;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":378
 *                         return False
 *         else:
 *             if group.radicalElectrons: return False             # <<<<<<<<<<<<<<
 *         if self.lonePairs:
 *             for lp1 in self.lonePairs:
 */
    __pyx_t_7 = (__pyx_v_group->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_group->radicalElectrons) != 0);
    if (__pyx_t_7) {
      __pyx_r = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L10:;

  /* "rmgpy/molecule/group.py":379
 *         else:
 *             if group.radicalElectrons: return False
 *         if self.lonePairs:             # <<<<<<<<<<<<<<
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:
 */
  __pyx_t_7 = (__pyx_v_self->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_self->lonePairs) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":380
 *             if group.radicalElectrons: return False
 *         if self.lonePairs:
 *             for lp1 in self.lonePairs:             # <<<<<<<<<<<<<<
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:
 */
    if (unlikely(__pyx_v_self->lonePairs == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 380; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->lonePairs; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
    for (;;) {
      if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 380; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 380; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 380; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_lp1 = __pyx_t_10;

      /* "rmgpy/molecule/group.py":381
 *         if self.lonePairs:
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:             # <<<<<<<<<<<<<<
 *                     for lp2 in group.lonePairs:
 *                         if lp1 == lp2: break
 */
      __pyx_t_7 = (__pyx_v_group->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_group->lonePairs) != 0);
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":382
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                         if lp1 == lp2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->lonePairs == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 382; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->lonePairs; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
        for (;;) {
          if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 382; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 382; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 382; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_v_lp2 = __pyx_t_10;

          /* "rmgpy/molecule/group.py":383
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:
 *                         if lp1 == lp2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return False
 */
          __pyx_t_7 = ((__pyx_v_lp1 == __pyx_v_lp2) != 0);
          if (__pyx_t_7) {
            goto __pyx_L24_break;
          }

          /* "rmgpy/molecule/group.py":382
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                         if lp1 == lp2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":385
 *                         if lp1 == lp2: break
 *                     else:
 *                         return False             # <<<<<<<<<<<<<<
 *         else:
 *             if group.lonePairs: return False
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":382
 *             for lp1 in self.lonePairs:
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:             # <<<<<<<<<<<<<<
 *                         if lp1 == lp2: break
 *                     else:
 */
        __pyx_L24_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L22;
      }
      __pyx_L22:;

      /* "rmgpy/molecule/group.py":380
 *             if group.radicalElectrons: return False
 *         if self.lonePairs:
 *             for lp1 in self.lonePairs:             # <<<<<<<<<<<<<<
 *                 if group.lonePairs:
 *                     for lp2 in group.lonePairs:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L19;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":387
 *                         return False
 *         else:
 *             if group.lonePairs: return False             # <<<<<<<<<<<<<<
 *         #Each charge in self must have an equivalent in other
 *         if self.charge:
 */
    __pyx_t_7 = (__pyx_v_group->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_group->lonePairs) != 0);
    if (__pyx_t_7) {
      __pyx_r = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L19:;

  /* "rmgpy/molecule/group.py":389
 *             if group.lonePairs: return False
 *         #Each charge in self must have an equivalent in other
 *         if self.charge:             # <<<<<<<<<<<<<<
 *             for charge1 in self.charge:
 *                 if group.charge:
 */
  __pyx_t_7 = (__pyx_v_self->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_self->charge) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":390
 *         #Each charge in self must have an equivalent in other
 *         if self.charge:
 *             for charge1 in self.charge:             # <<<<<<<<<<<<<<
 *                 if group.charge:
 *                     for charge2 in group.charge:
 */
    if (unlikely(__pyx_v_self->charge == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->charge; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
    for (;;) {
      if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 390; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_charge1 = __pyx_t_10;

      /* "rmgpy/molecule/group.py":391
 *         if self.charge:
 *             for charge1 in self.charge:
 *                 if group.charge:             # <<<<<<<<<<<<<<
 *                     for charge2 in group.charge:
 *                         if charge1 == charge2: break
 */
      __pyx_t_7 = (__pyx_v_group->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_group->charge) != 0);
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":392
 *             for charge1 in self.charge:
 *                 if group.charge:
 *                     for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                         if charge1 == charge2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->charge == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 392; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->charge; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
        for (;;) {
          if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 392; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 392; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_10 = __Pyx_PyInt_As_short(__pyx_t_3); if (unlikely((__pyx_t_10 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 392; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_v_charge2 = __pyx_t_10;

          /* "rmgpy/molecule/group.py":393
 *                 if group.charge:
 *                     for charge2 in group.charge:
 *                         if charge1 == charge2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return False
 */
          __pyx_t_7 = ((__pyx_v_charge1 == __pyx_v_charge2) != 0);
          if (__pyx_t_7) {
            goto __pyx_L33_break;
          }

          /* "rmgpy/molecule/group.py":392
 *             for charge1 in self.charge:
 *                 if group.charge:
 *                     for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                         if charge1 == charge2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":395
 *                         if charge1 == charge2: break
 *                     else:
 *                         return False             # <<<<<<<<<<<<<<
 *         else:
 *             if group.charge: return False
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":392
 *             for charge1 in self.charge:
 *                 if group.charge:
 *                     for charge2 in group.charge:             # <<<<<<<<<<<<<<
 *                         if charge1 == charge2: break
 *                     else:
 */
        __pyx_L33_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L31;
      }
      __pyx_L31:;

      /* "rmgpy/molecule/group.py":390
 *         #Each charge in self must have an equivalent in other
 *         if self.charge:
 *             for charge1 in self.charge:             # <<<<<<<<<<<<<<
 *                 if group.charge:
 *                     for charge2 in group.charge:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L28;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":397
 *                         return False
 *         else:
 *             if group.charge: return False             # <<<<<<<<<<<<<<
 *         # Otherwise self is in fact a specific case of other
 *         return True
 */
    __pyx_t_7 = (__pyx_v_group->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_group->charge) != 0);
    if (__pyx_t_7) {
      __pyx_r = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L28:;

  /* "rmgpy/molecule/group.py":399
 *             if group.charge: return False
 *         # Otherwise self is in fact a specific case of other
 *         return True             # <<<<<<<<<<<<<<
 * 
 *     def isOxygen(self):
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":346
 *         return True
 * 
 *     def isSpecificCaseOf(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is the same as `self` or is a more
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.isSpecificCaseOf", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_group);
  __Pyx_XDECREF((PyObject *)__pyx_v_atomType1);
  __Pyx_XDECREF(__pyx_v_atomType2);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf[] = "\n        Returns ``True`` if `other` is the same as `self` or is a more\n        specific case of `self`. Returns ``False`` if some of `self` is not\n        included in `other` or they are mutually exclusive. \n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSpecificCaseOf (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Vertex, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSpecificCaseOf", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSpecificCaseOf(__pyx_v_self, __pyx_v_other, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 346; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.isSpecificCaseOf", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":401
 *         return True
 * 
 *     def isOxygen(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom represents an oxygen atom or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isOxygen(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_allOxygens = NULL;
  PyObject *__pyx_v_checkList = NULL;
  PyObject *__pyx_v_x = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isOxygen", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isOxygen); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 401; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 401; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 401; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 401; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":406
 *         not.
 *         """
 *         allOxygens = [atomTypes['O']] + atomTypes['O'].specific             # <<<<<<<<<<<<<<
 *         checkList=[x in allOxygens for x in self.atomType]
 * 
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_O); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_t_2);
  __Pyx_GIVEREF(__pyx_t_2);
  __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_O); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_specific); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = PyNumber_Add(__pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 406; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_allOxygens = __pyx_t_3;
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":407
 *         """
 *         allOxygens = [atomTypes['O']] + atomTypes['O'].specific
 *         checkList=[x in allOxygens for x in self.atomType]             # <<<<<<<<<<<<<<
 * 
 *         return all(checkList)
 */
  __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 407; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 407; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 407; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 407; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_x, __pyx_v_allOxygens, Py_EQ)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 407; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_t_5); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 407; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_3, (PyObject*)__pyx_t_1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 407; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_checkList = ((PyObject*)__pyx_t_3);
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":409
 *         checkList=[x in allOxygens for x in self.atomType]
 * 
 *         return all(checkList)             # <<<<<<<<<<<<<<
 * 
 *     def isSulfur(self):
 */
  __pyx_t_3 = PyTuple_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 409; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_v_checkList);
  PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_v_checkList);
  __Pyx_GIVEREF(__pyx_v_checkList);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_all, __pyx_t_3, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 409; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 409; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":401
 *         return True
 * 
 *     def isOxygen(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom represents an oxygen atom or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.GroupAtom.isOxygen", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_allOxygens);
  __Pyx_XDECREF(__pyx_v_checkList);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen[] = "\n        Return ``True`` if the atom represents an oxygen atom or ``False`` if\n        not.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isOxygen (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isOxygen", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isOxygen(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 401; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.isOxygen", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":411
 *         return all(checkList)
 * 
 *     def isSulfur(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom represents an sulfur atom or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSulfur(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_allSulfur = NULL;
  PyObject *__pyx_v_checkList = NULL;
  PyObject *__pyx_v_x = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSulfur", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSulfur); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":416
 *         not.
 *         """
 *         allSulfur = [atomTypes['S']] + atomTypes['S'].specific             # <<<<<<<<<<<<<<
 *         checkList=[x in allSulfur for x in self.atomType]
 * 
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 416; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_S); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 416; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 416; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_t_2);
  __Pyx_GIVEREF(__pyx_t_2);
  __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 416; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_S); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 416; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_specific); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 416; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = PyNumber_Add(__pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 416; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_allSulfur = __pyx_t_3;
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":417
 *         """
 *         allSulfur = [atomTypes['S']] + atomTypes['S'].specific
 *         checkList=[x in allSulfur for x in self.atomType]             # <<<<<<<<<<<<<<
 * 
 *         return all(checkList)
 */
  __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 417; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 417; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __pyx_v_self->atomType; __Pyx_INCREF(__pyx_t_2); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_2)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 417; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 417; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_x, __pyx_v_allSulfur, Py_EQ)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 417; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_t_5); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 417; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_3, (PyObject*)__pyx_t_1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 417; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_checkList = ((PyObject*)__pyx_t_3);
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":419
 *         checkList=[x in allSulfur for x in self.atomType]
 * 
 *         return all(checkList)             # <<<<<<<<<<<<<<
 * 
 *     def hasWildcards(self):
 */
  __pyx_t_3 = PyTuple_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 419; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_v_checkList);
  PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_v_checkList);
  __Pyx_GIVEREF(__pyx_v_checkList);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_all, __pyx_t_3, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 419; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 419; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":411
 *         return all(checkList)
 * 
 *     def isSulfur(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom represents an sulfur atom or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.GroupAtom.isSulfur", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_allSulfur);
  __Pyx_XDECREF(__pyx_v_checkList);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur[] = "\n        Return ``True`` if the atom represents an sulfur atom or ``False`` if\n        not.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSulfur (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSulfur", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSulfur(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 411; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.isSulfur", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":421
 *         return all(checkList)
 * 
 *     def hasWildcards(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom has wildcards in any of the attributes:
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_39hasWildcards(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_hasWildcards(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_bond = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_t_7;
  PyObject *(*__pyx_t_8)(PyObject *);
  Py_ssize_t __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasWildcards", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_hasWildcards); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 421; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_39hasWildcards)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 421; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 421; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 421; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":428
 *         """
 * 
 *         if len(self.atomType) > 1:             # <<<<<<<<<<<<<<
 *             return True
 *         elif len(self.radicalElectrons) > 1 or len(self.radicalElectrons) == 0:
 */
  __pyx_t_1 = __pyx_v_self->atomType;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_6 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_5 = ((__pyx_t_6 > 1) != 0);
  if (__pyx_t_5) {

    /* "rmgpy/molecule/group.py":429
 * 
 *         if len(self.atomType) > 1:
 *             return True             # <<<<<<<<<<<<<<
 *         elif len(self.radicalElectrons) > 1 or len(self.radicalElectrons) == 0:
 *             return True
 */
    __pyx_r = 1;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":430
 *         if len(self.atomType) > 1:
 *             return True
 *         elif len(self.radicalElectrons) > 1 or len(self.radicalElectrons) == 0:             # <<<<<<<<<<<<<<
 *             return True
 *         elif len(self.lonePairs) > 1:
 */
  __pyx_t_1 = __pyx_v_self->radicalElectrons;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_6 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_7 = ((__pyx_t_6 > 1) != 0);
  if (!__pyx_t_7) {
    goto __pyx_L5_next_or;
  } else {
    __pyx_t_5 = __pyx_t_7;
    goto __pyx_L4_bool_binop_done;
  }
  __pyx_L5_next_or:;
  __pyx_t_1 = __pyx_v_self->radicalElectrons;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_6 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_7 = ((__pyx_t_6 == 0) != 0);
  __pyx_t_5 = __pyx_t_7;
  __pyx_L4_bool_binop_done:;
  if (__pyx_t_5) {

    /* "rmgpy/molecule/group.py":431
 *             return True
 *         elif len(self.radicalElectrons) > 1 or len(self.radicalElectrons) == 0:
 *             return True             # <<<<<<<<<<<<<<
 *         elif len(self.lonePairs) > 1:
 *             return True
 */
    __pyx_r = 1;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":432
 *         elif len(self.radicalElectrons) > 1 or len(self.radicalElectrons) == 0:
 *             return True
 *         elif len(self.lonePairs) > 1:             # <<<<<<<<<<<<<<
 *             return True
 *         for bond in self.bonds.values():
 */
  __pyx_t_1 = __pyx_v_self->lonePairs;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 432; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_6 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 432; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_5 = ((__pyx_t_6 > 1) != 0);
  if (__pyx_t_5) {

    /* "rmgpy/molecule/group.py":433
 *             return True
 *         elif len(self.lonePairs) > 1:
 *             return True             # <<<<<<<<<<<<<<
 *         for bond in self.bonds.values():
 *             if len(bond.order) > 1:
 */
    __pyx_r = 1;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":434
 *         elif len(self.lonePairs) > 1:
 *             return True
 *         for bond in self.bonds.values():             # <<<<<<<<<<<<<<
 *             if len(bond.order) > 1:
 *                 return True
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_bonds); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_values); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (__pyx_t_2) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
    __pyx_t_3 = __pyx_t_1; __Pyx_INCREF(__pyx_t_3); __pyx_t_6 = 0;
    __pyx_t_8 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_8 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (;;) {
    if (likely(!__pyx_t_8)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_1 = __pyx_t_8(__pyx_t_3);
      if (unlikely(!__pyx_t_1)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_1);
    }
    __Pyx_XDECREF_SET(__pyx_v_bond, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":435
 *             return True
 *         for bond in self.bonds.values():
 *             if len(bond.order) > 1:             # <<<<<<<<<<<<<<
 *                 return True
 * 
 */
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond, __pyx_n_s_order); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 435; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_9 = PyObject_Length(__pyx_t_1); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 435; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_5 = ((__pyx_t_9 > 1) != 0);
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":436
 *         for bond in self.bonds.values():
 *             if len(bond.order) > 1:
 *                 return True             # <<<<<<<<<<<<<<
 * 
 *         return False
 */
      __pyx_r = 1;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":434
 *         elif len(self.lonePairs) > 1:
 *             return True
 *         for bond in self.bonds.values():             # <<<<<<<<<<<<<<
 *             if len(bond.order) > 1:
 *                 return True
 */
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":438
 *                 return True
 * 
 *         return False             # <<<<<<<<<<<<<<
 * 
 *     def makeSampleAtom(self):
 */
  __pyx_r = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":421
 *         return all(checkList)
 * 
 *     def hasWildcards(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the atom has wildcards in any of the attributes:
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.GroupAtom.hasWildcards", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_bond);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_39hasWildcards(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_38hasWildcards[] = "\n        Return ``True`` if the atom has wildcards in any of the attributes:\n        atomtype, electronpairs, lone pairs, charge, and bond order. Returns\n        ''False'' if no attribute has wildcards.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_39hasWildcards(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("hasWildcards (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_38hasWildcards(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_38hasWildcards(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasWildcards", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_hasWildcards(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 421; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.hasWildcards", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":440
 *         return False
 * 
 *     def makeSampleAtom(self):             # <<<<<<<<<<<<<<
 *         """
 * 
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_41makeSampleAtom(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_makeSampleAtom(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atomtype = NULL;
  PyObject *__pyx_v_element = NULL;
  PyObject *__pyx_v_defaultLonePairs = NULL;
  PyObject *__pyx_v_elementLabel = NULL;
  PyObject *__pyx_v_subtype = NULL;
  PyObject *__pyx_v_defaultAtom = NULL;
  PyObject *__pyx_v_newAtom = NULL;
  PyObject *__pyx_v_x = NULL;
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *(*__pyx_t_6)(PyObject *);
  int __pyx_t_7;
  int __pyx_t_8;
  int __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  PyObject *(*__pyx_t_11)(PyObject *);
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("makeSampleAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_makeSampleAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 440; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_41makeSampleAtom)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 440; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 440; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8molecule_Atom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 440; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":451
 * 
 *         #Use the first atomtype to determine element, even if there is more than one atomtype
 *         atomtype = self.atomType[0]             # <<<<<<<<<<<<<<
 *         element = None
 * 
 */
  if (unlikely(__pyx_v_self->atomType == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 451; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->atomType, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 451; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomtype = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":452
 *         #Use the first atomtype to determine element, even if there is more than one atomtype
 *         atomtype = self.atomType[0]
 *         element = None             # <<<<<<<<<<<<<<
 * 
 *         defaultLonePairs={'H': 0,
 */
  __Pyx_INCREF(Py_None);
  __pyx_v_element = Py_None;

  /* "rmgpy/molecule/group.py":454
 *         element = None
 * 
 *         defaultLonePairs={'H': 0,             # <<<<<<<<<<<<<<
 *                           'D': 0,
 *                           'T': 0,
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_H, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_D, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_T, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_He, __pyx_int_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_C, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_O, __pyx_int_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_N, __pyx_int_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_Si, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_S, __pyx_int_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_Ne, __pyx_int_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_Cl, __pyx_int_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_Ar, __pyx_int_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 454; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_defaultLonePairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":468
 *         }
 * 
 *         for elementLabel in allElements:             # <<<<<<<<<<<<<<
 *             if atomtype is atomTypes[elementLabel] or atomtype in atomTypes[elementLabel].specific:
 *                 element = elementLabel
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_allElements); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
    __pyx_t_2 = __pyx_t_1; __Pyx_INCREF(__pyx_t_2); __pyx_t_5 = 0;
    __pyx_t_6 = NULL;
  } else {
    __pyx_t_5 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (;;) {
    if (likely(!__pyx_t_6)) {
      if (likely(PyList_CheckExact(__pyx_t_2))) {
        if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_1); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_5 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_1); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_1 = __pyx_t_6(__pyx_t_2);
      if (unlikely(!__pyx_t_1)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 468; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_1);
    }
    __Pyx_XDECREF_SET(__pyx_v_elementLabel, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":469
 * 
 *         for elementLabel in allElements:
 *             if atomtype is atomTypes[elementLabel] or atomtype in atomTypes[elementLabel].specific:             # <<<<<<<<<<<<<<
 *                 element = elementLabel
 *                 break
 */
    __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 469; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_3 = PyObject_GetItem(__pyx_t_1, __pyx_v_elementLabel); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 469; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_8 = (__pyx_v_atomtype == __pyx_t_3);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_9 = (__pyx_t_8 != 0);
    if (!__pyx_t_9) {
      goto __pyx_L7_next_or;
    } else {
      __pyx_t_7 = __pyx_t_9;
      goto __pyx_L6_bool_binop_done;
    }
    __pyx_L7_next_or:;
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 469; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_1 = PyObject_GetItem(__pyx_t_3, __pyx_v_elementLabel); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 469; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_specific); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 469; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_9 = (__Pyx_PySequence_Contains(__pyx_v_atomtype, __pyx_t_3, Py_EQ)); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 469; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_8 = (__pyx_t_9 != 0);
    __pyx_t_7 = __pyx_t_8;
    __pyx_L6_bool_binop_done:;
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":470
 *         for elementLabel in allElements:
 *             if atomtype is atomTypes[elementLabel] or atomtype in atomTypes[elementLabel].specific:
 *                 element = elementLabel             # <<<<<<<<<<<<<<
 *                 break
 *         else:
 */
      __Pyx_INCREF(__pyx_v_elementLabel);
      __Pyx_DECREF_SET(__pyx_v_element, __pyx_v_elementLabel);

      /* "rmgpy/molecule/group.py":471
 *             if atomtype is atomTypes[elementLabel] or atomtype in atomTypes[elementLabel].specific:
 *                 element = elementLabel
 *                 break             # <<<<<<<<<<<<<<
 *         else:
 *             #For types that correspond to more than one type of element, pick the first that appears in specific
 */
      goto __pyx_L4_break;
    }

    /* "rmgpy/molecule/group.py":468
 *         }
 * 
 *         for elementLabel in allElements:             # <<<<<<<<<<<<<<
 *             if atomtype is atomTypes[elementLabel] or atomtype in atomTypes[elementLabel].specific:
 *                 element = elementLabel
 */
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":474
 *         else:
 *             #For types that correspond to more than one type of element, pick the first that appears in specific
 *             for subtype in atomtype.specific:             # <<<<<<<<<<<<<<
 *                 if subtype.label in allElements:
 *                     element = subtype.label
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomtype, __pyx_n_s_specific); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 474; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    if (likely(PyList_CheckExact(__pyx_t_3)) || PyTuple_CheckExact(__pyx_t_3)) {
      __pyx_t_1 = __pyx_t_3; __Pyx_INCREF(__pyx_t_1); __pyx_t_10 = 0;
      __pyx_t_11 = NULL;
    } else {
      __pyx_t_10 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 474; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_11 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 474; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    for (;;) {
      if (likely(!__pyx_t_11)) {
        if (likely(PyList_CheckExact(__pyx_t_1))) {
          if (__pyx_t_10 >= PyList_GET_SIZE(__pyx_t_1)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_10); __Pyx_INCREF(__pyx_t_3); __pyx_t_10++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 474; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_10); __pyx_t_10++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 474; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        } else {
          if (__pyx_t_10 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_10); __Pyx_INCREF(__pyx_t_3); __pyx_t_10++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 474; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_10); __pyx_t_10++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 474; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        }
      } else {
        __pyx_t_3 = __pyx_t_11(__pyx_t_1);
        if (unlikely(!__pyx_t_3)) {
          PyObject* exc_type = PyErr_Occurred();
          if (exc_type) {
            if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
            else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 474; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          break;
        }
        __Pyx_GOTREF(__pyx_t_3);
      }
      __Pyx_XDECREF_SET(__pyx_v_subtype, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":475
 *             #For types that correspond to more than one type of element, pick the first that appears in specific
 *             for subtype in atomtype.specific:
 *                 if subtype.label in allElements:             # <<<<<<<<<<<<<<
 *                     element = subtype.label
 *                     break
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_subtype, __pyx_n_s_label); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 475; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_allElements); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 475; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_7 = (__Pyx_PySequence_Contains(__pyx_t_3, __pyx_t_4, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 475; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_8 = (__pyx_t_7 != 0);
      if (__pyx_t_8) {

        /* "rmgpy/molecule/group.py":476
 *             for subtype in atomtype.specific:
 *                 if subtype.label in allElements:
 *                     element = subtype.label             # <<<<<<<<<<<<<<
 *                     break
 * 
 */
        __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_subtype, __pyx_n_s_label); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 476; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF_SET(__pyx_v_element, __pyx_t_4);
        __pyx_t_4 = 0;

        /* "rmgpy/molecule/group.py":477
 *                 if subtype.label in allElements:
 *                     element = subtype.label
 *                     break             # <<<<<<<<<<<<<<
 * 
 *         #dummy defaultAtom to get default values
 */
        goto __pyx_L10_break;
      }

      /* "rmgpy/molecule/group.py":474
 *         else:
 *             #For types that correspond to more than one type of element, pick the first that appears in specific
 *             for subtype in atomtype.specific:             # <<<<<<<<<<<<<<
 *                 if subtype.label in allElements:
 *                     element = subtype.label
 */
    }
    __pyx_L10_break:;
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":468
 *         }
 * 
 *         for elementLabel in allElements:             # <<<<<<<<<<<<<<
 *             if atomtype is atomTypes[elementLabel] or atomtype in atomTypes[elementLabel].specific:
 *                 element = elementLabel
 */
  __pyx_L4_break:;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":480
 * 
 *         #dummy defaultAtom to get default values
 *         defaultAtom = mol.Atom()             # <<<<<<<<<<<<<<
 * 
 *         newAtom = mol.Atom(element = element,
 */
  __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_8molecule_Atom)), __pyx_empty_tuple, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 480; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_v_defaultAtom = __pyx_t_2;
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":482
 *         defaultAtom = mol.Atom()
 * 
 *         newAtom = mol.Atom(element = element,             # <<<<<<<<<<<<<<
 *                            radicalElectrons = self.radicalElectrons[0] if self.radicalElectrons else defaultAtom.radicalElectrons,
 *                            charge = self.charge[0] if self.charge else defaultAtom.charge,
 */
  __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_element, __pyx_v_element) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":483
 * 
 *         newAtom = mol.Atom(element = element,
 *                            radicalElectrons = self.radicalElectrons[0] if self.radicalElectrons else defaultAtom.radicalElectrons,             # <<<<<<<<<<<<<<
 *                            charge = self.charge[0] if self.charge else defaultAtom.charge,
 *                            lonePairs = self.lonePairs[0] if self.lonePairs else defaultAtom.lonePairs,
 */
  __pyx_t_8 = (__pyx_v_self->radicalElectrons != Py_None) && (PyList_GET_SIZE(__pyx_v_self->radicalElectrons) != 0);
  if (__pyx_t_8) {
    if (unlikely(__pyx_v_self->radicalElectrons == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 483; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_4 = __Pyx_GetItemInt_List(__pyx_v_self->radicalElectrons, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 483; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
  } else {
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_defaultAtom, __pyx_n_s_radicalElectrons); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 483; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
  }
  if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_radicalElectrons, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":484
 *         newAtom = mol.Atom(element = element,
 *                            radicalElectrons = self.radicalElectrons[0] if self.radicalElectrons else defaultAtom.radicalElectrons,
 *                            charge = self.charge[0] if self.charge else defaultAtom.charge,             # <<<<<<<<<<<<<<
 *                            lonePairs = self.lonePairs[0] if self.lonePairs else defaultAtom.lonePairs,
 *                            label = self.label if self.label else defaultAtom.label)
 */
  __pyx_t_8 = (__pyx_v_self->charge != Py_None) && (PyList_GET_SIZE(__pyx_v_self->charge) != 0);
  if (__pyx_t_8) {
    if (unlikely(__pyx_v_self->charge == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_4 = __Pyx_GetItemInt_List(__pyx_v_self->charge, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
  } else {
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_defaultAtom, __pyx_n_s_charge); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
  }
  if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_charge, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":485
 *                            radicalElectrons = self.radicalElectrons[0] if self.radicalElectrons else defaultAtom.radicalElectrons,
 *                            charge = self.charge[0] if self.charge else defaultAtom.charge,
 *                            lonePairs = self.lonePairs[0] if self.lonePairs else defaultAtom.lonePairs,             # <<<<<<<<<<<<<<
 *                            label = self.label if self.label else defaultAtom.label)
 * 
 */
  __pyx_t_8 = (__pyx_v_self->lonePairs != Py_None) && (PyList_GET_SIZE(__pyx_v_self->lonePairs) != 0);
  if (__pyx_t_8) {
    if (unlikely(__pyx_v_self->lonePairs == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 485; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_4 = __Pyx_GetItemInt_List(__pyx_v_self->lonePairs, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 485; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
  } else {
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_defaultAtom, __pyx_n_s_lonePairs); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 485; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
  }
  if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_lonePairs, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":486
 *                            charge = self.charge[0] if self.charge else defaultAtom.charge,
 *                            lonePairs = self.lonePairs[0] if self.lonePairs else defaultAtom.lonePairs,
 *                            label = self.label if self.label else defaultAtom.label)             # <<<<<<<<<<<<<<
 * 
 *         #For some reason the default when no lone pairs is set to -100,
 */
  __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_v_self->label); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 486; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_8) {
    __Pyx_INCREF(__pyx_v_self->label);
    __pyx_t_1 = __pyx_v_self->label;
  } else {
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_defaultAtom, __pyx_n_s_label); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 486; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
  }
  if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_label, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":482
 *         defaultAtom = mol.Atom()
 * 
 *         newAtom = mol.Atom(element = element,             # <<<<<<<<<<<<<<
 *                            radicalElectrons = self.radicalElectrons[0] if self.radicalElectrons else defaultAtom.radicalElectrons,
 *                            charge = self.charge[0] if self.charge else defaultAtom.charge,
 */
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_8molecule_Atom)), __pyx_empty_tuple, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 482; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_newAtom = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":492
 *         #Instead we will set it to 0 here
 * 
 *         if newAtom.lonePairs == -100:             # <<<<<<<<<<<<<<
 *             if atomtype in [atomTypes[x] for x in ['N5d', 'N5dd', 'N5t', 'N5b', 'N5s']]:
 *                 newAtom.lonePairs = 0
 */
  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_newAtom, __pyx_n_s_lonePairs); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 492; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_RichCompare(__pyx_t_1, __pyx_int_neg_100, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 492; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_8 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 492; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":493
 * 
 *         if newAtom.lonePairs == -100:
 *             if atomtype in [atomTypes[x] for x in ['N5d', 'N5dd', 'N5t', 'N5b', 'N5s']]:             # <<<<<<<<<<<<<<
 *                 newAtom.lonePairs = 0
 *                 newAtom.charge = 1
 */
    __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = __pyx_tuple__3; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
    for (;;) {
      if (__pyx_t_5 >= 5) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_4 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_4); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_4 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_4);
      __pyx_t_4 = 0;
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_GetItem(__pyx_t_4, __pyx_v_x); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_2, (PyObject*)__pyx_t_3))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_8 = (__Pyx_PySequence_Contains(__pyx_v_atomtype, __pyx_t_2, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_7 = (__pyx_t_8 != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":494
 *         if newAtom.lonePairs == -100:
 *             if atomtype in [atomTypes[x] for x in ['N5d', 'N5dd', 'N5t', 'N5b', 'N5s']]:
 *                 newAtom.lonePairs = 0             # <<<<<<<<<<<<<<
 *                 newAtom.charge = 1
 *             elif atomtype is atomTypes['N1d']:
 */
      if (__Pyx_PyObject_SetAttrStr(__pyx_v_newAtom, __pyx_n_s_lonePairs, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 494; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

      /* "rmgpy/molecule/group.py":495
 *             if atomtype in [atomTypes[x] for x in ['N5d', 'N5dd', 'N5t', 'N5b', 'N5s']]:
 *                 newAtom.lonePairs = 0
 *                 newAtom.charge = 1             # <<<<<<<<<<<<<<
 *             elif atomtype is atomTypes['N1d']:
 *                 newAtom.charge = -1
 */
      if (__Pyx_PyObject_SetAttrStr(__pyx_v_newAtom, __pyx_n_s_charge, __pyx_int_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 495; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L13;
    }

    /* "rmgpy/molecule/group.py":496
 *                 newAtom.lonePairs = 0
 *                 newAtom.charge = 1
 *             elif atomtype is atomTypes['N1d']:             # <<<<<<<<<<<<<<
 *                 newAtom.charge = -1
 *             else:
 */
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_N1d); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 496; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_7 = (__pyx_v_atomtype == __pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_8 = (__pyx_t_7 != 0);
    if (__pyx_t_8) {

      /* "rmgpy/molecule/group.py":497
 *                 newAtom.charge = 1
 *             elif atomtype is atomTypes['N1d']:
 *                 newAtom.charge = -1             # <<<<<<<<<<<<<<
 *             else:
 *                 newAtom.lonePairs = defaultLonePairs[newAtom.symbol]
 */
      if (__Pyx_PyObject_SetAttrStr(__pyx_v_newAtom, __pyx_n_s_charge, __pyx_int_neg_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 497; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L13;
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":499
 *                 newAtom.charge = -1
 *             else:
 *                 newAtom.lonePairs = defaultLonePairs[newAtom.symbol]             # <<<<<<<<<<<<<<
 * 
 *         return newAtom
 */
      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_newAtom, __pyx_n_s_symbol); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 499; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_2 = __Pyx_PyDict_GetItem(__pyx_v_defaultLonePairs, __pyx_t_1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 499; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      if (__Pyx_PyObject_SetAttrStr(__pyx_v_newAtom, __pyx_n_s_lonePairs, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 499; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    }
    __pyx_L13:;
    goto __pyx_L12;
  }
  __pyx_L12:;

  /* "rmgpy/molecule/group.py":501
 *                 newAtom.lonePairs = defaultLonePairs[newAtom.symbol]
 * 
 *         return newAtom             # <<<<<<<<<<<<<<
 * 
 *     def getBondOrdersForAtom(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  if (!(likely(__Pyx_TypeTest(__pyx_v_newAtom, __pyx_ptype_5rmgpy_8molecule_8molecule_Atom)))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 501; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_INCREF(__pyx_v_newAtom);
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *)__pyx_v_newAtom);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":440
 *         return False
 * 
 *     def makeSampleAtom(self):             # <<<<<<<<<<<<<<
 *         """
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.makeSampleAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomtype);
  __Pyx_XDECREF(__pyx_v_element);
  __Pyx_XDECREF(__pyx_v_defaultLonePairs);
  __Pyx_XDECREF(__pyx_v_elementLabel);
  __Pyx_XDECREF(__pyx_v_subtype);
  __Pyx_XDECREF(__pyx_v_defaultAtom);
  __Pyx_XDECREF(__pyx_v_newAtom);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_41makeSampleAtom(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_40makeSampleAtom[] = "\n\n        Returns: a class :Atom: object analagous to the GroupAtom\n\n        This makes a sample, so it takes the first element when there are multiple options inside of\n        self.atomtype, self.radicalElectrons, self.lonePairs, and self.charge\n\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_41makeSampleAtom(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("makeSampleAtom (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_40makeSampleAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_40makeSampleAtom(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("makeSampleAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_makeSampleAtom(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 440; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.makeSampleAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":503
 *         return newAtom
 * 
 *     def getBondOrdersForAtom(self):             # <<<<<<<<<<<<<<
 *         """
 *         This helper function is to help calculate total bond orders for an
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_43getBondOrdersForAtom(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_getBondOrdersForAtom(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_num_B_bond = NULL;
  PyObject *__pyx_v_order = NULL;
  CYTHON_UNUSED PyObject *__pyx_v__ = NULL;
  PyObject *__pyx_v_bond = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_t_7;
  int __pyx_t_8;
  int __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBondOrdersForAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getBondOrdersForAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_43getBondOrdersForAtom)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":512
 *         than three `B` bonds, the order for  each is 3/2.0
 *         """
 *         num_B_bond = 0             # <<<<<<<<<<<<<<
 *         order = 0
 *         for _, bond in self.bonds.iteritems():
 */
  __Pyx_INCREF(__pyx_int_0);
  __pyx_v_num_B_bond = __pyx_int_0;

  /* "rmgpy/molecule/group.py":513
 *         """
 *         num_B_bond = 0
 *         order = 0             # <<<<<<<<<<<<<<
 *         for _, bond in self.bonds.iteritems():
 *             if bond.order == 'B':
 */
  __Pyx_INCREF(__pyx_int_0);
  __pyx_v_order = __pyx_int_0;

  /* "rmgpy/molecule/group.py":514
 *         num_B_bond = 0
 *         order = 0
 *         for _, bond in self.bonds.iteritems():             # <<<<<<<<<<<<<<
 *             if bond.order == 'B':
 *                 num_B_bond += 1
 */
  __pyx_t_5 = 0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_bonds); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (unlikely(__pyx_t_2 == Py_None)) {
    PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_3 = __Pyx_dict_iterator(__pyx_t_2, 0, __pyx_n_s_iteritems, (&__pyx_t_6), (&__pyx_t_7)); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_XDECREF(__pyx_t_1);
  __pyx_t_1 = __pyx_t_3;
  __pyx_t_3 = 0;
  while (1) {
    __pyx_t_8 = __Pyx_dict_iter_next(__pyx_t_1, __pyx_t_6, &__pyx_t_5, &__pyx_t_3, &__pyx_t_2, NULL, __pyx_t_7);
    if (unlikely(__pyx_t_8 == 0)) break;
    if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 514; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_XDECREF_SET(__pyx_v__, __pyx_t_3);
    __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_bond, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":515
 *         order = 0
 *         for _, bond in self.bonds.iteritems():
 *             if bond.order == 'B':             # <<<<<<<<<<<<<<
 *                 num_B_bond += 1
 *             else:
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond, __pyx_n_s_order); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_9 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_n_s_B, Py_EQ)); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_9) {

      /* "rmgpy/molecule/group.py":516
 *         for _, bond in self.bonds.iteritems():
 *             if bond.order == 'B':
 *                 num_B_bond += 1             # <<<<<<<<<<<<<<
 *             else:
 *                 order += mol.bond_orders[bond.order]
 */
      __pyx_t_2 = PyNumber_InPlaceAdd(__pyx_v_num_B_bond, __pyx_int_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 516; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF_SET(__pyx_v_num_B_bond, __pyx_t_2);
      __pyx_t_2 = 0;
      goto __pyx_L5;
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":518
 *                 num_B_bond += 1
 *             else:
 *                 order += mol.bond_orders[bond.order]             # <<<<<<<<<<<<<<
 * 
 *         if num_B_bond == 3:
 */
      if (unlikely(__pyx_v_5rmgpy_8molecule_8molecule_bond_orders == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 518; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond, __pyx_n_s_order); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 518; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_3 = __Pyx_PyDict_GetItem(__pyx_v_5rmgpy_8molecule_8molecule_bond_orders, __pyx_t_2); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 518; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_2 = PyNumber_InPlaceAdd(__pyx_v_order, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 518; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_DECREF_SET(__pyx_v_order, __pyx_t_2);
      __pyx_t_2 = 0;
    }
    __pyx_L5:;
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":520
 *                 order += mol.bond_orders[bond.order]
 * 
 *         if num_B_bond == 3:             # <<<<<<<<<<<<<<
 *             order += num_B_bond * 4/3.0
 *         else:
 */
  __pyx_t_1 = PyObject_RichCompare(__pyx_v_num_B_bond, __pyx_int_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 520; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 520; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_9) {

    /* "rmgpy/molecule/group.py":521
 * 
 *         if num_B_bond == 3:
 *             order += num_B_bond * 4/3.0             # <<<<<<<<<<<<<<
 *         else:
 *             order += num_B_bond * 3/2.0
 */
    __pyx_t_1 = PyNumber_Multiply(__pyx_v_num_B_bond, __pyx_int_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 521; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_2 = __Pyx_PyNumber_Divide(__pyx_t_1, __pyx_float_3_0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 521; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = PyNumber_InPlaceAdd(__pyx_v_order, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 521; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF_SET(__pyx_v_order, __pyx_t_1);
    __pyx_t_1 = 0;
    goto __pyx_L6;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":523
 *             order += num_B_bond * 4/3.0
 *         else:
 *             order += num_B_bond * 3/2.0             # <<<<<<<<<<<<<<
 * 
 *         return order
 */
    __pyx_t_1 = PyNumber_Multiply(__pyx_v_num_B_bond, __pyx_int_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 523; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_2 = __Pyx_PyNumber_Divide(__pyx_t_1, __pyx_float_2_0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 523; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = PyNumber_InPlaceAdd(__pyx_v_order, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 523; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF_SET(__pyx_v_order, __pyx_t_1);
    __pyx_t_1 = 0;
  }
  __pyx_L6:;

  /* "rmgpy/molecule/group.py":525
 *             order += num_B_bond * 3/2.0
 * 
 *         return order             # <<<<<<<<<<<<<<
 * 
 * ################################################################################
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_order);
  __pyx_r = __pyx_v_order;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":503
 *         return newAtom
 * 
 *     def getBondOrdersForAtom(self):             # <<<<<<<<<<<<<<
 *         """
 *         This helper function is to help calculate total bond orders for an
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.getBondOrdersForAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_num_B_bond);
  __Pyx_XDECREF(__pyx_v_order);
  __Pyx_XDECREF(__pyx_v__);
  __Pyx_XDECREF(__pyx_v_bond);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_43getBondOrdersForAtom(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_42getBondOrdersForAtom[] = "\n        This helper function is to help calculate total bond orders for an\n        input atom.\n\n        Some special consideration for the order `B` bond. For atoms having\n        three `B` bonds, the order for each is 4/3.0, while for atoms having other\n        than three `B` bonds, the order for  each is 3/2.0\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_43getBondOrdersForAtom(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getBondOrdersForAtom (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_42getBondOrdersForAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_42getBondOrdersForAtom(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBondOrdersForAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupAtom_getBondOrdersForAtom(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 503; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.getBondOrdersForAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":35
 * cdef class GroupAtom(Vertex):
 * 
 *     cdef public list atomType             # <<<<<<<<<<<<<<
 *     cdef public list radicalElectrons
 *     cdef public list charge
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->atomType);
  __pyx_r = __pyx_v_self->atomType;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 35; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.atomType.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_8atomType_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->atomType);
  __Pyx_DECREF(__pyx_v_self->atomType);
  __pyx_v_self->atomType = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":36
 * 
 *     cdef public list atomType
 *     cdef public list radicalElectrons             # <<<<<<<<<<<<<<
 *     cdef public list charge
 *     cdef public str label
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->radicalElectrons);
  __pyx_r = __pyx_v_self->radicalElectrons;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 36; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.radicalElectrons.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->radicalElectrons);
  __Pyx_DECREF(__pyx_v_self->radicalElectrons);
  __pyx_v_self->radicalElectrons = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":37
 *     cdef public list atomType
 *     cdef public list radicalElectrons
 *     cdef public list charge             # <<<<<<<<<<<<<<
 *     cdef public str label
 *     cdef public list lonePairs
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->charge);
  __pyx_r = __pyx_v_self->charge;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 37; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->charge);
  __Pyx_DECREF(__pyx_v_self->charge);
  __pyx_v_self->charge = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.charge.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_6charge_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->charge);
  __Pyx_DECREF(__pyx_v_self->charge);
  __pyx_v_self->charge = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":38
 *     cdef public list radicalElectrons
 *     cdef public list charge
 *     cdef public str label             # <<<<<<<<<<<<<<
 *     cdef public list lonePairs
 * 
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->label);
  __pyx_r = __pyx_v_self->label;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyString_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 38; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->label);
  __Pyx_DECREF(__pyx_v_self->label);
  __pyx_v_self->label = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.label.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_5label_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->label);
  __Pyx_DECREF(__pyx_v_self->label);
  __pyx_v_self->label = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":39
 *     cdef public list charge
 *     cdef public str label
 *     cdef public list lonePairs             # <<<<<<<<<<<<<<
 * 
 *     cpdef Vertex copy(self)
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->lonePairs);
  __pyx_r = __pyx_v_self->lonePairs;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupAtom.lonePairs.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->lonePairs);
  __Pyx_DECREF(__pyx_v_self->lonePairs);
  __pyx_v_self->lonePairs = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":545
 *     """
 * 
 *     def __init__(self, atom1, atom2, order=None):             # <<<<<<<<<<<<<<
 *         Edge.__init__(self, atom1, atom2)
 *         self.order = order or []
 */

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_atom1 = 0;
  PyObject *__pyx_v_atom2 = 0;
  PyObject *__pyx_v_order = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__init__ (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atom1,&__pyx_n_s_atom2,&__pyx_n_s_order,0};
    PyObject* values[3] = {0,0,0};
    values[2] = ((PyObject *)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom1)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom2)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("__init__", 0, 2, 3, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 545; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
        case  2:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_order);
          if (value) { values[2] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__init__") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 545; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_atom1 = values[0];
    __pyx_v_atom2 = values[1];
    __pyx_v_order = values[2];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("__init__", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 545; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return -1;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond___init__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), __pyx_v_atom1, __pyx_v_atom2, __pyx_v_order);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond___init__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_atom1, PyObject *__pyx_v_atom2, PyObject *__pyx_v_order) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  Py_ssize_t __pyx_t_4;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__init__", 0);

  /* "rmgpy/molecule/group.py":546
 * 
 *     def __init__(self, atom1, atom2, order=None):
 *         Edge.__init__(self, atom1, atom2)             # <<<<<<<<<<<<<<
 *         self.order = order or []
 * 
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5graph_Edge)), __pyx_n_s_init); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 546; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  __pyx_t_4 = 0;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
      __pyx_t_4 = 1;
    }
  }
  __pyx_t_5 = PyTuple_New(3+__pyx_t_4); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 546; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  if (__pyx_t_3) {
    PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
  }
  __Pyx_INCREF(((PyObject *)__pyx_v_self));
  PyTuple_SET_ITEM(__pyx_t_5, 0+__pyx_t_4, ((PyObject *)__pyx_v_self));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
  __Pyx_INCREF(__pyx_v_atom1);
  PyTuple_SET_ITEM(__pyx_t_5, 1+__pyx_t_4, __pyx_v_atom1);
  __Pyx_GIVEREF(__pyx_v_atom1);
  __Pyx_INCREF(__pyx_v_atom2);
  PyTuple_SET_ITEM(__pyx_t_5, 2+__pyx_t_4, __pyx_v_atom2);
  __Pyx_GIVEREF(__pyx_v_atom2);
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 546; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":547
 *     def __init__(self, atom1, atom2, order=None):
 *         Edge.__init__(self, atom1, atom2)
 *         self.order = order or []             # <<<<<<<<<<<<<<
 * 
 *     def __str__(self):
 */
  __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_v_order); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 547; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (!__pyx_t_6) {
    goto __pyx_L4_next_or;
  } else {
    if (!(likely(PyList_CheckExact(__pyx_v_order))||((__pyx_v_order) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_order)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 547; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_order);
    __pyx_t_1 = __pyx_v_order;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_or:;
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 547; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_t_2);
  __pyx_t_1 = __pyx_t_2;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_L3_bool_binop_done:;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->order);
  __Pyx_DECREF(__pyx_v_self->order);
  __pyx_v_self->order = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":545
 *     """
 * 
 *     def __init__(self, atom1, atom2, order=None):             # <<<<<<<<<<<<<<
 *         Edge.__init__(self, atom1, atom2)
 *         self.order = order or []
 */

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":549
 *         self.order = order or []
 * 
 *     def __str__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a human-readable string representation of the object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_3__str__(PyObject *__pyx_v_self); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_2__str__[] = "\n        Return a human-readable string representation of the object.\n        ";
#if CYTHON_COMPILING_IN_CPYTHON
struct wrapperbase __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_2__str__;
#endif
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_3__str__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__str__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_2__str__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_2__str__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__str__", 0);

  /* "rmgpy/molecule/group.py":553
 *         Return a human-readable string representation of the object.
 *         """
 *         return str(self.order)             # <<<<<<<<<<<<<<
 * 
 *     def __repr__(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 553; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_self->order);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_self->order);
  __Pyx_GIVEREF(__pyx_v_self->order);
  __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyString_Type))), __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 553; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":549
 *         self.order = order or []
 * 
 *     def __str__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a human-readable string representation of the object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__str__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":555
 *         return str(self.order)
 * 
 *     def __repr__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a representation that can be used to reconstruct the object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5__repr__(PyObject *__pyx_v_self); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_4__repr__[] = "\n        Return a representation that can be used to reconstruct the object.\n        ";
#if CYTHON_COMPILING_IN_CPYTHON
struct wrapperbase __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_4__repr__;
#endif
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5__repr__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__repr__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_4__repr__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_4__repr__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__repr__", 0);

  /* "rmgpy/molecule/group.py":559
 *         Return a representation that can be used to reconstruct the object.
 *         """
 *         return "<GroupBond {0!r}>".format(self.order)             # <<<<<<<<<<<<<<
 * 
 *     def __reduce__(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_GroupBond_0_r, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 559; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_v_self->order); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 559; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 559; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(__pyx_v_self->order);
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_v_self->order);
    __Pyx_GIVEREF(__pyx_v_self->order);
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 559; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":555
 *         return str(self.order)
 * 
 *     def __repr__(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a representation that can be used to reconstruct the object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__repr__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":561
 *         return "<GroupBond {0!r}>".format(self.order)
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_7__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_6__reduce__[] = "\n        A helper function used when pickling an object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_7__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__reduce__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_6__reduce__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_6__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__reduce__", 0);

  /* "rmgpy/molecule/group.py":565
 *         A helper function used when pickling an object.
 *         """
 *         return (GroupBond, (self.vertex1, self.vertex2, self.order))             # <<<<<<<<<<<<<<
 * 
 *     def copy(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 565; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  PyTuple_SET_ITEM(__pyx_t_1, 0, ((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  __Pyx_INCREF(((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  PyTuple_SET_ITEM(__pyx_t_1, 1, ((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  __Pyx_INCREF(__pyx_v_self->order);
  PyTuple_SET_ITEM(__pyx_t_1, 2, __pyx_v_self->order);
  __Pyx_GIVEREF(__pyx_v_self->order);
  __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 565; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)));
  PyTuple_SET_ITEM(__pyx_t_2, 0, ((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)));
  __Pyx_GIVEREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)));
  PyTuple_SET_ITEM(__pyx_t_2, 1, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":561
 *         return "<GroupBond {0!r}>".format(self.order)
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__reduce__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":567
 *         return (GroupBond, (self.vertex1, self.vertex2, self.order))
 * 
 *     def copy(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a deep copy of the :class:`GroupBond` object. Modifying the
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_copy); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5graph_Edge))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":572
 *         attributes of the copy will not affect the original.
 *         """
 *         return GroupBond(self.vertex1, self.vertex2, self.order[:])             # <<<<<<<<<<<<<<
 * 
 *     def isSingle(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 572; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_PyList_GetSlice(__pyx_v_self->order, 0, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 572; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 572; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  PyTuple_SET_ITEM(__pyx_t_2, 0, ((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self->__pyx_base.vertex1));
  __Pyx_INCREF(((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  PyTuple_SET_ITEM(__pyx_t_2, 1, ((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self->__pyx_base.vertex2));
  PyTuple_SET_ITEM(__pyx_t_2, 2, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 572; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":567
 *         return (GroupBond, (self.vertex1, self.vertex2, self.order))
 * 
 *     def copy(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return a deep copy of the :class:`GroupBond` object. Modifying the
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_8copy[] = "\n        Return a deep copy of the :class:`GroupBond` object. Modifying the\n        attributes of the copy will not affect the original.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("copy (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_8copy(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_8copy(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_9GroupBond_copy(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 567; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":574
 *         return GroupBond(self.vertex1, self.vertex2, self.order[:])
 * 
 *     def isSingle(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a single bond or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSingle(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSingle", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSingle); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 574; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 574; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 574; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 574; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":579
 *         not. Bonds with any wildcards will return  ``False``.
 *         """
 *         return self.order[0] == 'S' and len(self.order) == 1             # <<<<<<<<<<<<<<
 * 
 *     def isDouble(self):
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 579; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 579; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_S, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 579; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {
    goto __pyx_L4_next_and;
  } else {
    __pyx_t_5 = __pyx_t_6;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_and:;
  __pyx_t_1 = __pyx_v_self->order;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 579; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 579; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_6 = ((__pyx_t_7 == 1) != 0);
  __pyx_t_5 = __pyx_t_6;
  __pyx_L3_bool_binop_done:;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":574
 *         return GroupBond(self.vertex1, self.vertex2, self.order[:])
 * 
 *     def isSingle(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a single bond or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isSingle", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_10isSingle[] = "\n        Return ``True`` if the bond represents a single bond or ``False`` if\n        not. Bonds with any wildcards will return  ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSingle (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_10isSingle(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_10isSingle(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSingle", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSingle(__pyx_v_self, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 574; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 574; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isSingle", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":581
 *         return self.order[0] == 'S' and len(self.order) == 1
 * 
 *     def isDouble(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a double bond or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isDouble(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isDouble", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isDouble); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 581; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 581; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 581; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 581; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":586
 *         not. Bonds with any wildcards will return  ``False``.
 *         """
 *         return self.order[0] == 'D' and len(self.order) == 1             # <<<<<<<<<<<<<<
 * 
 *     def isTriple(self):
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 586; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 586; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_D, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 586; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {
    goto __pyx_L4_next_and;
  } else {
    __pyx_t_5 = __pyx_t_6;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_and:;
  __pyx_t_1 = __pyx_v_self->order;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 586; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 586; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_6 = ((__pyx_t_7 == 1) != 0);
  __pyx_t_5 = __pyx_t_6;
  __pyx_L3_bool_binop_done:;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":581
 *         return self.order[0] == 'S' and len(self.order) == 1
 * 
 *     def isDouble(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a double bond or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isDouble", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_12isDouble[] = "\n        Return ``True`` if the bond represents a double bond or ``False`` if\n        not. Bonds with any wildcards will return  ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isDouble (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_12isDouble(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_12isDouble(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isDouble", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isDouble(__pyx_v_self, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 581; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 581; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isDouble", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":588
 *         return self.order[0] == 'D' and len(self.order) == 1
 * 
 *     def isTriple(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a triple bond or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isTriple(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isTriple", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isTriple); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":593
 *         not. Bonds with any wildcards will return  ``False``.
 *         """
 *         return self.order[0] == 'T' and len(self.order) == 1             # <<<<<<<<<<<<<<
 * 
 *     def isBenzene(self):
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 593; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 593; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_T, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 593; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {
    goto __pyx_L4_next_and;
  } else {
    __pyx_t_5 = __pyx_t_6;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_and:;
  __pyx_t_1 = __pyx_v_self->order;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 593; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 593; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_6 = ((__pyx_t_7 == 1) != 0);
  __pyx_t_5 = __pyx_t_6;
  __pyx_L3_bool_binop_done:;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":588
 *         return self.order[0] == 'D' and len(self.order) == 1
 * 
 *     def isTriple(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a triple bond or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isTriple", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_14isTriple[] = "\n        Return ``True`` if the bond represents a triple bond or ``False`` if\n        not. Bonds with any wildcards will return  ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isTriple (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_14isTriple(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_14isTriple(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isTriple", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isTriple(__pyx_v_self, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isTriple", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":595
 *         return self.order[0] == 'T' and len(self.order) == 1
 * 
 *     def isBenzene(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a benzene bond or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isBenzene", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 595; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 595; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 595; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 595; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":600
 *         not. Bonds with any wildcards will return  ``False``.
 *         """
 *         return self.order[0] == 'B' and len(self.order) == 1             # <<<<<<<<<<<<<<
 * 
 *     def __changeBond(self, order):
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_B, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {
    goto __pyx_L4_next_and;
  } else {
    __pyx_t_5 = __pyx_t_6;
    goto __pyx_L3_bool_binop_done;
  }
  __pyx_L4_next_and:;
  __pyx_t_1 = __pyx_v_self->order;
  __Pyx_INCREF(__pyx_t_1);
  if (unlikely(__pyx_t_1 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = PyList_GET_SIZE(__pyx_t_1); if (unlikely(__pyx_t_7 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_6 = ((__pyx_t_7 == 1) != 0);
  __pyx_t_5 = __pyx_t_6;
  __pyx_L3_bool_binop_done:;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":595
 *         return self.order[0] == 'T' and len(self.order) == 1
 * 
 *     def isBenzene(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the bond represents a benzene bond or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isBenzene", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_16isBenzene[] = "\n        Return ``True`` if the bond represents a benzene bond or ``False`` if\n        not. Bonds with any wildcards will return  ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isBenzene (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_16isBenzene(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_16isBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isBenzene", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isBenzene(__pyx_v_self, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 595; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 595; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isBenzene", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":602
 *         return self.order[0] == 'B' and len(self.order) == 1
 * 
 *     def __changeBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the bond group as a result of applying a CHANGE_BOND action,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond___changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, short __pyx_v_order, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_newOrder = NULL;
  PyObject *__pyx_v_bond = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  int __pyx_t_8;
  int __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  PyObject *__pyx_t_11 = NULL;
  PyObject *__pyx_t_12 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__changeBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_changeBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 602; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond)) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_3 = __Pyx_PyInt_From_short(__pyx_v_order); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 602; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 602; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 602; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 602; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":608
 *         in bond order, and should be 1 or -1.
 *         """
 *         newOrder = []             # <<<<<<<<<<<<<<
 *         for bond in self.order:
 *             if order == 1:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 608; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_newOrder = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":609
 *         """
 *         newOrder = []
 *         for bond in self.order:             # <<<<<<<<<<<<<<
 *             if order == 1:
 *                 if bond == 'S':         newOrder.append('D')
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 609; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->order; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 609; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 609; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_bond, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":615
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 *             elif order == -1:             # <<<<<<<<<<<<<<
 *                 if bond == 'D':         newOrder.append('S')
 *                 elif bond == 'T':       newOrder.append('D')
 */
    switch (__pyx_v_order) {

      /* "rmgpy/molecule/group.py":610
 *         newOrder = []
 *         for bond in self.order:
 *             if order == 1:             # <<<<<<<<<<<<<<
 *                 if bond == 'S':         newOrder.append('D')
 *                 elif bond == 'D':       newOrder.append('T')
 */
      case 1:

      /* "rmgpy/molecule/group.py":611
 *         for bond in self.order:
 *             if order == 1:
 *                 if bond == 'S':         newOrder.append('D')             # <<<<<<<<<<<<<<
 *                 elif bond == 'D':       newOrder.append('T')
 *                 else:
 */
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_v_bond, __pyx_n_s_S, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 611; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_8) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_newOrder, __pyx_n_s_D); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 611; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L5;
      }

      /* "rmgpy/molecule/group.py":612
 *             if order == 1:
 *                 if bond == 'S':         newOrder.append('D')
 *                 elif bond == 'D':       newOrder.append('T')             # <<<<<<<<<<<<<<
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 */
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_v_bond, __pyx_n_s_D, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 612; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_8) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_newOrder, __pyx_n_s_T); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 612; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L5;
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":614
 *                 elif bond == 'D':       newOrder.append('T')
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))             # <<<<<<<<<<<<<<
 *             elif order == -1:
 *                 if bond == 'D':         newOrder.append('S')
 */
        __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupBond_due_t, __pyx_n_s_format); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_5 = NULL;
        __pyx_t_10 = 0;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
          __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
          if (likely(__pyx_t_5)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
            __Pyx_INCREF(__pyx_t_5);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_3, function);
            __pyx_t_10 = 1;
          }
        }
        __pyx_t_11 = PyTuple_New(2+__pyx_t_10); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        if (__pyx_t_5) {
          PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        }
        __Pyx_INCREF(__pyx_v_bond);
        PyTuple_SET_ITEM(__pyx_t_11, 0+__pyx_t_10, __pyx_v_bond);
        __Pyx_GIVEREF(__pyx_v_bond);
        __Pyx_INCREF(__pyx_v_self->order);
        PyTuple_SET_ITEM(__pyx_t_11, 1+__pyx_t_10, __pyx_v_self->order);
        __Pyx_GIVEREF(__pyx_v_self->order);
        __pyx_t_6 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_11, NULL); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_3 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
          __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_4);
          if (likely(__pyx_t_3)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_3);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_4, function);
          }
        }
        if (!__pyx_t_3) {
          __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_6); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
          __Pyx_GOTREF(__pyx_t_2);
        } else {
          __pyx_t_11 = PyTuple_New(1+1); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_11);
          PyTuple_SET_ITEM(__pyx_t_11, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
          PyTuple_SET_ITEM(__pyx_t_11, 0+1, __pyx_t_6);
          __Pyx_GIVEREF(__pyx_t_6);
          __pyx_t_6 = 0;
          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_11, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
        }
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_Raise(__pyx_t_2, 0, 0, 0);
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_L5:;
      break;

      /* "rmgpy/molecule/group.py":615
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 *             elif order == -1:             # <<<<<<<<<<<<<<
 *                 if bond == 'D':         newOrder.append('S')
 *                 elif bond == 'T':       newOrder.append('D')
 */
      case -1:

      /* "rmgpy/molecule/group.py":616
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 *             elif order == -1:
 *                 if bond == 'D':         newOrder.append('S')             # <<<<<<<<<<<<<<
 *                 elif bond == 'T':       newOrder.append('D')
 *                 else:
 */
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_v_bond, __pyx_n_s_D, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 616; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_8) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_newOrder, __pyx_n_s_S); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 616; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L6;
      }

      /* "rmgpy/molecule/group.py":617
 *             elif order == -1:
 *                 if bond == 'D':         newOrder.append('S')
 *                 elif bond == 'T':       newOrder.append('D')             # <<<<<<<<<<<<<<
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 */
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_v_bond, __pyx_n_s_T, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 617; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_8) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_newOrder, __pyx_n_s_D); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 617; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L6;
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":619
 *                 elif bond == 'T':       newOrder.append('D')
 *                 else:
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))             # <<<<<<<<<<<<<<
 *             else:
 *                 raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 */
        __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupBond_due_t, __pyx_n_s_format); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        __pyx_t_3 = NULL;
        __pyx_t_10 = 0;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_6))) {
          __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_6);
          if (likely(__pyx_t_3)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_6);
            __Pyx_INCREF(__pyx_t_3);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_6, function);
            __pyx_t_10 = 1;
          }
        }
        __pyx_t_5 = PyTuple_New(2+__pyx_t_10); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        if (__pyx_t_3) {
          PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        }
        __Pyx_INCREF(__pyx_v_bond);
        PyTuple_SET_ITEM(__pyx_t_5, 0+__pyx_t_10, __pyx_v_bond);
        __Pyx_GIVEREF(__pyx_v_bond);
        __Pyx_INCREF(__pyx_v_self->order);
        PyTuple_SET_ITEM(__pyx_t_5, 1+__pyx_t_10, __pyx_v_self->order);
        __Pyx_GIVEREF(__pyx_v_self->order);
        __pyx_t_11 = __Pyx_PyObject_Call(__pyx_t_6, __pyx_t_5, NULL); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
        __pyx_t_6 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
          __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_4);
          if (likely(__pyx_t_6)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_6);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_4, function);
          }
        }
        if (!__pyx_t_6) {
          __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_11); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
          __Pyx_GOTREF(__pyx_t_2);
        } else {
          __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_5);
          PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
          PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_t_11);
          __Pyx_GIVEREF(__pyx_t_11);
          __pyx_t_11 = 0;
          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        }
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_Raise(__pyx_t_2, 0, 0, 0);
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_L6:;
      break;
      default:

      /* "rmgpy/molecule/group.py":621
 *                     raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid bond order "{0}" in set {1}".'.format(bond, self.order))
 *             else:
 *                 raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid order "{0}".'.format(order))             # <<<<<<<<<<<<<<
 *         # Set the new bond orders, removing any duplicates
 *         self.order = list(set(newOrder))
 */
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_11 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupBond_due_t_2, __pyx_n_s_format); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_6 = __Pyx_PyInt_From_short(__pyx_v_order); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __pyx_t_3 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_11))) {
        __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_11);
        if (likely(__pyx_t_3)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_11);
          __Pyx_INCREF(__pyx_t_3);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_11, function);
        }
      }
      if (!__pyx_t_3) {
        __pyx_t_5 = __Pyx_PyObject_CallOneArg(__pyx_t_11, __pyx_t_6); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
        __Pyx_GOTREF(__pyx_t_5);
      } else {
        __pyx_t_12 = PyTuple_New(1+1); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_12);
        PyTuple_SET_ITEM(__pyx_t_12, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
        PyTuple_SET_ITEM(__pyx_t_12, 0+1, __pyx_t_6);
        __Pyx_GIVEREF(__pyx_t_6);
        __pyx_t_6 = 0;
        __pyx_t_5 = __Pyx_PyObject_Call(__pyx_t_11, __pyx_t_12, NULL); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      }
      __Pyx_DECREF(__pyx_t_11); __pyx_t_11 = 0;
      __pyx_t_11 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_11 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_11)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_11);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_11) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_5); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_12 = PyTuple_New(1+1); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_12);
        PyTuple_SET_ITEM(__pyx_t_12, 0, __pyx_t_11); __Pyx_GIVEREF(__pyx_t_11); __pyx_t_11 = NULL;
        PyTuple_SET_ITEM(__pyx_t_12, 0+1, __pyx_t_5);
        __Pyx_GIVEREF(__pyx_t_5);
        __pyx_t_5 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_12, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_Raise(__pyx_t_2, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      break;
    }

    /* "rmgpy/molecule/group.py":609
 *         """
 *         newOrder = []
 *         for bond in self.order:             # <<<<<<<<<<<<<<
 *             if order == 1:
 *                 if bond == 'S':         newOrder.append('D')
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":623
 *                 raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
 *         # Set the new bond orders, removing any duplicates
 *         self.order = list(set(newOrder))             # <<<<<<<<<<<<<<
 * 
 *     def applyAction(self, action):
 */
  __pyx_t_1 = PySet_New(__pyx_v_newOrder); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 623; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 623; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)(&PyList_Type))), __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 623; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->order);
  __Pyx_DECREF(__pyx_v_self->order);
  __pyx_v_self->order = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":602
 *         return self.order[0] == 'B' and len(self.order) == 1
 * 
 *     def __changeBond(self, order):             # <<<<<<<<<<<<<<
 *         """
 *         Update the bond group as a result of applying a CHANGE_BOND action,
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_11);
  __Pyx_XDECREF(__pyx_t_12);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_newOrder);
  __Pyx_XDECREF(__pyx_v_bond);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_18__changeBond[] = "\n        Update the bond group as a result of applying a CHANGE_BOND action,\n        where `order` specifies whether the bond is incremented or decremented\n        in bond order, and should be 1 or -1.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond(PyObject *__pyx_v_self, PyObject *__pyx_arg_order) {
  short __pyx_v_order;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__changeBond (wrapper)", 0);
  assert(__pyx_arg_order); {
    __pyx_v_order = __Pyx_PyInt_As_short(__pyx_arg_order); if (unlikely((__pyx_v_order == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 602; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_18__changeBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((short)__pyx_v_order));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_18__changeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, short __pyx_v_order) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__changeBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond___changeBond(__pyx_v_self, __pyx_v_order, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 602; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.__changeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":625
 *         self.order = list(set(newOrder))
 * 
 *     def applyAction(self, action):             # <<<<<<<<<<<<<<
 *         """
 *         Update the bond group as a result of applying `action`, a tuple
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_action, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  short __pyx_t_7;
  PyObject *__pyx_t_8 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("applyAction", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_applyAction); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_action); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_action);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_action);
        __Pyx_GIVEREF(__pyx_v_action);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":632
 *         :ref:`here <reaction-recipe-actions>`.
 *         """
 *         if action[0].upper() == 'CHANGE_BOND':             # <<<<<<<<<<<<<<
 *             self.__changeBond(action[2])
 *         else:
 */
  if (unlikely(__pyx_v_action == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_action, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_upper); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (__pyx_t_2) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_1, __pyx_n_s_CHANGE_BOND, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 632; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":633
 *         """
 *         if action[0].upper() == 'CHANGE_BOND':
 *             self.__changeBond(action[2])             # <<<<<<<<<<<<<<
 *         else:
 *             raise ActionError('Unable to update GroupBond: Invalid action {0}".'.format(action))
 */
    if (unlikely(__pyx_v_action == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 633; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_action, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 633; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_7 = __Pyx_PyInt_As_short(__pyx_t_1); if (unlikely((__pyx_t_7 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 633; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx___changeBond(__pyx_v_self, __pyx_t_7, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 633; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L3;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":635
 *             self.__changeBond(action[2])
 *         else:
 *             raise ActionError('Unable to update GroupBond: Invalid action {0}".'.format(action))             # <<<<<<<<<<<<<<
 * 
 *     def equivalent(self, other):
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_ActionError); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Unable_to_update_GroupBond_Inval, __pyx_n_s_format); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_5);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_5, function);
      }
    }
    if (!__pyx_t_4) {
      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_5, __pyx_v_action); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      __Pyx_INCREF(__pyx_v_action);
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_v_action);
      __Pyx_GIVEREF(__pyx_v_action);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_8, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_8 = PyTuple_New(1+1); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_8);
      PyTuple_SET_ITEM(__pyx_t_8, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_8, 0+1, __pyx_t_2);
      __Pyx_GIVEREF(__pyx_t_2);
      __pyx_t_2 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_8, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_8); __pyx_t_8 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 635; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_L3:;

  /* "rmgpy/molecule/group.py":625
 *         self.order = list(set(newOrder))
 * 
 *     def applyAction(self, action):             # <<<<<<<<<<<<<<
 *         """
 *         Update the bond group as a result of applying `action`, a tuple
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_8);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.applyAction", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_20applyAction[] = "\n        Update the bond group as a result of applying `action`, a tuple\n        containing the name of the reaction recipe action along with any\n        required parameters. The available actions can be found\n        :ref:`here <reaction-recipe-actions>`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction(PyObject *__pyx_v_self, PyObject *__pyx_v_action) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("applyAction (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_action), (&PyList_Type), 1, "action", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_20applyAction(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((PyObject*)__pyx_v_action));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_20applyAction(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_action) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("applyAction", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_applyAction(__pyx_v_self, __pyx_v_action, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.applyAction", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":637
 *             raise ActionError('Unable to update GroupBond: Invalid action {0}".'.format(action))
 * 
 *     def equivalent(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_gb = 0;
  PyObject *__pyx_v_order1 = 0;
  PyObject *__pyx_v_order2 = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("equivalent", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_equivalent); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":644
 *         """
 *         cython.declare(gb=GroupBond)
 *         if not isinstance(other, GroupBond):             # <<<<<<<<<<<<<<
 *             # Let the equivalent method of other handle it
 *             # We expect self to be a Bond object, but can't test for it here
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":648
 *             # We expect self to be a Bond object, but can't test for it here
 *             # because that would create an import cycle
 *             return other.equivalent(self)             # <<<<<<<<<<<<<<
 *         gb = other
 * 
 */
    __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge *)__pyx_v_other->__pyx_vtab)->equivalent(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_self), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 648; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_r = __pyx_t_7;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":649
 *             # because that would create an import cycle
 *             return other.equivalent(self)
 *         gb = other             # <<<<<<<<<<<<<<
 * 
 *         cython.declare(order1=str, order2=str)
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_GroupBond))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 649; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_gb = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":654
 *         # Compare two bond groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:             # <<<<<<<<<<<<<<
 *             for order2 in gb.order:
 *                 if order1 == order2: break
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 654; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->order; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 654; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 654; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(PyString_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 654; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_order1, ((PyObject*)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":655
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:
 *             for order2 in gb.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    if (unlikely(__pyx_v_gb->order == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_gb->order; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      if (!(likely(PyString_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_XDECREF_SET(__pyx_v_order2, ((PyObject*)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":656
 *         for order1 in self.order:
 *             for order2 in gb.order:
 *                 if order1 == order2: break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_v_order1, __pyx_v_order2, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 656; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_6 = (__pyx_t_7 != 0);
      if (__pyx_t_6) {
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":655
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:
 *             for order2 in gb.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":658
 *                 if order1 == order2: break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         for order1 in gb.order:
 *             for order2 in self.order:
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":655
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:
 *             for order2 in gb.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":654
 *         # Compare two bond groups for equivalence
 *         # Each atom type in self must have an equivalent in other (and vice versa)
 *         for order1 in self.order:             # <<<<<<<<<<<<<<
 *             for order2 in gb.order:
 *                 if order1 == order2: break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":659
 *             else:
 *                 return False
 *         for order1 in gb.order:             # <<<<<<<<<<<<<<
 *             for order2 in self.order:
 *                 if order1 == order2: break
 */
  if (unlikely(__pyx_v_gb->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 659; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_gb->order; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 659; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 659; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(PyString_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 659; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_order1, ((PyObject*)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":660
 *                 return False
 *         for order1 in gb.order:
 *             for order2 in self.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    if (unlikely(__pyx_v_self->order == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 660; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_self->order; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 660; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 660; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      if (!(likely(PyString_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 660; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_XDECREF_SET(__pyx_v_order2, ((PyObject*)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":661
 *         for order1 in gb.order:
 *             for order2 in self.order:
 *                 if order1 == order2: break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_v_order1, __pyx_v_order2, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 661; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = (__pyx_t_6 != 0);
      if (__pyx_t_7) {
        goto __pyx_L13_break;
      }

      /* "rmgpy/molecule/group.py":660
 *                 return False
 *         for order1 in gb.order:
 *             for order2 in self.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":663
 *                 if order1 == order2: break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         # Otherwise the two bond groups are equivalent
 *         return True
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":660
 *                 return False
 *         for order1 in gb.order:
 *             for order2 in self.order:             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    __pyx_L13_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":659
 *             else:
 *                 return False
 *         for order1 in gb.order:             # <<<<<<<<<<<<<<
 *             for order2 in self.order:
 *                 if order1 == order2: break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":665
 *                 return False
 *         # Otherwise the two bond groups are equivalent
 *         return True             # <<<<<<<<<<<<<<
 * 
 *     def isSpecificCaseOf(self, other):
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":637
 *             raise ActionError('Unable to update GroupBond: Invalid action {0}".'.format(action))
 * 
 *     def equivalent(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.equivalent", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_gb);
  __Pyx_XDECREF(__pyx_v_order1);
  __Pyx_XDECREF(__pyx_v_order2);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_22equivalent[] = "\n        Returns ``True`` if `other` is equivalent to `self` or ``False`` if not,\n        where `other` can be either an :class:`Bond` or an :class:`GroupBond`\n        object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("equivalent (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Edge, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_22equivalent(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_22equivalent(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("equivalent", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_equivalent(__pyx_v_self, __pyx_v_other, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 637; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.equivalent", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":667
 *         return True
 * 
 *     def isSpecificCaseOf(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is the same as `self` or is a more
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_gb = 0;
  PyObject *__pyx_v_order1 = 0;
  PyObject *__pyx_v_order2 = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSpecificCaseOf", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSpecificCaseOf); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":674
 *         """
 *         cython.declare(gb=GroupBond)
 *         if not isinstance(other, GroupBond):             # <<<<<<<<<<<<<<
 *             # Let the isSpecificCaseOf method of other handle it
 *             # We expect self to be a Bond object, but can't test for it here
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":678
 *             # We expect self to be a Bond object, but can't test for it here
 *             # because that would create an import cycle
 *             return other.isSpecificCaseOf(self)             # <<<<<<<<<<<<<<
 *         gb = other
 * 
 */
    __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge *)__pyx_v_other->__pyx_vtab)->isSpecificCaseOf(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_self), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 678; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_r = __pyx_t_7;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":679
 *             # because that would create an import cycle
 *             return other.isSpecificCaseOf(self)
 *         gb = other             # <<<<<<<<<<<<<<
 * 
 *         cython.declare(order1=str, order2=str)
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_GroupBond))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 679; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_gb = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":684
 *         # Compare two bond groups for equivalence
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match             # <<<<<<<<<<<<<<
 *             for order2 in gb.order: # can match any of these
 *                 if order1 == order2: break
 */
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 684; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->order; __Pyx_INCREF(__pyx_t_1); __pyx_t_8 = 0;
  for (;;) {
    if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_8); __Pyx_INCREF(__pyx_t_2); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 684; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 684; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(PyString_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 684; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_order1, ((PyObject*)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":685
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match
 *             for order2 in gb.order: # can match any of these             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    if (unlikely(__pyx_v_gb->order == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 685; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_2 = __pyx_v_gb->order; __Pyx_INCREF(__pyx_t_2); __pyx_t_9 = 0;
    for (;;) {
      if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_2)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 685; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 685; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      if (!(likely(PyString_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "str", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 685; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_XDECREF_SET(__pyx_v_order2, ((PyObject*)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":686
 *         for order1 in self.order: # all these must match
 *             for order2 in gb.order: # can match any of these
 *                 if order1 == order2: break             # <<<<<<<<<<<<<<
 *             else:
 *                 return False
 */
      __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_v_order1, __pyx_v_order2, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 686; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_6 = (__pyx_t_7 != 0);
      if (__pyx_t_6) {
        goto __pyx_L7_break;
      }

      /* "rmgpy/molecule/group.py":685
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match
 *             for order2 in gb.order: # can match any of these             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":688
 *                 if order1 == order2: break
 *             else:
 *                 return False             # <<<<<<<<<<<<<<
 *         # Otherwise self is in fact a specific case of other
 *         return True
 */
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":685
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match
 *             for order2 in gb.order: # can match any of these             # <<<<<<<<<<<<<<
 *                 if order1 == order2: break
 *             else:
 */
    __pyx_L7_break:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":684
 *         # Compare two bond groups for equivalence
 *         # Each atom type in self must have an equivalent in other
 *         for order1 in self.order: # all these must match             # <<<<<<<<<<<<<<
 *             for order2 in gb.order: # can match any of these
 *                 if order1 == order2: break
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":690
 *                 return False
 *         # Otherwise self is in fact a specific case of other
 *         return True             # <<<<<<<<<<<<<<
 * 
 *     def makeBond(self, molecule, atom1, atom2):
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":667
 *         return True
 * 
 *     def isSpecificCaseOf(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is the same as `self` or is a more
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isSpecificCaseOf", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_gb);
  __Pyx_XDECREF(__pyx_v_order1);
  __Pyx_XDECREF(__pyx_v_order2);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf[] = "\n        Returns ``True`` if `other` is the same as `self` or is a more\n        specific case of `self`. Returns ``False`` if some of `self` is not\n        included in `other` or they are mutually exclusive.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSpecificCaseOf (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Edge, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSpecificCaseOf", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSpecificCaseOf(__pyx_v_self, __pyx_v_other, 1); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.isSpecificCaseOf", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":692
 *         return True
 * 
 *     def makeBond(self, molecule, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Creates a :class: Bond between atom1 and atom2 analogous to self
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_27makeBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_9GroupBond_makeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *__pyx_v_molecule, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom2, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_newBond = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("makeBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_makeBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_27makeBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(3+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_molecule));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_molecule));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_molecule));
      __Pyx_INCREF(((PyObject *)__pyx_v_atom1));
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, ((PyObject *)__pyx_v_atom1));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom1));
      __Pyx_INCREF(((PyObject *)__pyx_v_atom2));
      PyTuple_SET_ITEM(__pyx_t_6, 2+__pyx_t_5, ((PyObject *)__pyx_v_atom2));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom2));
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":702
 * 
 *         """
 *         newBond = mol.Bond(atom1, atom2, order = self.order[0])             # <<<<<<<<<<<<<<
 *         molecule.addBond(newBond)
 * ################################################################################
 */
  __pyx_t_1 = PyTuple_New(2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(((PyObject *)__pyx_v_atom1));
  PyTuple_SET_ITEM(__pyx_t_1, 0, ((PyObject *)__pyx_v_atom1));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_atom1));
  __Pyx_INCREF(((PyObject *)__pyx_v_atom2));
  PyTuple_SET_ITEM(__pyx_t_1, 1, ((PyObject *)__pyx_v_atom2));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_atom2));
  __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (unlikely(__pyx_v_self->order == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_3 = __Pyx_GetItemInt_List(__pyx_v_self->order, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_order, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_8molecule_Bond)), __pyx_t_1, __pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_newBond = __pyx_t_3;
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":703
 *         """
 *         newBond = mol.Bond(atom1, atom2, order = self.order[0])
 *         molecule.addBond(newBond)             # <<<<<<<<<<<<<<
 * ################################################################################
 * 
 */
  if (!(likely(__Pyx_TypeTest(__pyx_v_newBond, __pyx_ptype_5rmgpy_8molecule_8molecule_Bond)))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 703; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Molecule *)__pyx_v_molecule->__pyx_base.__pyx_vtab)->addBond(__pyx_v_molecule, ((struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond *)__pyx_v_newBond), 0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 703; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":692
 *         return True
 * 
 *     def makeBond(self, molecule, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Creates a :class: Bond between atom1 and atom2 analogous to self
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.makeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_newBond);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_27makeBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_26makeBond[] = "\n        Creates a :class: Bond between atom1 and atom2 analogous to self\n\n        The intended input arguments should be class :Atom: not class :GroupAtom:\n        Args:\n            atom1: First :class: Atom the bond connects\n            atom2: Second :class: Atom the bond connects\n\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_27makeBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *__pyx_v_molecule = 0;
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom1 = 0;
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom2 = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("makeBond (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_molecule,&__pyx_n_s_atom1,&__pyx_n_s_atom2,0};
    PyObject* values[3] = {0,0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_molecule)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom1)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("makeBond", 1, 3, 3, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
        case  2:
        if (likely((values[2] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom2)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("makeBond", 1, 3, 3, 2); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "makeBond") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else if (PyTuple_GET_SIZE(__pyx_args) != 3) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
      values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
    }
    __pyx_v_molecule = ((struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *)values[0]);
    __pyx_v_atom1 = ((struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *)values[1]);
    __pyx_v_atom2 = ((struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *)values[2]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("makeBond", 1, 3, 3, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.makeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_molecule), __pyx_ptype_5rmgpy_8molecule_8molecule_Molecule, 1, "molecule", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom1), __pyx_ptype_5rmgpy_8molecule_8molecule_Atom, 1, "atom1", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom2), __pyx_ptype_5rmgpy_8molecule_8molecule_Atom, 1, "atom2", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_26makeBond(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), __pyx_v_molecule, __pyx_v_atom1, __pyx_v_atom2);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_26makeBond(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *__pyx_v_molecule, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *__pyx_v_atom2) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("makeBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_9GroupBond_makeBond(__pyx_v_self, __pyx_v_molecule, __pyx_v_atom1, __pyx_v_atom2, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 692; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.makeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":76
 * cdef class GroupBond(Edge):
 * 
 *     cdef public list order             # <<<<<<<<<<<<<<
 * 
 *     cpdef Edge copy(self)
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order___get__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->order);
  __pyx_r = __pyx_v_self->order;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 76; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->order);
  __Pyx_DECREF(__pyx_v_self->order);
  __pyx_v_self->order = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.GroupBond.order.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_9GroupBond_5order_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->order);
  __Pyx_DECREF(__pyx_v_self->order);
  __pyx_v_self->order = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":715
 *     """
 * 
 *     def __init__(self, atoms=None, multiplicity=None):             # <<<<<<<<<<<<<<
 *         Graph.__init__(self, atoms)
 *         self.multiplicity = multiplicity if multiplicity else []
 */

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_1__init__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_atoms = 0;
  PyObject *__pyx_v_multiplicity = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__init__ (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atoms,&__pyx_n_s_multiplicity,0};
    PyObject* values[2] = {0,0};
    values[0] = ((PyObject *)Py_None);
    values[1] = ((PyObject *)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atoms);
          if (value) { values[0] = value; kw_args--; }
        }
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_multiplicity);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__init__") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 715; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_atoms = values[0];
    __pyx_v_multiplicity = values[1];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("__init__", 0, 0, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 715; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return -1;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group___init__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_atoms, __pyx_v_multiplicity);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group___init__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_atoms, PyObject *__pyx_v_multiplicity) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  Py_ssize_t __pyx_t_4;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__init__", 0);

  /* "rmgpy/molecule/group.py":716
 * 
 *     def __init__(self, atoms=None, multiplicity=None):
 *         Graph.__init__(self, atoms)             # <<<<<<<<<<<<<<
 *         self.multiplicity = multiplicity if multiplicity else []
 *         self.update()
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5graph_Graph)), __pyx_n_s_init); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 716; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  __pyx_t_4 = 0;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
      __pyx_t_4 = 1;
    }
  }
  __pyx_t_5 = PyTuple_New(2+__pyx_t_4); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 716; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  if (__pyx_t_3) {
    PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
  }
  __Pyx_INCREF(((PyObject *)__pyx_v_self));
  PyTuple_SET_ITEM(__pyx_t_5, 0+__pyx_t_4, ((PyObject *)__pyx_v_self));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
  __Pyx_INCREF(__pyx_v_atoms);
  PyTuple_SET_ITEM(__pyx_t_5, 1+__pyx_t_4, __pyx_v_atoms);
  __Pyx_GIVEREF(__pyx_v_atoms);
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 716; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":717
 *     def __init__(self, atoms=None, multiplicity=None):
 *         Graph.__init__(self, atoms)
 *         self.multiplicity = multiplicity if multiplicity else []             # <<<<<<<<<<<<<<
 *         self.update()
 * 
 */
  __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_v_multiplicity); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 717; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_6) {
    if (!(likely(PyList_CheckExact(__pyx_v_multiplicity))||((__pyx_v_multiplicity) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_multiplicity)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 717; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_INCREF(__pyx_v_multiplicity);
    __pyx_t_1 = __pyx_v_multiplicity;
  } else {
    __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 717; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = __pyx_t_2;
    __pyx_t_2 = 0;
  }
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->multiplicity);
  __Pyx_DECREF(__pyx_v_self->multiplicity);
  __pyx_v_self->multiplicity = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":718
 *         Graph.__init__(self, atoms)
 *         self.multiplicity = multiplicity if multiplicity else []
 *         self.update()             # <<<<<<<<<<<<<<
 * 
 *     def __reduce__(self):
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_update); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_5 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_5)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_5);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (__pyx_t_5) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_5); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":715
 *     """
 * 
 *     def __init__(self, atoms=None, multiplicity=None):             # <<<<<<<<<<<<<<
 *         Graph.__init__(self, atoms)
 *         self.multiplicity = multiplicity if multiplicity else []
 */

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":720
 *         self.update()
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_3__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_2__reduce__[] = "\n        A helper function used when pickling an object.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_3__reduce__(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__reduce__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_2__reduce__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_2__reduce__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__reduce__", 0);

  /* "rmgpy/molecule/group.py":724
 *         A helper function used when pickling an object.
 *         """
 *         return (Group, (self.vertices,))             # <<<<<<<<<<<<<<
 * 
 *     def __getAtoms(self): return self.vertices
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 724; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_self->__pyx_base.vertices);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_self->__pyx_base.vertices);
  __Pyx_GIVEREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 724; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)));
  PyTuple_SET_ITEM(__pyx_t_2, 0, ((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)));
  __Pyx_GIVEREF(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)));
  PyTuple_SET_ITEM(__pyx_t_2, 1, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":720
 *         self.update()
 * 
 *     def __reduce__(self):             # <<<<<<<<<<<<<<
 *         """
 *         A helper function used when pickling an object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.__reduce__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":726
 *         return (Group, (self.vertices,))
 * 
 *     def __getAtoms(self): return self.vertices             # <<<<<<<<<<<<<<
 *     def __setAtoms(self, atoms): self.vertices = atoms
 *     atoms = property(__getAtoms, __setAtoms)
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_5__getAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_5__getAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__getAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_4__getAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_4__getAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__getAtoms", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_r = __pyx_v_self->__pyx_base.vertices;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":727
 * 
 *     def __getAtoms(self): return self.vertices
 *     def __setAtoms(self, atoms): self.vertices = atoms             # <<<<<<<<<<<<<<
 *     atoms = property(__getAtoms, __setAtoms)
 * 
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_7__setAtoms(PyObject *__pyx_v_self, PyObject *__pyx_v_atoms); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_7__setAtoms(PyObject *__pyx_v_self, PyObject *__pyx_v_atoms) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__setAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_6__setAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_atoms));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_6__setAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_atoms) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__setAtoms", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_atoms))||((__pyx_v_atoms) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_atoms)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_atoms;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->__pyx_base.vertices);
  __Pyx_DECREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_v_self->__pyx_base.vertices = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.__setAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":730
 *     atoms = property(__getAtoms, __setAtoms)
 * 
 *     def addAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Add an `atom` to the graph. The atom is initialized with no bonds.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_addAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_addAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 730; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_atom)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 730; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 730; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_atom));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_atom));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_atom));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 730; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":734
 *         Add an `atom` to the graph. The atom is initialized with no bonds.
 *         """
 *         return self.addVertex(atom)             # <<<<<<<<<<<<<<
 * 
 *     def addBond(self, bond):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.addVertex(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom), 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 734; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":730
 *     atoms = property(__getAtoms, __setAtoms)
 * 
 *     def addAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Add an `atom` to the graph. The atom is initialized with no bonds.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_8addAtom[] = "\n        Add an `atom` to the graph. The atom is initialized with no bonds.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 730; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_8addAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_8addAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_addAtom(__pyx_v_self, __pyx_v_atom, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 730; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":736
 *         return self.addVertex(atom)
 * 
 *     def addBond(self, bond):             # <<<<<<<<<<<<<<
 *         """
 *         Add a `bond` to the graph as an edge connecting the two atoms `atom1`
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_addBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_addBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 736; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_bond)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 736; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 736; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_bond));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_bond));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_bond));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 736; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":741
 *         and `atom2`.
 *         """
 *         return self.addEdge(bond)             # <<<<<<<<<<<<<<
 * 
 *     def getBonds(self, atom):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.addEdge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_bond), 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 741; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":736
 *         return self.addVertex(atom)
 * 
 *     def addBond(self, bond):             # <<<<<<<<<<<<<<
 *         """
 *         Add a `bond` to the graph as an edge connecting the two atoms `atom1`
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_10addBond[] = "\n        Add a `bond` to the graph as an edge connecting the two atoms `atom1`\n        and `atom2`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addBond (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_bond), __pyx_ptype_5rmgpy_8molecule_5group_GroupBond, 1, "bond", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 736; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_10addBond(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_bond));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_10addBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_addBond(__pyx_v_self, __pyx_v_bond, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 736; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":743
 *         return self.addEdge(bond)
 * 
 *     def getBonds(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Return a list of the bonds involving the specified `atom`.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_getBonds(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBonds", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getBonds); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_atom)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_atom));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_atom));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_atom));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyDict_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "dict", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":747
 *         Return a list of the bonds involving the specified `atom`.
 *         """
 *         return self.getEdges(atom)             # <<<<<<<<<<<<<<
 * 
 *     def getBond(self, atom1, atom2):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.getEdges(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 747; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":743
 *         return self.addEdge(bond)
 * 
 *     def getBonds(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Return a list of the bonds involving the specified `atom`.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBonds", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_12getBonds[] = "\n        Return a list of the bonds involving the specified `atom`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds(PyObject *__pyx_v_self, PyObject *__pyx_v_atom) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getBonds (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12getBonds(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12getBonds(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBonds", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_getBonds(__pyx_v_self, __pyx_v_atom, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBonds", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":749
 *         return self.getEdges(atom)
 * 
 *     def getBond(self, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Returns the bond connecting atoms `atom1` and `atom2`.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_f_5rmgpy_8molecule_5group_5Group_getBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_atom1));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_atom1));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom1));
      __Pyx_INCREF(((PyObject *)__pyx_v_atom2));
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, ((PyObject *)__pyx_v_atom2));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom2));
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5group_GroupBond))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":753
 *         Returns the bond connecting atoms `atom1` and `atom2`.
 *         """
 *         return self.getEdge(atom1, atom2)             # <<<<<<<<<<<<<<
 * 
 *     def hasAtom(self, atom):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  __pyx_t_1 = ((PyObject *)((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.getEdge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom1), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom2), 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (!(likely(((__pyx_t_1) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_1, __pyx_ptype_5rmgpy_8molecule_5group_GroupBond))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":749
 *         return self.getEdges(atom)
 * 
 *     def getBond(self, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Returns the bond connecting atoms `atom1` and `atom2`.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_14getBond[] = "\n        Returns the bond connecting atoms `atom1` and `atom2`.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1 = 0;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2 = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getBond (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atom1,&__pyx_n_s_atom2,0};
    PyObject* values[2] = {0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom1)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom2)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("getBond", 1, 2, 2, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "getBond") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
    }
    __pyx_v_atom1 = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)values[0]);
    __pyx_v_atom2 = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("getBond", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom1), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom1", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom2), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom2", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_14getBond(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_atom1, __pyx_v_atom2);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_14getBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_getBond(__pyx_v_self, __pyx_v_atom1, __pyx_v_atom2, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":755
 *         return self.getEdge(atom1, atom2)
 * 
 *     def hasAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `atom` is an atom in the graph, or ``False`` if
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_hasAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_hasAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 755; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_atom)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 755; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 755; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_atom));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_atom));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_atom));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 755; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 755; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":760
 *         not.
 *         """
 *         return self.hasVertex(atom)             # <<<<<<<<<<<<<<
 * 
 *     def hasBond(self, atom1, atom2):
 */
  __pyx_t_6 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.hasVertex(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom), 0); if (unlikely(__pyx_t_6 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 760; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_t_6;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":755
 *         return self.getEdge(atom1, atom2)
 * 
 *     def hasAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `atom` is an atom in the graph, or ``False`` if
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.hasAtom", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_16hasAtom[] = "\n        Returns ``True`` if `atom` is an atom in the graph, or ``False`` if\n        not.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("hasAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 755; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_16hasAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_16hasAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_hasAtom(__pyx_v_self, __pyx_v_atom, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 755; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.hasAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":762
 *         return self.hasVertex(atom)
 * 
 *     def hasBond(self, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if atoms `atom1` and `atom2` are connected
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_hasBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_hasBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_atom1));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_atom1));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom1));
      __Pyx_INCREF(((PyObject *)__pyx_v_atom2));
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, ((PyObject *)__pyx_v_atom2));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_atom2));
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_7 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_7;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":767
 *         by an bond, or ``False`` if not.
 *         """
 *         return self.hasEdge(atom1, atom2)             # <<<<<<<<<<<<<<
 * 
 *     def removeAtom(self, atom):
 */
  __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.hasEdge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom1), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom2), 0); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 767; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_t_7;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":762
 *         return self.hasVertex(atom)
 * 
 *     def hasBond(self, atom1, atom2):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if atoms `atom1` and `atom2` are connected
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.hasBond", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_18hasBond[] = "\n        Returns ``True`` if atoms `atom1` and `atom2` are connected\n        by an bond, or ``False`` if not.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1 = 0;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2 = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("hasBond (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_atom1,&__pyx_n_s_atom2,0};
    PyObject* values[2] = {0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom1)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_atom2)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("hasBond", 1, 2, 2, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "hasBond") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
    }
    __pyx_v_atom1 = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)values[0]);
    __pyx_v_atom2 = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("hasBond", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.hasBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom1), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom1", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom2), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom2", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_18hasBond(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_atom1, __pyx_v_atom2);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18hasBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom1, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom2) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("hasBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_hasBond(__pyx_v_self, __pyx_v_atom1, __pyx_v_atom2, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 762; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.hasBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":769
 *         return self.hasEdge(atom1, atom2)
 * 
 *     def removeAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Remove `atom` and all bonds associated with it from the graph. Does
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_removeAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("removeAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_removeAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 769; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_atom)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 769; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 769; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_atom));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_atom));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_atom));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 769; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":775
 *         removal.
 *         """
 *         return self.removeVertex(atom)             # <<<<<<<<<<<<<<
 * 
 *     def removeBond(self, bond):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.removeVertex(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *)__pyx_v_atom), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 775; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":769
 *         return self.hasEdge(atom1, atom2)
 * 
 *     def removeAtom(self, atom):             # <<<<<<<<<<<<<<
 *         """
 *         Remove `atom` and all bonds associated with it from the graph. Does
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.removeAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_20removeAtom[] = "\n        Remove `atom` and all bonds associated with it from the graph. Does\n        not remove atoms that no longer have any bonds as a result of this\n        removal.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_atom) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("removeAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_atom), __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, 1, "atom", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 769; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_20removeAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_20removeAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("removeAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_removeAtom(__pyx_v_self, __pyx_v_atom, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 769; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.removeAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":777
 *         return self.removeVertex(atom)
 * 
 *     def removeBond(self, bond):             # <<<<<<<<<<<<<<
 *         """
 *         Remove the bond between atoms `atom1` and `atom2` from the graph.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_removeBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("removeBond", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_removeBond); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_bond)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_bond));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_bond));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_bond));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":783
 *         this removal.
 *         """
 *         return self.removeEdge(bond)             # <<<<<<<<<<<<<<
 * 
 *     def sortAtoms(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.removeEdge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *)__pyx_v_bond), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 783; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":777
 *         return self.removeVertex(atom)
 * 
 *     def removeBond(self, bond):             # <<<<<<<<<<<<<<
 *         """
 *         Remove the bond between atoms `atom1` and `atom2` from the graph.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.removeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_22removeBond[] = "\n        Remove the bond between atoms `atom1` and `atom2` from the graph.\n        Does not remove atoms that no longer have any bonds as a result of\n        this removal.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond(PyObject *__pyx_v_self, PyObject *__pyx_v_bond) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("removeBond (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_bond), __pyx_ptype_5rmgpy_8molecule_5group_GroupBond, 1, "bond", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_22removeBond(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_v_bond));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_22removeBond(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_bond) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("removeBond", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_removeBond(__pyx_v_self, __pyx_v_bond, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 777; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.removeBond", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":785
 *         return self.removeEdge(bond)
 * 
 *     def sortAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Sort the atoms in the graph. This can make certain operations, e.g.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_sortAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("sortAtoms", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_sortAtoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 785; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 785; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 785; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":790
 *         the isomorphism functions, much more efficient.
 *         """
 *         return self.sortVertices()             # <<<<<<<<<<<<<<
 * 
 *     def copy(self, deep=False):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.sortVertices(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 790; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":785
 *         return self.removeEdge(bond)
 * 
 *     def sortAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Sort the atoms in the graph. This can make certain operations, e.g.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.sortAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_24sortAtoms[] = "\n        Sort the atoms in the graph. This can make certain operations, e.g.\n        the isomorphism functions, much more efficient.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("sortAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_24sortAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_24sortAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("sortAtoms", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_sortAtoms(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 785; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.sortAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":792
 *         return self.sortVertices()
 * 
 *     def copy(self, deep=False):             # <<<<<<<<<<<<<<
 *         """
 *         Create a copy of the current graph. If `deep` is ``True``, a deep copy
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_f_5rmgpy_8molecule_5group_5Group_copy(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_copy *__pyx_optional_args) {
  int __pyx_v_deep = ((int)0);
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_other = 0;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_g = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_deep = __pyx_optional_args->deep;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_copy); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_v_deep); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_4 = __pyx_t_1; __pyx_t_5 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_5)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_5);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
        }
      }
      if (!__pyx_t_5) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_6 = PyTuple_New(1+1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
        PyTuple_SET_ITEM(__pyx_t_6, 0+1, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5graph_Graph))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":800
 *         """
 *         other = cython.declare(Group)
 *         g = Graph.copy(self, deep)             # <<<<<<<<<<<<<<
 *         other = Group(g.vertices)
 *         return other
 */
  __pyx_t_7.__pyx_n = 1;
  __pyx_t_7.deep = __pyx_v_deep;
  __pyx_t_1 = ((PyObject *)__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->copy(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 1, &__pyx_t_7)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 800; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_g = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":801
 *         other = cython.declare(Group)
 *         g = Graph.copy(self, deep)
 *         other = Group(g.vertices)             # <<<<<<<<<<<<<<
 *         return other
 * 
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 801; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_g->vertices);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_g->vertices);
  __Pyx_GIVEREF(__pyx_v_g->vertices);
  __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)), __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 801; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":802
 *         g = Graph.copy(self, deep)
 *         other = Group(g.vertices)
 *         return other             # <<<<<<<<<<<<<<
 * 
 *     def update(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  __Pyx_INCREF(((PyObject *)__pyx_v_other));
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_other);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":792
 *         return self.sortVertices()
 * 
 *     def copy(self, deep=False):             # <<<<<<<<<<<<<<
 *         """
 *         Create a copy of the current graph. If `deep` is ``True``, a deep copy
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_other);
  __Pyx_XDECREF((PyObject *)__pyx_v_g);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_26copy[] = "\n        Create a copy of the current graph. If `deep` is ``True``, a deep copy\n        is made: copies of the vertices and edges are used in the new graph.\n        If `deep` is ``False`` or not specified, a shallow copy is made: the\n        original vertices and edges are used in the new graph.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  int __pyx_v_deep;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("copy (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_deep,0};
    PyObject* values[1] = {0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_deep);
          if (value) { values[0] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "copy") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    if (values[0]) {
      __pyx_v_deep = __Pyx_PyObject_IsTrue(values[0]); if (unlikely((__pyx_v_deep == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
    } else {
      __pyx_v_deep = ((int)0);
    }
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("copy", 0, 0, 1, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_26copy(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_deep);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_26copy(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_v_deep) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("copy", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.deep = __pyx_v_deep;
  __pyx_t_1 = ((PyObject *)__pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.copy(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 1, &__pyx_t_2)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 792; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.copy", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":804
 *         return other
 * 
 *     def update(self):             # <<<<<<<<<<<<<<
 * 
 *         self.updateConnectivityValues()
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_29update(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_29update(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("update (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_28update(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_28update(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("update", 0);

  /* "rmgpy/molecule/group.py":806
 *     def update(self):
 * 
 *         self.updateConnectivityValues()             # <<<<<<<<<<<<<<
 *         self.updateFingerprint()
 * 
 */
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.updateConnectivityValues(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 806; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":807
 * 
 *         self.updateConnectivityValues()
 *         self.updateFingerprint()             # <<<<<<<<<<<<<<
 * 
 * 
 */
  __pyx_t_1 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->updateFingerprint(__pyx_v_self, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 807; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":804
 *         return other
 * 
 *     def update(self):             # <<<<<<<<<<<<<<
 * 
 *         self.updateConnectivityValues()
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.update", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":810
 * 
 * 
 *     def merge(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Merge two groups so as to store them in a single
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_f_5rmgpy_8molecule_5group_5Group_merge(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_g = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_molecule = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("merge", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_merge); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 810; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 810; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 810; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 810; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5graph_Graph))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 810; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":816
 *         object is returned.
 *         """
 *         g = Graph.merge(self, other)             # <<<<<<<<<<<<<<
 *         molecule = Group(atoms=g.vertices)
 *         return molecule
 */
  __pyx_t_1 = ((PyObject *)__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->merge(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 816; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_g = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":817
 *         """
 *         g = Graph.merge(self, other)
 *         molecule = Group(atoms=g.vertices)             # <<<<<<<<<<<<<<
 *         return molecule
 * 
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 817; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_atoms, __pyx_v_g->vertices) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 817; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)), __pyx_empty_tuple, __pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 817; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_molecule = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":818
 *         g = Graph.merge(self, other)
 *         molecule = Group(atoms=g.vertices)
 *         return molecule             # <<<<<<<<<<<<<<
 * 
 *     def split(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  __Pyx_INCREF(((PyObject *)__pyx_v_molecule));
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_molecule);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":810
 * 
 * 
 *     def merge(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Merge two groups so as to store them in a single
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.merge", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_g);
  __Pyx_XDECREF((PyObject *)__pyx_v_molecule);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_30merge[] = "\n        Merge two groups so as to store them in a single\n        :class:`Group` object. The merged :class:`Group`\n        object is returned.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("merge (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 810; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_30merge(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_30merge(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("merge", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_merge(__pyx_v_self, __pyx_v_other, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 810; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.merge", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":820
 *         return molecule
 * 
 *     def split(self):             # <<<<<<<<<<<<<<
 *         """
 *         Convert a single :class:`Group` object containing two or more
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_33split(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_split(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_graphs = NULL;
  PyObject *__pyx_v_molecules = NULL;
  PyObject *__pyx_v_g = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_molecule = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("split", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_split); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_33split)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyList_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":825
 *         unconnected groups into separate class:`Group` objects.
 *         """
 *         graphs = Graph.split(self)             # <<<<<<<<<<<<<<
 *         molecules = []
 *         for g in graphs:
 */
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->split(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 825; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_graphs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":826
 *         """
 *         graphs = Graph.split(self)
 *         molecules = []             # <<<<<<<<<<<<<<
 *         for g in graphs:
 *             molecule = Group(atoms=g.vertices)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 826; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_molecules = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":827
 *         graphs = Graph.split(self)
 *         molecules = []
 *         for g in graphs:             # <<<<<<<<<<<<<<
 *             molecule = Group(atoms=g.vertices)
 *             molecules.append(molecule)
 */
  if (unlikely(__pyx_v_graphs == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 827; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_graphs; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 827; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 827; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_g, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":828
 *         molecules = []
 *         for g in graphs:
 *             molecule = Group(atoms=g.vertices)             # <<<<<<<<<<<<<<
 *             molecules.append(molecule)
 *         return molecules
 */
    __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 828; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_g, __pyx_n_s_vertices); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 828; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_atoms, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 828; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)), __pyx_empty_tuple, __pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 828; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_XDECREF_SET(__pyx_v_molecule, ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_3));
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":829
 *         for g in graphs:
 *             molecule = Group(atoms=g.vertices)
 *             molecules.append(molecule)             # <<<<<<<<<<<<<<
 *         return molecules
 * 
 */
    __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_molecules, ((PyObject *)__pyx_v_molecule)); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 829; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

    /* "rmgpy/molecule/group.py":827
 *         graphs = Graph.split(self)
 *         molecules = []
 *         for g in graphs:             # <<<<<<<<<<<<<<
 *             molecule = Group(atoms=g.vertices)
 *             molecules.append(molecule)
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":830
 *             molecule = Group(atoms=g.vertices)
 *             molecules.append(molecule)
 *         return molecules             # <<<<<<<<<<<<<<
 * 
 *     def clearLabeledAtoms(self):
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_molecules);
  __pyx_r = __pyx_v_molecules;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":820
 *         return molecule
 * 
 *     def split(self):             # <<<<<<<<<<<<<<
 *         """
 *         Convert a single :class:`Group` object containing two or more
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.split", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_graphs);
  __Pyx_XDECREF(__pyx_v_molecules);
  __Pyx_XDECREF(__pyx_v_g);
  __Pyx_XDECREF((PyObject *)__pyx_v_molecule);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_33split(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_32split[] = "\n        Convert a single :class:`Group` object containing two or more\n        unconnected groups into separate class:`Group` objects.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_33split(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("split (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_32split(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_32split(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("split", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_split(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 820; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.split", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":832
 *         return molecules
 * 
 *     def clearLabeledAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Remove the labels from all atoms in the molecular group.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_clearLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("clearLabeledAtoms", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_clearLabeledAtoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 832; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 832; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 832; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":836
 *         Remove the labels from all atoms in the molecular group.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             atom.label = ''
 * 
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 836; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 836; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 836; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":837
 *         """
 *         for atom in self.vertices:
 *             atom.label = ''             # <<<<<<<<<<<<<<
 * 
 *     def containsLabeledAtom(self, label):
 */
    if (__Pyx_PyObject_SetAttrStr(__pyx_v_atom, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

    /* "rmgpy/molecule/group.py":836
 *         Remove the labels from all atoms in the molecular group.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             atom.label = ''
 * 
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":832
 *         return molecules
 * 
 *     def clearLabeledAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Remove the labels from all atoms in the molecular group.
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.clearLabeledAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms[] = "\n        Remove the labels from all atoms in the molecular group.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("clearLabeledAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("clearLabeledAtoms", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_clearLabeledAtoms(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 832; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.clearLabeledAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":839
 *             atom.label = ''
 * 
 *     def containsLabeledAtom(self, label):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the group contains an atom with the label
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_containsLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atom = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("containsLabeledAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_containsLabeledAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_label);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_label);
        __Pyx_GIVEREF(__pyx_v_label);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":844
 *         `label` and ``False`` otherwise.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label == label: return True
 *         return False
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 844; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_7 = 0;
  for (;;) {
    if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 844; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 844; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":845
 *         """
 *         for atom in self.vertices:
 *             if atom.label == label: return True             # <<<<<<<<<<<<<<
 *         return False
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 845; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_v_label, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 845; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_6) {
      __pyx_r = 1;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":844
 *         `label` and ``False`` otherwise.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label == label: return True
 *         return False
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":846
 *         for atom in self.vertices:
 *             if atom.label == label: return True
 *         return False             # <<<<<<<<<<<<<<
 * 
 *     def getLabeledAtom(self, label):
 */
  __pyx_r = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":839
 *             atom.label = ''
 * 
 *     def containsLabeledAtom(self, label):             # <<<<<<<<<<<<<<
 *         """
 *         Return ``True`` if the group contains an atom with the label
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.containsLabeledAtom", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom[] = "\n        Return ``True`` if the group contains an atom with the label\n        `label` and ``False`` otherwise.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("containsLabeledAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_label), (&PyString_Type), 1, "label", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject*)__pyx_v_label));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("containsLabeledAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_containsLabeledAtom(__pyx_v_self, __pyx_v_label, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.containsLabeledAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":848
 *         return False
 * 
 *     def getLabeledAtom(self, label):             # <<<<<<<<<<<<<<
 *         """
 *         Return the atom in the group that is labeled with the given `label`.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_atom = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  Py_ssize_t __pyx_t_6;
  int __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getLabeledAtom", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getLabeledAtom); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 848; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 848; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 848; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_label);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_label);
        __Pyx_GIVEREF(__pyx_v_label);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 848; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 848; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":853
 *         Raises :class:`ValueError` if no atom in the group has that label.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label == label: return atom
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 853; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 853; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 853; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":854
 *         """
 *         for atom in self.vertices:
 *             if atom.label == label: return atom             # <<<<<<<<<<<<<<
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 854; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_v_label, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 854; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_7) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      if (!(likely(((__pyx_v_atom) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_atom, __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 854; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_INCREF(__pyx_v_atom);
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_v_atom);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":853
 *         Raises :class:`ValueError` if no atom in the group has that label.
 *         """
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label == label: return atom
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":855
 *         for atom in self.vertices:
 *             if atom.label == label: return atom
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))             # <<<<<<<<<<<<<<
 * 
 *     def getLabeledAtoms(self):
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_No_atom_in_the_functional_group, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_v_label); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(__pyx_v_label);
    PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_label);
    __Pyx_GIVEREF(__pyx_v_label);
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_5, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_ValueError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_Raise(__pyx_t_1, 0, 0, 0);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":848
 *         return False
 * 
 *     def getLabeledAtom(self, label):             # <<<<<<<<<<<<<<
 *         """
 *         Return the atom in the group that is labeled with the given `label`.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getLabeledAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_38getLabeledAtom[] = "\n        Return the atom in the group that is labeled with the given `label`.\n        Raises :class:`ValueError` if no atom in the group has that label.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom(PyObject *__pyx_v_self, PyObject *__pyx_v_label) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getLabeledAtom (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_label), (&PyString_Type), 1, "label", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 848; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_38getLabeledAtom(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject*)__pyx_v_label));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_38getLabeledAtom(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getLabeledAtom", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtom(__pyx_v_self, __pyx_v_label, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 848; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getLabeledAtom", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":857
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 * 
 *     def getLabeledAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return the labeled atoms as a ``dict`` with the keys being the labels
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_labeled = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  int __pyx_t_6;
  int __pyx_t_7;
  int __pyx_t_8;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getLabeledAtoms", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_getLabeledAtoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyDict_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "dict", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":863
 *         same label, the value is converted to a list of these atoms.
 *         """
 *         labeled = {}             # <<<<<<<<<<<<<<
 *         for atom in self.vertices:
 *             if atom.label != '':
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 863; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_labeled = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":864
 *         """
 *         labeled = {}
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label != '':
 *                 if atom.label in labeled:
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 864; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 864; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 864; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":865
 *         labeled = {}
 *         for atom in self.vertices:
 *             if atom.label != '':             # <<<<<<<<<<<<<<
 *                 if atom.label in labeled:
 *                     if isinstance(labeled[atom.label],list):
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 865; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_kp_s_, Py_NE)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 865; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_6) {

      /* "rmgpy/molecule/group.py":866
 *         for atom in self.vertices:
 *             if atom.label != '':
 *                 if atom.label in labeled:             # <<<<<<<<<<<<<<
 *                     if isinstance(labeled[atom.label],list):
 *                         labeled[atom.label].append(atom)
 */
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_6 = (__Pyx_PyDict_Contains(__pyx_t_2, __pyx_v_labeled, Py_EQ)); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 866; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_7 = (__pyx_t_6 != 0);
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":867
 *             if atom.label != '':
 *                 if atom.label in labeled:
 *                     if isinstance(labeled[atom.label],list):             # <<<<<<<<<<<<<<
 *                         labeled[atom.label].append(atom)
 *                     else:
 */
        __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 867; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __pyx_t_3 = __Pyx_PyDict_GetItem(__pyx_v_labeled, __pyx_t_2); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 867; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_3);
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        __pyx_t_7 = PyList_Check(__pyx_t_3); 
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_6 = (__pyx_t_7 != 0);
        if (__pyx_t_6) {

          /* "rmgpy/molecule/group.py":868
 *                 if atom.label in labeled:
 *                     if isinstance(labeled[atom.label],list):
 *                         labeled[atom.label].append(atom)             # <<<<<<<<<<<<<<
 *                     else:
 *                         labeled[atom.label] = [labeled[atom.label]]
 */
          __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 868; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          __pyx_t_2 = __Pyx_PyDict_GetItem(__pyx_v_labeled, __pyx_t_3); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 868; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_2);
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_t_8 = __Pyx_PyObject_Append(__pyx_t_2, __pyx_v_atom); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 868; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L7;
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":870
 *                         labeled[atom.label].append(atom)
 *                     else:
 *                         labeled[atom.label] = [labeled[atom.label]]             # <<<<<<<<<<<<<<
 *                         labeled[atom.label].append(atom)
 *                 else:
 */
          __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 870; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __pyx_t_3 = __Pyx_PyDict_GetItem(__pyx_v_labeled, __pyx_t_2); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 870; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 870; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          PyList_SET_ITEM(__pyx_t_2, 0, __pyx_t_3);
          __Pyx_GIVEREF(__pyx_t_3);
          __pyx_t_3 = 0;
          __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 870; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          if (unlikely(PyDict_SetItem(__pyx_v_labeled, __pyx_t_3, __pyx_t_2) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 870; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

          /* "rmgpy/molecule/group.py":871
 *                     else:
 *                         labeled[atom.label] = [labeled[atom.label]]
 *                         labeled[atom.label].append(atom)             # <<<<<<<<<<<<<<
 *                 else:
 *                     labeled[atom.label] = atom
 */
          __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 871; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __pyx_t_3 = __Pyx_PyDict_GetItem(__pyx_v_labeled, __pyx_t_2); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 871; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          __pyx_t_8 = __Pyx_PyObject_Append(__pyx_t_3, __pyx_v_atom); if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 871; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        }
        __pyx_L7:;
        goto __pyx_L6;
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":873
 *                         labeled[atom.label].append(atom)
 *                 else:
 *                     labeled[atom.label] = atom             # <<<<<<<<<<<<<<
 *         return labeled
 * 
 */
        __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_label); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 873; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        if (unlikely(PyDict_SetItem(__pyx_v_labeled, __pyx_t_3, __pyx_v_atom) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 873; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      }
      __pyx_L6:;
      goto __pyx_L5;
    }
    __pyx_L5:;

    /* "rmgpy/molecule/group.py":864
 *         """
 *         labeled = {}
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if atom.label != '':
 *                 if atom.label in labeled:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":874
 *                 else:
 *                     labeled[atom.label] = atom
 *         return labeled             # <<<<<<<<<<<<<<
 * 
 *     def fromAdjacencyList(self, adjlist):
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_labeled);
  __pyx_r = __pyx_v_labeled;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":857
 *         raise ValueError('No atom in the functional group has the label "{0}".'.format(label))
 * 
 *     def getLabeledAtoms(self):             # <<<<<<<<<<<<<<
 *         """
 *         Return the labeled atoms as a ``dict`` with the keys being the labels
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getLabeledAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_labeled);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms[] = "\n        Return the labeled atoms as a ``dict`` with the keys being the labels\n        and the values the atoms themselves. If two or more atoms have the\n        same label, the value is converted to a list of these atoms.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("getLabeledAtoms (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("getLabeledAtoms", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtoms(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 857; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.getLabeledAtoms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":876
 *         return labeled
 * 
 *     def fromAdjacencyList(self, adjlist):             # <<<<<<<<<<<<<<
 *         """
 *         Convert a string adjacency list `adjlist` to a molecular structure.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_v_adjlist); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_fromAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_adjlist, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_fromAdjacencyList = NULL;
  PyObject *__pyx_v_multiplicity = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *(*__pyx_t_6)(PyObject *);
  int __pyx_t_7;
  int __pyx_t_8;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("fromAdjacencyList", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_fromAdjacencyList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 876; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_adjlist); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 876; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 876; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_adjlist);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_adjlist);
        __Pyx_GIVEREF(__pyx_v_adjlist);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 876; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":882
 *         ``False``.
 *         """
 *         from .adjlist import fromAdjacencyList             # <<<<<<<<<<<<<<
 *         self.vertices, multiplicity = fromAdjacencyList(adjlist, group=True)
 *         if multiplicity is not None:
 */
  __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 882; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_n_s_fromAdjacencyList);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_n_s_fromAdjacencyList);
  __Pyx_GIVEREF(__pyx_n_s_fromAdjacencyList);
  __pyx_t_2 = __Pyx_Import(__pyx_n_s_adjlist, __pyx_t_1, 1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 882; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_ImportFrom(__pyx_t_2, __pyx_n_s_fromAdjacencyList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 882; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_fromAdjacencyList = __pyx_t_1;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":883
 *         """
 *         from .adjlist import fromAdjacencyList
 *         self.vertices, multiplicity = fromAdjacencyList(adjlist, group=True)             # <<<<<<<<<<<<<<
 *         if multiplicity is not None:
 *             self.multiplicity = multiplicity
 */
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_v_adjlist);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_adjlist);
  __Pyx_GIVEREF(__pyx_v_adjlist);
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_group, Py_True) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = __Pyx_PyObject_Call(__pyx_v_fromAdjacencyList, __pyx_t_2, __pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if ((likely(PyTuple_CheckExact(__pyx_t_3))) || (PyList_CheckExact(__pyx_t_3))) {
    PyObject* sequence = __pyx_t_3;
    #if CYTHON_COMPILING_IN_CPYTHON
    Py_ssize_t size = Py_SIZE(sequence);
    #else
    Py_ssize_t size = PySequence_Size(sequence);
    #endif
    if (unlikely(size != 2)) {
      if (size > 2) __Pyx_RaiseTooManyValuesError(2);
      else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    #if CYTHON_COMPILING_IN_CPYTHON
    if (likely(PyTuple_CheckExact(sequence))) {
      __pyx_t_1 = PyTuple_GET_ITEM(sequence, 0); 
      __pyx_t_2 = PyTuple_GET_ITEM(sequence, 1); 
    } else {
      __pyx_t_1 = PyList_GET_ITEM(sequence, 0); 
      __pyx_t_2 = PyList_GET_ITEM(sequence, 1); 
    }
    __Pyx_INCREF(__pyx_t_1);
    __Pyx_INCREF(__pyx_t_2);
    #else
    __pyx_t_1 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_2 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    #endif
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  } else {
    Py_ssize_t index = -1;
    __pyx_t_5 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_6 = Py_TYPE(__pyx_t_5)->tp_iternext;
    index = 0; __pyx_t_1 = __pyx_t_6(__pyx_t_5); if (unlikely(!__pyx_t_1)) goto __pyx_L3_unpacking_failed;
    __Pyx_GOTREF(__pyx_t_1);
    index = 1; __pyx_t_2 = __pyx_t_6(__pyx_t_5); if (unlikely(!__pyx_t_2)) goto __pyx_L3_unpacking_failed;
    __Pyx_GOTREF(__pyx_t_2);
    if (__Pyx_IternextUnpackEndCheck(__pyx_t_6(__pyx_t_5), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_6 = NULL;
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    goto __pyx_L4_unpacking_done;
    __pyx_L3_unpacking_failed:;
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_6 = NULL;
    if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_L4_unpacking_done:;
  }
  if (!(likely(PyList_CheckExact(__pyx_t_1))||((__pyx_t_1) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_t_1)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 883; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->__pyx_base.vertices);
  __Pyx_DECREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_v_self->__pyx_base.vertices = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  __pyx_v_multiplicity = __pyx_t_2;
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":884
 *         from .adjlist import fromAdjacencyList
 *         self.vertices, multiplicity = fromAdjacencyList(adjlist, group=True)
 *         if multiplicity is not None:             # <<<<<<<<<<<<<<
 *             self.multiplicity = multiplicity
 *         self.update()
 */
  __pyx_t_7 = (__pyx_v_multiplicity != Py_None);
  __pyx_t_8 = (__pyx_t_7 != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":885
 *         self.vertices, multiplicity = fromAdjacencyList(adjlist, group=True)
 *         if multiplicity is not None:
 *             self.multiplicity = multiplicity             # <<<<<<<<<<<<<<
 *         self.update()
 *         return self
 */
    if (!(likely(PyList_CheckExact(__pyx_v_multiplicity))||((__pyx_v_multiplicity) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_multiplicity)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 885; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_3 = __pyx_v_multiplicity;
    __Pyx_INCREF(__pyx_t_3);
    __Pyx_GIVEREF(__pyx_t_3);
    __Pyx_GOTREF(__pyx_v_self->multiplicity);
    __Pyx_DECREF(__pyx_v_self->multiplicity);
    __pyx_v_self->multiplicity = ((PyObject*)__pyx_t_3);
    __pyx_t_3 = 0;
    goto __pyx_L5;
  }
  __pyx_L5:;

  /* "rmgpy/molecule/group.py":886
 *         if multiplicity is not None:
 *             self.multiplicity = multiplicity
 *         self.update()             # <<<<<<<<<<<<<<
 *         return self
 * 
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_update); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 886; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_1 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_1 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_1)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_1);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (__pyx_t_1) {
    __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 886; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  } else {
    __pyx_t_3 = __Pyx_PyObject_CallNoArg(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 886; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":887
 *             self.multiplicity = multiplicity
 *         self.update()
 *         return self             # <<<<<<<<<<<<<<
 * 
 *     def toAdjacencyList(self, label=''):
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(((PyObject *)__pyx_v_self));
  __pyx_r = ((PyObject *)__pyx_v_self);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":876
 *         return labeled
 * 
 *     def fromAdjacencyList(self, adjlist):             # <<<<<<<<<<<<<<
 *         """
 *         Convert a string adjacency list `adjlist` to a molecular structure.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.fromAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_fromAdjacencyList);
  __Pyx_XDECREF(__pyx_v_multiplicity);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_v_adjlist); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList[] = "\n        Convert a string adjacency list `adjlist` to a molecular structure.\n        Skips the first line (assuming it's a label) unless `withLabel` is\n        ``False``.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_v_adjlist) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("fromAdjacencyList (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_adjlist), (&PyString_Type), 1, "adjlist", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 876; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject*)__pyx_v_adjlist));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_adjlist) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("fromAdjacencyList", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_fromAdjacencyList(__pyx_v_self, __pyx_v_adjlist, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 876; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.fromAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":889
 *         return self
 * 
 *     def toAdjacencyList(self, label=''):             # <<<<<<<<<<<<<<
 *         """
 *         Convert the molecular structure to a string adjacency list.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_toAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList *__pyx_optional_args) {
  PyObject *__pyx_v_label = ((PyObject*)__pyx_kp_s_);
  PyObject *__pyx_v_toAdjacencyList = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("toAdjacencyList", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_label = __pyx_optional_args->label;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_toAdjacencyList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 889; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 889; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 889; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(__pyx_v_label);
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_label);
        __Pyx_GIVEREF(__pyx_v_label);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 889; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":893
 *         Convert the molecular structure to a string adjacency list.
 *         """
 *         from .adjlist import toAdjacencyList             # <<<<<<<<<<<<<<
 *         return toAdjacencyList(self.vertices, multiplicity=self.multiplicity, label='', group=True)
 * 
 */
  __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 893; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_n_s_toAdjacencyList);
  PyList_SET_ITEM(__pyx_t_1, 0, __pyx_n_s_toAdjacencyList);
  __Pyx_GIVEREF(__pyx_n_s_toAdjacencyList);
  __pyx_t_2 = __Pyx_Import(__pyx_n_s_adjlist, __pyx_t_1, 1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 893; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_ImportFrom(__pyx_t_2, __pyx_n_s_toAdjacencyList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 893; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_toAdjacencyList = __pyx_t_1;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":894
 *         """
 *         from .adjlist import toAdjacencyList
 *         return toAdjacencyList(self.vertices, multiplicity=self.multiplicity, label='', group=True)             # <<<<<<<<<<<<<<
 * 
 * 
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 894; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_v_self->__pyx_base.vertices);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_self->__pyx_base.vertices);
  __Pyx_GIVEREF(__pyx_v_self->__pyx_base.vertices);
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 894; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_multiplicity, __pyx_v_self->multiplicity) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 894; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 894; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyDict_SetItem(__pyx_t_1, __pyx_n_s_group, Py_True) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 894; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = __Pyx_PyObject_Call(__pyx_v_toAdjacencyList, __pyx_t_2, __pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 894; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_r = __pyx_t_3;
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":889
 *         return self
 * 
 *     def toAdjacencyList(self, label=''):             # <<<<<<<<<<<<<<
 *         """
 *         Convert the molecular structure to a string adjacency list.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.toAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_toAdjacencyList);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_44toAdjacencyList[] = "\n        Convert the molecular structure to a string adjacency list.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_label = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("toAdjacencyList (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_label,0};
    PyObject* values[1] = {0};
    values[0] = ((PyObject*)__pyx_kp_s_);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_label);
          if (value) { values[0] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "toAdjacencyList") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 889; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_label = ((PyObject*)values[0]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("toAdjacencyList", 0, 0, 1, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 889; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.toAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_label), (&PyString_Type), 1, "label", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 889; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_44toAdjacencyList(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_label);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_44toAdjacencyList(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_label) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("toAdjacencyList", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.label = __pyx_v_label;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->toAdjacencyList(__pyx_v_self, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 889; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.toAdjacencyList", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":897
 * 
 * 
 *     def updateFingerprint(self):             # <<<<<<<<<<<<<<
 *         """
 *         Update the molecular fingerprint used to accelerate the subgraph
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_updateFingerprint(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_atom = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_atomType = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_carbon = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_nitrogen = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_oxygen = 0;
  struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *__pyx_v_sulfur = 0;
  int __pyx_v_isCarbon;
  int __pyx_v_isNitrogen;
  int __pyx_v_isOxygen;
  int __pyx_v_isSulfur;
  int __pyx_v_radical;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_t_7;
  int __pyx_t_8;
  int __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("updateFingerprint", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_updateFingerprint); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 897; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 897; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 897; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_r = __pyx_t_2;
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":906
 *         cython.declare(isCarbon=cython.bint, isNitrogen=cython.bint, isOxygen=cython.bint, isSulfur=cython.bint, radical=cython.int)
 * 
 *         carbon   = atomTypes['C']             # <<<<<<<<<<<<<<
 *         nitrogen = atomTypes['N']
 *         oxygen   = atomTypes['O']
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 906; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_C); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 906; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 906; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_carbon = ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":907
 * 
 *         carbon   = atomTypes['C']
 *         nitrogen = atomTypes['N']             # <<<<<<<<<<<<<<
 *         oxygen   = atomTypes['O']
 *         sulfur   = atomTypes['S']
 */
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 907; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_1 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_N); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 907; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  if (!(likely(((__pyx_t_1) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_1, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 907; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_nitrogen = ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":908
 *         carbon   = atomTypes['C']
 *         nitrogen = atomTypes['N']
 *         oxygen   = atomTypes['O']             # <<<<<<<<<<<<<<
 *         sulfur   = atomTypes['S']
 * 
 */
  __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 908; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_O); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 908; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 908; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_oxygen = ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":909
 *         nitrogen = atomTypes['N']
 *         oxygen   = atomTypes['O']
 *         sulfur   = atomTypes['S']             # <<<<<<<<<<<<<<
 * 
 *         self.carbonCount   = 0
 */
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 909; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_1 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_S); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 909; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  if (!(likely(((__pyx_t_1) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_1, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 909; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_sulfur = ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":911
 *         sulfur   = atomTypes['S']
 * 
 *         self.carbonCount   = 0             # <<<<<<<<<<<<<<
 *         self.nitrogenCount = 0
 *         self.oxygenCount   = 0
 */
  __pyx_v_self->carbonCount = 0;

  /* "rmgpy/molecule/group.py":912
 * 
 *         self.carbonCount   = 0
 *         self.nitrogenCount = 0             # <<<<<<<<<<<<<<
 *         self.oxygenCount   = 0
 *         self.sulfurCount   = 0
 */
  __pyx_v_self->nitrogenCount = 0;

  /* "rmgpy/molecule/group.py":913
 *         self.carbonCount   = 0
 *         self.nitrogenCount = 0
 *         self.oxygenCount   = 0             # <<<<<<<<<<<<<<
 *         self.sulfurCount   = 0
 *         self.radicalCount  = 0
 */
  __pyx_v_self->oxygenCount = 0;

  /* "rmgpy/molecule/group.py":914
 *         self.nitrogenCount = 0
 *         self.oxygenCount   = 0
 *         self.sulfurCount   = 0             # <<<<<<<<<<<<<<
 *         self.radicalCount  = 0
 *         for atom in self.vertices:
 */
  __pyx_v_self->sulfurCount = 0;

  /* "rmgpy/molecule/group.py":915
 *         self.oxygenCount   = 0
 *         self.sulfurCount   = 0
 *         self.radicalCount  = 0             # <<<<<<<<<<<<<<
 *         for atom in self.vertices:
 *             if len(atom.atomType) == 1:
 */
  __pyx_v_self->radicalCount = 0;

  /* "rmgpy/molecule/group.py":916
 *         self.sulfurCount   = 0
 *         self.radicalCount  = 0
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if len(atom.atomType) == 1:
 *                 atomType   = atom.atomType[0]
 */
  if (unlikely(__pyx_v_self->__pyx_base.vertices == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 916; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_self->__pyx_base.vertices; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
  for (;;) {
    if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 916; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 916; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 916; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_atom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_2));
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":917
 *         self.radicalCount  = 0
 *         for atom in self.vertices:
 *             if len(atom.atomType) == 1:             # <<<<<<<<<<<<<<
 *                 atomType   = atom.atomType[0]
 *                 isCarbon   = atomType.equivalent(carbon)
 */
    __pyx_t_2 = __pyx_v_atom->atomType;
    __Pyx_INCREF(__pyx_t_2);
    if (unlikely(__pyx_t_2 == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 917; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_6 = PyList_GET_SIZE(__pyx_t_2); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 917; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_7 = ((__pyx_t_6 == 1) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":918
 *         for atom in self.vertices:
 *             if len(atom.atomType) == 1:
 *                 atomType   = atom.atomType[0]             # <<<<<<<<<<<<<<
 *                 isCarbon   = atomType.equivalent(carbon)
 *                 isNitrogen = atomType.equivalent(nitrogen)
 */
      if (unlikely(__pyx_v_atom->atomType == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 918; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_atom->atomType, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 918; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 918; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_XDECREF_SET(__pyx_v_atomType, ((struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_t_2));
      __pyx_t_2 = 0;

      /* "rmgpy/molecule/group.py":919
 *             if len(atom.atomType) == 1:
 *                 atomType   = atom.atomType[0]
 *                 isCarbon   = atomType.equivalent(carbon)             # <<<<<<<<<<<<<<
 *                 isNitrogen = atomType.equivalent(nitrogen)
 *                 isOxygen   = atomType.equivalent(oxygen)
 */
      __pyx_v_isCarbon = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType->__pyx_vtab)->equivalent(__pyx_v_atomType, __pyx_v_carbon, 0);

      /* "rmgpy/molecule/group.py":920
 *                 atomType   = atom.atomType[0]
 *                 isCarbon   = atomType.equivalent(carbon)
 *                 isNitrogen = atomType.equivalent(nitrogen)             # <<<<<<<<<<<<<<
 *                 isOxygen   = atomType.equivalent(oxygen)
 *                 isSulfur   = atomType.equivalent(sulfur)
 */
      __pyx_v_isNitrogen = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType->__pyx_vtab)->equivalent(__pyx_v_atomType, __pyx_v_nitrogen, 0);

      /* "rmgpy/molecule/group.py":921
 *                 isCarbon   = atomType.equivalent(carbon)
 *                 isNitrogen = atomType.equivalent(nitrogen)
 *                 isOxygen   = atomType.equivalent(oxygen)             # <<<<<<<<<<<<<<
 *                 isSulfur   = atomType.equivalent(sulfur)
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:
 */
      __pyx_v_isOxygen = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType->__pyx_vtab)->equivalent(__pyx_v_atomType, __pyx_v_oxygen, 0);

      /* "rmgpy/molecule/group.py":922
 *                 isNitrogen = atomType.equivalent(nitrogen)
 *                 isOxygen   = atomType.equivalent(oxygen)
 *                 isSulfur   = atomType.equivalent(sulfur)             # <<<<<<<<<<<<<<
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:
 *                     self.carbonCount += 1
 */
      __pyx_v_isSulfur = ((struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType *)__pyx_v_atomType->__pyx_vtab)->equivalent(__pyx_v_atomType, __pyx_v_sulfur, 0);

      /* "rmgpy/molecule/group.py":923
 *                 isOxygen   = atomType.equivalent(oxygen)
 *                 isSulfur   = atomType.equivalent(sulfur)
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:             # <<<<<<<<<<<<<<
 *                     self.carbonCount += 1
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:
 */
      __pyx_t_8 = (__pyx_v_isCarbon != 0);
      if (__pyx_t_8) {
        goto __pyx_L8_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L7_bool_binop_done;
      }
      __pyx_L8_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isNitrogen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L9_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L7_bool_binop_done;
      }
      __pyx_L9_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isOxygen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L10_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L7_bool_binop_done;
      }
      __pyx_L10_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isSulfur != 0)) != 0);
      __pyx_t_7 = __pyx_t_8;
      __pyx_L7_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":924
 *                 isSulfur   = atomType.equivalent(sulfur)
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:
 *                     self.carbonCount += 1             # <<<<<<<<<<<<<<
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:
 *                     self.nitrogenCount += 1
 */
        __pyx_v_self->carbonCount = (__pyx_v_self->carbonCount + 1);
        goto __pyx_L6;
      }

      /* "rmgpy/molecule/group.py":925
 *                 if isCarbon and not isNitrogen and not isOxygen and not isSulfur:
 *                     self.carbonCount += 1
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:             # <<<<<<<<<<<<<<
 *                     self.nitrogenCount += 1
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:
 */
      __pyx_t_8 = (__pyx_v_isNitrogen != 0);
      if (__pyx_t_8) {
        goto __pyx_L12_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L11_bool_binop_done;
      }
      __pyx_L12_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isCarbon != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L13_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L11_bool_binop_done;
      }
      __pyx_L13_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isOxygen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L14_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L11_bool_binop_done;
      }
      __pyx_L14_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isSulfur != 0)) != 0);
      __pyx_t_7 = __pyx_t_8;
      __pyx_L11_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":926
 *                     self.carbonCount += 1
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:
 *                     self.nitrogenCount += 1             # <<<<<<<<<<<<<<
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:
 *                     self.oxygenCount += 1
 */
        __pyx_v_self->nitrogenCount = (__pyx_v_self->nitrogenCount + 1);
        goto __pyx_L6;
      }

      /* "rmgpy/molecule/group.py":927
 *                 elif isNitrogen and not isCarbon and not isOxygen and not isSulfur:
 *                     self.nitrogenCount += 1
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:             # <<<<<<<<<<<<<<
 *                     self.oxygenCount += 1
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:
 */
      __pyx_t_8 = (__pyx_v_isOxygen != 0);
      if (__pyx_t_8) {
        goto __pyx_L16_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L15_bool_binop_done;
      }
      __pyx_L16_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isCarbon != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L17_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L15_bool_binop_done;
      }
      __pyx_L17_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isNitrogen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L18_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L15_bool_binop_done;
      }
      __pyx_L18_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isSulfur != 0)) != 0);
      __pyx_t_7 = __pyx_t_8;
      __pyx_L15_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":928
 *                     self.nitrogenCount += 1
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:
 *                     self.oxygenCount += 1             # <<<<<<<<<<<<<<
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:
 *                     self.sulfurCount += 1
 */
        __pyx_v_self->oxygenCount = (__pyx_v_self->oxygenCount + 1);
        goto __pyx_L6;
      }

      /* "rmgpy/molecule/group.py":929
 *                 elif isOxygen and not isCarbon and not isNitrogen and not isSulfur:
 *                     self.oxygenCount += 1
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:             # <<<<<<<<<<<<<<
 *                     self.sulfurCount += 1
 *             if len(atom.radicalElectrons) == 1:
 */
      __pyx_t_8 = (__pyx_v_isSulfur != 0);
      if (__pyx_t_8) {
        goto __pyx_L20_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L19_bool_binop_done;
      }
      __pyx_L20_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isCarbon != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L21_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L19_bool_binop_done;
      }
      __pyx_L21_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isNitrogen != 0)) != 0);
      if (__pyx_t_8) {
        goto __pyx_L22_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L19_bool_binop_done;
      }
      __pyx_L22_next_and:;
      __pyx_t_8 = ((!(__pyx_v_isOxygen != 0)) != 0);
      __pyx_t_7 = __pyx_t_8;
      __pyx_L19_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":930
 *                     self.oxygenCount += 1
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:
 *                     self.sulfurCount += 1             # <<<<<<<<<<<<<<
 *             if len(atom.radicalElectrons) == 1:
 *                 radical = atom.radicalElectrons[0]
 */
        __pyx_v_self->sulfurCount = (__pyx_v_self->sulfurCount + 1);
        goto __pyx_L6;
      }
      __pyx_L6:;
      goto __pyx_L5;
    }
    __pyx_L5:;

    /* "rmgpy/molecule/group.py":931
 *                 elif isSulfur and not isCarbon and not isNitrogen and not isOxygen:
 *                     self.sulfurCount += 1
 *             if len(atom.radicalElectrons) == 1:             # <<<<<<<<<<<<<<
 *                 radical = atom.radicalElectrons[0]
 *                 self.radicalCount += radical
 */
    __pyx_t_2 = __pyx_v_atom->radicalElectrons;
    __Pyx_INCREF(__pyx_t_2);
    if (unlikely(__pyx_t_2 == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "object of type 'NoneType' has no len()");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 931; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_6 = PyList_GET_SIZE(__pyx_t_2); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 931; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_7 = ((__pyx_t_6 == 1) != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":932
 *                     self.sulfurCount += 1
 *             if len(atom.radicalElectrons) == 1:
 *                 radical = atom.radicalElectrons[0]             # <<<<<<<<<<<<<<
 *                 self.radicalCount += radical
 * 
 */
      if (unlikely(__pyx_v_atom->radicalElectrons == Py_None)) {
        PyErr_SetString(PyExc_TypeError, "'NoneType' object is not subscriptable");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 932; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_2 = __Pyx_GetItemInt_List(__pyx_v_atom->radicalElectrons, 0, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 932; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_9 = __Pyx_PyInt_As_int(__pyx_t_2); if (unlikely((__pyx_t_9 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 932; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_radical = __pyx_t_9;

      /* "rmgpy/molecule/group.py":933
 *             if len(atom.radicalElectrons) == 1:
 *                 radical = atom.radicalElectrons[0]
 *                 self.radicalCount += radical             # <<<<<<<<<<<<<<
 * 
 *     def isIsomorphic(self, other, initialMap=None):
 */
      __pyx_v_self->radicalCount = (__pyx_v_self->radicalCount + __pyx_v_radical);
      goto __pyx_L23;
    }
    __pyx_L23:;

    /* "rmgpy/molecule/group.py":916
 *         self.sulfurCount   = 0
 *         self.radicalCount  = 0
 *         for atom in self.vertices:             # <<<<<<<<<<<<<<
 *             if len(atom.atomType) == 1:
 *                 atomType   = atom.atomType[0]
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":897
 * 
 * 
 *     def updateFingerprint(self):             # <<<<<<<<<<<<<<
 *         """
 *         Update the molecular fingerprint used to accelerate the subgraph
 */

  /* function exit code */
  __pyx_r = Py_None; __Pyx_INCREF(Py_None);
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.updateFingerprint", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_atom);
  __Pyx_XDECREF((PyObject *)__pyx_v_atomType);
  __Pyx_XDECREF((PyObject *)__pyx_v_carbon);
  __Pyx_XDECREF((PyObject *)__pyx_v_nitrogen);
  __Pyx_XDECREF((PyObject *)__pyx_v_oxygen);
  __Pyx_XDECREF((PyObject *)__pyx_v_sulfur);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_46updateFingerprint[] = "\n        Update the molecular fingerprint used to accelerate the subgraph\n        isomorphism checks.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("updateFingerprint (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_46updateFingerprint(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_46updateFingerprint(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("updateFingerprint", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __pyx_f_5rmgpy_8molecule_5group_5Group_updateFingerprint(__pyx_v_self, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 897; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.updateFingerprint", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":935
 *                 self.radicalCount += radical
 * 
 *     def isIsomorphic(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if two graphs are isomorphic and ``False``
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isIsomorphic *__pyx_optional_args) {
  PyObject *__pyx_v_initialMap = ((PyObject*)Py_None);
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_t_8;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isIsomorphic", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_initialMap = __pyx_optional_args->initialMap;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isIsomorphic); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_other));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_other));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
      __Pyx_INCREF(__pyx_v_initialMap);
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_initialMap);
      __Pyx_GIVEREF(__pyx_v_initialMap);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_7 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_7;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":945
 *         # It only makes sense to compare a Group to a Group for full
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # Do the isomorphism comparison
 */
  __pyx_t_7 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_8 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":946
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         # Do the isomorphism comparison
 *         return Graph.isIsomorphic(self, other, initialMap)
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":948
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # Do the isomorphism comparison
 *         return Graph.isIsomorphic(self, other, initialMap)             # <<<<<<<<<<<<<<
 * 
 *     def findIsomorphism(self, other, initialMap=None):
 */
  __pyx_t_9.__pyx_n = 1;
  __pyx_t_9.initialMap = __pyx_v_initialMap;
  __pyx_t_8 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->isIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_9); if (unlikely(__pyx_t_8 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 948; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_t_8;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":935
 *                 self.radicalCount += radical
 * 
 *     def isIsomorphic(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if two graphs are isomorphic and ``False``
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_48isIsomorphic[] = "\n        Returns ``True`` if two graphs are isomorphic and ``False``\n        otherwise. The `initialMap` attribute can be used to specify a required\n        mapping from `self` to `other` (i.e. the atoms of `self` are the keys,\n        while the atoms of `other` are the values). The `other` parameter must\n        be a :class:`Group` object, or a :class:`TypeError` is raised.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other = 0;
  PyObject *__pyx_v_initialMap = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isIsomorphic (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_other,&__pyx_n_s_initialMap,0};
    PyObject* values[2] = {0,0};
    values[1] = ((PyObject*)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_other)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_initialMap);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "isIsomorphic") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)values[0]);
    __pyx_v_initialMap = ((PyObject*)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("isIsomorphic", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_initialMap), (&PyDict_Type), 1, "initialMap", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_48isIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_other, __pyx_v_initialMap);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_48isIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic __pyx_t_2;
  PyObject *__pyx_t_3 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isIsomorphic", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.isIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_2); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 935; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_r = __pyx_t_3;
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":950
 *         return Graph.isIsomorphic(self, other, initialMap)
 * 
 *     def findIsomorphism(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is isomorphic and ``False``
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_findIsomorphism(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findIsomorphism *__pyx_optional_args) {
  PyObject *__pyx_v_initialMap = ((PyObject*)Py_None);
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_t_8;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism __pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("findIsomorphism", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_initialMap = __pyx_optional_args->initialMap;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_findIsomorphism); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_other));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_other));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
      __Pyx_INCREF(__pyx_v_initialMap);
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_initialMap);
      __Pyx_GIVEREF(__pyx_v_initialMap);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyList_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":962
 *         # It only makes sense to compare a Group to a Group for full
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # Do the isomorphism comparison
 */
  __pyx_t_7 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_8 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":963
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         # Do the isomorphism comparison
 *         return Graph.findIsomorphism(self, other, initialMap)
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 963; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":965
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # Do the isomorphism comparison
 *         return Graph.findIsomorphism(self, other, initialMap)             # <<<<<<<<<<<<<<
 * 
 *     def isSubgraphIsomorphic(self, other, initialMap=None):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_9.__pyx_n = 1;
  __pyx_t_9.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->findIsomorphism(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_9); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 965; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":950
 *         return Graph.isIsomorphic(self, other, initialMap)
 * 
 *     def findIsomorphism(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is isomorphic and ``False``
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findIsomorphism", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_50findIsomorphism[] = "\n        Returns ``True`` if `other` is isomorphic and ``False``\n        otherwise, and the matching mapping. The `initialMap` attribute can be\n        used to specify a required mapping from `self` to `other` (i.e. the\n        atoms of `self` are the keys, while the atoms of `other` are the\n        values). The returned mapping also uses the atoms of `self` for the keys\n        and the atoms of `other` for the values. The `other` parameter must\n        be a :class:`Group` object, or a :class:`TypeError` is raised.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other = 0;
  PyObject *__pyx_v_initialMap = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("findIsomorphism (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_other,&__pyx_n_s_initialMap,0};
    PyObject* values[2] = {0,0};
    values[1] = ((PyObject*)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_other)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_initialMap);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "findIsomorphism") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)values[0]);
    __pyx_v_initialMap = ((PyObject*)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("findIsomorphism", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findIsomorphism", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_initialMap), (&PyDict_Type), 1, "initialMap", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_50findIsomorphism(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_other, __pyx_v_initialMap);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_50findIsomorphism(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("findIsomorphism", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.findIsomorphism(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 950; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findIsomorphism", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":967
 *         return Graph.findIsomorphism(self, other, initialMap)
 * 
 *     def isSubgraphIsomorphic(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is subgraph isomorphic and ``False``
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic *__pyx_optional_args) {
  PyObject *__pyx_v_initialMap = ((PyObject*)Py_None);
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_group = 0;
  short __pyx_v_mult1;
  short __pyx_v_mult2;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_t_8;
  short __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSubgraphIsomorphic", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_initialMap = __pyx_optional_args->initialMap;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isSubgraphIsomorphic); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_other));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_other));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
      __Pyx_INCREF(__pyx_v_initialMap);
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_initialMap);
      __Pyx_GIVEREF(__pyx_v_initialMap);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_7 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_7;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":980
 *         # It only makes sense to compare a Group to a Group for subgraph
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         group = other
 */
  __pyx_t_7 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_8 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":981
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         group = other
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 981; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 981; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 981; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 981; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 981; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 981; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 981; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 981; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":982
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         group = other             # <<<<<<<<<<<<<<
 * 
 *         if self.multiplicity:
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 982; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_group = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":984
 *         group = other
 * 
 *         if self.multiplicity:             # <<<<<<<<<<<<<<
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 */
  __pyx_t_8 = (__pyx_v_self->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_self->multiplicity) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":985
 * 
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:             # <<<<<<<<<<<<<<
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 */
    if (unlikely(__pyx_v_self->multiplicity == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->multiplicity; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
    for (;;) {
      if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_9 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_9 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 985; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_mult1 = __pyx_t_9;

      /* "rmgpy/molecule/group.py":986
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:             # <<<<<<<<<<<<<<
 *                     for mult2 in group.multiplicity:
 *                         if mult1 == mult2: break
 */
      __pyx_t_8 = (__pyx_v_group->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_group->multiplicity) != 0);
      if (__pyx_t_8) {

        /* "rmgpy/molecule/group.py":987
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->multiplicity == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 987; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->multiplicity; __Pyx_INCREF(__pyx_t_2); __pyx_t_10 = 0;
        for (;;) {
          if (__pyx_t_10 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_10); __Pyx_INCREF(__pyx_t_4); __pyx_t_10++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 987; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_10); __pyx_t_10++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 987; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_9 = __Pyx_PyInt_As_short(__pyx_t_4); if (unlikely((__pyx_t_9 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 987; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
          __pyx_v_mult2 = __pyx_t_9;

          /* "rmgpy/molecule/group.py":988
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 *                         if mult1 == mult2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return False
 */
          __pyx_t_8 = ((__pyx_v_mult1 == __pyx_v_mult2) != 0);
          if (__pyx_t_8) {
            goto __pyx_L9_break;
          }

          /* "rmgpy/molecule/group.py":987
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":990
 *                         if mult1 == mult2: break
 *                     else:
 *                         return False             # <<<<<<<<<<<<<<
 *         else:
 *             if group.multiplicity: return False
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":987
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        __pyx_L9_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L7;
      }
      __pyx_L7:;

      /* "rmgpy/molecule/group.py":985
 * 
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:             # <<<<<<<<<<<<<<
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L4;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":992
 *                         return False
 *         else:
 *             if group.multiplicity: return False             # <<<<<<<<<<<<<<
 *         # Do the isomorphism comparison
 *         return Graph.isSubgraphIsomorphic(self, other, initialMap)
 */
    __pyx_t_8 = (__pyx_v_group->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_group->multiplicity) != 0);
    if (__pyx_t_8) {
      __pyx_r = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L4:;

  /* "rmgpy/molecule/group.py":994
 *             if group.multiplicity: return False
 *         # Do the isomorphism comparison
 *         return Graph.isSubgraphIsomorphic(self, other, initialMap)             # <<<<<<<<<<<<<<
 * 
 *     def findSubgraphIsomorphisms(self, other, initialMap=None):
 */
  __pyx_t_11.__pyx_n = 1;
  __pyx_t_11.initialMap = __pyx_v_initialMap;
  __pyx_t_8 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->isSubgraphIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_11); if (unlikely(__pyx_t_8 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 994; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_t_8;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":967
 *         return Graph.findIsomorphism(self, other, initialMap)
 * 
 *     def isSubgraphIsomorphic(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is subgraph isomorphic and ``False``
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isSubgraphIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -2;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_group);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic[] = "\n        Returns ``True`` if `other` is subgraph isomorphic and ``False``\n        otherwise. In other words, return ``True`` if self is more specific than other.\n        The `initialMap` attribute can be used to specify a required\n        mapping from `self` to `other` (i.e. the atoms of `self` are the keys,\n        while the atoms of `other` are the values). The `other` parameter must\n        be a :class:`Group` object, or a :class:`TypeError` is raised.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other = 0;
  PyObject *__pyx_v_initialMap = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isSubgraphIsomorphic (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_other,&__pyx_n_s_initialMap,0};
    PyObject* values[2] = {0,0};
    values[1] = ((PyObject*)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_other)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_initialMap);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "isSubgraphIsomorphic") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)values[0]);
    __pyx_v_initialMap = ((PyObject*)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("isSubgraphIsomorphic", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isSubgraphIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_initialMap), (&PyDict_Type), 1, "initialMap", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_other, __pyx_v_initialMap);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic __pyx_t_2;
  PyObject *__pyx_t_3 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isSubgraphIsomorphic", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.isSubgraphIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_2); if (unlikely(__pyx_t_1 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = __Pyx_PyBool_FromLong(__pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 967; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_r = __pyx_t_3;
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isSubgraphIsomorphic", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":996
 *         return Graph.isSubgraphIsomorphic(self, other, initialMap)
 * 
 *     def findSubgraphIsomorphisms(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is subgraph isomorphic and ``False``
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static PyObject *__pyx_f_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms *__pyx_optional_args) {
  PyObject *__pyx_v_initialMap = ((PyObject*)Py_None);
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_group = 0;
  short __pyx_v_mult1;
  short __pyx_v_mult2;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_t_7;
  int __pyx_t_8;
  short __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms __pyx_t_11;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("findSubgraphIsomorphisms", 0);
  if (__pyx_optional_args) {
    if (__pyx_optional_args->__pyx_n > 0) {
      __pyx_v_initialMap = __pyx_optional_args->initialMap;
    }
  }
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_findSubgraphIsomorphisms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms)) {
      __Pyx_XDECREF(__pyx_r);
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      __pyx_t_5 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
          __pyx_t_5 = 1;
        }
      }
      __pyx_t_6 = PyTuple_New(2+__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      if (__pyx_t_4) {
        PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
      }
      __Pyx_INCREF(((PyObject *)__pyx_v_other));
      PyTuple_SET_ITEM(__pyx_t_6, 0+__pyx_t_5, ((PyObject *)__pyx_v_other));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
      __Pyx_INCREF(__pyx_v_initialMap);
      PyTuple_SET_ITEM(__pyx_t_6, 1+__pyx_t_5, __pyx_v_initialMap);
      __Pyx_GIVEREF(__pyx_v_initialMap);
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_6, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(PyList_CheckExact(__pyx_t_2))||((__pyx_t_2) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_t_2)->tp_name), 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((PyObject*)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1013
 *         # It only makes sense to compare a Group to a Group for subgraph
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         group = other
 */
  __pyx_t_7 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_8 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":1014
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         group = other
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1014; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1014; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_6 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_6 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_6)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_6);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_6) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1014; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1014; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_6); __Pyx_GIVEREF(__pyx_t_6); __pyx_t_6 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1014; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1014; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1014; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1014; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":1015
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         group = other             # <<<<<<<<<<<<<<
 * 
 *         if self.multiplicity:
 */
  if (!(likely(((((PyObject *)__pyx_v_other)) == Py_None) || likely(__Pyx_TypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1015; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = ((PyObject *)__pyx_v_other);
  __Pyx_INCREF(__pyx_t_1);
  __pyx_v_group = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1017
 *         group = other
 * 
 *         if self.multiplicity:             # <<<<<<<<<<<<<<
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 */
  __pyx_t_8 = (__pyx_v_self->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_self->multiplicity) != 0);
  if (__pyx_t_8) {

    /* "rmgpy/molecule/group.py":1018
 * 
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:             # <<<<<<<<<<<<<<
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 */
    if (unlikely(__pyx_v_self->multiplicity == Py_None)) {
      PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1018; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_1 = __pyx_v_self->multiplicity; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
    for (;;) {
      if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1018; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1018; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __pyx_t_9 = __Pyx_PyInt_As_short(__pyx_t_2); if (unlikely((__pyx_t_9 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1018; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_v_mult1 = __pyx_t_9;

      /* "rmgpy/molecule/group.py":1019
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:             # <<<<<<<<<<<<<<
 *                     for mult2 in group.multiplicity:
 *                         if mult1 == mult2: break
 */
      __pyx_t_8 = (__pyx_v_group->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_group->multiplicity) != 0);
      if (__pyx_t_8) {

        /* "rmgpy/molecule/group.py":1020
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        if (unlikely(__pyx_v_group->multiplicity == Py_None)) {
          PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1020; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __pyx_t_2 = __pyx_v_group->multiplicity; __Pyx_INCREF(__pyx_t_2); __pyx_t_10 = 0;
        for (;;) {
          if (__pyx_t_10 >= PyList_GET_SIZE(__pyx_t_2)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_4 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_10); __Pyx_INCREF(__pyx_t_4); __pyx_t_10++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1020; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_4 = PySequence_ITEM(__pyx_t_2, __pyx_t_10); __pyx_t_10++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1020; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
          __pyx_t_9 = __Pyx_PyInt_As_short(__pyx_t_4); if (unlikely((__pyx_t_9 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1020; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
          __pyx_v_mult2 = __pyx_t_9;

          /* "rmgpy/molecule/group.py":1021
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 *                         if mult1 == mult2: break             # <<<<<<<<<<<<<<
 *                     else:
 *                         return []
 */
          __pyx_t_8 = ((__pyx_v_mult1 == __pyx_v_mult2) != 0);
          if (__pyx_t_8) {
            goto __pyx_L9_break;
          }

          /* "rmgpy/molecule/group.py":1020
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":1023
 *                         if mult1 == mult2: break
 *                     else:
 *                         return []             # <<<<<<<<<<<<<<
 *         else:
 *             if group.multiplicity: return []
 */
          __Pyx_XDECREF(__pyx_r);
          __pyx_t_4 = PyList_New(0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1023; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_4);
          __pyx_r = ((PyObject*)__pyx_t_4);
          __pyx_t_4 = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }

        /* "rmgpy/molecule/group.py":1020
 *             for mult1 in self.multiplicity:
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:             # <<<<<<<<<<<<<<
 *                         if mult1 == mult2: break
 *                     else:
 */
        __pyx_L9_break:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L7;
      }
      __pyx_L7:;

      /* "rmgpy/molecule/group.py":1018
 * 
 *         if self.multiplicity:
 *             for mult1 in self.multiplicity:             # <<<<<<<<<<<<<<
 *                 if group.multiplicity:
 *                     for mult2 in group.multiplicity:
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    goto __pyx_L4;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":1025
 *                         return []
 *         else:
 *             if group.multiplicity: return []             # <<<<<<<<<<<<<<
 * 
 *         # Do the isomorphism comparison
 */
    __pyx_t_8 = (__pyx_v_group->multiplicity != Py_None) && (PyList_GET_SIZE(__pyx_v_group->multiplicity) != 0);
    if (__pyx_t_8) {
      __Pyx_XDECREF(__pyx_r);
      __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1025; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_r = ((PyObject*)__pyx_t_1);
      __pyx_t_1 = 0;
      goto __pyx_L0;
    }
  }
  __pyx_L4:;

  /* "rmgpy/molecule/group.py":1028
 * 
 *         # Do the isomorphism comparison
 *         return Graph.findSubgraphIsomorphisms(self, other, initialMap)             # <<<<<<<<<<<<<<
 * 
 *     def isIdentical(self, other):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_11.__pyx_n = 1;
  __pyx_t_11.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph->findSubgraphIsomorphisms(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_11); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1028; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":996
 *         return Graph.isSubgraphIsomorphic(self, other, initialMap)
 * 
 *     def findSubgraphIsomorphisms(self, other, initialMap=None):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is subgraph isomorphic and ``False``
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findSubgraphIsomorphisms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_group);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms[] = "\n        Returns ``True`` if `other` is subgraph isomorphic and ``False``\n        otherwise. In other words, return ``True`` is self is more specific than other.\n        Also returns the lists all of valid mappings. The\n        `initialMap` attribute can be used to specify a required mapping from\n        `self` to `other` (i.e. the atoms of `self` are the keys, while the\n        atoms of `other` are the values). The returned mappings also use the\n        atoms of `self` for the keys and the atoms of `other` for the values.\n        The `other` parameter must be a :class:`Group` object, or a\n        :class:`TypeError` is raised.\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other = 0;
  PyObject *__pyx_v_initialMap = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("findSubgraphIsomorphisms (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_other,&__pyx_n_s_initialMap,0};
    PyObject* values[2] = {0,0};
    values[1] = ((PyObject*)Py_None);
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_other)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_initialMap);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "findSubgraphIsomorphisms") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_other = ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)values[0]);
    __pyx_v_initialMap = ((PyObject*)values[1]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("findSubgraphIsomorphisms", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findSubgraphIsomorphisms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_initialMap), (&PyDict_Type), 1, "initialMap", 1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), __pyx_v_other, __pyx_v_initialMap);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, PyObject *__pyx_v_initialMap) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("findSubgraphIsomorphisms", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2.__pyx_n = 1;
  __pyx_t_2.initialMap = __pyx_v_initialMap;
  __pyx_t_1 = __pyx_vtabptr_5rmgpy_8molecule_5group_Group->__pyx_base.findSubgraphIsomorphisms(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 1, &__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 996; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.findSubgraphIsomorphisms", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1030
 *         return Graph.findSubgraphIsomorphisms(self, other, initialMap)
 * 
 *     def isIdentical(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is identical and ``False`` otherwise.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isIdentical(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other, int __pyx_skip_dispatch) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_t_6;
  int __pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isIdentical", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isIdentical); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (!__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_other)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4); __Pyx_GIVEREF(__pyx_t_4); __pyx_t_4 = NULL;
        __Pyx_INCREF(((PyObject *)__pyx_v_other));
        PyTuple_SET_ITEM(__pyx_t_5, 0+1, ((PyObject *)__pyx_v_other));
        __Pyx_GIVEREF(((PyObject *)__pyx_v_other));
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_6 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_6;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1039
 *         # It only makes sense to compare a Group to a Group for full
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):             # <<<<<<<<<<<<<<
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))
 *         # An identical group is always a child of itself and
 */
  __pyx_t_6 = __Pyx_TypeCheck(((PyObject *)__pyx_v_other), ((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)); 
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":1040
 *         # isomorphism, so raise an exception if this is not what was requested
 *         if not isinstance(other, Group):
 *             raise TypeError('Got a {0} object for parameter "other", when a Group object is required.'.format(other.__class__))             # <<<<<<<<<<<<<<
 *         # An identical group is always a child of itself and
 *         # is the only case where that is true. Therefore
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_n_s_format); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1040; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_other), __pyx_n_s_class); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1040; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
      __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_2);
      if (likely(__pyx_t_5)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_5);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_2, function);
      }
    }
    if (!__pyx_t_5) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1040; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_GOTREF(__pyx_t_1);
    } else {
      __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1040; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
      PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1040; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1040; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_TypeError, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1040; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_Raise(__pyx_t_1, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1040; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "rmgpy/molecule/group.py":1045
 *         # if we do both directions of isSubgraphIsmorphic, we need
 *         # to get True twice for it to be identical
 *         if not self.isSubgraphIsomorphic(other):             # <<<<<<<<<<<<<<
 *             return False
 *         elif not other.isSubgraphIsomorphic(self):
 */
  __pyx_t_7 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.isSubgraphIsomorphic(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), __pyx_v_other, 0, NULL); if (unlikely(__pyx_t_7 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1045; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_6 = ((!(__pyx_t_7 != 0)) != 0);
  if (__pyx_t_6) {

    /* "rmgpy/molecule/group.py":1046
 *         # to get True twice for it to be identical
 *         if not self.isSubgraphIsomorphic(other):
 *             return False             # <<<<<<<<<<<<<<
 *         elif not other.isSubgraphIsomorphic(self):
 *             return False
 */
    __pyx_r = 0;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":1047
 *         if not self.isSubgraphIsomorphic(other):
 *             return False
 *         elif not other.isSubgraphIsomorphic(self):             # <<<<<<<<<<<<<<
 *             return False
 *         else:
 */
  __pyx_t_6 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph *)__pyx_v_other->__pyx_vtab)->isSubgraphIsomorphic(__pyx_v_other, ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 0, NULL); if (unlikely(__pyx_t_6 == -2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1047; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_7 = ((!(__pyx_t_6 != 0)) != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":1048
 *             return False
 *         elif not other.isSubgraphIsomorphic(self):
 *             return False             # <<<<<<<<<<<<<<
 *         else:
 *             return True
 */
    __pyx_r = 0;
    goto __pyx_L0;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":1050
 *             return False
 *         else:
 *             return True             # <<<<<<<<<<<<<<
 * 
 *     def isAromaticRing(self):
 */
    __pyx_r = 1;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":1030
 *         return Graph.findSubgraphIsomorphisms(self, other, initialMap)
 * 
 *     def isIdentical(self, other):             # <<<<<<<<<<<<<<
 *         """
 *         Returns ``True`` if `other` is identical and ``False`` otherwise.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.isIdentical", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical(PyObject *__pyx_v_self, PyObject *__pyx_v_other); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_56isIdentical[] = "\n        Returns ``True`` if `other` is identical and ``False`` otherwise.\n        The function `isIsomorphic` respects wildcards, while this function\n        does not, make it more useful for checking groups to groups (as\n        opposed to molecules to groups)\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical(PyObject *__pyx_v_self, PyObject *__pyx_v_other) {
  CYTHON_UNUSED int __pyx_lineno = 0;
  CYTHON_UNUSED const char *__pyx_filename = NULL;
  CYTHON_UNUSED int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isIdentical (wrapper)", 0);
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_other), __pyx_ptype_5rmgpy_8molecule_5graph_Graph, 1, "other", 0))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_56isIdentical(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_other));

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_56isIdentical(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *__pyx_v_other) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isIdentical", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_isIdentical(__pyx_v_self, __pyx_v_other, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1030; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isIdentical", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1052
 *             return True
 * 
 *     def isAromaticRing(self):             # <<<<<<<<<<<<<<
 *         """
 *         This method returns a boolean telling if the group has a 5 or 6 cyclic with
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_59isAromaticRing(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isAromaticRing(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  Py_ssize_t __pyx_v_ring_size;
  PyObject *__pyx_v_ringAtom = NULL;
  PyObject *__pyx_v_bondedAtom = NULL;
  PyObject *__pyx_v_bond = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  int __pyx_t_7;
  PyObject *(*__pyx_t_8)(PyObject *);
  Py_ssize_t __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  int __pyx_t_11;
  int __pyx_t_12;
  PyObject *__pyx_t_13 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isAromaticRing", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isAromaticRing); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1052; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_59isAromaticRing)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1052; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1052; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1052; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1058
 *         """
 * 
 *         ring_size = len(self.atoms)             # <<<<<<<<<<<<<<
 *         if ring_size not in [5, 6]:
 *             return False
 */
  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1058; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = PyObject_Length(__pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1058; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_ring_size = __pyx_t_6;

  /* "rmgpy/molecule/group.py":1059
 * 
 *         ring_size = len(self.atoms)
 *         if ring_size not in [5, 6]:             # <<<<<<<<<<<<<<
 *             return False
 *         for ringAtom in self.atoms:
 */
  switch (__pyx_v_ring_size) {
    case 5:
    case 6:
    __pyx_t_5 = 0;
    break;
    default:
    __pyx_t_5 = 1;
    break;
  }
  __pyx_t_7 = (__pyx_t_5 != 0);
  if (__pyx_t_7) {

    /* "rmgpy/molecule/group.py":1060
 *         ring_size = len(self.atoms)
 *         if ring_size not in [5, 6]:
 *             return False             # <<<<<<<<<<<<<<
 *         for ringAtom in self.atoms:
 *             for bondedAtom, bond in ringAtom.edges.iteritems():
 */
    __pyx_r = 0;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":1061
 *         if ring_size not in [5, 6]:
 *             return False
 *         for ringAtom in self.atoms:             # <<<<<<<<<<<<<<
 *             for bondedAtom, bond in ringAtom.edges.iteritems():
 *                 if bondedAtom in self.atoms:
 */
  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1061; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
    __pyx_t_2 = __pyx_t_1; __Pyx_INCREF(__pyx_t_2); __pyx_t_6 = 0;
    __pyx_t_8 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1061; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_8 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1061; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (;;) {
    if (likely(!__pyx_t_8)) {
      if (likely(PyList_CheckExact(__pyx_t_2))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1061; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1061; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1061; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1061; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_1 = __pyx_t_8(__pyx_t_2);
      if (unlikely(!__pyx_t_1)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1061; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_1);
    }
    __Pyx_XDECREF_SET(__pyx_v_ringAtom, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1062
 *             return False
 *         for ringAtom in self.atoms:
 *             for bondedAtom, bond in ringAtom.edges.iteritems():             # <<<<<<<<<<<<<<
 *                 if bondedAtom in self.atoms:
 *                     if not bond.isBenzene():
 */
    __pyx_t_9 = 0;
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_ringAtom, __pyx_n_s_edges); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1062; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    if (unlikely(__pyx_t_3 == Py_None)) {
      PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1062; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_4 = __Pyx_dict_iterator(__pyx_t_3, 0, __pyx_n_s_iteritems, (&__pyx_t_10), (&__pyx_t_11)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1062; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF(__pyx_t_1);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
    while (1) {
      __pyx_t_12 = __Pyx_dict_iter_next(__pyx_t_1, __pyx_t_10, &__pyx_t_9, &__pyx_t_4, &__pyx_t_3, NULL, __pyx_t_11);
      if (unlikely(__pyx_t_12 == 0)) break;
      if (unlikely(__pyx_t_12 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1062; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_XDECREF_SET(__pyx_v_bondedAtom, __pyx_t_4);
      __pyx_t_4 = 0;
      __Pyx_XDECREF_SET(__pyx_v_bond, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1063
 *         for ringAtom in self.atoms:
 *             for bondedAtom, bond in ringAtom.edges.iteritems():
 *                 if bondedAtom in self.atoms:             # <<<<<<<<<<<<<<
 *                     if not bond.isBenzene():
 *                         return False
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1063; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_7 = (__Pyx_PySequence_Contains(__pyx_v_bondedAtom, __pyx_t_3, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1063; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = (__pyx_t_7 != 0);
      if (__pyx_t_5) {

        /* "rmgpy/molecule/group.py":1064
 *             for bondedAtom, bond in ringAtom.edges.iteritems():
 *                 if bondedAtom in self.atoms:
 *                     if not bond.isBenzene():             # <<<<<<<<<<<<<<
 *                         return False
 *         return True
 */
        __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond, __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1064; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_13 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
          __pyx_t_13 = PyMethod_GET_SELF(__pyx_t_4);
          if (likely(__pyx_t_13)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_13);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_4, function);
          }
        }
        if (__pyx_t_13) {
          __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_13); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1064; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
        } else {
          __pyx_t_3 = __Pyx_PyObject_CallNoArg(__pyx_t_4); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1064; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_3);
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1064; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_7 = ((!__pyx_t_5) != 0);
        if (__pyx_t_7) {

          /* "rmgpy/molecule/group.py":1065
 *                 if bondedAtom in self.atoms:
 *                     if not bond.isBenzene():
 *                         return False             # <<<<<<<<<<<<<<
 *         return True
 * 
 */
          __pyx_r = 0;
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          goto __pyx_L0;
        }
        goto __pyx_L8;
      }
      __pyx_L8:;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1061
 *         if ring_size not in [5, 6]:
 *             return False
 *         for ringAtom in self.atoms:             # <<<<<<<<<<<<<<
 *             for bondedAtom, bond in ringAtom.edges.iteritems():
 *                 if bondedAtom in self.atoms:
 */
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1066
 *                     if not bond.isBenzene():
 *                         return False
 *         return True             # <<<<<<<<<<<<<<
 * 
 *     def standardizeAtomType(self):
 */
  __pyx_r = 1;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1052
 *             return True
 * 
 *     def isAromaticRing(self):             # <<<<<<<<<<<<<<
 *         """
 *         This method returns a boolean telling if the group has a 5 or 6 cyclic with
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_13);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.isAromaticRing", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_ringAtom);
  __Pyx_XDECREF(__pyx_v_bondedAtom);
  __Pyx_XDECREF(__pyx_v_bond);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_59isAromaticRing(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_58isAromaticRing[] = "\n        This method returns a boolean telling if the group has a 5 or 6 cyclic with\n        benzene bonds exclusively\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_59isAromaticRing(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isAromaticRing (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_58isAromaticRing(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_58isAromaticRing(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isAromaticRing", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_isAromaticRing(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1052; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isAromaticRing", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1068
 *         return True
 * 
 *     def standardizeAtomType(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function changes the atomTypes in a group if the atom must
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_61standardizeAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_standardizeAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_v_modified;
  CYTHON_UNUSED PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_specifics = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_index = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_claimedAtomType = NULL;
  PyObject *__pyx_v_newAtomType = NULL;
  PyObject *__pyx_v_element = NULL;
  PyObject *__pyx_v_elementLabel = NULL;
  PyObject *__pyx_v_groupFeatures = NULL;
  PyObject *__pyx_v_bondOrder = NULL;
  PyObject *__pyx_v_filledValency = NULL;
  PyObject *__pyx_v_specificAtomType = NULL;
  PyObject *__pyx_v_atomtypeFeatureList = NULL;
  PyObject *__pyx_v_molFeature = NULL;
  PyObject *__pyx_v_atomtypeFeature = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  Py_ssize_t __pyx_t_7;
  int __pyx_t_8;
  int __pyx_t_9;
  PyObject *(*__pyx_t_10)(PyObject *);
  int __pyx_t_11;
  PyObject *__pyx_t_12 = NULL;
  PyObject *__pyx_t_13 = NULL;
  int __pyx_t_14;
  PyObject *__pyx_t_15 = NULL;
  PyObject *(*__pyx_t_16)(PyObject *);
  Py_ssize_t __pyx_t_17;
  PyObject *(*__pyx_t_18)(PyObject *);
  PyObject *__pyx_t_19 = NULL;
  PyObject *(*__pyx_t_20)(PyObject *);
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("standardizeAtomType", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_standardizeAtomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1068; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_61standardizeAtomType)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1068; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1068; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1068; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1085
 *         Returns a 'True' if the group was modified otherwise returns 'False'
 *         """
 *         modified = False             # <<<<<<<<<<<<<<
 * 
 *         #If this atom or any of its ligands has wild cards, then don't try to standardize
 */
  __pyx_v_modified = 0;

  /* "rmgpy/molecule/group.py":1088
 * 
 *         #If this atom or any of its ligands has wild cards, then don't try to standardize
 *         if self.hasWildCards: return modified             # <<<<<<<<<<<<<<
 *         for bond12, atom2 in self.bonds.iteritems():
 *             if atom2.hasWildCards: return modified
 */
  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_hasWildCards); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1088; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1088; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if (__pyx_t_5) {
    __pyx_r = __pyx_v_modified;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":1089
 *         #If this atom or any of its ligands has wild cards, then don't try to standardize
 *         if self.hasWildCards: return modified
 *         for bond12, atom2 in self.bonds.iteritems():             # <<<<<<<<<<<<<<
 *             if atom2.hasWildCards: return modified
 * 
 */
  __pyx_t_6 = 0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_bonds); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1089; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (unlikely(__pyx_t_2 == Py_None)) {
    PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1089; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_3 = __Pyx_dict_iterator(__pyx_t_2, 0, __pyx_n_s_iteritems, (&__pyx_t_7), (&__pyx_t_8)); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1089; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_XDECREF(__pyx_t_1);
  __pyx_t_1 = __pyx_t_3;
  __pyx_t_3 = 0;
  while (1) {
    __pyx_t_9 = __Pyx_dict_iter_next(__pyx_t_1, __pyx_t_7, &__pyx_t_6, &__pyx_t_3, &__pyx_t_2, NULL, __pyx_t_8);
    if (unlikely(__pyx_t_9 == 0)) break;
    if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1089; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_3);
    __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1090
 *         if self.hasWildCards: return modified
 *         for bond12, atom2 in self.bonds.iteritems():
 *             if atom2.hasWildCards: return modified             # <<<<<<<<<<<<<<
 * 
 *         #list of :class:AtomType which are elements with more sub-divided atomtypes beneath them
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom2, __pyx_n_s_hasWildCards); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1090; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1090; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_5) {
      __pyx_r = __pyx_v_modified;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1093
 * 
 *         #list of :class:AtomType which are elements with more sub-divided atomtypes beneath them
 *         specifics= [elementLabel for elementLabel in allElements if elementLabel not in nonSpecifics]             # <<<<<<<<<<<<<<
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_allElements); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_3 = __pyx_t_2; __Pyx_INCREF(__pyx_t_3); __pyx_t_7 = 0;
    __pyx_t_10 = NULL;
  } else {
    __pyx_t_7 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_10 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_10)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_7 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_7); __Pyx_INCREF(__pyx_t_2); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_10(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_elementLabel, __pyx_t_2);
    __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_nonSpecifics); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_elementLabel, __pyx_t_2, Py_NE)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_11 = (__pyx_t_5 != 0);
    if (__pyx_t_11) {
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_v_elementLabel))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1093; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L9;
    }
    __pyx_L9:;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_v_specifics = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1094
 *         #list of :class:AtomType which are elements with more sub-divided atomtypes beneath them
 *         specifics= [elementLabel for elementLabel in allElements if elementLabel not in nonSpecifics]
 *         for index, atom in enumerate(self.atoms):             # <<<<<<<<<<<<<<
 *             claimedAtomType = atom.atomType[0]
 *             newAtomType = None
 */
  __Pyx_INCREF(__pyx_int_0);
  __pyx_t_1 = __pyx_int_0;
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (likely(PyList_CheckExact(__pyx_t_3)) || PyTuple_CheckExact(__pyx_t_3)) {
    __pyx_t_2 = __pyx_t_3; __Pyx_INCREF(__pyx_t_2); __pyx_t_7 = 0;
    __pyx_t_10 = NULL;
  } else {
    __pyx_t_7 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_10 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  for (;;) {
    if (likely(!__pyx_t_10)) {
      if (likely(PyList_CheckExact(__pyx_t_2))) {
        if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_3); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_7 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_7); __Pyx_INCREF(__pyx_t_3); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_3 = __pyx_t_10(__pyx_t_2);
      if (unlikely(!__pyx_t_3)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_3);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_3);
    __pyx_t_3 = 0;
    __Pyx_INCREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_1);
    __pyx_t_3 = PyNumber_Add(__pyx_t_1, __pyx_int_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_1);
    __pyx_t_1 = __pyx_t_3;
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1095
 *         specifics= [elementLabel for elementLabel in allElements if elementLabel not in nonSpecifics]
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]             # <<<<<<<<<<<<<<
 *             newAtomType = None
 *             element = None
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1095; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = __Pyx_GetItemInt(__pyx_t_3, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1095; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_claimedAtomType, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1096
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]
 *             newAtomType = None             # <<<<<<<<<<<<<<
 *             element = None
 *             #Ignore elements that do not have more than one atomtype
 */
    __Pyx_INCREF(Py_None);
    __Pyx_XDECREF_SET(__pyx_v_newAtomType, Py_None);

    /* "rmgpy/molecule/group.py":1097
 *             claimedAtomType = atom.atomType[0]
 *             newAtomType = None
 *             element = None             # <<<<<<<<<<<<<<
 *             #Ignore elements that do not have more than one atomtype
 *             if claimedAtomType.label in nonSpecifics: continue
 */
    __Pyx_INCREF(Py_None);
    __Pyx_XDECREF_SET(__pyx_v_element, Py_None);

    /* "rmgpy/molecule/group.py":1099
 *             element = None
 *             #Ignore elements that do not have more than one atomtype
 *             if claimedAtomType.label in nonSpecifics: continue             # <<<<<<<<<<<<<<
 *             for elementLabel in specifics:
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 */
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_claimedAtomType, __pyx_n_s_label); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1099; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_nonSpecifics); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1099; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_11 = (__Pyx_PySequence_Contains(__pyx_t_4, __pyx_t_3, Py_EQ)); if (unlikely(__pyx_t_11 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1099; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_5 = (__pyx_t_11 != 0);
    if (__pyx_t_5) {
      goto __pyx_L10_continue;
    }

    /* "rmgpy/molecule/group.py":1100
 *             #Ignore elements that do not have more than one atomtype
 *             if claimedAtomType.label in nonSpecifics: continue
 *             for elementLabel in specifics:             # <<<<<<<<<<<<<<
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 *                     element = atomTypes[elementLabel]
 */
    __pyx_t_3 = __pyx_v_specifics; __Pyx_INCREF(__pyx_t_3); __pyx_t_6 = 0;
    for (;;) {
      if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_3)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_4 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_4); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_4 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1100; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_elementLabel, __pyx_t_4);
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1101
 *             if claimedAtomType.label in nonSpecifics: continue
 *             for elementLabel in specifics:
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:             # <<<<<<<<<<<<<<
 *                     element = atomTypes[elementLabel]
 *                     break
 */
      __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_claimedAtomType, __pyx_n_s_label); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_12 = PyObject_RichCompare(__pyx_t_4, __pyx_v_elementLabel, Py_EQ); __Pyx_XGOTREF(__pyx_t_12); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_11 = __Pyx_PyObject_IsTrue(__pyx_t_12); if (unlikely(__pyx_t_11 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      if (!__pyx_t_11) {
        goto __pyx_L17_next_or;
      } else {
        __pyx_t_5 = __pyx_t_11;
        goto __pyx_L16_bool_binop_done;
      }
      __pyx_L17_next_or:;
      __pyx_t_12 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_12);
      __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_claimedAtomType, __pyx_n_s_label); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_13 = PyObject_GetItem(__pyx_t_12, __pyx_t_4); if (unlikely(__pyx_t_13 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_13);
      __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_12 = PyObject_GetItem(__pyx_t_4, __pyx_v_elementLabel); if (unlikely(__pyx_t_12 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_12);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_t_12, __pyx_n_s_specific); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      __pyx_t_11 = (__Pyx_PySequence_Contains(__pyx_t_13, __pyx_t_4, Py_EQ)); if (unlikely(__pyx_t_11 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1101; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_14 = (__pyx_t_11 != 0);
      __pyx_t_5 = __pyx_t_14;
      __pyx_L16_bool_binop_done:;
      if (__pyx_t_5) {

        /* "rmgpy/molecule/group.py":1102
 *             for elementLabel in specifics:
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 *                     element = atomTypes[elementLabel]             # <<<<<<<<<<<<<<
 *                     break
 * 
 */
        __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1102; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_13 = PyObject_GetItem(__pyx_t_4, __pyx_v_elementLabel); if (unlikely(__pyx_t_13 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1102; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_13);
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_DECREF_SET(__pyx_v_element, __pyx_t_13);
        __pyx_t_13 = 0;

        /* "rmgpy/molecule/group.py":1103
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 *                     element = atomTypes[elementLabel]
 *                     break             # <<<<<<<<<<<<<<
 * 
 *             #claimedAtomType is not in one of the specified elements
 */
        goto __pyx_L14_break;
      }

      /* "rmgpy/molecule/group.py":1100
 *             #Ignore elements that do not have more than one atomtype
 *             if claimedAtomType.label in nonSpecifics: continue
 *             for elementLabel in specifics:             # <<<<<<<<<<<<<<
 *                 if claimedAtomType.label == elementLabel or atomTypes[claimedAtomType.label] in atomTypes[elementLabel].specific:
 *                     element = atomTypes[elementLabel]
 */
    }
    __pyx_L14_break:;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1106
 * 
 *             #claimedAtomType is not in one of the specified elements
 *             if not element: continue             # <<<<<<<<<<<<<<
 *             #Don't standardize atomtypes for nitrogen for now
 *             #The work on the nitrogen atomtypes is still incomplete
 */
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_v_element); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1106; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_14 = ((!__pyx_t_5) != 0);
    if (__pyx_t_14) {
      goto __pyx_L10_continue;
    }

    /* "rmgpy/molecule/group.py":1109
 *             #Don't standardize atomtypes for nitrogen for now
 *             #The work on the nitrogen atomtypes is still incomplete
 *             elif element is atomTypes['N']: continue             # <<<<<<<<<<<<<<
 * 
 *             groupFeatures = getFeatures(atom, atom.bonds)
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1109; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_13 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_N); if (unlikely(__pyx_t_13 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1109; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_13);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_14 = (__pyx_v_element == __pyx_t_13);
    __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
    __pyx_t_5 = (__pyx_t_14 != 0);
    if (__pyx_t_5) {
      goto __pyx_L10_continue;
    }

    /* "rmgpy/molecule/group.py":1111
 *             elif element is atomTypes['N']: continue
 * 
 *             groupFeatures = getFeatures(atom, atom.bonds)             # <<<<<<<<<<<<<<
 * 
 *             bondOrder = atom.getBondOrdersForAtom()
 */
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_getFeatures); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_12 = NULL;
    __pyx_t_6 = 0;
    if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_12 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_12)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_12);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
        __pyx_t_6 = 1;
      }
    }
    __pyx_t_15 = PyTuple_New(2+__pyx_t_6); if (unlikely(!__pyx_t_15)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_15);
    if (__pyx_t_12) {
      PyTuple_SET_ITEM(__pyx_t_15, 0, __pyx_t_12); __Pyx_GIVEREF(__pyx_t_12); __pyx_t_12 = NULL;
    }
    __Pyx_INCREF(__pyx_v_atom);
    PyTuple_SET_ITEM(__pyx_t_15, 0+__pyx_t_6, __pyx_v_atom);
    __Pyx_GIVEREF(__pyx_v_atom);
    PyTuple_SET_ITEM(__pyx_t_15, 1+__pyx_t_6, __pyx_t_4);
    __Pyx_GIVEREF(__pyx_t_4);
    __pyx_t_4 = 0;
    __pyx_t_13 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_15, NULL); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1111; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_13);
    __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_groupFeatures, __pyx_t_13);
    __pyx_t_13 = 0;

    /* "rmgpy/molecule/group.py":1113
 *             groupFeatures = getFeatures(atom, atom.bonds)
 * 
 *             bondOrder = atom.getBondOrdersForAtom()             # <<<<<<<<<<<<<<
 *             filledValency =  atom.radicalElectrons[0] + bondOrder
 * 
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_getBondOrdersForAtom); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1113; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_15 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_15 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_15)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_15);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (__pyx_t_15) {
      __pyx_t_13 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_15); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1113; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
    } else {
      __pyx_t_13 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1113; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_GOTREF(__pyx_t_13);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_bondOrder, __pyx_t_13);
    __pyx_t_13 = 0;

    /* "rmgpy/molecule/group.py":1114
 * 
 *             bondOrder = atom.getBondOrdersForAtom()
 *             filledValency =  atom.radicalElectrons[0] + bondOrder             # <<<<<<<<<<<<<<
 * 
 *             #For an atomtype to be known for certain, the valency must be filled
 */
    __pyx_t_13 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_radicalElectrons); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1114; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_13);
    __pyx_t_3 = __Pyx_GetItemInt(__pyx_t_13, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1114; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
    __pyx_t_13 = PyNumber_Add(__pyx_t_3, __pyx_v_bondOrder); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1114; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_13);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_filledValency, __pyx_t_13);
    __pyx_t_13 = 0;

    /* "rmgpy/molecule/group.py":1118
 *             #For an atomtype to be known for certain, the valency must be filled
 *             #within 1 of the total valency available
 *             if filledValency >= PeriodicSystem.valence_electrons[self.symbol][element] - 1:             # <<<<<<<<<<<<<<
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 */
    __pyx_t_13 = __Pyx_GetModuleGlobalName(__pyx_n_s_PeriodicSystem); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1118; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_13);
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_13, __pyx_n_s_valence_electrons); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1118; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
    __pyx_t_13 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_symbol); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1118; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_13);
    __pyx_t_15 = PyObject_GetItem(__pyx_t_3, __pyx_t_13); if (unlikely(__pyx_t_15 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1118; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_15);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
    __pyx_t_13 = PyObject_GetItem(__pyx_t_15, __pyx_v_element); if (unlikely(__pyx_t_13 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1118; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_13);
    __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
    __pyx_t_15 = PyNumber_Subtract(__pyx_t_13, __pyx_int_1); if (unlikely(!__pyx_t_15)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1118; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_15);
    __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
    __pyx_t_13 = PyObject_RichCompare(__pyx_v_filledValency, __pyx_t_15, Py_GE); __Pyx_XGOTREF(__pyx_t_13); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1118; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_13); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1118; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1119
 *             #within 1 of the total valency available
 *             if filledValency >= PeriodicSystem.valence_electrons[self.symbol][element] - 1:
 *                 for specificAtomType in element.specific:             # <<<<<<<<<<<<<<
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 */
      __pyx_t_13 = __Pyx_PyObject_GetAttrStr(__pyx_v_element, __pyx_n_s_specific); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_13);
      if (likely(PyList_CheckExact(__pyx_t_13)) || PyTuple_CheckExact(__pyx_t_13)) {
        __pyx_t_15 = __pyx_t_13; __Pyx_INCREF(__pyx_t_15); __pyx_t_6 = 0;
        __pyx_t_16 = NULL;
      } else {
        __pyx_t_6 = -1; __pyx_t_15 = PyObject_GetIter(__pyx_t_13); if (unlikely(!__pyx_t_15)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_15);
        __pyx_t_16 = Py_TYPE(__pyx_t_15)->tp_iternext; if (unlikely(!__pyx_t_16)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
      for (;;) {
        if (likely(!__pyx_t_16)) {
          if (likely(PyList_CheckExact(__pyx_t_15))) {
            if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_15)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_13 = PyList_GET_ITEM(__pyx_t_15, __pyx_t_6); __Pyx_INCREF(__pyx_t_13); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_13 = PySequence_ITEM(__pyx_t_15, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          } else {
            if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_15)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_13 = PyTuple_GET_ITEM(__pyx_t_15, __pyx_t_6); __Pyx_INCREF(__pyx_t_13); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_13 = PySequence_ITEM(__pyx_t_15, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          }
        } else {
          __pyx_t_13 = __pyx_t_16(__pyx_t_15);
          if (unlikely(!__pyx_t_13)) {
            PyObject* exc_type = PyErr_Occurred();
            if (exc_type) {
              if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
              else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1119; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            break;
          }
          __Pyx_GOTREF(__pyx_t_13);
        }
        __Pyx_XDECREF_SET(__pyx_v_specificAtomType, __pyx_t_13);
        __pyx_t_13 = 0;

        /* "rmgpy/molecule/group.py":1120
 *             if filledValency >= PeriodicSystem.valence_electrons[self.symbol][element] - 1:
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()             # <<<<<<<<<<<<<<
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 *                         if atomtypeFeature == []:
 */
        __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_specificAtomType, __pyx_n_s_getFeatures); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1120; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_4 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
          __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
          if (likely(__pyx_t_4)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
            __Pyx_INCREF(__pyx_t_4);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_3, function);
          }
        }
        if (__pyx_t_4) {
          __pyx_t_13 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1120; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        } else {
          __pyx_t_13 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1120; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_13);
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_XDECREF_SET(__pyx_v_atomtypeFeatureList, __pyx_t_13);
        __pyx_t_13 = 0;

        /* "rmgpy/molecule/group.py":1121
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):             # <<<<<<<<<<<<<<
 *                         if atomtypeFeature == []:
 *                             continue
 */
        __pyx_t_13 = PyTuple_New(2); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_13);
        __Pyx_INCREF(__pyx_v_groupFeatures);
        PyTuple_SET_ITEM(__pyx_t_13, 0, __pyx_v_groupFeatures);
        __Pyx_GIVEREF(__pyx_v_groupFeatures);
        __Pyx_INCREF(__pyx_v_atomtypeFeatureList);
        PyTuple_SET_ITEM(__pyx_t_13, 1, __pyx_v_atomtypeFeatureList);
        __Pyx_GIVEREF(__pyx_v_atomtypeFeatureList);
        __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_13, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
        if (likely(PyList_CheckExact(__pyx_t_3)) || PyTuple_CheckExact(__pyx_t_3)) {
          __pyx_t_13 = __pyx_t_3; __Pyx_INCREF(__pyx_t_13); __pyx_t_17 = 0;
          __pyx_t_18 = NULL;
        } else {
          __pyx_t_17 = -1; __pyx_t_13 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_13);
          __pyx_t_18 = Py_TYPE(__pyx_t_13)->tp_iternext; if (unlikely(!__pyx_t_18)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        for (;;) {
          if (likely(!__pyx_t_18)) {
            if (likely(PyList_CheckExact(__pyx_t_13))) {
              if (__pyx_t_17 >= PyList_GET_SIZE(__pyx_t_13)) break;
              #if CYTHON_COMPILING_IN_CPYTHON
              __pyx_t_3 = PyList_GET_ITEM(__pyx_t_13, __pyx_t_17); __Pyx_INCREF(__pyx_t_3); __pyx_t_17++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              #else
              __pyx_t_3 = PySequence_ITEM(__pyx_t_13, __pyx_t_17); __pyx_t_17++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              #endif
            } else {
              if (__pyx_t_17 >= PyTuple_GET_SIZE(__pyx_t_13)) break;
              #if CYTHON_COMPILING_IN_CPYTHON
              __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_13, __pyx_t_17); __Pyx_INCREF(__pyx_t_3); __pyx_t_17++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              #else
              __pyx_t_3 = PySequence_ITEM(__pyx_t_13, __pyx_t_17); __pyx_t_17++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              #endif
            }
          } else {
            __pyx_t_3 = __pyx_t_18(__pyx_t_13);
            if (unlikely(!__pyx_t_3)) {
              PyObject* exc_type = PyErr_Occurred();
              if (exc_type) {
                if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
                else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              }
              break;
            }
            __Pyx_GOTREF(__pyx_t_3);
          }
          if ((likely(PyTuple_CheckExact(__pyx_t_3))) || (PyList_CheckExact(__pyx_t_3))) {
            PyObject* sequence = __pyx_t_3;
            #if CYTHON_COMPILING_IN_CPYTHON
            Py_ssize_t size = Py_SIZE(sequence);
            #else
            Py_ssize_t size = PySequence_Size(sequence);
            #endif
            if (unlikely(size != 2)) {
              if (size > 2) __Pyx_RaiseTooManyValuesError(2);
              else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
              {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            #if CYTHON_COMPILING_IN_CPYTHON
            if (likely(PyTuple_CheckExact(sequence))) {
              __pyx_t_4 = PyTuple_GET_ITEM(sequence, 0); 
              __pyx_t_12 = PyTuple_GET_ITEM(sequence, 1); 
            } else {
              __pyx_t_4 = PyList_GET_ITEM(sequence, 0); 
              __pyx_t_12 = PyList_GET_ITEM(sequence, 1); 
            }
            __Pyx_INCREF(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_12);
            #else
            __pyx_t_4 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_4);
            __pyx_t_12 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_12);
            #endif
            __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          } else {
            Py_ssize_t index = -1;
            __pyx_t_19 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_19)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_19);
            __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
            __pyx_t_20 = Py_TYPE(__pyx_t_19)->tp_iternext;
            index = 0; __pyx_t_4 = __pyx_t_20(__pyx_t_19); if (unlikely(!__pyx_t_4)) goto __pyx_L24_unpacking_failed;
            __Pyx_GOTREF(__pyx_t_4);
            index = 1; __pyx_t_12 = __pyx_t_20(__pyx_t_19); if (unlikely(!__pyx_t_12)) goto __pyx_L24_unpacking_failed;
            __Pyx_GOTREF(__pyx_t_12);
            if (__Pyx_IternextUnpackEndCheck(__pyx_t_20(__pyx_t_19), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __pyx_t_20 = NULL;
            __Pyx_DECREF(__pyx_t_19); __pyx_t_19 = 0;
            goto __pyx_L25_unpacking_done;
            __pyx_L24_unpacking_failed:;
            __Pyx_DECREF(__pyx_t_19); __pyx_t_19 = 0;
            __pyx_t_20 = NULL;
            if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
            {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1121; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __pyx_L25_unpacking_done:;
          }
          __Pyx_XDECREF_SET(__pyx_v_molFeature, __pyx_t_4);
          __pyx_t_4 = 0;
          __Pyx_XDECREF_SET(__pyx_v_atomtypeFeature, __pyx_t_12);
          __pyx_t_12 = 0;

          /* "rmgpy/molecule/group.py":1122
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 *                         if atomtypeFeature == []:             # <<<<<<<<<<<<<<
 *                             continue
 *                         elif molFeature not in atomtypeFeature:
 */
          __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1122; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          __pyx_t_12 = PyObject_RichCompare(__pyx_v_atomtypeFeature, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_12); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1122; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_12); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1122; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
          if (__pyx_t_5) {

            /* "rmgpy/molecule/group.py":1123
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 *                         if atomtypeFeature == []:
 *                             continue             # <<<<<<<<<<<<<<
 *                         elif molFeature not in atomtypeFeature:
 *                             break
 */
            goto __pyx_L22_continue;
          }

          /* "rmgpy/molecule/group.py":1124
 *                         if atomtypeFeature == []:
 *                             continue
 *                         elif molFeature not in atomtypeFeature:             # <<<<<<<<<<<<<<
 *                             break
 *                     else:
 */
          __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_molFeature, __pyx_v_atomtypeFeature, Py_NE)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1124; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __pyx_t_14 = (__pyx_t_5 != 0);
          if (__pyx_t_14) {

            /* "rmgpy/molecule/group.py":1125
 *                             continue
 *                         elif molFeature not in atomtypeFeature:
 *                             break             # <<<<<<<<<<<<<<
 *                     else:
 *                         if specificAtomType is atomTypes['Oa'] or specificAtomType is atomTypes['Sa']:
 */
            goto __pyx_L23_break;
          }

          /* "rmgpy/molecule/group.py":1121
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):             # <<<<<<<<<<<<<<
 *                         if atomtypeFeature == []:
 *                             continue
 */
          __pyx_L22_continue:;
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":1127
 *                             break
 *                     else:
 *                         if specificAtomType is atomTypes['Oa'] or specificAtomType is atomTypes['Sa']:             # <<<<<<<<<<<<<<
 *                             if atom.lonePairs == 3 or atom.radicalElectrons == 2:
 *                                 newAtomType = specificAtomType
 */
          __pyx_t_12 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1127; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_12);
          __pyx_t_3 = PyObject_GetItem(__pyx_t_12, __pyx_n_s_Oa); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1127; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
          __pyx_t_5 = (__pyx_v_specificAtomType == __pyx_t_3);
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_t_11 = (__pyx_t_5 != 0);
          if (!__pyx_t_11) {
            goto __pyx_L30_next_or;
          } else {
            __pyx_t_14 = __pyx_t_11;
            goto __pyx_L29_bool_binop_done;
          }
          __pyx_L30_next_or:;
          __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1127; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          __pyx_t_12 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_Sa); if (unlikely(__pyx_t_12 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1127; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_12);
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_t_11 = (__pyx_v_specificAtomType == __pyx_t_12);
          __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
          __pyx_t_5 = (__pyx_t_11 != 0);
          __pyx_t_14 = __pyx_t_5;
          __pyx_L29_bool_binop_done:;
          if (__pyx_t_14) {

            /* "rmgpy/molecule/group.py":1128
 *                     else:
 *                         if specificAtomType is atomTypes['Oa'] or specificAtomType is atomTypes['Sa']:
 *                             if atom.lonePairs == 3 or atom.radicalElectrons == 2:             # <<<<<<<<<<<<<<
 *                                 newAtomType = specificAtomType
 *                                 break
 */
            __pyx_t_12 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_lonePairs); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_12);
            __pyx_t_3 = PyObject_RichCompare(__pyx_t_12, __pyx_int_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
            __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
            if (!__pyx_t_5) {
              goto __pyx_L33_next_or;
            } else {
              __pyx_t_14 = __pyx_t_5;
              goto __pyx_L32_bool_binop_done;
            }
            __pyx_L33_next_or:;
            __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_radicalElectrons); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_3);
            __pyx_t_12 = PyObject_RichCompare(__pyx_t_3, __pyx_int_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_12); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
            __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_12); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1128; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
            __pyx_t_14 = __pyx_t_5;
            __pyx_L32_bool_binop_done:;
            if (__pyx_t_14) {

              /* "rmgpy/molecule/group.py":1129
 *                         if specificAtomType is atomTypes['Oa'] or specificAtomType is atomTypes['Sa']:
 *                             if atom.lonePairs == 3 or atom.radicalElectrons == 2:
 *                                 newAtomType = specificAtomType             # <<<<<<<<<<<<<<
 *                                 break
 *                         else:
 */
              __Pyx_INCREF(__pyx_v_specificAtomType);
              __Pyx_DECREF_SET(__pyx_v_newAtomType, __pyx_v_specificAtomType);

              /* "rmgpy/molecule/group.py":1130
 *                             if atom.lonePairs == 3 or atom.radicalElectrons == 2:
 *                                 newAtomType = specificAtomType
 *                                 break             # <<<<<<<<<<<<<<
 *                         else:
 *                             newAtomType = specificAtomType
 */
              goto __pyx_L21_break;
            }
            goto __pyx_L28;
          }
          /*else*/ {

            /* "rmgpy/molecule/group.py":1132
 *                                 break
 *                         else:
 *                             newAtomType = specificAtomType             # <<<<<<<<<<<<<<
 *                             break
 * 
 */
            __Pyx_INCREF(__pyx_v_specificAtomType);
            __Pyx_DECREF_SET(__pyx_v_newAtomType, __pyx_v_specificAtomType);

            /* "rmgpy/molecule/group.py":1133
 *                         else:
 *                             newAtomType = specificAtomType
 *                             break             # <<<<<<<<<<<<<<
 * 
 *             #set the new atom type if the algorithm found one
 */
            goto __pyx_L21_break;
          }
          __pyx_L28:;
        }

        /* "rmgpy/molecule/group.py":1121
 *                 for specificAtomType in element.specific:
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):             # <<<<<<<<<<<<<<
 *                         if atomtypeFeature == []:
 *                             continue
 */
        __pyx_L23_break:;
        __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;

        /* "rmgpy/molecule/group.py":1119
 *             #within 1 of the total valency available
 *             if filledValency >= PeriodicSystem.valence_electrons[self.symbol][element] - 1:
 *                 for specificAtomType in element.specific:             # <<<<<<<<<<<<<<
 *                     atomtypeFeatureList = specificAtomType.getFeatures()
 *                     for molFeature, atomtypeFeature in zip(groupFeatures, atomtypeFeatureList):
 */
      }
      __pyx_L21_break:;
      __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
      goto __pyx_L19;
    }
    __pyx_L19:;

    /* "rmgpy/molecule/group.py":1136
 * 
 *             #set the new atom type if the algorithm found one
 *             if newAtomType and newAtomType is not claimedAtomType:             # <<<<<<<<<<<<<<
 *                 atom.atomType[0] = newAtomType
 *                 modified = True
 */
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_v_newAtomType); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1136; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (__pyx_t_5) {
      goto __pyx_L36_next_and;
    } else {
      __pyx_t_14 = __pyx_t_5;
      goto __pyx_L35_bool_binop_done;
    }
    __pyx_L36_next_and:;
    __pyx_t_5 = (__pyx_v_newAtomType != __pyx_v_claimedAtomType);
    __pyx_t_11 = (__pyx_t_5 != 0);
    __pyx_t_14 = __pyx_t_11;
    __pyx_L35_bool_binop_done:;
    if (__pyx_t_14) {

      /* "rmgpy/molecule/group.py":1137
 *             #set the new atom type if the algorithm found one
 *             if newAtomType and newAtomType is not claimedAtomType:
 *                 atom.atomType[0] = newAtomType             # <<<<<<<<<<<<<<
 *                 modified = True
 * 
 */
      __pyx_t_15 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_15)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1137; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_15);
      if (unlikely(__Pyx_SetItemInt(__pyx_t_15, 0, __pyx_v_newAtomType, long, 1, __Pyx_PyInt_From_long, 0, 0, 1) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1137; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;

      /* "rmgpy/molecule/group.py":1138
 *             if newAtomType and newAtomType is not claimedAtomType:
 *                 atom.atomType[0] = newAtomType
 *                 modified = True             # <<<<<<<<<<<<<<
 * 
 *         return modified
 */
      __pyx_v_modified = 1;
      goto __pyx_L34;
    }
    __pyx_L34:;

    /* "rmgpy/molecule/group.py":1094
 *         #list of :class:AtomType which are elements with more sub-divided atomtypes beneath them
 *         specifics= [elementLabel for elementLabel in allElements if elementLabel not in nonSpecifics]
 *         for index, atom in enumerate(self.atoms):             # <<<<<<<<<<<<<<
 *             claimedAtomType = atom.atomType[0]
 *             newAtomType = None
 */
    __pyx_L10_continue:;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1140
 *                 modified = True
 * 
 *         return modified             # <<<<<<<<<<<<<<
 * 
 *     def addExplicitLigands(self):
 */
  __pyx_r = __pyx_v_modified;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1068
 *         return True
 * 
 *     def standardizeAtomType(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function changes the atomTypes in a group if the atom must
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_12);
  __Pyx_XDECREF(__pyx_t_13);
  __Pyx_XDECREF(__pyx_t_15);
  __Pyx_XDECREF(__pyx_t_19);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.standardizeAtomType", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_specifics);
  __Pyx_XDECREF(__pyx_v_index);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_claimedAtomType);
  __Pyx_XDECREF(__pyx_v_newAtomType);
  __Pyx_XDECREF(__pyx_v_element);
  __Pyx_XDECREF(__pyx_v_elementLabel);
  __Pyx_XDECREF(__pyx_v_groupFeatures);
  __Pyx_XDECREF(__pyx_v_bondOrder);
  __Pyx_XDECREF(__pyx_v_filledValency);
  __Pyx_XDECREF(__pyx_v_specificAtomType);
  __Pyx_XDECREF(__pyx_v_atomtypeFeatureList);
  __Pyx_XDECREF(__pyx_v_molFeature);
  __Pyx_XDECREF(__pyx_v_atomtypeFeature);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_61standardizeAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_60standardizeAtomType[] = "\n        This function changes the atomTypes in a group if the atom must\n        be a specific atomType based on its bonds and valency.\n\n        Currently only standardizes oxygen, carbon and sulfur atomTypes\n\n        We also only check when there is exactly one atomType,\n        one bondType, one radical setting.\n        For any group where there are wildcards or multiple attributes,\n        we cannot apply this check.\n\n        In the case where the atomType is ambigious based on bonds\n        and valency, this function will not change the type.\n\n        Returns a 'True' if the group was modified otherwise returns 'False'\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_61standardizeAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("standardizeAtomType (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_60standardizeAtomType(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_60standardizeAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("standardizeAtomType", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_standardizeAtomType(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1068; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.standardizeAtomType", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1142
 *         return modified
 * 
 *     def addExplicitLigands(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function Od/Sd ligand to CO or CS atomtypes if they are not already there.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_63addExplicitLigands(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_addExplicitLigands(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  int __pyx_v_modified;
  PyObject *__pyx_v_atomsToAddTo = NULL;
  PyObject *__pyx_v_index = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_claimedAtomType = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_v_atomIndex = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_newAtom = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_newBond = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  PyObject *(*__pyx_t_7)(PyObject *);
  int __pyx_t_8;
  int __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  Py_ssize_t __pyx_t_11;
  int __pyx_t_12;
  PyObject *__pyx_t_13 = NULL;
  int __pyx_t_14;
  PyObject *__pyx_t_15 = NULL;
  int __pyx_t_16;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addExplicitLigands", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_addExplicitLigands); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1142; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_63addExplicitLigands)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1142; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1142; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1142; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1149
 *         """
 * 
 *         modified = False             # <<<<<<<<<<<<<<
 * 
 *         atomsToAddTo=[]
 */
  __pyx_v_modified = 0;

  /* "rmgpy/molecule/group.py":1151
 *         modified = False
 * 
 *         atomsToAddTo=[]             # <<<<<<<<<<<<<<
 * 
 *         for index, atom in enumerate(self.atoms):
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1151; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomsToAddTo = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1153
 *         atomsToAddTo=[]
 * 
 *         for index, atom in enumerate(self.atoms):             # <<<<<<<<<<<<<<
 *             claimedAtomType = atom.atomType[0]
 *             #Do not perform is this atom has wildCards
 */
  __Pyx_INCREF(__pyx_int_0);
  __pyx_t_1 = __pyx_int_0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_3 = __pyx_t_2; __Pyx_INCREF(__pyx_t_3); __pyx_t_6 = 0;
    __pyx_t_7 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_7 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_7)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_7(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;
    __Pyx_INCREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_1);
    __pyx_t_2 = PyNumber_Add(__pyx_t_1, __pyx_int_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1153; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1);
    __pyx_t_1 = __pyx_t_2;
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1154
 * 
 *         for index, atom in enumerate(self.atoms):
 *             claimedAtomType = atom.atomType[0]             # <<<<<<<<<<<<<<
 *             #Do not perform is this atom has wildCards
 *             if atom.hasWildCards: continue
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1154; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_4 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1154; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_XDECREF_SET(__pyx_v_claimedAtomType, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1156
 *             claimedAtomType = atom.atomType[0]
 *             #Do not perform is this atom has wildCards
 *             if atom.hasWildCards: continue             # <<<<<<<<<<<<<<
 *             elif claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:
 *                 for atom2, bond12 in atom.bonds.iteritems():
 */
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_hasWildCards); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1156; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1156; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    if (__pyx_t_5) {
      goto __pyx_L3_continue;
    }

    /* "rmgpy/molecule/group.py":1157
 *             #Do not perform is this atom has wildCards
 *             if atom.hasWildCards: continue
 *             elif claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:             # <<<<<<<<<<<<<<
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isDouble():
 */
    __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1157; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_2 = PyObject_GetItem(__pyx_t_4, __pyx_n_s_CO); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1157; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_8 = (__pyx_v_claimedAtomType == __pyx_t_2);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_9 = (__pyx_t_8 != 0);
    if (!__pyx_t_9) {
      goto __pyx_L7_next_or;
    } else {
      __pyx_t_5 = __pyx_t_9;
      goto __pyx_L6_bool_binop_done;
    }
    __pyx_L7_next_or:;
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1157; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_4 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_CS); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1157; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_9 = (__pyx_v_claimedAtomType == __pyx_t_4);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_8 = (__pyx_t_9 != 0);
    __pyx_t_5 = __pyx_t_8;
    __pyx_L6_bool_binop_done:;
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1158
 *             if atom.hasWildCards: continue
 *             elif claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:
 *                 for atom2, bond12 in atom.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                     if bond12.isDouble():
 *                         break
 */
      __pyx_t_10 = 0;
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1158; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      if (unlikely(__pyx_t_2 == Py_None)) {
        PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1158; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_13 = __Pyx_dict_iterator(__pyx_t_2, 0, __pyx_n_s_iteritems, (&__pyx_t_11), (&__pyx_t_12)); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1158; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_13);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_XDECREF(__pyx_t_4);
      __pyx_t_4 = __pyx_t_13;
      __pyx_t_13 = 0;
      while (1) {
        __pyx_t_14 = __Pyx_dict_iter_next(__pyx_t_4, __pyx_t_11, &__pyx_t_10, &__pyx_t_13, &__pyx_t_2, NULL, __pyx_t_12);
        if (unlikely(__pyx_t_14 == 0)) break;
        if (unlikely(__pyx_t_14 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1158; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_13);
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_13);
        __pyx_t_13 = 0;
        __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_2);
        __pyx_t_2 = 0;

        /* "rmgpy/molecule/group.py":1159
 *             elif claimedAtomType is atomTypes['CO'] or claimedAtomType is atomTypes['CS']:
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isDouble():             # <<<<<<<<<<<<<<
 *                         break
 *                 else: atomsToAddTo.append(index)
 */
        __pyx_t_13 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isDouble); if (unlikely(!__pyx_t_13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_13);
        __pyx_t_15 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_13))) {
          __pyx_t_15 = PyMethod_GET_SELF(__pyx_t_13);
          if (likely(__pyx_t_15)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_13);
            __Pyx_INCREF(__pyx_t_15);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_13, function);
          }
        }
        if (__pyx_t_15) {
          __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_13, __pyx_t_15); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
        } else {
          __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_13); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_13); __pyx_t_13 = 0;
        __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1159; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        if (__pyx_t_5) {

          /* "rmgpy/molecule/group.py":1160
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isDouble():
 *                         break             # <<<<<<<<<<<<<<
 *                 else: atomsToAddTo.append(index)
 * 
 */
          goto __pyx_L9_break;
        }
      }

      /* "rmgpy/molecule/group.py":1161
 *                     if bond12.isDouble():
 *                         break
 *                 else: atomsToAddTo.append(index)             # <<<<<<<<<<<<<<
 * 
 *         for atomIndex in atomsToAddTo:
 */
      /*else*/ {
        __pyx_t_16 = __Pyx_PyList_Append(__pyx_v_atomsToAddTo, __pyx_v_index); if (unlikely(__pyx_t_16 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1161; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_L9_break:;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      goto __pyx_L5;
    }
    __pyx_L5:;

    /* "rmgpy/molecule/group.py":1153
 *         atomsToAddTo=[]
 * 
 *         for index, atom in enumerate(self.atoms):             # <<<<<<<<<<<<<<
 *             claimedAtomType = atom.atomType[0]
 *             #Do not perform is this atom has wildCards
 */
    __pyx_L3_continue:;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1163
 *                 else: atomsToAddTo.append(index)
 * 
 *         for atomIndex in atomsToAddTo:             # <<<<<<<<<<<<<<
 *             modified = True
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 */
  __pyx_t_1 = __pyx_v_atomsToAddTo; __Pyx_INCREF(__pyx_t_1); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_3 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_6); __Pyx_INCREF(__pyx_t_3); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1163; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1163; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atomIndex, __pyx_t_3);
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1164
 * 
 *         for atomIndex in atomsToAddTo:
 *             modified = True             # <<<<<<<<<<<<<<
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
    __pyx_v_modified = 1;

    /* "rmgpy/molecule/group.py":1165
 *         for atomIndex in atomsToAddTo:
 *             modified = True
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = PyObject_GetItem(__pyx_t_3, __pyx_v_atomIndex); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1165; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_4, __pyx_n_s_atomType); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_GetItemInt(__pyx_t_3, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1165; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_2 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_CO); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1165; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_5 = (__pyx_t_4 == __pyx_t_2);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_8 = (__pyx_t_5 != 0);
    if (__pyx_t_8) {

      /* "rmgpy/molecule/group.py":1166
 *             modified = True
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
      __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_GetItem(__pyx_t_4, __pyx_n_s_Od); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyList_SET_ITEM(__pyx_t_4, 0, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_atomType, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_INCREF(__pyx_int_0);
      PyList_SET_ITEM(__pyx_t_4, 0, __pyx_int_0);
      __Pyx_GIVEREF(__pyx_int_0);
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_radicalElectrons, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_charge, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_2); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1166; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_4));
      __pyx_t_4 = 0;
      goto __pyx_L13;
    }

    /* "rmgpy/molecule/group.py":1167
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             self.addAtom(newAtom)
 */
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1167; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_2 = PyObject_GetItem(__pyx_t_4, __pyx_v_atomIndex); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1167; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1167; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_GetItemInt(__pyx_t_4, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1167; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1167; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_3 = PyObject_GetItem(__pyx_t_4, __pyx_n_s_CS); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1167; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_8 = (__pyx_t_2 == __pyx_t_3);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_5 = (__pyx_t_8 != 0);
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1168
 *                 newAtom = GroupAtom(atomType=[atomTypes['Od']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *             self.addAtom(newAtom)
 *             newBond = GroupBond(self.atoms[atomIndex], newAtom, order=['D'])
 */
      __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_4 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_Sd); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      PyList_SET_ITEM(__pyx_t_2, 0, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_atomType, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_INCREF(__pyx_int_0);
      PyList_SET_ITEM(__pyx_t_2, 0, __pyx_int_0);
      __Pyx_GIVEREF(__pyx_int_0);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_radicalElectrons, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_charge, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1168; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_2));
      __pyx_t_2 = 0;
      goto __pyx_L13;
    }
    __pyx_L13:;

    /* "rmgpy/molecule/group.py":1169
 *             elif self.atoms[atomIndex].atomType[0] is atomTypes['CS']:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             self.addAtom(newAtom)             # <<<<<<<<<<<<<<
 *             newBond = GroupBond(self.atoms[atomIndex], newAtom, order=['D'])
 *             self.addBond(newBond)
 */
    if (unlikely(!__pyx_v_newAtom)) { __Pyx_RaiseUnboundLocalError("newAtom"); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1169; __pyx_clineno = __LINE__; goto __pyx_L1_error;} }
    __pyx_t_2 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->addAtom(__pyx_v_self, __pyx_v_newAtom, 0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1169; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1170
 *                 newAtom = GroupAtom(atomType=[atomTypes['Sd']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             self.addAtom(newAtom)
 *             newBond = GroupBond(self.atoms[atomIndex], newAtom, order=['D'])             # <<<<<<<<<<<<<<
 *             self.addBond(newBond)
 * 
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_v_atomIndex); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1170; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (unlikely(!__pyx_v_newAtom)) { __Pyx_RaiseUnboundLocalError("newAtom"); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1170; __pyx_clineno = __LINE__; goto __pyx_L1_error;} }
    __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_3);
    __Pyx_GIVEREF(__pyx_t_3);
    __Pyx_INCREF(((PyObject *)__pyx_v_newAtom));
    PyTuple_SET_ITEM(__pyx_t_2, 1, ((PyObject *)__pyx_v_newAtom));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_newAtom));
    __pyx_t_3 = 0;
    __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_INCREF(__pyx_n_s_D);
    PyList_SET_ITEM(__pyx_t_4, 0, __pyx_n_s_D);
    __Pyx_GIVEREF(__pyx_n_s_D);
    if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_order, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1170; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_newBond, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_4));
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1171
 *             self.addAtom(newAtom)
 *             newBond = GroupBond(self.atoms[atomIndex], newAtom, order=['D'])
 *             self.addBond(newBond)             # <<<<<<<<<<<<<<
 * 
 *         return modified
 */
    __pyx_t_4 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->addBond(__pyx_v_self, __pyx_v_newBond, 0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1171; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1163
 *                 else: atomsToAddTo.append(index)
 * 
 *         for atomIndex in atomsToAddTo:             # <<<<<<<<<<<<<<
 *             modified = True
 *             if self.atoms[atomIndex].atomType[0] is atomTypes['CO']:
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1173
 *             self.addBond(newBond)
 * 
 *         return modified             # <<<<<<<<<<<<<<
 * 
 *     def standardizeGroup(self):
 */
  __pyx_r = __pyx_v_modified;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1142
 *         return modified
 * 
 *     def addExplicitLigands(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function Od/Sd ligand to CO or CS atomtypes if they are not already there.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_13);
  __Pyx_XDECREF(__pyx_t_15);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.addExplicitLigands", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_atomsToAddTo);
  __Pyx_XDECREF(__pyx_v_index);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_claimedAtomType);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XDECREF(__pyx_v_atomIndex);
  __Pyx_XDECREF((PyObject *)__pyx_v_newAtom);
  __Pyx_XDECREF((PyObject *)__pyx_v_newBond);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_63addExplicitLigands(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_62addExplicitLigands[] = "\n        This function Od/Sd ligand to CO or CS atomtypes if they are not already there.\n\n        Returns a 'True' if the group was modified otherwise returns 'False'\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_63addExplicitLigands(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addExplicitLigands (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_62addExplicitLigands(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_62addExplicitLigands(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addExplicitLigands", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_addExplicitLigands(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1142; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addExplicitLigands", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1175
 *         return modified
 * 
 *     def standardizeGroup(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function modifies groups to make them have a standard AdjList form.
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_65standardizeGroup(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_standardizeGroup(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  CYTHON_UNUSED int __pyx_v_modified;
  PyObject *__pyx_v_checkList = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("standardizeGroup", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_standardizeGroup); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1175; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_65standardizeGroup)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1175; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1175; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1175; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1185
 *         """
 * 
 *         modified = False             # <<<<<<<<<<<<<<
 * 
 *         #If viable then we apply current conventions:
 */
  __pyx_v_modified = 0;

  /* "rmgpy/molecule/group.py":1188
 * 
 *         #If viable then we apply current conventions:
 *         checkList=[]             # <<<<<<<<<<<<<<
 *         checkList.append(self.standardizeAtomType())
 *         checkList.append(self.addExplicitLigands())
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1188; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_checkList = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1189
 *         #If viable then we apply current conventions:
 *         checkList=[]
 *         checkList.append(self.standardizeAtomType())             # <<<<<<<<<<<<<<
 *         checkList.append(self.addExplicitLigands())
 *         return any(checkList)
 */
  __pyx_t_1 = __Pyx_PyBool_FromLong(((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->standardizeAtomType(__pyx_v_self, 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1189; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_checkList, __pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1189; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1190
 *         checkList=[]
 *         checkList.append(self.standardizeAtomType())
 *         checkList.append(self.addExplicitLigands())             # <<<<<<<<<<<<<<
 *         return any(checkList)
 * 
 */
  __pyx_t_1 = __Pyx_PyBool_FromLong(((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->addExplicitLigands(__pyx_v_self, 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1190; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_checkList, __pyx_t_1); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1190; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1191
 *         checkList.append(self.standardizeAtomType())
 *         checkList.append(self.addExplicitLigands())
 *         return any(checkList)             # <<<<<<<<<<<<<<
 * 
 * 
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1191; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_checkList);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_checkList);
  __Pyx_GIVEREF(__pyx_v_checkList);
  __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_any, __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1191; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1191; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_5;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1175
 *         return modified
 * 
 *     def standardizeGroup(self):             # <<<<<<<<<<<<<<
 *         """
 *         This function modifies groups to make them have a standard AdjList form.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.standardizeGroup", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_checkList);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_65standardizeGroup(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_64standardizeGroup[] = "\n        This function modifies groups to make them have a standard AdjList form.\n\n        Currently it makes atomtypes as specific as possible and makes CO/CS atomtypes\n        have explicit Od/Sd ligands. Other functions can be added as necessary\n\n        Returns a 'True' if the group was modified otherwise returns 'False'\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_65standardizeGroup(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("standardizeGroup (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_64standardizeGroup(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_64standardizeGroup(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("standardizeGroup", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_standardizeGroup(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1175; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.standardizeGroup", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1194
 * 
 * 
 *     def addImplicitAtomsFromAtomType(self):             # <<<<<<<<<<<<<<
 *         """
 * 
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_67addImplicitAtomsFromAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_f_5rmgpy_8molecule_5group_5Group_addImplicitAtomsFromAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_implicitAtoms = NULL;
  PyObject *__pyx_v_lonePairsRequired = NULL;
  PyObject *__pyx_v_copyGroup = NULL;
  PyObject *__pyx_v_atom1 = NULL;
  PyObject *__pyx_v_atomtypeFeatureList = NULL;
  PyObject *__pyx_v_allDoubleRequired = NULL;
  PyObject *__pyx_v_rDoubleRequired = NULL;
  PyObject *__pyx_v_oDoubleRequired = NULL;
  PyObject *__pyx_v_sDoubleRequired = NULL;
  PyObject *__pyx_v_tripleRequired = NULL;
  PyObject *__pyx_v_single = NULL;
  PyObject *__pyx_v_rDouble = NULL;
  PyObject *__pyx_v_oDouble = NULL;
  PyObject *__pyx_v_sDouble = NULL;
  PyObject *__pyx_v_triple = NULL;
  PyObject *__pyx_v_benzene = NULL;
  PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_bond12 = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *__pyx_v_newAtom = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_newBond = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_bond = NULL;
  PyObject *__pyx_v_lonePair = NULL;
  PyObject *__pyx_v_featureList = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *(*__pyx_t_6)(PyObject *);
  Py_ssize_t __pyx_t_7;
  PyObject *(*__pyx_t_8)(PyObject *);
  int __pyx_t_9;
  PyObject *__pyx_t_10 = NULL;
  Py_ssize_t __pyx_t_11;
  int __pyx_t_12;
  int __pyx_t_13;
  int __pyx_t_14;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addImplicitAtomsFromAtomType", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_addImplicitAtomsFromAtomType); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1194; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_67addImplicitAtomsFromAtomType)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1194; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1194; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1194; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1204
 * 
 *         #dictionary of implicit atoms and their bonds
 *         implicitAtoms = {}             # <<<<<<<<<<<<<<
 *         lonePairsRequired = {}
 * 
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1204; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_implicitAtoms = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1205
 *         #dictionary of implicit atoms and their bonds
 *         implicitAtoms = {}
 *         lonePairsRequired = {}             # <<<<<<<<<<<<<<
 * 
 *         copyGroup = deepcopy(self)
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1205; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_lonePairsRequired = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1207
 *         lonePairsRequired = {}
 * 
 *         copyGroup = deepcopy(self)             # <<<<<<<<<<<<<<
 * 
 *         for atom1 in copyGroup.atoms:
 */
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_deepcopy); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, ((PyObject *)__pyx_v_self)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(((PyObject *)__pyx_v_self));
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, ((PyObject *)__pyx_v_self));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1207; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_copyGroup = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1209
 *         copyGroup = deepcopy(self)
 * 
 *         for atom1 in copyGroup.atoms:             # <<<<<<<<<<<<<<
 *             atomtypeFeatureList = atom1.atomType[0].getFeatures()
 *             lonePairsRequired[atom1]=atomtypeFeatureList[7]
 */
  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_copyGroup, __pyx_n_s_atoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
    __pyx_t_2 = __pyx_t_1; __Pyx_INCREF(__pyx_t_2); __pyx_t_5 = 0;
    __pyx_t_6 = NULL;
  } else {
    __pyx_t_5 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (;;) {
    if (likely(!__pyx_t_6)) {
      if (likely(PyList_CheckExact(__pyx_t_2))) {
        if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_1); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_5 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_1); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_1 = __pyx_t_6(__pyx_t_2);
      if (unlikely(!__pyx_t_1)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1209; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_1);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom1, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1210
 * 
 *         for atom1 in copyGroup.atoms:
 *             atomtypeFeatureList = atom1.atomType[0].getFeatures()             # <<<<<<<<<<<<<<
 *             lonePairsRequired[atom1]=atomtypeFeatureList[7]
 * 
 */
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom1, __pyx_n_s_atomType); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_3 = __Pyx_GetItemInt(__pyx_t_4, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1210; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_getFeatures); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_4);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_4, function);
      }
    }
    if (__pyx_t_3) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    } else {
      __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1210; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_XDECREF_SET(__pyx_v_atomtypeFeatureList, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1211
 *         for atom1 in copyGroup.atoms:
 *             atomtypeFeatureList = atom1.atomType[0].getFeatures()
 *             lonePairsRequired[atom1]=atomtypeFeatureList[7]             # <<<<<<<<<<<<<<
 * 
 *             #set to 0 required if empty list
 */
    __pyx_t_1 = __Pyx_GetItemInt(__pyx_v_atomtypeFeatureList, 7, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1211; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    if (unlikely(PyDict_SetItem(__pyx_v_lonePairsRequired, __pyx_v_atom1, __pyx_t_1) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1211; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1214
 * 
 *             #set to 0 required if empty list
 *             atomtypeFeatureList = [featureList if featureList else [0] for featureList in atomtypeFeatureList]             # <<<<<<<<<<<<<<
 *             allDoubleRequired = atomtypeFeatureList[1]
 *             rDoubleRequired = atomtypeFeatureList[2]
 */
    __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (likely(PyList_CheckExact(__pyx_v_atomtypeFeatureList)) || PyTuple_CheckExact(__pyx_v_atomtypeFeatureList)) {
      __pyx_t_4 = __pyx_v_atomtypeFeatureList; __Pyx_INCREF(__pyx_t_4); __pyx_t_7 = 0;
      __pyx_t_8 = NULL;
    } else {
      __pyx_t_7 = -1; __pyx_t_4 = PyObject_GetIter(__pyx_v_atomtypeFeatureList); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_8 = Py_TYPE(__pyx_t_4)->tp_iternext; if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    for (;;) {
      if (likely(!__pyx_t_8)) {
        if (likely(PyList_CheckExact(__pyx_t_4))) {
          if (__pyx_t_7 >= PyList_GET_SIZE(__pyx_t_4)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_7); __Pyx_INCREF(__pyx_t_3); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_4, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        } else {
          if (__pyx_t_7 >= PyTuple_GET_SIZE(__pyx_t_4)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_4, __pyx_t_7); __Pyx_INCREF(__pyx_t_3); __pyx_t_7++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_3 = PySequence_ITEM(__pyx_t_4, __pyx_t_7); __pyx_t_7++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        }
      } else {
        __pyx_t_3 = __pyx_t_8(__pyx_t_4);
        if (unlikely(!__pyx_t_3)) {
          PyObject* exc_type = PyErr_Occurred();
          if (exc_type) {
            if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
            else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          break;
        }
        __Pyx_GOTREF(__pyx_t_3);
      }
      __Pyx_XDECREF_SET(__pyx_v_featureList, __pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_v_featureList); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_9) {
        __Pyx_INCREF(__pyx_v_featureList);
        __pyx_t_3 = __pyx_v_featureList;
      } else {
        __pyx_t_10 = PyList_New(1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_10);
        __Pyx_INCREF(__pyx_int_0);
        PyList_SET_ITEM(__pyx_t_10, 0, __pyx_int_0);
        __Pyx_GIVEREF(__pyx_int_0);
        __pyx_t_3 = __pyx_t_10;
        __pyx_t_10 = 0;
      }
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_3))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1214; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    }
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF_SET(__pyx_v_atomtypeFeatureList, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1215
 *             #set to 0 required if empty list
 *             atomtypeFeatureList = [featureList if featureList else [0] for featureList in atomtypeFeatureList]
 *             allDoubleRequired = atomtypeFeatureList[1]             # <<<<<<<<<<<<<<
 *             rDoubleRequired = atomtypeFeatureList[2]
 *             oDoubleRequired = atomtypeFeatureList[3]
 */
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_atomtypeFeatureList, 1, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1215; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_allDoubleRequired, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1216
 *             atomtypeFeatureList = [featureList if featureList else [0] for featureList in atomtypeFeatureList]
 *             allDoubleRequired = atomtypeFeatureList[1]
 *             rDoubleRequired = atomtypeFeatureList[2]             # <<<<<<<<<<<<<<
 *             oDoubleRequired = atomtypeFeatureList[3]
 *             sDoubleRequired = atomtypeFeatureList[4]
 */
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_atomtypeFeatureList, 2, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1216; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_rDoubleRequired, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1217
 *             allDoubleRequired = atomtypeFeatureList[1]
 *             rDoubleRequired = atomtypeFeatureList[2]
 *             oDoubleRequired = atomtypeFeatureList[3]             # <<<<<<<<<<<<<<
 *             sDoubleRequired = atomtypeFeatureList[4]
 *             tripleRequired = atomtypeFeatureList[5]
 */
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_atomtypeFeatureList, 3, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1217; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_oDoubleRequired, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1218
 *             rDoubleRequired = atomtypeFeatureList[2]
 *             oDoubleRequired = atomtypeFeatureList[3]
 *             sDoubleRequired = atomtypeFeatureList[4]             # <<<<<<<<<<<<<<
 *             tripleRequired = atomtypeFeatureList[5]
 * 
 */
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_atomtypeFeatureList, 4, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1218; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_sDoubleRequired, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1219
 *             oDoubleRequired = atomtypeFeatureList[3]
 *             sDoubleRequired = atomtypeFeatureList[4]
 *             tripleRequired = atomtypeFeatureList[5]             # <<<<<<<<<<<<<<
 * 
 *             #count up number of bonds
 */
    __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_atomtypeFeatureList, 5, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1219; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_XDECREF_SET(__pyx_v_tripleRequired, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1222
 * 
 *             #count up number of bonds
 *             single = 0; rDouble = 0; oDouble = 0; sDouble = 0; triple = 0; benzene = 0             # <<<<<<<<<<<<<<
 *             for atom2, bond12 in atom1.bonds.iteritems():
 *                 # Count numbers of each higher-order bond type
 */
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_single, __pyx_int_0);
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_rDouble, __pyx_int_0);
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_oDouble, __pyx_int_0);
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_sDouble, __pyx_int_0);
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_triple, __pyx_int_0);
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_benzene, __pyx_int_0);

    /* "rmgpy/molecule/group.py":1223
 *             #count up number of bonds
 *             single = 0; rDouble = 0; oDouble = 0; sDouble = 0; triple = 0; benzene = 0
 *             for atom2, bond12 in atom1.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                 # Count numbers of each higher-order bond type
 *                 if bond12.isSingle():
 */
    __pyx_t_7 = 0;
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom1, __pyx_n_s_bonds); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    if (unlikely(__pyx_t_4 == Py_None)) {
      PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_3 = __Pyx_dict_iterator(__pyx_t_4, 0, __pyx_n_s_iteritems, (&__pyx_t_11), (&__pyx_t_12)); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_XDECREF(__pyx_t_1);
    __pyx_t_1 = __pyx_t_3;
    __pyx_t_3 = 0;
    while (1) {
      __pyx_t_13 = __Pyx_dict_iter_next(__pyx_t_1, __pyx_t_11, &__pyx_t_7, &__pyx_t_3, &__pyx_t_4, NULL, __pyx_t_12);
      if (unlikely(__pyx_t_13 == 0)) break;
      if (unlikely(__pyx_t_13 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1223; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_3);
      __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_4);
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1225
 *             for atom2, bond12 in atom1.bonds.iteritems():
 *                 # Count numbers of each higher-order bond type
 *                 if bond12.isSingle():             # <<<<<<<<<<<<<<
 *                     single += 1
 *                 elif bond12.isDouble():
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isSingle); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_10 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_10)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_10);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_10) {
        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_10); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      } else {
        __pyx_t_4 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1225; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_9) {

        /* "rmgpy/molecule/group.py":1226
 *                 # Count numbers of each higher-order bond type
 *                 if bond12.isSingle():
 *                     single += 1             # <<<<<<<<<<<<<<
 *                 elif bond12.isDouble():
 *                     if atom2.isOxygen():
 */
        __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_single, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1226; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF_SET(__pyx_v_single, __pyx_t_4);
        __pyx_t_4 = 0;
        goto __pyx_L9;
      }

      /* "rmgpy/molecule/group.py":1227
 *                 if bond12.isSingle():
 *                     single += 1
 *                 elif bond12.isDouble():             # <<<<<<<<<<<<<<
 *                     if atom2.isOxygen():
 *                         oDouble += 1
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isDouble); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1227; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_10 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_10)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_10);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_10) {
        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_10); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1227; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      } else {
        __pyx_t_4 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1227; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1227; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_9) {

        /* "rmgpy/molecule/group.py":1228
 *                     single += 1
 *                 elif bond12.isDouble():
 *                     if atom2.isOxygen():             # <<<<<<<<<<<<<<
 *                         oDouble += 1
 *                     elif atom2.isSulfur():
 */
        __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom2, __pyx_n_s_isOxygen); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_10 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
          __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_3);
          if (likely(__pyx_t_10)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
            __Pyx_INCREF(__pyx_t_10);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_3, function);
          }
        }
        if (__pyx_t_10) {
          __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_10); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
        } else {
          __pyx_t_4 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1228; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        if (__pyx_t_9) {

          /* "rmgpy/molecule/group.py":1229
 *                 elif bond12.isDouble():
 *                     if atom2.isOxygen():
 *                         oDouble += 1             # <<<<<<<<<<<<<<
 *                     elif atom2.isSulfur():
 *                         sDouble += 1
 */
          __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_oDouble, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1229; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_4);
          __Pyx_DECREF_SET(__pyx_v_oDouble, __pyx_t_4);
          __pyx_t_4 = 0;
          goto __pyx_L10;
        }

        /* "rmgpy/molecule/group.py":1230
 *                     if atom2.isOxygen():
 *                         oDouble += 1
 *                     elif atom2.isSulfur():             # <<<<<<<<<<<<<<
 *                         sDouble += 1
 *                     else:
 */
        __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom2, __pyx_n_s_isSulfur); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_10 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
          __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_3);
          if (likely(__pyx_t_10)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
            __Pyx_INCREF(__pyx_t_10);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_3, function);
          }
        }
        if (__pyx_t_10) {
          __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_10); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
        } else {
          __pyx_t_4 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1230; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        if (__pyx_t_9) {

          /* "rmgpy/molecule/group.py":1231
 *                         oDouble += 1
 *                     elif atom2.isSulfur():
 *                         sDouble += 1             # <<<<<<<<<<<<<<
 *                     else:
 *                         # rDouble is for double bonds NOT to oxygen or Sulfur
 */
          __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_sDouble, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1231; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_4);
          __Pyx_DECREF_SET(__pyx_v_sDouble, __pyx_t_4);
          __pyx_t_4 = 0;
          goto __pyx_L10;
        }
        /*else*/ {

          /* "rmgpy/molecule/group.py":1234
 *                     else:
 *                         # rDouble is for double bonds NOT to oxygen or Sulfur
 *                         rDouble += 1             # <<<<<<<<<<<<<<
 *                 elif bond12.isTriple(): triple += 1
 *                 elif bond12.isBenzene(): benzene += 1
 */
          __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_rDouble, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1234; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_4);
          __Pyx_DECREF_SET(__pyx_v_rDouble, __pyx_t_4);
          __pyx_t_4 = 0;
        }
        __pyx_L10:;
        goto __pyx_L9;
      }

      /* "rmgpy/molecule/group.py":1235
 *                         # rDouble is for double bonds NOT to oxygen or Sulfur
 *                         rDouble += 1
 *                 elif bond12.isTriple(): triple += 1             # <<<<<<<<<<<<<<
 *                 elif bond12.isBenzene(): benzene += 1
 * 
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isTriple); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_10 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_10)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_10);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_10) {
        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_10); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      } else {
        __pyx_t_4 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_9) {
        __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_triple, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1235; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF_SET(__pyx_v_triple, __pyx_t_4);
        __pyx_t_4 = 0;
        goto __pyx_L9;
      }

      /* "rmgpy/molecule/group.py":1236
 *                         rDouble += 1
 *                 elif bond12.isTriple(): triple += 1
 *                 elif bond12.isBenzene(): benzene += 1             # <<<<<<<<<<<<<<
 * 
 * 
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_10 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_10)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_10);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_10) {
        __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_10); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      } else {
        __pyx_t_4 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_9) {
        __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_benzene, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF_SET(__pyx_v_benzene, __pyx_t_4);
        __pyx_t_4 = 0;
        goto __pyx_L9;
      }
      __pyx_L9:;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1239
 * 
 * 
 *             while oDouble < oDoubleRequired[0]:             # <<<<<<<<<<<<<<
 *                 oDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['O']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
    while (1) {
      __pyx_t_1 = __Pyx_GetItemInt(__pyx_v_oDoubleRequired, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1239; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_oDouble, __pyx_t_1, Py_LT); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1239; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1239; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (!__pyx_t_9) break;

      /* "rmgpy/molecule/group.py":1240
 * 
 *             while oDouble < oDoubleRequired[0]:
 *                 oDouble +=1             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['O']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 */
      __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_oDouble, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1240; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF_SET(__pyx_v_oDouble, __pyx_t_4);
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1241
 *             while oDouble < oDoubleRequired[0]:
 *                 oDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['O']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond
 */
      __pyx_t_4 = PyDict_New(); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_3 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_O); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      PyList_SET_ITEM(__pyx_t_1, 0, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_atomType, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_INCREF(__pyx_int_0);
      PyList_SET_ITEM(__pyx_t_1, 0, __pyx_int_0);
      __Pyx_GIVEREF(__pyx_int_0);
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_radicalElectrons, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_charge, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1241; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_1));
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1242
 *                 oDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['O']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])             # <<<<<<<<<<<<<<
 *                 implicitAtoms[newAtom] = newBond
 *             while sDouble < sDoubleRequired[0]:
 */
      __pyx_t_1 = PyTuple_New(2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1242; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_INCREF(__pyx_v_atom1);
      PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_atom1);
      __Pyx_GIVEREF(__pyx_v_atom1);
      __Pyx_INCREF(((PyObject *)__pyx_v_newAtom));
      PyTuple_SET_ITEM(__pyx_t_1, 1, ((PyObject *)__pyx_v_newAtom));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_newAtom));
      __pyx_t_4 = PyDict_New(); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1242; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1242; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_n_s_D);
      PyList_SET_ITEM(__pyx_t_3, 0, __pyx_n_s_D);
      __Pyx_GIVEREF(__pyx_n_s_D);
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_order, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1242; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_1, __pyx_t_4); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1242; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newBond, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1243
 *                 newAtom = GroupAtom(atomType=[atomTypes['O']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond             # <<<<<<<<<<<<<<
 *             while sDouble < sDoubleRequired[0]:
 *                 sDouble +=1
 */
      if (unlikely(PyDict_SetItem(__pyx_v_implicitAtoms, ((PyObject *)__pyx_v_newAtom), ((PyObject *)__pyx_v_newBond)) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1243; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "rmgpy/molecule/group.py":1244
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond
 *             while sDouble < sDoubleRequired[0]:             # <<<<<<<<<<<<<<
 *                 sDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['S']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
    while (1) {
      __pyx_t_3 = __Pyx_GetItemInt(__pyx_v_sDoubleRequired, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1244; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_sDouble, __pyx_t_3, Py_LT); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1244; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1244; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (!__pyx_t_9) break;

      /* "rmgpy/molecule/group.py":1245
 *                 implicitAtoms[newAtom] = newBond
 *             while sDouble < sDoubleRequired[0]:
 *                 sDouble +=1             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['S']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 */
      __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_sDouble, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1245; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF_SET(__pyx_v_sDouble, __pyx_t_4);
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1246
 *             while sDouble < sDoubleRequired[0]:
 *                 sDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['S']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond
 */
      __pyx_t_4 = PyDict_New(); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_1 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_S); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      PyList_SET_ITEM(__pyx_t_3, 0, __pyx_t_1);
      __Pyx_GIVEREF(__pyx_t_1);
      __pyx_t_1 = 0;
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_atomType, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_int_0);
      PyList_SET_ITEM(__pyx_t_3, 0, __pyx_int_0);
      __Pyx_GIVEREF(__pyx_int_0);
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_radicalElectrons, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_charge, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_4); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1246; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1247
 *                 sDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['S']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])             # <<<<<<<<<<<<<<
 *                 implicitAtoms[newAtom] = newBond
 *             while rDouble < rDoubleRequired[0] or rDouble + oDouble + sDouble < allDoubleRequired[0]:
 */
      __pyx_t_3 = PyTuple_New(2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_v_atom1);
      PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_v_atom1);
      __Pyx_GIVEREF(__pyx_v_atom1);
      __Pyx_INCREF(((PyObject *)__pyx_v_newAtom));
      PyTuple_SET_ITEM(__pyx_t_3, 1, ((PyObject *)__pyx_v_newAtom));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_newAtom));
      __pyx_t_4 = PyDict_New(); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_INCREF(__pyx_n_s_D);
      PyList_SET_ITEM(__pyx_t_1, 0, __pyx_n_s_D);
      __Pyx_GIVEREF(__pyx_n_s_D);
      if (PyDict_SetItem(__pyx_t_4, __pyx_n_s_order, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1247; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newBond, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_1));
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1248
 *                 newAtom = GroupAtom(atomType=[atomTypes['S']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond             # <<<<<<<<<<<<<<
 *             while rDouble < rDoubleRequired[0] or rDouble + oDouble + sDouble < allDoubleRequired[0]:
 *                 rDouble +=1
 */
      if (unlikely(PyDict_SetItem(__pyx_v_implicitAtoms, ((PyObject *)__pyx_v_newAtom), ((PyObject *)__pyx_v_newBond)) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1248; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "rmgpy/molecule/group.py":1249
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond
 *             while rDouble < rDoubleRequired[0] or rDouble + oDouble + sDouble < allDoubleRequired[0]:             # <<<<<<<<<<<<<<
 *                 rDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
    while (1) {
      __pyx_t_1 = __Pyx_GetItemInt(__pyx_v_rDoubleRequired, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1249; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_rDouble, __pyx_t_1, Py_LT); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_14 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_14 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (!__pyx_t_14) {
        goto __pyx_L18_next_or;
      } else {
        __pyx_t_9 = __pyx_t_14;
        goto __pyx_L17_bool_binop_done;
      }
      __pyx_L18_next_or:;
      __pyx_t_4 = PyNumber_Add(__pyx_v_rDouble, __pyx_v_oDouble); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_1 = PyNumber_Add(__pyx_t_4, __pyx_v_sDouble); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = __Pyx_GetItemInt(__pyx_v_allDoubleRequired, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1249; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_t_1, __pyx_t_4, Py_LT); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_14 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_14 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1249; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_9 = __pyx_t_14;
      __pyx_L17_bool_binop_done:;
      if (!__pyx_t_9) break;

      /* "rmgpy/molecule/group.py":1250
 *                 implicitAtoms[newAtom] = newBond
 *             while rDouble < rDoubleRequired[0] or rDouble + oDouble + sDouble < allDoubleRequired[0]:
 *                 rDouble +=1             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 */
      __pyx_t_3 = PyNumber_InPlaceAdd(__pyx_v_rDouble, __pyx_int_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1250; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF_SET(__pyx_v_rDouble, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1251
 *             while rDouble < rDoubleRequired[0] or rDouble + oDouble + sDouble < allDoubleRequired[0]:
 *                 rDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond
 */
      __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_1 = PyObject_GetItem(__pyx_t_4, __pyx_n_s_C); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyList_SET_ITEM(__pyx_t_4, 0, __pyx_t_1);
      __Pyx_GIVEREF(__pyx_t_1);
      __pyx_t_1 = 0;
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_atomType, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_INCREF(__pyx_int_0);
      PyList_SET_ITEM(__pyx_t_4, 0, __pyx_int_0);
      __Pyx_GIVEREF(__pyx_int_0);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_radicalElectrons, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = PyList_New(0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_charge, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1251; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_4));
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1252
 *                 rDouble +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])             # <<<<<<<<<<<<<<
 *                 implicitAtoms[newAtom] = newBond
 *             while triple < tripleRequired[0]:
 */
      __pyx_t_4 = PyTuple_New(2); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_INCREF(__pyx_v_atom1);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_v_atom1);
      __Pyx_GIVEREF(__pyx_v_atom1);
      __Pyx_INCREF(((PyObject *)__pyx_v_newAtom));
      PyTuple_SET_ITEM(__pyx_t_4, 1, ((PyObject *)__pyx_v_newAtom));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_newAtom));
      __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_INCREF(__pyx_n_s_D);
      PyList_SET_ITEM(__pyx_t_1, 0, __pyx_n_s_D);
      __Pyx_GIVEREF(__pyx_n_s_D);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_order, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_4, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1252; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newBond, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_1));
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1253
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond             # <<<<<<<<<<<<<<
 *             while triple < tripleRequired[0]:
 *                 triple +=1
 */
      if (unlikely(PyDict_SetItem(__pyx_v_implicitAtoms, ((PyObject *)__pyx_v_newAtom), ((PyObject *)__pyx_v_newBond)) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1253; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "rmgpy/molecule/group.py":1254
 *                 newBond = GroupBond(atom1, newAtom, order=['D'])
 *                 implicitAtoms[newAtom] = newBond
 *             while triple < tripleRequired[0]:             # <<<<<<<<<<<<<<
 *                 triple +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
    while (1) {
      __pyx_t_1 = __Pyx_GetItemInt(__pyx_v_tripleRequired, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1254; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_triple, __pyx_t_1, Py_LT); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1254; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1254; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!__pyx_t_9) break;

      /* "rmgpy/molecule/group.py":1255
 *                 implicitAtoms[newAtom] = newBond
 *             while triple < tripleRequired[0]:
 *                 triple +=1             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['T'])
 */
      __pyx_t_3 = PyNumber_InPlaceAdd(__pyx_v_triple, __pyx_int_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1255; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF_SET(__pyx_v_triple, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1256
 *             while triple < tripleRequired[0]:
 *                 triple +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *                 newBond = GroupBond(atom1, newAtom, order=['T'])
 *                 implicitAtoms[newAtom] = newBond
 */
      __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_4 = PyObject_GetItem(__pyx_t_1, __pyx_n_s_C); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      PyList_SET_ITEM(__pyx_t_1, 0, __pyx_t_4);
      __Pyx_GIVEREF(__pyx_t_4);
      __pyx_t_4 = 0;
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_atomType, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_1 = PyList_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_INCREF(__pyx_int_0);
      PyList_SET_ITEM(__pyx_t_1, 0, __pyx_int_0);
      __Pyx_GIVEREF(__pyx_int_0);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_radicalElectrons, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_charge, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1256; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)__pyx_t_1));
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1257
 *                 triple +=1
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['T'])             # <<<<<<<<<<<<<<
 *                 implicitAtoms[newAtom] = newBond
 * 
 */
      __pyx_t_1 = PyTuple_New(2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_INCREF(__pyx_v_atom1);
      PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_atom1);
      __Pyx_GIVEREF(__pyx_v_atom1);
      __Pyx_INCREF(((PyObject *)__pyx_v_newAtom));
      PyTuple_SET_ITEM(__pyx_t_1, 1, ((PyObject *)__pyx_v_newAtom));
      __Pyx_GIVEREF(((PyObject *)__pyx_v_newAtom));
      __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_INCREF(__pyx_n_s_T);
      PyList_SET_ITEM(__pyx_t_4, 0, __pyx_n_s_T);
      __Pyx_GIVEREF(__pyx_n_s_T);
      if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_order, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_1, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newBond, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_4));
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1258
 *                 newAtom = GroupAtom(atomType=[atomTypes['C']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *                 newBond = GroupBond(atom1, newAtom, order=['T'])
 *                 implicitAtoms[newAtom] = newBond             # <<<<<<<<<<<<<<
 * 
 *         for atom, bond in implicitAtoms.iteritems():
 */
      if (unlikely(PyDict_SetItem(__pyx_v_implicitAtoms, ((PyObject *)__pyx_v_newAtom), ((PyObject *)__pyx_v_newBond)) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1258; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "rmgpy/molecule/group.py":1209
 *         copyGroup = deepcopy(self)
 * 
 *         for atom1 in copyGroup.atoms:             # <<<<<<<<<<<<<<
 *             atomtypeFeatureList = atom1.atomType[0].getFeatures()
 *             lonePairsRequired[atom1]=atomtypeFeatureList[7]
 */
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1260
 *                 implicitAtoms[newAtom] = newBond
 * 
 *         for atom, bond in implicitAtoms.iteritems():             # <<<<<<<<<<<<<<
 *             copyGroup.addAtom(atom)
 *             copyGroup.addBond(bond)
 */
  __pyx_t_5 = 0;
  __pyx_t_4 = __Pyx_dict_iterator(__pyx_v_implicitAtoms, 1, __pyx_n_s_iteritems, (&__pyx_t_11), (&__pyx_t_12)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1260; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_2);
  __pyx_t_2 = __pyx_t_4;
  __pyx_t_4 = 0;
  while (1) {
    __pyx_t_13 = __Pyx_dict_iter_next(__pyx_t_2, __pyx_t_11, &__pyx_t_5, &__pyx_t_4, &__pyx_t_3, NULL, __pyx_t_12);
    if (unlikely(__pyx_t_13 == 0)) break;
    if (unlikely(__pyx_t_13 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1260; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_4);
    __pyx_t_4 = 0;
    __Pyx_XDECREF_SET(__pyx_v_bond, __pyx_t_3);
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1261
 * 
 *         for atom, bond in implicitAtoms.iteritems():
 *             copyGroup.addAtom(atom)             # <<<<<<<<<<<<<<
 *             copyGroup.addBond(bond)
 * 
 */
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_copyGroup, __pyx_n_s_addAtom); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1261; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
      __pyx_t_1 = PyMethod_GET_SELF(__pyx_t_4);
      if (likely(__pyx_t_1)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
        __Pyx_INCREF(__pyx_t_1);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_4, function);
      }
    }
    if (!__pyx_t_1) {
      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_v_atom); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1261; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
    } else {
      __pyx_t_10 = PyTuple_New(1+1); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1261; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_10);
      PyTuple_SET_ITEM(__pyx_t_10, 0, __pyx_t_1); __Pyx_GIVEREF(__pyx_t_1); __pyx_t_1 = NULL;
      __Pyx_INCREF(__pyx_v_atom);
      PyTuple_SET_ITEM(__pyx_t_10, 0+1, __pyx_v_atom);
      __Pyx_GIVEREF(__pyx_v_atom);
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_10, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1261; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
    }
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1262
 *         for atom, bond in implicitAtoms.iteritems():
 *             copyGroup.addAtom(atom)
 *             copyGroup.addBond(bond)             # <<<<<<<<<<<<<<
 * 
 *         for atom, lonePair in lonePairsRequired.iteritems():
 */
    __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_copyGroup, __pyx_n_s_addBond); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1262; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_10 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
      __pyx_t_10 = PyMethod_GET_SELF(__pyx_t_4);
      if (likely(__pyx_t_10)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
        __Pyx_INCREF(__pyx_t_10);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_4, function);
      }
    }
    if (!__pyx_t_10) {
      __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_v_bond); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1262; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
    } else {
      __pyx_t_1 = PyTuple_New(1+1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1262; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_t_10); __Pyx_GIVEREF(__pyx_t_10); __pyx_t_10 = NULL;
      __Pyx_INCREF(__pyx_v_bond);
      PyTuple_SET_ITEM(__pyx_t_1, 0+1, __pyx_v_bond);
      __Pyx_GIVEREF(__pyx_v_bond);
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_1, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1262; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    }
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1264
 *             copyGroup.addBond(bond)
 * 
 *         for atom, lonePair in lonePairsRequired.iteritems():             # <<<<<<<<<<<<<<
 *             if lonePair: atom.lonePairs = lonePair
 * 
 */
  __pyx_t_11 = 0;
  __pyx_t_3 = __Pyx_dict_iterator(__pyx_v_lonePairsRequired, 1, __pyx_n_s_iteritems, (&__pyx_t_5), (&__pyx_t_12)); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1264; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_2);
  __pyx_t_2 = __pyx_t_3;
  __pyx_t_3 = 0;
  while (1) {
    __pyx_t_13 = __Pyx_dict_iter_next(__pyx_t_2, __pyx_t_5, &__pyx_t_11, &__pyx_t_3, &__pyx_t_4, NULL, __pyx_t_12);
    if (unlikely(__pyx_t_13 == 0)) break;
    if (unlikely(__pyx_t_13 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1264; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_3);
    __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_lonePair, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1265
 * 
 *         for atom, lonePair in lonePairsRequired.iteritems():
 *             if lonePair: atom.lonePairs = lonePair             # <<<<<<<<<<<<<<
 * 
 *         return copyGroup
 */
    __pyx_t_9 = __Pyx_PyObject_IsTrue(__pyx_v_lonePair); if (unlikely(__pyx_t_9 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (__pyx_t_9) {
      if (__Pyx_PyObject_SetAttrStr(__pyx_v_atom, __pyx_n_s_lonePairs, __pyx_v_lonePair) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1265; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L25;
    }
    __pyx_L25:;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1267
 *             if lonePair: atom.lonePairs = lonePair
 * 
 *         return copyGroup             # <<<<<<<<<<<<<<
 * 
 *     def addImplicitBenzene(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  if (!(likely(((__pyx_v_copyGroup) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_copyGroup, __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1267; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_INCREF(__pyx_v_copyGroup);
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_copyGroup);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1194
 * 
 * 
 *     def addImplicitAtomsFromAtomType(self):             # <<<<<<<<<<<<<<
 *         """
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_10);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitAtomsFromAtomType", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_implicitAtoms);
  __Pyx_XDECREF(__pyx_v_lonePairsRequired);
  __Pyx_XDECREF(__pyx_v_copyGroup);
  __Pyx_XDECREF(__pyx_v_atom1);
  __Pyx_XDECREF(__pyx_v_atomtypeFeatureList);
  __Pyx_XDECREF(__pyx_v_allDoubleRequired);
  __Pyx_XDECREF(__pyx_v_rDoubleRequired);
  __Pyx_XDECREF(__pyx_v_oDoubleRequired);
  __Pyx_XDECREF(__pyx_v_sDoubleRequired);
  __Pyx_XDECREF(__pyx_v_tripleRequired);
  __Pyx_XDECREF(__pyx_v_single);
  __Pyx_XDECREF(__pyx_v_rDouble);
  __Pyx_XDECREF(__pyx_v_oDouble);
  __Pyx_XDECREF(__pyx_v_sDouble);
  __Pyx_XDECREF(__pyx_v_triple);
  __Pyx_XDECREF(__pyx_v_benzene);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XDECREF((PyObject *)__pyx_v_newAtom);
  __Pyx_XDECREF((PyObject *)__pyx_v_newBond);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_bond);
  __Pyx_XDECREF(__pyx_v_lonePair);
  __Pyx_XDECREF(__pyx_v_featureList);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_67addImplicitAtomsFromAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_66addImplicitAtomsFromAtomType[] = "\n\n        Returns: a modified group with implicit atoms added\n        Add implicit double/triple bonded atoms O, S or R, for which we will use a C\n\n        Not designed to work with wildcards\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_67addImplicitAtomsFromAtomType(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addImplicitAtomsFromAtomType (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_66addImplicitAtomsFromAtomType(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_66addImplicitAtomsFromAtomType(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addImplicitAtomsFromAtomType", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_addImplicitAtomsFromAtomType(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1194; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitAtomsFromAtomType", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1283
 *         """
 * 
 *         def classifyBenzeneCarbons(group, partners = []):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_12__defaults__(CYTHON_UNUSED PyObject *__pyx_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__defaults__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__Pyx_CyFunction_Defaults(__pyx_defaults, __pyx_self)->__pyx_arg_partners);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __Pyx_CyFunction_Defaults(__pyx_defaults, __pyx_self)->__pyx_arg_partners);
  __Pyx_GIVEREF(__Pyx_CyFunction_Defaults(__pyx_defaults, __pyx_self)->__pyx_arg_partners);
  __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_INCREF(Py_None);
  PyTuple_SET_ITEM(__pyx_t_2, 1, Py_None);
  __Pyx_GIVEREF(Py_None);
  __pyx_t_1 = 0;
  __pyx_r = __pyx_t_2;
  __pyx_t_2 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.__defaults__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_1classifyBenzeneCarbons(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_classifyBenzeneCarbons[] = "\n            Args:\n                group: :class:Group with atoms to classify\n                partners: dictionary of partnered up atoms, which must be a cbf atom\n\n            Returns: tuple with lists of each atom classification\n            ";
static PyMethodDef __pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_1classifyBenzeneCarbons = {"classifyBenzeneCarbons", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_1classifyBenzeneCarbons, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_classifyBenzeneCarbons};
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_1classifyBenzeneCarbons(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_group = 0;
  PyObject *__pyx_v_partners = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("classifyBenzeneCarbons (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_group,&__pyx_n_s_partners,0};
    PyObject* values[2] = {0,0};
    __pyx_defaults *__pyx_dynamic_args = __Pyx_CyFunction_Defaults(__pyx_defaults, __pyx_self);
    values[1] = __pyx_dynamic_args->__pyx_arg_partners;
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_group)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_partners);
          if (value) { values[1] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "classifyBenzeneCarbons") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_group = values[0];
    __pyx_v_partners = values[1];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("classifyBenzeneCarbons", 0, 1, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.classifyBenzeneCarbons", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_classifyBenzeneCarbons(__pyx_self, __pyx_v_group, __pyx_v_partners);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_classifyBenzeneCarbons(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_group, PyObject *__pyx_v_partners) {
  PyObject *__pyx_v_cbAtomList = NULL;
  PyObject *__pyx_v_cbfAtomList = NULL;
  PyObject *__pyx_v_cbfAtomList1 = NULL;
  PyObject *__pyx_v_cbfAtomList2 = NULL;
  PyObject *__pyx_v_connectedCbfs = NULL;
  PyObject *__pyx_v_labelsOfCarbonAtomTypes = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_benzeneBonds = NULL;
  PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_v_cbfAtom = NULL;
  PyObject *__pyx_v_fbBonds = NULL;
  PyObject *__pyx_v_bond = NULL;
  PyObject *__pyx_v_x = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  Py_ssize_t __pyx_t_4;
  PyObject *(*__pyx_t_5)(PyObject *);
  int __pyx_t_6;
  int __pyx_t_7;
  int __pyx_t_8;
  Py_ssize_t __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  int __pyx_t_11;
  PyObject *__pyx_t_12 = NULL;
  int __pyx_t_13;
  PyObject *__pyx_t_14 = NULL;
  int __pyx_t_15;
  PyObject *__pyx_t_16 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("classifyBenzeneCarbons", 0);

  /* "rmgpy/molecule/group.py":1291
 *             Returns: tuple with lists of each atom classification
 *             """
 *             cbAtomList = []             # <<<<<<<<<<<<<<
 *             cbfAtomList = [] #All Cbf Atoms
 *             cbfAtomList1 = [] #Cbf Atoms that are bonded to exactly one other Cbf (part of 2 rings)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1291; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_cbAtomList = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1292
 *             """
 *             cbAtomList = []
 *             cbfAtomList = [] #All Cbf Atoms             # <<<<<<<<<<<<<<
 *             cbfAtomList1 = [] #Cbf Atoms that are bonded to exactly one other Cbf (part of 2 rings)
 *             cbfAtomList2 = [] #Cbf that are sandwiched between two other Cbf (part of 2 rings)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1292; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_cbfAtomList = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1293
 *             cbAtomList = []
 *             cbfAtomList = [] #All Cbf Atoms
 *             cbfAtomList1 = [] #Cbf Atoms that are bonded to exactly one other Cbf (part of 2 rings)             # <<<<<<<<<<<<<<
 *             cbfAtomList2 = [] #Cbf that are sandwiched between two other Cbf (part of 2 rings)
 *             connectedCbfs={} #dictionary of connections to other cbfAtoms
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1293; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_cbfAtomList1 = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1294
 *             cbfAtomList = [] #All Cbf Atoms
 *             cbfAtomList1 = [] #Cbf Atoms that are bonded to exactly one other Cbf (part of 2 rings)
 *             cbfAtomList2 = [] #Cbf that are sandwiched between two other Cbf (part of 2 rings)             # <<<<<<<<<<<<<<
 *             connectedCbfs={} #dictionary of connections to other cbfAtoms
 * 
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1294; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_cbfAtomList2 = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1295
 *             cbfAtomList1 = [] #Cbf Atoms that are bonded to exactly one other Cbf (part of 2 rings)
 *             cbfAtomList2 = [] #Cbf that are sandwiched between two other Cbf (part of 2 rings)
 *             connectedCbfs={} #dictionary of connections to other cbfAtoms             # <<<<<<<<<<<<<<
 * 
 *             #Only want to work with benzene bonds on carbon
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1295; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_connectedCbfs = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1298
 * 
 *             #Only want to work with benzene bonds on carbon
 *             labelsOfCarbonAtomTypes = [x.label for x in atomTypes['C'].specific] + ['C']             # <<<<<<<<<<<<<<
 *             #Also allow with R!H and some nitrogen groups
 *             labelsOfCarbonAtomTypes.extend(['R!H', 'N5b', 'N3b'])
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_C); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_specific); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_3 = __pyx_t_2; __Pyx_INCREF(__pyx_t_3); __pyx_t_4 = 0;
    __pyx_t_5 = NULL;
  } else {
    __pyx_t_4 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_5)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_4 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_4); __Pyx_INCREF(__pyx_t_2); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_4 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_4); __Pyx_INCREF(__pyx_t_2); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_5(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_2);
    __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_x, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_2))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_n_s_C);
  PyList_SET_ITEM(__pyx_t_3, 0, __pyx_n_s_C);
  __Pyx_GIVEREF(__pyx_n_s_C);
  __pyx_t_2 = PyNumber_Add(__pyx_t_1, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1298; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_v_labelsOfCarbonAtomTypes = ((PyObject*)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1300
 *             labelsOfCarbonAtomTypes = [x.label for x in atomTypes['C'].specific] + ['C']
 *             #Also allow with R!H and some nitrogen groups
 *             labelsOfCarbonAtomTypes.extend(['R!H', 'N5b', 'N3b'])             # <<<<<<<<<<<<<<
 * 
 *             for atom in group.atoms:
 */
  __pyx_t_2 = PyList_New(3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1300; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_kp_s_R_H);
  PyList_SET_ITEM(__pyx_t_2, 0, __pyx_kp_s_R_H);
  __Pyx_GIVEREF(__pyx_kp_s_R_H);
  __Pyx_INCREF(__pyx_n_s_N5b);
  PyList_SET_ITEM(__pyx_t_2, 1, __pyx_n_s_N5b);
  __Pyx_GIVEREF(__pyx_n_s_N5b);
  __Pyx_INCREF(__pyx_n_s_N3b);
  PyList_SET_ITEM(__pyx_t_2, 2, __pyx_n_s_N3b);
  __Pyx_GIVEREF(__pyx_n_s_N3b);
  __pyx_t_6 = __Pyx_PyList_Extend(__pyx_v_labelsOfCarbonAtomTypes, __pyx_t_2); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1300; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1302
 *             labelsOfCarbonAtomTypes.extend(['R!H', 'N5b', 'N3b'])
 * 
 *             for atom in group.atoms:             # <<<<<<<<<<<<<<
 *                 if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue
 *                 elif atom.atomType[0].label in ['Cb', 'N3b']: #Make Cb and N3b into normal cb atoms
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_group, __pyx_n_s_atoms); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_3 = __pyx_t_2; __Pyx_INCREF(__pyx_t_3); __pyx_t_4 = 0;
    __pyx_t_5 = NULL;
  } else {
    __pyx_t_4 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_5)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_4 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_4); __Pyx_INCREF(__pyx_t_2); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_4 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_4); __Pyx_INCREF(__pyx_t_2); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_5(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1302; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1303
 * 
 *             for atom in group.atoms:
 *                 if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue             # <<<<<<<<<<<<<<
 *                 elif atom.atomType[0].label in ['Cb', 'N3b']: #Make Cb and N3b into normal cb atoms
 *                     cbAtomList.append(atom)
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1303; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1303; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1303; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_7 = (__Pyx_PySequence_Contains(__pyx_t_2, __pyx_v_labelsOfCarbonAtomTypes, Py_NE)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1303; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_8 = (__pyx_t_7 != 0);
    if (__pyx_t_8) {
      goto __pyx_L5_continue;
    }

    /* "rmgpy/molecule/group.py":1304
 *             for atom in group.atoms:
 *                 if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue
 *                 elif atom.atomType[0].label in ['Cb', 'N3b']: #Make Cb and N3b into normal cb atoms             # <<<<<<<<<<<<<<
 *                     cbAtomList.append(atom)
 *                 elif atom.atomType[0].label == 'Cbf':
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1304; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1304; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1304; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_n_s_Cb, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1304; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (!__pyx_t_7) {
      goto __pyx_L9_next_or;
    } else {
      __pyx_t_8 = __pyx_t_7;
      goto __pyx_L8_bool_binop_done;
    }
    __pyx_L9_next_or:;
    __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_n_s_N3b, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1304; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_8 = __pyx_t_7;
    __pyx_L8_bool_binop_done:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_7 = (__pyx_t_8 != 0);
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":1305
 *                 if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue
 *                 elif atom.atomType[0].label in ['Cb', 'N3b']: #Make Cb and N3b into normal cb atoms
 *                     cbAtomList.append(atom)             # <<<<<<<<<<<<<<
 *                 elif atom.atomType[0].label == 'Cbf':
 *                     cbfAtomList.append(atom)
 */
      __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_cbAtomList, __pyx_v_atom); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1305; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L7;
    }

    /* "rmgpy/molecule/group.py":1306
 *                 elif atom.atomType[0].label in ['Cb', 'N3b']: #Make Cb and N3b into normal cb atoms
 *                     cbAtomList.append(atom)
 *                 elif atom.atomType[0].label == 'Cbf':             # <<<<<<<<<<<<<<
 *                     cbfAtomList.append(atom)
 *                 else:
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1306; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1306; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1306; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_7 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_n_s_Cbf, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1306; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_7) {

      /* "rmgpy/molecule/group.py":1307
 *                     cbAtomList.append(atom)
 *                 elif atom.atomType[0].label == 'Cbf':
 *                     cbfAtomList.append(atom)             # <<<<<<<<<<<<<<
 *                 else:
 *                     benzeneBonds = 0
 */
      __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_cbfAtomList, __pyx_v_atom); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1307; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L7;
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":1309
 *                     cbfAtomList.append(atom)
 *                 else:
 *                     benzeneBonds = 0             # <<<<<<<<<<<<<<
 *                     for atom2, bond12 in atom.bonds.iteritems():
 *                         if bond12.isBenzene(): benzeneBonds+=1
 */
      __Pyx_INCREF(__pyx_int_0);
      __Pyx_XDECREF_SET(__pyx_v_benzeneBonds, __pyx_int_0);

      /* "rmgpy/molecule/group.py":1310
 *                 else:
 *                     benzeneBonds = 0
 *                     for atom2, bond12 in atom.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                         if bond12.isBenzene(): benzeneBonds+=1
 *                     if benzeneBonds > 2: cbfAtomList.append(atom)
 */
      __pyx_t_9 = 0;
      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1310; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      if (unlikely(__pyx_t_1 == Py_None)) {
        PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1310; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_12 = __Pyx_dict_iterator(__pyx_t_1, 0, __pyx_n_s_iteritems, (&__pyx_t_10), (&__pyx_t_11)); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1310; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_12);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_XDECREF(__pyx_t_2);
      __pyx_t_2 = __pyx_t_12;
      __pyx_t_12 = 0;
      while (1) {
        __pyx_t_13 = __Pyx_dict_iter_next(__pyx_t_2, __pyx_t_10, &__pyx_t_9, &__pyx_t_12, &__pyx_t_1, NULL, __pyx_t_11);
        if (unlikely(__pyx_t_13 == 0)) break;
        if (unlikely(__pyx_t_13 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1310; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_12);
        __Pyx_GOTREF(__pyx_t_1);
        __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_12);
        __pyx_t_12 = 0;
        __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_1);
        __pyx_t_1 = 0;

        /* "rmgpy/molecule/group.py":1311
 *                     benzeneBonds = 0
 *                     for atom2, bond12 in atom.bonds.iteritems():
 *                         if bond12.isBenzene(): benzeneBonds+=1             # <<<<<<<<<<<<<<
 *                     if benzeneBonds > 2: cbfAtomList.append(atom)
 *                     elif benzeneBonds >0: cbAtomList.append(atom)
 */
        __pyx_t_12 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1311; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_12);
        __pyx_t_14 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_12))) {
          __pyx_t_14 = PyMethod_GET_SELF(__pyx_t_12);
          if (likely(__pyx_t_14)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_12);
            __Pyx_INCREF(__pyx_t_14);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_12, function);
          }
        }
        if (__pyx_t_14) {
          __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_12, __pyx_t_14); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1311; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_14); __pyx_t_14 = 0;
        } else {
          __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_12); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1311; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_1);
        __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
        __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1311; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        if (__pyx_t_7) {
          __pyx_t_1 = PyNumber_InPlaceAdd(__pyx_v_benzeneBonds, __pyx_int_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1311; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_1);
          __Pyx_DECREF_SET(__pyx_v_benzeneBonds, __pyx_t_1);
          __pyx_t_1 = 0;
          goto __pyx_L12;
        }
        __pyx_L12:;
      }
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

      /* "rmgpy/molecule/group.py":1312
 *                     for atom2, bond12 in atom.bonds.iteritems():
 *                         if bond12.isBenzene(): benzeneBonds+=1
 *                     if benzeneBonds > 2: cbfAtomList.append(atom)             # <<<<<<<<<<<<<<
 *                     elif benzeneBonds >0: cbAtomList.append(atom)
 * 
 */
      __pyx_t_2 = PyObject_RichCompare(__pyx_v_benzeneBonds, __pyx_int_2, Py_GT); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1312; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1312; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (__pyx_t_7) {
        __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_cbfAtomList, __pyx_v_atom); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1312; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L13;
      }

      /* "rmgpy/molecule/group.py":1313
 *                         if bond12.isBenzene(): benzeneBonds+=1
 *                     if benzeneBonds > 2: cbfAtomList.append(atom)
 *                     elif benzeneBonds >0: cbAtomList.append(atom)             # <<<<<<<<<<<<<<
 * 
 *             #further sort the cbf atoms
 */
      __pyx_t_2 = PyObject_RichCompare(__pyx_v_benzeneBonds, __pyx_int_0, Py_GT); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1313; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1313; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (__pyx_t_7) {
        __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_cbAtomList, __pyx_v_atom); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1313; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L13;
      }
      __pyx_L13:;
    }
    __pyx_L7:;

    /* "rmgpy/molecule/group.py":1302
 *             labelsOfCarbonAtomTypes.extend(['R!H', 'N5b', 'N3b'])
 * 
 *             for atom in group.atoms:             # <<<<<<<<<<<<<<
 *                 if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue
 *                 elif atom.atomType[0].label in ['Cb', 'N3b']: #Make Cb and N3b into normal cb atoms
 */
    __pyx_L5_continue:;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":1316
 * 
 *             #further sort the cbf atoms
 *             for cbfAtom in cbfAtomList:             # <<<<<<<<<<<<<<
 *                 fbBonds = 0
 *                 connectedCbfs[cbfAtom] = []
 */
  __pyx_t_3 = __pyx_v_cbfAtomList; __Pyx_INCREF(__pyx_t_3); __pyx_t_4 = 0;
  for (;;) {
    if (__pyx_t_4 >= PyList_GET_SIZE(__pyx_t_3)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_4); __Pyx_INCREF(__pyx_t_2); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1316; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1316; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_cbfAtom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1317
 *             #further sort the cbf atoms
 *             for cbfAtom in cbfAtomList:
 *                 fbBonds = 0             # <<<<<<<<<<<<<<
 *                 connectedCbfs[cbfAtom] = []
 *                 for atom2, bond in cbfAtom.bonds.iteritems():
 */
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_fbBonds, __pyx_int_0);

    /* "rmgpy/molecule/group.py":1318
 *             for cbfAtom in cbfAtomList:
 *                 fbBonds = 0
 *                 connectedCbfs[cbfAtom] = []             # <<<<<<<<<<<<<<
 *                 for atom2, bond in cbfAtom.bonds.iteritems():
 *                     if bond.order[0] == 'B' and atom2 in cbfAtomList:
 */
    __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1318; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    if (unlikely(PyDict_SetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom, __pyx_t_2) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1318; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1319
 *                 fbBonds = 0
 *                 connectedCbfs[cbfAtom] = []
 *                 for atom2, bond in cbfAtom.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                     if bond.order[0] == 'B' and atom2 in cbfAtomList:
 *                         fbBonds +=1
 */
    __pyx_t_10 = 0;
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_cbfAtom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1319; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (unlikely(__pyx_t_1 == Py_None)) {
      PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1319; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_12 = __Pyx_dict_iterator(__pyx_t_1, 0, __pyx_n_s_iteritems, (&__pyx_t_9), (&__pyx_t_11)); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1319; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_12);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_XDECREF(__pyx_t_2);
    __pyx_t_2 = __pyx_t_12;
    __pyx_t_12 = 0;
    while (1) {
      __pyx_t_13 = __Pyx_dict_iter_next(__pyx_t_2, __pyx_t_9, &__pyx_t_10, &__pyx_t_12, &__pyx_t_1, NULL, __pyx_t_11);
      if (unlikely(__pyx_t_13 == 0)) break;
      if (unlikely(__pyx_t_13 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1319; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_12);
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_12);
      __pyx_t_12 = 0;
      __Pyx_XDECREF_SET(__pyx_v_bond, __pyx_t_1);
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1320
 *                 connectedCbfs[cbfAtom] = []
 *                 for atom2, bond in cbfAtom.bonds.iteritems():
 *                     if bond.order[0] == 'B' and atom2 in cbfAtomList:             # <<<<<<<<<<<<<<
 *                         fbBonds +=1
 *                         connectedCbfs[cbfAtom].append(atom2)
 */
      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond, __pyx_n_s_order); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_12 = __Pyx_GetItemInt(__pyx_t_1, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_12 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1320; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_12);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_8 = (__Pyx_PyString_Equals(__pyx_t_12, __pyx_n_s_B, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      if (__pyx_t_8) {
        goto __pyx_L20_next_and;
      } else {
        __pyx_t_7 = __pyx_t_8;
        goto __pyx_L19_bool_binop_done;
      }
      __pyx_L20_next_and:;
      __pyx_t_8 = (__Pyx_PySequence_Contains(__pyx_v_atom2, __pyx_v_cbfAtomList, Py_EQ)); if (unlikely(__pyx_t_8 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1320; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_15 = (__pyx_t_8 != 0);
      __pyx_t_7 = __pyx_t_15;
      __pyx_L19_bool_binop_done:;
      if (__pyx_t_7) {

        /* "rmgpy/molecule/group.py":1321
 *                 for atom2, bond in cbfAtom.bonds.iteritems():
 *                     if bond.order[0] == 'B' and atom2 in cbfAtomList:
 *                         fbBonds +=1             # <<<<<<<<<<<<<<
 *                         connectedCbfs[cbfAtom].append(atom2)
 *                 if fbBonds < 2: cbfAtomList1.append(cbfAtom)
 */
        __pyx_t_12 = PyNumber_InPlaceAdd(__pyx_v_fbBonds, __pyx_int_1); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1321; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_12);
        __Pyx_DECREF_SET(__pyx_v_fbBonds, __pyx_t_12);
        __pyx_t_12 = 0;

        /* "rmgpy/molecule/group.py":1322
 *                     if bond.order[0] == 'B' and atom2 in cbfAtomList:
 *                         fbBonds +=1
 *                         connectedCbfs[cbfAtom].append(atom2)             # <<<<<<<<<<<<<<
 *                 if fbBonds < 2: cbfAtomList1.append(cbfAtom)
 *                 elif fbBonds == 2: cbfAtomList2.append(cbfAtom)
 */
        __pyx_t_12 = __Pyx_PyDict_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_12 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1322; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_12);
        __pyx_t_6 = __Pyx_PyObject_Append(__pyx_t_12, __pyx_v_atom2); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1322; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
        goto __pyx_L18;
      }
      __pyx_L18:;
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1323
 *                         fbBonds +=1
 *                         connectedCbfs[cbfAtom].append(atom2)
 *                 if fbBonds < 2: cbfAtomList1.append(cbfAtom)             # <<<<<<<<<<<<<<
 *                 elif fbBonds == 2: cbfAtomList2.append(cbfAtom)
 *                 elif fbBonds == 3: pass #leaving here in case we ever want to handle Cbf3 atoms
 */
    __pyx_t_2 = PyObject_RichCompare(__pyx_v_fbBonds, __pyx_int_2, Py_LT); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1323; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1323; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_7) {
      __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_cbfAtomList1, __pyx_v_cbfAtom); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1323; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L21;
    }

    /* "rmgpy/molecule/group.py":1324
 *                         connectedCbfs[cbfAtom].append(atom2)
 *                 if fbBonds < 2: cbfAtomList1.append(cbfAtom)
 *                 elif fbBonds == 2: cbfAtomList2.append(cbfAtom)             # <<<<<<<<<<<<<<
 *                 elif fbBonds == 3: pass #leaving here in case we ever want to handle Cbf3 atoms
 * 
 */
    __pyx_t_2 = PyObject_RichCompare(__pyx_v_fbBonds, __pyx_int_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1324; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1324; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_7) {
      __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_cbfAtomList2, __pyx_v_cbfAtom); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1324; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L21;
    }

    /* "rmgpy/molecule/group.py":1325
 *                 if fbBonds < 2: cbfAtomList1.append(cbfAtom)
 *                 elif fbBonds == 2: cbfAtomList2.append(cbfAtom)
 *                 elif fbBonds == 3: pass #leaving here in case we ever want to handle Cbf3 atoms             # <<<<<<<<<<<<<<
 * 
 *             #reclassify any atoms with partners as cbf1 atoms
 */
    __pyx_t_2 = PyObject_RichCompare(__pyx_v_fbBonds, __pyx_int_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1325; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_7 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1325; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_7) {
      goto __pyx_L21;
    }
    __pyx_L21:;

    /* "rmgpy/molecule/group.py":1316
 * 
 *             #further sort the cbf atoms
 *             for cbfAtom in cbfAtomList:             # <<<<<<<<<<<<<<
 *                 fbBonds = 0
 *                 connectedCbfs[cbfAtom] = []
 */
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":1328
 * 
 *             #reclassify any atoms with partners as cbf1 atoms
 *             for cbfAtom in partners:             # <<<<<<<<<<<<<<
 *                 if cbfAtom in cbAtomList:
 *                     cbAtomList.remove(cbfAtom)
 */
  if (likely(PyList_CheckExact(__pyx_v_partners)) || PyTuple_CheckExact(__pyx_v_partners)) {
    __pyx_t_3 = __pyx_v_partners; __Pyx_INCREF(__pyx_t_3); __pyx_t_4 = 0;
    __pyx_t_5 = NULL;
  } else {
    __pyx_t_4 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_v_partners); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_5 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  for (;;) {
    if (likely(!__pyx_t_5)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_4 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_4); __Pyx_INCREF(__pyx_t_2); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_4 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_4); __Pyx_INCREF(__pyx_t_2); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_5(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1328; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_cbfAtom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1329
 *             #reclassify any atoms with partners as cbf1 atoms
 *             for cbfAtom in partners:
 *                 if cbfAtom in cbAtomList:             # <<<<<<<<<<<<<<
 *                     cbAtomList.remove(cbfAtom)
 *                     cbfAtomList.append(cbfAtom)
 */
    __pyx_t_7 = (__Pyx_PySequence_Contains(__pyx_v_cbfAtom, __pyx_v_cbAtomList, Py_EQ)); if (unlikely(__pyx_t_7 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1329; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_15 = (__pyx_t_7 != 0);
    if (__pyx_t_15) {

      /* "rmgpy/molecule/group.py":1330
 *             for cbfAtom in partners:
 *                 if cbfAtom in cbAtomList:
 *                     cbAtomList.remove(cbfAtom)             # <<<<<<<<<<<<<<
 *                     cbfAtomList.append(cbfAtom)
 *                     cbfAtomList1.append(cbfAtom)
 */
      __pyx_t_12 = __Pyx_PyObject_GetAttrStr(__pyx_v_cbAtomList, __pyx_n_s_remove); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1330; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_12);
      __pyx_t_1 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_12))) {
        __pyx_t_1 = PyMethod_GET_SELF(__pyx_t_12);
        if (likely(__pyx_t_1)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_12);
          __Pyx_INCREF(__pyx_t_1);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_12, function);
        }
      }
      if (!__pyx_t_1) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_12, __pyx_v_cbfAtom); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1330; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
      } else {
        __pyx_t_14 = PyTuple_New(1+1); if (unlikely(!__pyx_t_14)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1330; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_14);
        PyTuple_SET_ITEM(__pyx_t_14, 0, __pyx_t_1); __Pyx_GIVEREF(__pyx_t_1); __pyx_t_1 = NULL;
        __Pyx_INCREF(__pyx_v_cbfAtom);
        PyTuple_SET_ITEM(__pyx_t_14, 0+1, __pyx_v_cbfAtom);
        __Pyx_GIVEREF(__pyx_v_cbfAtom);
        __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_12, __pyx_t_14, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1330; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_14); __pyx_t_14 = 0;
      }
      __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

      /* "rmgpy/molecule/group.py":1331
 *                 if cbfAtom in cbAtomList:
 *                     cbAtomList.remove(cbfAtom)
 *                     cbfAtomList.append(cbfAtom)             # <<<<<<<<<<<<<<
 *                     cbfAtomList1.append(cbfAtom)
 * 
 */
      __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_cbfAtomList, __pyx_v_cbfAtom); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1331; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

      /* "rmgpy/molecule/group.py":1332
 *                     cbAtomList.remove(cbfAtom)
 *                     cbfAtomList.append(cbfAtom)
 *                     cbfAtomList1.append(cbfAtom)             # <<<<<<<<<<<<<<
 * 
 *             #check that cbfAtoms only have benzene bonds
 */
      __pyx_t_6 = __Pyx_PyList_Append(__pyx_v_cbfAtomList1, __pyx_v_cbfAtom); if (unlikely(__pyx_t_6 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1332; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L24;
    }
    __pyx_L24:;

    /* "rmgpy/molecule/group.py":1328
 * 
 *             #reclassify any atoms with partners as cbf1 atoms
 *             for cbfAtom in partners:             # <<<<<<<<<<<<<<
 *                 if cbfAtom in cbAtomList:
 *                     cbAtomList.remove(cbfAtom)
 */
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":1335
 * 
 *             #check that cbfAtoms only have benzene bonds
 *             for cbfAtom in cbfAtomList:             # <<<<<<<<<<<<<<
 *                 for atom2, bond12 in cbfAtom.bonds.iteritems():
 *                     assert bond12.isBenzene(), "Cbf atom in {0} has a bond with an order other than 'B'".format(group)
 */
  __pyx_t_3 = __pyx_v_cbfAtomList; __Pyx_INCREF(__pyx_t_3); __pyx_t_4 = 0;
  for (;;) {
    if (__pyx_t_4 >= PyList_GET_SIZE(__pyx_t_3)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_4); __Pyx_INCREF(__pyx_t_2); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1335; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_cbfAtom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1336
 *             #check that cbfAtoms only have benzene bonds
 *             for cbfAtom in cbfAtomList:
 *                 for atom2, bond12 in cbfAtom.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                     assert bond12.isBenzene(), "Cbf atom in {0} has a bond with an order other than 'B'".format(group)
 * 
 */
    __pyx_t_9 = 0;
    __pyx_t_12 = __Pyx_PyObject_GetAttrStr(__pyx_v_cbfAtom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1336; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_12);
    if (unlikely(__pyx_t_12 == Py_None)) {
      PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1336; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_14 = __Pyx_dict_iterator(__pyx_t_12, 0, __pyx_n_s_iteritems, (&__pyx_t_10), (&__pyx_t_11)); if (unlikely(!__pyx_t_14)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1336; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_14);
    __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
    __Pyx_XDECREF(__pyx_t_2);
    __pyx_t_2 = __pyx_t_14;
    __pyx_t_14 = 0;
    while (1) {
      __pyx_t_13 = __Pyx_dict_iter_next(__pyx_t_2, __pyx_t_10, &__pyx_t_9, &__pyx_t_14, &__pyx_t_12, NULL, __pyx_t_11);
      if (unlikely(__pyx_t_13 == 0)) break;
      if (unlikely(__pyx_t_13 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1336; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_14);
      __Pyx_GOTREF(__pyx_t_12);
      __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_14);
      __pyx_t_14 = 0;
      __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_12);
      __pyx_t_12 = 0;

      /* "rmgpy/molecule/group.py":1337
 *             for cbfAtom in cbfAtomList:
 *                 for atom2, bond12 in cbfAtom.bonds.iteritems():
 *                     assert bond12.isBenzene(), "Cbf atom in {0} has a bond with an order other than 'B'".format(group)             # <<<<<<<<<<<<<<
 * 
 *             return (cbAtomList, cbfAtomList, cbfAtomList1, cbfAtomList2, connectedCbfs)
 */
      #ifndef CYTHON_WITHOUT_ASSERTIONS
      if (unlikely(!Py_OptimizeFlag)) {
        __pyx_t_14 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_14)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_14);
        __pyx_t_1 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_14))) {
          __pyx_t_1 = PyMethod_GET_SELF(__pyx_t_14);
          if (likely(__pyx_t_1)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_14);
            __Pyx_INCREF(__pyx_t_1);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_14, function);
          }
        }
        if (__pyx_t_1) {
          __pyx_t_12 = __Pyx_PyObject_CallOneArg(__pyx_t_14, __pyx_t_1); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        } else {
          __pyx_t_12 = __Pyx_PyObject_CallNoArg(__pyx_t_14); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_12);
        __Pyx_DECREF(__pyx_t_14); __pyx_t_14 = 0;
        __pyx_t_15 = __Pyx_PyObject_IsTrue(__pyx_t_12); if (unlikely(__pyx_t_15 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
        if (unlikely(!__pyx_t_15)) {
          __pyx_t_14 = __Pyx_PyObject_GetAttrStr(__pyx_kp_s_Cbf_atom_in_0_has_a_bond_with_an, __pyx_n_s_format); if (unlikely(!__pyx_t_14)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_14);
          __pyx_t_1 = NULL;
          if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_14))) {
            __pyx_t_1 = PyMethod_GET_SELF(__pyx_t_14);
            if (likely(__pyx_t_1)) {
              PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_14);
              __Pyx_INCREF(__pyx_t_1);
              __Pyx_INCREF(function);
              __Pyx_DECREF_SET(__pyx_t_14, function);
            }
          }
          if (!__pyx_t_1) {
            __pyx_t_12 = __Pyx_PyObject_CallOneArg(__pyx_t_14, __pyx_v_group); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_12);
          } else {
            __pyx_t_16 = PyTuple_New(1+1); if (unlikely(!__pyx_t_16)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_16);
            PyTuple_SET_ITEM(__pyx_t_16, 0, __pyx_t_1); __Pyx_GIVEREF(__pyx_t_1); __pyx_t_1 = NULL;
            __Pyx_INCREF(__pyx_v_group);
            PyTuple_SET_ITEM(__pyx_t_16, 0+1, __pyx_v_group);
            __Pyx_GIVEREF(__pyx_v_group);
            __pyx_t_12 = __Pyx_PyObject_Call(__pyx_t_14, __pyx_t_16, NULL); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_12);
            __Pyx_DECREF(__pyx_t_16); __pyx_t_16 = 0;
          }
          __Pyx_DECREF(__pyx_t_14); __pyx_t_14 = 0;
          __pyx_t_14 = PyTuple_Pack(1, __pyx_t_12); if (unlikely(!__pyx_t_14)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_14);
          __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
          PyErr_SetObject(PyExc_AssertionError, __pyx_t_14);
          __Pyx_DECREF(__pyx_t_14); __pyx_t_14 = 0;
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1337; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
      }
      #endif
    }
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1335
 * 
 *             #check that cbfAtoms only have benzene bonds
 *             for cbfAtom in cbfAtomList:             # <<<<<<<<<<<<<<
 *                 for atom2, bond12 in cbfAtom.bonds.iteritems():
 *                     assert bond12.isBenzene(), "Cbf atom in {0} has a bond with an order other than 'B'".format(group)
 */
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":1339
 *                     assert bond12.isBenzene(), "Cbf atom in {0} has a bond with an order other than 'B'".format(group)
 * 
 *             return (cbAtomList, cbfAtomList, cbfAtomList1, cbfAtomList2, connectedCbfs)             # <<<<<<<<<<<<<<
 * 
 *         def addBenzeneAtomToGroup(group, connectingAtom, cbf=False):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_3 = PyTuple_New(5); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1339; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_v_cbAtomList);
  PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_v_cbAtomList);
  __Pyx_GIVEREF(__pyx_v_cbAtomList);
  __Pyx_INCREF(__pyx_v_cbfAtomList);
  PyTuple_SET_ITEM(__pyx_t_3, 1, __pyx_v_cbfAtomList);
  __Pyx_GIVEREF(__pyx_v_cbfAtomList);
  __Pyx_INCREF(__pyx_v_cbfAtomList1);
  PyTuple_SET_ITEM(__pyx_t_3, 2, __pyx_v_cbfAtomList1);
  __Pyx_GIVEREF(__pyx_v_cbfAtomList1);
  __Pyx_INCREF(__pyx_v_cbfAtomList2);
  PyTuple_SET_ITEM(__pyx_t_3, 3, __pyx_v_cbfAtomList2);
  __Pyx_GIVEREF(__pyx_v_cbfAtomList2);
  __Pyx_INCREF(__pyx_v_connectedCbfs);
  PyTuple_SET_ITEM(__pyx_t_3, 4, __pyx_v_connectedCbfs);
  __Pyx_GIVEREF(__pyx_v_connectedCbfs);
  __pyx_r = __pyx_t_3;
  __pyx_t_3 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1283
 *         """
 * 
 *         def classifyBenzeneCarbons(group, partners = []):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_12);
  __Pyx_XDECREF(__pyx_t_14);
  __Pyx_XDECREF(__pyx_t_16);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.classifyBenzeneCarbons", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_cbAtomList);
  __Pyx_XDECREF(__pyx_v_cbfAtomList);
  __Pyx_XDECREF(__pyx_v_cbfAtomList1);
  __Pyx_XDECREF(__pyx_v_cbfAtomList2);
  __Pyx_XDECREF(__pyx_v_connectedCbfs);
  __Pyx_XDECREF(__pyx_v_labelsOfCarbonAtomTypes);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_benzeneBonds);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XDECREF(__pyx_v_cbfAtom);
  __Pyx_XDECREF(__pyx_v_fbBonds);
  __Pyx_XDECREF(__pyx_v_bond);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1341
 *             return (cbAtomList, cbfAtomList, cbfAtomList1, cbfAtomList2, connectedCbfs)
 * 
 *         def addBenzeneAtomToGroup(group, connectingAtom, cbf=False):             # <<<<<<<<<<<<<<
 *             """
 *             This function adds a benzene atom to the group
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_3addBenzeneAtomToGroup(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_2addBenzeneAtomToGroup[] = "\n            This function adds a benzene atom to the group\n\n            Args:\n                group: :class:Group that we want to add a benzene atom to\n                connectingAtom: :class:GroupAtom that is connected to the new benzene atom\n                cbf: boolean indicating if the new atom should be a cbf\n            Returns: a tuple containing the modified group and the new atom\n\n            ";
static PyMethodDef __pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_3addBenzeneAtomToGroup = {"addBenzeneAtomToGroup", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_3addBenzeneAtomToGroup, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_2addBenzeneAtomToGroup};
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_3addBenzeneAtomToGroup(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_group = 0;
  PyObject *__pyx_v_connectingAtom = 0;
  PyObject *__pyx_v_cbf = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addBenzeneAtomToGroup (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_group,&__pyx_n_s_connectingAtom,&__pyx_n_s_cbf,0};
    PyObject* values[3] = {0,0,0};
    values[2] = ((PyObject *)((PyObject *)Py_False));
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_group)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_connectingAtom)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("addBenzeneAtomToGroup", 0, 2, 3, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1341; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
        case  2:
        if (kw_args > 0) {
          PyObject* value = PyDict_GetItem(__pyx_kwds, __pyx_n_s_cbf);
          if (value) { values[2] = value; kw_args--; }
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "addBenzeneAtomToGroup") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1341; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else {
      switch (PyTuple_GET_SIZE(__pyx_args)) {
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        break;
        default: goto __pyx_L5_argtuple_error;
      }
    }
    __pyx_v_group = values[0];
    __pyx_v_connectingAtom = values[1];
    __pyx_v_cbf = values[2];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("addBenzeneAtomToGroup", 0, 2, 3, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1341; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.addBenzeneAtomToGroup", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_2addBenzeneAtomToGroup(__pyx_self, __pyx_v_group, __pyx_v_connectingAtom, __pyx_v_cbf);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_2addBenzeneAtomToGroup(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_group, PyObject *__pyx_v_connectingAtom, PyObject *__pyx_v_cbf) {
  PyObject *__pyx_v_newAtom = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_newBond = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addBenzeneAtomToGroup", 0);

  /* "rmgpy/molecule/group.py":1352
 * 
 *             """
 *             newAtom = None             # <<<<<<<<<<<<<<
 *             if cbf:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Cbf']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
  __Pyx_INCREF(Py_None);
  __pyx_v_newAtom = Py_None;

  /* "rmgpy/molecule/group.py":1353
 *             """
 *             newAtom = None
 *             if cbf:             # <<<<<<<<<<<<<<
 *                 newAtom = GroupAtom(atomType=[atomTypes['Cbf']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             else:
 */
  __pyx_t_1 = __Pyx_PyObject_IsTrue(__pyx_v_cbf); if (unlikely(__pyx_t_1 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1353; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__pyx_t_1) {

    /* "rmgpy/molecule/group.py":1354
 *             newAtom = None
 *             if cbf:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Cbf']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *             else:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Cb']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 */
    __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = PyObject_GetItem(__pyx_t_3, __pyx_n_s_Cbf); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    PyList_SET_ITEM(__pyx_t_3, 0, __pyx_t_4);
    __Pyx_GIVEREF(__pyx_t_4);
    __pyx_t_4 = 0;
    if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_atomType, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_INCREF(__pyx_int_0);
    PyList_SET_ITEM(__pyx_t_3, 0, __pyx_int_0);
    __Pyx_GIVEREF(__pyx_int_0);
    if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_radicalElectrons, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_charge, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1354; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __Pyx_DECREF_SET(__pyx_v_newAtom, __pyx_t_3);
    __pyx_t_3 = 0;
    goto __pyx_L3;
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":1356
 *                 newAtom = GroupAtom(atomType=[atomTypes['Cbf']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             else:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Cb']], radicalElectrons=[0], charge=[], label='', lonePairs=None)             # <<<<<<<<<<<<<<
 *             newBond = GroupBond(connectingAtom, newAtom, order=['B'])
 *             group.addAtom(newAtom)
 */
    __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_4 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_Cb); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyList_SET_ITEM(__pyx_t_2, 0, __pyx_t_4);
    __Pyx_GIVEREF(__pyx_t_4);
    __pyx_t_4 = 0;
    if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_atomType, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_INCREF(__pyx_int_0);
    PyList_SET_ITEM(__pyx_t_2, 0, __pyx_int_0);
    __Pyx_GIVEREF(__pyx_int_0);
    if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_radicalElectrons, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_charge, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_label, __pyx_kp_s_) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_lonePairs, Py_None) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom)), __pyx_empty_tuple, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1356; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_DECREF_SET(__pyx_v_newAtom, __pyx_t_2);
    __pyx_t_2 = 0;
  }
  __pyx_L3:;

  /* "rmgpy/molecule/group.py":1357
 *             else:
 *                 newAtom = GroupAtom(atomType=[atomTypes['Cb']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             newBond = GroupBond(connectingAtom, newAtom, order=['B'])             # <<<<<<<<<<<<<<
 *             group.addAtom(newAtom)
 *             group.addBond(newBond)
 */
  __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1357; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_INCREF(__pyx_v_connectingAtom);
  PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_connectingAtom);
  __Pyx_GIVEREF(__pyx_v_connectingAtom);
  __Pyx_INCREF(__pyx_v_newAtom);
  PyTuple_SET_ITEM(__pyx_t_2, 1, __pyx_v_newAtom);
  __Pyx_GIVEREF(__pyx_v_newAtom);
  __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1357; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1357; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_INCREF(__pyx_n_s_B);
  PyList_SET_ITEM(__pyx_t_4, 0, __pyx_n_s_B);
  __Pyx_GIVEREF(__pyx_n_s_B);
  if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_order, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1357; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1357; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_v_newBond = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":1358
 *                 newAtom = GroupAtom(atomType=[atomTypes['Cb']], radicalElectrons=[0], charge=[], label='', lonePairs=None)
 *             newBond = GroupBond(connectingAtom, newAtom, order=['B'])
 *             group.addAtom(newAtom)             # <<<<<<<<<<<<<<
 *             group.addBond(newBond)
 * 
 */
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_group, __pyx_n_s_addAtom); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1358; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_2 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (!__pyx_t_2) {
    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_v_newAtom); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1358; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
  } else {
    __pyx_t_5 = PyTuple_New(1+1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1358; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_2); __Pyx_GIVEREF(__pyx_t_2); __pyx_t_2 = NULL;
    __Pyx_INCREF(__pyx_v_newAtom);
    PyTuple_SET_ITEM(__pyx_t_5, 0+1, __pyx_v_newAtom);
    __Pyx_GIVEREF(__pyx_v_newAtom);
    __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_5, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1358; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":1359
 *             newBond = GroupBond(connectingAtom, newAtom, order=['B'])
 *             group.addAtom(newAtom)
 *             group.addBond(newBond)             # <<<<<<<<<<<<<<
 * 
 *             return (group, newAtom)
 */
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_group, __pyx_n_s_addBond); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1359; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_5 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_5)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_5);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
    }
  }
  if (!__pyx_t_5) {
    __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_3, ((PyObject *)__pyx_v_newBond)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1359; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
  } else {
    __pyx_t_2 = PyTuple_New(1+1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1359; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
    __Pyx_INCREF(((PyObject *)__pyx_v_newBond));
    PyTuple_SET_ITEM(__pyx_t_2, 0+1, ((PyObject *)__pyx_v_newBond));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_newBond));
    __pyx_t_4 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_t_2, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1359; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":1361
 *             group.addBond(newBond)
 * 
 *             return (group, newAtom)             # <<<<<<<<<<<<<<
 * 
 *         def sortByConnectivity(atomList):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_4 = PyTuple_New(2); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1361; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_INCREF(__pyx_v_group);
  PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_v_group);
  __Pyx_GIVEREF(__pyx_v_group);
  __Pyx_INCREF(__pyx_v_newAtom);
  PyTuple_SET_ITEM(__pyx_t_4, 1, __pyx_v_newAtom);
  __Pyx_GIVEREF(__pyx_v_newAtom);
  __pyx_r = __pyx_t_4;
  __pyx_t_4 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1341
 *             return (cbAtomList, cbfAtomList, cbfAtomList1, cbfAtomList2, connectedCbfs)
 * 
 *         def addBenzeneAtomToGroup(group, connectingAtom, cbf=False):             # <<<<<<<<<<<<<<
 *             """
 *             This function adds a benzene atom to the group
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.addBenzeneAtomToGroup", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_newAtom);
  __Pyx_XDECREF((PyObject *)__pyx_v_newBond);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1363
 *             return (group, newAtom)
 * 
 *         def sortByConnectivity(atomList):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_5sortByConnectivity(PyObject *__pyx_self, PyObject *__pyx_v_atomList); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_4sortByConnectivity[] = "\n            Args:\n                atomList: input list of atoms\n\n            Returns: a sorted list of atoms where each atom is connected to a previous\n            atom in the list if possible\n            ";
static PyMethodDef __pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_5sortByConnectivity = {"sortByConnectivity", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_5sortByConnectivity, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_4sortByConnectivity};
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_5sortByConnectivity(PyObject *__pyx_self, PyObject *__pyx_v_atomList) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("sortByConnectivity (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_4sortByConnectivity(__pyx_self, ((PyObject *)__pyx_v_atomList));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_4sortByConnectivity(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_atomList) {
  PyObject *__pyx_v_sortedAtomList = NULL;
  Py_ssize_t __pyx_v_previousLength;
  PyObject *__pyx_v_atom1 = NULL;
  int __pyx_v_added;
  PyObject *__pyx_v_atom2 = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_t_2;
  int __pyx_t_3;
  Py_ssize_t __pyx_t_4;
  PyObject *(*__pyx_t_5)(PyObject *);
  PyObject *__pyx_t_6 = NULL;
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  PyObject *__pyx_t_10 = NULL;
  PyObject *__pyx_t_11 = NULL;
  int __pyx_t_12;
  int __pyx_t_13;
  PyObject *__pyx_t_14 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("sortByConnectivity", 0);

  /* "rmgpy/molecule/group.py":1371
 *             atom in the list if possible
 *             """
 *             sortedAtomList=[]             # <<<<<<<<<<<<<<
 *             while atomList:
 *                 sortedAtomList.append(atomList.pop(0))
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1371; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_sortedAtomList = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1372
 *             """
 *             sortedAtomList=[]
 *             while atomList:             # <<<<<<<<<<<<<<
 *                 sortedAtomList.append(atomList.pop(0))
 *                 previousLength = len(sortedAtomList)
 */
  while (1) {
    __pyx_t_2 = __Pyx_PyObject_IsTrue(__pyx_v_atomList); if (unlikely(__pyx_t_2 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1372; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (!__pyx_t_2) break;

    /* "rmgpy/molecule/group.py":1373
 *             sortedAtomList=[]
 *             while atomList:
 *                 sortedAtomList.append(atomList.pop(0))             # <<<<<<<<<<<<<<
 *                 previousLength = len(sortedAtomList)
 *                 for atom1 in atomList:
 */
    __pyx_t_1 = __Pyx_PyObject_PopIndex(__pyx_v_atomList, 0, 1, Py_ssize_t, PyInt_FromSsize_t); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1373; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_3 = __Pyx_PyList_Append(__pyx_v_sortedAtomList, __pyx_t_1); if (unlikely(__pyx_t_3 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1373; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1374
 *             while atomList:
 *                 sortedAtomList.append(atomList.pop(0))
 *                 previousLength = len(sortedAtomList)             # <<<<<<<<<<<<<<
 *                 for atom1 in atomList:
 *                     added = False
 */
    __pyx_t_4 = PyList_GET_SIZE(__pyx_v_sortedAtomList); if (unlikely(__pyx_t_4 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1374; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_v_previousLength = __pyx_t_4;

    /* "rmgpy/molecule/group.py":1375
 *                 sortedAtomList.append(atomList.pop(0))
 *                 previousLength = len(sortedAtomList)
 *                 for atom1 in atomList:             # <<<<<<<<<<<<<<
 *                     added = False
 *                     for atom2, bond12 in atom1.bonds.iteritems():
 */
    if (likely(PyList_CheckExact(__pyx_v_atomList)) || PyTuple_CheckExact(__pyx_v_atomList)) {
      __pyx_t_1 = __pyx_v_atomList; __Pyx_INCREF(__pyx_t_1); __pyx_t_4 = 0;
      __pyx_t_5 = NULL;
    } else {
      __pyx_t_4 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_v_atomList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1375; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_5 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1375; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    for (;;) {
      if (likely(!__pyx_t_5)) {
        if (likely(PyList_CheckExact(__pyx_t_1))) {
          if (__pyx_t_4 >= PyList_GET_SIZE(__pyx_t_1)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_6 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_4); __Pyx_INCREF(__pyx_t_6); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1375; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_6 = PySequence_ITEM(__pyx_t_1, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1375; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        } else {
          if (__pyx_t_4 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_6 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_4); __Pyx_INCREF(__pyx_t_6); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1375; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_6 = PySequence_ITEM(__pyx_t_1, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1375; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        }
      } else {
        __pyx_t_6 = __pyx_t_5(__pyx_t_1);
        if (unlikely(!__pyx_t_6)) {
          PyObject* exc_type = PyErr_Occurred();
          if (exc_type) {
            if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
            else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1375; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          break;
        }
        __Pyx_GOTREF(__pyx_t_6);
      }
      __Pyx_XDECREF_SET(__pyx_v_atom1, __pyx_t_6);
      __pyx_t_6 = 0;

      /* "rmgpy/molecule/group.py":1376
 *                 previousLength = len(sortedAtomList)
 *                 for atom1 in atomList:
 *                     added = False             # <<<<<<<<<<<<<<
 *                     for atom2, bond12 in atom1.bonds.iteritems():
 *                         if atom2 in sortedAtomList:
 */
      __pyx_v_added = 0;

      /* "rmgpy/molecule/group.py":1377
 *                 for atom1 in atomList:
 *                     added = False
 *                     for atom2, bond12 in atom1.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                         if atom2 in sortedAtomList:
 *                             sortedAtomList.append(atom1)
 */
      __pyx_t_7 = 0;
      __pyx_t_10 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom1, __pyx_n_s_bonds); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1377; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_10);
      if (unlikely(__pyx_t_10 == Py_None)) {
        PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1377; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_11 = __Pyx_dict_iterator(__pyx_t_10, 0, __pyx_n_s_iteritems, (&__pyx_t_8), (&__pyx_t_9)); if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1377; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_11);
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      __Pyx_XDECREF(__pyx_t_6);
      __pyx_t_6 = __pyx_t_11;
      __pyx_t_11 = 0;
      while (1) {
        __pyx_t_12 = __Pyx_dict_iter_next(__pyx_t_6, __pyx_t_8, &__pyx_t_7, &__pyx_t_11, &__pyx_t_10, NULL, __pyx_t_9);
        if (unlikely(__pyx_t_12 == 0)) break;
        if (unlikely(__pyx_t_12 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1377; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_11);
        __Pyx_GOTREF(__pyx_t_10);
        __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_11);
        __pyx_t_11 = 0;
        __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_10);
        __pyx_t_10 = 0;

        /* "rmgpy/molecule/group.py":1378
 *                     added = False
 *                     for atom2, bond12 in atom1.bonds.iteritems():
 *                         if atom2 in sortedAtomList:             # <<<<<<<<<<<<<<
 *                             sortedAtomList.append(atom1)
 *                             added = True
 */
        __pyx_t_2 = (__Pyx_PySequence_Contains(__pyx_v_atom2, __pyx_v_sortedAtomList, Py_EQ)); if (unlikely(__pyx_t_2 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1378; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __pyx_t_13 = (__pyx_t_2 != 0);
        if (__pyx_t_13) {

          /* "rmgpy/molecule/group.py":1379
 *                     for atom2, bond12 in atom1.bonds.iteritems():
 *                         if atom2 in sortedAtomList:
 *                             sortedAtomList.append(atom1)             # <<<<<<<<<<<<<<
 *                             added = True
 *                             break
 */
          __pyx_t_3 = __Pyx_PyList_Append(__pyx_v_sortedAtomList, __pyx_v_atom1); if (unlikely(__pyx_t_3 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1379; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

          /* "rmgpy/molecule/group.py":1380
 *                         if atom2 in sortedAtomList:
 *                             sortedAtomList.append(atom1)
 *                             added = True             # <<<<<<<<<<<<<<
 *                             break
 *                     if added : continue
 */
          __pyx_v_added = 1;

          /* "rmgpy/molecule/group.py":1381
 *                             sortedAtomList.append(atom1)
 *                             added = True
 *                             break             # <<<<<<<<<<<<<<
 *                     if added : continue
 *                 for atom1 in sortedAtomList[previousLength:]:
 */
          goto __pyx_L8_break;
        }
      }
      __pyx_L8_break:;
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;

      /* "rmgpy/molecule/group.py":1382
 *                             added = True
 *                             break
 *                     if added : continue             # <<<<<<<<<<<<<<
 *                 for atom1 in sortedAtomList[previousLength:]:
 *                     atomList.remove(atom1)
 */
      __pyx_t_13 = (__pyx_v_added != 0);
      if (__pyx_t_13) {
        goto __pyx_L5_continue;
      }

      /* "rmgpy/molecule/group.py":1375
 *                 sortedAtomList.append(atomList.pop(0))
 *                 previousLength = len(sortedAtomList)
 *                 for atom1 in atomList:             # <<<<<<<<<<<<<<
 *                     added = False
 *                     for atom2, bond12 in atom1.bonds.iteritems():
 */
      __pyx_L5_continue:;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1383
 *                             break
 *                     if added : continue
 *                 for atom1 in sortedAtomList[previousLength:]:             # <<<<<<<<<<<<<<
 *                     atomList.remove(atom1)
 * 
 */
    __pyx_t_1 = __Pyx_PyList_GetSlice(__pyx_v_sortedAtomList, __pyx_v_previousLength, PY_SSIZE_T_MAX); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1383; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_6 = __pyx_t_1; __Pyx_INCREF(__pyx_t_6); __pyx_t_4 = 0;
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    for (;;) {
      if (__pyx_t_4 >= PyList_GET_SIZE(__pyx_t_6)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_1 = PyList_GET_ITEM(__pyx_t_6, __pyx_t_4); __Pyx_INCREF(__pyx_t_1); __pyx_t_4++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1383; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_1 = PySequence_ITEM(__pyx_t_6, __pyx_t_4); __pyx_t_4++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1383; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_atom1, __pyx_t_1);
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1384
 *                     if added : continue
 *                 for atom1 in sortedAtomList[previousLength:]:
 *                     atomList.remove(atom1)             # <<<<<<<<<<<<<<
 * 
 *             return sortedAtomList
 */
      __pyx_t_10 = __Pyx_PyObject_GetAttrStr(__pyx_v_atomList, __pyx_n_s_remove); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1384; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_10);
      __pyx_t_11 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_10))) {
        __pyx_t_11 = PyMethod_GET_SELF(__pyx_t_10);
        if (likely(__pyx_t_11)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_10);
          __Pyx_INCREF(__pyx_t_11);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_10, function);
        }
      }
      if (!__pyx_t_11) {
        __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_10, __pyx_v_atom1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1384; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_1);
      } else {
        __pyx_t_14 = PyTuple_New(1+1); if (unlikely(!__pyx_t_14)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1384; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_14);
        PyTuple_SET_ITEM(__pyx_t_14, 0, __pyx_t_11); __Pyx_GIVEREF(__pyx_t_11); __pyx_t_11 = NULL;
        __Pyx_INCREF(__pyx_v_atom1);
        PyTuple_SET_ITEM(__pyx_t_14, 0+1, __pyx_v_atom1);
        __Pyx_GIVEREF(__pyx_v_atom1);
        __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_10, __pyx_t_14, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1384; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_1);
        __Pyx_DECREF(__pyx_t_14); __pyx_t_14 = 0;
      }
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1383
 *                             break
 *                     if added : continue
 *                 for atom1 in sortedAtomList[previousLength:]:             # <<<<<<<<<<<<<<
 *                     atomList.remove(atom1)
 * 
 */
    }
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  }

  /* "rmgpy/molecule/group.py":1386
 *                     atomList.remove(atom1)
 * 
 *             return sortedAtomList             # <<<<<<<<<<<<<<
 * 
 *         def checkSet(superList, subList):
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_sortedAtomList);
  __pyx_r = __pyx_v_sortedAtomList;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1363
 *             return (group, newAtom)
 * 
 *         def sortByConnectivity(atomList):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_10);
  __Pyx_XDECREF(__pyx_t_11);
  __Pyx_XDECREF(__pyx_t_14);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.sortByConnectivity", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_sortedAtomList);
  __Pyx_XDECREF(__pyx_v_atom1);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1388
 *             return sortedAtomList
 * 
 *         def checkSet(superList, subList):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_7checkSet(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_6checkSet[] = "\n            Args:\n                superList: list to check if superset of partList\n                subList:  list to check if subset of superList\n\n            Returns: Boolean to see if superList is a superset of subList\n\n            ";
static PyMethodDef __pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_7checkSet = {"checkSet", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_7checkSet, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_6checkSet};
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_7checkSet(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_superList = 0;
  PyObject *__pyx_v_subList = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("checkSet (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_superList,&__pyx_n_s_subList,0};
    PyObject* values[2] = {0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_superList)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_subList)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("checkSet", 1, 2, 2, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1388; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "checkSet") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1388; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
    }
    __pyx_v_superList = values[0];
    __pyx_v_subList = values[1];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("checkSet", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1388; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.checkSet", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_6checkSet(__pyx_self, __pyx_v_superList, __pyx_v_subList);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_6checkSet(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_superList, PyObject *__pyx_v_subList) {
  PyObject *__pyx_v_superSet = NULL;
  PyObject *__pyx_v_subSet = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("checkSet", 0);

  /* "rmgpy/molecule/group.py":1397
 * 
 *             """
 *             superSet = set(superList)             # <<<<<<<<<<<<<<
 *             subSet = set(subList)
 *             return superSet.issuperset(subSet)
 */
  __pyx_t_1 = PySet_New(__pyx_v_superList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1397; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_superSet = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1398
 *             """
 *             superSet = set(superList)
 *             subSet = set(subList)             # <<<<<<<<<<<<<<
 *             return superSet.issuperset(subSet)
 * 
 */
  __pyx_t_1 = PySet_New(__pyx_v_subList); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1398; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_subSet = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1399
 *             superSet = set(superList)
 *             subSet = set(subList)
 *             return superSet.issuperset(subSet)             # <<<<<<<<<<<<<<
 * 
 *         def mergeOverlappingBenzeneRings(ring1, ring2, od):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_superSet, __pyx_n_s_issuperset); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1399; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_v_subSet); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1399; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1399; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(__pyx_v_subSet);
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, __pyx_v_subSet);
    __Pyx_GIVEREF(__pyx_v_subSet);
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1399; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1388
 *             return sortedAtomList
 * 
 *         def checkSet(superList, subList):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.checkSet", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_superSet);
  __Pyx_XDECREF(__pyx_v_subSet);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1401
 *             return superSet.issuperset(subSet)
 * 
 *         def mergeOverlappingBenzeneRings(ring1, ring2, od):             # <<<<<<<<<<<<<<
 *             """
 *             The input arguements of rings are always in the order that the atoms appear
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_9mergeOverlappingBenzeneRings(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_8mergeOverlappingBenzeneRings[] = "\n            The input arguements of rings are always in the order that the atoms appear\n            inside the ring. That is, each atom is connected to the ones adjacent on the\n            list.\n\n            Args:\n                ring1: list of :class:GroupAtoms representing first partial ring to merge\n                ring2: list :class:GroupAtoms representing second partial ring to merge\n                od: in for overlap distance\n\n            This function tries to see if the beginning or ends of each list have the\n            same atom objects, i.e the two part rings should be merged together.\n\n            Returns: If rings are mergable, returns a new list of the merged ring, otherwise\n            an empty list\n\n            ";
static PyMethodDef __pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_9mergeOverlappingBenzeneRings = {"mergeOverlappingBenzeneRings", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_9mergeOverlappingBenzeneRings, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_8mergeOverlappingBenzeneRings};
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_9mergeOverlappingBenzeneRings(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_ring1 = 0;
  PyObject *__pyx_v_ring2 = 0;
  PyObject *__pyx_v_od = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("mergeOverlappingBenzeneRings (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_ring1,&__pyx_n_s_ring2,&__pyx_n_s_od,0};
    PyObject* values[3] = {0,0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_ring1)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_ring2)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("mergeOverlappingBenzeneRings", 1, 3, 3, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1401; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
        case  2:
        if (likely((values[2] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_od)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("mergeOverlappingBenzeneRings", 1, 3, 3, 2); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1401; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "mergeOverlappingBenzeneRings") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1401; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else if (PyTuple_GET_SIZE(__pyx_args) != 3) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
      values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
    }
    __pyx_v_ring1 = values[0];
    __pyx_v_ring2 = values[1];
    __pyx_v_od = values[2];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("mergeOverlappingBenzeneRings", 1, 3, 3, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1401; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.mergeOverlappingBenzeneRings", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_8mergeOverlappingBenzeneRings(__pyx_self, __pyx_v_ring1, __pyx_v_ring2, __pyx_v_od);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_8mergeOverlappingBenzeneRings(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_ring1, PyObject *__pyx_v_ring2, PyObject *__pyx_v_od) {
  PyObject *__pyx_v_newRing = NULL;
  PyObject *__pyx_v_matchList1 = NULL;
  PyObject *__pyx_v_matchList2 = NULL;
  PyObject *__pyx_v_matchList3 = NULL;
  PyObject *__pyx_v_matchList4 = NULL;
  PyObject *__pyx_v_x1 = NULL;
  PyObject *__pyx_v_x2 = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_t_2;
  Py_ssize_t __pyx_t_3;
  int __pyx_t_4;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  PyObject *__pyx_t_7 = NULL;
  PyObject *(*__pyx_t_8)(PyObject *);
  PyObject *__pyx_t_9 = NULL;
  PyObject *__pyx_t_10 = NULL;
  PyObject *(*__pyx_t_11)(PyObject *);
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("mergeOverlappingBenzeneRings", 0);

  /* "rmgpy/molecule/group.py":1419
 * 
 *             """
 *             newRing = []             # <<<<<<<<<<<<<<
 *             #ring already complete
 *             if len(ring1) ==6 or len(ring2) == 6: return newRing
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1419; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_newRing = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1421
 *             newRing = []
 *             #ring already complete
 *             if len(ring1) ==6 or len(ring2) == 6: return newRing             # <<<<<<<<<<<<<<
 * 
 *             #start of ring1 matches end of ring2
 */
  __pyx_t_3 = PyObject_Length(__pyx_v_ring1); if (unlikely(__pyx_t_3 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1421; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_4 = ((__pyx_t_3 == 6) != 0);
  if (!__pyx_t_4) {
    goto __pyx_L5_next_or;
  } else {
    __pyx_t_2 = __pyx_t_4;
    goto __pyx_L4_bool_binop_done;
  }
  __pyx_L5_next_or:;
  __pyx_t_3 = PyObject_Length(__pyx_v_ring2); if (unlikely(__pyx_t_3 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1421; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_4 = ((__pyx_t_3 == 6) != 0);
  __pyx_t_2 = __pyx_t_4;
  __pyx_L4_bool_binop_done:;
  if (__pyx_t_2) {
    __Pyx_XDECREF(__pyx_r);
    __Pyx_INCREF(__pyx_v_newRing);
    __pyx_r = __pyx_v_newRing;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":1424
 * 
 *             #start of ring1 matches end of ring2
 *             matchList1 = [x1 is x2 for x1,x2 in zip(ring1[-od:],ring2[:od])]             # <<<<<<<<<<<<<<
 *             #end of ring1 matches end of ring2
 *             matchList2 = [x1 is x2 for x1,x2 in zip(ring1[-od:],ring2[:od-1:-1])]
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_5 = PyNumber_Negative(__pyx_v_od); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  __pyx_t_6 = __Pyx_PyObject_GetSlice(__pyx_v_ring1, 0, 0, &__pyx_t_5, NULL, NULL, 0, 0, 1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  __pyx_t_5 = __Pyx_PyObject_GetSlice(__pyx_v_ring2, 0, 0, NULL, &__pyx_v_od, NULL, 0, 0, 1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  __pyx_t_7 = PyTuple_New(2); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_7);
  PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_6);
  __Pyx_GIVEREF(__pyx_t_6);
  PyTuple_SET_ITEM(__pyx_t_7, 1, __pyx_t_5);
  __Pyx_GIVEREF(__pyx_t_5);
  __pyx_t_6 = 0;
  __pyx_t_5 = 0;
  __pyx_t_5 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_7, NULL); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  if (likely(PyList_CheckExact(__pyx_t_5)) || PyTuple_CheckExact(__pyx_t_5)) {
    __pyx_t_7 = __pyx_t_5; __Pyx_INCREF(__pyx_t_7); __pyx_t_3 = 0;
    __pyx_t_8 = NULL;
  } else {
    __pyx_t_3 = -1; __pyx_t_7 = PyObject_GetIter(__pyx_t_5); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    __pyx_t_8 = Py_TYPE(__pyx_t_7)->tp_iternext; if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  for (;;) {
    if (likely(!__pyx_t_8)) {
      if (likely(PyList_CheckExact(__pyx_t_7))) {
        if (__pyx_t_3 >= PyList_GET_SIZE(__pyx_t_7)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_5 = PyList_GET_ITEM(__pyx_t_7, __pyx_t_3); __Pyx_INCREF(__pyx_t_5); __pyx_t_3++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_5 = PySequence_ITEM(__pyx_t_7, __pyx_t_3); __pyx_t_3++; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_3 >= PyTuple_GET_SIZE(__pyx_t_7)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_5 = PyTuple_GET_ITEM(__pyx_t_7, __pyx_t_3); __Pyx_INCREF(__pyx_t_5); __pyx_t_3++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_5 = PySequence_ITEM(__pyx_t_7, __pyx_t_3); __pyx_t_3++; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_5 = __pyx_t_8(__pyx_t_7);
      if (unlikely(!__pyx_t_5)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_5);
    }
    if ((likely(PyTuple_CheckExact(__pyx_t_5))) || (PyList_CheckExact(__pyx_t_5))) {
      PyObject* sequence = __pyx_t_5;
      #if CYTHON_COMPILING_IN_CPYTHON
      Py_ssize_t size = Py_SIZE(sequence);
      #else
      Py_ssize_t size = PySequence_Size(sequence);
      #endif
      if (unlikely(size != 2)) {
        if (size > 2) __Pyx_RaiseTooManyValuesError(2);
        else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      #if CYTHON_COMPILING_IN_CPYTHON
      if (likely(PyTuple_CheckExact(sequence))) {
        __pyx_t_6 = PyTuple_GET_ITEM(sequence, 0); 
        __pyx_t_9 = PyTuple_GET_ITEM(sequence, 1); 
      } else {
        __pyx_t_6 = PyList_GET_ITEM(sequence, 0); 
        __pyx_t_9 = PyList_GET_ITEM(sequence, 1); 
      }
      __Pyx_INCREF(__pyx_t_6);
      __Pyx_INCREF(__pyx_t_9);
      #else
      __pyx_t_6 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __pyx_t_9 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_9);
      #endif
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    } else {
      Py_ssize_t index = -1;
      __pyx_t_10 = PyObject_GetIter(__pyx_t_5); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_10);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      __pyx_t_11 = Py_TYPE(__pyx_t_10)->tp_iternext;
      index = 0; __pyx_t_6 = __pyx_t_11(__pyx_t_10); if (unlikely(!__pyx_t_6)) goto __pyx_L8_unpacking_failed;
      __Pyx_GOTREF(__pyx_t_6);
      index = 1; __pyx_t_9 = __pyx_t_11(__pyx_t_10); if (unlikely(!__pyx_t_9)) goto __pyx_L8_unpacking_failed;
      __Pyx_GOTREF(__pyx_t_9);
      if (__Pyx_IternextUnpackEndCheck(__pyx_t_11(__pyx_t_10), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_11 = NULL;
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      goto __pyx_L9_unpacking_done;
      __pyx_L8_unpacking_failed:;
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      __pyx_t_11 = NULL;
      if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_L9_unpacking_done:;
    }
    __Pyx_XDECREF_SET(__pyx_v_x1, __pyx_t_6);
    __pyx_t_6 = 0;
    __Pyx_XDECREF_SET(__pyx_v_x2, __pyx_t_9);
    __pyx_t_9 = 0;
    __pyx_t_2 = (__pyx_v_x1 == __pyx_v_x2);
    __pyx_t_5 = __Pyx_PyBool_FromLong(__pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_5))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1424; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  }
  __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  __pyx_v_matchList1 = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1426
 *             matchList1 = [x1 is x2 for x1,x2 in zip(ring1[-od:],ring2[:od])]
 *             #end of ring1 matches end of ring2
 *             matchList2 = [x1 is x2 for x1,x2 in zip(ring1[-od:],ring2[:od-1:-1])]             # <<<<<<<<<<<<<<
 *             #start of ring1 matches end of ring2
 *             matchList3 = [x1 is x2 for x1,x2 in zip(ring1[:od],ring2[-od:])]
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_7 = PyNumber_Negative(__pyx_v_od); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_7);
  __pyx_t_5 = __Pyx_PyObject_GetSlice(__pyx_v_ring1, 0, 0, &__pyx_t_7, NULL, NULL, 0, 0, 1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  __pyx_t_7 = PyNumber_Subtract(__pyx_v_od, __pyx_int_1); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_7);
  __pyx_t_9 = PySlice_New(Py_None, __pyx_t_7, __pyx_int_neg_1); if (unlikely(!__pyx_t_9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_9);
  __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  __pyx_t_7 = PyObject_GetItem(__pyx_v_ring2, __pyx_t_9); if (unlikely(__pyx_t_7 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_7);
  __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
  __pyx_t_9 = PyTuple_New(2); if (unlikely(!__pyx_t_9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_9);
  PyTuple_SET_ITEM(__pyx_t_9, 0, __pyx_t_5);
  __Pyx_GIVEREF(__pyx_t_5);
  PyTuple_SET_ITEM(__pyx_t_9, 1, __pyx_t_7);
  __Pyx_GIVEREF(__pyx_t_7);
  __pyx_t_5 = 0;
  __pyx_t_7 = 0;
  __pyx_t_7 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_9, NULL); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_7);
  __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
  if (likely(PyList_CheckExact(__pyx_t_7)) || PyTuple_CheckExact(__pyx_t_7)) {
    __pyx_t_9 = __pyx_t_7; __Pyx_INCREF(__pyx_t_9); __pyx_t_3 = 0;
    __pyx_t_8 = NULL;
  } else {
    __pyx_t_3 = -1; __pyx_t_9 = PyObject_GetIter(__pyx_t_7); if (unlikely(!__pyx_t_9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_9);
    __pyx_t_8 = Py_TYPE(__pyx_t_9)->tp_iternext; if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  for (;;) {
    if (likely(!__pyx_t_8)) {
      if (likely(PyList_CheckExact(__pyx_t_9))) {
        if (__pyx_t_3 >= PyList_GET_SIZE(__pyx_t_9)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_7 = PyList_GET_ITEM(__pyx_t_9, __pyx_t_3); __Pyx_INCREF(__pyx_t_7); __pyx_t_3++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_7 = PySequence_ITEM(__pyx_t_9, __pyx_t_3); __pyx_t_3++; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_3 >= PyTuple_GET_SIZE(__pyx_t_9)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_7 = PyTuple_GET_ITEM(__pyx_t_9, __pyx_t_3); __Pyx_INCREF(__pyx_t_7); __pyx_t_3++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_7 = PySequence_ITEM(__pyx_t_9, __pyx_t_3); __pyx_t_3++; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_7 = __pyx_t_8(__pyx_t_9);
      if (unlikely(!__pyx_t_7)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_7);
    }
    if ((likely(PyTuple_CheckExact(__pyx_t_7))) || (PyList_CheckExact(__pyx_t_7))) {
      PyObject* sequence = __pyx_t_7;
      #if CYTHON_COMPILING_IN_CPYTHON
      Py_ssize_t size = Py_SIZE(sequence);
      #else
      Py_ssize_t size = PySequence_Size(sequence);
      #endif
      if (unlikely(size != 2)) {
        if (size > 2) __Pyx_RaiseTooManyValuesError(2);
        else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      #if CYTHON_COMPILING_IN_CPYTHON
      if (likely(PyTuple_CheckExact(sequence))) {
        __pyx_t_5 = PyTuple_GET_ITEM(sequence, 0); 
        __pyx_t_6 = PyTuple_GET_ITEM(sequence, 1); 
      } else {
        __pyx_t_5 = PyList_GET_ITEM(sequence, 0); 
        __pyx_t_6 = PyList_GET_ITEM(sequence, 1); 
      }
      __Pyx_INCREF(__pyx_t_5);
      __Pyx_INCREF(__pyx_t_6);
      #else
      __pyx_t_5 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      __pyx_t_6 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      #endif
      __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
    } else {
      Py_ssize_t index = -1;
      __pyx_t_10 = PyObject_GetIter(__pyx_t_7); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_10);
      __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
      __pyx_t_11 = Py_TYPE(__pyx_t_10)->tp_iternext;
      index = 0; __pyx_t_5 = __pyx_t_11(__pyx_t_10); if (unlikely(!__pyx_t_5)) goto __pyx_L12_unpacking_failed;
      __Pyx_GOTREF(__pyx_t_5);
      index = 1; __pyx_t_6 = __pyx_t_11(__pyx_t_10); if (unlikely(!__pyx_t_6)) goto __pyx_L12_unpacking_failed;
      __Pyx_GOTREF(__pyx_t_6);
      if (__Pyx_IternextUnpackEndCheck(__pyx_t_11(__pyx_t_10), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_11 = NULL;
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      goto __pyx_L13_unpacking_done;
      __pyx_L12_unpacking_failed:;
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      __pyx_t_11 = NULL;
      if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_L13_unpacking_done:;
    }
    __Pyx_XDECREF_SET(__pyx_v_x1, __pyx_t_5);
    __pyx_t_5 = 0;
    __Pyx_XDECREF_SET(__pyx_v_x2, __pyx_t_6);
    __pyx_t_6 = 0;
    __pyx_t_2 = (__pyx_v_x1 == __pyx_v_x2);
    __pyx_t_7 = __Pyx_PyBool_FromLong(__pyx_t_2); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_7))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1426; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  }
  __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
  __pyx_v_matchList2 = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1428
 *             matchList2 = [x1 is x2 for x1,x2 in zip(ring1[-od:],ring2[:od-1:-1])]
 *             #start of ring1 matches end of ring2
 *             matchList3 = [x1 is x2 for x1,x2 in zip(ring1[:od],ring2[-od:])]             # <<<<<<<<<<<<<<
 *             #start of ring1 matches start of ring2
 *             matchList4 = [x1 is x2 for x1,x2 in zip(ring1[:od],ring2[od::-1])]
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_9 = __Pyx_PyObject_GetSlice(__pyx_v_ring1, 0, 0, NULL, &__pyx_v_od, NULL, 0, 0, 1); if (unlikely(!__pyx_t_9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_9);
  __pyx_t_7 = PyNumber_Negative(__pyx_v_od); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_7);
  __pyx_t_6 = __Pyx_PyObject_GetSlice(__pyx_v_ring2, 0, 0, &__pyx_t_7, NULL, NULL, 0, 0, 1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  __pyx_t_7 = PyTuple_New(2); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_7);
  PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_9);
  __Pyx_GIVEREF(__pyx_t_9);
  PyTuple_SET_ITEM(__pyx_t_7, 1, __pyx_t_6);
  __Pyx_GIVEREF(__pyx_t_6);
  __pyx_t_9 = 0;
  __pyx_t_6 = 0;
  __pyx_t_6 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_7, NULL); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  if (likely(PyList_CheckExact(__pyx_t_6)) || PyTuple_CheckExact(__pyx_t_6)) {
    __pyx_t_7 = __pyx_t_6; __Pyx_INCREF(__pyx_t_7); __pyx_t_3 = 0;
    __pyx_t_8 = NULL;
  } else {
    __pyx_t_3 = -1; __pyx_t_7 = PyObject_GetIter(__pyx_t_6); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    __pyx_t_8 = Py_TYPE(__pyx_t_7)->tp_iternext; if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  for (;;) {
    if (likely(!__pyx_t_8)) {
      if (likely(PyList_CheckExact(__pyx_t_7))) {
        if (__pyx_t_3 >= PyList_GET_SIZE(__pyx_t_7)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_6 = PyList_GET_ITEM(__pyx_t_7, __pyx_t_3); __Pyx_INCREF(__pyx_t_6); __pyx_t_3++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_6 = PySequence_ITEM(__pyx_t_7, __pyx_t_3); __pyx_t_3++; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_3 >= PyTuple_GET_SIZE(__pyx_t_7)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_6 = PyTuple_GET_ITEM(__pyx_t_7, __pyx_t_3); __Pyx_INCREF(__pyx_t_6); __pyx_t_3++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_6 = PySequence_ITEM(__pyx_t_7, __pyx_t_3); __pyx_t_3++; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_6 = __pyx_t_8(__pyx_t_7);
      if (unlikely(!__pyx_t_6)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_6);
    }
    if ((likely(PyTuple_CheckExact(__pyx_t_6))) || (PyList_CheckExact(__pyx_t_6))) {
      PyObject* sequence = __pyx_t_6;
      #if CYTHON_COMPILING_IN_CPYTHON
      Py_ssize_t size = Py_SIZE(sequence);
      #else
      Py_ssize_t size = PySequence_Size(sequence);
      #endif
      if (unlikely(size != 2)) {
        if (size > 2) __Pyx_RaiseTooManyValuesError(2);
        else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      #if CYTHON_COMPILING_IN_CPYTHON
      if (likely(PyTuple_CheckExact(sequence))) {
        __pyx_t_9 = PyTuple_GET_ITEM(sequence, 0); 
        __pyx_t_5 = PyTuple_GET_ITEM(sequence, 1); 
      } else {
        __pyx_t_9 = PyList_GET_ITEM(sequence, 0); 
        __pyx_t_5 = PyList_GET_ITEM(sequence, 1); 
      }
      __Pyx_INCREF(__pyx_t_9);
      __Pyx_INCREF(__pyx_t_5);
      #else
      __pyx_t_9 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_9);
      __pyx_t_5 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      #endif
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    } else {
      Py_ssize_t index = -1;
      __pyx_t_10 = PyObject_GetIter(__pyx_t_6); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_10);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __pyx_t_11 = Py_TYPE(__pyx_t_10)->tp_iternext;
      index = 0; __pyx_t_9 = __pyx_t_11(__pyx_t_10); if (unlikely(!__pyx_t_9)) goto __pyx_L16_unpacking_failed;
      __Pyx_GOTREF(__pyx_t_9);
      index = 1; __pyx_t_5 = __pyx_t_11(__pyx_t_10); if (unlikely(!__pyx_t_5)) goto __pyx_L16_unpacking_failed;
      __Pyx_GOTREF(__pyx_t_5);
      if (__Pyx_IternextUnpackEndCheck(__pyx_t_11(__pyx_t_10), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_11 = NULL;
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      goto __pyx_L17_unpacking_done;
      __pyx_L16_unpacking_failed:;
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      __pyx_t_11 = NULL;
      if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_L17_unpacking_done:;
    }
    __Pyx_XDECREF_SET(__pyx_v_x1, __pyx_t_9);
    __pyx_t_9 = 0;
    __Pyx_XDECREF_SET(__pyx_v_x2, __pyx_t_5);
    __pyx_t_5 = 0;
    __pyx_t_2 = (__pyx_v_x1 == __pyx_v_x2);
    __pyx_t_6 = __Pyx_PyBool_FromLong(__pyx_t_2); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_6))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1428; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  }
  __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  __pyx_v_matchList3 = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1430
 *             matchList3 = [x1 is x2 for x1,x2 in zip(ring1[:od],ring2[-od:])]
 *             #start of ring1 matches start of ring2
 *             matchList4 = [x1 is x2 for x1,x2 in zip(ring1[:od],ring2[od::-1])]             # <<<<<<<<<<<<<<
 *             if not False in matchList1:
 *                 newRing = ring1 +ring2[od:]
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_7 = __Pyx_PyObject_GetSlice(__pyx_v_ring1, 0, 0, NULL, &__pyx_v_od, NULL, 0, 0, 1); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_7);
  __pyx_t_6 = PySlice_New(__pyx_v_od, Py_None, __pyx_int_neg_1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __pyx_t_5 = PyObject_GetItem(__pyx_v_ring2, __pyx_t_6); if (unlikely(__pyx_t_5 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_5);
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  __pyx_t_6 = PyTuple_New(2); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  PyTuple_SET_ITEM(__pyx_t_6, 0, __pyx_t_7);
  __Pyx_GIVEREF(__pyx_t_7);
  PyTuple_SET_ITEM(__pyx_t_6, 1, __pyx_t_5);
  __Pyx_GIVEREF(__pyx_t_5);
  __pyx_t_7 = 0;
  __pyx_t_5 = 0;
  __pyx_t_5 = __Pyx_PyObject_Call(__pyx_builtin_zip, __pyx_t_6, NULL); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  if (likely(PyList_CheckExact(__pyx_t_5)) || PyTuple_CheckExact(__pyx_t_5)) {
    __pyx_t_6 = __pyx_t_5; __Pyx_INCREF(__pyx_t_6); __pyx_t_3 = 0;
    __pyx_t_8 = NULL;
  } else {
    __pyx_t_3 = -1; __pyx_t_6 = PyObject_GetIter(__pyx_t_5); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_8 = Py_TYPE(__pyx_t_6)->tp_iternext; if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  for (;;) {
    if (likely(!__pyx_t_8)) {
      if (likely(PyList_CheckExact(__pyx_t_6))) {
        if (__pyx_t_3 >= PyList_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_5 = PyList_GET_ITEM(__pyx_t_6, __pyx_t_3); __Pyx_INCREF(__pyx_t_5); __pyx_t_3++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_5 = PySequence_ITEM(__pyx_t_6, __pyx_t_3); __pyx_t_3++; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_3 >= PyTuple_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_5 = PyTuple_GET_ITEM(__pyx_t_6, __pyx_t_3); __Pyx_INCREF(__pyx_t_5); __pyx_t_3++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_5 = PySequence_ITEM(__pyx_t_6, __pyx_t_3); __pyx_t_3++; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_5 = __pyx_t_8(__pyx_t_6);
      if (unlikely(!__pyx_t_5)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_5);
    }
    if ((likely(PyTuple_CheckExact(__pyx_t_5))) || (PyList_CheckExact(__pyx_t_5))) {
      PyObject* sequence = __pyx_t_5;
      #if CYTHON_COMPILING_IN_CPYTHON
      Py_ssize_t size = Py_SIZE(sequence);
      #else
      Py_ssize_t size = PySequence_Size(sequence);
      #endif
      if (unlikely(size != 2)) {
        if (size > 2) __Pyx_RaiseTooManyValuesError(2);
        else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      #if CYTHON_COMPILING_IN_CPYTHON
      if (likely(PyTuple_CheckExact(sequence))) {
        __pyx_t_7 = PyTuple_GET_ITEM(sequence, 0); 
        __pyx_t_9 = PyTuple_GET_ITEM(sequence, 1); 
      } else {
        __pyx_t_7 = PyList_GET_ITEM(sequence, 0); 
        __pyx_t_9 = PyList_GET_ITEM(sequence, 1); 
      }
      __Pyx_INCREF(__pyx_t_7);
      __Pyx_INCREF(__pyx_t_9);
      #else
      __pyx_t_7 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_7);
      __pyx_t_9 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_9);
      #endif
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    } else {
      Py_ssize_t index = -1;
      __pyx_t_10 = PyObject_GetIter(__pyx_t_5); if (unlikely(!__pyx_t_10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_10);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      __pyx_t_11 = Py_TYPE(__pyx_t_10)->tp_iternext;
      index = 0; __pyx_t_7 = __pyx_t_11(__pyx_t_10); if (unlikely(!__pyx_t_7)) goto __pyx_L20_unpacking_failed;
      __Pyx_GOTREF(__pyx_t_7);
      index = 1; __pyx_t_9 = __pyx_t_11(__pyx_t_10); if (unlikely(!__pyx_t_9)) goto __pyx_L20_unpacking_failed;
      __Pyx_GOTREF(__pyx_t_9);
      if (__Pyx_IternextUnpackEndCheck(__pyx_t_11(__pyx_t_10), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_11 = NULL;
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      goto __pyx_L21_unpacking_done;
      __pyx_L20_unpacking_failed:;
      __Pyx_DECREF(__pyx_t_10); __pyx_t_10 = 0;
      __pyx_t_11 = NULL;
      if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_L21_unpacking_done:;
    }
    __Pyx_XDECREF_SET(__pyx_v_x1, __pyx_t_7);
    __pyx_t_7 = 0;
    __Pyx_XDECREF_SET(__pyx_v_x2, __pyx_t_9);
    __pyx_t_9 = 0;
    __pyx_t_2 = (__pyx_v_x1 == __pyx_v_x2);
    __pyx_t_5 = __Pyx_PyBool_FromLong(__pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_5);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_5))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1430; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  }
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  __pyx_v_matchList4 = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1431
 *             #start of ring1 matches start of ring2
 *             matchList4 = [x1 is x2 for x1,x2 in zip(ring1[:od],ring2[od::-1])]
 *             if not False in matchList1:             # <<<<<<<<<<<<<<
 *                 newRing = ring1 +ring2[od:]
 *             elif not False in matchList2:
 */
  __pyx_t_2 = (__Pyx_PySequence_Contains(Py_False, __pyx_v_matchList1, Py_NE)); if (unlikely(__pyx_t_2 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1431; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_4 = (__pyx_t_2 != 0);
  if (__pyx_t_4) {

    /* "rmgpy/molecule/group.py":1432
 *             matchList4 = [x1 is x2 for x1,x2 in zip(ring1[:od],ring2[od::-1])]
 *             if not False in matchList1:
 *                 newRing = ring1 +ring2[od:]             # <<<<<<<<<<<<<<
 *             elif not False in matchList2:
 *                 newRing = ring1 + ring2[-od-1::-1]
 */
    __pyx_t_1 = __Pyx_PyObject_GetSlice(__pyx_v_ring2, 0, 0, &__pyx_v_od, NULL, NULL, 0, 0, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1432; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_6 = PyNumber_Add(__pyx_v_ring1, __pyx_t_1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1432; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_DECREF_SET(__pyx_v_newRing, __pyx_t_6);
    __pyx_t_6 = 0;
    goto __pyx_L22;
  }

  /* "rmgpy/molecule/group.py":1433
 *             if not False in matchList1:
 *                 newRing = ring1 +ring2[od:]
 *             elif not False in matchList2:             # <<<<<<<<<<<<<<
 *                 newRing = ring1 + ring2[-od-1::-1]
 *             elif not False in matchList3:
 */
  __pyx_t_4 = (__Pyx_PySequence_Contains(Py_False, __pyx_v_matchList2, Py_NE)); if (unlikely(__pyx_t_4 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1433; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = (__pyx_t_4 != 0);
  if (__pyx_t_2) {

    /* "rmgpy/molecule/group.py":1434
 *                 newRing = ring1 +ring2[od:]
 *             elif not False in matchList2:
 *                 newRing = ring1 + ring2[-od-1::-1]             # <<<<<<<<<<<<<<
 *             elif not False in matchList3:
 *                 newRing = ring2[:-od] + ring1
 */
    __pyx_t_6 = PyNumber_Negative(__pyx_v_od); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_1 = PyNumber_Subtract(__pyx_t_6, __pyx_int_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = PySlice_New(__pyx_t_1, Py_None, __pyx_int_neg_1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = PyObject_GetItem(__pyx_v_ring2, __pyx_t_6); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1434; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = PyNumber_Add(__pyx_v_ring1, __pyx_t_1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1434; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_DECREF_SET(__pyx_v_newRing, __pyx_t_6);
    __pyx_t_6 = 0;
    goto __pyx_L22;
  }

  /* "rmgpy/molecule/group.py":1435
 *             elif not False in matchList2:
 *                 newRing = ring1 + ring2[-od-1::-1]
 *             elif not False in matchList3:             # <<<<<<<<<<<<<<
 *                 newRing = ring2[:-od] + ring1
 *             elif not False in matchList4:
 */
  __pyx_t_2 = (__Pyx_PySequence_Contains(Py_False, __pyx_v_matchList3, Py_NE)); if (unlikely(__pyx_t_2 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1435; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_4 = (__pyx_t_2 != 0);
  if (__pyx_t_4) {

    /* "rmgpy/molecule/group.py":1436
 *                 newRing = ring1 + ring2[-od-1::-1]
 *             elif not False in matchList3:
 *                 newRing = ring2[:-od] + ring1             # <<<<<<<<<<<<<<
 *             elif not False in matchList4:
 *                 newRing = ring2[:od-1:-1] + ring1
 */
    __pyx_t_6 = PyNumber_Negative(__pyx_v_od); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1436; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_1 = __Pyx_PyObject_GetSlice(__pyx_v_ring2, 0, 0, NULL, &__pyx_t_6, NULL, 0, 0, 1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1436; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = PyNumber_Add(__pyx_t_1, __pyx_v_ring1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1436; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_DECREF_SET(__pyx_v_newRing, __pyx_t_6);
    __pyx_t_6 = 0;
    goto __pyx_L22;
  }

  /* "rmgpy/molecule/group.py":1437
 *             elif not False in matchList3:
 *                 newRing = ring2[:-od] + ring1
 *             elif not False in matchList4:             # <<<<<<<<<<<<<<
 *                 newRing = ring2[:od-1:-1] + ring1
 * 
 */
  __pyx_t_4 = (__Pyx_PySequence_Contains(Py_False, __pyx_v_matchList4, Py_NE)); if (unlikely(__pyx_t_4 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1437; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = (__pyx_t_4 != 0);
  if (__pyx_t_2) {

    /* "rmgpy/molecule/group.py":1438
 *                 newRing = ring2[:-od] + ring1
 *             elif not False in matchList4:
 *                 newRing = ring2[:od-1:-1] + ring1             # <<<<<<<<<<<<<<
 * 
 *             return newRing
 */
    __pyx_t_6 = PyNumber_Subtract(__pyx_v_od, __pyx_int_1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1438; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_1 = PySlice_New(Py_None, __pyx_t_6, __pyx_int_neg_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1438; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_6 = PyObject_GetItem(__pyx_v_ring2, __pyx_t_1); if (unlikely(__pyx_t_6 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1438; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_1 = PyNumber_Add(__pyx_t_6, __pyx_v_ring1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1438; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __Pyx_DECREF_SET(__pyx_v_newRing, __pyx_t_1);
    __pyx_t_1 = 0;
    goto __pyx_L22;
  }
  __pyx_L22:;

  /* "rmgpy/molecule/group.py":1440
 *                 newRing = ring2[:od-1:-1] + ring1
 * 
 *             return newRing             # <<<<<<<<<<<<<<
 * 
 *         def addCbAtomToRing(ring, cbAtom):
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_newRing);
  __pyx_r = __pyx_v_newRing;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1401
 *             return superSet.issuperset(subSet)
 * 
 *         def mergeOverlappingBenzeneRings(ring1, ring2, od):             # <<<<<<<<<<<<<<
 *             """
 *             The input arguements of rings are always in the order that the atoms appear
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_7);
  __Pyx_XDECREF(__pyx_t_9);
  __Pyx_XDECREF(__pyx_t_10);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.mergeOverlappingBenzeneRings", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_newRing);
  __Pyx_XDECREF(__pyx_v_matchList1);
  __Pyx_XDECREF(__pyx_v_matchList2);
  __Pyx_XDECREF(__pyx_v_matchList3);
  __Pyx_XDECREF(__pyx_v_matchList4);
  __Pyx_XDECREF(__pyx_v_x1);
  __Pyx_XDECREF(__pyx_v_x2);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1442
 *             return newRing
 * 
 *         def addCbAtomToRing(ring, cbAtom):             # <<<<<<<<<<<<<<
 *             """
 *             Every 'Cb' atom belongs in exactly one benzene ring. This function checks
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_11addCbAtomToRing(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_10addCbAtomToRing[] = "\n            Every 'Cb' atom belongs in exactly one benzene ring. This function checks\n            adds the cbAtom to the ring (in connectivity order) if the cbAtom is connected\n            to any the last or first atom in the partial ring.\n\n            Args:\n                ring: list of :class:GroupAtoms representing a partial ring to merge\n                cbAtom: :class:GroupAtom with atomtype 'Cb'\n\n            Returns: If cbAtom connects to the beginning or end of ring, returns a\n            new list of the merged ring, otherwise an empty list\n\n            ";
static PyMethodDef __pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_11addCbAtomToRing = {"addCbAtomToRing", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_11addCbAtomToRing, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_10addCbAtomToRing};
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_11addCbAtomToRing(PyObject *__pyx_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  PyObject *__pyx_v_ring = 0;
  PyObject *__pyx_v_cbAtom = 0;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addCbAtomToRing (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_ring,&__pyx_n_s_cbAtom,0};
    PyObject* values[2] = {0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        case  0: break;
        default: goto __pyx_L5_argtuple_error;
      }
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_ring)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        case  1:
        if (likely((values[1] = PyDict_GetItem(__pyx_kwds, __pyx_n_s_cbAtom)) != 0)) kw_args--;
        else {
          __Pyx_RaiseArgtupleInvalid("addCbAtomToRing", 1, 2, 2, 1); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1442; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
        }
      }
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "addCbAtomToRing") < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1442; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
      }
    } else if (PyTuple_GET_SIZE(__pyx_args) != 2) {
      goto __pyx_L5_argtuple_error;
    } else {
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
    }
    __pyx_v_ring = values[0];
    __pyx_v_cbAtom = values[1];
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("addCbAtomToRing", 1, 2, 2, PyTuple_GET_SIZE(__pyx_args)); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1442; __pyx_clineno = __LINE__; goto __pyx_L3_error;}
  __pyx_L3_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.addCbAtomToRing", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return NULL;
  __pyx_L4_argument_unpacking_done:;
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_10addCbAtomToRing(__pyx_self, __pyx_v_ring, __pyx_v_cbAtom);

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_10addCbAtomToRing(CYTHON_UNUSED PyObject *__pyx_self, PyObject *__pyx_v_ring, PyObject *__pyx_v_cbAtom) {
  PyObject *__pyx_v_mergedRing = NULL;
  PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  Py_ssize_t __pyx_t_2;
  int __pyx_t_3;
  Py_ssize_t __pyx_t_4;
  int __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  PyObject *__pyx_t_7 = NULL;
  int __pyx_t_8;
  PyObject *__pyx_t_9 = NULL;
  int __pyx_t_10;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addCbAtomToRing", 0);

  /* "rmgpy/molecule/group.py":1457
 *             """
 * 
 *             mergedRing = []             # <<<<<<<<<<<<<<
 *             #ring already complete
 *             if len(ring) == 6 : return mergedRing
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1457; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_mergedRing = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1459
 *             mergedRing = []
 *             #ring already complete
 *             if len(ring) == 6 : return mergedRing             # <<<<<<<<<<<<<<
 *             for atom2, bond12 in cbAtom.bonds.iteritems():
 *                 if bond12.isBenzene():
 */
  __pyx_t_2 = PyObject_Length(__pyx_v_ring); if (unlikely(__pyx_t_2 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1459; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_3 = ((__pyx_t_2 == 6) != 0);
  if (__pyx_t_3) {
    __Pyx_XDECREF(__pyx_r);
    __Pyx_INCREF(__pyx_v_mergedRing);
    __pyx_r = __pyx_v_mergedRing;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":1460
 *             #ring already complete
 *             if len(ring) == 6 : return mergedRing
 *             for atom2, bond12 in cbAtom.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                 if bond12.isBenzene():
 *                     if atom2 is ring[-1]: mergedRing = ring+[cbAtom]
 */
  __pyx_t_2 = 0;
  __pyx_t_6 = __Pyx_PyObject_GetAttrStr(__pyx_v_cbAtom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1460; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  if (unlikely(__pyx_t_6 == Py_None)) {
    PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1460; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_7 = __Pyx_dict_iterator(__pyx_t_6, 0, __pyx_n_s_iteritems, (&__pyx_t_4), (&__pyx_t_5)); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1460; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_7);
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  __Pyx_XDECREF(__pyx_t_1);
  __pyx_t_1 = __pyx_t_7;
  __pyx_t_7 = 0;
  while (1) {
    __pyx_t_8 = __Pyx_dict_iter_next(__pyx_t_1, __pyx_t_4, &__pyx_t_2, &__pyx_t_7, &__pyx_t_6, NULL, __pyx_t_5);
    if (unlikely(__pyx_t_8 == 0)) break;
    if (unlikely(__pyx_t_8 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1460; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_7);
    __pyx_t_7 = 0;
    __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_6);
    __pyx_t_6 = 0;

    /* "rmgpy/molecule/group.py":1461
 *             if len(ring) == 6 : return mergedRing
 *             for atom2, bond12 in cbAtom.bonds.iteritems():
 *                 if bond12.isBenzene():             # <<<<<<<<<<<<<<
 *                     if atom2 is ring[-1]: mergedRing = ring+[cbAtom]
 *                     elif atom2 is ring[0]: mergedRing = [cbAtom] +ring
 */
    __pyx_t_7 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1461; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    __pyx_t_9 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_7))) {
      __pyx_t_9 = PyMethod_GET_SELF(__pyx_t_7);
      if (likely(__pyx_t_9)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_7);
        __Pyx_INCREF(__pyx_t_9);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_7, function);
      }
    }
    if (__pyx_t_9) {
      __pyx_t_6 = __Pyx_PyObject_CallOneArg(__pyx_t_7, __pyx_t_9); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1461; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_9); __pyx_t_9 = 0;
    } else {
      __pyx_t_6 = __Pyx_PyObject_CallNoArg(__pyx_t_7); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1461; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_GOTREF(__pyx_t_6);
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
    __pyx_t_3 = __Pyx_PyObject_IsTrue(__pyx_t_6); if (unlikely(__pyx_t_3 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1461; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    if (__pyx_t_3) {

      /* "rmgpy/molecule/group.py":1462
 *             for atom2, bond12 in cbAtom.bonds.iteritems():
 *                 if bond12.isBenzene():
 *                     if atom2 is ring[-1]: mergedRing = ring+[cbAtom]             # <<<<<<<<<<<<<<
 *                     elif atom2 is ring[0]: mergedRing = [cbAtom] +ring
 * 
 */
      __pyx_t_6 = __Pyx_GetItemInt(__pyx_v_ring, -1, long, 1, __Pyx_PyInt_From_long, 0, 1, 1); if (unlikely(__pyx_t_6 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1462; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_6);
      __pyx_t_3 = (__pyx_v_atom2 == __pyx_t_6);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      __pyx_t_10 = (__pyx_t_3 != 0);
      if (__pyx_t_10) {
        __pyx_t_6 = PyList_New(1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1462; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        __Pyx_INCREF(__pyx_v_cbAtom);
        PyList_SET_ITEM(__pyx_t_6, 0, __pyx_v_cbAtom);
        __Pyx_GIVEREF(__pyx_v_cbAtom);
        __pyx_t_7 = PyNumber_Add(__pyx_v_ring, __pyx_t_6); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1462; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_7);
        __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
        __Pyx_DECREF_SET(__pyx_v_mergedRing, __pyx_t_7);
        __pyx_t_7 = 0;
        goto __pyx_L7;
      }

      /* "rmgpy/molecule/group.py":1463
 *                 if bond12.isBenzene():
 *                     if atom2 is ring[-1]: mergedRing = ring+[cbAtom]
 *                     elif atom2 is ring[0]: mergedRing = [cbAtom] +ring             # <<<<<<<<<<<<<<
 * 
 *             return mergedRing
 */
      __pyx_t_7 = __Pyx_GetItemInt(__pyx_v_ring, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_7 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1463; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_7);
      __pyx_t_10 = (__pyx_v_atom2 == __pyx_t_7);
      __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
      __pyx_t_3 = (__pyx_t_10 != 0);
      if (__pyx_t_3) {
        __pyx_t_7 = PyList_New(1); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1463; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_7);
        __Pyx_INCREF(__pyx_v_cbAtom);
        PyList_SET_ITEM(__pyx_t_7, 0, __pyx_v_cbAtom);
        __Pyx_GIVEREF(__pyx_v_cbAtom);
        __pyx_t_6 = PyNumber_Add(__pyx_t_7, __pyx_v_ring); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1463; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_6);
        __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
        __Pyx_DECREF_SET(__pyx_v_mergedRing, __pyx_t_6);
        __pyx_t_6 = 0;
        goto __pyx_L7;
      }
      __pyx_L7:;
      goto __pyx_L6;
    }
    __pyx_L6:;
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1465
 *                     elif atom2 is ring[0]: mergedRing = [cbAtom] +ring
 * 
 *             return mergedRing             # <<<<<<<<<<<<<<
 * 
 *         #######################################################################################
 */
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_mergedRing);
  __pyx_r = __pyx_v_mergedRing;
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1442
 *             return newRing
 * 
 *         def addCbAtomToRing(ring, cbAtom):             # <<<<<<<<<<<<<<
 *             """
 *             Every 'Cb' atom belongs in exactly one benzene ring. This function checks
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_7);
  __Pyx_XDECREF(__pyx_t_9);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene.addCbAtomToRing", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_mergedRing);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1269
 *         return copyGroup
 * 
 *     def addImplicitBenzene(self):             # <<<<<<<<<<<<<<
 *         """
 *         Returns: A modified group with any implicit benzene rings added
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_69addImplicitBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_f_5rmgpy_8molecule_5group_5Group_addImplicitBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *classifyBenzeneCarbons = 0;
  PyObject *addBenzeneAtomToGroup = 0;
  PyObject *sortByConnectivity = 0;
  PyObject *checkSet = 0;
  PyObject *mergeOverlappingBenzeneRings = 0;
  PyObject *addCbAtomToRing = 0;
  PyObject *__pyx_v_copyGroup = NULL;
  PyObject *__pyx_v_cbAtomList = NULL;
  PyObject *__pyx_v_cbfAtomList = NULL;
  PyObject *__pyx_v_cbfAtomList1 = NULL;
  PyObject *__pyx_v_cbfAtomList2 = NULL;
  PyObject *__pyx_v_connectedCbfs = NULL;
  PyObject *__pyx_v_partners = NULL;
  PyObject *__pyx_v_cbfAtom = NULL;
  PyObject *__pyx_v_potentialPartner = NULL;
  PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_v_bondsAreBenzene = NULL;
  PyObject *__pyx_v_newAtom = NULL;
  PyObject *__pyx_v_rings = NULL;
  PyObject *__pyx_v_otherCbf = NULL;
  PyObject *__pyx_v_newRingSeeds = NULL;
  PyObject *__pyx_v_allLigands = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_index1 = NULL;
  PyObject *__pyx_v_ring1 = NULL;
  PyObject *__pyx_v_mergeRingDict = NULL;
  PyObject *__pyx_v_index = NULL;
  PyObject *__pyx_v_ring2 = NULL;
  PyObject *__pyx_v_mergeRing = NULL;
  PyObject *__pyx_v_newRingSeed = NULL;
  PyObject *__pyx_v_inRing = NULL;
  PyObject *__pyx_v_ring = NULL;
  PyObject *__pyx_v_x = NULL;
  PyObject *__pyx_v_cbAtom = NULL;
  PyObject *__pyx_v_mergedRingDict = NULL;
  PyObject *__pyx_v_carbonsToGrow = NULL;
  PyObject *__pyx_v_lastAtom = NULL;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *__pyx_v_newBond = NULL;
  PyObject *__pyx_v_bond23 = NULL;
  PyObject *__pyx_v_cycle = NULL;
  PyObject *__defaults__ = 0;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  PyObject *__pyx_t_7 = NULL;
  PyObject *(*__pyx_t_8)(PyObject *);
  Py_ssize_t __pyx_t_9;
  int __pyx_t_10;
  PyObject *(*__pyx_t_11)(PyObject *);
  int __pyx_t_12;
  int __pyx_t_13;
  Py_ssize_t __pyx_t_14;
  Py_ssize_t __pyx_t_15;
  int __pyx_t_16;
  int __pyx_t_17;
  Py_ssize_t __pyx_t_18;
  PyObject *(*__pyx_t_19)(PyObject *);
  PyObject *__pyx_t_20 = NULL;
  int __pyx_t_21;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addImplicitBenzene", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_addImplicitBenzene); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1269; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_69addImplicitBenzene)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1269; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1269; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1269; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1283
 *         """
 * 
 *         def classifyBenzeneCarbons(group, partners = []):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */
  __pyx_t_1 = __Pyx_CyFunction_NewEx(&__pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_1classifyBenzeneCarbons, 0, __pyx_n_s_Group_addImplicitBenzene_locals, NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_d, ((PyObject *)__pyx_codeobj__5)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (!__Pyx_CyFunction_InitDefaults(__pyx_t_1, sizeof(__pyx_defaults), 1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_CyFunction_Defaults(__pyx_defaults, __pyx_t_1)->__pyx_arg_partners = __pyx_t_2;
  __Pyx_GIVEREF(__pyx_t_2);
  __pyx_t_2 = 0;
  __Pyx_CyFunction_SetDefaultsGetter(__pyx_t_1, __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_12__defaults__);
  classifyBenzeneCarbons = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1341
 *             return (cbAtomList, cbfAtomList, cbfAtomList1, cbfAtomList2, connectedCbfs)
 * 
 *         def addBenzeneAtomToGroup(group, connectingAtom, cbf=False):             # <<<<<<<<<<<<<<
 *             """
 *             This function adds a benzene atom to the group
 */
  __pyx_t_1 = __Pyx_CyFunction_NewEx(&__pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_3addBenzeneAtomToGroup, 0, __pyx_n_s_Group_addImplicitBenzene_locals_2, NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_d, ((PyObject *)__pyx_codeobj__7)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1341; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_CyFunction_SetDefaultsTuple(__pyx_t_1, __pyx_tuple__8);
  addBenzeneAtomToGroup = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1363
 *             return (group, newAtom)
 * 
 *         def sortByConnectivity(atomList):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */
  __pyx_t_1 = __Pyx_CyFunction_NewEx(&__pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_5sortByConnectivity, 0, __pyx_n_s_Group_addImplicitBenzene_locals_3, NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_d, ((PyObject *)__pyx_codeobj__10)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1363; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  sortByConnectivity = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1388
 *             return sortedAtomList
 * 
 *         def checkSet(superList, subList):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */
  __pyx_t_1 = __Pyx_CyFunction_NewEx(&__pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_7checkSet, 0, __pyx_n_s_Group_addImplicitBenzene_locals_4, NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_d, ((PyObject *)__pyx_codeobj__12)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1388; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  checkSet = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1401
 *             return superSet.issuperset(subSet)
 * 
 *         def mergeOverlappingBenzeneRings(ring1, ring2, od):             # <<<<<<<<<<<<<<
 *             """
 *             The input arguements of rings are always in the order that the atoms appear
 */
  __pyx_t_1 = __Pyx_CyFunction_NewEx(&__pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_9mergeOverlappingBenzeneRings, 0, __pyx_n_s_Group_addImplicitBenzene_locals_5, NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_d, ((PyObject *)__pyx_codeobj__14)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1401; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  mergeOverlappingBenzeneRings = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1442
 *             return newRing
 * 
 *         def addCbAtomToRing(ring, cbAtom):             # <<<<<<<<<<<<<<
 *             """
 *             Every 'Cb' atom belongs in exactly one benzene ring. This function checks
 */
  __pyx_t_1 = __Pyx_CyFunction_NewEx(&__pyx_mdef_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_11addCbAtomToRing, 0, __pyx_n_s_Group_addImplicitBenzene_locals_6, NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_d, ((PyObject *)__pyx_codeobj__16)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1442; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  addCbAtomToRing = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1469
 *         #######################################################################################
 *         #start of main algorithm
 *         copyGroup = deepcopy(self)             # <<<<<<<<<<<<<<
 *         """
 *         Step 1. Classify all atoms as Cb, Cbf1, Cbf2, Cbf3, ignoring all non-benzene carbons
 */
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_deepcopy); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1469; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (!__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, ((PyObject *)__pyx_v_self)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1469; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
  } else {
    __pyx_t_4 = PyTuple_New(1+1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1469; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3); __Pyx_GIVEREF(__pyx_t_3); __pyx_t_3 = NULL;
    __Pyx_INCREF(((PyObject *)__pyx_v_self));
    PyTuple_SET_ITEM(__pyx_t_4, 0+1, ((PyObject *)__pyx_v_self));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1469; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_copyGroup = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1484
 *         Currently we only allow 3 Cbf atoms, so Cbf3 is not possible.
 *         """
 *         (cbAtomList, cbfAtomList, cbfAtomList1, cbfAtomList2, connectedCbfs) = classifyBenzeneCarbons(copyGroup)             # <<<<<<<<<<<<<<
 * 
 *         #check that there are less than three Cbf atoms
 */
  __pyx_t_1 = PyTuple_New(1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_INCREF(__pyx_v_copyGroup);
  PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_v_copyGroup);
  __Pyx_GIVEREF(__pyx_v_copyGroup);
  __pyx_t_2 = __Pyx_PyObject_Call(classifyBenzeneCarbons, __pyx_t_1, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  if ((likely(PyTuple_CheckExact(__pyx_t_2))) || (PyList_CheckExact(__pyx_t_2))) {
    PyObject* sequence = __pyx_t_2;
    #if CYTHON_COMPILING_IN_CPYTHON
    Py_ssize_t size = Py_SIZE(sequence);
    #else
    Py_ssize_t size = PySequence_Size(sequence);
    #endif
    if (unlikely(size != 5)) {
      if (size > 5) __Pyx_RaiseTooManyValuesError(5);
      else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    #if CYTHON_COMPILING_IN_CPYTHON
    if (likely(PyTuple_CheckExact(sequence))) {
      __pyx_t_1 = PyTuple_GET_ITEM(sequence, 0); 
      __pyx_t_4 = PyTuple_GET_ITEM(sequence, 1); 
      __pyx_t_3 = PyTuple_GET_ITEM(sequence, 2); 
      __pyx_t_5 = PyTuple_GET_ITEM(sequence, 3); 
      __pyx_t_6 = PyTuple_GET_ITEM(sequence, 4); 
    } else {
      __pyx_t_1 = PyList_GET_ITEM(sequence, 0); 
      __pyx_t_4 = PyList_GET_ITEM(sequence, 1); 
      __pyx_t_3 = PyList_GET_ITEM(sequence, 2); 
      __pyx_t_5 = PyList_GET_ITEM(sequence, 3); 
      __pyx_t_6 = PyList_GET_ITEM(sequence, 4); 
    }
    __Pyx_INCREF(__pyx_t_1);
    __Pyx_INCREF(__pyx_t_4);
    __Pyx_INCREF(__pyx_t_3);
    __Pyx_INCREF(__pyx_t_5);
    __Pyx_INCREF(__pyx_t_6);
    #else
    {
      Py_ssize_t i;
      PyObject** temps[5] = {&__pyx_t_1,&__pyx_t_4,&__pyx_t_3,&__pyx_t_5,&__pyx_t_6};
      for (i=0; i < 5; i++) {
        PyObject* item = PySequence_ITEM(sequence, i); if (unlikely(!item)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(item);
        *(temps[i]) = item;
      }
    }
    #endif
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  } else {
    Py_ssize_t index = -1;
    PyObject** temps[5] = {&__pyx_t_1,&__pyx_t_4,&__pyx_t_3,&__pyx_t_5,&__pyx_t_6};
    __pyx_t_7 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_8 = Py_TYPE(__pyx_t_7)->tp_iternext;
    for (index=0; index < 5; index++) {
      PyObject* item = __pyx_t_8(__pyx_t_7); if (unlikely(!item)) goto __pyx_L3_unpacking_failed;
      __Pyx_GOTREF(item);
      *(temps[index]) = item;
    }
    if (__Pyx_IternextUnpackEndCheck(__pyx_t_8(__pyx_t_7), 5) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_8 = NULL;
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
    goto __pyx_L4_unpacking_done;
    __pyx_L3_unpacking_failed:;
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
    __pyx_t_8 = NULL;
    if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1484; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_L4_unpacking_done:;
  }
  __pyx_v_cbAtomList = __pyx_t_1;
  __pyx_t_1 = 0;
  __pyx_v_cbfAtomList = __pyx_t_4;
  __pyx_t_4 = 0;
  __pyx_v_cbfAtomList1 = __pyx_t_3;
  __pyx_t_3 = 0;
  __pyx_v_cbfAtomList2 = __pyx_t_5;
  __pyx_t_5 = 0;
  __pyx_v_connectedCbfs = __pyx_t_6;
  __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1487
 * 
 *         #check that there are less than three Cbf atoms
 *         if len(cbfAtomList) > 3:             # <<<<<<<<<<<<<<
 *             if not self.isBenzeneExplicit():
 *                 raise ImplicitBenzeneError("{0} has more than three Cbf atoms and does not have fully explicit benzene rings.")
 */
  __pyx_t_9 = PyObject_Length(__pyx_v_cbfAtomList); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1487; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_10 = ((__pyx_t_9 > 3) != 0);
  if (__pyx_t_10) {

    /* "rmgpy/molecule/group.py":1488
 *         #check that there are less than three Cbf atoms
 *         if len(cbfAtomList) > 3:
 *             if not self.isBenzeneExplicit():             # <<<<<<<<<<<<<<
 *                 raise ImplicitBenzeneError("{0} has more than three Cbf atoms and does not have fully explicit benzene rings.")
 *             #already fully explicit if it passes isBenzeneExplict()
 */
    __pyx_t_10 = ((!(((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->isBenzeneExplicit(__pyx_v_self, 0) != 0)) != 0);
    if (__pyx_t_10) {

      /* "rmgpy/molecule/group.py":1489
 *         if len(cbfAtomList) > 3:
 *             if not self.isBenzeneExplicit():
 *                 raise ImplicitBenzeneError("{0} has more than three Cbf atoms and does not have fully explicit benzene rings.")             # <<<<<<<<<<<<<<
 *             #already fully explicit if it passes isBenzeneExplict()
 *             else: return copyGroup
 */
      __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_ImplicitBenzeneError); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1489; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_6 = __Pyx_PyObject_Call(__pyx_t_2, __pyx_tuple__17, NULL); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1489; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_Raise(__pyx_t_6, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1489; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":1491
 *                 raise ImplicitBenzeneError("{0} has more than three Cbf atoms and does not have fully explicit benzene rings.")
 *             #already fully explicit if it passes isBenzeneExplict()
 *             else: return copyGroup             # <<<<<<<<<<<<<<
 *         """
 *         #Step 2. Partner up each Cbf1 and Cbf2 atom
 */
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      if (!(likely(((__pyx_v_copyGroup) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_copyGroup, __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1491; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_INCREF(__pyx_v_copyGroup);
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_copyGroup);
      goto __pyx_L0;
    }
  }

  /* "rmgpy/molecule/group.py":1505
 *         to partner up Cbf2 and Cbf3 atoms.
 *         """
 *         partners = {} #dictionary of exclusive partners, has 1:2 and 2:1             # <<<<<<<<<<<<<<
 *         for cbfAtom in cbfAtomList1:
 *             if cbfAtom in partners: continue
 */
  __pyx_t_6 = PyDict_New(); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1505; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __pyx_v_partners = ((PyObject*)__pyx_t_6);
  __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1506
 *         """
 *         partners = {} #dictionary of exclusive partners, has 1:2 and 2:1
 *         for cbfAtom in cbfAtomList1:             # <<<<<<<<<<<<<<
 *             if cbfAtom in partners: continue
 *             #if cbfAtom has a connected cbf it must be the partner
 */
  if (likely(PyList_CheckExact(__pyx_v_cbfAtomList1)) || PyTuple_CheckExact(__pyx_v_cbfAtomList1)) {
    __pyx_t_6 = __pyx_v_cbfAtomList1; __Pyx_INCREF(__pyx_t_6); __pyx_t_9 = 0;
    __pyx_t_11 = NULL;
  } else {
    __pyx_t_9 = -1; __pyx_t_6 = PyObject_GetIter(__pyx_v_cbfAtomList1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1506; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_11 = Py_TYPE(__pyx_t_6)->tp_iternext; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1506; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  for (;;) {
    if (likely(!__pyx_t_11)) {
      if (likely(PyList_CheckExact(__pyx_t_6))) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_6, __pyx_t_9); __Pyx_INCREF(__pyx_t_2); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1506; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_6, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1506; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_9 >= PyTuple_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_6, __pyx_t_9); __Pyx_INCREF(__pyx_t_2); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1506; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_6, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1506; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_11(__pyx_t_6);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1506; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_cbfAtom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1507
 *         partners = {} #dictionary of exclusive partners, has 1:2 and 2:1
 *         for cbfAtom in cbfAtomList1:
 *             if cbfAtom in partners: continue             # <<<<<<<<<<<<<<
 *             #if cbfAtom has a connected cbf it must be the partner
 *             elif connectedCbfs[cbfAtom] and connectedCbfs[cbfAtom][0] not in partners:
 */
    __pyx_t_10 = (__Pyx_PyDict_Contains(__pyx_v_cbfAtom, __pyx_v_partners, Py_EQ)); if (unlikely(__pyx_t_10 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1507; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_12 = (__pyx_t_10 != 0);
    if (__pyx_t_12) {
      goto __pyx_L7_continue;
    }

    /* "rmgpy/molecule/group.py":1509
 *             if cbfAtom in partners: continue
 *             #if cbfAtom has a connected cbf it must be the partner
 *             elif connectedCbfs[cbfAtom] and connectedCbfs[cbfAtom][0] not in partners:             # <<<<<<<<<<<<<<
 *                 partners[cbfAtom] = connectedCbfs[cbfAtom][0]
 *                 partners[connectedCbfs[cbfAtom][0]] = cbfAtom
 */
    __pyx_t_2 = PyObject_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1509; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_10 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_10 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1509; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (__pyx_t_10) {
      goto __pyx_L11_next_and;
    } else {
      __pyx_t_12 = __pyx_t_10;
      goto __pyx_L10_bool_binop_done;
    }
    __pyx_L11_next_and:;
    __pyx_t_2 = PyObject_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1509; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_5 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_5 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1509; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_5);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_10 = (__Pyx_PyDict_Contains(__pyx_t_5, __pyx_v_partners, Py_NE)); if (unlikely(__pyx_t_10 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1509; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
    __pyx_t_13 = (__pyx_t_10 != 0);
    __pyx_t_12 = __pyx_t_13;
    __pyx_L10_bool_binop_done:;
    if (__pyx_t_12) {

      /* "rmgpy/molecule/group.py":1510
 *             #if cbfAtom has a connected cbf it must be the partner
 *             elif connectedCbfs[cbfAtom] and connectedCbfs[cbfAtom][0] not in partners:
 *                 partners[cbfAtom] = connectedCbfs[cbfAtom][0]             # <<<<<<<<<<<<<<
 *                 partners[connectedCbfs[cbfAtom][0]] = cbfAtom
 *             else:
 */
      __pyx_t_5 = PyObject_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_5 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1510; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_5);
      __pyx_t_2 = __Pyx_GetItemInt(__pyx_t_5, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1510; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      if (unlikely(PyDict_SetItem(__pyx_v_partners, __pyx_v_cbfAtom, __pyx_t_2) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1510; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

      /* "rmgpy/molecule/group.py":1511
 *             elif connectedCbfs[cbfAtom] and connectedCbfs[cbfAtom][0] not in partners:
 *                 partners[cbfAtom] = connectedCbfs[cbfAtom][0]
 *                 partners[connectedCbfs[cbfAtom][0]] = cbfAtom             # <<<<<<<<<<<<<<
 *             else:
 *                 #search for a potential partner out of atoms benzene bonded atoms
 */
      __pyx_t_2 = PyObject_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1511; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_5 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_5 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1511; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_5);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (unlikely(PyDict_SetItem(__pyx_v_partners, __pyx_t_5, __pyx_v_cbfAtom) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1511; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      goto __pyx_L9;
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":1514
 *             else:
 *                 #search for a potential partner out of atoms benzene bonded atoms
 *                 potentialPartner = None             # <<<<<<<<<<<<<<
 *                 for atom2, bond12 in cbfAtom.bonds.iteritems():
 *                     if atom2 in partners: continue
 */
      __Pyx_INCREF(Py_None);
      __Pyx_XDECREF_SET(__pyx_v_potentialPartner, Py_None);

      /* "rmgpy/molecule/group.py":1515
 *                 #search for a potential partner out of atoms benzene bonded atoms
 *                 potentialPartner = None
 *                 for atom2, bond12 in cbfAtom.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                     if atom2 in partners: continue
 *                     #Potential partner must not have any bonds except benzene bonds
 */
      __pyx_t_14 = 0;
      __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_cbfAtom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      if (unlikely(__pyx_t_2 == Py_None)) {
        PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_3 = __Pyx_dict_iterator(__pyx_t_2, 0, __pyx_n_s_iteritems, (&__pyx_t_15), (&__pyx_t_16)); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_XDECREF(__pyx_t_5);
      __pyx_t_5 = __pyx_t_3;
      __pyx_t_3 = 0;
      while (1) {
        __pyx_t_17 = __Pyx_dict_iter_next(__pyx_t_5, __pyx_t_15, &__pyx_t_14, &__pyx_t_3, &__pyx_t_2, NULL, __pyx_t_16);
        if (unlikely(__pyx_t_17 == 0)) break;
        if (unlikely(__pyx_t_17 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1515; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_3);
        __pyx_t_3 = 0;
        __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_2);
        __pyx_t_2 = 0;

        /* "rmgpy/molecule/group.py":1516
 *                 potentialPartner = None
 *                 for atom2, bond12 in cbfAtom.bonds.iteritems():
 *                     if atom2 in partners: continue             # <<<<<<<<<<<<<<
 *                     #Potential partner must not have any bonds except benzene bonds
 *                     elif bond12.isBenzene():
 */
        __pyx_t_12 = (__Pyx_PyDict_Contains(__pyx_v_atom2, __pyx_v_partners, Py_EQ)); if (unlikely(__pyx_t_12 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1516; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __pyx_t_13 = (__pyx_t_12 != 0);
        if (__pyx_t_13) {
          goto __pyx_L12_continue;
        }

        /* "rmgpy/molecule/group.py":1518
 *                     if atom2 in partners: continue
 *                     #Potential partner must not have any bonds except benzene bonds
 *                     elif bond12.isBenzene():             # <<<<<<<<<<<<<<
 *                         bondsAreBenzene = [True if bond23.isBenzene() else False for bond23 in atom2.bonds.values()]
 *                         if all(bondsAreBenzene):
 */
        __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1518; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_4 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
          __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
          if (likely(__pyx_t_4)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
            __Pyx_INCREF(__pyx_t_4);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_3, function);
          }
        }
        if (__pyx_t_4) {
          __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1518; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        } else {
          __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1518; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1518; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        if (__pyx_t_13) {

          /* "rmgpy/molecule/group.py":1519
 *                     #Potential partner must not have any bonds except benzene bonds
 *                     elif bond12.isBenzene():
 *                         bondsAreBenzene = [True if bond23.isBenzene() else False for bond23 in atom2.bonds.values()]             # <<<<<<<<<<<<<<
 *                         if all(bondsAreBenzene):
 *                             potentialPartner= atom2
 */
          __pyx_t_2 = PyList_New(0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom2, __pyx_n_s_bonds); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_4);
          __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_4, __pyx_n_s_values); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_1);
          __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
          __pyx_t_4 = NULL;
          if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_1))) {
            __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_1);
            if (likely(__pyx_t_4)) {
              PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_1);
              __Pyx_INCREF(__pyx_t_4);
              __Pyx_INCREF(function);
              __Pyx_DECREF_SET(__pyx_t_1, function);
            }
          }
          if (__pyx_t_4) {
            __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_1, __pyx_t_4); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
          } else {
            __pyx_t_3 = __Pyx_PyObject_CallNoArg(__pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          if (likely(PyList_CheckExact(__pyx_t_3)) || PyTuple_CheckExact(__pyx_t_3)) {
            __pyx_t_1 = __pyx_t_3; __Pyx_INCREF(__pyx_t_1); __pyx_t_18 = 0;
            __pyx_t_19 = NULL;
          } else {
            __pyx_t_18 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_1);
            __pyx_t_19 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_19)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          for (;;) {
            if (likely(!__pyx_t_19)) {
              if (likely(PyList_CheckExact(__pyx_t_1))) {
                if (__pyx_t_18 >= PyList_GET_SIZE(__pyx_t_1)) break;
                #if CYTHON_COMPILING_IN_CPYTHON
                __pyx_t_3 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_18); __Pyx_INCREF(__pyx_t_3); __pyx_t_18++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
                #else
                __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_18); __pyx_t_18++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
                #endif
              } else {
                if (__pyx_t_18 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
                #if CYTHON_COMPILING_IN_CPYTHON
                __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_18); __Pyx_INCREF(__pyx_t_3); __pyx_t_18++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
                #else
                __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_18); __pyx_t_18++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
                #endif
              }
            } else {
              __pyx_t_3 = __pyx_t_19(__pyx_t_1);
              if (unlikely(!__pyx_t_3)) {
                PyObject* exc_type = PyErr_Occurred();
                if (exc_type) {
                  if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
                  else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
                }
                break;
              }
              __Pyx_GOTREF(__pyx_t_3);
            }
            __Pyx_XDECREF_SET(__pyx_v_bond23, __pyx_t_3);
            __pyx_t_3 = 0;
            __pyx_t_7 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond23, __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_GOTREF(__pyx_t_7);
            __pyx_t_20 = NULL;
            if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_7))) {
              __pyx_t_20 = PyMethod_GET_SELF(__pyx_t_7);
              if (likely(__pyx_t_20)) {
                PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_7);
                __Pyx_INCREF(__pyx_t_20);
                __Pyx_INCREF(function);
                __Pyx_DECREF_SET(__pyx_t_7, function);
              }
            }
            if (__pyx_t_20) {
              __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_7, __pyx_t_20); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
              __Pyx_DECREF(__pyx_t_20); __pyx_t_20 = 0;
            } else {
              __pyx_t_4 = __Pyx_PyObject_CallNoArg(__pyx_t_7); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            __Pyx_GOTREF(__pyx_t_4);
            __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
            __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
            if (__pyx_t_13) {
              __Pyx_INCREF(Py_True);
              __pyx_t_3 = Py_True;
            } else {
              __Pyx_INCREF(Py_False);
              __pyx_t_3 = Py_False;
            }
            if (unlikely(__Pyx_ListComp_Append(__pyx_t_2, (PyObject*)__pyx_t_3))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1519; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          }
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          __Pyx_XDECREF_SET(__pyx_v_bondsAreBenzene, ((PyObject*)__pyx_t_2));
          __pyx_t_2 = 0;

          /* "rmgpy/molecule/group.py":1520
 *                     elif bond12.isBenzene():
 *                         bondsAreBenzene = [True if bond23.isBenzene() else False for bond23 in atom2.bonds.values()]
 *                         if all(bondsAreBenzene):             # <<<<<<<<<<<<<<
 *                             potentialPartner= atom2
 *                 #Make a Cb atom the partner, now marking it as a Cbfatom
 */
          __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1520; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __Pyx_INCREF(__pyx_v_bondsAreBenzene);
          PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_bondsAreBenzene);
          __Pyx_GIVEREF(__pyx_v_bondsAreBenzene);
          __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_all, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1520; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_1);
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
          __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1520; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
          if (__pyx_t_13) {

            /* "rmgpy/molecule/group.py":1521
 *                         bondsAreBenzene = [True if bond23.isBenzene() else False for bond23 in atom2.bonds.values()]
 *                         if all(bondsAreBenzene):
 *                             potentialPartner= atom2             # <<<<<<<<<<<<<<
 *                 #Make a Cb atom the partner, now marking it as a Cbfatom
 *                 if potentialPartner:
 */
            __Pyx_INCREF(__pyx_v_atom2);
            __Pyx_DECREF_SET(__pyx_v_potentialPartner, __pyx_v_atom2);
            goto __pyx_L17;
          }
          __pyx_L17:;
          goto __pyx_L14;
        }
        __pyx_L14:;
        __pyx_L12_continue:;
      }
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;

      /* "rmgpy/molecule/group.py":1523
 *                             potentialPartner= atom2
 *                 #Make a Cb atom the partner, now marking it as a Cbfatom
 *                 if potentialPartner:             # <<<<<<<<<<<<<<
 *                     partners[cbfAtom] = potentialPartner
 *                     partners[potentialPartner]= cbfAtom
 */
      __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_v_potentialPartner); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1523; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_13) {

        /* "rmgpy/molecule/group.py":1524
 *                 #Make a Cb atom the partner, now marking it as a Cbfatom
 *                 if potentialPartner:
 *                     partners[cbfAtom] = potentialPartner             # <<<<<<<<<<<<<<
 *                     partners[potentialPartner]= cbfAtom
 *                 #otherwise create a new atom to be the partner
 */
        if (unlikely(PyDict_SetItem(__pyx_v_partners, __pyx_v_cbfAtom, __pyx_v_potentialPartner) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1524; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

        /* "rmgpy/molecule/group.py":1525
 *                 if potentialPartner:
 *                     partners[cbfAtom] = potentialPartner
 *                     partners[potentialPartner]= cbfAtom             # <<<<<<<<<<<<<<
 *                 #otherwise create a new atom to be the partner
 *                 else:
 */
        if (unlikely(PyDict_SetItem(__pyx_v_partners, __pyx_v_potentialPartner, __pyx_v_cbfAtom) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1525; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L18;
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":1528
 *                 #otherwise create a new atom to be the partner
 *                 else:
 *                     (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, cbfAtom, True)             # <<<<<<<<<<<<<<
 *                     partners[cbfAtom] = newAtom
 *                     partners[newAtom] = cbfAtom
 */
        __pyx_t_5 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_2addBenzeneAtomToGroup(addBenzeneAtomToGroup, __pyx_v_copyGroup, __pyx_v_cbfAtom, Py_True); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1528; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        if ((likely(PyTuple_CheckExact(__pyx_t_5))) || (PyList_CheckExact(__pyx_t_5))) {
          PyObject* sequence = __pyx_t_5;
          #if CYTHON_COMPILING_IN_CPYTHON
          Py_ssize_t size = Py_SIZE(sequence);
          #else
          Py_ssize_t size = PySequence_Size(sequence);
          #endif
          if (unlikely(size != 2)) {
            if (size > 2) __Pyx_RaiseTooManyValuesError(2);
            else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
            {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1528; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          #if CYTHON_COMPILING_IN_CPYTHON
          if (likely(PyTuple_CheckExact(sequence))) {
            __pyx_t_1 = PyTuple_GET_ITEM(sequence, 0); 
            __pyx_t_2 = PyTuple_GET_ITEM(sequence, 1); 
          } else {
            __pyx_t_1 = PyList_GET_ITEM(sequence, 0); 
            __pyx_t_2 = PyList_GET_ITEM(sequence, 1); 
          }
          __Pyx_INCREF(__pyx_t_1);
          __Pyx_INCREF(__pyx_t_2);
          #else
          __pyx_t_1 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1528; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_1);
          __pyx_t_2 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1528; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          #endif
          __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        } else {
          Py_ssize_t index = -1;
          __pyx_t_3 = PyObject_GetIter(__pyx_t_5); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1528; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
          __pyx_t_8 = Py_TYPE(__pyx_t_3)->tp_iternext;
          index = 0; __pyx_t_1 = __pyx_t_8(__pyx_t_3); if (unlikely(!__pyx_t_1)) goto __pyx_L19_unpacking_failed;
          __Pyx_GOTREF(__pyx_t_1);
          index = 1; __pyx_t_2 = __pyx_t_8(__pyx_t_3); if (unlikely(!__pyx_t_2)) goto __pyx_L19_unpacking_failed;
          __Pyx_GOTREF(__pyx_t_2);
          if (__Pyx_IternextUnpackEndCheck(__pyx_t_8(__pyx_t_3), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1528; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __pyx_t_8 = NULL;
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          goto __pyx_L20_unpacking_done;
          __pyx_L19_unpacking_failed:;
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
          __pyx_t_8 = NULL;
          if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1528; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __pyx_L20_unpacking_done:;
        }
        __Pyx_DECREF_SET(__pyx_v_copyGroup, __pyx_t_1);
        __pyx_t_1 = 0;
        __Pyx_XDECREF_SET(__pyx_v_newAtom, __pyx_t_2);
        __pyx_t_2 = 0;

        /* "rmgpy/molecule/group.py":1529
 *                 else:
 *                     (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, cbfAtom, True)
 *                     partners[cbfAtom] = newAtom             # <<<<<<<<<<<<<<
 *                     partners[newAtom] = cbfAtom
 * 
 */
        if (unlikely(PyDict_SetItem(__pyx_v_partners, __pyx_v_cbfAtom, __pyx_v_newAtom) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1529; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

        /* "rmgpy/molecule/group.py":1530
 *                     (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, cbfAtom, True)
 *                     partners[cbfAtom] = newAtom
 *                     partners[newAtom] = cbfAtom             # <<<<<<<<<<<<<<
 * 
 *         #reclassify all atoms since we may have added new ones
 */
        if (unlikely(PyDict_SetItem(__pyx_v_partners, __pyx_v_newAtom, __pyx_v_cbfAtom) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1530; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_L18:;
    }
    __pyx_L9:;

    /* "rmgpy/molecule/group.py":1506
 *         """
 *         partners = {} #dictionary of exclusive partners, has 1:2 and 2:1
 *         for cbfAtom in cbfAtomList1:             # <<<<<<<<<<<<<<
 *             if cbfAtom in partners: continue
 *             #if cbfAtom has a connected cbf it must be the partner
 */
    __pyx_L7_continue:;
  }
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1533
 * 
 *         #reclassify all atoms since we may have added new ones
 *         (cbAtomList, cbfAtomList, cbfAtomList1, cbfAtomList2, connectedCbfs) = classifyBenzeneCarbons(copyGroup, partners)             # <<<<<<<<<<<<<<
 * 
 *         """
 */
  __pyx_t_6 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_classifyBenzeneCarbons(classifyBenzeneCarbons, __pyx_v_copyGroup, __pyx_v_partners); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1533; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  if ((likely(PyTuple_CheckExact(__pyx_t_6))) || (PyList_CheckExact(__pyx_t_6))) {
    PyObject* sequence = __pyx_t_6;
    #if CYTHON_COMPILING_IN_CPYTHON
    Py_ssize_t size = Py_SIZE(sequence);
    #else
    Py_ssize_t size = PySequence_Size(sequence);
    #endif
    if (unlikely(size != 5)) {
      if (size > 5) __Pyx_RaiseTooManyValuesError(5);
      else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1533; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    #if CYTHON_COMPILING_IN_CPYTHON
    if (likely(PyTuple_CheckExact(sequence))) {
      __pyx_t_5 = PyTuple_GET_ITEM(sequence, 0); 
      __pyx_t_2 = PyTuple_GET_ITEM(sequence, 1); 
      __pyx_t_1 = PyTuple_GET_ITEM(sequence, 2); 
      __pyx_t_3 = PyTuple_GET_ITEM(sequence, 3); 
      __pyx_t_4 = PyTuple_GET_ITEM(sequence, 4); 
    } else {
      __pyx_t_5 = PyList_GET_ITEM(sequence, 0); 
      __pyx_t_2 = PyList_GET_ITEM(sequence, 1); 
      __pyx_t_1 = PyList_GET_ITEM(sequence, 2); 
      __pyx_t_3 = PyList_GET_ITEM(sequence, 3); 
      __pyx_t_4 = PyList_GET_ITEM(sequence, 4); 
    }
    __Pyx_INCREF(__pyx_t_5);
    __Pyx_INCREF(__pyx_t_2);
    __Pyx_INCREF(__pyx_t_1);
    __Pyx_INCREF(__pyx_t_3);
    __Pyx_INCREF(__pyx_t_4);
    #else
    {
      Py_ssize_t i;
      PyObject** temps[5] = {&__pyx_t_5,&__pyx_t_2,&__pyx_t_1,&__pyx_t_3,&__pyx_t_4};
      for (i=0; i < 5; i++) {
        PyObject* item = PySequence_ITEM(sequence, i); if (unlikely(!item)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1533; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(item);
        *(temps[i]) = item;
      }
    }
    #endif
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  } else {
    Py_ssize_t index = -1;
    PyObject** temps[5] = {&__pyx_t_5,&__pyx_t_2,&__pyx_t_1,&__pyx_t_3,&__pyx_t_4};
    __pyx_t_7 = PyObject_GetIter(__pyx_t_6); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1533; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_7);
    __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    __pyx_t_8 = Py_TYPE(__pyx_t_7)->tp_iternext;
    for (index=0; index < 5; index++) {
      PyObject* item = __pyx_t_8(__pyx_t_7); if (unlikely(!item)) goto __pyx_L21_unpacking_failed;
      __Pyx_GOTREF(item);
      *(temps[index]) = item;
    }
    if (__Pyx_IternextUnpackEndCheck(__pyx_t_8(__pyx_t_7), 5) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1533; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_8 = NULL;
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
    goto __pyx_L22_unpacking_done;
    __pyx_L21_unpacking_failed:;
    __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
    __pyx_t_8 = NULL;
    if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1533; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_L22_unpacking_done:;
  }
  __Pyx_DECREF_SET(__pyx_v_cbAtomList, __pyx_t_5);
  __pyx_t_5 = 0;
  __Pyx_DECREF_SET(__pyx_v_cbfAtomList, __pyx_t_2);
  __pyx_t_2 = 0;
  __Pyx_DECREF_SET(__pyx_v_cbfAtomList1, __pyx_t_1);
  __pyx_t_1 = 0;
  __Pyx_DECREF_SET(__pyx_v_cbfAtomList2, __pyx_t_3);
  __pyx_t_3 = 0;
  __Pyx_DECREF_SET(__pyx_v_connectedCbfs, __pyx_t_4);
  __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":1543
 *         before completely disconnected atoms. Otherwise, we will erroneously create new rings.
 *         """
 *         cbAtomList = sortByConnectivity(cbAtomList)             # <<<<<<<<<<<<<<
 *         cbfAtomList1 = sortByConnectivity(cbfAtomList1)
 *         cbfAtomList2 = sortByConnectivity(cbfAtomList2)
 */
  __pyx_t_6 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_4sortByConnectivity(sortByConnectivity, __pyx_v_cbAtomList); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1543; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __Pyx_DECREF_SET(__pyx_v_cbAtomList, __pyx_t_6);
  __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1544
 *         """
 *         cbAtomList = sortByConnectivity(cbAtomList)
 *         cbfAtomList1 = sortByConnectivity(cbfAtomList1)             # <<<<<<<<<<<<<<
 *         cbfAtomList2 = sortByConnectivity(cbfAtomList2)
 * 
 */
  __pyx_t_6 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_4sortByConnectivity(sortByConnectivity, __pyx_v_cbfAtomList1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1544; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __Pyx_DECREF_SET(__pyx_v_cbfAtomList1, __pyx_t_6);
  __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1545
 *         cbAtomList = sortByConnectivity(cbAtomList)
 *         cbfAtomList1 = sortByConnectivity(cbfAtomList1)
 *         cbfAtomList2 = sortByConnectivity(cbfAtomList2)             # <<<<<<<<<<<<<<
 * 
 *         """
 */
  __pyx_t_6 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_4sortByConnectivity(sortByConnectivity, __pyx_v_cbfAtomList2); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1545; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __Pyx_DECREF_SET(__pyx_v_cbfAtomList2, __pyx_t_6);
  __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1556
 * 
 *         """
 *         rings=[cycle for cycle in copyGroup.getAllCyclesOfSize(6) if Group(atoms = cycle).isAromaticRing()]             # <<<<<<<<<<<<<<
 * 
 *         """
 */
  __pyx_t_6 = PyList_New(0); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_copyGroup, __pyx_n_s_getAllCyclesOfSize); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_tuple__18, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  if (likely(PyList_CheckExact(__pyx_t_3)) || PyTuple_CheckExact(__pyx_t_3)) {
    __pyx_t_4 = __pyx_t_3; __Pyx_INCREF(__pyx_t_4); __pyx_t_9 = 0;
    __pyx_t_11 = NULL;
  } else {
    __pyx_t_9 = -1; __pyx_t_4 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_11 = Py_TYPE(__pyx_t_4)->tp_iternext; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  for (;;) {
    if (likely(!__pyx_t_11)) {
      if (likely(PyList_CheckExact(__pyx_t_4))) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_4)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_4, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_9 >= PyTuple_GET_SIZE(__pyx_t_4)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_4, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_3 = PySequence_ITEM(__pyx_t_4, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_3 = __pyx_t_11(__pyx_t_4);
      if (unlikely(!__pyx_t_3)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_3);
    }
    __Pyx_XDECREF_SET(__pyx_v_cycle, __pyx_t_3);
    __pyx_t_3 = 0;
    __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_atoms, __pyx_v_cycle) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)), __pyx_empty_tuple, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_13 = (((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_1)->__pyx_base.__pyx_vtab)->isAromaticRing(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_1), 0) != 0);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    if (__pyx_t_13) {
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_6, (PyObject*)__pyx_v_cycle))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L25;
    }
    __pyx_L25:;
  }
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_v_rings = ((PyObject*)__pyx_t_6);
  __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1569
 *         is not available.
 *         """
 *         for cbfAtom in cbfAtomList2:             # <<<<<<<<<<<<<<
 *             if connectedCbfs[cbfAtom][0] is partners[cbfAtom]: otherCbf = connectedCbfs[cbfAtom][1]
 *             else: otherCbf = connectedCbfs[cbfAtom][0]
 */
  if (likely(PyList_CheckExact(__pyx_v_cbfAtomList2)) || PyTuple_CheckExact(__pyx_v_cbfAtomList2)) {
    __pyx_t_6 = __pyx_v_cbfAtomList2; __Pyx_INCREF(__pyx_t_6); __pyx_t_9 = 0;
    __pyx_t_11 = NULL;
  } else {
    __pyx_t_9 = -1; __pyx_t_6 = PyObject_GetIter(__pyx_v_cbfAtomList2); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1569; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_11 = Py_TYPE(__pyx_t_6)->tp_iternext; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1569; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  for (;;) {
    if (likely(!__pyx_t_11)) {
      if (likely(PyList_CheckExact(__pyx_t_6))) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_4 = PyList_GET_ITEM(__pyx_t_6, __pyx_t_9); __Pyx_INCREF(__pyx_t_4); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1569; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_4 = PySequence_ITEM(__pyx_t_6, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1569; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_9 >= PyTuple_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_4 = PyTuple_GET_ITEM(__pyx_t_6, __pyx_t_9); __Pyx_INCREF(__pyx_t_4); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1569; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_4 = PySequence_ITEM(__pyx_t_6, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1569; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_4 = __pyx_t_11(__pyx_t_6);
      if (unlikely(!__pyx_t_4)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1569; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_4);
    }
    __Pyx_XDECREF_SET(__pyx_v_cbfAtom, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1570
 *         """
 *         for cbfAtom in cbfAtomList2:
 *             if connectedCbfs[cbfAtom][0] is partners[cbfAtom]: otherCbf = connectedCbfs[cbfAtom][1]             # <<<<<<<<<<<<<<
 *             else: otherCbf = connectedCbfs[cbfAtom][0]
 *             #These two ring seeds represent the two unique rings
 */
    __pyx_t_4 = PyObject_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1570; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = __Pyx_GetItemInt(__pyx_t_4, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1570; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_4 = __Pyx_PyDict_GetItem(__pyx_v_partners, __pyx_v_cbfAtom); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1570; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_13 = (__pyx_t_1 == __pyx_t_4);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_12 = (__pyx_t_13 != 0);
    if (__pyx_t_12) {
      __pyx_t_4 = PyObject_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1570; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_1 = __Pyx_GetItemInt(__pyx_t_4, 1, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1570; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_XDECREF_SET(__pyx_v_otherCbf, __pyx_t_1);
      __pyx_t_1 = 0;
      goto __pyx_L28;
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":1571
 *         for cbfAtom in cbfAtomList2:
 *             if connectedCbfs[cbfAtom][0] is partners[cbfAtom]: otherCbf = connectedCbfs[cbfAtom][1]
 *             else: otherCbf = connectedCbfs[cbfAtom][0]             # <<<<<<<<<<<<<<
 *             #These two ring seeds represent the two unique rings
 *             newRingSeeds = [[partners[cbfAtom], cbfAtom, otherCbf],
 */
      __pyx_t_1 = PyObject_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1571; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_4 = __Pyx_GetItemInt(__pyx_t_1, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1571; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_XDECREF_SET(__pyx_v_otherCbf, __pyx_t_4);
      __pyx_t_4 = 0;
    }
    __pyx_L28:;

    /* "rmgpy/molecule/group.py":1573
 *             else: otherCbf = connectedCbfs[cbfAtom][0]
 *             #These two ring seeds represent the two unique rings
 *             newRingSeeds = [[partners[cbfAtom], cbfAtom, otherCbf],             # <<<<<<<<<<<<<<
 *                             [partners[cbfAtom], cbfAtom]]
 *             allLigands = cbfAtom.bonds.keys()
 */
    __pyx_t_4 = __Pyx_PyDict_GetItem(__pyx_v_partners, __pyx_v_cbfAtom); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1573; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_1 = PyList_New(3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1573; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    PyList_SET_ITEM(__pyx_t_1, 0, __pyx_t_4);
    __Pyx_GIVEREF(__pyx_t_4);
    __Pyx_INCREF(__pyx_v_cbfAtom);
    PyList_SET_ITEM(__pyx_t_1, 1, __pyx_v_cbfAtom);
    __Pyx_GIVEREF(__pyx_v_cbfAtom);
    __Pyx_INCREF(__pyx_v_otherCbf);
    PyList_SET_ITEM(__pyx_t_1, 2, __pyx_v_otherCbf);
    __Pyx_GIVEREF(__pyx_v_otherCbf);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1574
 *             #These two ring seeds represent the two unique rings
 *             newRingSeeds = [[partners[cbfAtom], cbfAtom, otherCbf],
 *                             [partners[cbfAtom], cbfAtom]]             # <<<<<<<<<<<<<<
 *             allLigands = cbfAtom.bonds.keys()
 *             #add a new cb atom to the second newRing seed
 */
    __pyx_t_4 = __Pyx_PyDict_GetItem(__pyx_v_partners, __pyx_v_cbfAtom); if (unlikely(__pyx_t_4 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1574; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_3 = PyList_New(2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1574; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    PyList_SET_ITEM(__pyx_t_3, 0, __pyx_t_4);
    __Pyx_GIVEREF(__pyx_t_4);
    __Pyx_INCREF(__pyx_v_cbfAtom);
    PyList_SET_ITEM(__pyx_t_3, 1, __pyx_v_cbfAtom);
    __Pyx_GIVEREF(__pyx_v_cbfAtom);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1573
 *             else: otherCbf = connectedCbfs[cbfAtom][0]
 *             #These two ring seeds represent the two unique rings
 *             newRingSeeds = [[partners[cbfAtom], cbfAtom, otherCbf],             # <<<<<<<<<<<<<<
 *                             [partners[cbfAtom], cbfAtom]]
 *             allLigands = cbfAtom.bonds.keys()
 */
    __pyx_t_4 = PyList_New(2); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1573; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyList_SET_ITEM(__pyx_t_4, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    PyList_SET_ITEM(__pyx_t_4, 1, __pyx_t_3);
    __Pyx_GIVEREF(__pyx_t_3);
    __pyx_t_1 = 0;
    __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_newRingSeeds, ((PyObject*)__pyx_t_4));
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1575
 *             newRingSeeds = [[partners[cbfAtom], cbfAtom, otherCbf],
 *                             [partners[cbfAtom], cbfAtom]]
 *             allLigands = cbfAtom.bonds.keys()             # <<<<<<<<<<<<<<
 *             #add a new cb atom to the second newRing seed
 *             if len(allLigands) == 2:
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_cbfAtom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1575; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_keys); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1575; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_3 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_1))) {
      __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_1);
      if (likely(__pyx_t_3)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_1);
        __Pyx_INCREF(__pyx_t_3);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_1, function);
      }
    }
    if (__pyx_t_3) {
      __pyx_t_4 = __Pyx_PyObject_CallOneArg(__pyx_t_1, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1575; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    } else {
      __pyx_t_4 = __Pyx_PyObject_CallNoArg(__pyx_t_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1575; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_XDECREF_SET(__pyx_v_allLigands, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1577
 *             allLigands = cbfAtom.bonds.keys()
 *             #add a new cb atom to the second newRing seed
 *             if len(allLigands) == 2:             # <<<<<<<<<<<<<<
 *                 (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, cbfAtom)
 *                 newRingSeeds[1].append(newAtom)
 */
    __pyx_t_15 = PyObject_Length(__pyx_v_allLigands); if (unlikely(__pyx_t_15 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1577; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_12 = ((__pyx_t_15 == 2) != 0);
    if (__pyx_t_12) {

      /* "rmgpy/molecule/group.py":1578
 *             #add a new cb atom to the second newRing seed
 *             if len(allLigands) == 2:
 *                 (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, cbfAtom)             # <<<<<<<<<<<<<<
 *                 newRingSeeds[1].append(newAtom)
 *             #join the existing atom to the ringSeed
 */
      __pyx_t_4 = PyTuple_New(2); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1578; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_INCREF(__pyx_v_copyGroup);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_v_copyGroup);
      __Pyx_GIVEREF(__pyx_v_copyGroup);
      __Pyx_INCREF(__pyx_v_cbfAtom);
      PyTuple_SET_ITEM(__pyx_t_4, 1, __pyx_v_cbfAtom);
      __Pyx_GIVEREF(__pyx_v_cbfAtom);
      __pyx_t_1 = __Pyx_PyObject_Call(addBenzeneAtomToGroup, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1578; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if ((likely(PyTuple_CheckExact(__pyx_t_1))) || (PyList_CheckExact(__pyx_t_1))) {
        PyObject* sequence = __pyx_t_1;
        #if CYTHON_COMPILING_IN_CPYTHON
        Py_ssize_t size = Py_SIZE(sequence);
        #else
        Py_ssize_t size = PySequence_Size(sequence);
        #endif
        if (unlikely(size != 2)) {
          if (size > 2) __Pyx_RaiseTooManyValuesError(2);
          else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1578; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        #if CYTHON_COMPILING_IN_CPYTHON
        if (likely(PyTuple_CheckExact(sequence))) {
          __pyx_t_4 = PyTuple_GET_ITEM(sequence, 0); 
          __pyx_t_3 = PyTuple_GET_ITEM(sequence, 1); 
        } else {
          __pyx_t_4 = PyList_GET_ITEM(sequence, 0); 
          __pyx_t_3 = PyList_GET_ITEM(sequence, 1); 
        }
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(__pyx_t_3);
        #else
        __pyx_t_4 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1578; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_3 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1578; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        #endif
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      } else {
        Py_ssize_t index = -1;
        __pyx_t_2 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1578; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __pyx_t_8 = Py_TYPE(__pyx_t_2)->tp_iternext;
        index = 0; __pyx_t_4 = __pyx_t_8(__pyx_t_2); if (unlikely(!__pyx_t_4)) goto __pyx_L30_unpacking_failed;
        __Pyx_GOTREF(__pyx_t_4);
        index = 1; __pyx_t_3 = __pyx_t_8(__pyx_t_2); if (unlikely(!__pyx_t_3)) goto __pyx_L30_unpacking_failed;
        __Pyx_GOTREF(__pyx_t_3);
        if (__Pyx_IternextUnpackEndCheck(__pyx_t_8(__pyx_t_2), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1578; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __pyx_t_8 = NULL;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L31_unpacking_done;
        __pyx_L30_unpacking_failed:;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        __pyx_t_8 = NULL;
        if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1578; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __pyx_L31_unpacking_done:;
      }
      __Pyx_DECREF_SET(__pyx_v_copyGroup, __pyx_t_4);
      __pyx_t_4 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1579
 *             if len(allLigands) == 2:
 *                 (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, cbfAtom)
 *                 newRingSeeds[1].append(newAtom)             # <<<<<<<<<<<<<<
 *             #join the existing atom to the ringSeed
 *             elif len(allLigands) == 3:
 */
      __pyx_t_1 = __Pyx_GetItemInt_List(__pyx_v_newRingSeeds, 1, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1579; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_21 = __Pyx_PyObject_Append(__pyx_t_1, __pyx_v_newAtom); if (unlikely(__pyx_t_21 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1579; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L29;
    }

    /* "rmgpy/molecule/group.py":1581
 *                 newRingSeeds[1].append(newAtom)
 *             #join the existing atom to the ringSeed
 *             elif len(allLigands) == 3:             # <<<<<<<<<<<<<<
 *                 for atom2 in allLigands:
 *                     if atom2 not in connectedCbfs[cbfAtom]:
 */
    __pyx_t_15 = PyObject_Length(__pyx_v_allLigands); if (unlikely(__pyx_t_15 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1581; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_12 = ((__pyx_t_15 == 3) != 0);
    if (__pyx_t_12) {

      /* "rmgpy/molecule/group.py":1582
 *             #join the existing atom to the ringSeed
 *             elif len(allLigands) == 3:
 *                 for atom2 in allLigands:             # <<<<<<<<<<<<<<
 *                     if atom2 not in connectedCbfs[cbfAtom]:
 *                         newRingSeeds[1].append(atom2)
 */
      if (likely(PyList_CheckExact(__pyx_v_allLigands)) || PyTuple_CheckExact(__pyx_v_allLigands)) {
        __pyx_t_1 = __pyx_v_allLigands; __Pyx_INCREF(__pyx_t_1); __pyx_t_15 = 0;
        __pyx_t_19 = NULL;
      } else {
        __pyx_t_15 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_v_allLigands); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1582; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_1);
        __pyx_t_19 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_19)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1582; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      for (;;) {
        if (likely(!__pyx_t_19)) {
          if (likely(PyList_CheckExact(__pyx_t_1))) {
            if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_1)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_3 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_3); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1582; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1582; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          } else {
            if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_3); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1582; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1582; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          }
        } else {
          __pyx_t_3 = __pyx_t_19(__pyx_t_1);
          if (unlikely(!__pyx_t_3)) {
            PyObject* exc_type = PyErr_Occurred();
            if (exc_type) {
              if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
              else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1582; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            break;
          }
          __Pyx_GOTREF(__pyx_t_3);
        }
        __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_3);
        __pyx_t_3 = 0;

        /* "rmgpy/molecule/group.py":1583
 *             elif len(allLigands) == 3:
 *                 for atom2 in allLigands:
 *                     if atom2 not in connectedCbfs[cbfAtom]:             # <<<<<<<<<<<<<<
 *                         newRingSeeds[1].append(atom2)
 *                         break
 */
        __pyx_t_3 = PyObject_GetItem(__pyx_v_connectedCbfs, __pyx_v_cbfAtom); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1583; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_12 = (__Pyx_PySequence_Contains(__pyx_v_atom2, __pyx_t_3, Py_NE)); if (unlikely(__pyx_t_12 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1583; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __pyx_t_13 = (__pyx_t_12 != 0);
        if (__pyx_t_13) {

          /* "rmgpy/molecule/group.py":1584
 *                 for atom2 in allLigands:
 *                     if atom2 not in connectedCbfs[cbfAtom]:
 *                         newRingSeeds[1].append(atom2)             # <<<<<<<<<<<<<<
 *                         break
 *             #Check for duplicates, merge or create new rings
 */
          __pyx_t_3 = __Pyx_GetItemInt_List(__pyx_v_newRingSeeds, 1, long, 1, __Pyx_PyInt_From_long, 1, 0, 1); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1584; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_3);
          __pyx_t_21 = __Pyx_PyObject_Append(__pyx_t_3, __pyx_v_atom2); if (unlikely(__pyx_t_21 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1584; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

          /* "rmgpy/molecule/group.py":1585
 *                     if atom2 not in connectedCbfs[cbfAtom]:
 *                         newRingSeeds[1].append(atom2)
 *                         break             # <<<<<<<<<<<<<<
 *             #Check for duplicates, merge or create new rings
 *             for index1, ring1 in enumerate(newRingSeeds):
 */
          goto __pyx_L33_break;
        }

        /* "rmgpy/molecule/group.py":1582
 *             #join the existing atom to the ringSeed
 *             elif len(allLigands) == 3:
 *                 for atom2 in allLigands:             # <<<<<<<<<<<<<<
 *                     if atom2 not in connectedCbfs[cbfAtom]:
 *                         newRingSeeds[1].append(atom2)
 */
      }
      __pyx_L33_break:;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L29;
    }
    __pyx_L29:;

    /* "rmgpy/molecule/group.py":1587
 *                         break
 *             #Check for duplicates, merge or create new rings
 *             for index1, ring1 in enumerate(newRingSeeds):             # <<<<<<<<<<<<<<
 *                 mergeRingDict={}
 *                 for index, ring2 in enumerate(rings):
 */
    __Pyx_INCREF(__pyx_int_0);
    __pyx_t_1 = __pyx_int_0;
    __pyx_t_3 = __pyx_v_newRingSeeds; __Pyx_INCREF(__pyx_t_3); __pyx_t_15 = 0;
    for (;;) {
      if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_3)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_4 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_15); __Pyx_INCREF(__pyx_t_4); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1587; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_4 = PySequence_ITEM(__pyx_t_3, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1587; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_ring1, __pyx_t_4);
      __pyx_t_4 = 0;
      __Pyx_INCREF(__pyx_t_1);
      __Pyx_XDECREF_SET(__pyx_v_index1, __pyx_t_1);
      __pyx_t_4 = PyNumber_Add(__pyx_t_1, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1587; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_1);
      __pyx_t_1 = __pyx_t_4;
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1588
 *             #Check for duplicates, merge or create new rings
 *             for index1, ring1 in enumerate(newRingSeeds):
 *                 mergeRingDict={}             # <<<<<<<<<<<<<<
 *                 for index, ring2 in enumerate(rings):
 *                     #check if a duplicate of a fully created ring
 */
      __pyx_t_4 = PyDict_New(); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1588; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_XDECREF_SET(__pyx_v_mergeRingDict, ((PyObject*)__pyx_t_4));
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1589
 *             for index1, ring1 in enumerate(newRingSeeds):
 *                 mergeRingDict={}
 *                 for index, ring2 in enumerate(rings):             # <<<<<<<<<<<<<<
 *                     #check if a duplicate of a fully created ring
 *                     if checkSet(ring2, ring1): break
 */
      __Pyx_INCREF(__pyx_int_0);
      __pyx_t_4 = __pyx_int_0;
      __pyx_t_2 = __pyx_v_rings; __Pyx_INCREF(__pyx_t_2); __pyx_t_14 = 0;
      for (;;) {
        if (__pyx_t_14 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_5 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_14); __Pyx_INCREF(__pyx_t_5); __pyx_t_14++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1589; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_5 = PySequence_ITEM(__pyx_t_2, __pyx_t_14); __pyx_t_14++; if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1589; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
        __Pyx_XDECREF_SET(__pyx_v_ring2, __pyx_t_5);
        __pyx_t_5 = 0;
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_4);
        __pyx_t_5 = PyNumber_Add(__pyx_t_4, __pyx_int_1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1589; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __Pyx_DECREF(__pyx_t_4);
        __pyx_t_4 = __pyx_t_5;
        __pyx_t_5 = 0;

        /* "rmgpy/molecule/group.py":1591
 *                 for index, ring2 in enumerate(rings):
 *                     #check if a duplicate of a fully created ring
 *                     if checkSet(ring2, ring1): break             # <<<<<<<<<<<<<<
 *                     #Next try to merge the ringseed into rings
 *                     mergeRing = mergeOverlappingBenzeneRings(ring2, ring1, 2)
 */
        __pyx_t_5 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_6checkSet(checkSet, __pyx_v_ring2, __pyx_v_ring1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1591; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1591; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        if (__pyx_t_13) {
          goto __pyx_L38_break;
        }

        /* "rmgpy/molecule/group.py":1593
 *                     if checkSet(ring2, ring1): break
 *                     #Next try to merge the ringseed into rings
 *                     mergeRing = mergeOverlappingBenzeneRings(ring2, ring1, 2)             # <<<<<<<<<<<<<<
 *                     if mergeRing:
 *                         mergeRingDict[index] = mergeRing
 */
        __pyx_t_5 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_8mergeOverlappingBenzeneRings(mergeOverlappingBenzeneRings, __pyx_v_ring2, __pyx_v_ring1, __pyx_int_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1593; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __Pyx_XDECREF_SET(__pyx_v_mergeRing, __pyx_t_5);
        __pyx_t_5 = 0;

        /* "rmgpy/molecule/group.py":1594
 *                     #Next try to merge the ringseed into rings
 *                     mergeRing = mergeOverlappingBenzeneRings(ring2, ring1, 2)
 *                     if mergeRing:             # <<<<<<<<<<<<<<
 *                         mergeRingDict[index] = mergeRing
 *                         break
 */
        __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_v_mergeRing); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1594; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        if (__pyx_t_13) {

          /* "rmgpy/molecule/group.py":1595
 *                     mergeRing = mergeOverlappingBenzeneRings(ring2, ring1, 2)
 *                     if mergeRing:
 *                         mergeRingDict[index] = mergeRing             # <<<<<<<<<<<<<<
 *                         break
 *                 #otherwise add this ringSeed because it represents a completely new ring
 */
          if (unlikely(PyDict_SetItem(__pyx_v_mergeRingDict, __pyx_v_index, __pyx_v_mergeRing) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1595; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

          /* "rmgpy/molecule/group.py":1596
 *                     if mergeRing:
 *                         mergeRingDict[index] = mergeRing
 *                         break             # <<<<<<<<<<<<<<
 *                 #otherwise add this ringSeed because it represents a completely new ring
 *                 else: rings.append(ring1)
 */
          goto __pyx_L38_break;
        }

        /* "rmgpy/molecule/group.py":1589
 *             for index1, ring1 in enumerate(newRingSeeds):
 *                 mergeRingDict={}
 *                 for index, ring2 in enumerate(rings):             # <<<<<<<<<<<<<<
 *                     #check if a duplicate of a fully created ring
 *                     if checkSet(ring2, ring1): break
 */
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":1598
 *                         break
 *                 #otherwise add this ringSeed because it represents a completely new ring
 *                 else: rings.append(ring1)             # <<<<<<<<<<<<<<
 *                 #if we merged a ring, we need to remove the old ring from rings and add the merged ring
 *                 rings = [rings[index] if not index in mergeRingDict else mergeRingDict[index] for index in range(len(rings))]
 */
        __pyx_t_21 = __Pyx_PyList_Append(__pyx_v_rings, __pyx_v_ring1); if (unlikely(__pyx_t_21 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1598; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }

      /* "rmgpy/molecule/group.py":1589
 *             for index1, ring1 in enumerate(newRingSeeds):
 *                 mergeRingDict={}
 *                 for index, ring2 in enumerate(rings):             # <<<<<<<<<<<<<<
 *                     #check if a duplicate of a fully created ring
 *                     if checkSet(ring2, ring1): break
 */
      __pyx_L38_break:;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1600
 *                 else: rings.append(ring1)
 *                 #if we merged a ring, we need to remove the old ring from rings and add the merged ring
 *                 rings = [rings[index] if not index in mergeRingDict else mergeRingDict[index] for index in range(len(rings))]             # <<<<<<<<<<<<<<
 * 
 *         """
 */
      __pyx_t_4 = PyList_New(0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_14 = PyList_GET_SIZE(__pyx_v_rings); if (unlikely(__pyx_t_14 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_2 = PyInt_FromSsize_t(__pyx_t_14); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __pyx_t_5 = PyTuple_New(1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_5);
      PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_2);
      __Pyx_GIVEREF(__pyx_t_2);
      __pyx_t_2 = 0;
      __pyx_t_2 = __Pyx_PyObject_Call(__pyx_builtin_range, __pyx_t_5, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
        __pyx_t_5 = __pyx_t_2; __Pyx_INCREF(__pyx_t_5); __pyx_t_14 = 0;
        __pyx_t_19 = NULL;
      } else {
        __pyx_t_14 = -1; __pyx_t_5 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __pyx_t_19 = Py_TYPE(__pyx_t_5)->tp_iternext; if (unlikely(!__pyx_t_19)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      for (;;) {
        if (likely(!__pyx_t_19)) {
          if (likely(PyList_CheckExact(__pyx_t_5))) {
            if (__pyx_t_14 >= PyList_GET_SIZE(__pyx_t_5)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_2 = PyList_GET_ITEM(__pyx_t_5, __pyx_t_14); __Pyx_INCREF(__pyx_t_2); __pyx_t_14++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_2 = PySequence_ITEM(__pyx_t_5, __pyx_t_14); __pyx_t_14++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          } else {
            if (__pyx_t_14 >= PyTuple_GET_SIZE(__pyx_t_5)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_5, __pyx_t_14); __Pyx_INCREF(__pyx_t_2); __pyx_t_14++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_2 = PySequence_ITEM(__pyx_t_5, __pyx_t_14); __pyx_t_14++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          }
        } else {
          __pyx_t_2 = __pyx_t_19(__pyx_t_5);
          if (unlikely(!__pyx_t_2)) {
            PyObject* exc_type = PyErr_Occurred();
            if (exc_type) {
              if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
              else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            break;
          }
          __Pyx_GOTREF(__pyx_t_2);
        }
        __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_2);
        __pyx_t_2 = 0;
        __pyx_t_13 = (__Pyx_PyDict_Contains(__pyx_v_index, __pyx_v_mergeRingDict, Py_NE)); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        if ((__pyx_t_13 != 0)) {
          __pyx_t_7 = PyObject_GetItem(__pyx_v_rings, __pyx_v_index); if (unlikely(__pyx_t_7 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_7);
          __pyx_t_2 = __pyx_t_7;
          __pyx_t_7 = 0;
        } else {
          __pyx_t_7 = __Pyx_PyDict_GetItem(__pyx_v_mergeRingDict, __pyx_v_index); if (unlikely(__pyx_t_7 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
          __Pyx_GOTREF(__pyx_t_7);
          __pyx_t_2 = __pyx_t_7;
          __pyx_t_7 = 0;
        }
        if (unlikely(__Pyx_ListComp_Append(__pyx_t_4, (PyObject*)__pyx_t_2))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1600; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      }
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      __Pyx_DECREF_SET(__pyx_v_rings, ((PyObject*)__pyx_t_4));
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1587
 *                         break
 *             #Check for duplicates, merge or create new rings
 *             for index1, ring1 in enumerate(newRingSeeds):             # <<<<<<<<<<<<<<
 *                 mergeRingDict={}
 *                 for index, ring2 in enumerate(rings):
 */
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1569
 *         is not available.
 *         """
 *         for cbfAtom in cbfAtomList2:             # <<<<<<<<<<<<<<
 *             if connectedCbfs[cbfAtom][0] is partners[cbfAtom]: otherCbf = connectedCbfs[cbfAtom][1]
 *             else: otherCbf = connectedCbfs[cbfAtom][0]
 */
  }
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1608
 *         twice to the rings.
 *         """
 *         for cbfAtom in cbfAtomList1:             # <<<<<<<<<<<<<<
 *             newRingSeed = [partners[cbfAtom], cbfAtom]
 *             inRing = 0
 */
  if (likely(PyList_CheckExact(__pyx_v_cbfAtomList1)) || PyTuple_CheckExact(__pyx_v_cbfAtomList1)) {
    __pyx_t_6 = __pyx_v_cbfAtomList1; __Pyx_INCREF(__pyx_t_6); __pyx_t_9 = 0;
    __pyx_t_11 = NULL;
  } else {
    __pyx_t_9 = -1; __pyx_t_6 = PyObject_GetIter(__pyx_v_cbfAtomList1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1608; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_11 = Py_TYPE(__pyx_t_6)->tp_iternext; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1608; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  for (;;) {
    if (likely(!__pyx_t_11)) {
      if (likely(PyList_CheckExact(__pyx_t_6))) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyList_GET_ITEM(__pyx_t_6, __pyx_t_9); __Pyx_INCREF(__pyx_t_1); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1608; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_6, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1608; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_9 >= PyTuple_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_6, __pyx_t_9); __Pyx_INCREF(__pyx_t_1); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1608; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_6, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1608; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_1 = __pyx_t_11(__pyx_t_6);
      if (unlikely(!__pyx_t_1)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1608; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_1);
    }
    __Pyx_XDECREF_SET(__pyx_v_cbfAtom, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1609
 *         """
 *         for cbfAtom in cbfAtomList1:
 *             newRingSeed = [partners[cbfAtom], cbfAtom]             # <<<<<<<<<<<<<<
 *             inRing = 0
 *             #check to see if duplicate of an existing ring
 */
    __pyx_t_1 = __Pyx_PyDict_GetItem(__pyx_v_partners, __pyx_v_cbfAtom); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1609; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_3 = PyList_New(2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1609; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    PyList_SET_ITEM(__pyx_t_3, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __Pyx_INCREF(__pyx_v_cbfAtom);
    PyList_SET_ITEM(__pyx_t_3, 1, __pyx_v_cbfAtom);
    __Pyx_GIVEREF(__pyx_v_cbfAtom);
    __pyx_t_1 = 0;
    __Pyx_XDECREF_SET(__pyx_v_newRingSeed, ((PyObject*)__pyx_t_3));
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1610
 *         for cbfAtom in cbfAtomList1:
 *             newRingSeed = [partners[cbfAtom], cbfAtom]
 *             inRing = 0             # <<<<<<<<<<<<<<
 *             #check to see if duplicate of an existing ring
 *             for ring in rings:
 */
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_inRing, __pyx_int_0);

    /* "rmgpy/molecule/group.py":1612
 *             inRing = 0
 *             #check to see if duplicate of an existing ring
 *             for ring in rings:             # <<<<<<<<<<<<<<
 *                 if checkSet(ring, newRingSeed):
 *                     inRing +=1
 */
    __pyx_t_3 = __pyx_v_rings; __Pyx_INCREF(__pyx_t_3); __pyx_t_15 = 0;
    for (;;) {
      if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_3)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_1 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_15); __Pyx_INCREF(__pyx_t_1); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1612; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_1 = PySequence_ITEM(__pyx_t_3, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1612; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_ring, __pyx_t_1);
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1613
 *             #check to see if duplicate of an existing ring
 *             for ring in rings:
 *                 if checkSet(ring, newRingSeed):             # <<<<<<<<<<<<<<
 *                     inRing +=1
 *             #move on to next cbfAtom if we found two rings
 */
      __pyx_t_1 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_6checkSet(checkSet, __pyx_v_ring, __pyx_v_newRingSeed); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1613; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      if (__pyx_t_13) {

        /* "rmgpy/molecule/group.py":1614
 *             for ring in rings:
 *                 if checkSet(ring, newRingSeed):
 *                     inRing +=1             # <<<<<<<<<<<<<<
 *             #move on to next cbfAtom if we found two rings
 *             if inRing ==2: continue
 */
        __pyx_t_1 = PyNumber_InPlaceAdd(__pyx_v_inRing, __pyx_int_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1614; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_1);
        __Pyx_DECREF_SET(__pyx_v_inRing, __pyx_t_1);
        __pyx_t_1 = 0;
        goto __pyx_L48;
      }
      __pyx_L48:;

      /* "rmgpy/molecule/group.py":1612
 *             inRing = 0
 *             #check to see if duplicate of an existing ring
 *             for ring in rings:             # <<<<<<<<<<<<<<
 *                 if checkSet(ring, newRingSeed):
 *                     inRing +=1
 */
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1616
 *                     inRing +=1
 *             #move on to next cbfAtom if we found two rings
 *             if inRing ==2: continue             # <<<<<<<<<<<<<<
 *             #try to merge into existing rings, if cbf1 is connected
 *             for index, ring in enumerate(rings):
 */
    __pyx_t_3 = PyObject_RichCompare(__pyx_v_inRing, __pyx_int_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1616; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1616; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    if (__pyx_t_13) {
      goto __pyx_L44_continue;
    }

    /* "rmgpy/molecule/group.py":1618
 *             if inRing ==2: continue
 *             #try to merge into existing rings, if cbf1 is connected
 *             for index, ring in enumerate(rings):             # <<<<<<<<<<<<<<
 *                 mergeRingDict={}
 *                 mergeRing = mergeOverlappingBenzeneRings(ring, newRingSeed, 1)
 */
    __Pyx_INCREF(__pyx_int_0);
    __pyx_t_3 = __pyx_int_0;
    __pyx_t_1 = __pyx_v_rings; __Pyx_INCREF(__pyx_t_1); __pyx_t_15 = 0;
    for (;;) {
      if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_4 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_4); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1618; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_4 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1618; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_ring, __pyx_t_4);
      __pyx_t_4 = 0;
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_3);
      __pyx_t_4 = PyNumber_Add(__pyx_t_3, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1618; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_3);
      __pyx_t_3 = __pyx_t_4;
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1619
 *             #try to merge into existing rings, if cbf1 is connected
 *             for index, ring in enumerate(rings):
 *                 mergeRingDict={}             # <<<<<<<<<<<<<<
 *                 mergeRing = mergeOverlappingBenzeneRings(ring, newRingSeed, 1)
 *                 if mergeRing:
 */
      __pyx_t_4 = PyDict_New(); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1619; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_XDECREF_SET(__pyx_v_mergeRingDict, ((PyObject*)__pyx_t_4));
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1620
 *             for index, ring in enumerate(rings):
 *                 mergeRingDict={}
 *                 mergeRing = mergeOverlappingBenzeneRings(ring, newRingSeed, 1)             # <<<<<<<<<<<<<<
 *                 if mergeRing:
 *                     inRing+=1
 */
      __pyx_t_4 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_8mergeOverlappingBenzeneRings(mergeOverlappingBenzeneRings, __pyx_v_ring, __pyx_v_newRingSeed, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1620; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_XDECREF_SET(__pyx_v_mergeRing, __pyx_t_4);
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1621
 *                 mergeRingDict={}
 *                 mergeRing = mergeOverlappingBenzeneRings(ring, newRingSeed, 1)
 *                 if mergeRing:             # <<<<<<<<<<<<<<
 *                     inRing+=1
 *                     mergeRingDict[index]=mergeRing
 */
      __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_v_mergeRing); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1621; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_13) {

        /* "rmgpy/molecule/group.py":1622
 *                 mergeRing = mergeOverlappingBenzeneRings(ring, newRingSeed, 1)
 *                 if mergeRing:
 *                     inRing+=1             # <<<<<<<<<<<<<<
 *                     mergeRingDict[index]=mergeRing
 *                     #move on to next cbfAtom if we found two rings
 */
        __pyx_t_4 = PyNumber_InPlaceAdd(__pyx_v_inRing, __pyx_int_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1622; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF_SET(__pyx_v_inRing, __pyx_t_4);
        __pyx_t_4 = 0;

        /* "rmgpy/molecule/group.py":1623
 *                 if mergeRing:
 *                     inRing+=1
 *                     mergeRingDict[index]=mergeRing             # <<<<<<<<<<<<<<
 *                     #move on to next cbfAtom if we found two rings
 *                     if inRing ==2: break
 */
        if (unlikely(PyDict_SetItem(__pyx_v_mergeRingDict, __pyx_v_index, __pyx_v_mergeRing) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1623; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

        /* "rmgpy/molecule/group.py":1625
 *                     mergeRingDict[index]=mergeRing
 *                     #move on to next cbfAtom if we found two rings
 *                     if inRing ==2: break             # <<<<<<<<<<<<<<
 *             #if we merged a ring, we need to remove the old ring from rings and add the merged ring
 *             rings = [rings[index] if not index in mergeRingDict else mergeRingDict[index] for index in range(len(rings))]
 */
        __pyx_t_4 = PyObject_RichCompare(__pyx_v_inRing, __pyx_int_2, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1625; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        if (__pyx_t_13) {
          goto __pyx_L51_break;
        }
        goto __pyx_L52;
      }
      __pyx_L52:;

      /* "rmgpy/molecule/group.py":1618
 *             if inRing ==2: continue
 *             #try to merge into existing rings, if cbf1 is connected
 *             for index, ring in enumerate(rings):             # <<<<<<<<<<<<<<
 *                 mergeRingDict={}
 *                 mergeRing = mergeOverlappingBenzeneRings(ring, newRingSeed, 1)
 */
    }
    __pyx_L51_break:;
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1627
 *                     if inRing ==2: break
 *             #if we merged a ring, we need to remove the old ring from rings and add the merged ring
 *             rings = [rings[index] if not index in mergeRingDict else mergeRingDict[index] for index in range(len(rings))]             # <<<<<<<<<<<<<<
 *             #if we still dont have two ring, we create a completely new ring
 *             if inRing < 2:
 */
    __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_15 = PyList_GET_SIZE(__pyx_v_rings); if (unlikely(__pyx_t_15 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = PyInt_FromSsize_t(__pyx_t_15); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_4 = PyTuple_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_range, __pyx_t_4, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
      __pyx_t_4 = __pyx_t_1; __Pyx_INCREF(__pyx_t_4); __pyx_t_15 = 0;
      __pyx_t_19 = NULL;
    } else {
      __pyx_t_15 = -1; __pyx_t_4 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_19 = Py_TYPE(__pyx_t_4)->tp_iternext; if (unlikely(!__pyx_t_19)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    for (;;) {
      if (likely(!__pyx_t_19)) {
        if (likely(PyList_CheckExact(__pyx_t_4))) {
          if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_4)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_1 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_15); __Pyx_INCREF(__pyx_t_1); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_1 = PySequence_ITEM(__pyx_t_4, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        } else {
          if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_4)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_4, __pyx_t_15); __Pyx_INCREF(__pyx_t_1); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_1 = PySequence_ITEM(__pyx_t_4, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        }
      } else {
        __pyx_t_1 = __pyx_t_19(__pyx_t_4);
        if (unlikely(!__pyx_t_1)) {
          PyObject* exc_type = PyErr_Occurred();
          if (exc_type) {
            if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
            else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          break;
        }
        __Pyx_GOTREF(__pyx_t_1);
      }
      __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_1);
      __pyx_t_1 = 0;
      if (unlikely(!__pyx_v_mergeRingDict)) { __Pyx_RaiseUnboundLocalError("mergeRingDict"); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;} }
      __pyx_t_13 = (__Pyx_PyDict_Contains(__pyx_v_index, __pyx_v_mergeRingDict, Py_NE)); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if ((__pyx_t_13 != 0)) {
        __pyx_t_5 = PyObject_GetItem(__pyx_v_rings, __pyx_v_index); if (unlikely(__pyx_t_5 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_5);
        __pyx_t_1 = __pyx_t_5;
        __pyx_t_5 = 0;
      } else {
        if (unlikely(!__pyx_v_mergeRingDict)) { __Pyx_RaiseUnboundLocalError("mergeRingDict"); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;} }
        __pyx_t_5 = __Pyx_PyDict_GetItem(__pyx_v_mergeRingDict, __pyx_v_index); if (unlikely(__pyx_t_5 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_5);
        __pyx_t_1 = __pyx_t_5;
        __pyx_t_5 = 0;
      }
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_3, (PyObject*)__pyx_t_1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1627; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    }
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __Pyx_DECREF_SET(__pyx_v_rings, ((PyObject*)__pyx_t_3));
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1629
 *             rings = [rings[index] if not index in mergeRingDict else mergeRingDict[index] for index in range(len(rings))]
 *             #if we still dont have two ring, we create a completely new ring
 *             if inRing < 2:             # <<<<<<<<<<<<<<
 *                 for x in range(2-inRing):
 *                     rings.append(copy(newRingSeed))
 */
    __pyx_t_3 = PyObject_RichCompare(__pyx_v_inRing, __pyx_int_2, Py_LT); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1629; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1629; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    if (__pyx_t_13) {

      /* "rmgpy/molecule/group.py":1630
 *             #if we still dont have two ring, we create a completely new ring
 *             if inRing < 2:
 *                 for x in range(2-inRing):             # <<<<<<<<<<<<<<
 *                     rings.append(copy(newRingSeed))
 * 
 */
      __pyx_t_3 = PyNumber_Subtract(__pyx_int_2, __pyx_v_inRing); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyTuple_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3);
      __Pyx_GIVEREF(__pyx_t_3);
      __pyx_t_3 = 0;
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_range, __pyx_t_4, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (likely(PyList_CheckExact(__pyx_t_3)) || PyTuple_CheckExact(__pyx_t_3)) {
        __pyx_t_4 = __pyx_t_3; __Pyx_INCREF(__pyx_t_4); __pyx_t_15 = 0;
        __pyx_t_19 = NULL;
      } else {
        __pyx_t_15 = -1; __pyx_t_4 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_19 = Py_TYPE(__pyx_t_4)->tp_iternext; if (unlikely(!__pyx_t_19)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      for (;;) {
        if (likely(!__pyx_t_19)) {
          if (likely(PyList_CheckExact(__pyx_t_4))) {
            if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_4)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_3 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_15); __Pyx_INCREF(__pyx_t_3); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_3 = PySequence_ITEM(__pyx_t_4, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          } else {
            if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_4)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_4, __pyx_t_15); __Pyx_INCREF(__pyx_t_3); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_3 = PySequence_ITEM(__pyx_t_4, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          }
        } else {
          __pyx_t_3 = __pyx_t_19(__pyx_t_4);
          if (unlikely(!__pyx_t_3)) {
            PyObject* exc_type = PyErr_Occurred();
            if (exc_type) {
              if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
              else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1630; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            break;
          }
          __Pyx_GOTREF(__pyx_t_3);
        }
        __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_3);
        __pyx_t_3 = 0;

        /* "rmgpy/molecule/group.py":1631
 *             if inRing < 2:
 *                 for x in range(2-inRing):
 *                     rings.append(copy(newRingSeed))             # <<<<<<<<<<<<<<
 * 
 *         """
 */
        __pyx_t_1 = __Pyx_GetModuleGlobalName(__pyx_n_s_copy); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1631; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_1);
        __pyx_t_5 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_1))) {
          __pyx_t_5 = PyMethod_GET_SELF(__pyx_t_1);
          if (likely(__pyx_t_5)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_1);
            __Pyx_INCREF(__pyx_t_5);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_1, function);
          }
        }
        if (!__pyx_t_5) {
          __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_1, __pyx_v_newRingSeed); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1631; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
        } else {
          __pyx_t_2 = PyTuple_New(1+1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1631; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_t_5); __Pyx_GIVEREF(__pyx_t_5); __pyx_t_5 = NULL;
          __Pyx_INCREF(__pyx_v_newRingSeed);
          PyTuple_SET_ITEM(__pyx_t_2, 0+1, __pyx_v_newRingSeed);
          __Pyx_GIVEREF(__pyx_v_newRingSeed);
          __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_1, __pyx_t_2, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1631; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        }
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __pyx_t_21 = __Pyx_PyList_Append(__pyx_v_rings, __pyx_t_3); if (unlikely(__pyx_t_21 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1631; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

        /* "rmgpy/molecule/group.py":1630
 *             #if we still dont have two ring, we create a completely new ring
 *             if inRing < 2:
 *                 for x in range(2-inRing):             # <<<<<<<<<<<<<<
 *                     rings.append(copy(newRingSeed))
 * 
 */
      }
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      goto __pyx_L56;
    }
    __pyx_L56:;

    /* "rmgpy/molecule/group.py":1608
 *         twice to the rings.
 *         """
 *         for cbfAtom in cbfAtomList1:             # <<<<<<<<<<<<<<
 *             newRingSeed = [partners[cbfAtom], cbfAtom]
 *             inRing = 0
 */
    __pyx_L44_continue:;
  }
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1639
 *         a new ring for each Cb atom.
 *         """
 *         for cbAtom in cbAtomList:             # <<<<<<<<<<<<<<
 *             inRing = 0
 *             #check to see if already in a ring
 */
  if (likely(PyList_CheckExact(__pyx_v_cbAtomList)) || PyTuple_CheckExact(__pyx_v_cbAtomList)) {
    __pyx_t_6 = __pyx_v_cbAtomList; __Pyx_INCREF(__pyx_t_6); __pyx_t_9 = 0;
    __pyx_t_11 = NULL;
  } else {
    __pyx_t_9 = -1; __pyx_t_6 = PyObject_GetIter(__pyx_v_cbAtomList); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_6);
    __pyx_t_11 = Py_TYPE(__pyx_t_6)->tp_iternext; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  for (;;) {
    if (likely(!__pyx_t_11)) {
      if (likely(PyList_CheckExact(__pyx_t_6))) {
        if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_4 = PyList_GET_ITEM(__pyx_t_6, __pyx_t_9); __Pyx_INCREF(__pyx_t_4); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_4 = PySequence_ITEM(__pyx_t_6, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_9 >= PyTuple_GET_SIZE(__pyx_t_6)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_4 = PyTuple_GET_ITEM(__pyx_t_6, __pyx_t_9); __Pyx_INCREF(__pyx_t_4); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_4 = PySequence_ITEM(__pyx_t_6, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_4 = __pyx_t_11(__pyx_t_6);
      if (unlikely(!__pyx_t_4)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1639; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_4);
    }
    __Pyx_XDECREF_SET(__pyx_v_cbAtom, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1640
 *         """
 *         for cbAtom in cbAtomList:
 *             inRing = 0             # <<<<<<<<<<<<<<
 *             #check to see if already in a ring
 *             for ring in rings:
 */
    __Pyx_INCREF(__pyx_int_0);
    __Pyx_XDECREF_SET(__pyx_v_inRing, __pyx_int_0);

    /* "rmgpy/molecule/group.py":1642
 *             inRing = 0
 *             #check to see if already in a ring
 *             for ring in rings:             # <<<<<<<<<<<<<<
 *                 if checkSet(ring, [cbAtom]): inRing +=1
 *             #move on to next ring cbAtom if in a ring
 */
    __pyx_t_4 = __pyx_v_rings; __Pyx_INCREF(__pyx_t_4); __pyx_t_15 = 0;
    for (;;) {
      if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_4)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_15); __Pyx_INCREF(__pyx_t_3); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1642; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_4, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1642; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_ring, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1643
 *             #check to see if already in a ring
 *             for ring in rings:
 *                 if checkSet(ring, [cbAtom]): inRing +=1             # <<<<<<<<<<<<<<
 *             #move on to next ring cbAtom if in a ring
 *             if inRing == 1 : continue
 */
      __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1643; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_v_cbAtom);
      PyList_SET_ITEM(__pyx_t_3, 0, __pyx_v_cbAtom);
      __Pyx_GIVEREF(__pyx_v_cbAtom);
      __pyx_t_1 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_6checkSet(checkSet, __pyx_v_ring, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1643; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1643; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      if (__pyx_t_13) {
        __pyx_t_1 = PyNumber_InPlaceAdd(__pyx_v_inRing, __pyx_int_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1643; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_1);
        __Pyx_DECREF_SET(__pyx_v_inRing, __pyx_t_1);
        __pyx_t_1 = 0;
        goto __pyx_L63;
      }
      __pyx_L63:;

      /* "rmgpy/molecule/group.py":1642
 *             inRing = 0
 *             #check to see if already in a ring
 *             for ring in rings:             # <<<<<<<<<<<<<<
 *                 if checkSet(ring, [cbAtom]): inRing +=1
 *             #move on to next ring cbAtom if in a ring
 */
    }
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1645
 *                 if checkSet(ring, [cbAtom]): inRing +=1
 *             #move on to next ring cbAtom if in a ring
 *             if inRing == 1 : continue             # <<<<<<<<<<<<<<
 *             #check to see if can be merged to an existing ring
 *             for index, ring in enumerate(rings):
 */
    __pyx_t_4 = PyObject_RichCompare(__pyx_v_inRing, __pyx_int_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1645; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1645; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    if (__pyx_t_13) {
      goto __pyx_L59_continue;
    }

    /* "rmgpy/molecule/group.py":1647
 *             if inRing == 1 : continue
 *             #check to see if can be merged to an existing ring
 *             for index, ring in enumerate(rings):             # <<<<<<<<<<<<<<
 *                 mergeRingDict={}
 *                 mergeRing = addCbAtomToRing(ring, cbAtom)
 */
    __Pyx_INCREF(__pyx_int_0);
    __pyx_t_4 = __pyx_int_0;
    __pyx_t_1 = __pyx_v_rings; __Pyx_INCREF(__pyx_t_1); __pyx_t_15 = 0;
    for (;;) {
      if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_15); __Pyx_INCREF(__pyx_t_3); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_ring, __pyx_t_3);
      __pyx_t_3 = 0;
      __Pyx_INCREF(__pyx_t_4);
      __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_4);
      __pyx_t_3 = PyNumber_Add(__pyx_t_4, __pyx_int_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1647; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_4);
      __pyx_t_4 = __pyx_t_3;
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1648
 *             #check to see if can be merged to an existing ring
 *             for index, ring in enumerate(rings):
 *                 mergeRingDict={}             # <<<<<<<<<<<<<<
 *                 mergeRing = addCbAtomToRing(ring, cbAtom)
 *                 if mergeRing:
 */
      __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1648; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_XDECREF_SET(__pyx_v_mergeRingDict, ((PyObject*)__pyx_t_3));
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1649
 *             for index, ring in enumerate(rings):
 *                 mergeRingDict={}
 *                 mergeRing = addCbAtomToRing(ring, cbAtom)             # <<<<<<<<<<<<<<
 *                 if mergeRing:
 *                     inRing+=1
 */
      __pyx_t_3 = __pyx_pf_5rmgpy_8molecule_5group_5Group_18addImplicitBenzene_10addCbAtomToRing(addCbAtomToRing, __pyx_v_ring, __pyx_v_cbAtom); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1649; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_XDECREF_SET(__pyx_v_mergeRing, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1650
 *                 mergeRingDict={}
 *                 mergeRing = addCbAtomToRing(ring, cbAtom)
 *                 if mergeRing:             # <<<<<<<<<<<<<<
 *                     inRing+=1
 *                     mergeRingDict[index]=mergeRing
 */
      __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_v_mergeRing); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1650; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if (__pyx_t_13) {

        /* "rmgpy/molecule/group.py":1651
 *                 mergeRing = addCbAtomToRing(ring, cbAtom)
 *                 if mergeRing:
 *                     inRing+=1             # <<<<<<<<<<<<<<
 *                     mergeRingDict[index]=mergeRing
 *                     break
 */
        __pyx_t_3 = PyNumber_InPlaceAdd(__pyx_v_inRing, __pyx_int_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1651; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __Pyx_DECREF_SET(__pyx_v_inRing, __pyx_t_3);
        __pyx_t_3 = 0;

        /* "rmgpy/molecule/group.py":1652
 *                 if mergeRing:
 *                     inRing+=1
 *                     mergeRingDict[index]=mergeRing             # <<<<<<<<<<<<<<
 *                     break
 *             #if we merged a ring, we need to remove the old ring from rings and add the merged ring
 */
        if (unlikely(PyDict_SetItem(__pyx_v_mergeRingDict, __pyx_v_index, __pyx_v_mergeRing) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1652; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

        /* "rmgpy/molecule/group.py":1653
 *                     inRing+=1
 *                     mergeRingDict[index]=mergeRing
 *                     break             # <<<<<<<<<<<<<<
 *             #if we merged a ring, we need to remove the old ring from rings and add the merged ring
 *             rings = [rings[index] if not index in mergeRingDict else mergeRingDict[index] for index in range(len(rings))]
 */
        goto __pyx_L66_break;
      }

      /* "rmgpy/molecule/group.py":1647
 *             if inRing == 1 : continue
 *             #check to see if can be merged to an existing ring
 *             for index, ring in enumerate(rings):             # <<<<<<<<<<<<<<
 *                 mergeRingDict={}
 *                 mergeRing = addCbAtomToRing(ring, cbAtom)
 */
    }
    __pyx_L66_break:;
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1655
 *                     break
 *             #if we merged a ring, we need to remove the old ring from rings and add the merged ring
 *             rings = [rings[index] if not index in mergeRingDict else mergeRingDict[index] for index in range(len(rings))]             # <<<<<<<<<<<<<<
 *             #Start completely new ring if not any of above true
 *             if inRing == 0:
 */
    __pyx_t_4 = PyList_New(0); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_15 = PyList_GET_SIZE(__pyx_v_rings); if (unlikely(__pyx_t_15 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_1 = PyInt_FromSsize_t(__pyx_t_15); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_3 = PyTuple_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_1);
    __Pyx_GIVEREF(__pyx_t_1);
    __pyx_t_1 = 0;
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_range, __pyx_t_3, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
      __pyx_t_3 = __pyx_t_1; __Pyx_INCREF(__pyx_t_3); __pyx_t_15 = 0;
      __pyx_t_19 = NULL;
    } else {
      __pyx_t_15 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_19 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_19)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    for (;;) {
      if (likely(!__pyx_t_19)) {
        if (likely(PyList_CheckExact(__pyx_t_3))) {
          if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_3)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_1 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_15); __Pyx_INCREF(__pyx_t_1); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_1 = PySequence_ITEM(__pyx_t_3, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        } else {
          if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_15); __Pyx_INCREF(__pyx_t_1); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_1 = PySequence_ITEM(__pyx_t_3, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        }
      } else {
        __pyx_t_1 = __pyx_t_19(__pyx_t_3);
        if (unlikely(!__pyx_t_1)) {
          PyObject* exc_type = PyErr_Occurred();
          if (exc_type) {
            if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
            else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          break;
        }
        __Pyx_GOTREF(__pyx_t_1);
      }
      __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_1);
      __pyx_t_1 = 0;
      if (unlikely(!__pyx_v_mergeRingDict)) { __Pyx_RaiseUnboundLocalError("mergeRingDict"); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;} }
      __pyx_t_13 = (__Pyx_PyDict_Contains(__pyx_v_index, __pyx_v_mergeRingDict, Py_NE)); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      if ((__pyx_t_13 != 0)) {
        __pyx_t_2 = PyObject_GetItem(__pyx_v_rings, __pyx_v_index); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_2);
        __pyx_t_1 = __pyx_t_2;
        __pyx_t_2 = 0;
      } else {
        if (unlikely(!__pyx_v_mergeRingDict)) { __Pyx_RaiseUnboundLocalError("mergeRingDict"); {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;} }
        __pyx_t_2 = __Pyx_PyDict_GetItem(__pyx_v_mergeRingDict, __pyx_v_index); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_2);
        __pyx_t_1 = __pyx_t_2;
        __pyx_t_2 = 0;
      }
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_4, (PyObject*)__pyx_t_1))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1655; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_DECREF_SET(__pyx_v_rings, ((PyObject*)__pyx_t_4));
    __pyx_t_4 = 0;

    /* "rmgpy/molecule/group.py":1657
 *             rings = [rings[index] if not index in mergeRingDict else mergeRingDict[index] for index in range(len(rings))]
 *             #Start completely new ring if not any of above true
 *             if inRing == 0:             # <<<<<<<<<<<<<<
 *                 rings.append([cbAtom])
 * 
 */
    __pyx_t_4 = PyObject_RichCompare(__pyx_v_inRing, __pyx_int_0, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1657; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1657; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    if (__pyx_t_13) {

      /* "rmgpy/molecule/group.py":1658
 *             #Start completely new ring if not any of above true
 *             if inRing == 0:
 *                 rings.append([cbAtom])             # <<<<<<<<<<<<<<
 * 
 *         """
 */
      __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1658; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_INCREF(__pyx_v_cbAtom);
      PyList_SET_ITEM(__pyx_t_4, 0, __pyx_v_cbAtom);
      __Pyx_GIVEREF(__pyx_v_cbAtom);
      __pyx_t_21 = __Pyx_PyList_Append(__pyx_v_rings, __pyx_t_4); if (unlikely(__pyx_t_21 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1658; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      goto __pyx_L70;
    }
    __pyx_L70:;

    /* "rmgpy/molecule/group.py":1639
 *         a new ring for each Cb atom.
 *         """
 *         for cbAtom in cbAtomList:             # <<<<<<<<<<<<<<
 *             inRing = 0
 *             #check to see if already in a ring
 */
    __pyx_L59_continue:;
  }
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1665
 *         In this step we create new Cb atoms and add them to any rings which do not have 6 atoms.
 *         """
 *         mergedRingDict={}             # <<<<<<<<<<<<<<
 *         for index, ring in enumerate(rings):
 *             carbonsToGrow = 6-len(ring)
 */
  __pyx_t_6 = PyDict_New(); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1665; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  __pyx_v_mergedRingDict = ((PyObject*)__pyx_t_6);
  __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1666
 *         """
 *         mergedRingDict={}
 *         for index, ring in enumerate(rings):             # <<<<<<<<<<<<<<
 *             carbonsToGrow = 6-len(ring)
 *             mergedRingDict[index] = []
 */
  __Pyx_INCREF(__pyx_int_0);
  __pyx_t_6 = __pyx_int_0;
  __pyx_t_4 = __pyx_v_rings; __Pyx_INCREF(__pyx_t_4); __pyx_t_9 = 0;
  for (;;) {
    if (__pyx_t_9 >= PyList_GET_SIZE(__pyx_t_4)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_3 = PyList_GET_ITEM(__pyx_t_4, __pyx_t_9); __Pyx_INCREF(__pyx_t_3); __pyx_t_9++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1666; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_3 = PySequence_ITEM(__pyx_t_4, __pyx_t_9); __pyx_t_9++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1666; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_ring, __pyx_t_3);
    __pyx_t_3 = 0;
    __Pyx_INCREF(__pyx_t_6);
    __Pyx_XDECREF_SET(__pyx_v_index, __pyx_t_6);
    __pyx_t_3 = PyNumber_Add(__pyx_t_6, __pyx_int_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1666; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_6);
    __pyx_t_6 = __pyx_t_3;
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1667
 *         mergedRingDict={}
 *         for index, ring in enumerate(rings):
 *             carbonsToGrow = 6-len(ring)             # <<<<<<<<<<<<<<
 *             mergedRingDict[index] = []
 *             for x in range(carbonsToGrow):
 */
    __pyx_t_15 = PyObject_Length(__pyx_v_ring); if (unlikely(__pyx_t_15 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_3 = PyInt_FromSsize_t((6 - __pyx_t_15)); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1667; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_XDECREF_SET(__pyx_v_carbonsToGrow, __pyx_t_3);
    __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1668
 *         for index, ring in enumerate(rings):
 *             carbonsToGrow = 6-len(ring)
 *             mergedRingDict[index] = []             # <<<<<<<<<<<<<<
 *             for x in range(carbonsToGrow):
 *                 if x ==0: lastAtom = ring[-1]
 */
    __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1668; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    if (unlikely(PyDict_SetItem(__pyx_v_mergedRingDict, __pyx_v_index, __pyx_t_3) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1668; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1669
 *             carbonsToGrow = 6-len(ring)
 *             mergedRingDict[index] = []
 *             for x in range(carbonsToGrow):             # <<<<<<<<<<<<<<
 *                 if x ==0: lastAtom = ring[-1]
 *                 else: lastAtom = mergedRingDict[index][-1]
 */
    __pyx_t_3 = PyTuple_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_INCREF(__pyx_v_carbonsToGrow);
    PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_v_carbonsToGrow);
    __Pyx_GIVEREF(__pyx_v_carbonsToGrow);
    __pyx_t_1 = __Pyx_PyObject_Call(__pyx_builtin_range, __pyx_t_3, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
      __pyx_t_3 = __pyx_t_1; __Pyx_INCREF(__pyx_t_3); __pyx_t_15 = 0;
      __pyx_t_11 = NULL;
    } else {
      __pyx_t_15 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_11 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    for (;;) {
      if (likely(!__pyx_t_11)) {
        if (likely(PyList_CheckExact(__pyx_t_3))) {
          if (__pyx_t_15 >= PyList_GET_SIZE(__pyx_t_3)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_1 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_15); __Pyx_INCREF(__pyx_t_1); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_1 = PySequence_ITEM(__pyx_t_3, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        } else {
          if (__pyx_t_15 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
          #if CYTHON_COMPILING_IN_CPYTHON
          __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_15); __Pyx_INCREF(__pyx_t_1); __pyx_t_15++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #else
          __pyx_t_1 = PySequence_ITEM(__pyx_t_3, __pyx_t_15); __pyx_t_15++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          #endif
        }
      } else {
        __pyx_t_1 = __pyx_t_11(__pyx_t_3);
        if (unlikely(!__pyx_t_1)) {
          PyObject* exc_type = PyErr_Occurred();
          if (exc_type) {
            if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
            else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1669; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          }
          break;
        }
        __Pyx_GOTREF(__pyx_t_1);
      }
      __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_1);
      __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1670
 *             mergedRingDict[index] = []
 *             for x in range(carbonsToGrow):
 *                 if x ==0: lastAtom = ring[-1]             # <<<<<<<<<<<<<<
 *                 else: lastAtom = mergedRingDict[index][-1]
 *                 #add a new atom to the ring and the group
 */
      __pyx_t_1 = PyObject_RichCompare(__pyx_v_x, __pyx_int_0, Py_EQ); __Pyx_XGOTREF(__pyx_t_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1670; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1670; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      if (__pyx_t_13) {
        __pyx_t_1 = __Pyx_GetItemInt(__pyx_v_ring, -1, long, 1, __Pyx_PyInt_From_long, 0, 1, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1670; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_1);
        __Pyx_XDECREF_SET(__pyx_v_lastAtom, __pyx_t_1);
        __pyx_t_1 = 0;
        goto __pyx_L75;
      }
      /*else*/ {

        /* "rmgpy/molecule/group.py":1671
 *             for x in range(carbonsToGrow):
 *                 if x ==0: lastAtom = ring[-1]
 *                 else: lastAtom = mergedRingDict[index][-1]             # <<<<<<<<<<<<<<
 *                 #add a new atom to the ring and the group
 *                 (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, lastAtom)
 */
        __pyx_t_1 = __Pyx_PyDict_GetItem(__pyx_v_mergedRingDict, __pyx_v_index); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1671; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_1);
        __pyx_t_2 = __Pyx_GetItemInt(__pyx_t_1, -1, long, 1, __Pyx_PyInt_From_long, 0, 1, 1); if (unlikely(__pyx_t_2 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1671; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_XDECREF_SET(__pyx_v_lastAtom, __pyx_t_2);
        __pyx_t_2 = 0;
      }
      __pyx_L75:;

      /* "rmgpy/molecule/group.py":1673
 *                 else: lastAtom = mergedRingDict[index][-1]
 *                 #add a new atom to the ring and the group
 *                 (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, lastAtom)             # <<<<<<<<<<<<<<
 *                 mergedRingDict[index].append(newAtom)
 *                 #At the end attach to the other endpoint
 */
      __pyx_t_2 = PyTuple_New(2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1673; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_INCREF(__pyx_v_copyGroup);
      PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_copyGroup);
      __Pyx_GIVEREF(__pyx_v_copyGroup);
      __Pyx_INCREF(__pyx_v_lastAtom);
      PyTuple_SET_ITEM(__pyx_t_2, 1, __pyx_v_lastAtom);
      __Pyx_GIVEREF(__pyx_v_lastAtom);
      __pyx_t_1 = __Pyx_PyObject_Call(addBenzeneAtomToGroup, __pyx_t_2, NULL); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1673; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if ((likely(PyTuple_CheckExact(__pyx_t_1))) || (PyList_CheckExact(__pyx_t_1))) {
        PyObject* sequence = __pyx_t_1;
        #if CYTHON_COMPILING_IN_CPYTHON
        Py_ssize_t size = Py_SIZE(sequence);
        #else
        Py_ssize_t size = PySequence_Size(sequence);
        #endif
        if (unlikely(size != 2)) {
          if (size > 2) __Pyx_RaiseTooManyValuesError(2);
          else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
          {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1673; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        #if CYTHON_COMPILING_IN_CPYTHON
        if (likely(PyTuple_CheckExact(sequence))) {
          __pyx_t_2 = PyTuple_GET_ITEM(sequence, 0); 
          __pyx_t_5 = PyTuple_GET_ITEM(sequence, 1); 
        } else {
          __pyx_t_2 = PyList_GET_ITEM(sequence, 0); 
          __pyx_t_5 = PyList_GET_ITEM(sequence, 1); 
        }
        __Pyx_INCREF(__pyx_t_2);
        __Pyx_INCREF(__pyx_t_5);
        #else
        __pyx_t_2 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1673; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __pyx_t_5 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1673; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        #endif
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      } else {
        Py_ssize_t index = -1;
        __pyx_t_7 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1673; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_7);
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __pyx_t_8 = Py_TYPE(__pyx_t_7)->tp_iternext;
        index = 0; __pyx_t_2 = __pyx_t_8(__pyx_t_7); if (unlikely(!__pyx_t_2)) goto __pyx_L76_unpacking_failed;
        __Pyx_GOTREF(__pyx_t_2);
        index = 1; __pyx_t_5 = __pyx_t_8(__pyx_t_7); if (unlikely(!__pyx_t_5)) goto __pyx_L76_unpacking_failed;
        __Pyx_GOTREF(__pyx_t_5);
        if (__Pyx_IternextUnpackEndCheck(__pyx_t_8(__pyx_t_7), 2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1673; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __pyx_t_8 = NULL;
        __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
        goto __pyx_L77_unpacking_done;
        __pyx_L76_unpacking_failed:;
        __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
        __pyx_t_8 = NULL;
        if (__Pyx_IterFinish() == 0) __Pyx_RaiseNeedMoreValuesError(index);
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1673; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __pyx_L77_unpacking_done:;
      }
      __Pyx_DECREF_SET(__pyx_v_copyGroup, __pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_XDECREF_SET(__pyx_v_newAtom, __pyx_t_5);
      __pyx_t_5 = 0;

      /* "rmgpy/molecule/group.py":1674
 *                 #add a new atom to the ring and the group
 *                 (copyGroup, newAtom) = addBenzeneAtomToGroup(copyGroup, lastAtom)
 *                 mergedRingDict[index].append(newAtom)             # <<<<<<<<<<<<<<
 *                 #At the end attach to the other endpoint
 *                 if x == carbonsToGrow -1:
 */
      __pyx_t_1 = __Pyx_PyDict_GetItem(__pyx_v_mergedRingDict, __pyx_v_index); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1674; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_21 = __Pyx_PyObject_Append(__pyx_t_1, __pyx_v_newAtom); if (unlikely(__pyx_t_21 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1674; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

      /* "rmgpy/molecule/group.py":1676
 *                 mergedRingDict[index].append(newAtom)
 *                 #At the end attach to the other endpoint
 *                 if x == carbonsToGrow -1:             # <<<<<<<<<<<<<<
 *                     newBond = GroupBond(ring[0], newAtom, order=['B'])
 *                     copyGroup.addBond(newBond)
 */
      __pyx_t_1 = PyNumber_Subtract(__pyx_v_carbonsToGrow, __pyx_int_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1676; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      __pyx_t_5 = PyObject_RichCompare(__pyx_v_x, __pyx_t_1, Py_EQ); __Pyx_XGOTREF(__pyx_t_5); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1676; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __pyx_t_13 = __Pyx_PyObject_IsTrue(__pyx_t_5); if (unlikely(__pyx_t_13 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1676; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
      if (__pyx_t_13) {

        /* "rmgpy/molecule/group.py":1677
 *                 #At the end attach to the other endpoint
 *                 if x == carbonsToGrow -1:
 *                     newBond = GroupBond(ring[0], newAtom, order=['B'])             # <<<<<<<<<<<<<<
 *                     copyGroup.addBond(newBond)
 * 
 */
        __pyx_t_5 = __Pyx_GetItemInt(__pyx_v_ring, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_5 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1677; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
        __Pyx_GOTREF(__pyx_t_5);
        __pyx_t_1 = PyTuple_New(2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1677; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_1);
        PyTuple_SET_ITEM(__pyx_t_1, 0, __pyx_t_5);
        __Pyx_GIVEREF(__pyx_t_5);
        __Pyx_INCREF(__pyx_v_newAtom);
        PyTuple_SET_ITEM(__pyx_t_1, 1, __pyx_v_newAtom);
        __Pyx_GIVEREF(__pyx_v_newAtom);
        __pyx_t_5 = 0;
        __pyx_t_5 = PyDict_New(); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1677; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __pyx_t_2 = PyList_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1677; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_INCREF(__pyx_n_s_B);
        PyList_SET_ITEM(__pyx_t_2, 0, __pyx_n_s_B);
        __Pyx_GIVEREF(__pyx_n_s_B);
        if (PyDict_SetItem(__pyx_t_5, __pyx_n_s_order, __pyx_t_2) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1677; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        __pyx_t_2 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_GroupBond)), __pyx_t_1, __pyx_t_5); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1677; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        __Pyx_XDECREF_SET(__pyx_v_newBond, ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)__pyx_t_2));
        __pyx_t_2 = 0;

        /* "rmgpy/molecule/group.py":1678
 *                 if x == carbonsToGrow -1:
 *                     newBond = GroupBond(ring[0], newAtom, order=['B'])
 *                     copyGroup.addBond(newBond)             # <<<<<<<<<<<<<<
 * 
 *         return copyGroup
 */
        __pyx_t_5 = __Pyx_PyObject_GetAttrStr(__pyx_v_copyGroup, __pyx_n_s_addBond); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1678; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_5);
        __pyx_t_1 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_5))) {
          __pyx_t_1 = PyMethod_GET_SELF(__pyx_t_5);
          if (likely(__pyx_t_1)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_5);
            __Pyx_INCREF(__pyx_t_1);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_5, function);
          }
        }
        if (!__pyx_t_1) {
          __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_5, ((PyObject *)__pyx_v_newBond)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1678; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
        } else {
          __pyx_t_7 = PyTuple_New(1+1); if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1678; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_7);
          PyTuple_SET_ITEM(__pyx_t_7, 0, __pyx_t_1); __Pyx_GIVEREF(__pyx_t_1); __pyx_t_1 = NULL;
          __Pyx_INCREF(((PyObject *)__pyx_v_newBond));
          PyTuple_SET_ITEM(__pyx_t_7, 0+1, ((PyObject *)__pyx_v_newBond));
          __Pyx_GIVEREF(((PyObject *)__pyx_v_newBond));
          __pyx_t_2 = __Pyx_PyObject_Call(__pyx_t_5, __pyx_t_7, NULL); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1678; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_2);
          __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
        }
        __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
        __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
        goto __pyx_L78;
      }
      __pyx_L78:;

      /* "rmgpy/molecule/group.py":1669
 *             carbonsToGrow = 6-len(ring)
 *             mergedRingDict[index] = []
 *             for x in range(carbonsToGrow):             # <<<<<<<<<<<<<<
 *                 if x ==0: lastAtom = ring[-1]
 *                 else: lastAtom = mergedRingDict[index][-1]
 */
    }
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

    /* "rmgpy/molecule/group.py":1666
 *         """
 *         mergedRingDict={}
 *         for index, ring in enumerate(rings):             # <<<<<<<<<<<<<<
 *             carbonsToGrow = 6-len(ring)
 *             mergedRingDict[index] = []
 */
  }
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;

  /* "rmgpy/molecule/group.py":1680
 *                     copyGroup.addBond(newBond)
 * 
 *         return copyGroup             # <<<<<<<<<<<<<<
 * 
 *     def makeSampleMolecule(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  if (!(likely(((__pyx_v_copyGroup) == Py_None) || likely(__Pyx_TypeTest(__pyx_v_copyGroup, __pyx_ptype_5rmgpy_8molecule_5group_Group))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1680; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_INCREF(__pyx_v_copyGroup);
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_copyGroup);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1269
 *         return copyGroup
 * 
 *     def addImplicitBenzene(self):             # <<<<<<<<<<<<<<
 *         """
 *         Returns: A modified group with any implicit benzene rings added
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_XDECREF(__pyx_t_7);
  __Pyx_XDECREF(__pyx_t_20);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(classifyBenzeneCarbons);
  __Pyx_XDECREF(addBenzeneAtomToGroup);
  __Pyx_XDECREF(sortByConnectivity);
  __Pyx_XDECREF(checkSet);
  __Pyx_XDECREF(mergeOverlappingBenzeneRings);
  __Pyx_XDECREF(addCbAtomToRing);
  __Pyx_XDECREF(__pyx_v_copyGroup);
  __Pyx_XDECREF(__pyx_v_cbAtomList);
  __Pyx_XDECREF(__pyx_v_cbfAtomList);
  __Pyx_XDECREF(__pyx_v_cbfAtomList1);
  __Pyx_XDECREF(__pyx_v_cbfAtomList2);
  __Pyx_XDECREF(__pyx_v_connectedCbfs);
  __Pyx_XDECREF(__pyx_v_partners);
  __Pyx_XDECREF(__pyx_v_cbfAtom);
  __Pyx_XDECREF(__pyx_v_potentialPartner);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XDECREF(__pyx_v_bondsAreBenzene);
  __Pyx_XDECREF(__pyx_v_newAtom);
  __Pyx_XDECREF(__pyx_v_rings);
  __Pyx_XDECREF(__pyx_v_otherCbf);
  __Pyx_XDECREF(__pyx_v_newRingSeeds);
  __Pyx_XDECREF(__pyx_v_allLigands);
  __Pyx_XDECREF(__pyx_v_index1);
  __Pyx_XDECREF(__pyx_v_ring1);
  __Pyx_XDECREF(__pyx_v_mergeRingDict);
  __Pyx_XDECREF(__pyx_v_index);
  __Pyx_XDECREF(__pyx_v_ring2);
  __Pyx_XDECREF(__pyx_v_mergeRing);
  __Pyx_XDECREF(__pyx_v_newRingSeed);
  __Pyx_XDECREF(__pyx_v_inRing);
  __Pyx_XDECREF(__pyx_v_ring);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_XDECREF(__pyx_v_cbAtom);
  __Pyx_XDECREF(__pyx_v_mergedRingDict);
  __Pyx_XDECREF(__pyx_v_carbonsToGrow);
  __Pyx_XDECREF(__pyx_v_lastAtom);
  __Pyx_XDECREF((PyObject *)__pyx_v_newBond);
  __Pyx_XDECREF(__pyx_v_bond23);
  __Pyx_XDECREF(__pyx_v_cycle);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_69addImplicitBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_68addImplicitBenzene[] = "\n        Returns: A modified group with any implicit benzene rings added\n\n        This method currently does not if there are wildcards in atomtypes or bond orders\n        The current algorithm also requires that all Cb and Cbf are atomtyped\n\n        There are other cases where the algorithm doesn't work. For example whenever there\n        are many dangling Cb or Cbf atoms not in a ring, it is likely fail. In the database test\n        (the only use thus far), we will require that any group with more than 3 Cbfs have\n        complete rings. This is much stricter than this method can handle, but right now\n        this method cannot handle very general cases, so it is better to be conservative. \n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_69addImplicitBenzene(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("addImplicitBenzene (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_68addImplicitBenzene(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_68addImplicitBenzene(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("addImplicitBenzene", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_addImplicitBenzene(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1269; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.addImplicitBenzene", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1682
 *         return copyGroup
 * 
 *     def makeSampleMolecule(self):             # <<<<<<<<<<<<<<
 *         """
 *         Returns: A sample class :Molecule: from the group
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_71makeSampleMolecule(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *__pyx_f_5rmgpy_8molecule_5group_5Group_makeSampleMolecule(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_modifiedGroup = NULL;
  PyObject *__pyx_v_atomDict = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_newMolecule = NULL;
  PyObject *__pyx_v_atom1 = NULL;
  PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_v_statedCharge = NULL;
  PyObject *__pyx_v_hydrogenNeeded = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_x = NULL;
  PyObject *__pyx_v_newH = NULL;
  PyObject *__pyx_v_newBond = NULL;
  struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  Py_ssize_t __pyx_t_5;
  PyObject *(*__pyx_t_6)(PyObject *);
  Py_ssize_t __pyx_t_7;
  Py_ssize_t __pyx_t_8;
  int __pyx_t_9;
  int __pyx_t_10;
  PyObject *__pyx_t_11 = NULL;
  PyObject *__pyx_t_12 = NULL;
  PyObject *__pyx_t_13 = NULL;
  Py_ssize_t __pyx_t_14;
  PyObject *__pyx_t_15 = NULL;
  int __pyx_t_16;
  PyObject *(*__pyx_t_17)(PyObject *);
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("makeSampleMolecule", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_makeSampleMolecule); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1682; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_71makeSampleMolecule)) {
      __Pyx_XDECREF(((PyObject *)__pyx_r));
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1682; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1682; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5rmgpy_8molecule_8molecule_Molecule))))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1682; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *)__pyx_t_2);
      __pyx_t_2 = 0;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1691
 * 
 *         #Add implicit atoms
 *         modifiedGroup = self.addImplicitAtomsFromAtomType()             # <<<<<<<<<<<<<<
 *         #Add implicit benzene rings
 *         modifiedGroup = modifiedGroup.addImplicitBenzene()
 */
  __pyx_t_1 = ((PyObject *)((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->addImplicitAtomsFromAtomType(__pyx_v_self, 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1691; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_modifiedGroup = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1693
 *         modifiedGroup = self.addImplicitAtomsFromAtomType()
 *         #Add implicit benzene rings
 *         modifiedGroup = modifiedGroup.addImplicitBenzene()             # <<<<<<<<<<<<<<
 *         #Make dictionary of :GroupAtoms: to :Atoms:
 *         atomDict = {}
 */
  __pyx_t_1 = ((PyObject *)((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_modifiedGroup->__pyx_base.__pyx_vtab)->addImplicitBenzene(__pyx_v_modifiedGroup, 0)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1693; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF_SET(__pyx_v_modifiedGroup, ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_1));
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1695
 *         modifiedGroup = modifiedGroup.addImplicitBenzene()
 *         #Make dictionary of :GroupAtoms: to :Atoms:
 *         atomDict = {}             # <<<<<<<<<<<<<<
 *         for atom in modifiedGroup.atoms:
 *             atomDict[atom] = atom.makeSampleAtom()
 */
  __pyx_t_1 = PyDict_New(); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1695; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_atomDict = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1696
 *         #Make dictionary of :GroupAtoms: to :Atoms:
 *         atomDict = {}
 *         for atom in modifiedGroup.atoms:             # <<<<<<<<<<<<<<
 *             atomDict[atom] = atom.makeSampleAtom()
 * 
 */
  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_modifiedGroup), __pyx_n_s_atoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
    __pyx_t_2 = __pyx_t_1; __Pyx_INCREF(__pyx_t_2); __pyx_t_5 = 0;
    __pyx_t_6 = NULL;
  } else {
    __pyx_t_5 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (;;) {
    if (likely(!__pyx_t_6)) {
      if (likely(PyList_CheckExact(__pyx_t_2))) {
        if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_1); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_5 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_1); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_1 = __pyx_t_6(__pyx_t_2);
      if (unlikely(!__pyx_t_1)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1696; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_1);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1697
 *         atomDict = {}
 *         for atom in modifiedGroup.atoms:
 *             atomDict[atom] = atom.makeSampleAtom()             # <<<<<<<<<<<<<<
 * 
 *         #create the molecule
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_makeSampleAtom); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1697; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (__pyx_t_4) {
      __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1697; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    } else {
      __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1697; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    if (unlikely(PyDict_SetItem(__pyx_v_atomDict, __pyx_v_atom, __pyx_t_1) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1697; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1696
 *         #Make dictionary of :GroupAtoms: to :Atoms:
 *         atomDict = {}
 *         for atom in modifiedGroup.atoms:             # <<<<<<<<<<<<<<
 *             atomDict[atom] = atom.makeSampleAtom()
 * 
 */
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1700
 * 
 *         #create the molecule
 *         newMolecule = mol.Molecule(atoms = atomDict.values())             # <<<<<<<<<<<<<<
 *         #Add explicit bonds to :Atoms:
 *         for atom1 in modifiedGroup.atoms:
 */
  __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1700; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_1 = __Pyx_PyDict_Values(__pyx_v_atomDict); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1700; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_atoms, __pyx_t_1) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1700; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_1 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_8molecule_Molecule)), __pyx_empty_tuple, __pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1700; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_v_newMolecule = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1702
 *         newMolecule = mol.Molecule(atoms = atomDict.values())
 *         #Add explicit bonds to :Atoms:
 *         for atom1 in modifiedGroup.atoms:             # <<<<<<<<<<<<<<
 *             for atom2, bond12 in atom1.bonds.iteritems():
 *                 bond12.makeBond(newMolecule, atomDict[atom1], atomDict[atom2])
 */
  __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_modifiedGroup), __pyx_n_s_atoms); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  if (likely(PyList_CheckExact(__pyx_t_1)) || PyTuple_CheckExact(__pyx_t_1)) {
    __pyx_t_2 = __pyx_t_1; __Pyx_INCREF(__pyx_t_2); __pyx_t_5 = 0;
    __pyx_t_6 = NULL;
  } else {
    __pyx_t_5 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_6 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  for (;;) {
    if (likely(!__pyx_t_6)) {
      if (likely(PyList_CheckExact(__pyx_t_2))) {
        if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_1); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_5 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_1 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_5); __Pyx_INCREF(__pyx_t_1); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_1 = PySequence_ITEM(__pyx_t_2, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_1 = __pyx_t_6(__pyx_t_2);
      if (unlikely(!__pyx_t_1)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1702; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_1);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom1, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1703
 *         #Add explicit bonds to :Atoms:
 *         for atom1 in modifiedGroup.atoms:
 *             for atom2, bond12 in atom1.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                 bond12.makeBond(newMolecule, atomDict[atom1], atomDict[atom2])
 * 
 */
    __pyx_t_7 = 0;
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom1, __pyx_n_s_bonds); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1703; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    if (unlikely(__pyx_t_3 == Py_None)) {
      PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
      {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1703; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __pyx_t_4 = __Pyx_dict_iterator(__pyx_t_3, 0, __pyx_n_s_iteritems, (&__pyx_t_8), (&__pyx_t_9)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1703; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_XDECREF(__pyx_t_1);
    __pyx_t_1 = __pyx_t_4;
    __pyx_t_4 = 0;
    while (1) {
      __pyx_t_10 = __Pyx_dict_iter_next(__pyx_t_1, __pyx_t_8, &__pyx_t_7, &__pyx_t_4, &__pyx_t_3, NULL, __pyx_t_9);
      if (unlikely(__pyx_t_10 == 0)) break;
      if (unlikely(__pyx_t_10 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1703; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_4);
      __pyx_t_4 = 0;
      __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_3);
      __pyx_t_3 = 0;

      /* "rmgpy/molecule/group.py":1704
 *         for atom1 in modifiedGroup.atoms:
 *             for atom2, bond12 in atom1.bonds.iteritems():
 *                 bond12.makeBond(newMolecule, atomDict[atom1], atomDict[atom2])             # <<<<<<<<<<<<<<
 * 
 * 
 */
      __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_makeBond); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1704; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_11 = __Pyx_PyDict_GetItem(__pyx_v_atomDict, __pyx_v_atom1); if (unlikely(__pyx_t_11 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1704; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_11);
      __pyx_t_12 = __Pyx_PyDict_GetItem(__pyx_v_atomDict, __pyx_v_atom2); if (unlikely(__pyx_t_12 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1704; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
      __Pyx_GOTREF(__pyx_t_12);
      __pyx_t_13 = NULL;
      __pyx_t_14 = 0;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
        __pyx_t_13 = PyMethod_GET_SELF(__pyx_t_4);
        if (likely(__pyx_t_13)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
          __Pyx_INCREF(__pyx_t_13);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_4, function);
          __pyx_t_14 = 1;
        }
      }
      __pyx_t_15 = PyTuple_New(3+__pyx_t_14); if (unlikely(!__pyx_t_15)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1704; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_15);
      if (__pyx_t_13) {
        PyTuple_SET_ITEM(__pyx_t_15, 0, __pyx_t_13); __Pyx_GIVEREF(__pyx_t_13); __pyx_t_13 = NULL;
      }
      __Pyx_INCREF(__pyx_v_newMolecule);
      PyTuple_SET_ITEM(__pyx_t_15, 0+__pyx_t_14, __pyx_v_newMolecule);
      __Pyx_GIVEREF(__pyx_v_newMolecule);
      PyTuple_SET_ITEM(__pyx_t_15, 1+__pyx_t_14, __pyx_t_11);
      __Pyx_GIVEREF(__pyx_t_11);
      PyTuple_SET_ITEM(__pyx_t_15, 2+__pyx_t_14, __pyx_t_12);
      __Pyx_GIVEREF(__pyx_t_12);
      __pyx_t_11 = 0;
      __pyx_t_12 = 0;
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_15, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1704; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1702
 *         newMolecule = mol.Molecule(atoms = atomDict.values())
 *         #Add explicit bonds to :Atoms:
 *         for atom1 in modifiedGroup.atoms:             # <<<<<<<<<<<<<<
 *             for atom2, bond12 in atom1.bonds.iteritems():
 *                 bond12.makeBond(newMolecule, atomDict[atom1], atomDict[atom2])
 */
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1708
 * 
 *         #Saturate up to expected valency
 *         for atom in newMolecule.atoms:             # <<<<<<<<<<<<<<
 *             statedCharge = atom.charge
 *             atom.updateCharge()
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_newMolecule, __pyx_n_s_atoms); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1708; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_5 = 0;
    __pyx_t_6 = NULL;
  } else {
    __pyx_t_5 = -1; __pyx_t_1 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1708; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    __pyx_t_6 = Py_TYPE(__pyx_t_1)->tp_iternext; if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1708; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_6)) {
      if (likely(PyList_CheckExact(__pyx_t_1))) {
        if (__pyx_t_5 >= PyList_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1708; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1708; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_5 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_5); __Pyx_INCREF(__pyx_t_2); __pyx_t_5++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1708; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_5); __pyx_t_5++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1708; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_6(__pyx_t_1);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1708; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1709
 *         #Saturate up to expected valency
 *         for atom in newMolecule.atoms:
 *             statedCharge = atom.charge             # <<<<<<<<<<<<<<
 *             atom.updateCharge()
 *             if atom.charge - statedCharge:
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_charge); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1709; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_XDECREF_SET(__pyx_v_statedCharge, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1710
 *         for atom in newMolecule.atoms:
 *             statedCharge = atom.charge
 *             atom.updateCharge()             # <<<<<<<<<<<<<<
 *             if atom.charge - statedCharge:
 *                 hydrogenNeeded = atom.charge - statedCharge
 */
    __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_updateCharge); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1710; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_4 = NULL;
    if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
      __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
      if (likely(__pyx_t_4)) {
        PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
        __Pyx_INCREF(__pyx_t_4);
        __Pyx_INCREF(function);
        __Pyx_DECREF_SET(__pyx_t_3, function);
      }
    }
    if (__pyx_t_4) {
      __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1710; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    } else {
      __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1710; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1711
 *             statedCharge = atom.charge
 *             atom.updateCharge()
 *             if atom.charge - statedCharge:             # <<<<<<<<<<<<<<
 *                 hydrogenNeeded = atom.charge - statedCharge
 *                 for x in range(hydrogenNeeded):
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_charge); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1711; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_3 = PyNumber_Subtract(__pyx_t_2, __pyx_v_statedCharge); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1711; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_16 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_16 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1711; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    if (__pyx_t_16) {

      /* "rmgpy/molecule/group.py":1712
 *             atom.updateCharge()
 *             if atom.charge - statedCharge:
 *                 hydrogenNeeded = atom.charge - statedCharge             # <<<<<<<<<<<<<<
 *                 for x in range(hydrogenNeeded):
 *                     newH = mol.Atom('H', radicalElectrons=0, lonePairs=0, charge=0)
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_charge); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1712; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_2 = PyNumber_Subtract(__pyx_t_3, __pyx_v_statedCharge); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1712; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_XDECREF_SET(__pyx_v_hydrogenNeeded, __pyx_t_2);
      __pyx_t_2 = 0;

      /* "rmgpy/molecule/group.py":1713
 *             if atom.charge - statedCharge:
 *                 hydrogenNeeded = atom.charge - statedCharge
 *                 for x in range(hydrogenNeeded):             # <<<<<<<<<<<<<<
 *                     newH = mol.Atom('H', radicalElectrons=0, lonePairs=0, charge=0)
 *                     newBond = mol.Bond(atom, newH, 'S')
 */
      __pyx_t_2 = PyTuple_New(1); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_INCREF(__pyx_v_hydrogenNeeded);
      PyTuple_SET_ITEM(__pyx_t_2, 0, __pyx_v_hydrogenNeeded);
      __Pyx_GIVEREF(__pyx_v_hydrogenNeeded);
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_range, __pyx_t_2, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (likely(PyList_CheckExact(__pyx_t_3)) || PyTuple_CheckExact(__pyx_t_3)) {
        __pyx_t_2 = __pyx_t_3; __Pyx_INCREF(__pyx_t_2); __pyx_t_8 = 0;
        __pyx_t_17 = NULL;
      } else {
        __pyx_t_8 = -1; __pyx_t_2 = PyObject_GetIter(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_2);
        __pyx_t_17 = Py_TYPE(__pyx_t_2)->tp_iternext; if (unlikely(!__pyx_t_17)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      for (;;) {
        if (likely(!__pyx_t_17)) {
          if (likely(PyList_CheckExact(__pyx_t_2))) {
            if (__pyx_t_8 >= PyList_GET_SIZE(__pyx_t_2)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_3 = PyList_GET_ITEM(__pyx_t_2, __pyx_t_8); __Pyx_INCREF(__pyx_t_3); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          } else {
            if (__pyx_t_8 >= PyTuple_GET_SIZE(__pyx_t_2)) break;
            #if CYTHON_COMPILING_IN_CPYTHON
            __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_2, __pyx_t_8); __Pyx_INCREF(__pyx_t_3); __pyx_t_8++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #else
            __pyx_t_3 = PySequence_ITEM(__pyx_t_2, __pyx_t_8); __pyx_t_8++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            #endif
          }
        } else {
          __pyx_t_3 = __pyx_t_17(__pyx_t_2);
          if (unlikely(!__pyx_t_3)) {
            PyObject* exc_type = PyErr_Occurred();
            if (exc_type) {
              if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
              else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1713; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
            }
            break;
          }
          __Pyx_GOTREF(__pyx_t_3);
        }
        __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_3);
        __pyx_t_3 = 0;

        /* "rmgpy/molecule/group.py":1714
 *                 hydrogenNeeded = atom.charge - statedCharge
 *                 for x in range(hydrogenNeeded):
 *                     newH = mol.Atom('H', radicalElectrons=0, lonePairs=0, charge=0)             # <<<<<<<<<<<<<<
 *                     newBond = mol.Bond(atom, newH, 'S')
 *                     newMolecule.addAtom(newH)
 */
        __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_radicalElectrons, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_lonePairs, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        if (PyDict_SetItem(__pyx_t_3, __pyx_n_s_charge, __pyx_int_0) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_8molecule_Atom)), __pyx_tuple__19, __pyx_t_3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        __Pyx_XDECREF_SET(__pyx_v_newH, __pyx_t_4);
        __pyx_t_4 = 0;

        /* "rmgpy/molecule/group.py":1715
 *                 for x in range(hydrogenNeeded):
 *                     newH = mol.Atom('H', radicalElectrons=0, lonePairs=0, charge=0)
 *                     newBond = mol.Bond(atom, newH, 'S')             # <<<<<<<<<<<<<<
 *                     newMolecule.addAtom(newH)
 *                     newMolecule.addBond(newBond)
 */
        __pyx_t_4 = PyTuple_New(3); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1715; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_INCREF(__pyx_v_atom);
        PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_v_atom);
        __Pyx_GIVEREF(__pyx_v_atom);
        __Pyx_INCREF(__pyx_v_newH);
        PyTuple_SET_ITEM(__pyx_t_4, 1, __pyx_v_newH);
        __Pyx_GIVEREF(__pyx_v_newH);
        __Pyx_INCREF(__pyx_n_s_S);
        PyTuple_SET_ITEM(__pyx_t_4, 2, __pyx_n_s_S);
        __Pyx_GIVEREF(__pyx_n_s_S);
        __pyx_t_3 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_8molecule_Bond)), __pyx_t_4, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1715; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_XDECREF_SET(__pyx_v_newBond, __pyx_t_3);
        __pyx_t_3 = 0;

        /* "rmgpy/molecule/group.py":1716
 *                     newH = mol.Atom('H', radicalElectrons=0, lonePairs=0, charge=0)
 *                     newBond = mol.Bond(atom, newH, 'S')
 *                     newMolecule.addAtom(newH)             # <<<<<<<<<<<<<<
 *                     newMolecule.addBond(newBond)
 *                 atom.updateCharge()
 */
        __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_newMolecule, __pyx_n_s_addAtom); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1716; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_15 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
          __pyx_t_15 = PyMethod_GET_SELF(__pyx_t_4);
          if (likely(__pyx_t_15)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_15);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_4, function);
          }
        }
        if (!__pyx_t_15) {
          __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_v_newH); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1716; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
        } else {
          __pyx_t_12 = PyTuple_New(1+1); if (unlikely(!__pyx_t_12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1716; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_12);
          PyTuple_SET_ITEM(__pyx_t_12, 0, __pyx_t_15); __Pyx_GIVEREF(__pyx_t_15); __pyx_t_15 = NULL;
          __Pyx_INCREF(__pyx_v_newH);
          PyTuple_SET_ITEM(__pyx_t_12, 0+1, __pyx_v_newH);
          __Pyx_GIVEREF(__pyx_v_newH);
          __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_12, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1716; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_12); __pyx_t_12 = 0;
        }
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

        /* "rmgpy/molecule/group.py":1717
 *                     newBond = mol.Bond(atom, newH, 'S')
 *                     newMolecule.addAtom(newH)
 *                     newMolecule.addBond(newBond)             # <<<<<<<<<<<<<<
 *                 atom.updateCharge()
 *             # print statedCharge, atom.charge, type(atom.charge)
 */
        __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_newMolecule, __pyx_n_s_addBond); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1717; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_12 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
          __pyx_t_12 = PyMethod_GET_SELF(__pyx_t_4);
          if (likely(__pyx_t_12)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_12);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_4, function);
          }
        }
        if (!__pyx_t_12) {
          __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_v_newBond); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1717; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
        } else {
          __pyx_t_15 = PyTuple_New(1+1); if (unlikely(!__pyx_t_15)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1717; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_15);
          PyTuple_SET_ITEM(__pyx_t_15, 0, __pyx_t_12); __Pyx_GIVEREF(__pyx_t_12); __pyx_t_12 = NULL;
          __Pyx_INCREF(__pyx_v_newBond);
          PyTuple_SET_ITEM(__pyx_t_15, 0+1, __pyx_v_newBond);
          __Pyx_GIVEREF(__pyx_v_newBond);
          __pyx_t_3 = __Pyx_PyObject_Call(__pyx_t_4, __pyx_t_15, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1717; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_3);
          __Pyx_DECREF(__pyx_t_15); __pyx_t_15 = 0;
        }
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

        /* "rmgpy/molecule/group.py":1713
 *             if atom.charge - statedCharge:
 *                 hydrogenNeeded = atom.charge - statedCharge
 *                 for x in range(hydrogenNeeded):             # <<<<<<<<<<<<<<
 *                     newH = mol.Atom('H', radicalElectrons=0, lonePairs=0, charge=0)
 *                     newBond = mol.Bond(atom, newH, 'S')
 */
      }
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

      /* "rmgpy/molecule/group.py":1718
 *                     newMolecule.addAtom(newH)
 *                     newMolecule.addBond(newBond)
 *                 atom.updateCharge()             # <<<<<<<<<<<<<<
 *             # print statedCharge, atom.charge, type(atom.charge)
 * 
 */
      __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_updateCharge); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1718; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      goto __pyx_L11;
    }
    __pyx_L11:;

    /* "rmgpy/molecule/group.py":1708
 * 
 *         #Saturate up to expected valency
 *         for atom in newMolecule.atoms:             # <<<<<<<<<<<<<<
 *             statedCharge = atom.charge
 *             atom.updateCharge()
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1721
 *             # print statedCharge, atom.charge, type(atom.charge)
 * 
 *         newMolecule.update()             # <<<<<<<<<<<<<<
 *         # print newMolecule.atoms
 * 
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_newMolecule, __pyx_n_s_update); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1721; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = NULL;
  if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_2))) {
    __pyx_t_3 = PyMethod_GET_SELF(__pyx_t_2);
    if (likely(__pyx_t_3)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_2);
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_2, function);
    }
  }
  if (__pyx_t_3) {
    __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_2, __pyx_t_3); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1721; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  } else {
    __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_2); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1721; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1725
 * 
 * 
 *         return newMolecule             # <<<<<<<<<<<<<<
 * 
 *     def isBenzeneExplicit(self):
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  if (!(likely(__Pyx_TypeTest(__pyx_v_newMolecule, __pyx_ptype_5rmgpy_8molecule_8molecule_Molecule)))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1725; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_INCREF(__pyx_v_newMolecule);
  __pyx_r = ((struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *)__pyx_v_newMolecule);
  goto __pyx_L0;

  /* "rmgpy/molecule/group.py":1682
 *         return copyGroup
 * 
 *     def makeSampleMolecule(self):             # <<<<<<<<<<<<<<
 *         """
 *         Returns: A sample class :Molecule: from the group
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_11);
  __Pyx_XDECREF(__pyx_t_12);
  __Pyx_XDECREF(__pyx_t_13);
  __Pyx_XDECREF(__pyx_t_15);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.makeSampleMolecule", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_modifiedGroup);
  __Pyx_XDECREF(__pyx_v_atomDict);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_newMolecule);
  __Pyx_XDECREF(__pyx_v_atom1);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XDECREF(__pyx_v_statedCharge);
  __Pyx_XDECREF(__pyx_v_hydrogenNeeded);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_XDECREF(__pyx_v_newH);
  __Pyx_XDECREF(__pyx_v_newBond);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_71makeSampleMolecule(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_70makeSampleMolecule[] = "\n        Returns: A sample class :Molecule: from the group\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_71makeSampleMolecule(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("makeSampleMolecule (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_70makeSampleMolecule(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_70makeSampleMolecule(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("makeSampleMolecule", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = ((PyObject *)__pyx_f_5rmgpy_8molecule_5group_5Group_makeSampleMolecule(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1682; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.makeSampleMolecule", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.py":1727
 *         return newMolecule
 * 
 *     def isBenzeneExplicit(self):             # <<<<<<<<<<<<<<
 *         """
 * 
 */

static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_73isBenzeneExplicit(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static int __pyx_f_5rmgpy_8molecule_5group_5Group_isBenzeneExplicit(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, int __pyx_skip_dispatch) {
  PyObject *__pyx_v_cbAtomList = NULL;
  PyObject *__pyx_v_labelsOfCarbonAtomTypes = NULL;
  PyObject *__pyx_v_atom = NULL;
  PyObject *__pyx_v_benzeneBonds = NULL;
  CYTHON_UNUSED PyObject *__pyx_v_atom2 = NULL;
  PyObject *__pyx_v_bond12 = NULL;
  PyObject *__pyx_v_rings = NULL;
  int __pyx_v_inRing;
  PyObject *__pyx_v_ring = NULL;
  PyObject *__pyx_v_x = NULL;
  PyObject *__pyx_v_cycle = NULL;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  Py_ssize_t __pyx_t_6;
  PyObject *(*__pyx_t_7)(PyObject *);
  int __pyx_t_8;
  int __pyx_t_9;
  Py_ssize_t __pyx_t_10;
  Py_ssize_t __pyx_t_11;
  int __pyx_t_12;
  int __pyx_t_13;
  PyObject *__pyx_t_14 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isBenzeneExplicit", 0);
  /* Check if called by wrapper */
  if (unlikely(__pyx_skip_dispatch)) ;
  /* Check if overridden in Python */
  else if (unlikely(Py_TYPE(((PyObject *)__pyx_v_self))->tp_dictoffset != 0)) {
    __pyx_t_1 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_isBenzeneExplicit); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_1);
    if (!PyCFunction_Check(__pyx_t_1) || (PyCFunction_GET_FUNCTION(__pyx_t_1) != (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_73isBenzeneExplicit)) {
      __Pyx_INCREF(__pyx_t_1);
      __pyx_t_3 = __pyx_t_1; __pyx_t_4 = NULL;
      if (CYTHON_COMPILING_IN_CPYTHON && unlikely(PyMethod_Check(__pyx_t_3))) {
        __pyx_t_4 = PyMethod_GET_SELF(__pyx_t_3);
        if (likely(__pyx_t_4)) {
          PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
          __Pyx_INCREF(__pyx_t_4);
          __Pyx_INCREF(function);
          __Pyx_DECREF_SET(__pyx_t_3, function);
        }
      }
      if (__pyx_t_4) {
        __pyx_t_2 = __Pyx_PyObject_CallOneArg(__pyx_t_3, __pyx_t_4); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      } else {
        __pyx_t_2 = __Pyx_PyObject_CallNoArg(__pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __Pyx_GOTREF(__pyx_t_2);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      __pyx_r = __pyx_t_5;
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      goto __pyx_L0;
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  }

  /* "rmgpy/molecule/group.py":1737
 * 
 *         #classify atoms
 *         cbAtomList = []             # <<<<<<<<<<<<<<
 * 
 *         #only want to work with carbon atoms
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1737; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_cbAtomList = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "rmgpy/molecule/group.py":1740
 * 
 *         #only want to work with carbon atoms
 *         labelsOfCarbonAtomTypes = [x.label for x in atomTypes['C'].specific] + ['C']             # <<<<<<<<<<<<<<
 * 
 *         for atom in self.atoms:
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_t_2 = __Pyx_GetModuleGlobalName(__pyx_n_s_atomTypes); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = PyObject_GetItem(__pyx_t_2, __pyx_n_s_C); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_3, __pyx_n_s_specific); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_3 = __pyx_t_2; __Pyx_INCREF(__pyx_t_3); __pyx_t_6 = 0;
    __pyx_t_7 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_7 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_7)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_7(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_x, __pyx_t_2);
    __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_x, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    if (unlikely(__Pyx_ListComp_Append(__pyx_t_1, (PyObject*)__pyx_t_2))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_n_s_C);
  PyList_SET_ITEM(__pyx_t_3, 0, __pyx_n_s_C);
  __Pyx_GIVEREF(__pyx_n_s_C);
  __pyx_t_2 = PyNumber_Add(__pyx_t_1, __pyx_t_3); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1740; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_v_labelsOfCarbonAtomTypes = ((PyObject*)__pyx_t_2);
  __pyx_t_2 = 0;

  /* "rmgpy/molecule/group.py":1742
 *         labelsOfCarbonAtomTypes = [x.label for x in atomTypes['C'].specific] + ['C']
 * 
 *         for atom in self.atoms:             # <<<<<<<<<<<<<<
 *             if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue
 *             elif atom.atomType[0].label in ['Cb', 'Cbf']: #Make Cb and N3b into normal cb atoms
 */
  __pyx_t_2 = __Pyx_PyObject_GetAttrStr(((PyObject *)__pyx_v_self), __pyx_n_s_atoms); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1742; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (likely(PyList_CheckExact(__pyx_t_2)) || PyTuple_CheckExact(__pyx_t_2)) {
    __pyx_t_3 = __pyx_t_2; __Pyx_INCREF(__pyx_t_3); __pyx_t_6 = 0;
    __pyx_t_7 = NULL;
  } else {
    __pyx_t_6 = -1; __pyx_t_3 = PyObject_GetIter(__pyx_t_2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1742; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __pyx_t_7 = Py_TYPE(__pyx_t_3)->tp_iternext; if (unlikely(!__pyx_t_7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1742; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (likely(!__pyx_t_7)) {
      if (likely(PyList_CheckExact(__pyx_t_3))) {
        if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1742; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1742; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      } else {
        if (__pyx_t_6 >= PyTuple_GET_SIZE(__pyx_t_3)) break;
        #if CYTHON_COMPILING_IN_CPYTHON
        __pyx_t_2 = PyTuple_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1742; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #else
        __pyx_t_2 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1742; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        #endif
      }
    } else {
      __pyx_t_2 = __pyx_t_7(__pyx_t_3);
      if (unlikely(!__pyx_t_2)) {
        PyObject* exc_type = PyErr_Occurred();
        if (exc_type) {
          if (likely(exc_type == PyExc_StopIteration || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration))) PyErr_Clear();
          else {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1742; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        break;
      }
      __Pyx_GOTREF(__pyx_t_2);
    }
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_2);
    __pyx_t_2 = 0;

    /* "rmgpy/molecule/group.py":1743
 * 
 *         for atom in self.atoms:
 *             if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue             # <<<<<<<<<<<<<<
 *             elif atom.atomType[0].label in ['Cb', 'Cbf']: #Make Cb and N3b into normal cb atoms
 *                 cbAtomList.append(atom)
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1743; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_t_2, __pyx_v_labelsOfCarbonAtomTypes, Py_NE)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1743; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_8 = (__pyx_t_5 != 0);
    if (__pyx_t_8) {
      goto __pyx_L5_continue;
    }

    /* "rmgpy/molecule/group.py":1744
 *         for atom in self.atoms:
 *             if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue
 *             elif atom.atomType[0].label in ['Cb', 'Cbf']: #Make Cb and N3b into normal cb atoms             # <<<<<<<<<<<<<<
 *                 cbAtomList.append(atom)
 *             else:
 */
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_atomType); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1744; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __pyx_t_1 = __Pyx_GetItemInt(__pyx_t_2, 0, long, 1, __Pyx_PyInt_From_long, 0, 0, 1); if (unlikely(__pyx_t_1 == NULL)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1744; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_2 = __Pyx_PyObject_GetAttrStr(__pyx_t_1, __pyx_n_s_label); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1744; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
    __pyx_t_5 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_n_s_Cb, Py_EQ)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1744; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (!__pyx_t_5) {
      goto __pyx_L9_next_or;
    } else {
      __pyx_t_8 = __pyx_t_5;
      goto __pyx_L8_bool_binop_done;
    }
    __pyx_L9_next_or:;
    __pyx_t_5 = (__Pyx_PyString_Equals(__pyx_t_2, __pyx_n_s_Cbf, Py_EQ)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1744; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_8 = __pyx_t_5;
    __pyx_L8_bool_binop_done:;
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_5 = (__pyx_t_8 != 0);
    if (__pyx_t_5) {

      /* "rmgpy/molecule/group.py":1745
 *             if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue
 *             elif atom.atomType[0].label in ['Cb', 'Cbf']: #Make Cb and N3b into normal cb atoms
 *                 cbAtomList.append(atom)             # <<<<<<<<<<<<<<
 *             else:
 *                 benzeneBonds = 0
 */
      __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_cbAtomList, __pyx_v_atom); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1745; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L7;
    }
    /*else*/ {

      /* "rmgpy/molecule/group.py":1747
 *                 cbAtomList.append(atom)
 *             else:
 *                 benzeneBonds = 0             # <<<<<<<<<<<<<<
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isBenzene(): benzeneBonds+=1
 */
      __Pyx_INCREF(__pyx_int_0);
      __Pyx_XDECREF_SET(__pyx_v_benzeneBonds, __pyx_int_0);

      /* "rmgpy/molecule/group.py":1748
 *             else:
 *                 benzeneBonds = 0
 *                 for atom2, bond12 in atom.bonds.iteritems():             # <<<<<<<<<<<<<<
 *                     if bond12.isBenzene(): benzeneBonds+=1
 *                 if benzeneBonds >0: cbAtomList.append(atom)
 */
      __pyx_t_10 = 0;
      __pyx_t_1 = __Pyx_PyObject_GetAttrStr(__pyx_v_atom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1748; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_1);
      if (unlikely(__pyx_t_1 == Py_None)) {
        PyErr_Format(PyExc_AttributeError, "'NoneType' object has no attribute '%s'", "iteritems");
        {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1748; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_t_4 = __Pyx_dict_iterator(__pyx_t_1, 0, __pyx_n_s_iteritems, (&__pyx_t_11), (&__pyx_t_12)); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1748; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
      __Pyx_XDECREF(__pyx_t_2);
      __pyx_t_2 = __pyx_t_4;
      __pyx_t_4 = 0;
      while (1) {
        __pyx_t_13 = __Pyx_dict_iter_next(__pyx_t_2, __pyx_t_11, &__pyx_t_10, &__pyx_t_4, &__pyx_t_1, NULL, __pyx_t_12);
        if (unlikely(__pyx_t_13 == 0)) break;
        if (unlikely(__pyx_t_13 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1748; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_GOTREF(__pyx_t_1);
        __Pyx_XDECREF_SET(__pyx_v_atom2, __pyx_t_4);
        __pyx_t_4 = 0;
        __Pyx_XDECREF_SET(__pyx_v_bond12, __pyx_t_1);
        __pyx_t_1 = 0;

        /* "rmgpy/molecule/group.py":1749
 *                 benzeneBonds = 0
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isBenzene(): benzeneBonds+=1             # <<<<<<<<<<<<<<
 *                 if benzeneBonds >0: cbAtomList.append(atom)
 * 
 */
        __pyx_t_4 = __Pyx_PyObject_GetAttrStr(__pyx_v_bond12, __pyx_n_s_isBenzene); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __pyx_t_14 = NULL;
        if (CYTHON_COMPILING_IN_CPYTHON && likely(PyMethod_Check(__pyx_t_4))) {
          __pyx_t_14 = PyMethod_GET_SELF(__pyx_t_4);
          if (likely(__pyx_t_14)) {
            PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_4);
            __Pyx_INCREF(__pyx_t_14);
            __Pyx_INCREF(function);
            __Pyx_DECREF_SET(__pyx_t_4, function);
          }
        }
        if (__pyx_t_14) {
          __pyx_t_1 = __Pyx_PyObject_CallOneArg(__pyx_t_4, __pyx_t_14); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_DECREF(__pyx_t_14); __pyx_t_14 = 0;
        } else {
          __pyx_t_1 = __Pyx_PyObject_CallNoArg(__pyx_t_4); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        }
        __Pyx_GOTREF(__pyx_t_1);
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_1); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
        if (__pyx_t_5) {
          __pyx_t_1 = PyNumber_InPlaceAdd(__pyx_v_benzeneBonds, __pyx_int_1); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1749; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
          __Pyx_GOTREF(__pyx_t_1);
          __Pyx_DECREF_SET(__pyx_v_benzeneBonds, __pyx_t_1);
          __pyx_t_1 = 0;
          goto __pyx_L12;
        }
        __pyx_L12:;
      }
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;

      /* "rmgpy/molecule/group.py":1750
 *                 for atom2, bond12 in atom.bonds.iteritems():
 *                     if bond12.isBenzene(): benzeneBonds+=1
 *                 if benzeneBonds >0: cbAtomList.append(atom)             # <<<<<<<<<<<<<<
 * 
 *         #get all explicit benzene rings
 */
      __pyx_t_2 = PyObject_RichCompare(__pyx_v_benzeneBonds, __pyx_int_0, Py_GT); __Pyx_XGOTREF(__pyx_t_2); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1750; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_5 = __Pyx_PyObject_IsTrue(__pyx_t_2); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1750; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
      if (__pyx_t_5) {
        __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_cbAtomList, __pyx_v_atom); if (unlikely(__pyx_t_9 == -1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1750; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        goto __pyx_L13;
      }
      __pyx_L13:;
    }
    __pyx_L7:;

    /* "rmgpy/molecule/group.py":1742
 *         labelsOfCarbonAtomTypes = [x.label for x in atomTypes['C'].specific] + ['C']
 * 
 *         for atom in self.atoms:             # <<<<<<<<<<<<<<
 *             if not atom.atomType[0].label in labelsOfCarbonAtomTypes: continue
 *             elif atom.atomType[0].label in ['Cb', 'Cbf']: #Make Cb and N3b into normal cb atoms
 */
    __pyx_L5_continue:;
  }
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":1753
 * 
 *         #get all explicit benzene rings
 *         rings=[cycle for cycle in self.getAllCyclesOfSize(6) if Group(atoms = cycle).isAromaticRing()]             # <<<<<<<<<<<<<<
 * 
 *         #test that all benzene atoms are in benzene rings
 */
  __pyx_t_3 = PyList_New(0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_2 = ((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)__pyx_v_self->__pyx_base.__pyx_vtab)->__pyx_base.getAllCyclesOfSize(((struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *)__pyx_v_self), 6, 0); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_2);
  if (unlikely(__pyx_t_2 == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_t_2; __Pyx_INCREF(__pyx_t_1); __pyx_t_6 = 0;
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_2 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_6); __Pyx_INCREF(__pyx_t_2); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_2 = PySequence_ITEM(__pyx_t_1, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_cycle, __pyx_t_2);
    __pyx_t_2 = 0;
    __pyx_t_2 = PyDict_New(); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_2);
    if (PyDict_SetItem(__pyx_t_2, __pyx_n_s_atoms, __pyx_v_cycle) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_t_4 = __Pyx_PyObject_Call(((PyObject *)((PyObject*)__pyx_ptype_5rmgpy_8molecule_5group_Group)), __pyx_empty_tuple, __pyx_t_2); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
    __pyx_t_5 = (((struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group *)((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_4)->__pyx_base.__pyx_vtab)->isAromaticRing(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_t_4), 0) != 0);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    if (__pyx_t_5) {
      if (unlikely(__Pyx_ListComp_Append(__pyx_t_3, (PyObject*)__pyx_v_cycle))) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1753; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      goto __pyx_L16;
    }
    __pyx_L16:;
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_v_rings = ((PyObject*)__pyx_t_3);
  __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":1756
 * 
 *         #test that all benzene atoms are in benzene rings
 *         for atom in cbAtomList:             # <<<<<<<<<<<<<<
 *             inRing = False
 *             for ring in rings:
 */
  __pyx_t_3 = __pyx_v_cbAtomList; __Pyx_INCREF(__pyx_t_3); __pyx_t_6 = 0;
  for (;;) {
    if (__pyx_t_6 >= PyList_GET_SIZE(__pyx_t_3)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_1 = PyList_GET_ITEM(__pyx_t_3, __pyx_t_6); __Pyx_INCREF(__pyx_t_1); __pyx_t_6++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1756; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_1 = PySequence_ITEM(__pyx_t_3, __pyx_t_6); __pyx_t_6++; if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1756; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_atom, __pyx_t_1);
    __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1757
 *         #test that all benzene atoms are in benzene rings
 *         for atom in cbAtomList:
 *             inRing = False             # <<<<<<<<<<<<<<
 *             for ring in rings:
 *                 if atom in ring: inRing = True
 */
    __pyx_v_inRing = 0;

    /* "rmgpy/molecule/group.py":1758
 *         for atom in cbAtomList:
 *             inRing = False
 *             for ring in rings:             # <<<<<<<<<<<<<<
 *                 if atom in ring: inRing = True
 *             if not inRing: return False
 */
    __pyx_t_1 = __pyx_v_rings; __Pyx_INCREF(__pyx_t_1); __pyx_t_11 = 0;
    for (;;) {
      if (__pyx_t_11 >= PyList_GET_SIZE(__pyx_t_1)) break;
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_4 = PyList_GET_ITEM(__pyx_t_1, __pyx_t_11); __Pyx_INCREF(__pyx_t_4); __pyx_t_11++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1758; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #else
      __pyx_t_4 = PySequence_ITEM(__pyx_t_1, __pyx_t_11); __pyx_t_11++; if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1758; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      #endif
      __Pyx_XDECREF_SET(__pyx_v_ring, __pyx_t_4);
      __pyx_t_4 = 0;

      /* "rmgpy/molecule/group.py":1759
 *             inRing = False
 *             for ring in rings:
 *                 if atom in ring: inRing = True             # <<<<<<<<<<<<<<
 *             if not inRing: return False
 *         else: return True
 */
      __pyx_t_5 = (__Pyx_PySequence_Contains(__pyx_v_atom, __pyx_v_ring, Py_EQ)); if (unlikely(__pyx_t_5 < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1759; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_t_8 = (__pyx_t_5 != 0);
      if (__pyx_t_8) {
        __pyx_v_inRing = 1;
        goto __pyx_L21;
      }
      __pyx_L21:;

      /* "rmgpy/molecule/group.py":1758
 *         for atom in cbAtomList:
 *             inRing = False
 *             for ring in rings:             # <<<<<<<<<<<<<<
 *                 if atom in ring: inRing = True
 *             if not inRing: return False
 */
    }
    __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

    /* "rmgpy/molecule/group.py":1760
 *             for ring in rings:
 *                 if atom in ring: inRing = True
 *             if not inRing: return False             # <<<<<<<<<<<<<<
 *         else: return True
 */
    __pyx_t_8 = ((!(__pyx_v_inRing != 0)) != 0);
    if (__pyx_t_8) {
      __pyx_r = 0;
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      goto __pyx_L0;
    }

    /* "rmgpy/molecule/group.py":1756
 * 
 *         #test that all benzene atoms are in benzene rings
 *         for atom in cbAtomList:             # <<<<<<<<<<<<<<
 *             inRing = False
 *             for ring in rings:
 */
  }
  /*else*/ {

    /* "rmgpy/molecule/group.py":1761
 *                 if atom in ring: inRing = True
 *             if not inRing: return False
 *         else: return True             # <<<<<<<<<<<<<<
 */
    __pyx_r = 1;
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    goto __pyx_L0;
  }

  /* "rmgpy/molecule/group.py":1756
 * 
 *         #test that all benzene atoms are in benzene rings
 *         for atom in cbAtomList:             # <<<<<<<<<<<<<<
 *             inRing = False
 *             for ring in rings:
 */
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":1727
 *         return newMolecule
 * 
 *     def isBenzeneExplicit(self):             # <<<<<<<<<<<<<<
 *         """
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_14);
  __Pyx_WriteUnraisable("rmgpy.molecule.group.Group.isBenzeneExplicit", __pyx_clineno, __pyx_lineno, __pyx_filename, 0);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF(__pyx_v_cbAtomList);
  __Pyx_XDECREF(__pyx_v_labelsOfCarbonAtomTypes);
  __Pyx_XDECREF(__pyx_v_atom);
  __Pyx_XDECREF(__pyx_v_benzeneBonds);
  __Pyx_XDECREF(__pyx_v_atom2);
  __Pyx_XDECREF(__pyx_v_bond12);
  __Pyx_XDECREF(__pyx_v_rings);
  __Pyx_XDECREF(__pyx_v_ring);
  __Pyx_XDECREF(__pyx_v_x);
  __Pyx_XDECREF(__pyx_v_cycle);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_73isBenzeneExplicit(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused); /*proto*/
static char __pyx_doc_5rmgpy_8molecule_5group_5Group_72isBenzeneExplicit[] = "\n\n        Returns: 'True' if all Cb, Cbf atoms are in completely explicitly stated benzene rings.\n\n        Otherwise return 'False'\n\n        ";
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_73isBenzeneExplicit(PyObject *__pyx_v_self, CYTHON_UNUSED PyObject *unused) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("isBenzeneExplicit (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_72isBenzeneExplicit(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_72isBenzeneExplicit(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("isBenzeneExplicit", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyBool_FromLong(__pyx_f_5rmgpy_8molecule_5group_5Group_isBenzeneExplicit(__pyx_v_self, 1)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1727; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.isBenzeneExplicit", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":102
 * cdef class Group(Graph):
 * 
 *     cdef public list multiplicity             # <<<<<<<<<<<<<<
 * 
 *     # These read-only attribues act as a "fingerprint" for accelerating
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __Pyx_INCREF(__pyx_v_self->multiplicity);
  __pyx_r = __pyx_v_self->multiplicity;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  if (!(likely(PyList_CheckExact(__pyx_v_value))||((__pyx_v_value) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "list", Py_TYPE(__pyx_v_value)->tp_name), 0))) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 102; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_t_1 = __pyx_v_value;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->multiplicity);
  __Pyx_DECREF(__pyx_v_self->multiplicity);
  __pyx_v_self->multiplicity = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.multiplicity.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_5__del__(PyObject *__pyx_v_self); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_5__del__(PyObject *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_4__del__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12multiplicity_4__del__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__del__", 0);
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_self->multiplicity);
  __Pyx_DECREF(__pyx_v_self->multiplicity);
  __pyx_v_self->multiplicity = ((PyObject*)Py_None);

  /* function exit code */
  __pyx_r = 0;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":106
 *     # These read-only attribues act as a "fingerprint" for accelerating
 *     # subgraph isomorphism checks
 *     cdef public short carbonCount             # <<<<<<<<<<<<<<
 *     cdef public short nitrogenCount
 *     cdef public short oxygenCount
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->carbonCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 106; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.carbonCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11carbonCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 106; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->carbonCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.carbonCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":107
 *     # subgraph isomorphism checks
 *     cdef public short carbonCount
 *     cdef public short nitrogenCount             # <<<<<<<<<<<<<<
 *     cdef public short oxygenCount
 *     cdef public short sulfurCount
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->nitrogenCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 107; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.nitrogenCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_13nitrogenCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 107; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->nitrogenCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.nitrogenCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":108
 *     cdef public short carbonCount
 *     cdef public short nitrogenCount
 *     cdef public short oxygenCount             # <<<<<<<<<<<<<<
 *     cdef public short sulfurCount
 *     cdef public short radicalCount
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->oxygenCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 108; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.oxygenCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11oxygenCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 108; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->oxygenCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.oxygenCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":109
 *     cdef public short nitrogenCount
 *     cdef public short oxygenCount
 *     cdef public short sulfurCount             # <<<<<<<<<<<<<<
 *     cdef public short radicalCount
 * 
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->sulfurCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 109; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.sulfurCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_11sulfurCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 109; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->sulfurCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.sulfurCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "rmgpy/molecule/group.pxd":110
 *     cdef public short oxygenCount
 *     cdef public short sulfurCount
 *     cdef public short radicalCount             # <<<<<<<<<<<<<<
 * 
 *     cpdef addAtom(self, GroupAtom atom)
 */

/* Python wrapper */
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_1__get__(PyObject *__pyx_v_self); /*proto*/
static PyObject *__pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_1__get__(PyObject *__pyx_v_self) {
  PyObject *__pyx_r = 0;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__get__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount___get__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static PyObject *__pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount___get__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__get__", 0);
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = __Pyx_PyInt_From_short(__pyx_v_self->radicalCount); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 110; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.radicalCount.__get__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* Python wrapper */
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value); /*proto*/
static int __pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_3__set__(PyObject *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__set__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount_2__set__(((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)__pyx_v_self), ((PyObject *)__pyx_v_value));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5rmgpy_8molecule_5group_5Group_12radicalCount_2__set__(struct __pyx_obj_5rmgpy_8molecule_5group_Group *__pyx_v_self, PyObject *__pyx_v_value) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  short __pyx_t_1;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__set__", 0);
  __pyx_t_1 = __Pyx_PyInt_As_short(__pyx_v_value); if (unlikely((__pyx_t_1 == (short)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[1]; __pyx_lineno = 110; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_v_self->radicalCount = __pyx_t_1;

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_AddTraceback("rmgpy.molecule.group.Group.radicalCount.__set__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":194
 *         # experimental exception made for __getbuffer__ and __releasebuffer__
 *         # -- the details of this may change.
 *         def __getbuffer__(ndarray self, Py_buffer* info, int flags):             # <<<<<<<<<<<<<<
 *             # This implementation of getbuffer is geared towards Cython
 *             # requirements, and does not yet fullfill the PEP.
 */

/* Python wrapper */
static CYTHON_UNUSED int __pyx_pw_5numpy_7ndarray_1__getbuffer__(PyObject *__pyx_v_self, Py_buffer *__pyx_v_info, int __pyx_v_flags); /*proto*/
static CYTHON_UNUSED int __pyx_pw_5numpy_7ndarray_1__getbuffer__(PyObject *__pyx_v_self, Py_buffer *__pyx_v_info, int __pyx_v_flags) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__getbuffer__ (wrapper)", 0);
  __pyx_r = __pyx_pf_5numpy_7ndarray___getbuffer__(((PyArrayObject *)__pyx_v_self), ((Py_buffer *)__pyx_v_info), ((int)__pyx_v_flags));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_5numpy_7ndarray___getbuffer__(PyArrayObject *__pyx_v_self, Py_buffer *__pyx_v_info, int __pyx_v_flags) {
  int __pyx_v_copy_shape;
  int __pyx_v_i;
  int __pyx_v_ndim;
  int __pyx_v_endian_detector;
  int __pyx_v_little_endian;
  int __pyx_v_t;
  char *__pyx_v_f;
  PyArray_Descr *__pyx_v_descr = 0;
  int __pyx_v_offset;
  int __pyx_v_hasfields;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  int __pyx_t_2;
  PyObject *__pyx_t_3 = NULL;
  int __pyx_t_4;
  int __pyx_t_5;
  PyObject *__pyx_t_6 = NULL;
  char *__pyx_t_7;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__getbuffer__", 0);
  if (__pyx_v_info != NULL) {
    __pyx_v_info->obj = Py_None; __Pyx_INCREF(Py_None);
    __Pyx_GIVEREF(__pyx_v_info->obj);
  }

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":200
 *             # of flags
 * 
 *             if info == NULL: return             # <<<<<<<<<<<<<<
 * 
 *             cdef int copy_shape, i, ndim
 */
  __pyx_t_1 = ((__pyx_v_info == NULL) != 0);
  if (__pyx_t_1) {
    __pyx_r = 0;
    goto __pyx_L0;
  }

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":203
 * 
 *             cdef int copy_shape, i, ndim
 *             cdef int endian_detector = 1             # <<<<<<<<<<<<<<
 *             cdef bint little_endian = ((<char*>&endian_detector)[0] != 0)
 * 
 */
  __pyx_v_endian_detector = 1;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":204
 *             cdef int copy_shape, i, ndim
 *             cdef int endian_detector = 1
 *             cdef bint little_endian = ((<char*>&endian_detector)[0] != 0)             # <<<<<<<<<<<<<<
 * 
 *             ndim = PyArray_NDIM(self)
 */
  __pyx_v_little_endian = ((((char *)(&__pyx_v_endian_detector))[0]) != 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":206
 *             cdef bint little_endian = ((<char*>&endian_detector)[0] != 0)
 * 
 *             ndim = PyArray_NDIM(self)             # <<<<<<<<<<<<<<
 * 
 *             if sizeof(npy_intp) != sizeof(Py_ssize_t):
 */
  __pyx_v_ndim = PyArray_NDIM(__pyx_v_self);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":208
 *             ndim = PyArray_NDIM(self)
 * 
 *             if sizeof(npy_intp) != sizeof(Py_ssize_t):             # <<<<<<<<<<<<<<
 *                 copy_shape = 1
 *             else:
 */
  __pyx_t_1 = (((sizeof(npy_intp)) != (sizeof(Py_ssize_t))) != 0);
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":209
 * 
 *             if sizeof(npy_intp) != sizeof(Py_ssize_t):
 *                 copy_shape = 1             # <<<<<<<<<<<<<<
 *             else:
 *                 copy_shape = 0
 */
    __pyx_v_copy_shape = 1;
    goto __pyx_L4;
  }
  /*else*/ {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":211
 *                 copy_shape = 1
 *             else:
 *                 copy_shape = 0             # <<<<<<<<<<<<<<
 * 
 *             if ((flags & pybuf.PyBUF_C_CONTIGUOUS == pybuf.PyBUF_C_CONTIGUOUS)
 */
    __pyx_v_copy_shape = 0;
  }
  __pyx_L4:;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":213
 *                 copy_shape = 0
 * 
 *             if ((flags & pybuf.PyBUF_C_CONTIGUOUS == pybuf.PyBUF_C_CONTIGUOUS)             # <<<<<<<<<<<<<<
 *                 and not PyArray_CHKFLAGS(self, NPY_C_CONTIGUOUS)):
 *                 raise ValueError(u"ndarray is not C contiguous")
 */
  __pyx_t_2 = (((__pyx_v_flags & PyBUF_C_CONTIGUOUS) == PyBUF_C_CONTIGUOUS) != 0);
  if (__pyx_t_2) {
    goto __pyx_L7_next_and;
  } else {
    __pyx_t_1 = __pyx_t_2;
    goto __pyx_L6_bool_binop_done;
  }
  __pyx_L7_next_and:;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":214
 * 
 *             if ((flags & pybuf.PyBUF_C_CONTIGUOUS == pybuf.PyBUF_C_CONTIGUOUS)
 *                 and not PyArray_CHKFLAGS(self, NPY_C_CONTIGUOUS)):             # <<<<<<<<<<<<<<
 *                 raise ValueError(u"ndarray is not C contiguous")
 * 
 */
  __pyx_t_2 = ((!(PyArray_CHKFLAGS(__pyx_v_self, NPY_C_CONTIGUOUS) != 0)) != 0);
  __pyx_t_1 = __pyx_t_2;
  __pyx_L6_bool_binop_done:;
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":215
 *             if ((flags & pybuf.PyBUF_C_CONTIGUOUS == pybuf.PyBUF_C_CONTIGUOUS)
 *                 and not PyArray_CHKFLAGS(self, NPY_C_CONTIGUOUS)):
 *                 raise ValueError(u"ndarray is not C contiguous")             # <<<<<<<<<<<<<<
 * 
 *             if ((flags & pybuf.PyBUF_F_CONTIGUOUS == pybuf.PyBUF_F_CONTIGUOUS)
 */
    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_ValueError, __pyx_tuple__20, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 215; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_Raise(__pyx_t_3, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    {__pyx_filename = __pyx_f[2]; __pyx_lineno = 215; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":217
 *                 raise ValueError(u"ndarray is not C contiguous")
 * 
 *             if ((flags & pybuf.PyBUF_F_CONTIGUOUS == pybuf.PyBUF_F_CONTIGUOUS)             # <<<<<<<<<<<<<<
 *                 and not PyArray_CHKFLAGS(self, NPY_F_CONTIGUOUS)):
 *                 raise ValueError(u"ndarray is not Fortran contiguous")
 */
  __pyx_t_2 = (((__pyx_v_flags & PyBUF_F_CONTIGUOUS) == PyBUF_F_CONTIGUOUS) != 0);
  if (__pyx_t_2) {
    goto __pyx_L10_next_and;
  } else {
    __pyx_t_1 = __pyx_t_2;
    goto __pyx_L9_bool_binop_done;
  }
  __pyx_L10_next_and:;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":218
 * 
 *             if ((flags & pybuf.PyBUF_F_CONTIGUOUS == pybuf.PyBUF_F_CONTIGUOUS)
 *                 and not PyArray_CHKFLAGS(self, NPY_F_CONTIGUOUS)):             # <<<<<<<<<<<<<<
 *                 raise ValueError(u"ndarray is not Fortran contiguous")
 * 
 */
  __pyx_t_2 = ((!(PyArray_CHKFLAGS(__pyx_v_self, NPY_F_CONTIGUOUS) != 0)) != 0);
  __pyx_t_1 = __pyx_t_2;
  __pyx_L9_bool_binop_done:;
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":219
 *             if ((flags & pybuf.PyBUF_F_CONTIGUOUS == pybuf.PyBUF_F_CONTIGUOUS)
 *                 and not PyArray_CHKFLAGS(self, NPY_F_CONTIGUOUS)):
 *                 raise ValueError(u"ndarray is not Fortran contiguous")             # <<<<<<<<<<<<<<
 * 
 *             info.buf = PyArray_DATA(self)
 */
    __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_ValueError, __pyx_tuple__21, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 219; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_Raise(__pyx_t_3, 0, 0, 0);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    {__pyx_filename = __pyx_f[2]; __pyx_lineno = 219; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":221
 *                 raise ValueError(u"ndarray is not Fortran contiguous")
 * 
 *             info.buf = PyArray_DATA(self)             # <<<<<<<<<<<<<<
 *             info.ndim = ndim
 *             if copy_shape:
 */
  __pyx_v_info->buf = PyArray_DATA(__pyx_v_self);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":222
 * 
 *             info.buf = PyArray_DATA(self)
 *             info.ndim = ndim             # <<<<<<<<<<<<<<
 *             if copy_shape:
 *                 # Allocate new buffer for strides and shape info.
 */
  __pyx_v_info->ndim = __pyx_v_ndim;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":223
 *             info.buf = PyArray_DATA(self)
 *             info.ndim = ndim
 *             if copy_shape:             # <<<<<<<<<<<<<<
 *                 # Allocate new buffer for strides and shape info.
 *                 # This is allocated as one block, strides first.
 */
  __pyx_t_1 = (__pyx_v_copy_shape != 0);
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":226
 *                 # Allocate new buffer for strides and shape info.
 *                 # This is allocated as one block, strides first.
 *                 info.strides = <Py_ssize_t*>stdlib.malloc(sizeof(Py_ssize_t) * <size_t>ndim * 2)             # <<<<<<<<<<<<<<
 *                 info.shape = info.strides + ndim
 *                 for i in range(ndim):
 */
    __pyx_v_info->strides = ((Py_ssize_t *)malloc((((sizeof(Py_ssize_t)) * ((size_t)__pyx_v_ndim)) * 2)));

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":227
 *                 # This is allocated as one block, strides first.
 *                 info.strides = <Py_ssize_t*>stdlib.malloc(sizeof(Py_ssize_t) * <size_t>ndim * 2)
 *                 info.shape = info.strides + ndim             # <<<<<<<<<<<<<<
 *                 for i in range(ndim):
 *                     info.strides[i] = PyArray_STRIDES(self)[i]
 */
    __pyx_v_info->shape = (__pyx_v_info->strides + __pyx_v_ndim);

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":228
 *                 info.strides = <Py_ssize_t*>stdlib.malloc(sizeof(Py_ssize_t) * <size_t>ndim * 2)
 *                 info.shape = info.strides + ndim
 *                 for i in range(ndim):             # <<<<<<<<<<<<<<
 *                     info.strides[i] = PyArray_STRIDES(self)[i]
 *                     info.shape[i] = PyArray_DIMS(self)[i]
 */
    __pyx_t_4 = __pyx_v_ndim;
    for (__pyx_t_5 = 0; __pyx_t_5 < __pyx_t_4; __pyx_t_5+=1) {
      __pyx_v_i = __pyx_t_5;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":229
 *                 info.shape = info.strides + ndim
 *                 for i in range(ndim):
 *                     info.strides[i] = PyArray_STRIDES(self)[i]             # <<<<<<<<<<<<<<
 *                     info.shape[i] = PyArray_DIMS(self)[i]
 *             else:
 */
      (__pyx_v_info->strides[__pyx_v_i]) = (PyArray_STRIDES(__pyx_v_self)[__pyx_v_i]);

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":230
 *                 for i in range(ndim):
 *                     info.strides[i] = PyArray_STRIDES(self)[i]
 *                     info.shape[i] = PyArray_DIMS(self)[i]             # <<<<<<<<<<<<<<
 *             else:
 *                 info.strides = <Py_ssize_t*>PyArray_STRIDES(self)
 */
      (__pyx_v_info->shape[__pyx_v_i]) = (PyArray_DIMS(__pyx_v_self)[__pyx_v_i]);
    }
    goto __pyx_L11;
  }
  /*else*/ {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":232
 *                     info.shape[i] = PyArray_DIMS(self)[i]
 *             else:
 *                 info.strides = <Py_ssize_t*>PyArray_STRIDES(self)             # <<<<<<<<<<<<<<
 *                 info.shape = <Py_ssize_t*>PyArray_DIMS(self)
 *             info.suboffsets = NULL
 */
    __pyx_v_info->strides = ((Py_ssize_t *)PyArray_STRIDES(__pyx_v_self));

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":233
 *             else:
 *                 info.strides = <Py_ssize_t*>PyArray_STRIDES(self)
 *                 info.shape = <Py_ssize_t*>PyArray_DIMS(self)             # <<<<<<<<<<<<<<
 *             info.suboffsets = NULL
 *             info.itemsize = PyArray_ITEMSIZE(self)
 */
    __pyx_v_info->shape = ((Py_ssize_t *)PyArray_DIMS(__pyx_v_self));
  }
  __pyx_L11:;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":234
 *                 info.strides = <Py_ssize_t*>PyArray_STRIDES(self)
 *                 info.shape = <Py_ssize_t*>PyArray_DIMS(self)
 *             info.suboffsets = NULL             # <<<<<<<<<<<<<<
 *             info.itemsize = PyArray_ITEMSIZE(self)
 *             info.readonly = not PyArray_ISWRITEABLE(self)
 */
  __pyx_v_info->suboffsets = NULL;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":235
 *                 info.shape = <Py_ssize_t*>PyArray_DIMS(self)
 *             info.suboffsets = NULL
 *             info.itemsize = PyArray_ITEMSIZE(self)             # <<<<<<<<<<<<<<
 *             info.readonly = not PyArray_ISWRITEABLE(self)
 * 
 */
  __pyx_v_info->itemsize = PyArray_ITEMSIZE(__pyx_v_self);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":236
 *             info.suboffsets = NULL
 *             info.itemsize = PyArray_ITEMSIZE(self)
 *             info.readonly = not PyArray_ISWRITEABLE(self)             # <<<<<<<<<<<<<<
 * 
 *             cdef int t
 */
  __pyx_v_info->readonly = (!(PyArray_ISWRITEABLE(__pyx_v_self) != 0));

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":239
 * 
 *             cdef int t
 *             cdef char* f = NULL             # <<<<<<<<<<<<<<
 *             cdef dtype descr = self.descr
 *             cdef list stack
 */
  __pyx_v_f = NULL;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":240
 *             cdef int t
 *             cdef char* f = NULL
 *             cdef dtype descr = self.descr             # <<<<<<<<<<<<<<
 *             cdef list stack
 *             cdef int offset
 */
  __pyx_t_3 = ((PyObject *)__pyx_v_self->descr);
  __Pyx_INCREF(__pyx_t_3);
  __pyx_v_descr = ((PyArray_Descr *)__pyx_t_3);
  __pyx_t_3 = 0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":244
 *             cdef int offset
 * 
 *             cdef bint hasfields = PyDataType_HASFIELDS(descr)             # <<<<<<<<<<<<<<
 * 
 *             if not hasfields and not copy_shape:
 */
  __pyx_v_hasfields = PyDataType_HASFIELDS(__pyx_v_descr);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":246
 *             cdef bint hasfields = PyDataType_HASFIELDS(descr)
 * 
 *             if not hasfields and not copy_shape:             # <<<<<<<<<<<<<<
 *                 # do not call releasebuffer
 *                 info.obj = None
 */
  __pyx_t_2 = ((!(__pyx_v_hasfields != 0)) != 0);
  if (__pyx_t_2) {
    goto __pyx_L16_next_and;
  } else {
    __pyx_t_1 = __pyx_t_2;
    goto __pyx_L15_bool_binop_done;
  }
  __pyx_L16_next_and:;
  __pyx_t_2 = ((!(__pyx_v_copy_shape != 0)) != 0);
  __pyx_t_1 = __pyx_t_2;
  __pyx_L15_bool_binop_done:;
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":248
 *             if not hasfields and not copy_shape:
 *                 # do not call releasebuffer
 *                 info.obj = None             # <<<<<<<<<<<<<<
 *             else:
 *                 # need to call releasebuffer
 */
    __Pyx_INCREF(Py_None);
    __Pyx_GIVEREF(Py_None);
    __Pyx_GOTREF(__pyx_v_info->obj);
    __Pyx_DECREF(__pyx_v_info->obj);
    __pyx_v_info->obj = Py_None;
    goto __pyx_L14;
  }
  /*else*/ {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":251
 *             else:
 *                 # need to call releasebuffer
 *                 info.obj = self             # <<<<<<<<<<<<<<
 * 
 *             if not hasfields:
 */
    __Pyx_INCREF(((PyObject *)__pyx_v_self));
    __Pyx_GIVEREF(((PyObject *)__pyx_v_self));
    __Pyx_GOTREF(__pyx_v_info->obj);
    __Pyx_DECREF(__pyx_v_info->obj);
    __pyx_v_info->obj = ((PyObject *)__pyx_v_self);
  }
  __pyx_L14:;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":253
 *                 info.obj = self
 * 
 *             if not hasfields:             # <<<<<<<<<<<<<<
 *                 t = descr.type_num
 *                 if ((descr.byteorder == c'>' and little_endian) or
 */
  __pyx_t_1 = ((!(__pyx_v_hasfields != 0)) != 0);
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":254
 * 
 *             if not hasfields:
 *                 t = descr.type_num             # <<<<<<<<<<<<<<
 *                 if ((descr.byteorder == c'>' and little_endian) or
 *                     (descr.byteorder == c'<' and not little_endian)):
 */
    __pyx_t_4 = __pyx_v_descr->type_num;
    __pyx_v_t = __pyx_t_4;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":255
 *             if not hasfields:
 *                 t = descr.type_num
 *                 if ((descr.byteorder == c'>' and little_endian) or             # <<<<<<<<<<<<<<
 *                     (descr.byteorder == c'<' and not little_endian)):
 *                     raise ValueError(u"Non-native byte order not supported")
 */
    __pyx_t_2 = ((__pyx_v_descr->byteorder == '>') != 0);
    if (!__pyx_t_2) {
      goto __pyx_L20_next_or;
    } else {
      goto __pyx_L21_next_and;
    }
    __pyx_L21_next_and:;
    __pyx_t_2 = (__pyx_v_little_endian != 0);
    if (!__pyx_t_2) {
      goto __pyx_L20_next_or;
    } else {
      __pyx_t_1 = __pyx_t_2;
      goto __pyx_L19_bool_binop_done;
    }
    __pyx_L20_next_or:;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":256
 *                 t = descr.type_num
 *                 if ((descr.byteorder == c'>' and little_endian) or
 *                     (descr.byteorder == c'<' and not little_endian)):             # <<<<<<<<<<<<<<
 *                     raise ValueError(u"Non-native byte order not supported")
 *                 if   t == NPY_BYTE:        f = "b"
 */
    __pyx_t_2 = ((__pyx_v_descr->byteorder == '<') != 0);
    if (__pyx_t_2) {
      goto __pyx_L22_next_and;
    } else {
      __pyx_t_1 = __pyx_t_2;
      goto __pyx_L19_bool_binop_done;
    }
    __pyx_L22_next_and:;
    __pyx_t_2 = ((!(__pyx_v_little_endian != 0)) != 0);
    __pyx_t_1 = __pyx_t_2;
    __pyx_L19_bool_binop_done:;
    if (__pyx_t_1) {

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":257
 *                 if ((descr.byteorder == c'>' and little_endian) or
 *                     (descr.byteorder == c'<' and not little_endian)):
 *                     raise ValueError(u"Non-native byte order not supported")             # <<<<<<<<<<<<<<
 *                 if   t == NPY_BYTE:        f = "b"
 *                 elif t == NPY_UBYTE:       f = "B"
 */
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_ValueError, __pyx_tuple__22, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_Raise(__pyx_t_3, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      {__pyx_filename = __pyx_f[2]; __pyx_lineno = 257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":274
 *                 elif t == NPY_CDOUBLE:     f = "Zd"
 *                 elif t == NPY_CLONGDOUBLE: f = "Zg"
 *                 elif t == NPY_OBJECT:      f = "O"             # <<<<<<<<<<<<<<
 *                 else:
 *                     raise ValueError(u"unknown dtype code in numpy.pxd (%d)" % t)
 */
    switch (__pyx_v_t) {

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":258
 *                     (descr.byteorder == c'<' and not little_endian)):
 *                     raise ValueError(u"Non-native byte order not supported")
 *                 if   t == NPY_BYTE:        f = "b"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_UBYTE:       f = "B"
 *                 elif t == NPY_SHORT:       f = "h"
 */
      case NPY_BYTE:
      __pyx_v_f = __pyx_k_b;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":259
 *                     raise ValueError(u"Non-native byte order not supported")
 *                 if   t == NPY_BYTE:        f = "b"
 *                 elif t == NPY_UBYTE:       f = "B"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_SHORT:       f = "h"
 *                 elif t == NPY_USHORT:      f = "H"
 */
      case NPY_UBYTE:
      __pyx_v_f = __pyx_k_B;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":260
 *                 if   t == NPY_BYTE:        f = "b"
 *                 elif t == NPY_UBYTE:       f = "B"
 *                 elif t == NPY_SHORT:       f = "h"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_USHORT:      f = "H"
 *                 elif t == NPY_INT:         f = "i"
 */
      case NPY_SHORT:
      __pyx_v_f = __pyx_k_h;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":261
 *                 elif t == NPY_UBYTE:       f = "B"
 *                 elif t == NPY_SHORT:       f = "h"
 *                 elif t == NPY_USHORT:      f = "H"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_INT:         f = "i"
 *                 elif t == NPY_UINT:        f = "I"
 */
      case NPY_USHORT:
      __pyx_v_f = __pyx_k_H;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":262
 *                 elif t == NPY_SHORT:       f = "h"
 *                 elif t == NPY_USHORT:      f = "H"
 *                 elif t == NPY_INT:         f = "i"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_UINT:        f = "I"
 *                 elif t == NPY_LONG:        f = "l"
 */
      case NPY_INT:
      __pyx_v_f = __pyx_k_i;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":263
 *                 elif t == NPY_USHORT:      f = "H"
 *                 elif t == NPY_INT:         f = "i"
 *                 elif t == NPY_UINT:        f = "I"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_LONG:        f = "l"
 *                 elif t == NPY_ULONG:       f = "L"
 */
      case NPY_UINT:
      __pyx_v_f = __pyx_k_I;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":264
 *                 elif t == NPY_INT:         f = "i"
 *                 elif t == NPY_UINT:        f = "I"
 *                 elif t == NPY_LONG:        f = "l"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_ULONG:       f = "L"
 *                 elif t == NPY_LONGLONG:    f = "q"
 */
      case NPY_LONG:
      __pyx_v_f = __pyx_k_l;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":265
 *                 elif t == NPY_UINT:        f = "I"
 *                 elif t == NPY_LONG:        f = "l"
 *                 elif t == NPY_ULONG:       f = "L"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_LONGLONG:    f = "q"
 *                 elif t == NPY_ULONGLONG:   f = "Q"
 */
      case NPY_ULONG:
      __pyx_v_f = __pyx_k_L;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":266
 *                 elif t == NPY_LONG:        f = "l"
 *                 elif t == NPY_ULONG:       f = "L"
 *                 elif t == NPY_LONGLONG:    f = "q"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_ULONGLONG:   f = "Q"
 *                 elif t == NPY_FLOAT:       f = "f"
 */
      case NPY_LONGLONG:
      __pyx_v_f = __pyx_k_q;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":267
 *                 elif t == NPY_ULONG:       f = "L"
 *                 elif t == NPY_LONGLONG:    f = "q"
 *                 elif t == NPY_ULONGLONG:   f = "Q"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_FLOAT:       f = "f"
 *                 elif t == NPY_DOUBLE:      f = "d"
 */
      case NPY_ULONGLONG:
      __pyx_v_f = __pyx_k_Q;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":268
 *                 elif t == NPY_LONGLONG:    f = "q"
 *                 elif t == NPY_ULONGLONG:   f = "Q"
 *                 elif t == NPY_FLOAT:       f = "f"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_DOUBLE:      f = "d"
 *                 elif t == NPY_LONGDOUBLE:  f = "g"
 */
      case NPY_FLOAT:
      __pyx_v_f = __pyx_k_f;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":269
 *                 elif t == NPY_ULONGLONG:   f = "Q"
 *                 elif t == NPY_FLOAT:       f = "f"
 *                 elif t == NPY_DOUBLE:      f = "d"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_LONGDOUBLE:  f = "g"
 *                 elif t == NPY_CFLOAT:      f = "Zf"
 */
      case NPY_DOUBLE:
      __pyx_v_f = __pyx_k_d;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":270
 *                 elif t == NPY_FLOAT:       f = "f"
 *                 elif t == NPY_DOUBLE:      f = "d"
 *                 elif t == NPY_LONGDOUBLE:  f = "g"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_CFLOAT:      f = "Zf"
 *                 elif t == NPY_CDOUBLE:     f = "Zd"
 */
      case NPY_LONGDOUBLE:
      __pyx_v_f = __pyx_k_g;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":271
 *                 elif t == NPY_DOUBLE:      f = "d"
 *                 elif t == NPY_LONGDOUBLE:  f = "g"
 *                 elif t == NPY_CFLOAT:      f = "Zf"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_CDOUBLE:     f = "Zd"
 *                 elif t == NPY_CLONGDOUBLE: f = "Zg"
 */
      case NPY_CFLOAT:
      __pyx_v_f = __pyx_k_Zf;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":272
 *                 elif t == NPY_LONGDOUBLE:  f = "g"
 *                 elif t == NPY_CFLOAT:      f = "Zf"
 *                 elif t == NPY_CDOUBLE:     f = "Zd"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_CLONGDOUBLE: f = "Zg"
 *                 elif t == NPY_OBJECT:      f = "O"
 */
      case NPY_CDOUBLE:
      __pyx_v_f = __pyx_k_Zd;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":273
 *                 elif t == NPY_CFLOAT:      f = "Zf"
 *                 elif t == NPY_CDOUBLE:     f = "Zd"
 *                 elif t == NPY_CLONGDOUBLE: f = "Zg"             # <<<<<<<<<<<<<<
 *                 elif t == NPY_OBJECT:      f = "O"
 *                 else:
 */
      case NPY_CLONGDOUBLE:
      __pyx_v_f = __pyx_k_Zg;
      break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":274
 *                 elif t == NPY_CDOUBLE:     f = "Zd"
 *                 elif t == NPY_CLONGDOUBLE: f = "Zg"
 *                 elif t == NPY_OBJECT:      f = "O"             # <<<<<<<<<<<<<<
 *                 else:
 *                     raise ValueError(u"unknown dtype code in numpy.pxd (%d)" % t)
 */
      case NPY_OBJECT:
      __pyx_v_f = __pyx_k_O;
      break;
      default:

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":276
 *                 elif t == NPY_OBJECT:      f = "O"
 *                 else:
 *                     raise ValueError(u"unknown dtype code in numpy.pxd (%d)" % t)             # <<<<<<<<<<<<<<
 *                 info.format = f
 *                 return
 */
      __pyx_t_3 = __Pyx_PyInt_From_int(__pyx_v_t); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_6 = PyUnicode_Format(__pyx_kp_u_unknown_dtype_code_in_numpy_pxd, __pyx_t_3); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_3 = PyTuple_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_6);
      __Pyx_GIVEREF(__pyx_t_6);
      __pyx_t_6 = 0;
      __pyx_t_6 = __Pyx_PyObject_Call(__pyx_builtin_ValueError, __pyx_t_3, NULL); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_6);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __Pyx_Raise(__pyx_t_6, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
      {__pyx_filename = __pyx_f[2]; __pyx_lineno = 276; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      break;
    }

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":277
 *                 else:
 *                     raise ValueError(u"unknown dtype code in numpy.pxd (%d)" % t)
 *                 info.format = f             # <<<<<<<<<<<<<<
 *                 return
 *             else:
 */
    __pyx_v_info->format = __pyx_v_f;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":278
 *                     raise ValueError(u"unknown dtype code in numpy.pxd (%d)" % t)
 *                 info.format = f
 *                 return             # <<<<<<<<<<<<<<
 *             else:
 *                 info.format = <char*>stdlib.malloc(_buffer_format_string_len)
 */
    __pyx_r = 0;
    goto __pyx_L0;
  }
  /*else*/ {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":280
 *                 return
 *             else:
 *                 info.format = <char*>stdlib.malloc(_buffer_format_string_len)             # <<<<<<<<<<<<<<
 *                 info.format[0] = c'^' # Native data types, manual alignment
 *                 offset = 0
 */
    __pyx_v_info->format = ((char *)malloc(255));

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":281
 *             else:
 *                 info.format = <char*>stdlib.malloc(_buffer_format_string_len)
 *                 info.format[0] = c'^' # Native data types, manual alignment             # <<<<<<<<<<<<<<
 *                 offset = 0
 *                 f = _util_dtypestring(descr, info.format + 1,
 */
    (__pyx_v_info->format[0]) = '^';

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":282
 *                 info.format = <char*>stdlib.malloc(_buffer_format_string_len)
 *                 info.format[0] = c'^' # Native data types, manual alignment
 *                 offset = 0             # <<<<<<<<<<<<<<
 *                 f = _util_dtypestring(descr, info.format + 1,
 *                                       info.format + _buffer_format_string_len,
 */
    __pyx_v_offset = 0;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":283
 *                 info.format[0] = c'^' # Native data types, manual alignment
 *                 offset = 0
 *                 f = _util_dtypestring(descr, info.format + 1,             # <<<<<<<<<<<<<<
 *                                       info.format + _buffer_format_string_len,
 *                                       &offset)
 */
    __pyx_t_7 = __pyx_f_5numpy__util_dtypestring(__pyx_v_descr, (__pyx_v_info->format + 1), (__pyx_v_info->format + 255), (&__pyx_v_offset)); if (unlikely(__pyx_t_7 == NULL)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __pyx_v_f = __pyx_t_7;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":286
 *                                       info.format + _buffer_format_string_len,
 *                                       &offset)
 *                 f[0] = c'\0' # Terminate format string             # <<<<<<<<<<<<<<
 * 
 *         def __releasebuffer__(ndarray self, Py_buffer* info):
 */
    (__pyx_v_f[0]) = '\x00';
  }

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":194
 *         # experimental exception made for __getbuffer__ and __releasebuffer__
 *         # -- the details of this may change.
 *         def __getbuffer__(ndarray self, Py_buffer* info, int flags):             # <<<<<<<<<<<<<<
 *             # This implementation of getbuffer is geared towards Cython
 *             # requirements, and does not yet fullfill the PEP.
 */

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_6);
  __Pyx_AddTraceback("numpy.ndarray.__getbuffer__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  if (__pyx_v_info != NULL && __pyx_v_info->obj != NULL) {
    __Pyx_GOTREF(__pyx_v_info->obj);
    __Pyx_DECREF(__pyx_v_info->obj); __pyx_v_info->obj = NULL;
  }
  goto __pyx_L2;
  __pyx_L0:;
  if (__pyx_v_info != NULL && __pyx_v_info->obj == Py_None) {
    __Pyx_GOTREF(Py_None);
    __Pyx_DECREF(Py_None); __pyx_v_info->obj = NULL;
  }
  __pyx_L2:;
  __Pyx_XDECREF((PyObject *)__pyx_v_descr);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":288
 *                 f[0] = c'\0' # Terminate format string
 * 
 *         def __releasebuffer__(ndarray self, Py_buffer* info):             # <<<<<<<<<<<<<<
 *             if PyArray_HASFIELDS(self):
 *                 stdlib.free(info.format)
 */

/* Python wrapper */
static CYTHON_UNUSED void __pyx_pw_5numpy_7ndarray_3__releasebuffer__(PyObject *__pyx_v_self, Py_buffer *__pyx_v_info); /*proto*/
static CYTHON_UNUSED void __pyx_pw_5numpy_7ndarray_3__releasebuffer__(PyObject *__pyx_v_self, Py_buffer *__pyx_v_info) {
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__releasebuffer__ (wrapper)", 0);
  __pyx_pf_5numpy_7ndarray_2__releasebuffer__(((PyArrayObject *)__pyx_v_self), ((Py_buffer *)__pyx_v_info));

  /* function exit code */
  __Pyx_RefNannyFinishContext();
}

static void __pyx_pf_5numpy_7ndarray_2__releasebuffer__(PyArrayObject *__pyx_v_self, Py_buffer *__pyx_v_info) {
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  __Pyx_RefNannySetupContext("__releasebuffer__", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":289
 * 
 *         def __releasebuffer__(ndarray self, Py_buffer* info):
 *             if PyArray_HASFIELDS(self):             # <<<<<<<<<<<<<<
 *                 stdlib.free(info.format)
 *             if sizeof(npy_intp) != sizeof(Py_ssize_t):
 */
  __pyx_t_1 = (PyArray_HASFIELDS(__pyx_v_self) != 0);
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":290
 *         def __releasebuffer__(ndarray self, Py_buffer* info):
 *             if PyArray_HASFIELDS(self):
 *                 stdlib.free(info.format)             # <<<<<<<<<<<<<<
 *             if sizeof(npy_intp) != sizeof(Py_ssize_t):
 *                 stdlib.free(info.strides)
 */
    free(__pyx_v_info->format);
    goto __pyx_L3;
  }
  __pyx_L3:;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":291
 *             if PyArray_HASFIELDS(self):
 *                 stdlib.free(info.format)
 *             if sizeof(npy_intp) != sizeof(Py_ssize_t):             # <<<<<<<<<<<<<<
 *                 stdlib.free(info.strides)
 *                 # info.shape was stored after info.strides in the same block
 */
  __pyx_t_1 = (((sizeof(npy_intp)) != (sizeof(Py_ssize_t))) != 0);
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":292
 *                 stdlib.free(info.format)
 *             if sizeof(npy_intp) != sizeof(Py_ssize_t):
 *                 stdlib.free(info.strides)             # <<<<<<<<<<<<<<
 *                 # info.shape was stored after info.strides in the same block
 * 
 */
    free(__pyx_v_info->strides);
    goto __pyx_L4;
  }
  __pyx_L4:;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":288
 *                 f[0] = c'\0' # Terminate format string
 * 
 *         def __releasebuffer__(ndarray self, Py_buffer* info):             # <<<<<<<<<<<<<<
 *             if PyArray_HASFIELDS(self):
 *                 stdlib.free(info.format)
 */

  /* function exit code */
  __Pyx_RefNannyFinishContext();
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":768
 * ctypedef npy_cdouble     complex_t
 * 
 * cdef inline object PyArray_MultiIterNew1(a):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(1, <void*>a)
 * 
 */

static CYTHON_INLINE PyObject *__pyx_f_5numpy_PyArray_MultiIterNew1(PyObject *__pyx_v_a) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("PyArray_MultiIterNew1", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":769
 * 
 * cdef inline object PyArray_MultiIterNew1(a):
 *     return PyArray_MultiIterNew(1, <void*>a)             # <<<<<<<<<<<<<<
 * 
 * cdef inline object PyArray_MultiIterNew2(a, b):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyArray_MultiIterNew(1, ((void *)__pyx_v_a)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 769; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":768
 * ctypedef npy_cdouble     complex_t
 * 
 * cdef inline object PyArray_MultiIterNew1(a):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(1, <void*>a)
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("numpy.PyArray_MultiIterNew1", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":771
 *     return PyArray_MultiIterNew(1, <void*>a)
 * 
 * cdef inline object PyArray_MultiIterNew2(a, b):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)
 * 
 */

static CYTHON_INLINE PyObject *__pyx_f_5numpy_PyArray_MultiIterNew2(PyObject *__pyx_v_a, PyObject *__pyx_v_b) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("PyArray_MultiIterNew2", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":772
 * 
 * cdef inline object PyArray_MultiIterNew2(a, b):
 *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)             # <<<<<<<<<<<<<<
 * 
 * cdef inline object PyArray_MultiIterNew3(a, b, c):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyArray_MultiIterNew(2, ((void *)__pyx_v_a), ((void *)__pyx_v_b)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 772; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":771
 *     return PyArray_MultiIterNew(1, <void*>a)
 * 
 * cdef inline object PyArray_MultiIterNew2(a, b):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("numpy.PyArray_MultiIterNew2", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":774
 *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)
 * 
 * cdef inline object PyArray_MultiIterNew3(a, b, c):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)
 * 
 */

static CYTHON_INLINE PyObject *__pyx_f_5numpy_PyArray_MultiIterNew3(PyObject *__pyx_v_a, PyObject *__pyx_v_b, PyObject *__pyx_v_c) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("PyArray_MultiIterNew3", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":775
 * 
 * cdef inline object PyArray_MultiIterNew3(a, b, c):
 *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)             # <<<<<<<<<<<<<<
 * 
 * cdef inline object PyArray_MultiIterNew4(a, b, c, d):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyArray_MultiIterNew(3, ((void *)__pyx_v_a), ((void *)__pyx_v_b), ((void *)__pyx_v_c)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 775; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":774
 *     return PyArray_MultiIterNew(2, <void*>a, <void*>b)
 * 
 * cdef inline object PyArray_MultiIterNew3(a, b, c):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("numpy.PyArray_MultiIterNew3", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":777
 *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)
 * 
 * cdef inline object PyArray_MultiIterNew4(a, b, c, d):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)
 * 
 */

static CYTHON_INLINE PyObject *__pyx_f_5numpy_PyArray_MultiIterNew4(PyObject *__pyx_v_a, PyObject *__pyx_v_b, PyObject *__pyx_v_c, PyObject *__pyx_v_d) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("PyArray_MultiIterNew4", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":778
 * 
 * cdef inline object PyArray_MultiIterNew4(a, b, c, d):
 *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)             # <<<<<<<<<<<<<<
 * 
 * cdef inline object PyArray_MultiIterNew5(a, b, c, d, e):
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyArray_MultiIterNew(4, ((void *)__pyx_v_a), ((void *)__pyx_v_b), ((void *)__pyx_v_c), ((void *)__pyx_v_d)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 778; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":777
 *     return PyArray_MultiIterNew(3, <void*>a, <void*>b, <void*> c)
 * 
 * cdef inline object PyArray_MultiIterNew4(a, b, c, d):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("numpy.PyArray_MultiIterNew4", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":780
 *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)
 * 
 * cdef inline object PyArray_MultiIterNew5(a, b, c, d, e):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)
 * 
 */

static CYTHON_INLINE PyObject *__pyx_f_5numpy_PyArray_MultiIterNew5(PyObject *__pyx_v_a, PyObject *__pyx_v_b, PyObject *__pyx_v_c, PyObject *__pyx_v_d, PyObject *__pyx_v_e) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("PyArray_MultiIterNew5", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":781
 * 
 * cdef inline object PyArray_MultiIterNew5(a, b, c, d, e):
 *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)             # <<<<<<<<<<<<<<
 * 
 * cdef inline char* _util_dtypestring(dtype descr, char* f, char* end, int* offset) except NULL:
 */
  __Pyx_XDECREF(__pyx_r);
  __pyx_t_1 = PyArray_MultiIterNew(5, ((void *)__pyx_v_a), ((void *)__pyx_v_b), ((void *)__pyx_v_c), ((void *)__pyx_v_d), ((void *)__pyx_v_e)); if (unlikely(!__pyx_t_1)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 781; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_r = __pyx_t_1;
  __pyx_t_1 = 0;
  goto __pyx_L0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":780
 *     return PyArray_MultiIterNew(4, <void*>a, <void*>b, <void*>c, <void*> d)
 * 
 * cdef inline object PyArray_MultiIterNew5(a, b, c, d, e):             # <<<<<<<<<<<<<<
 *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)
 * 
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("numpy.PyArray_MultiIterNew5", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":783
 *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)
 * 
 * cdef inline char* _util_dtypestring(dtype descr, char* f, char* end, int* offset) except NULL:             # <<<<<<<<<<<<<<
 *     # Recursive utility function used in __getbuffer__ to get format
 *     # string. The new location in the format string is returned.
 */

static CYTHON_INLINE char *__pyx_f_5numpy__util_dtypestring(PyArray_Descr *__pyx_v_descr, char *__pyx_v_f, char *__pyx_v_end, int *__pyx_v_offset) {
  PyArray_Descr *__pyx_v_child = 0;
  int __pyx_v_endian_detector;
  int __pyx_v_little_endian;
  PyObject *__pyx_v_fields = 0;
  PyObject *__pyx_v_childname = NULL;
  PyObject *__pyx_v_new_offset = NULL;
  PyObject *__pyx_v_t = NULL;
  char *__pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  Py_ssize_t __pyx_t_2;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  int __pyx_t_5;
  int __pyx_t_6;
  int __pyx_t_7;
  long __pyx_t_8;
  char *__pyx_t_9;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("_util_dtypestring", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":790
 *     cdef int delta_offset
 *     cdef tuple i
 *     cdef int endian_detector = 1             # <<<<<<<<<<<<<<
 *     cdef bint little_endian = ((<char*>&endian_detector)[0] != 0)
 *     cdef tuple fields
 */
  __pyx_v_endian_detector = 1;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":791
 *     cdef tuple i
 *     cdef int endian_detector = 1
 *     cdef bint little_endian = ((<char*>&endian_detector)[0] != 0)             # <<<<<<<<<<<<<<
 *     cdef tuple fields
 * 
 */
  __pyx_v_little_endian = ((((char *)(&__pyx_v_endian_detector))[0]) != 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":794
 *     cdef tuple fields
 * 
 *     for childname in descr.names:             # <<<<<<<<<<<<<<
 *         fields = descr.fields[childname]
 *         child, new_offset = fields
 */
  if (unlikely(__pyx_v_descr->names == Py_None)) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
    {__pyx_filename = __pyx_f[2]; __pyx_lineno = 794; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  }
  __pyx_t_1 = __pyx_v_descr->names; __Pyx_INCREF(__pyx_t_1); __pyx_t_2 = 0;
  for (;;) {
    if (__pyx_t_2 >= PyTuple_GET_SIZE(__pyx_t_1)) break;
    #if CYTHON_COMPILING_IN_CPYTHON
    __pyx_t_3 = PyTuple_GET_ITEM(__pyx_t_1, __pyx_t_2); __Pyx_INCREF(__pyx_t_3); __pyx_t_2++; if (unlikely(0 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 794; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #else
    __pyx_t_3 = PySequence_ITEM(__pyx_t_1, __pyx_t_2); __pyx_t_2++; if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 794; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    #endif
    __Pyx_XDECREF_SET(__pyx_v_childname, __pyx_t_3);
    __pyx_t_3 = 0;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":795
 * 
 *     for childname in descr.names:
 *         fields = descr.fields[childname]             # <<<<<<<<<<<<<<
 *         child, new_offset = fields
 * 
 */
    __pyx_t_3 = PyObject_GetItem(__pyx_v_descr->fields, __pyx_v_childname); if (unlikely(__pyx_t_3 == NULL)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 795; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
    __Pyx_GOTREF(__pyx_t_3);
    if (!(likely(PyTuple_CheckExact(__pyx_t_3))||((__pyx_t_3) == Py_None)||(PyErr_Format(PyExc_TypeError, "Expected %.16s, got %.200s", "tuple", Py_TYPE(__pyx_t_3)->tp_name), 0))) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 795; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_fields, ((PyObject*)__pyx_t_3));
    __pyx_t_3 = 0;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":796
 *     for childname in descr.names:
 *         fields = descr.fields[childname]
 *         child, new_offset = fields             # <<<<<<<<<<<<<<
 * 
 *         if (end - f) - <int>(new_offset - offset[0]) < 15:
 */
    if (likely(__pyx_v_fields != Py_None)) {
      PyObject* sequence = __pyx_v_fields;
      #if CYTHON_COMPILING_IN_CPYTHON
      Py_ssize_t size = Py_SIZE(sequence);
      #else
      Py_ssize_t size = PySequence_Size(sequence);
      #endif
      if (unlikely(size != 2)) {
        if (size > 2) __Pyx_RaiseTooManyValuesError(2);
        else if (size >= 0) __Pyx_RaiseNeedMoreValuesError(size);
        {__pyx_filename = __pyx_f[2]; __pyx_lineno = 796; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      #if CYTHON_COMPILING_IN_CPYTHON
      __pyx_t_3 = PyTuple_GET_ITEM(sequence, 0); 
      __pyx_t_4 = PyTuple_GET_ITEM(sequence, 1); 
      __Pyx_INCREF(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_4);
      #else
      __pyx_t_3 = PySequence_ITEM(sequence, 0); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 796; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PySequence_ITEM(sequence, 1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 796; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      #endif
    } else {
      __Pyx_RaiseNoneNotIterableError(); {__pyx_filename = __pyx_f[2]; __pyx_lineno = 796; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
    if (!(likely(((__pyx_t_3) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_3, __pyx_ptype_5numpy_dtype))))) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 796; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_XDECREF_SET(__pyx_v_child, ((PyArray_Descr *)__pyx_t_3));
    __pyx_t_3 = 0;
    __Pyx_XDECREF_SET(__pyx_v_new_offset, __pyx_t_4);
    __pyx_t_4 = 0;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":798
 *         child, new_offset = fields
 * 
 *         if (end - f) - <int>(new_offset - offset[0]) < 15:             # <<<<<<<<<<<<<<
 *             raise RuntimeError(u"Format string allocated too short, see comment in numpy.pxd")
 * 
 */
    __pyx_t_4 = __Pyx_PyInt_From_int((__pyx_v_offset[0])); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 798; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_3 = PyNumber_Subtract(__pyx_v_new_offset, __pyx_t_4); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 798; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_GOTREF(__pyx_t_3);
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
    __pyx_t_5 = __Pyx_PyInt_As_int(__pyx_t_3); if (unlikely((__pyx_t_5 == (int)-1) && PyErr_Occurred())) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 798; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
    __pyx_t_6 = ((((__pyx_v_end - __pyx_v_f) - ((int)__pyx_t_5)) < 15) != 0);
    if (__pyx_t_6) {

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":799
 * 
 *         if (end - f) - <int>(new_offset - offset[0]) < 15:
 *             raise RuntimeError(u"Format string allocated too short, see comment in numpy.pxd")             # <<<<<<<<<<<<<<
 * 
 *         if ((child.byteorder == c'>' and little_endian) or
 */
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__23, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 799; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_Raise(__pyx_t_3, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      {__pyx_filename = __pyx_f[2]; __pyx_lineno = 799; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":801
 *             raise RuntimeError(u"Format string allocated too short, see comment in numpy.pxd")
 * 
 *         if ((child.byteorder == c'>' and little_endian) or             # <<<<<<<<<<<<<<
 *             (child.byteorder == c'<' and not little_endian)):
 *             raise ValueError(u"Non-native byte order not supported")
 */
    __pyx_t_7 = ((__pyx_v_child->byteorder == '>') != 0);
    if (!__pyx_t_7) {
      goto __pyx_L8_next_or;
    } else {
      goto __pyx_L9_next_and;
    }
    __pyx_L9_next_and:;
    __pyx_t_7 = (__pyx_v_little_endian != 0);
    if (!__pyx_t_7) {
      goto __pyx_L8_next_or;
    } else {
      __pyx_t_6 = __pyx_t_7;
      goto __pyx_L7_bool_binop_done;
    }
    __pyx_L8_next_or:;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":802
 * 
 *         if ((child.byteorder == c'>' and little_endian) or
 *             (child.byteorder == c'<' and not little_endian)):             # <<<<<<<<<<<<<<
 *             raise ValueError(u"Non-native byte order not supported")
 *             # One could encode it in the format string and have Cython
 */
    __pyx_t_7 = ((__pyx_v_child->byteorder == '<') != 0);
    if (__pyx_t_7) {
      goto __pyx_L10_next_and;
    } else {
      __pyx_t_6 = __pyx_t_7;
      goto __pyx_L7_bool_binop_done;
    }
    __pyx_L10_next_and:;
    __pyx_t_7 = ((!(__pyx_v_little_endian != 0)) != 0);
    __pyx_t_6 = __pyx_t_7;
    __pyx_L7_bool_binop_done:;
    if (__pyx_t_6) {

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":803
 *         if ((child.byteorder == c'>' and little_endian) or
 *             (child.byteorder == c'<' and not little_endian)):
 *             raise ValueError(u"Non-native byte order not supported")             # <<<<<<<<<<<<<<
 *             # One could encode it in the format string and have Cython
 *             # complain instead, BUT: < and > in format strings also imply
 */
      __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_ValueError, __pyx_tuple__24, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 803; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __Pyx_Raise(__pyx_t_3, 0, 0, 0);
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      {__pyx_filename = __pyx_f[2]; __pyx_lineno = 803; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":813
 * 
 *         # Output padding bytes
 *         while offset[0] < new_offset:             # <<<<<<<<<<<<<<
 *             f[0] = 120 # "x"; pad byte
 *             f += 1
 */
    while (1) {
      __pyx_t_3 = __Pyx_PyInt_From_int((__pyx_v_offset[0])); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 813; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_t_3, __pyx_v_new_offset, Py_LT); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 813; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 813; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (!__pyx_t_6) break;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":814
 *         # Output padding bytes
 *         while offset[0] < new_offset:
 *             f[0] = 120 # "x"; pad byte             # <<<<<<<<<<<<<<
 *             f += 1
 *             offset[0] += 1
 */
      (__pyx_v_f[0]) = 120;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":815
 *         while offset[0] < new_offset:
 *             f[0] = 120 # "x"; pad byte
 *             f += 1             # <<<<<<<<<<<<<<
 *             offset[0] += 1
 * 
 */
      __pyx_v_f = (__pyx_v_f + 1);

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":816
 *             f[0] = 120 # "x"; pad byte
 *             f += 1
 *             offset[0] += 1             # <<<<<<<<<<<<<<
 * 
 *         offset[0] += child.itemsize
 */
      __pyx_t_8 = 0;
      (__pyx_v_offset[__pyx_t_8]) = ((__pyx_v_offset[__pyx_t_8]) + 1);
    }

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":818
 *             offset[0] += 1
 * 
 *         offset[0] += child.itemsize             # <<<<<<<<<<<<<<
 * 
 *         if not PyDataType_HASFIELDS(child):
 */
    __pyx_t_8 = 0;
    (__pyx_v_offset[__pyx_t_8]) = ((__pyx_v_offset[__pyx_t_8]) + __pyx_v_child->elsize);

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":820
 *         offset[0] += child.itemsize
 * 
 *         if not PyDataType_HASFIELDS(child):             # <<<<<<<<<<<<<<
 *             t = child.type_num
 *             if end - f < 5:
 */
    __pyx_t_6 = ((!(PyDataType_HASFIELDS(__pyx_v_child) != 0)) != 0);
    if (__pyx_t_6) {

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":821
 * 
 *         if not PyDataType_HASFIELDS(child):
 *             t = child.type_num             # <<<<<<<<<<<<<<
 *             if end - f < 5:
 *                 raise RuntimeError(u"Format string allocated too short.")
 */
      __pyx_t_4 = __Pyx_PyInt_From_int(__pyx_v_child->type_num); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 821; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __Pyx_XDECREF_SET(__pyx_v_t, __pyx_t_4);
      __pyx_t_4 = 0;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":822
 *         if not PyDataType_HASFIELDS(child):
 *             t = child.type_num
 *             if end - f < 5:             # <<<<<<<<<<<<<<
 *                 raise RuntimeError(u"Format string allocated too short.")
 * 
 */
      __pyx_t_6 = (((__pyx_v_end - __pyx_v_f) < 5) != 0);
      if (__pyx_t_6) {

        /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":823
 *             t = child.type_num
 *             if end - f < 5:
 *                 raise RuntimeError(u"Format string allocated too short.")             # <<<<<<<<<<<<<<
 * 
 *             # Until ticket #99 is fixed, use integers to avoid warnings
 */
        __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_RuntimeError, __pyx_tuple__25, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 823; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        __Pyx_Raise(__pyx_t_4, 0, 0, 0);
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        {__pyx_filename = __pyx_f[2]; __pyx_lineno = 823; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":826
 * 
 *             # Until ticket #99 is fixed, use integers to avoid warnings
 *             if   t == NPY_BYTE:        f[0] =  98 #"b"             # <<<<<<<<<<<<<<
 *             elif t == NPY_UBYTE:       f[0] =  66 #"B"
 *             elif t == NPY_SHORT:       f[0] = 104 #"h"
 */
      __pyx_t_4 = PyInt_FromLong(NPY_BYTE); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 826; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 826; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 826; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 98;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":827
 *             # Until ticket #99 is fixed, use integers to avoid warnings
 *             if   t == NPY_BYTE:        f[0] =  98 #"b"
 *             elif t == NPY_UBYTE:       f[0] =  66 #"B"             # <<<<<<<<<<<<<<
 *             elif t == NPY_SHORT:       f[0] = 104 #"h"
 *             elif t == NPY_USHORT:      f[0] =  72 #"H"
 */
      __pyx_t_3 = PyInt_FromLong(NPY_UBYTE); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 827; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_t, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 827; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 827; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 66;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":828
 *             if   t == NPY_BYTE:        f[0] =  98 #"b"
 *             elif t == NPY_UBYTE:       f[0] =  66 #"B"
 *             elif t == NPY_SHORT:       f[0] = 104 #"h"             # <<<<<<<<<<<<<<
 *             elif t == NPY_USHORT:      f[0] =  72 #"H"
 *             elif t == NPY_INT:         f[0] = 105 #"i"
 */
      __pyx_t_4 = PyInt_FromLong(NPY_SHORT); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 828; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 828; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 828; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 104;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":829
 *             elif t == NPY_UBYTE:       f[0] =  66 #"B"
 *             elif t == NPY_SHORT:       f[0] = 104 #"h"
 *             elif t == NPY_USHORT:      f[0] =  72 #"H"             # <<<<<<<<<<<<<<
 *             elif t == NPY_INT:         f[0] = 105 #"i"
 *             elif t == NPY_UINT:        f[0] =  73 #"I"
 */
      __pyx_t_3 = PyInt_FromLong(NPY_USHORT); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 829; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_t, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 829; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 829; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 72;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":830
 *             elif t == NPY_SHORT:       f[0] = 104 #"h"
 *             elif t == NPY_USHORT:      f[0] =  72 #"H"
 *             elif t == NPY_INT:         f[0] = 105 #"i"             # <<<<<<<<<<<<<<
 *             elif t == NPY_UINT:        f[0] =  73 #"I"
 *             elif t == NPY_LONG:        f[0] = 108 #"l"
 */
      __pyx_t_4 = PyInt_FromLong(NPY_INT); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 830; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 830; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 830; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 105;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":831
 *             elif t == NPY_USHORT:      f[0] =  72 #"H"
 *             elif t == NPY_INT:         f[0] = 105 #"i"
 *             elif t == NPY_UINT:        f[0] =  73 #"I"             # <<<<<<<<<<<<<<
 *             elif t == NPY_LONG:        f[0] = 108 #"l"
 *             elif t == NPY_ULONG:       f[0] = 76  #"L"
 */
      __pyx_t_3 = PyInt_FromLong(NPY_UINT); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 831; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_t, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 831; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 831; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 73;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":832
 *             elif t == NPY_INT:         f[0] = 105 #"i"
 *             elif t == NPY_UINT:        f[0] =  73 #"I"
 *             elif t == NPY_LONG:        f[0] = 108 #"l"             # <<<<<<<<<<<<<<
 *             elif t == NPY_ULONG:       f[0] = 76  #"L"
 *             elif t == NPY_LONGLONG:    f[0] = 113 #"q"
 */
      __pyx_t_4 = PyInt_FromLong(NPY_LONG); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 832; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 832; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 832; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 108;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":833
 *             elif t == NPY_UINT:        f[0] =  73 #"I"
 *             elif t == NPY_LONG:        f[0] = 108 #"l"
 *             elif t == NPY_ULONG:       f[0] = 76  #"L"             # <<<<<<<<<<<<<<
 *             elif t == NPY_LONGLONG:    f[0] = 113 #"q"
 *             elif t == NPY_ULONGLONG:   f[0] = 81  #"Q"
 */
      __pyx_t_3 = PyInt_FromLong(NPY_ULONG); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_t, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 833; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 76;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":834
 *             elif t == NPY_LONG:        f[0] = 108 #"l"
 *             elif t == NPY_ULONG:       f[0] = 76  #"L"
 *             elif t == NPY_LONGLONG:    f[0] = 113 #"q"             # <<<<<<<<<<<<<<
 *             elif t == NPY_ULONGLONG:   f[0] = 81  #"Q"
 *             elif t == NPY_FLOAT:       f[0] = 102 #"f"
 */
      __pyx_t_4 = PyInt_FromLong(NPY_LONGLONG); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 834; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 834; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 834; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 113;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":835
 *             elif t == NPY_ULONG:       f[0] = 76  #"L"
 *             elif t == NPY_LONGLONG:    f[0] = 113 #"q"
 *             elif t == NPY_ULONGLONG:   f[0] = 81  #"Q"             # <<<<<<<<<<<<<<
 *             elif t == NPY_FLOAT:       f[0] = 102 #"f"
 *             elif t == NPY_DOUBLE:      f[0] = 100 #"d"
 */
      __pyx_t_3 = PyInt_FromLong(NPY_ULONGLONG); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 835; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_t, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 835; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 835; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 81;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":836
 *             elif t == NPY_LONGLONG:    f[0] = 113 #"q"
 *             elif t == NPY_ULONGLONG:   f[0] = 81  #"Q"
 *             elif t == NPY_FLOAT:       f[0] = 102 #"f"             # <<<<<<<<<<<<<<
 *             elif t == NPY_DOUBLE:      f[0] = 100 #"d"
 *             elif t == NPY_LONGDOUBLE:  f[0] = 103 #"g"
 */
      __pyx_t_4 = PyInt_FromLong(NPY_FLOAT); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 836; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 836; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 836; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 102;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":837
 *             elif t == NPY_ULONGLONG:   f[0] = 81  #"Q"
 *             elif t == NPY_FLOAT:       f[0] = 102 #"f"
 *             elif t == NPY_DOUBLE:      f[0] = 100 #"d"             # <<<<<<<<<<<<<<
 *             elif t == NPY_LONGDOUBLE:  f[0] = 103 #"g"
 *             elif t == NPY_CFLOAT:      f[0] = 90; f[1] = 102; f += 1 # Zf
 */
      __pyx_t_3 = PyInt_FromLong(NPY_DOUBLE); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_t, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 837; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 100;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":838
 *             elif t == NPY_FLOAT:       f[0] = 102 #"f"
 *             elif t == NPY_DOUBLE:      f[0] = 100 #"d"
 *             elif t == NPY_LONGDOUBLE:  f[0] = 103 #"g"             # <<<<<<<<<<<<<<
 *             elif t == NPY_CFLOAT:      f[0] = 90; f[1] = 102; f += 1 # Zf
 *             elif t == NPY_CDOUBLE:     f[0] = 90; f[1] = 100; f += 1 # Zd
 */
      __pyx_t_4 = PyInt_FromLong(NPY_LONGDOUBLE); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 838; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 838; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 838; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 103;
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":839
 *             elif t == NPY_DOUBLE:      f[0] = 100 #"d"
 *             elif t == NPY_LONGDOUBLE:  f[0] = 103 #"g"
 *             elif t == NPY_CFLOAT:      f[0] = 90; f[1] = 102; f += 1 # Zf             # <<<<<<<<<<<<<<
 *             elif t == NPY_CDOUBLE:     f[0] = 90; f[1] = 100; f += 1 # Zd
 *             elif t == NPY_CLONGDOUBLE: f[0] = 90; f[1] = 103; f += 1 # Zg
 */
      __pyx_t_3 = PyInt_FromLong(NPY_CFLOAT); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_t, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 839; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 90;
        (__pyx_v_f[1]) = 102;
        __pyx_v_f = (__pyx_v_f + 1);
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":840
 *             elif t == NPY_LONGDOUBLE:  f[0] = 103 #"g"
 *             elif t == NPY_CFLOAT:      f[0] = 90; f[1] = 102; f += 1 # Zf
 *             elif t == NPY_CDOUBLE:     f[0] = 90; f[1] = 100; f += 1 # Zd             # <<<<<<<<<<<<<<
 *             elif t == NPY_CLONGDOUBLE: f[0] = 90; f[1] = 103; f += 1 # Zg
 *             elif t == NPY_OBJECT:      f[0] = 79 #"O"
 */
      __pyx_t_4 = PyInt_FromLong(NPY_CDOUBLE); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 840; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 840; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 840; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 90;
        (__pyx_v_f[1]) = 100;
        __pyx_v_f = (__pyx_v_f + 1);
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":841
 *             elif t == NPY_CFLOAT:      f[0] = 90; f[1] = 102; f += 1 # Zf
 *             elif t == NPY_CDOUBLE:     f[0] = 90; f[1] = 100; f += 1 # Zd
 *             elif t == NPY_CLONGDOUBLE: f[0] = 90; f[1] = 103; f += 1 # Zg             # <<<<<<<<<<<<<<
 *             elif t == NPY_OBJECT:      f[0] = 79 #"O"
 *             else:
 */
      __pyx_t_3 = PyInt_FromLong(NPY_CLONGDOUBLE); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 841; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_3);
      __pyx_t_4 = PyObject_RichCompare(__pyx_v_t, __pyx_t_3, Py_EQ); __Pyx_XGOTREF(__pyx_t_4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 841; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_4); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 841; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 90;
        (__pyx_v_f[1]) = 103;
        __pyx_v_f = (__pyx_v_f + 1);
        goto __pyx_L15;
      }

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":842
 *             elif t == NPY_CDOUBLE:     f[0] = 90; f[1] = 100; f += 1 # Zd
 *             elif t == NPY_CLONGDOUBLE: f[0] = 90; f[1] = 103; f += 1 # Zg
 *             elif t == NPY_OBJECT:      f[0] = 79 #"O"             # <<<<<<<<<<<<<<
 *             else:
 *                 raise ValueError(u"unknown dtype code in numpy.pxd (%d)" % t)
 */
      __pyx_t_4 = PyInt_FromLong(NPY_OBJECT); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 842; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_GOTREF(__pyx_t_4);
      __pyx_t_3 = PyObject_RichCompare(__pyx_v_t, __pyx_t_4, Py_EQ); __Pyx_XGOTREF(__pyx_t_3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 842; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
      __pyx_t_6 = __Pyx_PyObject_IsTrue(__pyx_t_3); if (unlikely(__pyx_t_6 < 0)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 842; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
      if (__pyx_t_6) {
        (__pyx_v_f[0]) = 79;
        goto __pyx_L15;
      }
      /*else*/ {

        /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":844
 *             elif t == NPY_OBJECT:      f[0] = 79 #"O"
 *             else:
 *                 raise ValueError(u"unknown dtype code in numpy.pxd (%d)" % t)             # <<<<<<<<<<<<<<
 *             f += 1
 *         else:
 */
        __pyx_t_3 = PyUnicode_Format(__pyx_kp_u_unknown_dtype_code_in_numpy_pxd, __pyx_v_t); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 844; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __pyx_t_4 = PyTuple_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 844; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_4);
        PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_t_3);
        __Pyx_GIVEREF(__pyx_t_3);
        __pyx_t_3 = 0;
        __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_ValueError, __pyx_t_4, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 844; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
        __Pyx_GOTREF(__pyx_t_3);
        __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
        __Pyx_Raise(__pyx_t_3, 0, 0, 0);
        __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
        {__pyx_filename = __pyx_f[2]; __pyx_lineno = 844; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      }
      __pyx_L15:;

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":845
 *             else:
 *                 raise ValueError(u"unknown dtype code in numpy.pxd (%d)" % t)
 *             f += 1             # <<<<<<<<<<<<<<
 *         else:
 *             # Cython ignores struct boundary information ("T{...}"),
 */
      __pyx_v_f = (__pyx_v_f + 1);
      goto __pyx_L13;
    }
    /*else*/ {

      /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":849
 *             # Cython ignores struct boundary information ("T{...}"),
 *             # so don't output it
 *             f = _util_dtypestring(child, f, end, offset)             # <<<<<<<<<<<<<<
 *     return f
 * 
 */
      __pyx_t_9 = __pyx_f_5numpy__util_dtypestring(__pyx_v_child, __pyx_v_f, __pyx_v_end, __pyx_v_offset); if (unlikely(__pyx_t_9 == NULL)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 849; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
      __pyx_v_f = __pyx_t_9;
    }
    __pyx_L13:;

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":794
 *     cdef tuple fields
 * 
 *     for childname in descr.names:             # <<<<<<<<<<<<<<
 *         fields = descr.fields[childname]
 *         child, new_offset = fields
 */
  }
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":850
 *             # so don't output it
 *             f = _util_dtypestring(child, f, end, offset)
 *     return f             # <<<<<<<<<<<<<<
 * 
 * 
 */
  __pyx_r = __pyx_v_f;
  goto __pyx_L0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":783
 *     return PyArray_MultiIterNew(5, <void*>a, <void*>b, <void*>c, <void*> d, <void*> e)
 * 
 * cdef inline char* _util_dtypestring(dtype descr, char* f, char* end, int* offset) except NULL:             # <<<<<<<<<<<<<<
 *     # Recursive utility function used in __getbuffer__ to get format
 *     # string. The new location in the format string is returned.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_AddTraceback("numpy._util_dtypestring", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = NULL;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_child);
  __Pyx_XDECREF(__pyx_v_fields);
  __Pyx_XDECREF(__pyx_v_childname);
  __Pyx_XDECREF(__pyx_v_new_offset);
  __Pyx_XDECREF(__pyx_v_t);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":966
 * 
 * 
 * cdef inline void set_array_base(ndarray arr, object base):             # <<<<<<<<<<<<<<
 *      cdef PyObject* baseptr
 *      if base is None:
 */

static CYTHON_INLINE void __pyx_f_5numpy_set_array_base(PyArrayObject *__pyx_v_arr, PyObject *__pyx_v_base) {
  PyObject *__pyx_v_baseptr;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  int __pyx_t_2;
  __Pyx_RefNannySetupContext("set_array_base", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":968
 * cdef inline void set_array_base(ndarray arr, object base):
 *      cdef PyObject* baseptr
 *      if base is None:             # <<<<<<<<<<<<<<
 *          baseptr = NULL
 *      else:
 */
  __pyx_t_1 = (__pyx_v_base == Py_None);
  __pyx_t_2 = (__pyx_t_1 != 0);
  if (__pyx_t_2) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":969
 *      cdef PyObject* baseptr
 *      if base is None:
 *          baseptr = NULL             # <<<<<<<<<<<<<<
 *      else:
 *          Py_INCREF(base) # important to do this before decref below!
 */
    __pyx_v_baseptr = NULL;
    goto __pyx_L3;
  }
  /*else*/ {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":971
 *          baseptr = NULL
 *      else:
 *          Py_INCREF(base) # important to do this before decref below!             # <<<<<<<<<<<<<<
 *          baseptr = <PyObject*>base
 *      Py_XDECREF(arr.base)
 */
    Py_INCREF(__pyx_v_base);

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":972
 *      else:
 *          Py_INCREF(base) # important to do this before decref below!
 *          baseptr = <PyObject*>base             # <<<<<<<<<<<<<<
 *      Py_XDECREF(arr.base)
 *      arr.base = baseptr
 */
    __pyx_v_baseptr = ((PyObject *)__pyx_v_base);
  }
  __pyx_L3:;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":973
 *          Py_INCREF(base) # important to do this before decref below!
 *          baseptr = <PyObject*>base
 *      Py_XDECREF(arr.base)             # <<<<<<<<<<<<<<
 *      arr.base = baseptr
 * 
 */
  Py_XDECREF(__pyx_v_arr->base);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":974
 *          baseptr = <PyObject*>base
 *      Py_XDECREF(arr.base)
 *      arr.base = baseptr             # <<<<<<<<<<<<<<
 * 
 * cdef inline object get_array_base(ndarray arr):
 */
  __pyx_v_arr->base = __pyx_v_baseptr;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":966
 * 
 * 
 * cdef inline void set_array_base(ndarray arr, object base):             # <<<<<<<<<<<<<<
 *      cdef PyObject* baseptr
 *      if base is None:
 */

  /* function exit code */
  __Pyx_RefNannyFinishContext();
}

/* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":976
 *      arr.base = baseptr
 * 
 * cdef inline object get_array_base(ndarray arr):             # <<<<<<<<<<<<<<
 *     if arr.base is NULL:
 *         return None
 */

static CYTHON_INLINE PyObject *__pyx_f_5numpy_get_array_base(PyArrayObject *__pyx_v_arr) {
  PyObject *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  int __pyx_t_1;
  __Pyx_RefNannySetupContext("get_array_base", 0);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":977
 * 
 * cdef inline object get_array_base(ndarray arr):
 *     if arr.base is NULL:             # <<<<<<<<<<<<<<
 *         return None
 *     else:
 */
  __pyx_t_1 = ((__pyx_v_arr->base == NULL) != 0);
  if (__pyx_t_1) {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":978
 * cdef inline object get_array_base(ndarray arr):
 *     if arr.base is NULL:
 *         return None             # <<<<<<<<<<<<<<
 *     else:
 *         return <object>arr.base
 */
    __Pyx_XDECREF(__pyx_r);
    __Pyx_INCREF(Py_None);
    __pyx_r = Py_None;
    goto __pyx_L0;
  }
  /*else*/ {

    /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":980
 *         return None
 *     else:
 *         return <object>arr.base             # <<<<<<<<<<<<<<
 */
    __Pyx_XDECREF(__pyx_r);
    __Pyx_INCREF(((PyObject *)__pyx_v_arr->base));
    __pyx_r = ((PyObject *)__pyx_v_arr->base);
    goto __pyx_L0;
  }

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":976
 *      arr.base = baseptr
 * 
 * cdef inline object get_array_base(ndarray arr):             # <<<<<<<<<<<<<<
 *     if arr.base is NULL:
 *         return None
 */

  /* function exit code */
  __pyx_L0:;
  __Pyx_XGIVEREF(__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupAtom __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom;

static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_GroupAtom(PyTypeObject *t, PyObject *a, PyObject *k) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *p;
  PyObject *o = __pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_new(t, a, k);
  if (unlikely(!o)) return 0;
  p = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)o);
  p->__pyx_base.__pyx_vtab = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex*)__pyx_vtabptr_5rmgpy_8molecule_5group_GroupAtom;
  p->atomType = ((PyObject*)Py_None); Py_INCREF(Py_None);
  p->radicalElectrons = ((PyObject*)Py_None); Py_INCREF(Py_None);
  p->charge = ((PyObject*)Py_None); Py_INCREF(Py_None);
  p->label = ((PyObject*)Py_None); Py_INCREF(Py_None);
  p->lonePairs = ((PyObject*)Py_None); Py_INCREF(Py_None);
  return o;
}

static void __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupAtom(PyObject *o) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)o;
  #if PY_VERSION_HEX >= 0x030400a1
  if (unlikely(Py_TYPE(o)->tp_finalize) && !_PyGC_FINALIZED(o)) {
    if (PyObject_CallFinalizerFromDealloc(o)) return;
  }
  #endif
  PyObject_GC_UnTrack(o);
  Py_CLEAR(p->atomType);
  Py_CLEAR(p->radicalElectrons);
  Py_CLEAR(p->charge);
  Py_CLEAR(p->label);
  Py_CLEAR(p->lonePairs);
  PyObject_GC_Track(o);
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)) __pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_dealloc(o); else __Pyx_call_next_tp_dealloc(o, __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupAtom);
}

static int __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupAtom(PyObject *o, visitproc v, void *a) {
  int e;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)o;
  e = ((likely(__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)) ? ((__pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_traverse) ? __pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_traverse(o, v, a) : 0) : __Pyx_call_next_tp_traverse(o, v, a, __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupAtom)); if (e) return e;
  if (p->atomType) {
    e = (*v)(p->atomType, a); if (e) return e;
  }
  if (p->radicalElectrons) {
    e = (*v)(p->radicalElectrons, a); if (e) return e;
  }
  if (p->charge) {
    e = (*v)(p->charge, a); if (e) return e;
  }
  if (p->lonePairs) {
    e = (*v)(p->lonePairs, a); if (e) return e;
  }
  return 0;
}

static int __pyx_tp_clear_5rmgpy_8molecule_5group_GroupAtom(PyObject *o) {
  PyObject* tmp;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *)o;
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)) { if (__pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_clear) __pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_clear(o); } else __Pyx_call_next_tp_clear(o, __pyx_tp_clear_5rmgpy_8molecule_5group_GroupAtom);
  tmp = ((PyObject*)p->atomType);
  p->atomType = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  tmp = ((PyObject*)p->radicalElectrons);
  p->radicalElectrons = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  tmp = ((PyObject*)p->charge);
  p->charge = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  tmp = ((PyObject*)p->lonePairs);
  p->lonePairs = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  return 0;
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_atomType(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_atomType(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_8atomType_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_radicalElectrons(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_radicalElectrons(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_16radicalElectrons_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_charge(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_charge(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_6charge_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_label(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_label(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5label_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_lonePairs(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_lonePairs(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9lonePairs_5__del__(o);
  }
}

static PyMethodDef __pyx_methods_5rmgpy_8molecule_5group_GroupAtom[] = {
  {"__reduce__", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_3__reduce__, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_2__reduce__},
  {"__setstate__", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_5__setstate__, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_4__setstate__},
  {"bonds", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_11bonds, METH_NOARGS, 0},
  {"copy", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_13copy, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_12copy},
  {"__changeBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_15__changeBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_14__changeBond},
  {"__formBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_17__formBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_16__formBond},
  {"__breakBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_19__breakBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_18__breakBond},
  {"__gainRadical", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_21__gainRadical, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_20__gainRadical},
  {"__loseRadical", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_23__loseRadical, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_22__loseRadical},
  {"__gainPair", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_25__gainPair, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_24__gainPair},
  {"__losePair", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_27__losePair, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_26__losePair},
  {"applyAction", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_29applyAction, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_28applyAction},
  {"equivalent", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_31equivalent, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_30equivalent},
  {"isSpecificCaseOf", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_33isSpecificCaseOf, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_32isSpecificCaseOf},
  {"isOxygen", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_35isOxygen, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_34isOxygen},
  {"isSulfur", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_37isSulfur, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_36isSulfur},
  {"hasWildcards", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_39hasWildcards, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_38hasWildcards},
  {"makeSampleAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_41makeSampleAtom, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_40makeSampleAtom},
  {"getBondOrdersForAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_43getBondOrdersForAtom, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_42getBondOrdersForAtom},
  {0, 0, 0, 0}
};

static struct PyGetSetDef __pyx_getsets_5rmgpy_8molecule_5group_GroupAtom[] = {
  {(char *)"atomType", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_atomType, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_atomType, 0, 0},
  {(char *)"radicalElectrons", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_radicalElectrons, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_radicalElectrons, 0, 0},
  {(char *)"charge", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_charge, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_charge, 0, 0},
  {(char *)"label", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_label, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_label, 0, 0},
  {(char *)"lonePairs", __pyx_getprop_5rmgpy_8molecule_5group_9GroupAtom_lonePairs, __pyx_setprop_5rmgpy_8molecule_5group_9GroupAtom_lonePairs, 0, 0},
  {0, 0, 0, 0, 0}
};

static PyTypeObject __pyx_type_5rmgpy_8molecule_5group_GroupAtom = {
  PyVarObject_HEAD_INIT(0, 0)
  "rmgpy.molecule.group.GroupAtom", /*tp_name*/
  sizeof(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom), /*tp_basicsize*/
  0, /*tp_itemsize*/
  __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupAtom, /*tp_dealloc*/
  0, /*tp_print*/
  0, /*tp_getattr*/
  0, /*tp_setattr*/
  #if PY_MAJOR_VERSION < 3
  0, /*tp_compare*/
  #else
  0, /*reserved*/
  #endif
  __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_9__repr__, /*tp_repr*/
  0, /*tp_as_number*/
  0, /*tp_as_sequence*/
  0, /*tp_as_mapping*/
  0, /*tp_hash*/
  0, /*tp_call*/
  __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_7__str__, /*tp_str*/
  0, /*tp_getattro*/
  0, /*tp_setattro*/
  0, /*tp_as_buffer*/
  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE|Py_TPFLAGS_HAVE_GC, /*tp_flags*/
  "\n    An atom group. This class is based on the :class:`Atom` class, except that\n    it uses :ref:`atom types <atom-types>` instead of elements, and all\n    attributes are lists rather than individual values. The attributes are:\n\n    =================== =================== ====================================\n    Attribute           Type                Description\n    =================== =================== ====================================\n    `atomType`          ``list``            The allowed atom types (as :class:`AtomType` objects)\n    `radicalElectrons`  ``list``            The allowed numbers of radical electrons (as short integers)\n    `charge`            ``list``            The allowed formal charges (as short integers)\n    `label`             ``str``             A string label that can be used to tag individual atoms\n    `lonePairs`         ``list``            The number of lone electron pairs\n    =================== =================== ====================================\n\n    Each list represents a logical OR construct, i.e. an atom will match the\n    group if it matches *any* item in the list. However, the\n    `radicalElectrons`, and `charge` attributes are linked\n    such that an atom must match values from the same index in each of these in\n    order to match.\n    ", /*tp_doc*/
  __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupAtom, /*tp_traverse*/
  __pyx_tp_clear_5rmgpy_8molecule_5group_GroupAtom, /*tp_clear*/
  0, /*tp_richcompare*/
  0, /*tp_weaklistoffset*/
  0, /*tp_iter*/
  0, /*tp_iternext*/
  __pyx_methods_5rmgpy_8molecule_5group_GroupAtom, /*tp_methods*/
  0, /*tp_members*/
  __pyx_getsets_5rmgpy_8molecule_5group_GroupAtom, /*tp_getset*/
  0, /*tp_base*/
  0, /*tp_dict*/
  0, /*tp_descr_get*/
  0, /*tp_descr_set*/
  0, /*tp_dictoffset*/
  __pyx_pw_5rmgpy_8molecule_5group_9GroupAtom_1__init__, /*tp_init*/
  0, /*tp_alloc*/
  __pyx_tp_new_5rmgpy_8molecule_5group_GroupAtom, /*tp_new*/
  0, /*tp_free*/
  0, /*tp_is_gc*/
  0, /*tp_bases*/
  0, /*tp_mro*/
  0, /*tp_cache*/
  0, /*tp_subclasses*/
  0, /*tp_weaklist*/
  0, /*tp_del*/
  0, /*tp_version_tag*/
  #if PY_VERSION_HEX >= 0x030400a1
  0, /*tp_finalize*/
  #endif
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_GroupBond __pyx_vtable_5rmgpy_8molecule_5group_GroupBond;

static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_GroupBond(PyTypeObject *t, PyObject *a, PyObject *k) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *p;
  PyObject *o = __pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_new(t, a, k);
  if (unlikely(!o)) return 0;
  p = ((struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)o);
  p->__pyx_base.__pyx_vtab = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge*)__pyx_vtabptr_5rmgpy_8molecule_5group_GroupBond;
  p->order = ((PyObject*)Py_None); Py_INCREF(Py_None);
  return o;
}

static void __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupBond(PyObject *o) {
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)o;
  #if PY_VERSION_HEX >= 0x030400a1
  if (unlikely(Py_TYPE(o)->tp_finalize) && !_PyGC_FINALIZED(o)) {
    if (PyObject_CallFinalizerFromDealloc(o)) return;
  }
  #endif
  PyObject_GC_UnTrack(o);
  Py_CLEAR(p->order);
  PyObject_GC_Track(o);
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Edge)) __pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_dealloc(o); else __Pyx_call_next_tp_dealloc(o, __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupBond);
}

static int __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupBond(PyObject *o, visitproc v, void *a) {
  int e;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)o;
  e = ((likely(__pyx_ptype_5rmgpy_8molecule_5graph_Edge)) ? ((__pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_traverse) ? __pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_traverse(o, v, a) : 0) : __Pyx_call_next_tp_traverse(o, v, a, __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupBond)); if (e) return e;
  if (p->order) {
    e = (*v)(p->order, a); if (e) return e;
  }
  return 0;
}

static int __pyx_tp_clear_5rmgpy_8molecule_5group_GroupBond(PyObject *o) {
  PyObject* tmp;
  struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *p = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *)o;
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Edge)) { if (__pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_clear) __pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_clear(o); } else __Pyx_call_next_tp_clear(o, __pyx_tp_clear_5rmgpy_8molecule_5group_GroupBond);
  tmp = ((PyObject*)p->order);
  p->order = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  return 0;
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_9GroupBond_order(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_9GroupBond_order(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5order_5__del__(o);
  }
}

static PyMethodDef __pyx_methods_5rmgpy_8molecule_5group_GroupBond[] = {
  {"__reduce__", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_7__reduce__, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_6__reduce__},
  {"copy", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_9copy, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_8copy},
  {"isSingle", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_11isSingle, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_10isSingle},
  {"isDouble", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_13isDouble, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_12isDouble},
  {"isTriple", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_15isTriple, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_14isTriple},
  {"isBenzene", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_17isBenzene, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_16isBenzene},
  {"__changeBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_19__changeBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_18__changeBond},
  {"applyAction", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_21applyAction, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_20applyAction},
  {"equivalent", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_23equivalent, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_22equivalent},
  {"isSpecificCaseOf", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_25isSpecificCaseOf, METH_O, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_24isSpecificCaseOf},
  {"makeBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_9GroupBond_27makeBond, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_26makeBond},
  {0, 0, 0, 0}
};

static struct PyGetSetDef __pyx_getsets_5rmgpy_8molecule_5group_GroupBond[] = {
  {(char *)"order", __pyx_getprop_5rmgpy_8molecule_5group_9GroupBond_order, __pyx_setprop_5rmgpy_8molecule_5group_9GroupBond_order, 0, 0},
  {0, 0, 0, 0, 0}
};

static PyTypeObject __pyx_type_5rmgpy_8molecule_5group_GroupBond = {
  PyVarObject_HEAD_INIT(0, 0)
  "rmgpy.molecule.group.GroupBond", /*tp_name*/
  sizeof(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond), /*tp_basicsize*/
  0, /*tp_itemsize*/
  __pyx_tp_dealloc_5rmgpy_8molecule_5group_GroupBond, /*tp_dealloc*/
  0, /*tp_print*/
  0, /*tp_getattr*/
  0, /*tp_setattr*/
  #if PY_MAJOR_VERSION < 3
  0, /*tp_compare*/
  #else
  0, /*reserved*/
  #endif
  __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_5__repr__, /*tp_repr*/
  0, /*tp_as_number*/
  0, /*tp_as_sequence*/
  0, /*tp_as_mapping*/
  0, /*tp_hash*/
  0, /*tp_call*/
  __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_3__str__, /*tp_str*/
  0, /*tp_getattro*/
  0, /*tp_setattro*/
  0, /*tp_as_buffer*/
  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE|Py_TPFLAGS_HAVE_GC, /*tp_flags*/
  "\n    A bond group. This class is based on the :class:`Bond` class, except that\n    all attributes are lists rather than individual values. The allowed bond\n    types are given :ref:`here <bond-types>`. The attributes are:\n\n    =================== =================== ====================================\n    Attribute           Type                Description\n    =================== =================== ====================================\n    `order`             ``list``            The allowed bond orders (as character strings)\n    =================== =================== ====================================\n\n    Each list represents a logical OR construct, i.e. a bond will match the\n    group if it matches *any* item in the list.\n    ", /*tp_doc*/
  __pyx_tp_traverse_5rmgpy_8molecule_5group_GroupBond, /*tp_traverse*/
  __pyx_tp_clear_5rmgpy_8molecule_5group_GroupBond, /*tp_clear*/
  0, /*tp_richcompare*/
  0, /*tp_weaklistoffset*/
  0, /*tp_iter*/
  0, /*tp_iternext*/
  __pyx_methods_5rmgpy_8molecule_5group_GroupBond, /*tp_methods*/
  0, /*tp_members*/
  __pyx_getsets_5rmgpy_8molecule_5group_GroupBond, /*tp_getset*/
  0, /*tp_base*/
  0, /*tp_dict*/
  0, /*tp_descr_get*/
  0, /*tp_descr_set*/
  0, /*tp_dictoffset*/
  __pyx_pw_5rmgpy_8molecule_5group_9GroupBond_1__init__, /*tp_init*/
  0, /*tp_alloc*/
  __pyx_tp_new_5rmgpy_8molecule_5group_GroupBond, /*tp_new*/
  0, /*tp_free*/
  0, /*tp_is_gc*/
  0, /*tp_bases*/
  0, /*tp_mro*/
  0, /*tp_cache*/
  0, /*tp_subclasses*/
  0, /*tp_weaklist*/
  0, /*tp_del*/
  0, /*tp_version_tag*/
  #if PY_VERSION_HEX >= 0x030400a1
  0, /*tp_finalize*/
  #endif
};
static struct __pyx_vtabstruct_5rmgpy_8molecule_5group_Group __pyx_vtable_5rmgpy_8molecule_5group_Group;

static PyObject *__pyx_tp_new_5rmgpy_8molecule_5group_Group(PyTypeObject *t, PyObject *a, PyObject *k) {
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *p;
  PyObject *o = __pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_new(t, a, k);
  if (unlikely(!o)) return 0;
  p = ((struct __pyx_obj_5rmgpy_8molecule_5group_Group *)o);
  p->__pyx_base.__pyx_vtab = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph*)__pyx_vtabptr_5rmgpy_8molecule_5group_Group;
  p->multiplicity = ((PyObject*)Py_None); Py_INCREF(Py_None);
  return o;
}

static void __pyx_tp_dealloc_5rmgpy_8molecule_5group_Group(PyObject *o) {
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *p = (struct __pyx_obj_5rmgpy_8molecule_5group_Group *)o;
  #if PY_VERSION_HEX >= 0x030400a1
  if (unlikely(Py_TYPE(o)->tp_finalize) && !_PyGC_FINALIZED(o)) {
    if (PyObject_CallFinalizerFromDealloc(o)) return;
  }
  #endif
  PyObject_GC_UnTrack(o);
  Py_CLEAR(p->multiplicity);
  PyObject_GC_Track(o);
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Graph)) __pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_dealloc(o); else __Pyx_call_next_tp_dealloc(o, __pyx_tp_dealloc_5rmgpy_8molecule_5group_Group);
}

static int __pyx_tp_traverse_5rmgpy_8molecule_5group_Group(PyObject *o, visitproc v, void *a) {
  int e;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *p = (struct __pyx_obj_5rmgpy_8molecule_5group_Group *)o;
  e = ((likely(__pyx_ptype_5rmgpy_8molecule_5graph_Graph)) ? ((__pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_traverse) ? __pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_traverse(o, v, a) : 0) : __Pyx_call_next_tp_traverse(o, v, a, __pyx_tp_traverse_5rmgpy_8molecule_5group_Group)); if (e) return e;
  if (p->multiplicity) {
    e = (*v)(p->multiplicity, a); if (e) return e;
  }
  return 0;
}

static int __pyx_tp_clear_5rmgpy_8molecule_5group_Group(PyObject *o) {
  PyObject* tmp;
  struct __pyx_obj_5rmgpy_8molecule_5group_Group *p = (struct __pyx_obj_5rmgpy_8molecule_5group_Group *)o;
  if (likely(__pyx_ptype_5rmgpy_8molecule_5graph_Graph)) { if (__pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_clear) __pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_clear(o); } else __Pyx_call_next_tp_clear(o, __pyx_tp_clear_5rmgpy_8molecule_5group_Group);
  tmp = ((PyObject*)p->multiplicity);
  p->multiplicity = ((PyObject*)Py_None); Py_INCREF(Py_None);
  Py_XDECREF(tmp);
  return 0;
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_multiplicity(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_multiplicity(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_3__set__(o, v);
  }
  else {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_12multiplicity_5__del__(o);
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_carbonCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_carbonCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_11carbonCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_nitrogenCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_nitrogenCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_13nitrogenCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_oxygenCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_oxygenCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_11oxygenCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_sulfurCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_sulfurCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_11sulfurCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyObject *__pyx_getprop_5rmgpy_8molecule_5group_5Group_radicalCount(PyObject *o, CYTHON_UNUSED void *x) {
  return __pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_1__get__(o);
}

static int __pyx_setprop_5rmgpy_8molecule_5group_5Group_radicalCount(PyObject *o, PyObject *v, CYTHON_UNUSED void *x) {
  if (v) {
    return __pyx_pw_5rmgpy_8molecule_5group_5Group_12radicalCount_3__set__(o, v);
  }
  else {
    PyErr_SetString(PyExc_NotImplementedError, "__del__");
    return -1;
  }
}

static PyMethodDef __pyx_methods_5rmgpy_8molecule_5group_Group[] = {
  {"__reduce__", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_3__reduce__, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_2__reduce__},
  {"__getAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_5__getAtoms, METH_NOARGS, 0},
  {"__setAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_7__setAtoms, METH_O, 0},
  {"addAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_9addAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_8addAtom},
  {"addBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_11addBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_10addBond},
  {"getBonds", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_13getBonds, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_12getBonds},
  {"getBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_15getBond, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_14getBond},
  {"hasAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_17hasAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_16hasAtom},
  {"hasBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_19hasBond, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_18hasBond},
  {"removeAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_21removeAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_20removeAtom},
  {"removeBond", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_23removeBond, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_22removeBond},
  {"sortAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_25sortAtoms, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_24sortAtoms},
  {"copy", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_27copy, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_26copy},
  {"update", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_29update, METH_NOARGS, 0},
  {"merge", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_31merge, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_30merge},
  {"split", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_33split, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_32split},
  {"clearLabeledAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_35clearLabeledAtoms, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_34clearLabeledAtoms},
  {"containsLabeledAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_37containsLabeledAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_36containsLabeledAtom},
  {"getLabeledAtom", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_39getLabeledAtom, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_38getLabeledAtom},
  {"getLabeledAtoms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_41getLabeledAtoms, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_40getLabeledAtoms},
  {"fromAdjacencyList", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_43fromAdjacencyList, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_42fromAdjacencyList},
  {"toAdjacencyList", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_45toAdjacencyList, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_44toAdjacencyList},
  {"updateFingerprint", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_47updateFingerprint, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_46updateFingerprint},
  {"isIsomorphic", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_49isIsomorphic, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_48isIsomorphic},
  {"findIsomorphism", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_51findIsomorphism, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_50findIsomorphism},
  {"isSubgraphIsomorphic", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_53isSubgraphIsomorphic, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_52isSubgraphIsomorphic},
  {"findSubgraphIsomorphisms", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_55findSubgraphIsomorphisms, METH_VARARGS|METH_KEYWORDS, __pyx_doc_5rmgpy_8molecule_5group_5Group_54findSubgraphIsomorphisms},
  {"isIdentical", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_57isIdentical, METH_O, __pyx_doc_5rmgpy_8molecule_5group_5Group_56isIdentical},
  {"isAromaticRing", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_59isAromaticRing, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_58isAromaticRing},
  {"standardizeAtomType", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_61standardizeAtomType, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_60standardizeAtomType},
  {"addExplicitLigands", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_63addExplicitLigands, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_62addExplicitLigands},
  {"standardizeGroup", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_65standardizeGroup, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_64standardizeGroup},
  {"addImplicitAtomsFromAtomType", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_67addImplicitAtomsFromAtomType, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_66addImplicitAtomsFromAtomType},
  {"addImplicitBenzene", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_69addImplicitBenzene, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_68addImplicitBenzene},
  {"makeSampleMolecule", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_71makeSampleMolecule, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_70makeSampleMolecule},
  {"isBenzeneExplicit", (PyCFunction)__pyx_pw_5rmgpy_8molecule_5group_5Group_73isBenzeneExplicit, METH_NOARGS, __pyx_doc_5rmgpy_8molecule_5group_5Group_72isBenzeneExplicit},
  {0, 0, 0, 0}
};

static struct PyGetSetDef __pyx_getsets_5rmgpy_8molecule_5group_Group[] = {
  {(char *)"multiplicity", __pyx_getprop_5rmgpy_8molecule_5group_5Group_multiplicity, __pyx_setprop_5rmgpy_8molecule_5group_5Group_multiplicity, 0, 0},
  {(char *)"carbonCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_carbonCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_carbonCount, 0, 0},
  {(char *)"nitrogenCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_nitrogenCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_nitrogenCount, 0, 0},
  {(char *)"oxygenCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_oxygenCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_oxygenCount, 0, 0},
  {(char *)"sulfurCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_sulfurCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_sulfurCount, 0, 0},
  {(char *)"radicalCount", __pyx_getprop_5rmgpy_8molecule_5group_5Group_radicalCount, __pyx_setprop_5rmgpy_8molecule_5group_5Group_radicalCount, 0, 0},
  {0, 0, 0, 0, 0}
};

static PyTypeObject __pyx_type_5rmgpy_8molecule_5group_Group = {
  PyVarObject_HEAD_INIT(0, 0)
  "rmgpy.molecule.group.Group", /*tp_name*/
  sizeof(struct __pyx_obj_5rmgpy_8molecule_5group_Group), /*tp_basicsize*/
  0, /*tp_itemsize*/
  __pyx_tp_dealloc_5rmgpy_8molecule_5group_Group, /*tp_dealloc*/
  0, /*tp_print*/
  0, /*tp_getattr*/
  0, /*tp_setattr*/
  #if PY_MAJOR_VERSION < 3
  0, /*tp_compare*/
  #else
  0, /*reserved*/
  #endif
  0, /*tp_repr*/
  0, /*tp_as_number*/
  0, /*tp_as_sequence*/
  0, /*tp_as_mapping*/
  0, /*tp_hash*/
  0, /*tp_call*/
  0, /*tp_str*/
  0, /*tp_getattro*/
  0, /*tp_setattro*/
  0, /*tp_as_buffer*/
  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE|Py_TPFLAGS_HAVE_GC, /*tp_flags*/
  "\n    A representation of a molecular substructure group using a graph data\n    type, extending the :class:`Graph` class. The `atoms` and `bonds` attributes\n    are aliases for the `vertices` and `edges` attributes, and store \n    :class:`GroupAtom` and :class:`GroupBond` objects, respectively.\n    Corresponding alias methods have also been provided.\n    ", /*tp_doc*/
  __pyx_tp_traverse_5rmgpy_8molecule_5group_Group, /*tp_traverse*/
  __pyx_tp_clear_5rmgpy_8molecule_5group_Group, /*tp_clear*/
  0, /*tp_richcompare*/
  0, /*tp_weaklistoffset*/
  0, /*tp_iter*/
  0, /*tp_iternext*/
  __pyx_methods_5rmgpy_8molecule_5group_Group, /*tp_methods*/
  0, /*tp_members*/
  __pyx_getsets_5rmgpy_8molecule_5group_Group, /*tp_getset*/
  0, /*tp_base*/
  0, /*tp_dict*/
  0, /*tp_descr_get*/
  0, /*tp_descr_set*/
  0, /*tp_dictoffset*/
  __pyx_pw_5rmgpy_8molecule_5group_5Group_1__init__, /*tp_init*/
  0, /*tp_alloc*/
  __pyx_tp_new_5rmgpy_8molecule_5group_Group, /*tp_new*/
  0, /*tp_free*/
  0, /*tp_is_gc*/
  0, /*tp_bases*/
  0, /*tp_mro*/
  0, /*tp_cache*/
  0, /*tp_subclasses*/
  0, /*tp_weaklist*/
  0, /*tp_del*/
  0, /*tp_version_tag*/
  #if PY_VERSION_HEX >= 0x030400a1
  0, /*tp_finalize*/
  #endif
};

static PyMethodDef __pyx_methods[] = {
  {0, 0, 0, 0}
};

#if PY_MAJOR_VERSION >= 3
static struct PyModuleDef __pyx_moduledef = {
  #if PY_VERSION_HEX < 0x03020000
    { PyObject_HEAD_INIT(NULL) NULL, 0, NULL },
  #else
    PyModuleDef_HEAD_INIT,
  #endif
    "group",
    __pyx_k_This_module_provides_classes_an, /* m_doc */
    -1, /* m_size */
    __pyx_methods /* m_methods */,
    NULL, /* m_reload */
    NULL, /* m_traverse */
    NULL, /* m_clear */
    NULL /* m_free */
};
#endif

static __Pyx_StringTabEntry __pyx_string_tab[] = {
  {&__pyx_kp_s_, __pyx_k_, sizeof(__pyx_k_), 0, 0, 1, 0},
  {&__pyx_kp_s_0, __pyx_k_0, sizeof(__pyx_k_0), 0, 0, 1, 0},
  {&__pyx_kp_s_0_has_more_than_three_Cbf_atoms, __pyx_k_0_has_more_than_three_Cbf_atoms, sizeof(__pyx_k_0_has_more_than_three_Cbf_atoms), 0, 0, 1, 0},
  {&__pyx_n_s_ActionError, __pyx_k_ActionError, sizeof(__pyx_k_ActionError), 0, 0, 1, 1},
  {&__pyx_kp_s_An_exception_class_for_errors_t, __pyx_k_An_exception_class_for_errors_t, sizeof(__pyx_k_An_exception_class_for_errors_t), 0, 0, 1, 0},
  {&__pyx_kp_s_An_exception_class_when_encount, __pyx_k_An_exception_class_when_encount, sizeof(__pyx_k_An_exception_class_when_encount), 0, 0, 1, 0},
  {&__pyx_n_s_Ar, __pyx_k_Ar, sizeof(__pyx_k_Ar), 0, 0, 1, 1},
  {&__pyx_n_s_B, __pyx_k_B, sizeof(__pyx_k_B), 0, 0, 1, 1},
  {&__pyx_n_s_BREAK_BOND, __pyx_k_BREAK_BOND, sizeof(__pyx_k_BREAK_BOND), 0, 0, 1, 1},
  {&__pyx_n_s_C, __pyx_k_C, sizeof(__pyx_k_C), 0, 0, 1, 1},
  {&__pyx_n_s_CHANGE_BOND, __pyx_k_CHANGE_BOND, sizeof(__pyx_k_CHANGE_BOND), 0, 0, 1, 1},
  {&__pyx_n_s_CO, __pyx_k_CO, sizeof(__pyx_k_CO), 0, 0, 1, 1},
  {&__pyx_n_s_CS, __pyx_k_CS, sizeof(__pyx_k_CS), 0, 0, 1, 1},
  {&__pyx_n_s_Cb, __pyx_k_Cb, sizeof(__pyx_k_Cb), 0, 0, 1, 1},
  {&__pyx_n_s_Cbf, __pyx_k_Cbf, sizeof(__pyx_k_Cbf), 0, 0, 1, 1},
  {&__pyx_kp_s_Cbf_atom_in_0_has_a_bond_with_an, __pyx_k_Cbf_atom_in_0_has_a_bond_with_an, sizeof(__pyx_k_Cbf_atom_in_0_has_a_bond_with_an), 0, 0, 1, 0},
  {&__pyx_n_s_Cl, __pyx_k_Cl, sizeof(__pyx_k_Cl), 0, 0, 1, 1},
  {&__pyx_n_s_D, __pyx_k_D, sizeof(__pyx_k_D), 0, 0, 1, 1},
  {&__pyx_n_s_Edge, __pyx_k_Edge, sizeof(__pyx_k_Edge), 0, 0, 1, 1},
  {&__pyx_n_s_Exception, __pyx_k_Exception, sizeof(__pyx_k_Exception), 0, 0, 1, 1},
  {&__pyx_n_s_FORM_BOND, __pyx_k_FORM_BOND, sizeof(__pyx_k_FORM_BOND), 0, 0, 1, 1},
  {&__pyx_kp_u_Format_string_allocated_too_shor, __pyx_k_Format_string_allocated_too_shor, sizeof(__pyx_k_Format_string_allocated_too_shor), 0, 1, 0, 0},
  {&__pyx_kp_u_Format_string_allocated_too_shor_2, __pyx_k_Format_string_allocated_too_shor_2, sizeof(__pyx_k_Format_string_allocated_too_shor_2), 0, 1, 0, 0},
  {&__pyx_n_s_GAIN_PAIR, __pyx_k_GAIN_PAIR, sizeof(__pyx_k_GAIN_PAIR), 0, 0, 1, 1},
  {&__pyx_n_s_GAIN_RADICAL, __pyx_k_GAIN_RADICAL, sizeof(__pyx_k_GAIN_RADICAL), 0, 0, 1, 1},
  {&__pyx_kp_s_Got_a_0_object_for_parameter_oth, __pyx_k_Got_a_0_object_for_parameter_oth, sizeof(__pyx_k_Got_a_0_object_for_parameter_oth), 0, 0, 1, 0},
  {&__pyx_n_s_Graph, __pyx_k_Graph, sizeof(__pyx_k_Graph), 0, 0, 1, 1},
  {&__pyx_kp_s_GroupAtom_0_s, __pyx_k_GroupAtom_0_s, sizeof(__pyx_k_GroupAtom_0_s), 0, 0, 1, 0},
  {&__pyx_kp_s_GroupBond_0_r, __pyx_k_GroupBond_0_r, sizeof(__pyx_k_GroupBond_0_r), 0, 0, 1, 0},
  {&__pyx_n_s_Group_addImplicitBenzene_locals, __pyx_k_Group_addImplicitBenzene_locals, sizeof(__pyx_k_Group_addImplicitBenzene_locals), 0, 0, 1, 1},
  {&__pyx_n_s_Group_addImplicitBenzene_locals_2, __pyx_k_Group_addImplicitBenzene_locals_2, sizeof(__pyx_k_Group_addImplicitBenzene_locals_2), 0, 0, 1, 1},
  {&__pyx_n_s_Group_addImplicitBenzene_locals_3, __pyx_k_Group_addImplicitBenzene_locals_3, sizeof(__pyx_k_Group_addImplicitBenzene_locals_3), 0, 0, 1, 1},
  {&__pyx_n_s_Group_addImplicitBenzene_locals_4, __pyx_k_Group_addImplicitBenzene_locals_4, sizeof(__pyx_k_Group_addImplicitBenzene_locals_4), 0, 0, 1, 1},
  {&__pyx_n_s_Group_addImplicitBenzene_locals_5, __pyx_k_Group_addImplicitBenzene_locals_5, sizeof(__pyx_k_Group_addImplicitBenzene_locals_5), 0, 0, 1, 1},
  {&__pyx_n_s_Group_addImplicitBenzene_locals_6, __pyx_k_Group_addImplicitBenzene_locals_6, sizeof(__pyx_k_Group_addImplicitBenzene_locals_6), 0, 0, 1, 1},
  {&__pyx_n_s_H, __pyx_k_H, sizeof(__pyx_k_H), 0, 0, 1, 1},
  {&__pyx_n_s_He, __pyx_k_He, sizeof(__pyx_k_He), 0, 0, 1, 1},
  {&__pyx_n_s_ImplicitBenzeneError, __pyx_k_ImplicitBenzeneError, sizeof(__pyx_k_ImplicitBenzeneError), 0, 0, 1, 1},
  {&__pyx_n_s_LOSE_PAIR, __pyx_k_LOSE_PAIR, sizeof(__pyx_k_LOSE_PAIR), 0, 0, 1, 1},
  {&__pyx_n_s_LOSE_RADICAL, __pyx_k_LOSE_RADICAL, sizeof(__pyx_k_LOSE_RADICAL), 0, 0, 1, 1},
  {&__pyx_n_s_N, __pyx_k_N, sizeof(__pyx_k_N), 0, 0, 1, 1},
  {&__pyx_n_s_N1d, __pyx_k_N1d, sizeof(__pyx_k_N1d), 0, 0, 1, 1},
  {&__pyx_n_s_N3b, __pyx_k_N3b, sizeof(__pyx_k_N3b), 0, 0, 1, 1},
  {&__pyx_n_s_N5b, __pyx_k_N5b, sizeof(__pyx_k_N5b), 0, 0, 1, 1},
  {&__pyx_n_s_N5d, __pyx_k_N5d, sizeof(__pyx_k_N5d), 0, 0, 1, 1},
  {&__pyx_n_s_N5dd, __pyx_k_N5dd, sizeof(__pyx_k_N5dd), 0, 0, 1, 1},
  {&__pyx_n_s_N5s, __pyx_k_N5s, sizeof(__pyx_k_N5s), 0, 0, 1, 1},
  {&__pyx_n_s_N5t, __pyx_k_N5t, sizeof(__pyx_k_N5t), 0, 0, 1, 1},
  {&__pyx_n_s_Ne, __pyx_k_Ne, sizeof(__pyx_k_Ne), 0, 0, 1, 1},
  {&__pyx_kp_s_No_atom_in_the_functional_group, __pyx_k_No_atom_in_the_functional_group, sizeof(__pyx_k_No_atom_in_the_functional_group), 0, 0, 1, 0},
  {&__pyx_kp_u_Non_native_byte_order_not_suppor, __pyx_k_Non_native_byte_order_not_suppor, sizeof(__pyx_k_Non_native_byte_order_not_suppor), 0, 1, 0, 0},
  {&__pyx_n_s_O, __pyx_k_O, sizeof(__pyx_k_O), 0, 0, 1, 1},
  {&__pyx_n_s_Oa, __pyx_k_Oa, sizeof(__pyx_k_Oa), 0, 0, 1, 1},
  {&__pyx_n_s_Od, __pyx_k_Od, sizeof(__pyx_k_Od), 0, 0, 1, 1},
  {&__pyx_n_s_PeriodicSystem, __pyx_k_PeriodicSystem, sizeof(__pyx_k_PeriodicSystem), 0, 0, 1, 1},
  {&__pyx_kp_s_R_H, __pyx_k_R_H, sizeof(__pyx_k_R_H), 0, 0, 1, 0},
  {&__pyx_n_s_RuntimeError, __pyx_k_RuntimeError, sizeof(__pyx_k_RuntimeError), 0, 0, 1, 1},
  {&__pyx_n_s_S, __pyx_k_S, sizeof(__pyx_k_S), 0, 0, 1, 1},
  {&__pyx_n_s_Sa, __pyx_k_Sa, sizeof(__pyx_k_Sa), 0, 0, 1, 1},
  {&__pyx_n_s_Sd, __pyx_k_Sd, sizeof(__pyx_k_Sd), 0, 0, 1, 1},
  {&__pyx_n_s_Si, __pyx_k_Si, sizeof(__pyx_k_Si), 0, 0, 1, 1},
  {&__pyx_n_s_T, __pyx_k_T, sizeof(__pyx_k_T), 0, 0, 1, 1},
  {&__pyx_n_s_TypeError, __pyx_k_TypeError, sizeof(__pyx_k_TypeError), 0, 0, 1, 1},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_Inval, __pyx_k_Unable_to_update_GroupAtom_Inval, sizeof(__pyx_k_Unable_to_update_GroupAtom_Inval), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t, __pyx_k_Unable_to_update_GroupAtom_due_t, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_10, __pyx_k_Unable_to_update_GroupAtom_due_t_10, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_10), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_11, __pyx_k_Unable_to_update_GroupAtom_due_t_11, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_11), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_12, __pyx_k_Unable_to_update_GroupAtom_due_t_12, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_12), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_2, __pyx_k_Unable_to_update_GroupAtom_due_t_2, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_2), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_3, __pyx_k_Unable_to_update_GroupAtom_due_t_3, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_3), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_4, __pyx_k_Unable_to_update_GroupAtom_due_t_4, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_4), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_5, __pyx_k_Unable_to_update_GroupAtom_due_t_5, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_5), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_6, __pyx_k_Unable_to_update_GroupAtom_due_t_6, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_6), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_7, __pyx_k_Unable_to_update_GroupAtom_due_t_7, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_7), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_8, __pyx_k_Unable_to_update_GroupAtom_due_t_8, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_8), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupAtom_due_t_9, __pyx_k_Unable_to_update_GroupAtom_due_t_9, sizeof(__pyx_k_Unable_to_update_GroupAtom_due_t_9), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupBond_Inval, __pyx_k_Unable_to_update_GroupBond_Inval, sizeof(__pyx_k_Unable_to_update_GroupBond_Inval), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupBond_due_t, __pyx_k_Unable_to_update_GroupBond_due_t, sizeof(__pyx_k_Unable_to_update_GroupBond_due_t), 0, 0, 1, 0},
  {&__pyx_kp_s_Unable_to_update_GroupBond_due_t_2, __pyx_k_Unable_to_update_GroupBond_due_t_2, sizeof(__pyx_k_Unable_to_update_GroupBond_due_t_2), 0, 0, 1, 0},
  {&__pyx_kp_s_Users_agnes_PycharmProjects_RMG, __pyx_k_Users_agnes_PycharmProjects_RMG, sizeof(__pyx_k_Users_agnes_PycharmProjects_RMG), 0, 0, 1, 0},
  {&__pyx_n_s_ValueError, __pyx_k_ValueError, sizeof(__pyx_k_ValueError), 0, 0, 1, 1},
  {&__pyx_n_s_Vertex, __pyx_k_Vertex, sizeof(__pyx_k_Vertex), 0, 0, 1, 1},
  {&__pyx_kp_s__2, __pyx_k__2, sizeof(__pyx_k__2), 0, 0, 1, 0},
  {&__pyx_n_s__26, __pyx_k__26, sizeof(__pyx_k__26), 0, 0, 1, 1},
  {&__pyx_n_s_addAtom, __pyx_k_addAtom, sizeof(__pyx_k_addAtom), 0, 0, 1, 1},
  {&__pyx_n_s_addBenzeneAtomToGroup, __pyx_k_addBenzeneAtomToGroup, sizeof(__pyx_k_addBenzeneAtomToGroup), 0, 0, 1, 1},
  {&__pyx_n_s_addBond, __pyx_k_addBond, sizeof(__pyx_k_addBond), 0, 0, 1, 1},
  {&__pyx_n_s_addCbAtomToRing, __pyx_k_addCbAtomToRing, sizeof(__pyx_k_addCbAtomToRing), 0, 0, 1, 1},
  {&__pyx_n_s_addExplicitLigands, __pyx_k_addExplicitLigands, sizeof(__pyx_k_addExplicitLigands), 0, 0, 1, 1},
  {&__pyx_n_s_addImplicitAtomsFromAtomType, __pyx_k_addImplicitAtomsFromAtomType, sizeof(__pyx_k_addImplicitAtomsFromAtomType), 0, 0, 1, 1},
  {&__pyx_n_s_addImplicitBenzene, __pyx_k_addImplicitBenzene, sizeof(__pyx_k_addImplicitBenzene), 0, 0, 1, 1},
  {&__pyx_n_s_added, __pyx_k_added, sizeof(__pyx_k_added), 0, 0, 1, 1},
  {&__pyx_n_s_adjlist, __pyx_k_adjlist, sizeof(__pyx_k_adjlist), 0, 0, 1, 1},
  {&__pyx_n_s_all, __pyx_k_all, sizeof(__pyx_k_all), 0, 0, 1, 1},
  {&__pyx_n_s_allElements, __pyx_k_allElements, sizeof(__pyx_k_allElements), 0, 0, 1, 1},
  {&__pyx_n_s_any, __pyx_k_any, sizeof(__pyx_k_any), 0, 0, 1, 1},
  {&__pyx_n_s_append, __pyx_k_append, sizeof(__pyx_k_append), 0, 0, 1, 1},
  {&__pyx_n_s_applyAction, __pyx_k_applyAction, sizeof(__pyx_k_applyAction), 0, 0, 1, 1},
  {&__pyx_n_s_atom, __pyx_k_atom, sizeof(__pyx_k_atom), 0, 0, 1, 1},
  {&__pyx_n_s_atom1, __pyx_k_atom1, sizeof(__pyx_k_atom1), 0, 0, 1, 1},
  {&__pyx_n_s_atom2, __pyx_k_atom2, sizeof(__pyx_k_atom2), 0, 0, 1, 1},
  {&__pyx_n_s_atomList, __pyx_k_atomList, sizeof(__pyx_k_atomList), 0, 0, 1, 1},
  {&__pyx_n_s_atomType, __pyx_k_atomType, sizeof(__pyx_k_atomType), 0, 0, 1, 1},
  {&__pyx_n_s_atomTypes, __pyx_k_atomTypes, sizeof(__pyx_k_atomTypes), 0, 0, 1, 1},
  {&__pyx_n_s_atoms, __pyx_k_atoms, sizeof(__pyx_k_atoms), 0, 0, 1, 1},
  {&__pyx_n_s_atomtype, __pyx_k_atomtype, sizeof(__pyx_k_atomtype), 0, 0, 1, 1},
  {&__pyx_n_s_benzeneBonds, __pyx_k_benzeneBonds, sizeof(__pyx_k_benzeneBonds), 0, 0, 1, 1},
  {&__pyx_n_s_bond, __pyx_k_bond, sizeof(__pyx_k_bond), 0, 0, 1, 1},
  {&__pyx_n_s_bond12, __pyx_k_bond12, sizeof(__pyx_k_bond12), 0, 0, 1, 1},
  {&__pyx_n_s_bonds, __pyx_k_bonds, sizeof(__pyx_k_bonds), 0, 0, 1, 1},
  {&__pyx_n_s_breakBond, __pyx_k_breakBond, sizeof(__pyx_k_breakBond), 0, 0, 1, 1},
  {&__pyx_n_s_breakBond_2, __pyx_k_breakBond_2, sizeof(__pyx_k_breakBond_2), 0, 0, 1, 1},
  {&__pyx_n_s_cbAtom, __pyx_k_cbAtom, sizeof(__pyx_k_cbAtom), 0, 0, 1, 1},
  {&__pyx_n_s_cbAtomList, __pyx_k_cbAtomList, sizeof(__pyx_k_cbAtomList), 0, 0, 1, 1},
  {&__pyx_n_s_cbf, __pyx_k_cbf, sizeof(__pyx_k_cbf), 0, 0, 1, 1},
  {&__pyx_n_s_cbfAtom, __pyx_k_cbfAtom, sizeof(__pyx_k_cbfAtom), 0, 0, 1, 1},
  {&__pyx_n_s_cbfAtomList, __pyx_k_cbfAtomList, sizeof(__pyx_k_cbfAtomList), 0, 0, 1, 1},
  {&__pyx_n_s_cbfAtomList1, __pyx_k_cbfAtomList1, sizeof(__pyx_k_cbfAtomList1), 0, 0, 1, 1},
  {&__pyx_n_s_cbfAtomList2, __pyx_k_cbfAtomList2, sizeof(__pyx_k_cbfAtomList2), 0, 0, 1, 1},
  {&__pyx_n_s_changeBond, __pyx_k_changeBond, sizeof(__pyx_k_changeBond), 0, 0, 1, 1},
  {&__pyx_n_s_charge, __pyx_k_charge, sizeof(__pyx_k_charge), 0, 0, 1, 1},
  {&__pyx_n_s_checkSet, __pyx_k_checkSet, sizeof(__pyx_k_checkSet), 0, 0, 1, 1},
  {&__pyx_n_s_class, __pyx_k_class, sizeof(__pyx_k_class), 0, 0, 1, 1},
  {&__pyx_n_s_classifyBenzeneCarbons, __pyx_k_classifyBenzeneCarbons, sizeof(__pyx_k_classifyBenzeneCarbons), 0, 0, 1, 1},
  {&__pyx_n_s_clearLabeledAtoms, __pyx_k_clearLabeledAtoms, sizeof(__pyx_k_clearLabeledAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_connectedCbfs, __pyx_k_connectedCbfs, sizeof(__pyx_k_connectedCbfs), 0, 0, 1, 1},
  {&__pyx_n_s_connectingAtom, __pyx_k_connectingAtom, sizeof(__pyx_k_connectingAtom), 0, 0, 1, 1},
  {&__pyx_n_s_connectivity1, __pyx_k_connectivity1, sizeof(__pyx_k_connectivity1), 0, 0, 1, 1},
  {&__pyx_n_s_connectivity2, __pyx_k_connectivity2, sizeof(__pyx_k_connectivity2), 0, 0, 1, 1},
  {&__pyx_n_s_connectivity3, __pyx_k_connectivity3, sizeof(__pyx_k_connectivity3), 0, 0, 1, 1},
  {&__pyx_n_s_containsLabeledAtom, __pyx_k_containsLabeledAtom, sizeof(__pyx_k_containsLabeledAtom), 0, 0, 1, 1},
  {&__pyx_n_s_copy, __pyx_k_copy, sizeof(__pyx_k_copy), 0, 0, 1, 1},
  {&__pyx_n_s_decrementBond, __pyx_k_decrementBond, sizeof(__pyx_k_decrementBond), 0, 0, 1, 1},
  {&__pyx_n_s_decrementLonePair, __pyx_k_decrementLonePair, sizeof(__pyx_k_decrementLonePair), 0, 0, 1, 1},
  {&__pyx_n_s_decrementRadical, __pyx_k_decrementRadical, sizeof(__pyx_k_decrementRadical), 0, 0, 1, 1},
  {&__pyx_n_s_deep, __pyx_k_deep, sizeof(__pyx_k_deep), 0, 0, 1, 1},
  {&__pyx_n_s_deepcopy, __pyx_k_deepcopy, sizeof(__pyx_k_deepcopy), 0, 0, 1, 1},
  {&__pyx_n_s_doc, __pyx_k_doc, sizeof(__pyx_k_doc), 0, 0, 1, 1},
  {&__pyx_n_s_edges, __pyx_k_edges, sizeof(__pyx_k_edges), 0, 0, 1, 1},
  {&__pyx_n_s_element, __pyx_k_element, sizeof(__pyx_k_element), 0, 0, 1, 1},
  {&__pyx_n_s_enumerate, __pyx_k_enumerate, sizeof(__pyx_k_enumerate), 0, 0, 1, 1},
  {&__pyx_n_s_equivalent, __pyx_k_equivalent, sizeof(__pyx_k_equivalent), 0, 0, 1, 1},
  {&__pyx_n_s_fbBonds, __pyx_k_fbBonds, sizeof(__pyx_k_fbBonds), 0, 0, 1, 1},
  {&__pyx_n_s_findIsomorphism, __pyx_k_findIsomorphism, sizeof(__pyx_k_findIsomorphism), 0, 0, 1, 1},
  {&__pyx_n_s_findSubgraphIsomorphisms, __pyx_k_findSubgraphIsomorphisms, sizeof(__pyx_k_findSubgraphIsomorphisms), 0, 0, 1, 1},
  {&__pyx_n_s_formBond, __pyx_k_formBond, sizeof(__pyx_k_formBond), 0, 0, 1, 1},
  {&__pyx_n_s_formBond_2, __pyx_k_formBond_2, sizeof(__pyx_k_formBond_2), 0, 0, 1, 1},
  {&__pyx_n_s_format, __pyx_k_format, sizeof(__pyx_k_format), 0, 0, 1, 1},
  {&__pyx_n_s_fromAdjacencyList, __pyx_k_fromAdjacencyList, sizeof(__pyx_k_fromAdjacencyList), 0, 0, 1, 1},
  {&__pyx_n_s_gainPair, __pyx_k_gainPair, sizeof(__pyx_k_gainPair), 0, 0, 1, 1},
  {&__pyx_n_s_gainRadical, __pyx_k_gainRadical, sizeof(__pyx_k_gainRadical), 0, 0, 1, 1},
  {&__pyx_n_s_getAllCyclesOfSize, __pyx_k_getAllCyclesOfSize, sizeof(__pyx_k_getAllCyclesOfSize), 0, 0, 1, 1},
  {&__pyx_n_s_getAtoms, __pyx_k_getAtoms, sizeof(__pyx_k_getAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_getBond, __pyx_k_getBond, sizeof(__pyx_k_getBond), 0, 0, 1, 1},
  {&__pyx_n_s_getBondOrdersForAtom, __pyx_k_getBondOrdersForAtom, sizeof(__pyx_k_getBondOrdersForAtom), 0, 0, 1, 1},
  {&__pyx_n_s_getBonds, __pyx_k_getBonds, sizeof(__pyx_k_getBonds), 0, 0, 1, 1},
  {&__pyx_n_s_getFeatures, __pyx_k_getFeatures, sizeof(__pyx_k_getFeatures), 0, 0, 1, 1},
  {&__pyx_n_s_getLabeledAtom, __pyx_k_getLabeledAtom, sizeof(__pyx_k_getLabeledAtom), 0, 0, 1, 1},
  {&__pyx_n_s_getLabeledAtoms, __pyx_k_getLabeledAtoms, sizeof(__pyx_k_getLabeledAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_graph, __pyx_k_graph, sizeof(__pyx_k_graph), 0, 0, 1, 1},
  {&__pyx_n_s_group, __pyx_k_group, sizeof(__pyx_k_group), 0, 0, 1, 1},
  {&__pyx_n_s_hasAtom, __pyx_k_hasAtom, sizeof(__pyx_k_hasAtom), 0, 0, 1, 1},
  {&__pyx_n_s_hasBond, __pyx_k_hasBond, sizeof(__pyx_k_hasBond), 0, 0, 1, 1},
  {&__pyx_n_s_hasWildCards, __pyx_k_hasWildCards, sizeof(__pyx_k_hasWildCards), 0, 0, 1, 1},
  {&__pyx_n_s_hasWildcards, __pyx_k_hasWildcards, sizeof(__pyx_k_hasWildcards), 0, 0, 1, 1},
  {&__pyx_n_s_import, __pyx_k_import, sizeof(__pyx_k_import), 0, 0, 1, 1},
  {&__pyx_n_s_incrementBond, __pyx_k_incrementBond, sizeof(__pyx_k_incrementBond), 0, 0, 1, 1},
  {&__pyx_n_s_incrementLonePair, __pyx_k_incrementLonePair, sizeof(__pyx_k_incrementLonePair), 0, 0, 1, 1},
  {&__pyx_n_s_incrementRadical, __pyx_k_incrementRadical, sizeof(__pyx_k_incrementRadical), 0, 0, 1, 1},
  {&__pyx_n_s_init, __pyx_k_init, sizeof(__pyx_k_init), 0, 0, 1, 1},
  {&__pyx_n_s_initialMap, __pyx_k_initialMap, sizeof(__pyx_k_initialMap), 0, 0, 1, 1},
  {&__pyx_n_s_isAromaticRing, __pyx_k_isAromaticRing, sizeof(__pyx_k_isAromaticRing), 0, 0, 1, 1},
  {&__pyx_n_s_isBenzene, __pyx_k_isBenzene, sizeof(__pyx_k_isBenzene), 0, 0, 1, 1},
  {&__pyx_n_s_isBenzeneExplicit, __pyx_k_isBenzeneExplicit, sizeof(__pyx_k_isBenzeneExplicit), 0, 0, 1, 1},
  {&__pyx_n_s_isDouble, __pyx_k_isDouble, sizeof(__pyx_k_isDouble), 0, 0, 1, 1},
  {&__pyx_n_s_isIdentical, __pyx_k_isIdentical, sizeof(__pyx_k_isIdentical), 0, 0, 1, 1},
  {&__pyx_n_s_isIsomorphic, __pyx_k_isIsomorphic, sizeof(__pyx_k_isIsomorphic), 0, 0, 1, 1},
  {&__pyx_n_s_isOxygen, __pyx_k_isOxygen, sizeof(__pyx_k_isOxygen), 0, 0, 1, 1},
  {&__pyx_n_s_isSingle, __pyx_k_isSingle, sizeof(__pyx_k_isSingle), 0, 0, 1, 1},
  {&__pyx_n_s_isSpecificCaseOf, __pyx_k_isSpecificCaseOf, sizeof(__pyx_k_isSpecificCaseOf), 0, 0, 1, 1},
  {&__pyx_n_s_isSubgraphIsomorphic, __pyx_k_isSubgraphIsomorphic, sizeof(__pyx_k_isSubgraphIsomorphic), 0, 0, 1, 1},
  {&__pyx_n_s_isSulfur, __pyx_k_isSulfur, sizeof(__pyx_k_isSulfur), 0, 0, 1, 1},
  {&__pyx_n_s_isTriple, __pyx_k_isTriple, sizeof(__pyx_k_isTriple), 0, 0, 1, 1},
  {&__pyx_n_s_issuperset, __pyx_k_issuperset, sizeof(__pyx_k_issuperset), 0, 0, 1, 1},
  {&__pyx_n_s_iteritems, __pyx_k_iteritems, sizeof(__pyx_k_iteritems), 0, 0, 1, 1},
  {&__pyx_n_s_join, __pyx_k_join, sizeof(__pyx_k_join), 0, 0, 1, 1},
  {&__pyx_n_s_keys, __pyx_k_keys, sizeof(__pyx_k_keys), 0, 0, 1, 1},
  {&__pyx_n_s_label, __pyx_k_label, sizeof(__pyx_k_label), 0, 0, 1, 1},
  {&__pyx_n_s_labelsOfCarbonAtomTypes, __pyx_k_labelsOfCarbonAtomTypes, sizeof(__pyx_k_labelsOfCarbonAtomTypes), 0, 0, 1, 1},
  {&__pyx_n_s_lonePairs, __pyx_k_lonePairs, sizeof(__pyx_k_lonePairs), 0, 0, 1, 1},
  {&__pyx_n_s_losePair, __pyx_k_losePair, sizeof(__pyx_k_losePair), 0, 0, 1, 1},
  {&__pyx_n_s_loseRadical, __pyx_k_loseRadical, sizeof(__pyx_k_loseRadical), 0, 0, 1, 1},
  {&__pyx_n_s_main, __pyx_k_main, sizeof(__pyx_k_main), 0, 0, 1, 1},
  {&__pyx_n_s_makeBond, __pyx_k_makeBond, sizeof(__pyx_k_makeBond), 0, 0, 1, 1},
  {&__pyx_n_s_makeSampleAtom, __pyx_k_makeSampleAtom, sizeof(__pyx_k_makeSampleAtom), 0, 0, 1, 1},
  {&__pyx_n_s_makeSampleMolecule, __pyx_k_makeSampleMolecule, sizeof(__pyx_k_makeSampleMolecule), 0, 0, 1, 1},
  {&__pyx_n_s_matchList1, __pyx_k_matchList1, sizeof(__pyx_k_matchList1), 0, 0, 1, 1},
  {&__pyx_n_s_matchList2, __pyx_k_matchList2, sizeof(__pyx_k_matchList2), 0, 0, 1, 1},
  {&__pyx_n_s_matchList3, __pyx_k_matchList3, sizeof(__pyx_k_matchList3), 0, 0, 1, 1},
  {&__pyx_n_s_matchList4, __pyx_k_matchList4, sizeof(__pyx_k_matchList4), 0, 0, 1, 1},
  {&__pyx_n_s_merge, __pyx_k_merge, sizeof(__pyx_k_merge), 0, 0, 1, 1},
  {&__pyx_n_s_mergeOverlappingBenzeneRings, __pyx_k_mergeOverlappingBenzeneRings, sizeof(__pyx_k_mergeOverlappingBenzeneRings), 0, 0, 1, 1},
  {&__pyx_n_s_mergedRing, __pyx_k_mergedRing, sizeof(__pyx_k_mergedRing), 0, 0, 1, 1},
  {&__pyx_n_s_metaclass, __pyx_k_metaclass, sizeof(__pyx_k_metaclass), 0, 0, 1, 1},
  {&__pyx_n_s_module, __pyx_k_module, sizeof(__pyx_k_module), 0, 0, 1, 1},
  {&__pyx_n_s_mol, __pyx_k_mol, sizeof(__pyx_k_mol), 0, 0, 1, 1},
  {&__pyx_n_s_molecule, __pyx_k_molecule, sizeof(__pyx_k_molecule), 0, 0, 1, 1},
  {&__pyx_n_s_multiplicity, __pyx_k_multiplicity, sizeof(__pyx_k_multiplicity), 0, 0, 1, 1},
  {&__pyx_kp_u_ndarray_is_not_C_contiguous, __pyx_k_ndarray_is_not_C_contiguous, sizeof(__pyx_k_ndarray_is_not_C_contiguous), 0, 1, 0, 0},
  {&__pyx_kp_u_ndarray_is_not_Fortran_contiguou, __pyx_k_ndarray_is_not_Fortran_contiguou, sizeof(__pyx_k_ndarray_is_not_Fortran_contiguou), 0, 1, 0, 0},
  {&__pyx_n_s_newAtom, __pyx_k_newAtom, sizeof(__pyx_k_newAtom), 0, 0, 1, 1},
  {&__pyx_n_s_newBond, __pyx_k_newBond, sizeof(__pyx_k_newBond), 0, 0, 1, 1},
  {&__pyx_n_s_newRing, __pyx_k_newRing, sizeof(__pyx_k_newRing), 0, 0, 1, 1},
  {&__pyx_n_s_nonSpecifics, __pyx_k_nonSpecifics, sizeof(__pyx_k_nonSpecifics), 0, 0, 1, 1},
  {&__pyx_n_s_od, __pyx_k_od, sizeof(__pyx_k_od), 0, 0, 1, 1},
  {&__pyx_n_s_order, __pyx_k_order, sizeof(__pyx_k_order), 0, 0, 1, 1},
  {&__pyx_n_s_other, __pyx_k_other, sizeof(__pyx_k_other), 0, 0, 1, 1},
  {&__pyx_n_s_partners, __pyx_k_partners, sizeof(__pyx_k_partners), 0, 0, 1, 1},
  {&__pyx_n_s_pop, __pyx_k_pop, sizeof(__pyx_k_pop), 0, 0, 1, 1},
  {&__pyx_n_s_prepare, __pyx_k_prepare, sizeof(__pyx_k_prepare), 0, 0, 1, 1},
  {&__pyx_n_s_previousLength, __pyx_k_previousLength, sizeof(__pyx_k_previousLength), 0, 0, 1, 1},
  {&__pyx_n_s_property, __pyx_k_property, sizeof(__pyx_k_property), 0, 0, 1, 1},
  {&__pyx_n_s_pyx_vtable, __pyx_k_pyx_vtable, sizeof(__pyx_k_pyx_vtable), 0, 0, 1, 1},
  {&__pyx_n_s_qualname, __pyx_k_qualname, sizeof(__pyx_k_qualname), 0, 0, 1, 1},
  {&__pyx_n_s_radicalElectrons, __pyx_k_radicalElectrons, sizeof(__pyx_k_radicalElectrons), 0, 0, 1, 1},
  {&__pyx_n_s_range, __pyx_k_range, sizeof(__pyx_k_range), 0, 0, 1, 1},
  {&__pyx_n_s_remove, __pyx_k_remove, sizeof(__pyx_k_remove), 0, 0, 1, 1},
  {&__pyx_n_s_removeAtom, __pyx_k_removeAtom, sizeof(__pyx_k_removeAtom), 0, 0, 1, 1},
  {&__pyx_n_s_removeBond, __pyx_k_removeBond, sizeof(__pyx_k_removeBond), 0, 0, 1, 1},
  {&__pyx_n_s_ring, __pyx_k_ring, sizeof(__pyx_k_ring), 0, 0, 1, 1},
  {&__pyx_n_s_ring1, __pyx_k_ring1, sizeof(__pyx_k_ring1), 0, 0, 1, 1},
  {&__pyx_n_s_ring2, __pyx_k_ring2, sizeof(__pyx_k_ring2), 0, 0, 1, 1},
  {&__pyx_n_s_rmgpy_molecule_group, __pyx_k_rmgpy_molecule_group, sizeof(__pyx_k_rmgpy_molecule_group), 0, 0, 1, 1},
  {&__pyx_n_s_rmgpy_molecule_molecule, __pyx_k_rmgpy_molecule_molecule, sizeof(__pyx_k_rmgpy_molecule_molecule), 0, 0, 1, 1},
  {&__pyx_n_s_setAtoms, __pyx_k_setAtoms, sizeof(__pyx_k_setAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_sortAtoms, __pyx_k_sortAtoms, sizeof(__pyx_k_sortAtoms), 0, 0, 1, 1},
  {&__pyx_n_s_sortByConnectivity, __pyx_k_sortByConnectivity, sizeof(__pyx_k_sortByConnectivity), 0, 0, 1, 1},
  {&__pyx_n_s_sortedAtomList, __pyx_k_sortedAtomList, sizeof(__pyx_k_sortedAtomList), 0, 0, 1, 1},
  {&__pyx_n_s_sortingLabel, __pyx_k_sortingLabel, sizeof(__pyx_k_sortingLabel), 0, 0, 1, 1},
  {&__pyx_n_s_specific, __pyx_k_specific, sizeof(__pyx_k_specific), 0, 0, 1, 1},
  {&__pyx_n_s_split, __pyx_k_split, sizeof(__pyx_k_split), 0, 0, 1, 1},
  {&__pyx_n_s_standardizeAtomType, __pyx_k_standardizeAtomType, sizeof(__pyx_k_standardizeAtomType), 0, 0, 1, 1},
  {&__pyx_n_s_standardizeGroup, __pyx_k_standardizeGroup, sizeof(__pyx_k_standardizeGroup), 0, 0, 1, 1},
  {&__pyx_n_s_subList, __pyx_k_subList, sizeof(__pyx_k_subList), 0, 0, 1, 1},
  {&__pyx_n_s_subSet, __pyx_k_subSet, sizeof(__pyx_k_subSet), 0, 0, 1, 1},
  {&__pyx_n_s_superList, __pyx_k_superList, sizeof(__pyx_k_superList), 0, 0, 1, 1},
  {&__pyx_n_s_superSet, __pyx_k_superSet, sizeof(__pyx_k_superSet), 0, 0, 1, 1},
  {&__pyx_n_s_symbol, __pyx_k_symbol, sizeof(__pyx_k_symbol), 0, 0, 1, 1},
  {&__pyx_n_s_test, __pyx_k_test, sizeof(__pyx_k_test), 0, 0, 1, 1},
  {&__pyx_n_s_toAdjacencyList, __pyx_k_toAdjacencyList, sizeof(__pyx_k_toAdjacencyList), 0, 0, 1, 1},
  {&__pyx_kp_u_unknown_dtype_code_in_numpy_pxd, __pyx_k_unknown_dtype_code_in_numpy_pxd, sizeof(__pyx_k_unknown_dtype_code_in_numpy_pxd), 0, 1, 0, 0},
  {&__pyx_n_s_update, __pyx_k_update, sizeof(__pyx_k_update), 0, 0, 1, 1},
  {&__pyx_n_s_updateCharge, __pyx_k_updateCharge, sizeof(__pyx_k_updateCharge), 0, 0, 1, 1},
  {&__pyx_n_s_updateFingerprint, __pyx_k_updateFingerprint, sizeof(__pyx_k_updateFingerprint), 0, 0, 1, 1},
  {&__pyx_n_s_upper, __pyx_k_upper, sizeof(__pyx_k_upper), 0, 0, 1, 1},
  {&__pyx_n_s_valence_electrons, __pyx_k_valence_electrons, sizeof(__pyx_k_valence_electrons), 0, 0, 1, 1},
  {&__pyx_n_s_values, __pyx_k_values, sizeof(__pyx_k_values), 0, 0, 1, 1},
  {&__pyx_n_s_vertices, __pyx_k_vertices, sizeof(__pyx_k_vertices), 0, 0, 1, 1},
  {&__pyx_n_s_x, __pyx_k_x, sizeof(__pyx_k_x), 0, 0, 1, 1},
  {&__pyx_n_s_x1, __pyx_k_x1, sizeof(__pyx_k_x1), 0, 0, 1, 1},
  {&__pyx_n_s_x2, __pyx_k_x2, sizeof(__pyx_k_x2), 0, 0, 1, 1},
  {&__pyx_n_s_zip, __pyx_k_zip, sizeof(__pyx_k_zip), 0, 0, 1, 1},
  {0, 0, 0, 0, 0, 0, 0}
};
static int __Pyx_InitCachedBuiltins(void) {
  __pyx_builtin_Exception = __Pyx_GetBuiltinName(__pyx_n_s_Exception); if (!__pyx_builtin_Exception) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 46; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_property = __Pyx_GetBuiltinName(__pyx_n_s_property); if (!__pyx_builtin_property) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 136; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_range = __Pyx_GetBuiltinName(__pyx_n_s_range); if (!__pyx_builtin_range) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 90; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_any = __Pyx_GetBuiltinName(__pyx_n_s_any); if (!__pyx_builtin_any) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 203; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_zip = __Pyx_GetBuiltinName(__pyx_n_s_zip); if (!__pyx_builtin_zip) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 236; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_all = __Pyx_GetBuiltinName(__pyx_n_s_all); if (!__pyx_builtin_all) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 409; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_ValueError = __Pyx_GetBuiltinName(__pyx_n_s_ValueError); if (!__pyx_builtin_ValueError) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 855; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_TypeError = __Pyx_GetBuiltinName(__pyx_n_s_TypeError); if (!__pyx_builtin_TypeError) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 946; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_enumerate = __Pyx_GetBuiltinName(__pyx_n_s_enumerate); if (!__pyx_builtin_enumerate) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1094; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_builtin_RuntimeError = __Pyx_GetBuiltinName(__pyx_n_s_RuntimeError); if (!__pyx_builtin_RuntimeError) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 799; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  return 0;
  __pyx_L1_error:;
  return -1;
}

static int __Pyx_InitCachedConstants(void) {
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__Pyx_InitCachedConstants", 0);

  /* "rmgpy/molecule/group.py":493
 * 
 *         if newAtom.lonePairs == -100:
 *             if atomtype in [atomTypes[x] for x in ['N5d', 'N5dd', 'N5t', 'N5b', 'N5s']]:             # <<<<<<<<<<<<<<
 *                 newAtom.lonePairs = 0
 *                 newAtom.charge = 1
 */
  __pyx_tuple__3 = PyTuple_Pack(5, __pyx_n_s_N5d, __pyx_n_s_N5dd, __pyx_n_s_N5t, __pyx_n_s_N5b, __pyx_n_s_N5s); if (unlikely(!__pyx_tuple__3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 493; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__3);
  __Pyx_GIVEREF(__pyx_tuple__3);

  /* "rmgpy/molecule/group.py":1283
 *         """
 * 
 *         def classifyBenzeneCarbons(group, partners = []):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */
  __pyx_tuple__4 = PyTuple_Pack(16, __pyx_n_s_group, __pyx_n_s_partners, __pyx_n_s_cbAtomList, __pyx_n_s_cbfAtomList, __pyx_n_s_cbfAtomList1, __pyx_n_s_cbfAtomList2, __pyx_n_s_connectedCbfs, __pyx_n_s_labelsOfCarbonAtomTypes, __pyx_n_s_atom, __pyx_n_s_benzeneBonds, __pyx_n_s_atom2, __pyx_n_s_bond12, __pyx_n_s_cbfAtom, __pyx_n_s_fbBonds, __pyx_n_s_bond, __pyx_n_s_x); if (unlikely(!__pyx_tuple__4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__4);
  __Pyx_GIVEREF(__pyx_tuple__4);
  __pyx_codeobj__5 = (PyObject*)__Pyx_PyCode_New(2, 0, 16, 0, 0, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple__4, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_Users_agnes_PycharmProjects_RMG, __pyx_n_s_classifyBenzeneCarbons, 1283, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1283; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":1341
 *             return (cbAtomList, cbfAtomList, cbfAtomList1, cbfAtomList2, connectedCbfs)
 * 
 *         def addBenzeneAtomToGroup(group, connectingAtom, cbf=False):             # <<<<<<<<<<<<<<
 *             """
 *             This function adds a benzene atom to the group
 */
  __pyx_tuple__6 = PyTuple_Pack(5, __pyx_n_s_group, __pyx_n_s_connectingAtom, __pyx_n_s_cbf, __pyx_n_s_newAtom, __pyx_n_s_newBond); if (unlikely(!__pyx_tuple__6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1341; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__6);
  __Pyx_GIVEREF(__pyx_tuple__6);
  __pyx_codeobj__7 = (PyObject*)__Pyx_PyCode_New(3, 0, 5, 0, 0, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple__6, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_Users_agnes_PycharmProjects_RMG, __pyx_n_s_addBenzeneAtomToGroup, 1341, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__7)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1341; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_tuple__8 = PyTuple_Pack(1, ((PyObject *)Py_False)); if (unlikely(!__pyx_tuple__8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1341; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__8);
  __Pyx_GIVEREF(__pyx_tuple__8);

  /* "rmgpy/molecule/group.py":1363
 *             return (group, newAtom)
 * 
 *         def sortByConnectivity(atomList):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */
  __pyx_tuple__9 = PyTuple_Pack(7, __pyx_n_s_atomList, __pyx_n_s_sortedAtomList, __pyx_n_s_previousLength, __pyx_n_s_atom1, __pyx_n_s_added, __pyx_n_s_atom2, __pyx_n_s_bond12); if (unlikely(!__pyx_tuple__9)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1363; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__9);
  __Pyx_GIVEREF(__pyx_tuple__9);
  __pyx_codeobj__10 = (PyObject*)__Pyx_PyCode_New(1, 0, 7, 0, 0, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple__9, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_Users_agnes_PycharmProjects_RMG, __pyx_n_s_sortByConnectivity, 1363, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__10)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1363; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":1388
 *             return sortedAtomList
 * 
 *         def checkSet(superList, subList):             # <<<<<<<<<<<<<<
 *             """
 *             Args:
 */
  __pyx_tuple__11 = PyTuple_Pack(4, __pyx_n_s_superList, __pyx_n_s_subList, __pyx_n_s_superSet, __pyx_n_s_subSet); if (unlikely(!__pyx_tuple__11)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1388; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__11);
  __Pyx_GIVEREF(__pyx_tuple__11);
  __pyx_codeobj__12 = (PyObject*)__Pyx_PyCode_New(2, 0, 4, 0, 0, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple__11, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_Users_agnes_PycharmProjects_RMG, __pyx_n_s_checkSet, 1388, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__12)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1388; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":1401
 *             return superSet.issuperset(subSet)
 * 
 *         def mergeOverlappingBenzeneRings(ring1, ring2, od):             # <<<<<<<<<<<<<<
 *             """
 *             The input arguements of rings are always in the order that the atoms appear
 */
  __pyx_tuple__13 = PyTuple_Pack(10, __pyx_n_s_ring1, __pyx_n_s_ring2, __pyx_n_s_od, __pyx_n_s_newRing, __pyx_n_s_matchList1, __pyx_n_s_matchList2, __pyx_n_s_matchList3, __pyx_n_s_matchList4, __pyx_n_s_x1, __pyx_n_s_x2); if (unlikely(!__pyx_tuple__13)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1401; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__13);
  __Pyx_GIVEREF(__pyx_tuple__13);
  __pyx_codeobj__14 = (PyObject*)__Pyx_PyCode_New(3, 0, 10, 0, 0, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple__13, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_Users_agnes_PycharmProjects_RMG, __pyx_n_s_mergeOverlappingBenzeneRings, 1401, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__14)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1401; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":1442
 *             return newRing
 * 
 *         def addCbAtomToRing(ring, cbAtom):             # <<<<<<<<<<<<<<
 *             """
 *             Every 'Cb' atom belongs in exactly one benzene ring. This function checks
 */
  __pyx_tuple__15 = PyTuple_Pack(5, __pyx_n_s_ring, __pyx_n_s_cbAtom, __pyx_n_s_mergedRing, __pyx_n_s_atom2, __pyx_n_s_bond12); if (unlikely(!__pyx_tuple__15)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1442; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__15);
  __Pyx_GIVEREF(__pyx_tuple__15);
  __pyx_codeobj__16 = (PyObject*)__Pyx_PyCode_New(2, 0, 5, 0, 0, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple__15, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_Users_agnes_PycharmProjects_RMG, __pyx_n_s_addCbAtomToRing, 1442, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__16)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1442; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

  /* "rmgpy/molecule/group.py":1489
 *         if len(cbfAtomList) > 3:
 *             if not self.isBenzeneExplicit():
 *                 raise ImplicitBenzeneError("{0} has more than three Cbf atoms and does not have fully explicit benzene rings.")             # <<<<<<<<<<<<<<
 *             #already fully explicit if it passes isBenzeneExplict()
 *             else: return copyGroup
 */
  __pyx_tuple__17 = PyTuple_Pack(1, __pyx_kp_s_0_has_more_than_three_Cbf_atoms); if (unlikely(!__pyx_tuple__17)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1489; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__17);
  __Pyx_GIVEREF(__pyx_tuple__17);

  /* "rmgpy/molecule/group.py":1556
 * 
 *         """
 *         rings=[cycle for cycle in copyGroup.getAllCyclesOfSize(6) if Group(atoms = cycle).isAromaticRing()]             # <<<<<<<<<<<<<<
 * 
 *         """
 */
  __pyx_tuple__18 = PyTuple_Pack(1, __pyx_int_6); if (unlikely(!__pyx_tuple__18)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1556; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__18);
  __Pyx_GIVEREF(__pyx_tuple__18);

  /* "rmgpy/molecule/group.py":1714
 *                 hydrogenNeeded = atom.charge - statedCharge
 *                 for x in range(hydrogenNeeded):
 *                     newH = mol.Atom('H', radicalElectrons=0, lonePairs=0, charge=0)             # <<<<<<<<<<<<<<
 *                     newBond = mol.Bond(atom, newH, 'S')
 *                     newMolecule.addAtom(newH)
 */
  __pyx_tuple__19 = PyTuple_Pack(1, __pyx_n_s_H); if (unlikely(!__pyx_tuple__19)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1714; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__19);
  __Pyx_GIVEREF(__pyx_tuple__19);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":215
 *             if ((flags & pybuf.PyBUF_C_CONTIGUOUS == pybuf.PyBUF_C_CONTIGUOUS)
 *                 and not PyArray_CHKFLAGS(self, NPY_C_CONTIGUOUS)):
 *                 raise ValueError(u"ndarray is not C contiguous")             # <<<<<<<<<<<<<<
 * 
 *             if ((flags & pybuf.PyBUF_F_CONTIGUOUS == pybuf.PyBUF_F_CONTIGUOUS)
 */
  __pyx_tuple__20 = PyTuple_Pack(1, __pyx_kp_u_ndarray_is_not_C_contiguous); if (unlikely(!__pyx_tuple__20)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 215; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__20);
  __Pyx_GIVEREF(__pyx_tuple__20);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":219
 *             if ((flags & pybuf.PyBUF_F_CONTIGUOUS == pybuf.PyBUF_F_CONTIGUOUS)
 *                 and not PyArray_CHKFLAGS(self, NPY_F_CONTIGUOUS)):
 *                 raise ValueError(u"ndarray is not Fortran contiguous")             # <<<<<<<<<<<<<<
 * 
 *             info.buf = PyArray_DATA(self)
 */
  __pyx_tuple__21 = PyTuple_Pack(1, __pyx_kp_u_ndarray_is_not_Fortran_contiguou); if (unlikely(!__pyx_tuple__21)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 219; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__21);
  __Pyx_GIVEREF(__pyx_tuple__21);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":257
 *                 if ((descr.byteorder == c'>' and little_endian) or
 *                     (descr.byteorder == c'<' and not little_endian)):
 *                     raise ValueError(u"Non-native byte order not supported")             # <<<<<<<<<<<<<<
 *                 if   t == NPY_BYTE:        f = "b"
 *                 elif t == NPY_UBYTE:       f = "B"
 */
  __pyx_tuple__22 = PyTuple_Pack(1, __pyx_kp_u_Non_native_byte_order_not_suppor); if (unlikely(!__pyx_tuple__22)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 257; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__22);
  __Pyx_GIVEREF(__pyx_tuple__22);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":799
 * 
 *         if (end - f) - <int>(new_offset - offset[0]) < 15:
 *             raise RuntimeError(u"Format string allocated too short, see comment in numpy.pxd")             # <<<<<<<<<<<<<<
 * 
 *         if ((child.byteorder == c'>' and little_endian) or
 */
  __pyx_tuple__23 = PyTuple_Pack(1, __pyx_kp_u_Format_string_allocated_too_shor); if (unlikely(!__pyx_tuple__23)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 799; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__23);
  __Pyx_GIVEREF(__pyx_tuple__23);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":803
 *         if ((child.byteorder == c'>' and little_endian) or
 *             (child.byteorder == c'<' and not little_endian)):
 *             raise ValueError(u"Non-native byte order not supported")             # <<<<<<<<<<<<<<
 *             # One could encode it in the format string and have Cython
 *             # complain instead, BUT: < and > in format strings also imply
 */
  __pyx_tuple__24 = PyTuple_Pack(1, __pyx_kp_u_Non_native_byte_order_not_suppor); if (unlikely(!__pyx_tuple__24)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 803; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__24);
  __Pyx_GIVEREF(__pyx_tuple__24);

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":823
 *             t = child.type_num
 *             if end - f < 5:
 *                 raise RuntimeError(u"Format string allocated too short.")             # <<<<<<<<<<<<<<
 * 
 *             # Until ticket #99 is fixed, use integers to avoid warnings
 */
  __pyx_tuple__25 = PyTuple_Pack(1, __pyx_kp_u_Format_string_allocated_too_shor_2); if (unlikely(!__pyx_tuple__25)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 823; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_tuple__25);
  __Pyx_GIVEREF(__pyx_tuple__25);
  __Pyx_RefNannyFinishContext();
  return 0;
  __pyx_L1_error:;
  __Pyx_RefNannyFinishContext();
  return -1;
}

static int __Pyx_InitGlobals(void) {
  if (__Pyx_InitStrings(__pyx_string_tab) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  __pyx_float_2_0 = PyFloat_FromDouble(2.0); if (unlikely(!__pyx_float_2_0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_float_3_0 = PyFloat_FromDouble(3.0); if (unlikely(!__pyx_float_3_0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_0 = PyInt_FromLong(0); if (unlikely(!__pyx_int_0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_1 = PyInt_FromLong(1); if (unlikely(!__pyx_int_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_2 = PyInt_FromLong(2); if (unlikely(!__pyx_int_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_3 = PyInt_FromLong(3); if (unlikely(!__pyx_int_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_4 = PyInt_FromLong(4); if (unlikely(!__pyx_int_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_6 = PyInt_FromLong(6); if (unlikely(!__pyx_int_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_neg_1 = PyInt_FromLong(-1); if (unlikely(!__pyx_int_neg_1)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_int_neg_100 = PyInt_FromLong(-100); if (unlikely(!__pyx_int_neg_100)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  return 0;
  __pyx_L1_error:;
  return -1;
}

#if PY_MAJOR_VERSION < 3
PyMODINIT_FUNC initgroup(void); /*proto*/
PyMODINIT_FUNC initgroup(void)
#else
PyMODINIT_FUNC PyInit_group(void); /*proto*/
PyMODINIT_FUNC PyInit_group(void)
#endif
{
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;
  PyObject *__pyx_t_4 = NULL;
  PyObject *__pyx_t_5 = NULL;
  PyObject *__pyx_t_6 = NULL;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannyDeclarations
  #if CYTHON_REFNANNY
  __Pyx_RefNanny = __Pyx_RefNannyImportAPI("refnanny");
  if (!__Pyx_RefNanny) {
      PyErr_Clear();
      __Pyx_RefNanny = __Pyx_RefNannyImportAPI("Cython.Runtime.refnanny");
      if (!__Pyx_RefNanny)
          Py_FatalError("failed to import 'refnanny' module");
  }
  #endif
  __Pyx_RefNannySetupContext("PyMODINIT_FUNC PyInit_group(void)", 0);
  if ( __Pyx_check_binary_version() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_empty_tuple = PyTuple_New(0); if (unlikely(!__pyx_empty_tuple)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_empty_bytes = PyBytes_FromStringAndSize("", 0); if (unlikely(!__pyx_empty_bytes)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #ifdef __Pyx_CyFunction_USED
  if (__Pyx_CyFunction_init() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #endif
  #ifdef __Pyx_FusedFunction_USED
  if (__pyx_FusedFunction_init() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #endif
  #ifdef __Pyx_Generator_USED
  if (__pyx_Generator_init() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #endif
  /*--- Library function declarations ---*/
  /*--- Threads initialization code ---*/
  #if defined(__PYX_FORCE_INIT_THREADS) && __PYX_FORCE_INIT_THREADS
  #ifdef WITH_THREAD /* Python build with threading support? */
  PyEval_InitThreads();
  #endif
  #endif
  /*--- Module creation code ---*/
  #if PY_MAJOR_VERSION < 3
  __pyx_m = Py_InitModule4("group", __pyx_methods, __pyx_k_This_module_provides_classes_an, 0, PYTHON_API_VERSION); Py_XINCREF(__pyx_m);
  #else
  __pyx_m = PyModule_Create(&__pyx_moduledef);
  #endif
  if (unlikely(!__pyx_m)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_d = PyModule_GetDict(__pyx_m); if (unlikely(!__pyx_d)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  Py_INCREF(__pyx_d);
  __pyx_b = PyImport_AddModule(__Pyx_BUILTIN_MODULE_NAME); if (unlikely(!__pyx_b)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #if CYTHON_COMPILING_IN_PYPY
  Py_INCREF(__pyx_b);
  #endif
  if (PyObject_SetAttrString(__pyx_m, "__builtins__", __pyx_b) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  /*--- Initialize various global constants etc. ---*/
  if (unlikely(__Pyx_InitGlobals() < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #if PY_MAJOR_VERSION < 3 && (__PYX_DEFAULT_STRING_ENCODING_IS_ASCII || __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT)
  if (__Pyx_init_sys_getdefaultencoding_params() < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  #endif
  if (__pyx_module_is_main_rmgpy__molecule__group) {
    if (PyObject_SetAttrString(__pyx_m, "__name__", __pyx_n_s_main) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;};
  }
  #if PY_MAJOR_VERSION >= 3
  {
    PyObject *modules = PyImport_GetModuleDict(); if (unlikely(!modules)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (!PyDict_GetItemString(modules, "rmgpy.molecule.group")) {
      if (unlikely(PyDict_SetItemString(modules, "rmgpy.molecule.group", __pyx_m) < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    }
  }
  #endif
  /*--- Builtin init code ---*/
  if (unlikely(__Pyx_InitCachedBuiltins() < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  /*--- Constants init code ---*/
  if (unlikely(__Pyx_InitCachedConstants() < 0)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  /*--- Global init code ---*/
  /*--- Variable export code ---*/
  /*--- Function export code ---*/
  /*--- Type init code ---*/
  __pyx_ptype_5rmgpy_8molecule_5graph_Vertex = __Pyx_ImportType("rmgpy.molecule.graph", "Vertex", sizeof(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_5graph_Vertex)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5graph_Vertex = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Vertex*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_5graph_Vertex->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_5graph_Vertex)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5group_GroupAtom = &__pyx_vtable_5rmgpy_8molecule_5group_GroupAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx_base = *__pyx_vtabptr_5rmgpy_8molecule_5graph_Vertex;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx_base.copy = (struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_copy;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx_base.equivalent = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_equivalent;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx_base.isSpecificCaseOf = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, struct __pyx_obj_5rmgpy_8molecule_5graph_Vertex *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSpecificCaseOf;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___changeBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___changeBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___formBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___formBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___breakBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___breakBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___gainRadical = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainRadical;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___loseRadical = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___loseRadical;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___gainPair = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___gainPair;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.__pyx___losePair = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom___losePair;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.applyAction = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_applyAction;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.isOxygen = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isOxygen;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.isSulfur = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_isSulfur;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.hasWildcards = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_hasWildcards;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.makeSampleAtom = (struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_makeSampleAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupAtom.getBondOrdersForAtom = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupAtom_getBondOrdersForAtom;
  __pyx_type_5rmgpy_8molecule_5group_GroupAtom.tp_base = __pyx_ptype_5rmgpy_8molecule_5graph_Vertex;
  if (PyType_Ready(&__pyx_type_5rmgpy_8molecule_5group_GroupAtom) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 64; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_type_5rmgpy_8molecule_5group_GroupAtom.tp_print = 0;
  #if CYTHON_COMPILING_IN_CPYTHON
  {
    PyObject *wrapper = PyObject_GetAttrString((PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupAtom, "__str__"); if (unlikely(!wrapper)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 64; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (Py_TYPE(wrapper) == &PyWrapperDescr_Type) {
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_6__str__ = *((PyWrapperDescrObject *)wrapper)->d_base;
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_6__str__.doc = __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_6__str__;
      ((PyWrapperDescrObject *)wrapper)->d_base = &__pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_6__str__;
    }
  }
  #endif
  #if CYTHON_COMPILING_IN_CPYTHON
  {
    PyObject *wrapper = PyObject_GetAttrString((PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupAtom, "__repr__"); if (unlikely(!wrapper)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 64; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (Py_TYPE(wrapper) == &PyWrapperDescr_Type) {
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_8__repr__ = *((PyWrapperDescrObject *)wrapper)->d_base;
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_8__repr__.doc = __pyx_doc_5rmgpy_8molecule_5group_9GroupAtom_8__repr__;
      ((PyWrapperDescrObject *)wrapper)->d_base = &__pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupAtom_8__repr__;
    }
  }
  #endif
  if (__Pyx_SetVtable(__pyx_type_5rmgpy_8molecule_5group_GroupAtom.tp_dict, __pyx_vtabptr_5rmgpy_8molecule_5group_GroupAtom) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 64; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyObject_SetAttrString(__pyx_m, "GroupAtom", (PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupAtom) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 64; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_5group_GroupAtom = &__pyx_type_5rmgpy_8molecule_5group_GroupAtom;
  __pyx_ptype_5rmgpy_8molecule_5graph_Edge = __Pyx_ImportType("rmgpy.molecule.graph", "Edge", sizeof(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_5graph_Edge)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5graph_Edge = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Edge*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_5graph_Edge->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_5graph_Edge)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5group_GroupBond = &__pyx_vtable_5rmgpy_8molecule_5group_GroupBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx_base = *__pyx_vtabptr_5rmgpy_8molecule_5graph_Edge;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx_base.copy = (struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_copy;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx_base.equivalent = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_equivalent;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx_base.isSpecificCaseOf = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, struct __pyx_obj_5rmgpy_8molecule_5graph_Edge *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSpecificCaseOf;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.__pyx___changeBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, short, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond___changeBond;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.isSingle = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isSingle;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.isDouble = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isDouble;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.isTriple = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isTriple;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.isBenzene = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_isBenzene;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.applyAction = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_applyAction;
  __pyx_vtable_5rmgpy_8molecule_5group_GroupBond.makeBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_9GroupBond_makeBond;
  __pyx_type_5rmgpy_8molecule_5group_GroupBond.tp_base = __pyx_ptype_5rmgpy_8molecule_5graph_Edge;
  if (PyType_Ready(&__pyx_type_5rmgpy_8molecule_5group_GroupBond) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 529; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_type_5rmgpy_8molecule_5group_GroupBond.tp_print = 0;
  #if CYTHON_COMPILING_IN_CPYTHON
  {
    PyObject *wrapper = PyObject_GetAttrString((PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupBond, "__str__"); if (unlikely(!wrapper)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 529; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (Py_TYPE(wrapper) == &PyWrapperDescr_Type) {
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_2__str__ = *((PyWrapperDescrObject *)wrapper)->d_base;
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_2__str__.doc = __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_2__str__;
      ((PyWrapperDescrObject *)wrapper)->d_base = &__pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_2__str__;
    }
  }
  #endif
  #if CYTHON_COMPILING_IN_CPYTHON
  {
    PyObject *wrapper = PyObject_GetAttrString((PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupBond, "__repr__"); if (unlikely(!wrapper)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 529; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
    if (Py_TYPE(wrapper) == &PyWrapperDescr_Type) {
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_4__repr__ = *((PyWrapperDescrObject *)wrapper)->d_base;
      __pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_4__repr__.doc = __pyx_doc_5rmgpy_8molecule_5group_9GroupBond_4__repr__;
      ((PyWrapperDescrObject *)wrapper)->d_base = &__pyx_wrapperbase_5rmgpy_8molecule_5group_9GroupBond_4__repr__;
    }
  }
  #endif
  if (__Pyx_SetVtable(__pyx_type_5rmgpy_8molecule_5group_GroupBond.tp_dict, __pyx_vtabptr_5rmgpy_8molecule_5group_GroupBond) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 529; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyObject_SetAttrString(__pyx_m, "GroupBond", (PyObject *)&__pyx_type_5rmgpy_8molecule_5group_GroupBond) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 529; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_5group_GroupBond = &__pyx_type_5rmgpy_8molecule_5group_GroupBond;
  __pyx_ptype_5rmgpy_8molecule_5graph_Graph = __Pyx_ImportType("rmgpy.molecule.graph", "Graph", sizeof(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_5graph_Graph)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5graph_Graph = (struct __pyx_vtabstruct_5rmgpy_8molecule_5graph_Graph*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_5graph_Graph->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_5group_Group = &__pyx_vtable_5rmgpy_8molecule_5group_Group;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base = *__pyx_vtabptr_5rmgpy_8molecule_5graph_Graph;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.copy = (struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_copy *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_copy;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.merge = (struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_merge;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.split = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_split;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.isIsomorphic = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isIsomorphic *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_isIsomorphic;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.findIsomorphism = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findIsomorphism *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_findIsomorphism;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.isSubgraphIsomorphic = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_isSubgraphIsomorphic *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_isSubgraphIsomorphic;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.__pyx_base.findSubgraphIsomorphisms = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5graph_5Graph_findSubgraphIsomorphisms *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_findSubgraphIsomorphisms;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.addAtom = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_addAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.addBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_addBond;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.getBonds = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_getBonds;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.getBond = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_getBond;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.hasAtom = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_hasAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.hasBond = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_hasBond;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.removeAtom = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_removeAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.removeBond = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5group_GroupBond *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_removeBond;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.sortAtoms = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_sortAtoms;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.clearLabeledAtoms = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_clearLabeledAtoms;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.containsLabeledAtom = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_containsLabeledAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.getLabeledAtom = (struct __pyx_obj_5rmgpy_8molecule_5group_GroupAtom *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtom;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.getLabeledAtoms = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_getLabeledAtoms;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.fromAdjacencyList = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, PyObject *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_fromAdjacencyList;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.toAdjacencyList = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch, struct __pyx_opt_args_5rmgpy_8molecule_5group_5Group_toAdjacencyList *__pyx_optional_args))__pyx_f_5rmgpy_8molecule_5group_5Group_toAdjacencyList;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.updateFingerprint = (PyObject *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_updateFingerprint;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.isIdentical = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, struct __pyx_obj_5rmgpy_8molecule_5graph_Graph *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_isIdentical;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.isAromaticRing = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_isAromaticRing;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.standardizeAtomType = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_standardizeAtomType;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.addExplicitLigands = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_addExplicitLigands;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.standardizeGroup = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_standardizeGroup;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.addImplicitAtomsFromAtomType = (struct __pyx_obj_5rmgpy_8molecule_5group_Group *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_addImplicitAtomsFromAtomType;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.makeSampleMolecule = (struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_makeSampleMolecule;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.addImplicitBenzene = (struct __pyx_obj_5rmgpy_8molecule_5group_Group *(*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_addImplicitBenzene;
  __pyx_vtable_5rmgpy_8molecule_5group_Group.isBenzeneExplicit = (int (*)(struct __pyx_obj_5rmgpy_8molecule_5group_Group *, int __pyx_skip_dispatch))__pyx_f_5rmgpy_8molecule_5group_5Group_isBenzeneExplicit;
  __pyx_type_5rmgpy_8molecule_5group_Group.tp_base = __pyx_ptype_5rmgpy_8molecule_5graph_Graph;
  if (PyType_Ready(&__pyx_type_5rmgpy_8molecule_5group_Group) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 706; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_type_5rmgpy_8molecule_5group_Group.tp_print = 0;
  if (__Pyx_SetVtable(__pyx_type_5rmgpy_8molecule_5group_Group.tp_dict, __pyx_vtabptr_5rmgpy_8molecule_5group_Group) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 706; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (PyObject_SetAttrString(__pyx_m, "Group", (PyObject *)&__pyx_type_5rmgpy_8molecule_5group_Group) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 706; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_5group_Group = &__pyx_type_5rmgpy_8molecule_5group_Group;
  /*--- Type import code ---*/
  __pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType = __Pyx_ImportType("rmgpy.molecule.atomtype", "AtomType", sizeof(struct __pyx_obj_5rmgpy_8molecule_8atomtype_AtomType), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType)) {__pyx_filename = __pyx_f[3]; __pyx_lineno = 27; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_8atomtype_AtomType = (struct __pyx_vtabstruct_5rmgpy_8molecule_8atomtype_AtomType*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_8atomtype_AtomType->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_8atomtype_AtomType)) {__pyx_filename = __pyx_f[3]; __pyx_lineno = 27; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_7element_Element = __Pyx_ImportType("rmgpy.molecule.element", "Element", sizeof(struct __pyx_obj_5rmgpy_8molecule_7element_Element), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_7element_Element)) {__pyx_filename = __pyx_f[4]; __pyx_lineno = 27; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_7cpython_4type_type = __Pyx_ImportType(__Pyx_BUILTIN_MODULE_NAME, "type", 
  #if CYTHON_COMPILING_IN_PYPY
  sizeof(PyTypeObject),
  #else
  sizeof(PyHeapTypeObject),
  #endif
  0); if (unlikely(!__pyx_ptype_7cpython_4type_type)) {__pyx_filename = __pyx_f[5]; __pyx_lineno = 9; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5numpy_dtype = __Pyx_ImportType("numpy", "dtype", sizeof(PyArray_Descr), 0); if (unlikely(!__pyx_ptype_5numpy_dtype)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 155; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5numpy_flatiter = __Pyx_ImportType("numpy", "flatiter", sizeof(PyArrayIterObject), 0); if (unlikely(!__pyx_ptype_5numpy_flatiter)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 165; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5numpy_broadcast = __Pyx_ImportType("numpy", "broadcast", sizeof(PyArrayMultiIterObject), 0); if (unlikely(!__pyx_ptype_5numpy_broadcast)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 169; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5numpy_ndarray = __Pyx_ImportType("numpy", "ndarray", sizeof(PyArrayObject), 0); if (unlikely(!__pyx_ptype_5numpy_ndarray)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 178; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5numpy_ufunc = __Pyx_ImportType("numpy", "ufunc", sizeof(PyUFuncObject), 0); if (unlikely(!__pyx_ptype_5numpy_ufunc)) {__pyx_filename = __pyx_f[2]; __pyx_lineno = 861; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_8molecule_Atom = __Pyx_ImportType("rmgpy.molecule.molecule", "Atom", sizeof(struct __pyx_obj_5rmgpy_8molecule_8molecule_Atom), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_8molecule_Atom)) {__pyx_filename = __pyx_f[6]; __pyx_lineno = 37; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_8molecule_Atom = (struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Atom*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_8molecule_Atom->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_8molecule_Atom)) {__pyx_filename = __pyx_f[6]; __pyx_lineno = 37; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_8molecule_Bond = __Pyx_ImportType("rmgpy.molecule.molecule", "Bond", sizeof(struct __pyx_obj_5rmgpy_8molecule_8molecule_Bond), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_8molecule_Bond)) {__pyx_filename = __pyx_f[6]; __pyx_lineno = 82; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_8molecule_Bond = (struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Bond*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_8molecule_Bond->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_8molecule_Bond)) {__pyx_filename = __pyx_f[6]; __pyx_lineno = 82; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_ptype_5rmgpy_8molecule_8molecule_Molecule = __Pyx_ImportType("rmgpy.molecule.molecule", "Molecule", sizeof(struct __pyx_obj_5rmgpy_8molecule_8molecule_Molecule), 1); if (unlikely(!__pyx_ptype_5rmgpy_8molecule_8molecule_Molecule)) {__pyx_filename = __pyx_f[6]; __pyx_lineno = 106; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __pyx_vtabptr_5rmgpy_8molecule_8molecule_Molecule = (struct __pyx_vtabstruct_5rmgpy_8molecule_8molecule_Molecule*)__Pyx_GetVtable(__pyx_ptype_5rmgpy_8molecule_8molecule_Molecule->tp_dict); if (unlikely(!__pyx_vtabptr_5rmgpy_8molecule_8molecule_Molecule)) {__pyx_filename = __pyx_f[6]; __pyx_lineno = 106; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  /*--- Variable import code ---*/
  __pyx_t_1 = __Pyx_ImportModule("rmgpy.constants"); if (!__pyx_t_1) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "pi", (void **)&__pyx_vp_5rmgpy_9constants_pi, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "Na", (void **)&__pyx_vp_5rmgpy_9constants_Na, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "kB", (void **)&__pyx_vp_5rmgpy_9constants_kB, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "R", (void **)&__pyx_vp_5rmgpy_9constants_R, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "h", (void **)&__pyx_vp_5rmgpy_9constants_h, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "hbar", (void **)&__pyx_vp_5rmgpy_9constants_hbar, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "c", (void **)&__pyx_vp_5rmgpy_9constants_c, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "e", (void **)&__pyx_vp_5rmgpy_9constants_e, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "m_e", (void **)&__pyx_vp_5rmgpy_9constants_m_e, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "m_p", (void **)&__pyx_vp_5rmgpy_9constants_m_p, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "m_n", (void **)&__pyx_vp_5rmgpy_9constants_m_n, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "amu", (void **)&__pyx_vp_5rmgpy_9constants_amu, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "a0", (void **)&__pyx_vp_5rmgpy_9constants_a0, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_1, "E_h", (void **)&__pyx_vp_5rmgpy_9constants_E_h, "double") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  Py_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  __pyx_t_2 = __Pyx_ImportModule("rmgpy.molecule.molecule"); if (!__pyx_t_2) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  if (__Pyx_ImportVoidPtr(__pyx_t_2, "bond_orders", (void **)&__pyx_vp_5rmgpy_8molecule_8molecule_bond_orders, "PyObject *") < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  Py_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  /*--- Function import code ---*/
  /*--- Execution code ---*/

  /* "rmgpy/molecule/group.py":38
 * import cython
 * 
 * from .graph import Vertex, Edge, Graph             # <<<<<<<<<<<<<<
 * from .atomtype import atomTypes, allElements, nonSpecifics, getFeatures
 * from .element import PeriodicSystem
 */
  __pyx_t_3 = PyList_New(3); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 38; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_n_s_Vertex);
  PyList_SET_ITEM(__pyx_t_3, 0, __pyx_n_s_Vertex);
  __Pyx_GIVEREF(__pyx_n_s_Vertex);
  __Pyx_INCREF(__pyx_n_s_Edge);
  PyList_SET_ITEM(__pyx_t_3, 1, __pyx_n_s_Edge);
  __Pyx_GIVEREF(__pyx_n_s_Edge);
  __Pyx_INCREF(__pyx_n_s_Graph);
  PyList_SET_ITEM(__pyx_t_3, 2, __pyx_n_s_Graph);
  __Pyx_GIVEREF(__pyx_n_s_Graph);
  __pyx_t_4 = __Pyx_Import(__pyx_n_s_graph, __pyx_t_3, 1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 38; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":39
 * 
 * from .graph import Vertex, Edge, Graph
 * from .atomtype import atomTypes, allElements, nonSpecifics, getFeatures             # <<<<<<<<<<<<<<
 * from .element import PeriodicSystem
 * import rmgpy.molecule.molecule as mol
 */
  __pyx_t_4 = PyList_New(4); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_INCREF(__pyx_n_s_atomTypes);
  PyList_SET_ITEM(__pyx_t_4, 0, __pyx_n_s_atomTypes);
  __Pyx_GIVEREF(__pyx_n_s_atomTypes);
  __Pyx_INCREF(__pyx_n_s_allElements);
  PyList_SET_ITEM(__pyx_t_4, 1, __pyx_n_s_allElements);
  __Pyx_GIVEREF(__pyx_n_s_allElements);
  __Pyx_INCREF(__pyx_n_s_nonSpecifics);
  PyList_SET_ITEM(__pyx_t_4, 2, __pyx_n_s_nonSpecifics);
  __Pyx_GIVEREF(__pyx_n_s_nonSpecifics);
  __Pyx_INCREF(__pyx_n_s_getFeatures);
  PyList_SET_ITEM(__pyx_t_4, 3, __pyx_n_s_getFeatures);
  __Pyx_GIVEREF(__pyx_n_s_getFeatures);
  __pyx_t_3 = __Pyx_Import(__pyx_n_s_atomtype, __pyx_t_4, 1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_t_4 = __Pyx_ImportFrom(__pyx_t_3, __pyx_n_s_atomTypes); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_atomTypes, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_t_4 = __Pyx_ImportFrom(__pyx_t_3, __pyx_n_s_allElements); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_allElements, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_t_4 = __Pyx_ImportFrom(__pyx_t_3, __pyx_n_s_nonSpecifics); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_nonSpecifics, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __pyx_t_4 = __Pyx_ImportFrom(__pyx_t_3, __pyx_n_s_getFeatures); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_getFeatures, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 39; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":40
 * from .graph import Vertex, Edge, Graph
 * from .atomtype import atomTypes, allElements, nonSpecifics, getFeatures
 * from .element import PeriodicSystem             # <<<<<<<<<<<<<<
 * import rmgpy.molecule.molecule as mol
 * from copy import deepcopy, copy
 */
  __pyx_t_3 = PyList_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 40; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_n_s_PeriodicSystem);
  PyList_SET_ITEM(__pyx_t_3, 0, __pyx_n_s_PeriodicSystem);
  __Pyx_GIVEREF(__pyx_n_s_PeriodicSystem);
  __pyx_t_4 = __Pyx_Import(__pyx_n_s_element, __pyx_t_3, 1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 40; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = __Pyx_ImportFrom(__pyx_t_4, __pyx_n_s_PeriodicSystem); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 40; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_PeriodicSystem, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 40; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":41
 * from .atomtype import atomTypes, allElements, nonSpecifics, getFeatures
 * from .element import PeriodicSystem
 * import rmgpy.molecule.molecule as mol             # <<<<<<<<<<<<<<
 * from copy import deepcopy, copy
 * 
 */
  __pyx_t_4 = PyList_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 41; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_INCREF(__pyx_n_s__26);
  PyList_SET_ITEM(__pyx_t_4, 0, __pyx_n_s__26);
  __Pyx_GIVEREF(__pyx_n_s__26);
  __pyx_t_3 = __Pyx_Import(__pyx_n_s_rmgpy_molecule_molecule, __pyx_t_4, -1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 41; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_mol, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 41; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "rmgpy/molecule/group.py":42
 * from .element import PeriodicSystem
 * import rmgpy.molecule.molecule as mol
 * from copy import deepcopy, copy             # <<<<<<<<<<<<<<
 * 
 * ################################################################################
 */
  __pyx_t_3 = PyList_New(2); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 42; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_INCREF(__pyx_n_s_deepcopy);
  PyList_SET_ITEM(__pyx_t_3, 0, __pyx_n_s_deepcopy);
  __Pyx_GIVEREF(__pyx_n_s_deepcopy);
  __Pyx_INCREF(__pyx_n_s_copy);
  PyList_SET_ITEM(__pyx_t_3, 1, __pyx_n_s_copy);
  __Pyx_GIVEREF(__pyx_n_s_copy);
  __pyx_t_4 = __Pyx_Import(__pyx_n_s_copy, __pyx_t_3, -1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 42; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = __Pyx_ImportFrom(__pyx_t_4, __pyx_n_s_deepcopy); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 42; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_deepcopy, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 42; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __pyx_t_3 = __Pyx_ImportFrom(__pyx_t_4, __pyx_n_s_copy); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 42; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_copy, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 42; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":46
 * ################################################################################
 * 
 * class ActionError(Exception):             # <<<<<<<<<<<<<<
 *     """
 *     An exception class for errors that occur while applying reaction recipe
 */
  __pyx_t_4 = PyTuple_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 46; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_INCREF(__pyx_builtin_Exception);
  PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_builtin_Exception);
  __Pyx_GIVEREF(__pyx_builtin_Exception);
  __pyx_t_3 = __Pyx_CalculateMetaclass(NULL, __pyx_t_4); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 46; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_5 = __Pyx_Py3MetaclassPrepare(__pyx_t_3, __pyx_t_4, __pyx_n_s_ActionError, __pyx_n_s_ActionError, (PyObject *) NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_kp_s_An_exception_class_for_errors_t); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 46; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  __pyx_t_6 = __Pyx_Py3ClassCreate(__pyx_t_3, __pyx_n_s_ActionError, __pyx_t_4, __pyx_t_5, NULL, 0, 1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 46; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_ActionError, __pyx_t_6) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 46; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":54
 *     pass
 * 
 * class ImplicitBenzeneError(Exception):             # <<<<<<<<<<<<<<
 *     """
 *     An exception class when encountering a group with too many implicit benzene
 */
  __pyx_t_4 = PyTuple_New(1); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 54; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_INCREF(__pyx_builtin_Exception);
  PyTuple_SET_ITEM(__pyx_t_4, 0, __pyx_builtin_Exception);
  __Pyx_GIVEREF(__pyx_builtin_Exception);
  __pyx_t_3 = __Pyx_CalculateMetaclass(NULL, __pyx_t_4); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 54; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_5 = __Pyx_Py3MetaclassPrepare(__pyx_t_3, __pyx_t_4, __pyx_n_s_ImplicitBenzeneError, __pyx_n_s_ImplicitBenzeneError, (PyObject *) NULL, __pyx_n_s_rmgpy_molecule_group, __pyx_kp_s_An_exception_class_when_encount); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 54; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  __pyx_t_6 = __Pyx_Py3ClassCreate(__pyx_t_3, __pyx_n_s_ImplicitBenzeneError, __pyx_t_4, __pyx_t_5, NULL, 0, 1); if (unlikely(!__pyx_t_6)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 54; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_6);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_ImplicitBenzeneError, __pyx_t_6) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 54; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;

  /* "rmgpy/molecule/group.py":137
 * 
 *     @property
 *     def bonds(self): return self.edges             # <<<<<<<<<<<<<<
 * 
 *     def copy(self):
 */
  __pyx_t_4 = __Pyx_GetNameInClass((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom, __pyx_n_s_bonds); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 137; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);

  /* "rmgpy/molecule/group.py":136
 *         return "<GroupAtom {0!s}>".format(self)
 * 
 *     @property             # <<<<<<<<<<<<<<
 *     def bonds(self): return self.edges
 * 
 */
  __pyx_t_3 = PyTuple_New(1); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 136; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  PyTuple_SET_ITEM(__pyx_t_3, 0, __pyx_t_4);
  __Pyx_GIVEREF(__pyx_t_4);
  __pyx_t_4 = 0;
  __pyx_t_4 = __Pyx_PyObject_Call(__pyx_builtin_property, __pyx_t_3, NULL); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 136; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  if (PyDict_SetItem((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom->tp_dict, __pyx_n_s_bonds, __pyx_t_4) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 137; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  PyType_Modified(__pyx_ptype_5rmgpy_8molecule_5group_GroupAtom);

  /* "rmgpy/molecule/group.py":728
 *     def __getAtoms(self): return self.vertices
 *     def __setAtoms(self, atoms): self.vertices = atoms
 *     atoms = property(__getAtoms, __setAtoms)             # <<<<<<<<<<<<<<
 * 
 *     def addAtom(self, atom):
 */
  __pyx_t_4 = __Pyx_GetNameInClass((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_Group, __pyx_n_s_getAtoms); if (unlikely(!__pyx_t_4)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 728; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_4);
  __pyx_t_3 = __Pyx_GetNameInClass((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_Group, __pyx_n_s_setAtoms); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 728; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __pyx_t_5 = PyTuple_New(2); if (unlikely(!__pyx_t_5)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 728; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_5);
  PyTuple_SET_ITEM(__pyx_t_5, 0, __pyx_t_4);
  __Pyx_GIVEREF(__pyx_t_4);
  PyTuple_SET_ITEM(__pyx_t_5, 1, __pyx_t_3);
  __Pyx_GIVEREF(__pyx_t_3);
  __pyx_t_4 = 0;
  __pyx_t_3 = 0;
  __pyx_t_3 = __Pyx_PyObject_Call(__pyx_builtin_property, __pyx_t_5, NULL); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 728; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_5); __pyx_t_5 = 0;
  if (PyDict_SetItem((PyObject *)__pyx_ptype_5rmgpy_8molecule_5group_Group->tp_dict, __pyx_n_s_atoms, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 728; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  PyType_Modified(__pyx_ptype_5rmgpy_8molecule_5group_Group);

  /* "rmgpy/molecule/group.py":1
 * #!/usr/bin/env python             # <<<<<<<<<<<<<<
 * # encoding: utf-8
 * 
 */
  __pyx_t_3 = PyDict_New(); if (unlikely(!__pyx_t_3)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_GOTREF(__pyx_t_3);
  if (PyDict_SetItem(__pyx_d, __pyx_n_s_test, __pyx_t_3) < 0) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 1; __pyx_clineno = __LINE__; goto __pyx_L1_error;}
  __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;

  /* "../../../anaconda/envs/rmg_parallel_env/lib/python2.7/site-packages/Cython/Includes/numpy/__init__.pxd":976
 *      arr.base = baseptr
 * 
 * cdef inline object get_array_base(ndarray arr):             # <<<<<<<<<<<<<<
 *     if arr.base is NULL:
 *         return None
 */

  /*--- Wrapped vars code ---*/

  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_XDECREF(__pyx_t_4);
  __Pyx_XDECREF(__pyx_t_5);
  __Pyx_XDECREF(__pyx_t_6);
  if (__pyx_m) {
    if (__pyx_d) {
      __Pyx_AddTraceback("init rmgpy.molecule.group", __pyx_clineno, __pyx_lineno, __pyx_filename);
      Py_DECREF(__pyx_d); __pyx_d = 0;
    }
    Py_DECREF(__pyx_m); __pyx_m = 0;
  } else if (!PyErr_Occurred()) {
    PyErr_SetString(PyExc_ImportError, "init rmgpy.molecule.group");
  }
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  #if PY_MAJOR_VERSION < 3
  return;
  #else
  return __pyx_m;
  #endif
}

/* Runtime support code */
#if CYTHON_REFNANNY
static __Pyx_RefNannyAPIStruct *__Pyx_RefNannyImportAPI(const char *modname) {
    PyObject *m = NULL, *p = NULL;
    void *r = NULL;
    m = PyImport_ImportModule((char *)modname);
    if (!m) goto end;
    p = PyObject_GetAttrString(m, (char *)"RefNannyAPI");
    if (!p) goto end;
    r = PyLong_AsVoidPtr(p);
end:
    Py_XDECREF(p);
    Py_XDECREF(m);
    return (__Pyx_RefNannyAPIStruct *)r;
}
#endif

static PyObject *__Pyx_GetBuiltinName(PyObject *name) {
    PyObject* result = __Pyx_PyObject_GetAttrStr(__pyx_b, name);
    if (unlikely(!result)) {
        PyErr_Format(PyExc_NameError,
#if PY_MAJOR_VERSION >= 3
            "name '%U' is not defined", name);
#else
            "name '%.200s' is not defined", PyString_AS_STRING(name));
#endif
    }
    return result;
}

static void __Pyx_RaiseDoubleKeywordsError(
    const char* func_name,
    PyObject* kw_name)
{
    PyErr_Format(PyExc_TypeError,
        #if PY_MAJOR_VERSION >= 3
        "%s() got multiple values for keyword argument '%U'", func_name, kw_name);
        #else
        "%s() got multiple values for keyword argument '%s'", func_name,
        PyString_AsString(kw_name));
        #endif
}

static int __Pyx_ParseOptionalKeywords(
    PyObject *kwds,
    PyObject **argnames[],
    PyObject *kwds2,
    PyObject *values[],
    Py_ssize_t num_pos_args,
    const char* function_name)
{
    PyObject *key = 0, *value = 0;
    Py_ssize_t pos = 0;
    PyObject*** name;
    PyObject*** first_kw_arg = argnames + num_pos_args;
    while (PyDict_Next(kwds, &pos, &key, &value)) {
        name = first_kw_arg;
        while (*name && (**name != key)) name++;
        if (*name) {
            values[name-argnames] = value;
            continue;
        }
        name = first_kw_arg;
        #if PY_MAJOR_VERSION < 3
        if (likely(PyString_CheckExact(key)) || likely(PyString_Check(key))) {
            while (*name) {
                if ((CYTHON_COMPILING_IN_PYPY || PyString_GET_SIZE(**name) == PyString_GET_SIZE(key))
                        && _PyString_Eq(**name, key)) {
                    values[name-argnames] = value;
                    break;
                }
                name++;
            }
            if (*name) continue;
            else {
                PyObject*** argname = argnames;
                while (argname != first_kw_arg) {
                    if ((**argname == key) || (
                            (CYTHON_COMPILING_IN_PYPY || PyString_GET_SIZE(**argname) == PyString_GET_SIZE(key))
                             && _PyString_Eq(**argname, key))) {
                        goto arg_passed_twice;
                    }
                    argname++;
                }
            }
        } else
        #endif
        if (likely(PyUnicode_Check(key))) {
            while (*name) {
                int cmp = (**name == key) ? 0 :
                #if !CYTHON_COMPILING_IN_PYPY && PY_MAJOR_VERSION >= 3
                    (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :
                #endif
                    PyUnicode_Compare(**name, key);
                if (cmp < 0 && unlikely(PyErr_Occurred())) goto bad;
                if (cmp == 0) {
                    values[name-argnames] = value;
                    break;
                }
                name++;
            }
            if (*name) continue;
            else {
                PyObject*** argname = argnames;
                while (argname != first_kw_arg) {
                    int cmp = (**argname == key) ? 0 :
                    #if !CYTHON_COMPILING_IN_PYPY && PY_MAJOR_VERSION >= 3
                        (PyUnicode_GET_SIZE(**argname) != PyUnicode_GET_SIZE(key)) ? 1 :
                    #endif
                        PyUnicode_Compare(**argname, key);
                    if (cmp < 0 && unlikely(PyErr_Occurred())) goto bad;
                    if (cmp == 0) goto arg_passed_twice;
                    argname++;
                }
            }
        } else
            goto invalid_keyword_type;
        if (kwds2) {
            if (unlikely(PyDict_SetItem(kwds2, key, value))) goto bad;
        } else {
            goto invalid_keyword;
        }
    }
    return 0;
arg_passed_twice:
    __Pyx_RaiseDoubleKeywordsError(function_name, key);
    goto bad;
invalid_keyword_type:
    PyErr_Format(PyExc_TypeError,
        "%.200s() keywords must be strings", function_name);
    goto bad;
invalid_keyword:
    PyErr_Format(PyExc_TypeError,
    #if PY_MAJOR_VERSION < 3
        "%.200s() got an unexpected keyword argument '%.200s'",
        function_name, PyString_AsString(key));
    #else
        "%s() got an unexpected keyword argument '%U'",
        function_name, key);
    #endif
bad:
    return -1;
}

static void __Pyx_RaiseArgtupleInvalid(
    const char* func_name,
    int exact,
    Py_ssize_t num_min,
    Py_ssize_t num_max,
    Py_ssize_t num_found)
{
    Py_ssize_t num_expected;
    const char *more_or_less;
    if (num_found < num_min) {
        num_expected = num_min;
        more_or_less = "at least";
    } else {
        num_expected = num_max;
        more_or_less = "at most";
    }
    if (exact) {
        more_or_less = "exactly";
    }
    PyErr_Format(PyExc_TypeError,
                 "%.200s() takes %.8s %" CYTHON_FORMAT_SSIZE_T "d positional argument%.1s (%" CYTHON_FORMAT_SSIZE_T "d given)",
                 func_name, more_or_less, num_expected,
                 (num_expected == 1) ? "" : "s", num_found);
}

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_Call(PyObject *func, PyObject *arg, PyObject *kw) {
    PyObject *result;
    ternaryfunc call = func->ob_type->tp_call;
    if (unlikely(!call))
        return PyObject_Call(func, arg, kw);
    if (unlikely(Py_EnterRecursiveCall((char*)" while calling a Python object")))
        return NULL;
    result = (*call)(func, arg, kw);
    Py_LeaveRecursiveCall();
    if (unlikely(!result) && unlikely(!PyErr_Occurred())) {
        PyErr_SetString(
            PyExc_SystemError,
            "NULL result without error in PyObject_Call");
    }
    return result;
}
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallMethO(PyObject *func, PyObject *arg) {
    PyObject *self, *result;
    PyCFunction cfunc;
    cfunc = PyCFunction_GET_FUNCTION(func);
    self = PyCFunction_GET_SELF(func);
    if (unlikely(Py_EnterRecursiveCall((char*)" while calling a Python object")))
        return NULL;
    result = cfunc(self, arg);
    Py_LeaveRecursiveCall();
    if (unlikely(!result) && unlikely(!PyErr_Occurred())) {
        PyErr_SetString(
            PyExc_SystemError,
            "NULL result without error in PyObject_Call");
    }
    return result;
}
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static PyObject* __Pyx__PyObject_CallOneArg(PyObject *func, PyObject *arg) {
    PyObject *result;
    PyObject *args = PyTuple_New(1);
    if (unlikely(!args)) return NULL;
    Py_INCREF(arg);
    PyTuple_SET_ITEM(args, 0, arg);
    result = __Pyx_PyObject_Call(func, args, NULL);
    Py_DECREF(args);
    return result;
}
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallOneArg(PyObject *func, PyObject *arg) {
#ifdef __Pyx_CyFunction_USED
    if (likely(PyCFunction_Check(func) || PyObject_TypeCheck(func, __pyx_CyFunctionType))) {
#else
    if (likely(PyCFunction_Check(func))) {
#endif
        if (likely(PyCFunction_GET_FLAGS(func) & METH_O)) {
            return __Pyx_PyObject_CallMethO(func, arg);
        }
    }
    return __Pyx__PyObject_CallOneArg(func, arg);
}
#else
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallOneArg(PyObject *func, PyObject *arg) {
    PyObject* args = PyTuple_Pack(1, arg);
    return (likely(args)) ? __Pyx_PyObject_Call(func, args, NULL) : NULL;
}
#endif

static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Generic(PyObject *o, PyObject* j) {
    PyObject *r;
    if (!j) return NULL;
    r = PyObject_GetItem(o, j);
    Py_DECREF(j);
    return r;
}
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_List_Fast(PyObject *o, Py_ssize_t i,
                                                              int wraparound, int boundscheck) {
#if CYTHON_COMPILING_IN_CPYTHON
    if (wraparound & unlikely(i < 0)) i += PyList_GET_SIZE(o);
    if ((!boundscheck) || likely((0 <= i) & (i < PyList_GET_SIZE(o)))) {
        PyObject *r = PyList_GET_ITEM(o, i);
        Py_INCREF(r);
        return r;
    }
    return __Pyx_GetItemInt_Generic(o, PyInt_FromSsize_t(i));
#else
    return PySequence_GetItem(o, i);
#endif
}
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Tuple_Fast(PyObject *o, Py_ssize_t i,
                                                              int wraparound, int boundscheck) {
#if CYTHON_COMPILING_IN_CPYTHON
    if (wraparound & unlikely(i < 0)) i += PyTuple_GET_SIZE(o);
    if ((!boundscheck) || likely((0 <= i) & (i < PyTuple_GET_SIZE(o)))) {
        PyObject *r = PyTuple_GET_ITEM(o, i);
        Py_INCREF(r);
        return r;
    }
    return __Pyx_GetItemInt_Generic(o, PyInt_FromSsize_t(i));
#else
    return PySequence_GetItem(o, i);
#endif
}
static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Fast(PyObject *o, Py_ssize_t i,
                                                     int is_list, int wraparound, int boundscheck) {
#if CYTHON_COMPILING_IN_CPYTHON
    if (is_list || PyList_CheckExact(o)) {
        Py_ssize_t n = ((!wraparound) | likely(i >= 0)) ? i : i + PyList_GET_SIZE(o);
        if ((!boundscheck) || (likely((n >= 0) & (n < PyList_GET_SIZE(o))))) {
            PyObject *r = PyList_GET_ITEM(o, n);
            Py_INCREF(r);
            return r;
        }
    }
    else if (PyTuple_CheckExact(o)) {
        Py_ssize_t n = ((!wraparound) | likely(i >= 0)) ? i : i + PyTuple_GET_SIZE(o);
        if ((!boundscheck) || likely((n >= 0) & (n < PyTuple_GET_SIZE(o)))) {
            PyObject *r = PyTuple_GET_ITEM(o, n);
            Py_INCREF(r);
            return r;
        }
    } else {
        PySequenceMethods *m = Py_TYPE(o)->tp_as_sequence;
        if (likely(m && m->sq_item)) {
            if (wraparound && unlikely(i < 0) && likely(m->sq_length)) {
                Py_ssize_t l = m->sq_length(o);
                if (likely(l >= 0)) {
                    i += l;
                } else {
                    if (PyErr_ExceptionMatches(PyExc_OverflowError))
                        PyErr_Clear();
                    else
                        return NULL;
                }
            }
            return m->sq_item(o, i);
        }
    }
#else
    if (is_list || PySequence_Check(o)) {
        return PySequence_GetItem(o, i);
    }
#endif
    return __Pyx_GetItemInt_Generic(o, PyInt_FromSsize_t(i));
}

static CYTHON_INLINE PyObject *__Pyx_GetModuleGlobalName(PyObject *name) {
    PyObject *result;
#if CYTHON_COMPILING_IN_CPYTHON
    result = PyDict_GetItem(__pyx_d, name);
    if (likely(result)) {
        Py_INCREF(result);
    } else {
#else
    result = PyObject_GetItem(__pyx_d, name);
    if (!result) {
        PyErr_Clear();
#endif
        result = __Pyx_GetBuiltinName(name);
    }
    return result;
}

static CYTHON_INLINE int __Pyx_SetItemInt_Generic(PyObject *o, PyObject *j, PyObject *v) {
    int r;
    if (!j) return -1;
    r = PyObject_SetItem(o, j, v);
    Py_DECREF(j);
    return r;
}
static CYTHON_INLINE int __Pyx_SetItemInt_Fast(PyObject *o, Py_ssize_t i, PyObject *v,
                                               int is_list, int wraparound, int boundscheck) {
#if CYTHON_COMPILING_IN_CPYTHON
    if (is_list || PyList_CheckExact(o)) {
        Py_ssize_t n = (!wraparound) ? i : ((likely(i >= 0)) ? i : i + PyList_GET_SIZE(o));
        if ((!boundscheck) || likely((n >= 0) & (n < PyList_GET_SIZE(o)))) {
            PyObject* old = PyList_GET_ITEM(o, n);
            Py_INCREF(v);
            PyList_SET_ITEM(o, n, v);
            Py_DECREF(old);
            return 1;
        }
    } else {
        PySequenceMethods *m = Py_TYPE(o)->tp_as_sequence;
        if (likely(m && m->sq_ass_item)) {
            if (wraparound && unlikely(i < 0) && likely(m->sq_length)) {
                Py_ssize_t l = m->sq_length(o);
                if (likely(l >= 0)) {
                    i += l;
                } else {
                    if (PyErr_ExceptionMatches(PyExc_OverflowError))
                        PyErr_Clear();
                    else
                        return -1;
                }
            }
            return m->sq_ass_item(o, i, v);
        }
    }
#else
#if CYTHON_COMPILING_IN_PYPY
    if (is_list || (PySequence_Check(o) && !PyDict_Check(o))) {
#else
    if (is_list || PySequence_Check(o)) {
#endif
        return PySequence_SetItem(o, i, v);
    }
#endif
    return __Pyx_SetItemInt_Generic(o, PyInt_FromSsize_t(i), v);
}

#if !CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyBytes_Join(PyObject* sep, PyObject* values) {
    return PyObject_CallMethodObjArgs(sep, __pyx_n_s_join, values, NULL);
}
#endif

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE PyObject* __Pyx_PyObject_CallNoArg(PyObject *func) {
#ifdef __Pyx_CyFunction_USED
    if (likely(PyCFunction_Check(func) || PyObject_TypeCheck(func, __pyx_CyFunctionType))) {
#else
    if (likely(PyCFunction_Check(func))) {
#endif
        if (likely(PyCFunction_GET_FLAGS(func) & METH_NOARGS)) {
            return __Pyx_PyObject_CallMethO(func, NULL);
        }
    }
    return __Pyx_PyObject_Call(func, __pyx_empty_tuple, NULL);
}
#endif

static CYTHON_INLINE int __Pyx_TypeTest(PyObject *obj, PyTypeObject *type) {
    if (unlikely(!type)) {
        PyErr_SetString(PyExc_SystemError, "Missing type object");
        return 0;
    }
    if (likely(PyObject_TypeCheck(obj, type)))
        return 1;
    PyErr_Format(PyExc_TypeError, "Cannot convert %.200s to %.200s",
                 Py_TYPE(obj)->tp_name, type->tp_name);
    return 0;
}

#if CYTHON_COMPILING_IN_CPYTHON
static CYTHON_INLINE void __Pyx_crop_slice(Py_ssize_t* _start, Py_ssize_t* _stop, Py_ssize_t* _length) {
    Py_ssize_t start = *_start, stop = *_stop, length = *_length;
    if (start < 0) {
        start += length;
        if (start < 0)
            start = 0;
    }
    if (stop < 0)
        stop += length;
    else if (stop > length)
        stop = length;
    *_length = stop - start;
    *_start = start;
    *_stop = stop;
}
static CYTHON_INLINE void __Pyx_copy_object_array(PyObject** CYTHON_RESTRICT src, PyObject** CYTHON_RESTRICT dest, Py_ssize_t length) {
    PyObject *v;
    Py_ssize_t i;
    for (i = 0; i < length; i++) {
        v = dest[i] = src[i];
        Py_INCREF(v);
    }
}
static CYTHON_INLINE PyObject* __Pyx_PyList_GetSlice(
            PyObject* src, Py_ssize_t start, Py_ssize_t stop) {
    PyObject* dest;
    Py_ssize_t length = PyList_GET_SIZE(src);
    __Pyx_crop_slice(&start, &stop, &length);
    if (unlikely(length <= 0))
        return PyList_New(0);
    dest = PyList_New(length);
    if (unlikely(!dest))
        return NULL;
    __Pyx_copy_object_array(
        ((PyListObject*)src)->ob_item + start,
        ((PyListObject*)dest)->ob_item,
        length);
    return dest;
}
static CYTHON_INLINE PyObject* __Pyx_PyTuple_GetSlice(
            PyObject* src, Py_ssize_t start, Py_ssize_t stop) {
    PyObject* dest;
    Py_ssize_t length = PyTuple_GET_SIZE(src);
    __Pyx_crop_slice(&start, &stop, &length);
    if (unlikely(length <= 0))
        return PyTuple_New(0);
    dest = PyTuple_New(length);
    if (unlikely(!dest))
        return NULL;
    __Pyx_copy_object_array(
        ((PyTupleObject*)src)->ob_item + start,
        ((PyTupleObject*)dest)->ob_item,
        length);
    return dest;
}
#endif

static CYTHON_INLINE void __Pyx_ErrRestore(PyObject *type, PyObject *value, PyObject *tb) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyObject *tmp_type, *tmp_value, *tmp_tb;
    PyThreadState *tstate = PyThreadState_GET();
    tmp_type = tstate->curexc_type;
    tmp_value = tstate->curexc_value;
    tmp_tb = tstate->curexc_traceback;
    tstate->curexc_type = type;
    tstate->curexc_value = value;
    tstate->curexc_traceback = tb;
    Py_XDECREF(tmp_type);
    Py_XDECREF(tmp_value);
    Py_XDECREF(tmp_tb);
#else
    PyErr_Restore(type, value, tb);
#endif
}
static CYTHON_INLINE void __Pyx_ErrFetch(PyObject **type, PyObject **value, PyObject **tb) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyThreadState *tstate = PyThreadState_GET();
    *type = tstate->curexc_type;
    *value = tstate->curexc_value;
    *tb = tstate->curexc_traceback;
    tstate->curexc_type = 0;
    tstate->curexc_value = 0;
    tstate->curexc_traceback = 0;
#else
    PyErr_Fetch(type, value, tb);
#endif
}

#if PY_MAJOR_VERSION < 3
static void __Pyx_Raise(PyObject *type, PyObject *value, PyObject *tb,
                        CYTHON_UNUSED PyObject *cause) {
    Py_XINCREF(type);
    if (!value || value == Py_None)
        value = NULL;
    else
        Py_INCREF(value);
    if (!tb || tb == Py_None)
        tb = NULL;
    else {
        Py_INCREF(tb);
        if (!PyTraceBack_Check(tb)) {
            PyErr_SetString(PyExc_TypeError,
                "raise: arg 3 must be a traceback or None");
            goto raise_error;
        }
    }
    if (PyType_Check(type)) {
#if CYTHON_COMPILING_IN_PYPY
        if (!value) {
            Py_INCREF(Py_None);
            value = Py_None;
        }
#endif
        PyErr_NormalizeException(&type, &value, &tb);
    } else {
        if (value) {
            PyErr_SetString(PyExc_TypeError,
                "instance exception may not have a separate value");
            goto raise_error;
        }
        value = type;
        type = (PyObject*) Py_TYPE(type);
        Py_INCREF(type);
        if (!PyType_IsSubtype((PyTypeObject *)type, (PyTypeObject *)PyExc_BaseException)) {
            PyErr_SetString(PyExc_TypeError,
                "raise: exception class must be a subclass of BaseException");
            goto raise_error;
        }
    }
    __Pyx_ErrRestore(type, value, tb);
    return;
raise_error:
    Py_XDECREF(value);
    Py_XDECREF(type);
    Py_XDECREF(tb);
    return;
}
#else
static void __Pyx_Raise(PyObject *type, PyObject *value, PyObject *tb, PyObject *cause) {
    PyObject* owned_instance = NULL;
    if (tb == Py_None) {
        tb = 0;
    } else if (tb && !PyTraceBack_Check(tb)) {
        PyErr_SetString(PyExc_TypeError,
            "raise: arg 3 must be a traceback or None");
        goto bad;
    }
    if (value == Py_None)
        value = 0;
    if (PyExceptionInstance_Check(type)) {
        if (value) {
            PyErr_SetString(PyExc_TypeError,
                "instance exception may not have a separate value");
            goto bad;
        }
        value = type;
        type = (PyObject*) Py_TYPE(value);
    } else if (PyExceptionClass_Check(type)) {
        PyObject *instance_class = NULL;
        if (value && PyExceptionInstance_Check(value)) {
            instance_class = (PyObject*) Py_TYPE(value);
            if (instance_class != type) {
                if (PyObject_IsSubclass(instance_class, type)) {
                    type = instance_class;
                } else {
                    instance_class = NULL;
                }
            }
        }
        if (!instance_class) {
            PyObject *args;
            if (!value)
                args = PyTuple_New(0);
            else if (PyTuple_Check(value)) {
                Py_INCREF(value);
                args = value;
            } else
                args = PyTuple_Pack(1, value);
            if (!args)
                goto bad;
            owned_instance = PyObject_Call(type, args, NULL);
            Py_DECREF(args);
            if (!owned_instance)
                goto bad;
            value = owned_instance;
            if (!PyExceptionInstance_Check(value)) {
                PyErr_Format(PyExc_TypeError,
                             "calling %R should have returned an instance of "
                             "BaseException, not %R",
                             type, Py_TYPE(value));
                goto bad;
            }
        }
    } else {
        PyErr_SetString(PyExc_TypeError,
            "raise: exception class must be a subclass of BaseException");
        goto bad;
    }
#if PY_VERSION_HEX >= 0x03030000
    if (cause) {
#else
    if (cause && cause != Py_None) {
#endif
        PyObject *fixed_cause;
        if (cause == Py_None) {
            fixed_cause = NULL;
        } else if (PyExceptionClass_Check(cause)) {
            fixed_cause = PyObject_CallObject(cause, NULL);
            if (fixed_cause == NULL)
                goto bad;
        } else if (PyExceptionInstance_Check(cause)) {
            fixed_cause = cause;
            Py_INCREF(fixed_cause);
        } else {
            PyErr_SetString(PyExc_TypeError,
                            "exception causes must derive from "
                            "BaseException");
            goto bad;
        }
        PyException_SetCause(value, fixed_cause);
    }
    PyErr_SetObject(type, value);
    if (tb) {
        PyThreadState *tstate = PyThreadState_GET();
        PyObject* tmp_tb = tstate->curexc_traceback;
        if (tb != tmp_tb) {
            Py_INCREF(tb);
            tstate->curexc_traceback = tb;
            Py_XDECREF(tmp_tb);
        }
    }
bad:
    Py_XDECREF(owned_instance);
    return;
}
#endif

static CYTHON_INLINE int __Pyx_PyBytes_Equals(PyObject* s1, PyObject* s2, int equals) {
#if CYTHON_COMPILING_IN_PYPY
    return PyObject_RichCompareBool(s1, s2, equals);
#else
    if (s1 == s2) {
        return (equals == Py_EQ);
    } else if (PyBytes_CheckExact(s1) & PyBytes_CheckExact(s2)) {
        const char *ps1, *ps2;
        Py_ssize_t length = PyBytes_GET_SIZE(s1);
        if (length != PyBytes_GET_SIZE(s2))
            return (equals == Py_NE);
        ps1 = PyBytes_AS_STRING(s1);
        ps2 = PyBytes_AS_STRING(s2);
        if (ps1[0] != ps2[0]) {
            return (equals == Py_NE);
        } else if (length == 1) {
            return (equals == Py_EQ);
        } else {
            int result = memcmp(ps1, ps2, (size_t)length);
            return (equals == Py_EQ) ? (result == 0) : (result != 0);
        }
    } else if ((s1 == Py_None) & PyBytes_CheckExact(s2)) {
        return (equals == Py_NE);
    } else if ((s2 == Py_None) & PyBytes_CheckExact(s1)) {
        return (equals == Py_NE);
    } else {
        int result;
        PyObject* py_result = PyObject_RichCompare(s1, s2, equals);
        if (!py_result)
            return -1;
        result = __Pyx_PyObject_IsTrue(py_result);
        Py_DECREF(py_result);
        return result;
    }
#endif
}

static CYTHON_INLINE int __Pyx_PyUnicode_Equals(PyObject* s1, PyObject* s2, int equals) {
#if CYTHON_COMPILING_IN_PYPY
    return PyObject_RichCompareBool(s1, s2, equals);
#else
#if PY_MAJOR_VERSION < 3
    PyObject* owned_ref = NULL;
#endif
    int s1_is_unicode, s2_is_unicode;
    if (s1 == s2) {
        goto return_eq;
    }
    s1_is_unicode = PyUnicode_CheckExact(s1);
    s2_is_unicode = PyUnicode_CheckExact(s2);
#if PY_MAJOR_VERSION < 3
    if ((s1_is_unicode & (!s2_is_unicode)) && PyString_CheckExact(s2)) {
        owned_ref = PyUnicode_FromObject(s2);
        if (unlikely(!owned_ref))
            return -1;
        s2 = owned_ref;
        s2_is_unicode = 1;
    } else if ((s2_is_unicode & (!s1_is_unicode)) && PyString_CheckExact(s1)) {
        owned_ref = PyUnicode_FromObject(s1);
        if (unlikely(!owned_ref))
            return -1;
        s1 = owned_ref;
        s1_is_unicode = 1;
    } else if (((!s2_is_unicode) & (!s1_is_unicode))) {
        return __Pyx_PyBytes_Equals(s1, s2, equals);
    }
#endif
    if (s1_is_unicode & s2_is_unicode) {
        Py_ssize_t length;
        int kind;
        void *data1, *data2;
        if (unlikely(__Pyx_PyUnicode_READY(s1) < 0) || unlikely(__Pyx_PyUnicode_READY(s2) < 0))
            return -1;
        length = __Pyx_PyUnicode_GET_LENGTH(s1);
        if (length != __Pyx_PyUnicode_GET_LENGTH(s2)) {
            goto return_ne;
        }
        kind = __Pyx_PyUnicode_KIND(s1);
        if (kind != __Pyx_PyUnicode_KIND(s2)) {
            goto return_ne;
        }
        data1 = __Pyx_PyUnicode_DATA(s1);
        data2 = __Pyx_PyUnicode_DATA(s2);
        if (__Pyx_PyUnicode_READ(kind, data1, 0) != __Pyx_PyUnicode_READ(kind, data2, 0)) {
            goto return_ne;
        } else if (length == 1) {
            goto return_eq;
        } else {
            int result = memcmp(data1, data2, (size_t)(length * kind));
            #if PY_MAJOR_VERSION < 3
            Py_XDECREF(owned_ref);
            #endif
            return (equals == Py_EQ) ? (result == 0) : (result != 0);
        }
    } else if ((s1 == Py_None) & s2_is_unicode) {
        goto return_ne;
    } else if ((s2 == Py_None) & s1_is_unicode) {
        goto return_ne;
    } else {
        int result;
        PyObject* py_result = PyObject_RichCompare(s1, s2, equals);
        if (!py_result)
            return -1;
        result = __Pyx_PyObject_IsTrue(py_result);
        Py_DECREF(py_result);
        return result;
    }
return_eq:
    #if PY_MAJOR_VERSION < 3
    Py_XDECREF(owned_ref);
    #endif
    return (equals == Py_EQ);
return_ne:
    #if PY_MAJOR_VERSION < 3
    Py_XDECREF(owned_ref);
    #endif
    return (equals == Py_NE);
#endif
}

static void __Pyx_RaiseArgumentTypeInvalid(const char* name, PyObject *obj, PyTypeObject *type) {
    PyErr_Format(PyExc_TypeError,
        "Argument '%.200s' has incorrect type (expected %.200s, got %.200s)",
        name, type->tp_name, Py_TYPE(obj)->tp_name);
}
static CYTHON_INLINE int __Pyx_ArgTypeTest(PyObject *obj, PyTypeObject *type, int none_allowed,
    const char *name, int exact)
{
    if (unlikely(!type)) {
        PyErr_SetString(PyExc_SystemError, "Missing type object");
        return 0;
    }
    if (none_allowed && obj == Py_None) return 1;
    else if (exact) {
        if (likely(Py_TYPE(obj) == type)) return 1;
        #if PY_MAJOR_VERSION == 2
        else if ((type == &PyBaseString_Type) && likely(__Pyx_PyBaseString_CheckExact(obj))) return 1;
        #endif
    }
    else {
        if (likely(PyObject_TypeCheck(obj, type))) return 1;
    }
    __Pyx_RaiseArgumentTypeInvalid(name, obj, type);
    return 0;
}

static PyObject* __Pyx_PyObject_CallMethod1(PyObject* obj, PyObject* method_name, PyObject* arg) {
    PyObject *method, *result = NULL;
    method = __Pyx_PyObject_GetAttrStr(obj, method_name);
    if (unlikely(!method)) goto bad;
#if CYTHON_COMPILING_IN_CPYTHON
    if (likely(PyMethod_Check(method))) {
        PyObject *self = PyMethod_GET_SELF(method);
        if (likely(self)) {
            PyObject *args;
            PyObject *function = PyMethod_GET_FUNCTION(method);
            args = PyTuple_New(2);
            if (unlikely(!args)) goto bad;
            Py_INCREF(self);
            PyTuple_SET_ITEM(args, 0, self);
            Py_INCREF(arg);
            PyTuple_SET_ITEM(args, 1, arg);
            Py_INCREF(function);
            Py_DECREF(method); method = NULL;
            result = __Pyx_PyObject_Call(function, args, NULL);
            Py_DECREF(args);
            Py_DECREF(function);
            return result;
        }
    }
#endif
    result = __Pyx_PyObject_CallOneArg(method, arg);
bad:
    Py_XDECREF(method);
    return result;
}

static CYTHON_INLINE int __Pyx_PyObject_Append(PyObject* L, PyObject* x) {
    if (likely(PyList_CheckExact(L))) {
        if (unlikely(__Pyx_PyList_Append(L, x) < 0)) return -1;
    } else {
        PyObject* retval = __Pyx_PyObject_CallMethod1(L, __pyx_n_s_append, x);
        if (unlikely(!retval))
            return -1;
        Py_DECREF(retval);
    }
    return 0;
}

static void __Pyx_WriteUnraisable(const char *name, CYTHON_UNUSED int clineno,
                                  CYTHON_UNUSED int lineno, CYTHON_UNUSED const char *filename,
                                  int full_traceback) {
    PyObject *old_exc, *old_val, *old_tb;
    PyObject *ctx;
    __Pyx_ErrFetch(&old_exc, &old_val, &old_tb);
    if (full_traceback) {
        Py_XINCREF(old_exc);
        Py_XINCREF(old_val);
        Py_XINCREF(old_tb);
        __Pyx_ErrRestore(old_exc, old_val, old_tb);
        PyErr_PrintEx(1);
    }
    #if PY_MAJOR_VERSION < 3
    ctx = PyString_FromString(name);
    #else
    ctx = PyUnicode_FromString(name);
    #endif
    __Pyx_ErrRestore(old_exc, old_val, old_tb);
    if (!ctx) {
        PyErr_WriteUnraisable(Py_None);
    } else {
        PyErr_WriteUnraisable(ctx);
        Py_DECREF(ctx);
    }
}

static CYTHON_INLINE int __Pyx_IterFinish(void) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyThreadState *tstate = PyThreadState_GET();
    PyObject* exc_type = tstate->curexc_type;
    if (unlikely(exc_type)) {
        if (likely(exc_type == PyExc_StopIteration) || PyErr_GivenExceptionMatches(exc_type, PyExc_StopIteration)) {
            PyObject *exc_value, *exc_tb;
            exc_value = tstate->curexc_value;
            exc_tb = tstate->curexc_traceback;
            tstate->curexc_type = 0;
            tstate->curexc_value = 0;
            tstate->curexc_traceback = 0;
            Py_DECREF(exc_type);
            Py_XDECREF(exc_value);
            Py_XDECREF(exc_tb);
            return 0;
        } else {
            return -1;
        }
    }
    return 0;
#else
    if (unlikely(PyErr_Occurred())) {
        if (likely(PyErr_ExceptionMatches(PyExc_StopIteration))) {
            PyErr_Clear();
            return 0;
        } else {
            return -1;
        }
    }
    return 0;
#endif
}

static PyObject* __Pyx_PyObject_CallMethod0(PyObject* obj, PyObject* method_name) {
    PyObject *method, *result = NULL;
    method = __Pyx_PyObject_GetAttrStr(obj, method_name);
    if (unlikely(!method)) goto bad;
#if CYTHON_COMPILING_IN_CPYTHON
    if (likely(PyMethod_Check(method))) {
        PyObject *self = PyMethod_GET_SELF(method);
        if (likely(self)) {
            PyObject *function = PyMethod_GET_FUNCTION(method);
            result = __Pyx_PyObject_CallOneArg(function, self);
            Py_DECREF(method);
            return result;
        }
    }
#endif
    result = __Pyx_PyObject_CallNoArg(method);
    Py_DECREF(method);
bad:
    return result;
}

static CYTHON_INLINE void __Pyx_RaiseNeedMoreValuesError(Py_ssize_t index) {
    PyErr_Format(PyExc_ValueError,
                 "need more than %" CYTHON_FORMAT_SSIZE_T "d value%.1s to unpack",
                 index, (index == 1) ? "" : "s");
}

static CYTHON_INLINE void __Pyx_RaiseTooManyValuesError(Py_ssize_t expected) {
    PyErr_Format(PyExc_ValueError,
                 "too many values to unpack (expected %" CYTHON_FORMAT_SSIZE_T "d)", expected);
}

static int __Pyx_IternextUnpackEndCheck(PyObject *retval, Py_ssize_t expected) {
    if (unlikely(retval)) {
        Py_DECREF(retval);
        __Pyx_RaiseTooManyValuesError(expected);
        return -1;
    } else {
        return __Pyx_IterFinish();
    }
    return 0;
}

static CYTHON_INLINE void __Pyx_RaiseNoneNotIterableError(void) {
    PyErr_SetString(PyExc_TypeError, "'NoneType' object is not iterable");
}

static void __Pyx_UnpackTupleError(PyObject *t, Py_ssize_t index) {
    if (t == Py_None) {
      __Pyx_RaiseNoneNotIterableError();
    } else if (PyTuple_GET_SIZE(t) < index) {
      __Pyx_RaiseNeedMoreValuesError(PyTuple_GET_SIZE(t));
    } else {
      __Pyx_RaiseTooManyValuesError(index);
    }
}

static CYTHON_INLINE int __Pyx_unpack_tuple2(PyObject* tuple, PyObject** pvalue1, PyObject** pvalue2,
                                             int is_tuple, int has_known_size, int decref_tuple) {
    Py_ssize_t index;
    PyObject *value1 = NULL, *value2 = NULL, *iter = NULL;
    if (!is_tuple && unlikely(!PyTuple_Check(tuple))) {
        iternextfunc iternext;
        iter = PyObject_GetIter(tuple);
        if (unlikely(!iter)) goto bad;
        if (decref_tuple) { Py_DECREF(tuple); tuple = NULL; }
        iternext = Py_TYPE(iter)->tp_iternext;
        value1 = iternext(iter); if (unlikely(!value1)) { index = 0; goto unpacking_failed; }
        value2 = iternext(iter); if (unlikely(!value2)) { index = 1; goto unpacking_failed; }
        if (!has_known_size && unlikely(__Pyx_IternextUnpackEndCheck(iternext(iter), 2))) goto bad;
        Py_DECREF(iter);
    } else {
        if (!has_known_size && unlikely(PyTuple_GET_SIZE(tuple) != 2)) {
            __Pyx_UnpackTupleError(tuple, 2);
            goto bad;
        }
#if CYTHON_COMPILING_IN_PYPY
        value1 = PySequence_ITEM(tuple, 0);
        if (unlikely(!value1)) goto bad;
        value2 = PySequence_ITEM(tuple, 1);
        if (unlikely(!value2)) goto bad;
#else
        value1 = PyTuple_GET_ITEM(tuple, 0);
        value2 = PyTuple_GET_ITEM(tuple, 1);
        Py_INCREF(value1);
        Py_INCREF(value2);
#endif
        if (decref_tuple) { Py_DECREF(tuple); }
    }
    *pvalue1 = value1;
    *pvalue2 = value2;
    return 0;
unpacking_failed:
    if (!has_known_size && __Pyx_IterFinish() == 0)
        __Pyx_RaiseNeedMoreValuesError(index);
bad:
    Py_XDECREF(iter);
    Py_XDECREF(value1);
    Py_XDECREF(value2);
    if (decref_tuple) { Py_XDECREF(tuple); }
    return -1;
}

static CYTHON_INLINE PyObject* __Pyx_dict_iterator(PyObject* iterable, int is_dict, PyObject* method_name,
                                                   Py_ssize_t* p_orig_length, int* p_source_is_dict) {
    is_dict = is_dict || likely(PyDict_CheckExact(iterable));
    *p_source_is_dict = is_dict;
#if !CYTHON_COMPILING_IN_PYPY
    if (is_dict) {
        *p_orig_length = PyDict_Size(iterable);
        Py_INCREF(iterable);
        return iterable;
    }
#endif
    *p_orig_length = 0;
    if (method_name) {
        PyObject* iter;
        iterable = __Pyx_PyObject_CallMethod0(iterable, method_name);
        if (!iterable)
            return NULL;
#if !CYTHON_COMPILING_IN_PYPY
        if (PyTuple_CheckExact(iterable) || PyList_CheckExact(iterable))
            return iterable;
#endif
        iter = PyObject_GetIter(iterable);
        Py_DECREF(iterable);
        return iter;
    }
    return PyObject_GetIter(iterable);
}
static CYTHON_INLINE int __Pyx_dict_iter_next(PyObject* iter_obj, Py_ssize_t orig_length, Py_ssize_t* ppos,
                                              PyObject** pkey, PyObject** pvalue, PyObject** pitem, int source_is_dict) {
    PyObject* next_item;
#if !CYTHON_COMPILING_IN_PYPY
    if (source_is_dict) {
        PyObject *key, *value;
        if (unlikely(orig_length != PyDict_Size(iter_obj))) {
            PyErr_SetString(PyExc_RuntimeError, "dictionary changed size during iteration");
            return -1;
        }
        if (unlikely(!PyDict_Next(iter_obj, ppos, &key, &value))) {
            return 0;
        }
        if (pitem) {
            PyObject* tuple = PyTuple_New(2);
            if (unlikely(!tuple)) {
                return -1;
            }
            Py_INCREF(key);
            Py_INCREF(value);
            PyTuple_SET_ITEM(tuple, 0, key);
            PyTuple_SET_ITEM(tuple, 1, value);
            *pitem = tuple;
        } else {
            if (pkey) {
                Py_INCREF(key);
                *pkey = key;
            }
            if (pvalue) {
                Py_INCREF(value);
                *pvalue = value;
            }
        }
        return 1;
    } else if (PyTuple_CheckExact(iter_obj)) {
        Py_ssize_t pos = *ppos;
        if (unlikely(pos >= PyTuple_GET_SIZE(iter_obj))) return 0;
        *ppos = pos + 1;
        next_item = PyTuple_GET_ITEM(iter_obj, pos);
        Py_INCREF(next_item);
    } else if (PyList_CheckExact(iter_obj)) {
        Py_ssize_t pos = *ppos;
        if (unlikely(pos >= PyList_GET_SIZE(iter_obj))) return 0;
        *ppos = pos + 1;
        next_item = PyList_GET_ITEM(iter_obj, pos);
        Py_INCREF(next_item);
    } else
#endif
    {
        next_item = PyIter_Next(iter_obj);
        if (unlikely(!next_item)) {
            return __Pyx_IterFinish();
        }
    }
    if (pitem) {
        *pitem = next_item;
    } else if (pkey && pvalue) {
        if (__Pyx_unpack_tuple2(next_item, pkey, pvalue, source_is_dict, source_is_dict, 1))
            return -1;
    } else if (pkey) {
        *pkey = next_item;
    } else {
        *pvalue = next_item;
    }
    return 1;
}

static PyObject* __Pyx_ImportFrom(PyObject* module, PyObject* name) {
    PyObject* value = __Pyx_PyObject_GetAttrStr(module, name);
    if (unlikely(!value) && PyErr_ExceptionMatches(PyExc_AttributeError)) {
        PyErr_Format(PyExc_ImportError,
        #if PY_MAJOR_VERSION < 3
            "cannot import name %.230s", PyString_AS_STRING(name));
        #else
            "cannot import name %S", name);
        #endif
    }
    return value;
}

static CYTHON_INLINE void __Pyx_RaiseUnboundLocalError(const char *varname) {
    PyErr_Format(PyExc_UnboundLocalError, "local variable '%s' referenced before assignment", varname);
}

static PyObject* __Pyx__PyObject_PopIndex(PyObject* L, PyObject* py_ix) {
    PyObject *r;
    if (unlikely(!py_ix)) return NULL;
    r = __Pyx_PyObject_CallMethod1(L, __pyx_n_s_pop, py_ix);
    Py_DECREF(py_ix);
    return r;
}
static PyObject* __Pyx__PyList_PopIndex(PyObject* L, Py_ssize_t ix) {
#if CYTHON_COMPILING_IN_CPYTHON
    Py_ssize_t size = PyList_GET_SIZE(L);
    if (likely(size > (((PyListObject*)L)->allocated >> 1))) {
        Py_ssize_t cix = ix;
        if (cix < 0) {
            cix += size;
        }
        if (likely(0 <= cix && cix < size)) {
            PyObject* v = PyList_GET_ITEM(L, cix);
            Py_SIZE(L) -= 1;
            size -= 1;
            memmove(&PyList_GET_ITEM(L, cix), &PyList_GET_ITEM(L, cix+1), (size_t)(size-cix)*sizeof(PyObject*));
            return v;
        }
    }
#endif
    return __Pyx__PyObject_PopIndex(L, PyInt_FromSsize_t(ix));
}

static CYTHON_INLINE PyObject* __Pyx_PyObject_GetSlice(
        PyObject* obj, Py_ssize_t cstart, Py_ssize_t cstop,
        PyObject** _py_start, PyObject** _py_stop, PyObject** _py_slice,
        int has_cstart, int has_cstop, CYTHON_UNUSED int wraparound) {
#if CYTHON_COMPILING_IN_CPYTHON
    PyMappingMethods* mp;
#if PY_MAJOR_VERSION < 3
    PySequenceMethods* ms = Py_TYPE(obj)->tp_as_sequence;
    if (likely(ms && ms->sq_slice)) {
        if (!has_cstart) {
            if (_py_start && (*_py_start != Py_None)) {
                cstart = __Pyx_PyIndex_AsSsize_t(*_py_start);
                if ((cstart == (Py_ssize_t)-1) && PyErr_Occurred()) goto bad;
            } else
                cstart = 0;
        }
        if (!has_cstop) {
            if (_py_stop && (*_py_stop != Py_None)) {
                cstop = __Pyx_PyIndex_AsSsize_t(*_py_stop);
                if ((cstop == (Py_ssize_t)-1) && PyErr_Occurred()) goto bad;
            } else
                cstop = PY_SSIZE_T_MAX;
        }
        if (wraparound && unlikely((cstart < 0) | (cstop < 0)) && likely(ms->sq_length)) {
            Py_ssize_t l = ms->sq_length(obj);
            if (likely(l >= 0)) {
                if (cstop < 0) {
                    cstop += l;
                    if (cstop < 0) cstop = 0;
                }
                if (cstart < 0) {
                    cstart += l;
                    if (cstart < 0) cstart = 0;
                }
            } else {
                if (PyErr_ExceptionMatches(PyExc_OverflowError))
                    PyErr_Clear();
                else
                    goto bad;
            }
        }
        return ms->sq_slice(obj, cstart, cstop);
    }
#endif
    mp = Py_TYPE(obj)->tp_as_mapping;
    if (likely(mp && mp->mp_subscript))
#endif
    {
        PyObject* result;
        PyObject *py_slice, *py_start, *py_stop;
        if (_py_slice) {
            py_slice = *_py_slice;
        } else {
            PyObject* owned_start = NULL;
            PyObject* owned_stop = NULL;
            if (_py_start) {
                py_start = *_py_start;
            } else {
                if (has_cstart) {
                    owned_start = py_start = PyInt_FromSsize_t(cstart);
                    if (unlikely(!py_start)) goto bad;
                } else
                    py_start = Py_None;
            }
            if (_py_stop) {
                py_stop = *_py_stop;
            } else {
                if (has_cstop) {
                    owned_stop = py_stop = PyInt_FromSsize_t(cstop);
                    if (unlikely(!py_stop)) {
                        Py_XDECREF(owned_start);
                        goto bad;
                    }
                } else
                    py_stop = Py_None;
            }
            py_slice = PySlice_New(py_start, py_stop, Py_None);
            Py_XDECREF(owned_start);
            Py_XDECREF(owned_stop);
            if (unlikely(!py_slice)) goto bad;
        }
#if CYTHON_COMPILING_IN_CPYTHON
        result = mp->mp_subscript(obj, py_slice);
#else
        result = PyObject_GetItem(obj, py_slice);
#endif
        if (!_py_slice) {
            Py_DECREF(py_slice);
        }
        return result;
    }
    PyErr_Format(PyExc_TypeError,
        "'%.200s' object is unsliceable", Py_TYPE(obj)->tp_name);
bad:
    return NULL;
}

static PyTypeObject* __Pyx_FetchCommonType(PyTypeObject* type) {
    PyObject* fake_module;
    PyTypeObject* cached_type = NULL;
    fake_module = PyImport_AddModule((char*) "_cython_" CYTHON_ABI);
    if (!fake_module) return NULL;
    Py_INCREF(fake_module);
    cached_type = (PyTypeObject*) PyObject_GetAttrString(fake_module, type->tp_name);
    if (cached_type) {
        if (!PyType_Check((PyObject*)cached_type)) {
            PyErr_Format(PyExc_TypeError,
                "Shared Cython type %.200s is not a type object",
                type->tp_name);
            goto bad;
        }
        if (cached_type->tp_basicsize != type->tp_basicsize) {
            PyErr_Format(PyExc_TypeError,
                "Shared Cython type %.200s has the wrong size, try recompiling",
                type->tp_name);
            goto bad;
        }
    } else {
        if (!PyErr_ExceptionMatches(PyExc_AttributeError)) goto bad;
        PyErr_Clear();
        if (PyType_Ready(type) < 0) goto bad;
        if (PyObject_SetAttrString(fake_module, type->tp_name, (PyObject*) type) < 0)
            goto bad;
        Py_INCREF(type);
        cached_type = type;
    }
done:
    Py_DECREF(fake_module);
    return cached_type;
bad:
    Py_XDECREF(cached_type);
    cached_type = NULL;
    goto done;
}

static PyObject *
__Pyx_CyFunction_get_doc(__pyx_CyFunctionObject *op, CYTHON_UNUSED void *closure)
{
    if (unlikely(op->func_doc == NULL)) {
        if (op->func.m_ml->ml_doc) {
#if PY_MAJOR_VERSION >= 3
            op->func_doc = PyUnicode_FromString(op->func.m_ml->ml_doc);
#else
            op->func_doc = PyString_FromString(op->func.m_ml->ml_doc);
#endif
            if (unlikely(op->func_doc == NULL))
                return NULL;
        } else {
            Py_INCREF(Py_None);
            return Py_None;
        }
    }
    Py_INCREF(op->func_doc);
    return op->func_doc;
}
static int
__Pyx_CyFunction_set_doc(__pyx_CyFunctionObject *op, PyObject *value)
{
    PyObject *tmp = op->func_doc;
    if (value == NULL) {
        value = Py_None;
    }
    Py_INCREF(value);
    op->func_doc = value;
    Py_XDECREF(tmp);
    return 0;
}
static PyObject *
__Pyx_CyFunction_get_name(__pyx_CyFunctionObject *op)
{
    if (unlikely(op->func_name == NULL)) {
#if PY_MAJOR_VERSION >= 3
        op->func_name = PyUnicode_InternFromString(op->func.m_ml->ml_name);
#else
        op->func_name = PyString_InternFromString(op->func.m_ml->ml_name);
#endif
        if (unlikely(op->func_name == NULL))
            return NULL;
    }
    Py_INCREF(op->func_name);
    return op->func_name;
}
static int
__Pyx_CyFunction_set_name(__pyx_CyFunctionObject *op, PyObject *value)
{
    PyObject *tmp;
#if PY_MAJOR_VERSION >= 3
    if (unlikely(value == NULL || !PyUnicode_Check(value))) {
#else
    if (unlikely(value == NULL || !PyString_Check(value))) {
#endif
        PyErr_SetString(PyExc_TypeError,
                        "__name__ must be set to a string object");
        return -1;
    }
    tmp = op->func_name;
    Py_INCREF(value);
    op->func_name = value;
    Py_XDECREF(tmp);
    return 0;
}
static PyObject *
__Pyx_CyFunction_get_qualname(__pyx_CyFunctionObject *op)
{
    Py_INCREF(op->func_qualname);
    return op->func_qualname;
}
static int
__Pyx_CyFunction_set_qualname(__pyx_CyFunctionObject *op, PyObject *value)
{
    PyObject *tmp;
#if PY_MAJOR_VERSION >= 3
    if (unlikely(value == NULL || !PyUnicode_Check(value))) {
#else
    if (unlikely(value == NULL || !PyString_Check(value))) {
#endif
        PyErr_SetString(PyExc_TypeError,
                        "__qualname__ must be set to a string object");
        return -1;
    }
    tmp = op->func_qualname;
    Py_INCREF(value);
    op->func_qualname = value;
    Py_XDECREF(tmp);
    return 0;
}
static PyObject *
__Pyx_CyFunction_get_self(__pyx_CyFunctionObject *m, CYTHON_UNUSED void *closure)
{
    PyObject *self;
    self = m->func_closure;
    if (self == NULL)
        self = Py_None;
    Py_INCREF(self);
    return self;
}
static PyObject *
__Pyx_CyFunction_get_dict(__pyx_CyFunctionObject *op)
{
    if (unlikely(op->func_dict == NULL)) {
        op->func_dict = PyDict_New();
        if (unlikely(op->func_dict == NULL))
            return NULL;
    }
    Py_INCREF(op->func_dict);
    return op->func_dict;
}
static int
__Pyx_CyFunction_set_dict(__pyx_CyFunctionObject *op, PyObject *value)
{
    PyObject *tmp;
    if (unlikely(value == NULL)) {
        PyErr_SetString(PyExc_TypeError,
               "function's dictionary may not be deleted");
        return -1;
    }
    if (unlikely(!PyDict_Check(value))) {
        PyErr_SetString(PyExc_TypeError,
               "setting function's dictionary to a non-dict");
        return -1;
    }
    tmp = op->func_dict;
    Py_INCREF(value);
    op->func_dict = value;
    Py_XDECREF(tmp);
    return 0;
}
static PyObject *
__Pyx_CyFunction_get_globals(__pyx_CyFunctionObject *op)
{
    Py_INCREF(op->func_globals);
    return op->func_globals;
}
static PyObject *
__Pyx_CyFunction_get_closure(CYTHON_UNUSED __pyx_CyFunctionObject *op)
{
    Py_INCREF(Py_None);
    return Py_None;
}
static PyObject *
__Pyx_CyFunction_get_code(__pyx_CyFunctionObject *op)
{
    PyObject* result = (op->func_code) ? op->func_code : Py_None;
    Py_INCREF(result);
    return result;
}
static int
__Pyx_CyFunction_init_defaults(__pyx_CyFunctionObject *op) {
    PyObject *res = op->defaults_getter((PyObject *) op);
    if (unlikely(!res))
        return -1;
    op->defaults_tuple = PyTuple_GET_ITEM(res, 0);
    Py_INCREF(op->defaults_tuple);
    op->defaults_kwdict = PyTuple_GET_ITEM(res, 1);
    Py_INCREF(op->defaults_kwdict);
    Py_DECREF(res);
    return 0;
}
static int
__Pyx_CyFunction_set_defaults(__pyx_CyFunctionObject *op, PyObject* value) {
    PyObject* tmp;
    if (!value) {
        value = Py_None;
    } else if (value != Py_None && !PyTuple_Check(value)) {
        PyErr_SetString(PyExc_TypeError,
                        "__defaults__ must be set to a tuple object");
        return -1;
    }
    Py_INCREF(value);
    tmp = op->defaults_tuple;
    op->defaults_tuple = value;
    Py_XDECREF(tmp);
    return 0;
}
static PyObject *
__Pyx_CyFunction_get_defaults(__pyx_CyFunctionObject *op) {
    PyObject* result = op->defaults_tuple;
    if (unlikely(!result)) {
        if (op->defaults_getter) {
            if (__Pyx_CyFunction_init_defaults(op) < 0) return NULL;
            result = op->defaults_tuple;
        } else {
            result = Py_None;
        }
    }
    Py_INCREF(result);
    return result;
}
static int
__Pyx_CyFunction_set_kwdefaults(__pyx_CyFunctionObject *op, PyObject* value) {
    PyObject* tmp;
    if (!value) {
        value = Py_None;
    } else if (value != Py_None && !PyDict_Check(value)) {
        PyErr_SetString(PyExc_TypeError,
                        "__kwdefaults__ must be set to a dict object");
        return -1;
    }
    Py_INCREF(value);
    tmp = op->defaults_kwdict;
    op->defaults_kwdict = value;
    Py_XDECREF(tmp);
    return 0;
}
static PyObject *
__Pyx_CyFunction_get_kwdefaults(__pyx_CyFunctionObject *op) {
    PyObject* result = op->defaults_kwdict;
    if (unlikely(!result)) {
        if (op->defaults_getter) {
            if (__Pyx_CyFunction_init_defaults(op) < 0) return NULL;
            result = op->defaults_kwdict;
        } else {
            result = Py_None;
        }
    }
    Py_INCREF(result);
    return result;
}
static int
__Pyx_CyFunction_set_annotations(__pyx_CyFunctionObject *op, PyObject* value) {
    PyObject* tmp;
    if (!value || value == Py_None) {
        value = NULL;
    } else if (!PyDict_Check(value)) {
        PyErr_SetString(PyExc_TypeError,
                        "__annotations__ must be set to a dict object");
        return -1;
    }
    Py_XINCREF(value);
    tmp = op->func_annotations;
    op->func_annotations = value;
    Py_XDECREF(tmp);
    return 0;
}
static PyObject *
__Pyx_CyFunction_get_annotations(__pyx_CyFunctionObject *op) {
    PyObject* result = op->func_annotations;
    if (unlikely(!result)) {
        result = PyDict_New();
        if (unlikely(!result)) return NULL;
        op->func_annotations = result;
    }
    Py_INCREF(result);
    return result;
}
static PyGetSetDef __pyx_CyFunction_getsets[] = {
    {(char *) "func_doc", (getter)__Pyx_CyFunction_get_doc, (setter)__Pyx_CyFunction_set_doc, 0, 0},
    {(char *) "__doc__",  (getter)__Pyx_CyFunction_get_doc, (setter)__Pyx_CyFunction_set_doc, 0, 0},
    {(char *) "func_name", (getter)__Pyx_CyFunction_get_name, (setter)__Pyx_CyFunction_set_name, 0, 0},
    {(char *) "__name__", (getter)__Pyx_CyFunction_get_name, (setter)__Pyx_CyFunction_set_name, 0, 0},
    {(char *) "__qualname__", (getter)__Pyx_CyFunction_get_qualname, (setter)__Pyx_CyFunction_set_qualname, 0, 0},
    {(char *) "__self__", (getter)__Pyx_CyFunction_get_self, 0, 0, 0},
    {(char *) "func_dict", (getter)__Pyx_CyFunction_get_dict, (setter)__Pyx_CyFunction_set_dict, 0, 0},
    {(char *) "__dict__", (getter)__Pyx_CyFunction_get_dict, (setter)__Pyx_CyFunction_set_dict, 0, 0},
    {(char *) "func_globals", (getter)__Pyx_CyFunction_get_globals, 0, 0, 0},
    {(char *) "__globals__", (getter)__Pyx_CyFunction_get_globals, 0, 0, 0},
    {(char *) "func_closure", (getter)__Pyx_CyFunction_get_closure, 0, 0, 0},
    {(char *) "__closure__", (getter)__Pyx_CyFunction_get_closure, 0, 0, 0},
    {(char *) "func_code", (getter)__Pyx_CyFunction_get_code, 0, 0, 0},
    {(char *) "__code__", (getter)__Pyx_CyFunction_get_code, 0, 0, 0},
    {(char *) "func_defaults", (getter)__Pyx_CyFunction_get_defaults, (setter)__Pyx_CyFunction_set_defaults, 0, 0},
    {(char *) "__defaults__", (getter)__Pyx_CyFunction_get_defaults, (setter)__Pyx_CyFunction_set_defaults, 0, 0},
    {(char *) "__kwdefaults__", (getter)__Pyx_CyFunction_get_kwdefaults, (setter)__Pyx_CyFunction_set_kwdefaults, 0, 0},
    {(char *) "__annotations__", (getter)__Pyx_CyFunction_get_annotations, (setter)__Pyx_CyFunction_set_annotations, 0, 0},
    {0, 0, 0, 0, 0}
};
#ifndef PY_WRITE_RESTRICTED
#define PY_WRITE_RESTRICTED WRITE_RESTRICTED
#endif
static PyMemberDef __pyx_CyFunction_members[] = {
    {(char *) "__module__", T_OBJECT, offsetof(__pyx_CyFunctionObject, func.m_module), PY_WRITE_RESTRICTED, 0},
    {0, 0, 0,  0, 0}
};
static PyObject *
__Pyx_CyFunction_reduce(__pyx_CyFunctionObject *m, CYTHON_UNUSED PyObject *args)
{
#if PY_MAJOR_VERSION >= 3
    return PyUnicode_FromString(m->func.m_ml->ml_name);
#else
    return PyString_FromString(m->func.m_ml->ml_name);
#endif
}
static PyMethodDef __pyx_CyFunction_methods[] = {
    {"__reduce__", (PyCFunction)__Pyx_CyFunction_reduce, METH_VARARGS, 0},
    {0, 0, 0, 0}
};
#if PY_VERSION_HEX < 0x030500A0
#define __Pyx_CyFunction_weakreflist(cyfunc) ((cyfunc)->func_weakreflist)
#else
#define __Pyx_CyFunction_weakreflist(cyfunc) ((cyfunc)->func.m_weakreflist)
#endif
static PyObject *__Pyx_CyFunction_New(PyTypeObject *type, PyMethodDef *ml, int flags, PyObject* qualname,
                                      PyObject *closure, PyObject *module, PyObject* globals, PyObject* code) {
    __pyx_CyFunctionObject *op = PyObject_GC_New(__pyx_CyFunctionObject, type);
    if (op == NULL)
        return NULL;
    op->flags = flags;
    __Pyx_CyFunction_weakreflist(op) = NULL;
    op->func.m_ml = ml;
    op->func.m_self = (PyObject *) op;
    Py_XINCREF(closure);
    op->func_closure = closure;
    Py_XINCREF(module);
    op->func.m_module = module;
    op->func_dict = NULL;
    op->func_name = NULL;
    Py_INCREF(qualname);
    op->func_qualname = qualname;
    op->func_doc = NULL;
    op->func_classobj = NULL;
    op->func_globals = globals;
    Py_INCREF(op->func_globals);
    Py_XINCREF(code);
    op->func_code = code;
    op->defaults_pyobjects = 0;
    op->defaults = NULL;
    op->defaults_tuple = NULL;
    op->defaults_kwdict = NULL;
    op->defaults_getter = NULL;
    op->func_annotations = NULL;
    PyObject_GC_Track(op);
    return (PyObject *) op;
}
static int
__Pyx_CyFunction_clear(__pyx_CyFunctionObject *m)
{
    Py_CLEAR(m->func_closure);
    Py_CLEAR(m->func.m_module);
    Py_CLEAR(m->func_dict);
    Py_CLEAR(m->func_name);
    Py_CLEAR(m->func_qualname);
    Py_CLEAR(m->func_doc);
    Py_CLEAR(m->func_globals);
    Py_CLEAR(m->func_code);
    Py_CLEAR(m->func_classobj);
    Py_CLEAR(m->defaults_tuple);
    Py_CLEAR(m->defaults_kwdict);
    Py_CLEAR(m->func_annotations);
    if (m->defaults) {
        PyObject **pydefaults = __Pyx_CyFunction_Defaults(PyObject *, m);
        int i;
        for (i = 0; i < m->defaults_pyobjects; i++)
            Py_XDECREF(pydefaults[i]);
        PyMem_Free(m->defaults);
        m->defaults = NULL;
    }
    return 0;
}
static void __Pyx_CyFunction_dealloc(__pyx_CyFunctionObject *m)
{
    PyObject_GC_UnTrack(m);
    if (__Pyx_CyFunction_weakreflist(m) != NULL)
        PyObject_ClearWeakRefs((PyObject *) m);
    __Pyx_CyFunction_clear(m);
    PyObject_GC_Del(m);
}
static int __Pyx_CyFunction_traverse(__pyx_CyFunctionObject *m, visitproc visit, void *arg)
{
    Py_VISIT(m->func_closure);
    Py_VISIT(m->func.m_module);
    Py_VISIT(m->func_dict);
    Py_VISIT(m->func_name);
    Py_VISIT(m->func_qualname);
    Py_VISIT(m->func_doc);
    Py_VISIT(m->func_globals);
    Py_VISIT(m->func_code);
    Py_VISIT(m->func_classobj);
    Py_VISIT(m->defaults_tuple);
    Py_VISIT(m->defaults_kwdict);
    if (m->defaults) {
        PyObject **pydefaults = __Pyx_CyFunction_Defaults(PyObject *, m);
        int i;
        for (i = 0; i < m->defaults_pyobjects; i++)
            Py_VISIT(pydefaults[i]);
    }
    return 0;
}
static PyObject *__Pyx_CyFunction_descr_get(PyObject *func, PyObject *obj, PyObject *type)
{
    __pyx_CyFunctionObject *m = (__pyx_CyFunctionObject *) func;
    if (m->flags & __Pyx_CYFUNCTION_STATICMETHOD) {
        Py_INCREF(func);
        return func;
    }
    if (m->flags & __Pyx_CYFUNCTION_CLASSMETHOD) {
        if (type == NULL)
            type = (PyObject *)(Py_TYPE(obj));
        return PyMethod_New(func,
                            type, (PyObject *)(Py_TYPE(type)));
    }
    if (obj == Py_None)
        obj = NULL;
    return PyMethod_New(func, obj, type);
}
static PyObject*
__Pyx_CyFunction_repr(__pyx_CyFunctionObject *op)
{
#if PY_MAJOR_VERSION >= 3
    return PyUnicode_FromFormat("<cyfunction %U at %p>",
                                op->func_qualname, (void *)op);
#else
    return PyString_FromFormat("<cyfunction %s at %p>",
                               PyString_AsString(op->func_qualname), (void *)op);
#endif
}
#if CYTHON_COMPILING_IN_PYPY
static PyObject * __Pyx_CyFunction_Call(PyObject *func, PyObject *arg, PyObject *kw) {
    PyCFunctionObject* f = (PyCFunctionObject*)func;
    PyCFunction meth = PyCFunction_GET_FUNCTION(func);
    PyObject *self = PyCFunction_GET_SELF(func);
    Py_ssize_t size;
    switch (PyCFunction_GET_FLAGS(func) & ~(METH_CLASS | METH_STATIC | METH_COEXIST)) {
    case METH_VARARGS:
        if (likely(kw == NULL) || PyDict_Size(kw) == 0)
            return (*meth)(self, arg);
        break;
    case METH_VARARGS | METH_KEYWORDS:
        return (*(PyCFunctionWithKeywords)meth)(self, arg, kw);
    case METH_NOARGS:
        if (likely(kw == NULL) || PyDict_Size(kw) == 0) {
            size = PyTuple_GET_SIZE(arg);
            if (size == 0)
                return (*meth)(self, NULL);
            PyErr_Format(PyExc_TypeError,
                "%.200s() takes no arguments (%zd given)",
                f->m_ml->ml_name, size);
            return NULL;
        }
        break;
    case METH_O:
        if (likely(kw == NULL) || PyDict_Size(kw) == 0) {
            size = PyTuple_GET_SIZE(arg);
            if (size == 1)
                return (*meth)(self, PyTuple_GET_ITEM(arg, 0));
            PyErr_Format(PyExc_TypeError,
                "%.200s() takes exactly one argument (%zd given)",
                f->m_ml->ml_name, size);
            return NULL;
        }
        break;
    default:
        PyErr_SetString(PyExc_SystemError, "Bad call flags in "
                        "__Pyx_CyFunction_Call. METH_OLDARGS is no "
                        "longer supported!");
        return NULL;
    }
    PyErr_Format(PyExc_TypeError, "%.200s() takes no keyword arguments",
                 f->m_ml->ml_name);
    return NULL;
}
#else
static PyObject * __Pyx_CyFunction_Call(PyObject *func, PyObject *arg, PyObject *kw) {
	return PyCFunction_Call(func, arg, kw);
}
#endif
static PyTypeObject __pyx_CyFunctionType_type = {
    PyVarObject_HEAD_INIT(0, 0)
    "cython_function_or_method",
    sizeof(__pyx_CyFunctionObject),
    0,
    (destructor) __Pyx_CyFunction_dealloc,
    0,
    0,
    0,
#if PY_MAJOR_VERSION < 3
    0,
#else
    0,
#endif
    (reprfunc) __Pyx_CyFunction_repr,
    0,
    0,
    0,
    0,
    __Pyx_CyFunction_Call,
    0,
    0,
    0,
    0,
    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
    0,
    (traverseproc) __Pyx_CyFunction_traverse,
    (inquiry) __Pyx_CyFunction_clear,
    0,
#if PY_VERSION_HEX < 0x030500A0
    offsetof(__pyx_CyFunctionObject, func_weakreflist),
#else
    offsetof(PyCFunctionObject, m_weakreflist),
#endif
    0,
    0,
    __pyx_CyFunction_methods,
    __pyx_CyFunction_members,
    __pyx_CyFunction_getsets,
    0,
    0,
    __Pyx_CyFunction_descr_get,
    0,
    offsetof(__pyx_CyFunctionObject, func_dict),
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
#if PY_VERSION_HEX >= 0x030400a1
    0,
#endif
};
static int __Pyx_CyFunction_init(void) {
#if !CYTHON_COMPILING_IN_PYPY
    __pyx_CyFunctionType_type.tp_call = PyCFunction_Call;
#endif
    __pyx_CyFunctionType = __Pyx_FetchCommonType(&__pyx_CyFunctionType_type);
    if (__pyx_CyFunctionType == NULL) {
        return -1;
    }
    return 0;
}
static CYTHON_INLINE void *__Pyx_CyFunction_InitDefaults(PyObject *func, size_t size, int pyobjects) {
    __pyx_CyFunctionObject *m = (__pyx_CyFunctionObject *) func;
    m->defaults = PyMem_Malloc(size);
    if (!m->defaults)
        return PyErr_NoMemory();
    memset(m->defaults, 0, size);
    m->defaults_pyobjects = pyobjects;
    return m->defaults;
}
static CYTHON_INLINE void __Pyx_CyFunction_SetDefaultsTuple(PyObject *func, PyObject *tuple) {
    __pyx_CyFunctionObject *m = (__pyx_CyFunctionObject *) func;
    m->defaults_tuple = tuple;
    Py_INCREF(tuple);
}
static CYTHON_INLINE void __Pyx_CyFunction_SetDefaultsKwDict(PyObject *func, PyObject *dict) {
    __pyx_CyFunctionObject *m = (__pyx_CyFunctionObject *) func;
    m->defaults_kwdict = dict;
    Py_INCREF(dict);
}
static CYTHON_INLINE void __Pyx_CyFunction_SetAnnotationsDict(PyObject *func, PyObject *dict) {
    __pyx_CyFunctionObject *m = (__pyx_CyFunctionObject *) func;
    m->func_annotations = dict;
    Py_INCREF(dict);
}

static CYTHON_INLINE PyObject* __Pyx_PyDict_Values(PyObject* d) {
    if (PY_MAJOR_VERSION >= 3)
        return __Pyx_PyObject_CallMethod1((PyObject*)&PyDict_Type, __pyx_n_s_values, d);
    else
        return PyDict_Values(d);
}

static void __Pyx_call_next_tp_dealloc(PyObject* obj, destructor current_tp_dealloc) {
    PyTypeObject* type = Py_TYPE(obj);
    while (type && type->tp_dealloc != current_tp_dealloc)
        type = type->tp_base;
    while (type && type->tp_dealloc == current_tp_dealloc)
        type = type->tp_base;
    if (type)
        type->tp_dealloc(obj);
}

static int __Pyx_call_next_tp_traverse(PyObject* obj, visitproc v, void *a, traverseproc current_tp_traverse) {
    PyTypeObject* type = Py_TYPE(obj);
    while (type && type->tp_traverse != current_tp_traverse)
        type = type->tp_base;
    while (type && type->tp_traverse == current_tp_traverse)
        type = type->tp_base;
    if (type && type->tp_traverse)
        return type->tp_traverse(obj, v, a);
    return 0;
}

static void __Pyx_call_next_tp_clear(PyObject* obj, inquiry current_tp_clear) {
    PyTypeObject* type = Py_TYPE(obj);
    while (type && type->tp_clear != current_tp_clear)
        type = type->tp_base;
    while (type && type->tp_clear == current_tp_clear)
        type = type->tp_base;
    if (type && type->tp_clear)
        type->tp_clear(obj);
}

static void* __Pyx_GetVtable(PyObject *dict) {
    void* ptr;
    PyObject *ob = PyObject_GetItem(dict, __pyx_n_s_pyx_vtable);
    if (!ob)
        goto bad;
#if PY_VERSION_HEX >= 0x02070000
    ptr = PyCapsule_GetPointer(ob, 0);
#else
    ptr = PyCObject_AsVoidPtr(ob);
#endif
    if (!ptr && !PyErr_Occurred())
        PyErr_SetString(PyExc_RuntimeError, "invalid vtable found for imported type");
    Py_DECREF(ob);
    return ptr;
bad:
    Py_XDECREF(ob);
    return NULL;
}

static int __Pyx_SetVtable(PyObject *dict, void *vtable) {
#if PY_VERSION_HEX >= 0x02070000
    PyObject *ob = PyCapsule_New(vtable, 0, 0);
#else
    PyObject *ob = PyCObject_FromVoidPtr(vtable, 0);
#endif
    if (!ob)
        goto bad;
    if (PyDict_SetItem(dict, __pyx_n_s_pyx_vtable, ob) < 0)
        goto bad;
    Py_DECREF(ob);
    return 0;
bad:
    Py_XDECREF(ob);
    return -1;
}

static PyObject *__Pyx_CalculateMetaclass(PyTypeObject *metaclass, PyObject *bases) {
    Py_ssize_t i, nbases = PyTuple_GET_SIZE(bases);
    for (i=0; i < nbases; i++) {
        PyTypeObject *tmptype;
        PyObject *tmp = PyTuple_GET_ITEM(bases, i);
        tmptype = Py_TYPE(tmp);
#if PY_MAJOR_VERSION < 3
        if (tmptype == &PyClass_Type)
            continue;
#endif
        if (!metaclass) {
            metaclass = tmptype;
            continue;
        }
        if (PyType_IsSubtype(metaclass, tmptype))
            continue;
        if (PyType_IsSubtype(tmptype, metaclass)) {
            metaclass = tmptype;
            continue;
        }
        PyErr_SetString(PyExc_TypeError,
                        "metaclass conflict: "
                        "the metaclass of a derived class "
                        "must be a (non-strict) subclass "
                        "of the metaclasses of all its bases");
        return NULL;
    }
    if (!metaclass) {
#if PY_MAJOR_VERSION < 3
        metaclass = &PyClass_Type;
#else
        metaclass = &PyType_Type;
#endif
    }
    Py_INCREF((PyObject*) metaclass);
    return (PyObject*) metaclass;
}

static PyObject *__Pyx_Py3MetaclassPrepare(PyObject *metaclass, PyObject *bases, PyObject *name,
                                           PyObject *qualname, PyObject *mkw, PyObject *modname, PyObject *doc) {
    PyObject *ns;
    if (metaclass) {
        PyObject *prep = __Pyx_PyObject_GetAttrStr(metaclass, __pyx_n_s_prepare);
        if (prep) {
            PyObject *pargs = PyTuple_Pack(2, name, bases);
            if (unlikely(!pargs)) {
                Py_DECREF(prep);
                return NULL;
            }
            ns = PyObject_Call(prep, pargs, mkw);
            Py_DECREF(prep);
            Py_DECREF(pargs);
        } else {
            if (unlikely(!PyErr_ExceptionMatches(PyExc_AttributeError)))
                return NULL;
            PyErr_Clear();
            ns = PyDict_New();
        }
    } else {
        ns = PyDict_New();
    }
    if (unlikely(!ns))
        return NULL;
    if (unlikely(PyObject_SetItem(ns, __pyx_n_s_module, modname) < 0)) goto bad;
    if (unlikely(PyObject_SetItem(ns, __pyx_n_s_qualname, qualname) < 0)) goto bad;
    if (unlikely(doc && PyObject_SetItem(ns, __pyx_n_s_doc, doc) < 0)) goto bad;
    return ns;
bad:
    Py_DECREF(ns);
    return NULL;
}
static PyObject *__Pyx_Py3ClassCreate(PyObject *metaclass, PyObject *name, PyObject *bases,
                                      PyObject *dict, PyObject *mkw,
                                      int calculate_metaclass, int allow_py2_metaclass) {
    PyObject *result, *margs;
    PyObject *owned_metaclass = NULL;
    if (allow_py2_metaclass) {
        owned_metaclass = PyObject_GetItem(dict, __pyx_n_s_metaclass);
        if (owned_metaclass) {
            metaclass = owned_metaclass;
        } else if (likely(PyErr_ExceptionMatches(PyExc_KeyError))) {
            PyErr_Clear();
        } else {
            return NULL;
        }
    }
    if (calculate_metaclass && (!metaclass || PyType_Check(metaclass))) {
        metaclass = __Pyx_CalculateMetaclass((PyTypeObject*) metaclass, bases);
        Py_XDECREF(owned_metaclass);
        if (unlikely(!metaclass))
            return NULL;
        owned_metaclass = metaclass;
    }
    margs = PyTuple_Pack(3, name, bases, dict);
    if (unlikely(!margs)) {
        result = NULL;
    } else {
        result = PyObject_Call(metaclass, margs, mkw);
        Py_DECREF(margs);
    }
    Py_XDECREF(owned_metaclass);
    return result;
}

static PyObject *__Pyx_GetNameInClass(PyObject *nmspace, PyObject *name) {
    PyObject *result;
    result = __Pyx_PyObject_GetAttrStr(nmspace, name);
    if (!result)
        result = __Pyx_GetModuleGlobalName(name);
    return result;
}

static int __pyx_bisect_code_objects(__Pyx_CodeObjectCacheEntry* entries, int count, int code_line) {
    int start = 0, mid = 0, end = count - 1;
    if (end >= 0 && code_line > entries[end].code_line) {
        return count;
    }
    while (start < end) {
        mid = (start + end) / 2;
        if (code_line < entries[mid].code_line) {
            end = mid;
        } else if (code_line > entries[mid].code_line) {
             start = mid + 1;
        } else {
            return mid;
        }
    }
    if (code_line <= entries[mid].code_line) {
        return mid;
    } else {
        return mid + 1;
    }
}
static PyCodeObject *__pyx_find_code_object(int code_line) {
    PyCodeObject* code_object;
    int pos;
    if (unlikely(!code_line) || unlikely(!__pyx_code_cache.entries)) {
        return NULL;
    }
    pos = __pyx_bisect_code_objects(__pyx_code_cache.entries, __pyx_code_cache.count, code_line);
    if (unlikely(pos >= __pyx_code_cache.count) || unlikely(__pyx_code_cache.entries[pos].code_line != code_line)) {
        return NULL;
    }
    code_object = __pyx_code_cache.entries[pos].code_object;
    Py_INCREF(code_object);
    return code_object;
}
static void __pyx_insert_code_object(int code_line, PyCodeObject* code_object) {
    int pos, i;
    __Pyx_CodeObjectCacheEntry* entries = __pyx_code_cache.entries;
    if (unlikely(!code_line)) {
        return;
    }
    if (unlikely(!entries)) {
        entries = (__Pyx_CodeObjectCacheEntry*)PyMem_Malloc(64*sizeof(__Pyx_CodeObjectCacheEntry));
        if (likely(entries)) {
            __pyx_code_cache.entries = entries;
            __pyx_code_cache.max_count = 64;
            __pyx_code_cache.count = 1;
            entries[0].code_line = code_line;
            entries[0].code_object = code_object;
            Py_INCREF(code_object);
        }
        return;
    }
    pos = __pyx_bisect_code_objects(__pyx_code_cache.entries, __pyx_code_cache.count, code_line);
    if ((pos < __pyx_code_cache.count) && unlikely(__pyx_code_cache.entries[pos].code_line == code_line)) {
        PyCodeObject* tmp = entries[pos].code_object;
        entries[pos].code_object = code_object;
        Py_DECREF(tmp);
        return;
    }
    if (__pyx_code_cache.count == __pyx_code_cache.max_count) {
        int new_max = __pyx_code_cache.max_count + 64;
        entries = (__Pyx_CodeObjectCacheEntry*)PyMem_Realloc(
            __pyx_code_cache.entries, (size_t)new_max*sizeof(__Pyx_CodeObjectCacheEntry));
        if (unlikely(!entries)) {
            return;
        }
        __pyx_code_cache.entries = entries;
        __pyx_code_cache.max_count = new_max;
    }
    for (i=__pyx_code_cache.count; i>pos; i--) {
        entries[i] = entries[i-1];
    }
    entries[pos].code_line = code_line;
    entries[pos].code_object = code_object;
    __pyx_code_cache.count++;
    Py_INCREF(code_object);
}

#include "compile.h"
#include "frameobject.h"
#include "traceback.h"
static PyCodeObject* __Pyx_CreateCodeObjectForTraceback(
            const char *funcname, int c_line,
            int py_line, const char *filename) {
    PyCodeObject *py_code = 0;
    PyObject *py_srcfile = 0;
    PyObject *py_funcname = 0;
    #if PY_MAJOR_VERSION < 3
    py_srcfile = PyString_FromString(filename);
    #else
    py_srcfile = PyUnicode_FromString(filename);
    #endif
    if (!py_srcfile) goto bad;
    if (c_line) {
        #if PY_MAJOR_VERSION < 3
        py_funcname = PyString_FromFormat( "%s (%s:%d)", funcname, __pyx_cfilenm, c_line);
        #else
        py_funcname = PyUnicode_FromFormat( "%s (%s:%d)", funcname, __pyx_cfilenm, c_line);
        #endif
    }
    else {
        #if PY_MAJOR_VERSION < 3
        py_funcname = PyString_FromString(funcname);
        #else
        py_funcname = PyUnicode_FromString(funcname);
        #endif
    }
    if (!py_funcname) goto bad;
    py_code = __Pyx_PyCode_New(
        0,
        0,
        0,
        0,
        0,
        __pyx_empty_bytes, /*PyObject *code,*/
        __pyx_empty_tuple, /*PyObject *consts,*/
        __pyx_empty_tuple, /*PyObject *names,*/
        __pyx_empty_tuple, /*PyObject *varnames,*/
        __pyx_empty_tuple, /*PyObject *freevars,*/
        __pyx_empty_tuple, /*PyObject *cellvars,*/
        py_srcfile,   /*PyObject *filename,*/
        py_funcname,  /*PyObject *name,*/
        py_line,
        __pyx_empty_bytes  /*PyObject *lnotab*/
    );
    Py_DECREF(py_srcfile);
    Py_DECREF(py_funcname);
    return py_code;
bad:
    Py_XDECREF(py_srcfile);
    Py_XDECREF(py_funcname);
    return NULL;
}
static void __Pyx_AddTraceback(const char *funcname, int c_line,
                               int py_line, const char *filename) {
    PyCodeObject *py_code = 0;
    PyFrameObject *py_frame = 0;
    py_code = __pyx_find_code_object(c_line ? c_line : py_line);
    if (!py_code) {
        py_code = __Pyx_CreateCodeObjectForTraceback(
            funcname, c_line, py_line, filename);
        if (!py_code) goto bad;
        __pyx_insert_code_object(c_line ? c_line : py_line, py_code);
    }
    py_frame = PyFrame_New(
        PyThreadState_GET(), /*PyThreadState *tstate,*/
        py_code,             /*PyCodeObject *code,*/
        __pyx_d,      /*PyObject *globals,*/
        0                    /*PyObject *locals*/
    );
    if (!py_frame) goto bad;
    py_frame->f_lineno = py_line;
    PyTraceBack_Here(py_frame);
bad:
    Py_XDECREF(py_code);
    Py_XDECREF(py_frame);
}

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_short(short value) {
    const short neg_one = (short) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
    if (is_unsigned) {
        if (sizeof(short) < sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(short) <= sizeof(unsigned long)) {
            return PyLong_FromUnsignedLong((unsigned long) value);
        } else if (sizeof(short) <= sizeof(unsigned long long)) {
            return PyLong_FromUnsignedLongLong((unsigned long long) value);
        }
    } else {
        if (sizeof(short) <= sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(short) <= sizeof(long long)) {
            return PyLong_FromLongLong((long long) value);
        }
    }
    {
        int one = 1; int little = (int)*(unsigned char *)&one;
        unsigned char *bytes = (unsigned char *)&value;
        return _PyLong_FromByteArray(bytes, sizeof(short),
                                     little, !is_unsigned);
    }
}

#define __PYX_VERIFY_RETURN_INT(target_type, func_type, func_value)       \
    {                                                                     \
        func_type value = func_value;                                     \
        if (sizeof(target_type) < sizeof(func_type)) {                    \
            if (unlikely(value != (func_type) (target_type) value)) {     \
                func_type zero = 0;                                       \
                if (is_unsigned && unlikely(value < zero))                \
                    goto raise_neg_overflow;                              \
                else                                                      \
                    goto raise_overflow;                                  \
            }                                                             \
        }                                                                 \
        return (target_type) value;                                       \
    }

#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
  #include "longintrepr.h"
 #endif
#endif

static CYTHON_INLINE short __Pyx_PyInt_As_short(PyObject *x) {
    const short neg_one = (short) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
#if PY_MAJOR_VERSION < 3
    if (likely(PyInt_Check(x))) {
        if (sizeof(short) < sizeof(long)) {
            __PYX_VERIFY_RETURN_INT(short, long, PyInt_AS_LONG(x))
        } else {
            long val = PyInt_AS_LONG(x);
            if (is_unsigned && unlikely(val < 0)) {
                goto raise_neg_overflow;
            }
            return (short) val;
        }
    } else
#endif
    if (likely(PyLong_Check(x))) {
        if (is_unsigned) {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(short, digit, ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (unlikely(Py_SIZE(x) < 0)) {
                goto raise_neg_overflow;
            }
            if (sizeof(short) <= sizeof(unsigned long)) {
                __PYX_VERIFY_RETURN_INT(short, unsigned long, PyLong_AsUnsignedLong(x))
            } else if (sizeof(short) <= sizeof(unsigned long long)) {
                __PYX_VERIFY_RETURN_INT(short, unsigned long long, PyLong_AsUnsignedLongLong(x))
            }
        } else {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(short,  digit, +(((PyLongObject*)x)->ob_digit[0]));
                case -1: __PYX_VERIFY_RETURN_INT(short, sdigit, -(sdigit) ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (sizeof(short) <= sizeof(long)) {
                __PYX_VERIFY_RETURN_INT(short, long, PyLong_AsLong(x))
            } else if (sizeof(short) <= sizeof(long long)) {
                __PYX_VERIFY_RETURN_INT(short, long long, PyLong_AsLongLong(x))
            }
        }
        {
#if CYTHON_COMPILING_IN_PYPY && !defined(_PyLong_AsByteArray)
            PyErr_SetString(PyExc_RuntimeError,
                            "_PyLong_AsByteArray() not available in PyPy, cannot convert large numbers");
#else
            short val;
            PyObject *v = __Pyx_PyNumber_Int(x);
 #if PY_MAJOR_VERSION < 3
            if (likely(v) && !PyLong_Check(v)) {
                PyObject *tmp = v;
                v = PyNumber_Long(tmp);
                Py_DECREF(tmp);
            }
 #endif
            if (likely(v)) {
                int one = 1; int is_little = (int)*(unsigned char *)&one;
                unsigned char *bytes = (unsigned char *)&val;
                int ret = _PyLong_AsByteArray((PyLongObject *)v,
                                              bytes, sizeof(val),
                                              is_little, !is_unsigned);
                Py_DECREF(v);
                if (likely(!ret))
                    return val;
            }
#endif
            return (short) -1;
        }
    } else {
        short val;
        PyObject *tmp = __Pyx_PyNumber_Int(x);
        if (!tmp) return (short) -1;
        val = __Pyx_PyInt_As_short(tmp);
        Py_DECREF(tmp);
        return val;
    }
raise_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "value too large to convert to short");
    return (short) -1;
raise_neg_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "can't convert negative value to short");
    return (short) -1;
}

static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level) {
    PyObject *empty_list = 0;
    PyObject *module = 0;
    PyObject *global_dict = 0;
    PyObject *empty_dict = 0;
    PyObject *list;
    #if PY_VERSION_HEX < 0x03030000
    PyObject *py_import;
    py_import = __Pyx_PyObject_GetAttrStr(__pyx_b, __pyx_n_s_import);
    if (!py_import)
        goto bad;
    #endif
    if (from_list)
        list = from_list;
    else {
        empty_list = PyList_New(0);
        if (!empty_list)
            goto bad;
        list = empty_list;
    }
    global_dict = PyModule_GetDict(__pyx_m);
    if (!global_dict)
        goto bad;
    empty_dict = PyDict_New();
    if (!empty_dict)
        goto bad;
    {
        #if PY_MAJOR_VERSION >= 3
        if (level == -1) {
            if (strchr(__Pyx_MODULE_NAME, '.')) {
                #if PY_VERSION_HEX < 0x03030000
                PyObject *py_level = PyInt_FromLong(1);
                if (!py_level)
                    goto bad;
                module = PyObject_CallFunctionObjArgs(py_import,
                    name, global_dict, empty_dict, list, py_level, NULL);
                Py_DECREF(py_level);
                #else
                module = PyImport_ImportModuleLevelObject(
                    name, global_dict, empty_dict, list, 1);
                #endif
                if (!module) {
                    if (!PyErr_ExceptionMatches(PyExc_ImportError))
                        goto bad;
                    PyErr_Clear();
                }
            }
            level = 0;
        }
        #endif
        if (!module) {
            #if PY_VERSION_HEX < 0x03030000
            PyObject *py_level = PyInt_FromLong(level);
            if (!py_level)
                goto bad;
            module = PyObject_CallFunctionObjArgs(py_import,
                name, global_dict, empty_dict, list, py_level, NULL);
            Py_DECREF(py_level);
            #else
            module = PyImport_ImportModuleLevelObject(
                name, global_dict, empty_dict, list, level);
            #endif
        }
    }
bad:
    #if PY_VERSION_HEX < 0x03030000
    Py_XDECREF(py_import);
    #endif
    Py_XDECREF(empty_list);
    Py_XDECREF(empty_dict);
    return module;
}

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_long(long value) {
    const long neg_one = (long) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
    if (is_unsigned) {
        if (sizeof(long) < sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(long) <= sizeof(unsigned long)) {
            return PyLong_FromUnsignedLong((unsigned long) value);
        } else if (sizeof(long) <= sizeof(unsigned long long)) {
            return PyLong_FromUnsignedLongLong((unsigned long long) value);
        }
    } else {
        if (sizeof(long) <= sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(long) <= sizeof(long long)) {
            return PyLong_FromLongLong((long long) value);
        }
    }
    {
        int one = 1; int little = (int)*(unsigned char *)&one;
        unsigned char *bytes = (unsigned char *)&value;
        return _PyLong_FromByteArray(bytes, sizeof(long),
                                     little, !is_unsigned);
    }
}

static CYTHON_INLINE int __Pyx_PyInt_As_int(PyObject *x) {
    const int neg_one = (int) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
#if PY_MAJOR_VERSION < 3
    if (likely(PyInt_Check(x))) {
        if (sizeof(int) < sizeof(long)) {
            __PYX_VERIFY_RETURN_INT(int, long, PyInt_AS_LONG(x))
        } else {
            long val = PyInt_AS_LONG(x);
            if (is_unsigned && unlikely(val < 0)) {
                goto raise_neg_overflow;
            }
            return (int) val;
        }
    } else
#endif
    if (likely(PyLong_Check(x))) {
        if (is_unsigned) {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(int, digit, ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (unlikely(Py_SIZE(x) < 0)) {
                goto raise_neg_overflow;
            }
            if (sizeof(int) <= sizeof(unsigned long)) {
                __PYX_VERIFY_RETURN_INT(int, unsigned long, PyLong_AsUnsignedLong(x))
            } else if (sizeof(int) <= sizeof(unsigned long long)) {
                __PYX_VERIFY_RETURN_INT(int, unsigned long long, PyLong_AsUnsignedLongLong(x))
            }
        } else {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(int,  digit, +(((PyLongObject*)x)->ob_digit[0]));
                case -1: __PYX_VERIFY_RETURN_INT(int, sdigit, -(sdigit) ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (sizeof(int) <= sizeof(long)) {
                __PYX_VERIFY_RETURN_INT(int, long, PyLong_AsLong(x))
            } else if (sizeof(int) <= sizeof(long long)) {
                __PYX_VERIFY_RETURN_INT(int, long long, PyLong_AsLongLong(x))
            }
        }
        {
#if CYTHON_COMPILING_IN_PYPY && !defined(_PyLong_AsByteArray)
            PyErr_SetString(PyExc_RuntimeError,
                            "_PyLong_AsByteArray() not available in PyPy, cannot convert large numbers");
#else
            int val;
            PyObject *v = __Pyx_PyNumber_Int(x);
 #if PY_MAJOR_VERSION < 3
            if (likely(v) && !PyLong_Check(v)) {
                PyObject *tmp = v;
                v = PyNumber_Long(tmp);
                Py_DECREF(tmp);
            }
 #endif
            if (likely(v)) {
                int one = 1; int is_little = (int)*(unsigned char *)&one;
                unsigned char *bytes = (unsigned char *)&val;
                int ret = _PyLong_AsByteArray((PyLongObject *)v,
                                              bytes, sizeof(val),
                                              is_little, !is_unsigned);
                Py_DECREF(v);
                if (likely(!ret))
                    return val;
            }
#endif
            return (int) -1;
        }
    } else {
        int val;
        PyObject *tmp = __Pyx_PyNumber_Int(x);
        if (!tmp) return (int) -1;
        val = __Pyx_PyInt_As_int(tmp);
        Py_DECREF(tmp);
        return val;
    }
raise_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "value too large to convert to int");
    return (int) -1;
raise_neg_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "can't convert negative value to int");
    return (int) -1;
}

static CYTHON_INLINE PyObject* __Pyx_PyInt_From_int(int value) {
    const int neg_one = (int) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
    if (is_unsigned) {
        if (sizeof(int) < sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(int) <= sizeof(unsigned long)) {
            return PyLong_FromUnsignedLong((unsigned long) value);
        } else if (sizeof(int) <= sizeof(unsigned long long)) {
            return PyLong_FromUnsignedLongLong((unsigned long long) value);
        }
    } else {
        if (sizeof(int) <= sizeof(long)) {
            return PyInt_FromLong((long) value);
        } else if (sizeof(int) <= sizeof(long long)) {
            return PyLong_FromLongLong((long long) value);
        }
    }
    {
        int one = 1; int little = (int)*(unsigned char *)&one;
        unsigned char *bytes = (unsigned char *)&value;
        return _PyLong_FromByteArray(bytes, sizeof(int),
                                     little, !is_unsigned);
    }
}

#if CYTHON_CCOMPLEX
  #ifdef __cplusplus
    static CYTHON_INLINE __pyx_t_float_complex __pyx_t_float_complex_from_parts(float x, float y) {
      return ::std::complex< float >(x, y);
    }
  #else
    static CYTHON_INLINE __pyx_t_float_complex __pyx_t_float_complex_from_parts(float x, float y) {
      return x + y*(__pyx_t_float_complex)_Complex_I;
    }
  #endif
#else
    static CYTHON_INLINE __pyx_t_float_complex __pyx_t_float_complex_from_parts(float x, float y) {
      __pyx_t_float_complex z;
      z.real = x;
      z.imag = y;
      return z;
    }
#endif

#if CYTHON_CCOMPLEX
#else
    static CYTHON_INLINE int __Pyx_c_eqf(__pyx_t_float_complex a, __pyx_t_float_complex b) {
       return (a.real == b.real) && (a.imag == b.imag);
    }
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_sumf(__pyx_t_float_complex a, __pyx_t_float_complex b) {
        __pyx_t_float_complex z;
        z.real = a.real + b.real;
        z.imag = a.imag + b.imag;
        return z;
    }
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_difff(__pyx_t_float_complex a, __pyx_t_float_complex b) {
        __pyx_t_float_complex z;
        z.real = a.real - b.real;
        z.imag = a.imag - b.imag;
        return z;
    }
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_prodf(__pyx_t_float_complex a, __pyx_t_float_complex b) {
        __pyx_t_float_complex z;
        z.real = a.real * b.real - a.imag * b.imag;
        z.imag = a.real * b.imag + a.imag * b.real;
        return z;
    }
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_quotf(__pyx_t_float_complex a, __pyx_t_float_complex b) {
        __pyx_t_float_complex z;
        float denom = b.real * b.real + b.imag * b.imag;
        z.real = (a.real * b.real + a.imag * b.imag) / denom;
        z.imag = (a.imag * b.real - a.real * b.imag) / denom;
        return z;
    }
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_negf(__pyx_t_float_complex a) {
        __pyx_t_float_complex z;
        z.real = -a.real;
        z.imag = -a.imag;
        return z;
    }
    static CYTHON_INLINE int __Pyx_c_is_zerof(__pyx_t_float_complex a) {
       return (a.real == 0) && (a.imag == 0);
    }
    static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_conjf(__pyx_t_float_complex a) {
        __pyx_t_float_complex z;
        z.real =  a.real;
        z.imag = -a.imag;
        return z;
    }
    #if 1
        static CYTHON_INLINE float __Pyx_c_absf(__pyx_t_float_complex z) {
          #if !defined(HAVE_HYPOT) || defined(_MSC_VER)
            return sqrtf(z.real*z.real + z.imag*z.imag);
          #else
            return hypotf(z.real, z.imag);
          #endif
        }
        static CYTHON_INLINE __pyx_t_float_complex __Pyx_c_powf(__pyx_t_float_complex a, __pyx_t_float_complex b) {
            __pyx_t_float_complex z;
            float r, lnr, theta, z_r, z_theta;
            if (b.imag == 0 && b.real == (int)b.real) {
                if (b.real < 0) {
                    float denom = a.real * a.real + a.imag * a.imag;
                    a.real = a.real / denom;
                    a.imag = -a.imag / denom;
                    b.real = -b.real;
                }
                switch ((int)b.real) {
                    case 0:
                        z.real = 1;
                        z.imag = 0;
                        return z;
                    case 1:
                        return a;
                    case 2:
                        z = __Pyx_c_prodf(a, a);
                        return __Pyx_c_prodf(a, a);
                    case 3:
                        z = __Pyx_c_prodf(a, a);
                        return __Pyx_c_prodf(z, a);
                    case 4:
                        z = __Pyx_c_prodf(a, a);
                        return __Pyx_c_prodf(z, z);
                }
            }
            if (a.imag == 0) {
                if (a.real == 0) {
                    return a;
                }
                r = a.real;
                theta = 0;
            } else {
                r = __Pyx_c_absf(a);
                theta = atan2f(a.imag, a.real);
            }
            lnr = logf(r);
            z_r = expf(lnr * b.real - theta * b.imag);
            z_theta = theta * b.real + lnr * b.imag;
            z.real = z_r * cosf(z_theta);
            z.imag = z_r * sinf(z_theta);
            return z;
        }
    #endif
#endif

#if CYTHON_CCOMPLEX
  #ifdef __cplusplus
    static CYTHON_INLINE __pyx_t_double_complex __pyx_t_double_complex_from_parts(double x, double y) {
      return ::std::complex< double >(x, y);
    }
  #else
    static CYTHON_INLINE __pyx_t_double_complex __pyx_t_double_complex_from_parts(double x, double y) {
      return x + y*(__pyx_t_double_complex)_Complex_I;
    }
  #endif
#else
    static CYTHON_INLINE __pyx_t_double_complex __pyx_t_double_complex_from_parts(double x, double y) {
      __pyx_t_double_complex z;
      z.real = x;
      z.imag = y;
      return z;
    }
#endif

#if CYTHON_CCOMPLEX
#else
    static CYTHON_INLINE int __Pyx_c_eq(__pyx_t_double_complex a, __pyx_t_double_complex b) {
       return (a.real == b.real) && (a.imag == b.imag);
    }
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_sum(__pyx_t_double_complex a, __pyx_t_double_complex b) {
        __pyx_t_double_complex z;
        z.real = a.real + b.real;
        z.imag = a.imag + b.imag;
        return z;
    }
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_diff(__pyx_t_double_complex a, __pyx_t_double_complex b) {
        __pyx_t_double_complex z;
        z.real = a.real - b.real;
        z.imag = a.imag - b.imag;
        return z;
    }
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_prod(__pyx_t_double_complex a, __pyx_t_double_complex b) {
        __pyx_t_double_complex z;
        z.real = a.real * b.real - a.imag * b.imag;
        z.imag = a.real * b.imag + a.imag * b.real;
        return z;
    }
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_quot(__pyx_t_double_complex a, __pyx_t_double_complex b) {
        __pyx_t_double_complex z;
        double denom = b.real * b.real + b.imag * b.imag;
        z.real = (a.real * b.real + a.imag * b.imag) / denom;
        z.imag = (a.imag * b.real - a.real * b.imag) / denom;
        return z;
    }
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_neg(__pyx_t_double_complex a) {
        __pyx_t_double_complex z;
        z.real = -a.real;
        z.imag = -a.imag;
        return z;
    }
    static CYTHON_INLINE int __Pyx_c_is_zero(__pyx_t_double_complex a) {
       return (a.real == 0) && (a.imag == 0);
    }
    static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_conj(__pyx_t_double_complex a) {
        __pyx_t_double_complex z;
        z.real =  a.real;
        z.imag = -a.imag;
        return z;
    }
    #if 1
        static CYTHON_INLINE double __Pyx_c_abs(__pyx_t_double_complex z) {
          #if !defined(HAVE_HYPOT) || defined(_MSC_VER)
            return sqrt(z.real*z.real + z.imag*z.imag);
          #else
            return hypot(z.real, z.imag);
          #endif
        }
        static CYTHON_INLINE __pyx_t_double_complex __Pyx_c_pow(__pyx_t_double_complex a, __pyx_t_double_complex b) {
            __pyx_t_double_complex z;
            double r, lnr, theta, z_r, z_theta;
            if (b.imag == 0 && b.real == (int)b.real) {
                if (b.real < 0) {
                    double denom = a.real * a.real + a.imag * a.imag;
                    a.real = a.real / denom;
                    a.imag = -a.imag / denom;
                    b.real = -b.real;
                }
                switch ((int)b.real) {
                    case 0:
                        z.real = 1;
                        z.imag = 0;
                        return z;
                    case 1:
                        return a;
                    case 2:
                        z = __Pyx_c_prod(a, a);
                        return __Pyx_c_prod(a, a);
                    case 3:
                        z = __Pyx_c_prod(a, a);
                        return __Pyx_c_prod(z, a);
                    case 4:
                        z = __Pyx_c_prod(a, a);
                        return __Pyx_c_prod(z, z);
                }
            }
            if (a.imag == 0) {
                if (a.real == 0) {
                    return a;
                }
                r = a.real;
                theta = 0;
            } else {
                r = __Pyx_c_abs(a);
                theta = atan2(a.imag, a.real);
            }
            lnr = log(r);
            z_r = exp(lnr * b.real - theta * b.imag);
            z_theta = theta * b.real + lnr * b.imag;
            z.real = z_r * cos(z_theta);
            z.imag = z_r * sin(z_theta);
            return z;
        }
    #endif
#endif

static CYTHON_INLINE long __Pyx_PyInt_As_long(PyObject *x) {
    const long neg_one = (long) -1, const_zero = 0;
    const int is_unsigned = neg_one > const_zero;
#if PY_MAJOR_VERSION < 3
    if (likely(PyInt_Check(x))) {
        if (sizeof(long) < sizeof(long)) {
            __PYX_VERIFY_RETURN_INT(long, long, PyInt_AS_LONG(x))
        } else {
            long val = PyInt_AS_LONG(x);
            if (is_unsigned && unlikely(val < 0)) {
                goto raise_neg_overflow;
            }
            return (long) val;
        }
    } else
#endif
    if (likely(PyLong_Check(x))) {
        if (is_unsigned) {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(long, digit, ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (unlikely(Py_SIZE(x) < 0)) {
                goto raise_neg_overflow;
            }
            if (sizeof(long) <= sizeof(unsigned long)) {
                __PYX_VERIFY_RETURN_INT(long, unsigned long, PyLong_AsUnsignedLong(x))
            } else if (sizeof(long) <= sizeof(unsigned long long)) {
                __PYX_VERIFY_RETURN_INT(long, unsigned long long, PyLong_AsUnsignedLongLong(x))
            }
        } else {
#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
 #if CYTHON_USE_PYLONG_INTERNALS
            switch (Py_SIZE(x)) {
                case  0: return 0;
                case  1: __PYX_VERIFY_RETURN_INT(long,  digit, +(((PyLongObject*)x)->ob_digit[0]));
                case -1: __PYX_VERIFY_RETURN_INT(long, sdigit, -(sdigit) ((PyLongObject*)x)->ob_digit[0]);
            }
 #endif
#endif
            if (sizeof(long) <= sizeof(long)) {
                __PYX_VERIFY_RETURN_INT(long, long, PyLong_AsLong(x))
            } else if (sizeof(long) <= sizeof(long long)) {
                __PYX_VERIFY_RETURN_INT(long, long long, PyLong_AsLongLong(x))
            }
        }
        {
#if CYTHON_COMPILING_IN_PYPY && !defined(_PyLong_AsByteArray)
            PyErr_SetString(PyExc_RuntimeError,
                            "_PyLong_AsByteArray() not available in PyPy, cannot convert large numbers");
#else
            long val;
            PyObject *v = __Pyx_PyNumber_Int(x);
 #if PY_MAJOR_VERSION < 3
            if (likely(v) && !PyLong_Check(v)) {
                PyObject *tmp = v;
                v = PyNumber_Long(tmp);
                Py_DECREF(tmp);
            }
 #endif
            if (likely(v)) {
                int one = 1; int is_little = (int)*(unsigned char *)&one;
                unsigned char *bytes = (unsigned char *)&val;
                int ret = _PyLong_AsByteArray((PyLongObject *)v,
                                              bytes, sizeof(val),
                                              is_little, !is_unsigned);
                Py_DECREF(v);
                if (likely(!ret))
                    return val;
            }
#endif
            return (long) -1;
        }
    } else {
        long val;
        PyObject *tmp = __Pyx_PyNumber_Int(x);
        if (!tmp) return (long) -1;
        val = __Pyx_PyInt_As_long(tmp);
        Py_DECREF(tmp);
        return val;
    }
raise_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "value too large to convert to long");
    return (long) -1;
raise_neg_overflow:
    PyErr_SetString(PyExc_OverflowError,
        "can't convert negative value to long");
    return (long) -1;
}

static int __Pyx_check_binary_version(void) {
    char ctversion[4], rtversion[4];
    PyOS_snprintf(ctversion, 4, "%d.%d", PY_MAJOR_VERSION, PY_MINOR_VERSION);
    PyOS_snprintf(rtversion, 4, "%s", Py_GetVersion());
    if (ctversion[0] != rtversion[0] || ctversion[2] != rtversion[2]) {
        char message[200];
        PyOS_snprintf(message, sizeof(message),
                      "compiletime version %s of module '%.100s' "
                      "does not match runtime version %s",
                      ctversion, __Pyx_MODULE_NAME, rtversion);
        return PyErr_WarnEx(NULL, message, 1);
    }
    return 0;
}

#ifndef __PYX_HAVE_RT_ImportModule
#define __PYX_HAVE_RT_ImportModule
static PyObject *__Pyx_ImportModule(const char *name) {
    PyObject *py_name = 0;
    PyObject *py_module = 0;
    py_name = __Pyx_PyIdentifier_FromString(name);
    if (!py_name)
        goto bad;
    py_module = PyImport_Import(py_name);
    Py_DECREF(py_name);
    return py_module;
bad:
    Py_XDECREF(py_name);
    return 0;
}
#endif

#ifndef __PYX_HAVE_RT_ImportType
#define __PYX_HAVE_RT_ImportType
static PyTypeObject *__Pyx_ImportType(const char *module_name, const char *class_name,
    size_t size, int strict)
{
    PyObject *py_module = 0;
    PyObject *result = 0;
    PyObject *py_name = 0;
    char warning[200];
    Py_ssize_t basicsize;
#ifdef Py_LIMITED_API
    PyObject *py_basicsize;
#endif
    py_module = __Pyx_ImportModule(module_name);
    if (!py_module)
        goto bad;
    py_name = __Pyx_PyIdentifier_FromString(class_name);
    if (!py_name)
        goto bad;
    result = PyObject_GetAttr(py_module, py_name);
    Py_DECREF(py_name);
    py_name = 0;
    Py_DECREF(py_module);
    py_module = 0;
    if (!result)
        goto bad;
    if (!PyType_Check(result)) {
        PyErr_Format(PyExc_TypeError,
            "%.200s.%.200s is not a type object",
            module_name, class_name);
        goto bad;
    }
#ifndef Py_LIMITED_API
    basicsize = ((PyTypeObject *)result)->tp_basicsize;
#else
    py_basicsize = PyObject_GetAttrString(result, "__basicsize__");
    if (!py_basicsize)
        goto bad;
    basicsize = PyLong_AsSsize_t(py_basicsize);
    Py_DECREF(py_basicsize);
    py_basicsize = 0;
    if (basicsize == (Py_ssize_t)-1 && PyErr_Occurred())
        goto bad;
#endif
    if (!strict && (size_t)basicsize > size) {
        PyOS_snprintf(warning, sizeof(warning),
            "%s.%s size changed, may indicate binary incompatibility",
            module_name, class_name);
        if (PyErr_WarnEx(NULL, warning, 0) < 0) goto bad;
    }
    else if ((size_t)basicsize != size) {
        PyErr_Format(PyExc_ValueError,
            "%.200s.%.200s has the wrong size, try recompiling",
            module_name, class_name);
        goto bad;
    }
    return (PyTypeObject *)result;
bad:
    Py_XDECREF(py_module);
    Py_XDECREF(result);
    return NULL;
}
#endif

#ifndef __PYX_HAVE_RT_ImportVoidPtr
#define __PYX_HAVE_RT_ImportVoidPtr
static int __Pyx_ImportVoidPtr(PyObject *module, const char *name, void **p, const char *sig) {
    PyObject *d = 0;
    PyObject *cobj = 0;
    d = PyObject_GetAttrString(module, (char *)"__pyx_capi__");
    if (!d)
        goto bad;
    cobj = PyDict_GetItemString(d, name);
    if (!cobj) {
        PyErr_Format(PyExc_ImportError,
            "%.200s does not export expected C variable %.200s",
                PyModule_GetName(module), name);
        goto bad;
    }
#if PY_VERSION_HEX >= 0x02070000
    if (!PyCapsule_IsValid(cobj, sig)) {
        PyErr_Format(PyExc_TypeError,
            "C variable %.200s.%.200s has wrong signature (expected %.500s, got %.500s)",
             PyModule_GetName(module), name, sig, PyCapsule_GetName(cobj));
        goto bad;
    }
    *p = PyCapsule_GetPointer(cobj, sig);
#else
    {const char *desc, *s1, *s2;
    desc = (const char *)PyCObject_GetDesc(cobj);
    if (!desc)
        goto bad;
    s1 = desc; s2 = sig;
    while (*s1 != '\0' && *s1 == *s2) { s1++; s2++; }
    if (*s1 != *s2) {
        PyErr_Format(PyExc_TypeError,
            "C variable %.200s.%.200s has wrong signature (expected %.500s, got %.500s)",
             PyModule_GetName(module), name, sig, desc);
        goto bad;
    }
    *p = PyCObject_AsVoidPtr(cobj);}
#endif
    if (!(*p))
        goto bad;
    Py_DECREF(d);
    return 0;
bad:
    Py_XDECREF(d);
    return -1;
}
#endif

static int __Pyx_InitStrings(__Pyx_StringTabEntry *t) {
    while (t->p) {
        #if PY_MAJOR_VERSION < 3
        if (t->is_unicode) {
            *t->p = PyUnicode_DecodeUTF8(t->s, t->n - 1, NULL);
        } else if (t->intern) {
            *t->p = PyString_InternFromString(t->s);
        } else {
            *t->p = PyString_FromStringAndSize(t->s, t->n - 1);
        }
        #else
        if (t->is_unicode | t->is_str) {
            if (t->intern) {
                *t->p = PyUnicode_InternFromString(t->s);
            } else if (t->encoding) {
                *t->p = PyUnicode_Decode(t->s, t->n - 1, t->encoding, NULL);
            } else {
                *t->p = PyUnicode_FromStringAndSize(t->s, t->n - 1);
            }
        } else {
            *t->p = PyBytes_FromStringAndSize(t->s, t->n - 1);
        }
        #endif
        if (!*t->p)
            return -1;
        ++t;
    }
    return 0;
}

static CYTHON_INLINE PyObject* __Pyx_PyUnicode_FromString(const char* c_str) {
    return __Pyx_PyUnicode_FromStringAndSize(c_str, (Py_ssize_t)strlen(c_str));
}
static CYTHON_INLINE char* __Pyx_PyObject_AsString(PyObject* o) {
    Py_ssize_t ignore;
    return __Pyx_PyObject_AsStringAndSize(o, &ignore);
}
static CYTHON_INLINE char* __Pyx_PyObject_AsStringAndSize(PyObject* o, Py_ssize_t *length) {
#if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII || __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT
    if (
#if PY_MAJOR_VERSION < 3 && __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
            __Pyx_sys_getdefaultencoding_not_ascii &&
#endif
            PyUnicode_Check(o)) {
#if PY_VERSION_HEX < 0x03030000
        char* defenc_c;
        PyObject* defenc = _PyUnicode_AsDefaultEncodedString(o, NULL);
        if (!defenc) return NULL;
        defenc_c = PyBytes_AS_STRING(defenc);
#if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
        {
            char* end = defenc_c + PyBytes_GET_SIZE(defenc);
            char* c;
            for (c = defenc_c; c < end; c++) {
                if ((unsigned char) (*c) >= 128) {
                    PyUnicode_AsASCIIString(o);
                    return NULL;
                }
            }
        }
#endif
        *length = PyBytes_GET_SIZE(defenc);
        return defenc_c;
#else
        if (__Pyx_PyUnicode_READY(o) == -1) return NULL;
#if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
        if (PyUnicode_IS_ASCII(o)) {
            *length = PyUnicode_GET_LENGTH(o);
            return PyUnicode_AsUTF8(o);
        } else {
            PyUnicode_AsASCIIString(o);
            return NULL;
        }
#else
        return PyUnicode_AsUTF8AndSize(o, length);
#endif
#endif
    } else
#endif
#if !CYTHON_COMPILING_IN_PYPY
    if (PyByteArray_Check(o)) {
        *length = PyByteArray_GET_SIZE(o);
        return PyByteArray_AS_STRING(o);
    } else
#endif
    {
        char* result;
        int r = PyBytes_AsStringAndSize(o, &result, length);
        if (unlikely(r < 0)) {
            return NULL;
        } else {
            return result;
        }
    }
}
static CYTHON_INLINE int __Pyx_PyObject_IsTrue(PyObject* x) {
   int is_true = x == Py_True;
   if (is_true | (x == Py_False) | (x == Py_None)) return is_true;
   else return PyObject_IsTrue(x);
}
static CYTHON_INLINE PyObject* __Pyx_PyNumber_Int(PyObject* x) {
  PyNumberMethods *m;
  const char *name = NULL;
  PyObject *res = NULL;
#if PY_MAJOR_VERSION < 3
  if (PyInt_Check(x) || PyLong_Check(x))
#else
  if (PyLong_Check(x))
#endif
    return Py_INCREF(x), x;
  m = Py_TYPE(x)->tp_as_number;
#if PY_MAJOR_VERSION < 3
  if (m && m->nb_int) {
    name = "int";
    res = PyNumber_Int(x);
  }
  else if (m && m->nb_long) {
    name = "long";
    res = PyNumber_Long(x);
  }
#else
  if (m && m->nb_int) {
    name = "int";
    res = PyNumber_Long(x);
  }
#endif
  if (res) {
#if PY_MAJOR_VERSION < 3
    if (!PyInt_Check(res) && !PyLong_Check(res)) {
#else
    if (!PyLong_Check(res)) {
#endif
      PyErr_Format(PyExc_TypeError,
                   "__%.4s__ returned non-%.4s (type %.200s)",
                   name, name, Py_TYPE(res)->tp_name);
      Py_DECREF(res);
      return NULL;
    }
  }
  else if (!PyErr_Occurred()) {
    PyErr_SetString(PyExc_TypeError,
                    "an integer is required");
  }
  return res;
}
static CYTHON_INLINE Py_ssize_t __Pyx_PyIndex_AsSsize_t(PyObject* b) {
  Py_ssize_t ival;
  PyObject *x;
#if PY_MAJOR_VERSION < 3
  if (likely(PyInt_CheckExact(b)))
      return PyInt_AS_LONG(b);
#endif
  if (likely(PyLong_CheckExact(b))) {
    #if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION >= 3
     #if CYTHON_USE_PYLONG_INTERNALS
       switch (Py_SIZE(b)) {
       case -1: return -(sdigit)((PyLongObject*)b)->ob_digit[0];
       case  0: return 0;
       case  1: return ((PyLongObject*)b)->ob_digit[0];
       }
     #endif
    #endif
    return PyLong_AsSsize_t(b);
  }
  x = PyNumber_Index(b);
  if (!x) return -1;
  ival = PyInt_AsSsize_t(x);
  Py_DECREF(x);
  return ival;
}
static CYTHON_INLINE PyObject * __Pyx_PyInt_FromSize_t(size_t ival) {
    return PyInt_FromSize_t(ival);
}


#endif /* Py_PYTHON_H */
